{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PREPRINT – Discriminating Seagrasses From Green Macroalgae in European\n",
    "\n",
    "Intertidal areas using high resolution multispectral drone imagery – PREPRINT.\n",
    "\n",
    "Simon Oiry [![](data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==)](https://orcid.org/0000-0001-7161-5246) (Institut des Substances et Organismes de la Mer, ISOMer, Nantes Université, UR 2160, F-44000 Nantes, France)  \n",
    "Bede Ffinian Rowe Davies (Institut des Substances et Organismes de la Mer, ISOMer, Nantes Université, UR 2160, F-44000 Nantes, France)  \n",
    "Ana I. Sousa (CESAM – Centre for Environmental and Marine Studies, Department of Biology, University of Aveiro, Aveiro, Portugal)  \n",
    "Philippe Rosa (Institut des Substances et Organismes de la Mer, ISOMer, Nantes Université, UR 2160, F-44000 Nantes, France)  \n",
    "Maria Laura Zoffoli (Consiglio Nazionale delle Ricerche, Istituto di Scienze Marine (CNR-ISMAR), 00133 Rome, Italy)  \n",
    "Guillaume Brunier (BRGM French Geological Survey, Cayenne 97300, French Guiana)  \n",
    "Pierre Gernez (Institut des Substances et Organismes de la Mer, ISOMer, Nantes Université, UR 2160, F-44000 Nantes, France)  \n",
    "Laurent Barillé (Institut des Substances et Organismes de la Mer, ISOMer, Nantes Université, UR 2160, F-44000 Nantes, France)  \n",
    "March 12, 2024\n",
    "\n",
    "Coastal areas support seagrass meadows, which offer crucial ecosystem services including erosion control and carbon sequestration. However, these areas are increasingly impacted by human activities, leading to seagrass decline and habitat fragmentation. In situ surveys, traditionally performed to monitor these ecosystems face limitations on temporal and spatial coverage, particularly in intertidal zones, prompting the use of satellite data within monitoring programs. Yet, satellite remote sensing struggles with spatial and spectral resolution, making it difficult to discriminate seagrass from other macrophytes in highly heterogenous meadows. Drone images at a very high spatial resolution offer a promising solution to address challenges related to spatial heterogeneity and intrapixel mixture. This study focuses on using drone acquisitions with a ten spectral band sensor mirroring those of Sentinel-2, for mapping intertidal macrophytes and effectively discriminating between seagrass and green macroalgae. Nine drone flights were conducted at two different altitudes (12m and 120m) across heterogeneous intertidal European habitats in France and Portugal. Low altitude flights were used to train a Deep Learning classifier based on Neural Networks to discrimintate among five taxonomic classes of intertidal vegetation: Magnoliopsida (Seagrass), Chlorophyceae (Green macroalgae), Phaeophyceae (Brown algae), Rhodophyceae (Red macroalgae) and benthic Bacillariophyceae (Diatoms). Classification of drone imagery resulted in an overall accuracy of 94% across all the sites and images, covering a total area of 467 000 m². The model exhibited an accuracy of 96.4% in identifying seagrass. This work showed that although they share the same pigmentary composition, seagrass and green algae can be discriminated using a multispectral sensor with ten spectral bands at 444, 475, 531, 560, 650, 668, 705, 717, 740 and 840 nm. The algorithm should now be adapted for Sentinel-2 to see if this discrimination is still operational at a coarser spatial resolution."
   ],
   "id": "abfa4ffb-f455-4339-a72c-10baf4be8645"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──\n",
      "✔ dplyr     1.1.3     ✔ readr     2.1.4\n",
      "✔ forcats   1.0.0     ✔ stringr   1.5.0\n",
      "✔ ggplot2   3.4.3     ✔ tibble    3.2.1\n",
      "✔ lubridate 1.9.2     ✔ tidyr     1.3.0\n",
      "✔ purrr     1.0.2     \n",
      "── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "✖ purrr::compose() masks flextable::compose()\n",
      "✖ dplyr::filter()  masks stats::filter()\n",
      "✖ dplyr::lag()     masks stats::lag()\n",
      "ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: le package 'tidyterra' a été compilé avec la version R 4.3.2"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attachement du package : 'tidyterra'\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:stats':\n",
      "\n",
      "    filter"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "── Attaching packages ────────────────────────────────────── tidymodels 1.1.1 ──\n",
      "✔ broom        1.0.5     ✔ rsample      1.2.0\n",
      "✔ dials        1.2.0     ✔ tune         1.1.2\n",
      "✔ infer        1.0.5     ✔ workflows    1.1.3\n",
      "✔ modeldata    1.2.0     ✔ workflowsets 1.0.1\n",
      "✔ parsnip      1.1.1     ✔ yardstick    1.2.0\n",
      "✔ recipes      1.0.8     \n",
      "── Conflicts ───────────────────────────────────────── tidymodels_conflicts() ──\n",
      "✖ purrr::compose()    masks flextable::compose()\n",
      "✖ scales::discard()   masks purrr::discard()\n",
      "✖ tidyterra::filter() masks dplyr::filter(), stats::filter()\n",
      "✖ recipes::fixed()    masks stringr::fixed()\n",
      "✖ dplyr::lag()        masks stats::lag()\n",
      "✖ yardstick::spec()   masks readr::spec()\n",
      "✖ recipes::step()     masks stats::step()\n",
      "• Search for functions across packages at https://www.tidymodels.org/find/"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Le chargement a nécessité le package : cowplot\n",
      "\n",
      "Attachement du package : 'cowplot'\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:lubridate':\n",
      "\n",
      "    stamp\n",
      "\n",
      "Le chargement a nécessité le package : patchwork\n",
      "\n",
      "Attachement du package : 'patchwork'\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:cowplot':\n",
      "\n",
      "    align_plots\n",
      "\n",
      "Le chargement a nécessité le package : magrittr\n",
      "\n",
      "Attachement du package : 'magrittr'\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:purrr':\n",
      "\n",
      "    set_names\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:tidyr':\n",
      "\n",
      "    extract"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Linking to GEOS 3.11.2, GDAL 3.6.2, PROJ 9.2.0; sf_use_s2() is TRUE"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,\n",
      "which was just loaded, will retire in October 2023.\n",
      "Please refer to R-spatial evolution reports for details, especially\n",
      "https://r-spatial.org/r/2023/05/15/evolution4.html.\n",
      "It may be desirable to make the sf package available;\n",
      "package maintainers should consider adding sf to Suggests:.\n",
      "The sp package is now running under evolution status 2\n",
      "     (status 2 uses the sf package in place of rgdal)\n",
      "Support for Spatial objects (`sp`) will be deprecated in {rnaturalearth} and will be removed in a future release of the package. Please use `sf` objects with {rnaturalearth}. For example: `ne_download(returnclass = 'sf')`"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attachement du package : 'rnaturalearthdata'\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:rnaturalearth':\n",
      "\n",
      "    countries110"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: le package 'terra' a été compilé avec la version R 4.3.2"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "terra 1.7.55\n",
      "\n",
      "Attachement du package : 'terra'\n",
      "\n",
      "Les objets suivants sont masqués depuis 'package:magrittr':\n",
      "\n",
      "    extract, inset\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:patchwork':\n",
      "\n",
      "    area\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:scales':\n",
      "\n",
      "    rescale\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:tidyr':\n",
      "\n",
      "    extract\n",
      "\n",
      "Les objets suivants sont masqués depuis 'package:flextable':\n",
      "\n",
      "    align, colorize, rotate, width"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Linking to ImageMagick 6.9.12.93\n",
      "Enabled features: cairo, freetype, fftw, ghostscript, heic, lcms, pango, raw, rsvg, webp\n",
      "Disabled features: fontconfig, x11"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attachement du package : 'grid'\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:terra':\n",
      "\n",
      "    depth"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Le chargement a nécessité le package : abind"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Le chargement a nécessité le package : lattice\n",
      "\n",
      "Attachement du package : 'caret'\n",
      "\n",
      "Les objets suivants sont masqués depuis 'package:yardstick':\n",
      "\n",
      "    precision, recall, sensitivity, specificity\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:purrr':\n",
      "\n",
      "    lift"
     ]
    }
   ],
   "source": [
    "library(flextable)\n",
    "library(tidyverse)"
   ],
   "id": "c2617829-067c-44a7-af70-458474df39fb"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_comma<-scales::label_comma(accuracy = NULL, big.mark = \",\",decimal.mark = \".\")"
   ],
   "id": "f831b0de-7e65-4f19-bc4c-11ea54f841c0"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Introduction\n",
    "\n",
    "Coastal areas are vital hotspots for marine biodiversity, with intertidal seagrass meadows playing a crucial role at the interface between the land and oceans ([Unsworth et al., 2022](#ref-unsworth2022)). Seagrass meadows provide a myriad of ecosystem services to humanity, including carbon sequestration, oxygen production, protection against sea-level rise and coastline erosion, and limitation of eutrophication. They serve as vital habitats for a diverse array of marine and terrestrial species, providing living, breeding, and feeding grounds ([Gardner and Finlayson, 2018](#ref-gardner2018) ; [Zoffoli et al., 2022](#ref-Zoffoli2022) ; [Jankowska et al., 2019](#ref-jankowska2019)). Due to the concentration of human activities in coastal zones, seagrass meadows are directly exposed to and impacted by anthropogenic pressures. Global regression and fragmentation are currently observed due to climate change, diseases, urbanization, land reclamation, dredging, competition with alien species, and reduction in water quality ([Nguyen et al., 2021](#ref-nguyen2021) ; [Soissons et al., 2018](#ref-soissons2018) ; [Orth et al., 2006](#ref-orth2006) ; [Lin et al., 2018](#ref-lin2018) ; [Duffy et al., 2019](#ref-duffy2019) ; [Rasheed and Unsworth, 2011](#ref-rasheed2011long) ; [Chefaoui et al., 2018](#ref-chefaoui2018dramatic)). Both habitat fragmentation and reduction, in turn, can severely copromise the effectiveness of ecosystem services provided by seagrass meadows. While improvements in water quality have been recently reported in Europe, allowing an overall recovery of seagrass ecosystems at local scale, many coastal waters worldwide are still subjected to strong eutrophication processes ([Los Santos et al., 2019](#ref-deSantos2019) ; [Zoffoli et al., 2021](#ref-Zoffoli2021)). Coastal eutrophication has been ass ociated to excessive accumulation of green macroalgae, so-called green tides (Devlin and Brodie ([2023](#ref-devlin2023nutrients))). Green tides produce shade and suffocation over seagrass individuals, thus threatening the health of seagrass ecosystems ([Wang et al., 2022](#ref-wang2022)).\n",
    "\n",
    "The importance of seagrass meadows and the variety of ecosystem services they provide have led to the enhancement of both global and regional programs to monitor Essential Oceanic Variable (EOVs) such as seagrass composition ([Miloslavich et al., 2018](#ref-Miloslavich2018)), as well as Essential Biodiversity Variable (EBVs) such as seagrass taxonomic diversity, species distribution, population abundance, and phenology ([Pereira et al., 2013](#ref-Pereira2013)). Traditionally, indicators of seagrass status have been quantified using *in situ* measurements. The acquisition of field data in intertidal zones is however notoriously challenging. Intertidal seagrass meadows are only partially exposed during low tide and can be situated in difficult-to-reach mudflats, potentially leading to inaccurate and limited estimations with conventional sampling techniques ([Nijland et al., 2019](#ref-nijland2019)). Satellite observations have been proven effective in complementing in situ sampling, allowing for the near real-time and consistent retrieval of seagrass EOVs and EBVs over extensive meadows. ([Zoffoli et al., 2021](#ref-Zoffoli2021) ; [Xu et al., 2021](#ref-xu2021) ; [Traganos and Reinartz, 2018](#ref-Traganos2018) ; [Coffer et al., 2023](#ref-coffer2023))\n",
    "\n",
    "While satellite remote sensing (RS) provides temporally consistent observations over large spatial scales, its utilization over intertidal areas is limited by several constraints. Satellite missions with a high temporal resolution (e.g. daily MODIS observation) are limited by a coarse spatial resolution (\\>100m) over these restricted extension areas. Missions with a high spatial resolution such as Sentinel-2 (10m) or Landsat8/9 (30m) can be limited by low spectral resolution. The limited number of spectral bands is indeed a challenge to accurately discriminate seagrass from others co-existing macrophytes. In particular Chlorophyceae (green algae) and marine Magnoliopsida (seagrass) share the same pigment composition ([Ralph et al., 2002](#ref-ralph2002) ; [Douay et al., 2022](#ref-Douay2022)). Therefore, to someone not specialized in the field, their spectral signatures may appear to be alike ([Davies et al., 2023a](#ref-Davies2023) ; [Bannari et al., 2022](#ref-bannari2022)). Recently, using advanced machine-learning algorithms trained with a large hyperspectral library of more than 300 field reflectance spectra, Davies et al. ([2023a](#ref-Davies2023)) demonstrated that it was possible to discriminate Magnoliopsida from Chlorophyceae using radiometric data acquired at Sentinel-2 ’s spectral resolution. However the application of this approach to satellite RS remains to be validated. Moreover patches of green algae can develop at small spatial scales that are not observable using non-commercial satellite imagery ([Tuya et al., 2013](#ref-tuya2013)), especially during the initial stage of a green tide.\n",
    "\n",
    "Drones can potentially fill the gaps left by satellite RS and in situ measurements, due to their ability to provide spatially-explicit observations at very high spatial resolutions (pixel size from cm to mm) while capturing data at multi-spectral resolution ([Fairley et al., 2022](#ref-fairley2022drone) ; [Oh et al., 2017](#ref-oh2017use)). The versatility of drones allows for their application across a diverse thematic range , from coastal zone management ([Adade et al., 2021](#ref-adade2021) ; [Casella et al., 2020](#ref-casella2020) ; [Angnuureng et al., 2022](#ref-angnuureng2022)) to mapping species distribution ([Joyce et al., 2023](#ref-joyce2023) ; [Tallam et al., 2023](#ref-tallam2023) ; [Roca et al., 2022](#ref-Roca2022) ; [Román et al., 2021](#ref-Roman2021) ; [Brunier et al., 2022](#ref-Brunier2022Topographic)). However, when applied to coastal habitat mapping, most case studies are limited to a single flight, restricting the generalizability of their application over wider geographical scales ([Román et al., 2021](#ref-Roman2021) ; [Collin et al., 2019](#ref-collin2019improving) ; [Rossiter et al., 2020](#ref-rossiter2020uav) ; [Brunier et al., 2022](#ref-Brunier2022Topographic)). The present study aimed at analyzing the potential of multispectral drone RS to map intertidal macrophytes, with a particular focus on discriminating Magnoliopsida and Chlorophyceae (Seagrass and Green Algae respectively). Ten drone flights were performed over soft-bottom intertidal areas along two European countries (France and Portugal), covering a wide range of habitats, from monospecific seagrass meadows to meadows mixed with green, or red macroalgae. A deep learning algorithm was trained and validated for macrophyte discrimination, emphasizing applicability across diverse sites without a loss of prediction accuracy.\n",
    "\n",
    "# 2. Material & Methods\n",
    "\n",
    "## 2.1 Study sites\n",
    "\n",
    "Seven study sites distributed between France and Portugal were selected for their relatively extensive intertidal seagrass beds. Two sites are located in the Gulf of Morbihan (<a href=\"#fig-map\" class=\"quarto-xref\">Figure 1</a> A), France (47.5791°N, 2.8018°W). This gulf covers an area of 115 km² and is only connected to the sea through a 900m wide channel. A total of 53 small islands are scattered across the gulf leading to 250 km of shorelines. Patchy seagrass meadows can be found on a lot of these islands. Within the Gulf of Morbihan we chose two sites : one located on one of the gulf’s islands (Arz island) and the other located in the southern part of the gulf (Duer). Two other sites were located in Bourgneuf Bay, France (46.9849°N, 2.1488°W). This bay is 340 km² semi-enclosed macrotidal bay, protected from waves by Noirmoutier Island. Bourgneuf bay hosts a large intertidal seagrass meadow of about 6 km². Within this meadow, the sites observed by drones (L’Epine and Barbatre, <a href=\"#fig-map\" class=\"quarto-xref\">Figure 1</a> B) contained monospecific beds of *Nanozostera noltei* (dwarf eelgrass) with very little mixing with other macrophytes. Three sites were surveyed in the Ria de Aveiro Lagoon in Portugal (40.6887°N, 8.6810°W). The extent of this lagoon is about 80 km² with many narrow channels, large salt marshes and many mudflats that uncovers at low tide. It is connected to the open sea through a single channel, with a tidal lag between the North and the South of the lagoon. The southernmost site (Gafanha) was a mudflat located in a tributary river of the lagoon whereas the two other sites (Mataducos and Marinha Lanzarote) were situated in the middle of the lagoon and only accessible by boat (<a href=\"#fig-map\" class=\"quarto-xref\">Figure 1</a> C). These Portuguese sites are characterized by a more diverse intertidal vegetation, where patches of seagrass intermingle with red, brown and green macroalgae."
   ],
   "id": "c7bad61f-373e-4180-9215-2ead4ef52f65"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "world_map <- sovereignty10 %>% \n",
    "  st_as_sf() %>% \n",
    "  dplyr::filter(SOVEREIGNT%in%c(\"Spain\",\"France\",\"Portugal\",\n",
    "                                \"Italy\",\"Andorra\",\n",
    "                                \"United Kingdom\",\n",
    "                                \"Switzerland\",\"Belgium\",\n",
    "                                \"Germany\",\"Luxembourg\") ) \n",
    "\n",
    "bbox_europe <- st_bbox(c(xmin = -20, ymin = 34,\n",
    "                         xmax = 20, ymax = 55) ,\n",
    "                       crs = st_crs(world_map) ) \n",
    "\n",
    "world_map<-st_make_valid(world_map) \n",
    "\n",
    "european_union_map_cropped <- st_crop(world_map, bbox_europe)  %>% \n",
    "  st_transform(\"+proj=laea +lat_0=52 +lon_0=10 +x_0=4321000 +y_0=3210000 +ellps=GRS80 +units=m +no_defs \")  \n",
    "\n",
    "\n",
    "UnitedKingdom<-sovereignty10 %>% \n",
    "  st_as_sf() %>% \n",
    "  dplyr::filter(SOVEREIGNT%in%c(\"United Kingdom\") ) %>% \n",
    "  st_cast(\"POLYGON\") \n",
    "\n",
    "bbox_UK <- st_bbox(c(xmin = -20, ymin = 45,\n",
    "                         xmax = 20, ymax = 55) ,\n",
    "                       crs = st_crs(UnitedKingdom) ) \n",
    "\n",
    "UnitedKingdom<-st_make_valid(UnitedKingdom) \n",
    "\n",
    "UK_map_cropped <- st_crop(UnitedKingdom, bbox_UK)  %>% \n",
    "  st_transform(\"+proj=laea +lat_0=52 +lon_0=10 +x_0=4321000 +y_0=3210000 +ellps=GRS80 +units=m +no_defs \") \n",
    "\n",
    "Europe_sf<-european_union_map_cropped %>% \n",
    "  dplyr::bind_rows(UK_map_cropped) \n",
    "\n",
    "Projects<-data.frame(\n",
    "  Name=c(\n",
    "  \"Portugal - Aveiro Lagoon Flights\\nLower-Resolution: 4\\nHigher-Resolution: 1\",\n",
    "  \"France - Morbihan Gulf Flights\\nLower-Resolution: 3\\nHigher-Resolution: 3\",\n",
    "  \"France - Bourgneuf Bay Flights\\nLower-Resolution: 4\\nHigher-Resolution: 0\"\n",
    "  ) ,\n",
    "  Long=c(-8.745730,\n",
    "         -2.840592, \n",
    "         -2.111868) ,\n",
    "  Lat=c(40.617610, \n",
    "        47.578736,\n",
    "        47.024719) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(\"+proj=laea +lat_0=52 +lon_0=10 +x_0=4321000 +y_0=3210000 +ellps=GRS80 +units=m +no_defs \")  \n",
    "\n",
    "Projects_df<-Projects %>%\n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)  %>% \n",
    "  separate(Name, into=c(\"Country\",\"Bay\") ,sep=\" - \")  %>% \n",
    "  dplyr::mutate(Country=as.factor(Country) ,\n",
    "                Bay=as.factor(Bay) ) \n",
    "\n",
    "Miniworld_map <- sovereignty10 %>% \n",
    "  st_as_sf()\n",
    "\n",
    "sf_use_s2(FALSE)\n",
    "\n",
    "bbox_EU <- st_bbox(c(xmin = -30, ymin = 20,\n",
    "                         xmax = 50, ymax = 70) ,\n",
    "                       crs = st_crs(Miniworld_map) ) \n",
    "\n",
    "  MiniEU_map<-st_crop(Miniworld_map, bbox_EU)  %>% \n",
    "  st_transform(\"+proj=laea +lat_0=52 +lon_0=10 +x_0=4321000 +y_0=3210000 +ellps=GRS80 +units=m +no_defs \")  \n",
    "\n",
    "  df_P1<-Projects_df %>% \n",
    "    mutate(ID = c(1:nrow(.))) %>% \n",
    "    mutate(Site = case_when(ID == 1 ~ \"Aveiro Lagoon\",\n",
    "                            ID == 2 ~ \"Morbihan Gulf\",\n",
    "                            ID == 3 ~ \"Bourgneuf Bay\",\n",
    "                            TRUE ~ \"NA\"),\n",
    "           ID = case_when(ID == 1 ~ \"C\",\n",
    "                          ID == 2 ~ \"A\",\n",
    "                          TRUE ~ \"B\")) %>% \n",
    "    dplyr::select(c(lon,lat,ID,Site))\n",
    "    \n",
    "    scaleFUN <- function(x) paste0(sprintf(\"%.2f\", x),\"°N\")\n",
    "  \n",
    "\n",
    "p1  <-\n",
    "  ggplot(MiniEU_map) +\n",
    "  geom_sf(linewidth=0.5,alpha=0.93,\n",
    "          fill=\"#CFCFCF\",colour=\"grey30\")+\n",
    "    ggforce::geom_mark_ellipse(data=df_P1,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = ID,\n",
    "                     description=Site) ,\n",
    "                 linewidth=0.3,\n",
    "                 fill=\"goldenrod\",\n",
    "                 show.legend=F,\n",
    "                 label.hjust = 0.5,\n",
    "                 con.size = 2,\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 label.fontsize = c(25,20),\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") , \n",
    "  label.buffer = unit(5, \"mm\") ,\n",
    "  label.fill = \"grey90\")+\n",
    "  coord_sf(xlim=c(2600000,4100000) ,\n",
    "          ylim=c(1600000,3100000))+\n",
    "  theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  scale_y_continuous(labels=scaleFUN)+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\"),\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20) )\n",
    "\n",
    "\n",
    "\n",
    "ggsave(\"Figures/Figure1/Map_Drone_Sites.png\",p1,width= 10, height=10)  "
   ],
   "id": "3e4bbfcc-aadd-43d8-bfa9-d90f7c9db2b8"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filelist<-list.files(\"Data/shp/\", pattern = \".shp\", recursive = T,full.names = T)\n",
    "mask_aveiro <- filelist[str_detect(filelist,\"mask_land_intertidal_Aveiro\")] %>% \n",
    "  read_sf()\n",
    "\n",
    "Flight_aveiro<-data.frame(\n",
    "  Name=c(\n",
    "  \"Gafanha\",\n",
    "  \"Mataducos\",\n",
    "  \"Marinha Lanzarote\"\n",
    "  ) ,\n",
    "  Description = c(\"12m and 120m Flights\",\n",
    "                  \"120m Flight\",\n",
    "                  \"120m Flight\"),\n",
    "  Long=c(-8.742485,\n",
    "         -8.646980, \n",
    "         -8.690677) ,\n",
    "  Lat=c(40.597599, \n",
    "        40.666634,\n",
    "        40.665120) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(crs(mask_aveiro))  %>% \n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)\n",
    "\n",
    "aveiro_plot <-\n",
    "ggplot() +\n",
    "  scale_fill_manual(\"\", labels = c(\"Intertidal area\", \"Land area\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\"))+\n",
    "  geom_sf(data = mask_aveiro, mapping = aes(fill = Type),linewidth=0.05,alpha=0.93,colour=\"grey30\")+\n",
    "      coord_sf(xlim=c(519749,535233) ,\n",
    "          ylim=c(4493310,4508794) )+\n",
    "  ggforce::geom_mark_ellipse(data=Flight_aveiro,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = Name,\n",
    "                     group = Name),\n",
    "                     # description=Description),\n",
    "                 size=0.3,\n",
    "                 fill = \"goldenrod\",\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 show.legend=F,\n",
    "                 label.fontsize = 25,\n",
    "                 label.hjust = 0.5,\n",
    "                 con.size = 2,\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") ,\n",
    "  label.fill = \"grey90\",\n",
    "  label.buffer = unit(5, \"mm\")) +\n",
    "    # theme_void()+\n",
    "  theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  geom_label(aes(x = 521000, y = 4507794, label = \"C\"), size = 20)+\n",
    "  scale_x_continuous(breaks = seq(-8.70, -8.60, by = 0.10))+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\") ,\n",
    "        legend.position = \"none\",\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20))\n",
    "\n",
    "\n",
    "ggsave(\"Figures/Figure1/Map_Aveiro.png\",aveiro_plot,width= 10, height=10)  "
   ],
   "id": "33d8424a-510d-4f93-aff9-5c5412048aa6"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filelist<-list.files(\"Data/shp/\", pattern = \".shp\", recursive = T,full.names = T)\n",
    "mask_BB <- filelist[str_detect(filelist,\"mask_land_intertidal_BB\")] %>% \n",
    "  read_sf()\n",
    "\n",
    "Flight_BB<-data.frame(\n",
    "  Name=c(\n",
    "  \"Barbâtre\",\n",
    "  \"L'Epine\"\n",
    "  ) ,\n",
    "  Description = c(\"120m Flight\",\n",
    "                  \"120m Flight\"),\n",
    "  Long=c(-2.166313,\n",
    "         -2.228239) ,\n",
    "  Lat=c(46.956420, \n",
    "        46.989515) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(crs(mask_BB))  %>% \n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)\n",
    "\n",
    "BB_plot <- ggplot() +\n",
    "  scale_fill_manual(\"\", labels = c(\"Intertidal area\", \"Land area\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\"))+\n",
    "  geom_sf(data = mask_BB, mapping = aes(fill = Type),linewidth=0.05,alpha=0.93,colour=\"grey30\")+\n",
    "      coord_sf(xlim=c(550609,579539) ,\n",
    "          ylim=c(5189447,5218377) )+\n",
    "  ggforce::geom_mark_ellipse(data=Flight_BB,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = Name,\n",
    "                     group = Name),\n",
    "                     # description=Description),\n",
    "                 size=0.3,\n",
    "                 fill = \"goldenrod\",\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 show.legend=F,\n",
    "                 con.size = 2,\n",
    "                 label.fontsize = 25,\n",
    "                 label.hjust = 0.5,\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") ,\n",
    "  label.fill = \"grey90\",\n",
    "  label.buffer = unit(5, \"mm\")) +\n",
    "  theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  geom_label(aes(x = 553460, y = 5216442, label = \"B\"), size = 20)+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\") ,\n",
    "        legend.position = c(0.95,0.2),\n",
    "        legend.key.size =  unit(1, \"cm\"),\n",
    "        legend.text = element_text(size = 20),\n",
    "        legend.title=element_blank(),\n",
    "        legend.background = element_rect(linewidth = 0.5, \n",
    "                                         fill = \"white\",\n",
    "                                         linetype=\"solid\", \n",
    "                                  colour =\"black\"),\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20))\n",
    "\n",
    "\n",
    "ggsave(\"Figures/Figure1/Map_BB.png\",width= 10, height=10)  "
   ],
   "id": "2c5496c7-5405-4099-9d1f-7c434a665bce"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filelist<-list.files(\"Data/shp/\", pattern = \".shp\", recursive = T,full.names = T)\n",
    "mask_Morb <- filelist[str_detect(filelist,\"mask_land_intertidal_Morbihan\")] %>% \n",
    "  read_sf()\n",
    "\n",
    "Flight_Morbihan<-data.frame(\n",
    "  Name=c(\n",
    "  \"Arz\",\n",
    "  \"Duer\"\n",
    "  ) ,\n",
    "  Description = c(\"12m Flight\",\n",
    "                  \"12m and 120m Flight\"),\n",
    "  Long=c(-2.795961,\n",
    "         -2.746186) ,\n",
    "  Lat=c(47.603260, \n",
    "        47.543877) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(crs(mask_Morb))  %>% \n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)\n",
    "\n",
    "Morbihan_plot<-ggplot() +\n",
    "  scale_fill_manual(\"\", labels = c(\"Intertidal area\", \"Land mask\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\"))+\n",
    "  geom_sf(data = mask_Morb, mapping = aes(fill = Type),linewidth=0.05,alpha=0.93,colour=\"grey30\")+\n",
    "      coord_sf(xlim=c(503019,526177) ,\n",
    "          ylim=c(5257824,5280982))+\n",
    "  ggforce::geom_mark_ellipse(data=Flight_Morbihan,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = Name,\n",
    "                     group = Name),\n",
    "                     # description=Description),\n",
    "                 size=0.3,\n",
    "                 fill = \"goldenrod\",\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 show.legend=F,\n",
    "                 con.size = 2,\n",
    "                 label.hjust = 0.5,\n",
    "                 label.fontsize = 25,\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") ,\n",
    "  label.fill = \"grey90\",\n",
    "  label.buffer = unit(5, \"mm\")) +\n",
    "  theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  geom_label(aes(x = 505000, y = 5280982-1981, label = \"A\"), size = 20)+\n",
    "  scale_x_continuous(breaks = seq(-3, -2.6, by = 0.1))+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\"),\n",
    "        legend.position = \"none\",\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20))\n",
    "\n",
    "\n",
    "  ggsave(\"Figures/Figure1/Map_Morbihan.png\",width= 10, height=10)  "
   ],
   "id": "5ba5e56c-4e76-4f94-82b2-a974d75b24bf"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "layout <- c(\n",
    "  patchwork::area(t = 1, l = 1, b = 10, r = 10),\n",
    "  patchwork::area(t = 1, l = 11, b = 10, r = 20),\n",
    "  patchwork::area(t = 11, l = 1, b = 20, r = 10),\n",
    "  patchwork::area(t = 11, l = 11, b = 20, r = 20)\n",
    ")\n",
    "# plot(layout)\n",
    "\n",
    "Tot <- magick::image_ggplot(magick::image_read(\"Figures/Figure1/Map_Drone_Sites.png\"), interpolate = T)\n",
    "BB <- magick::image_ggplot(magick::image_read(\"Figures/Figure1/Map_BB.png\"), interpolate = T)\n",
    "Mor <- magick::image_ggplot(magick::image_read(\"Figures/Figure1/Map_Morbihan.png\"), interpolate = T)\n",
    "Av <- magick::image_ggplot(magick::image_read(\"Figures/Figure1/Map_Aveiro.png\"), interpolate = T)\n",
    "\n",
    "\n",
    "a<-Tot+Mor+BB+Av+patchwork::plot_layout(design=layout)\n",
    "\n",
    "ggsave(\"./Figures/Fig1_Map_Drone_Sites.png\",a, width= 10, height=10, dpi = 800)"
   ],
   "id": "e473cad7-f4d0-4411-be71-b1f532c21f0b"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Fig1_Map_Drone_Sites.png\")"
   ],
   "id": "cell-fig-map"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Field sampling\n",
    "\n",
    "### 2.2.1 Drone acquisition\n",
    "\n",
    "At each location, a DJI Matrice 200 quadcopter drone equipped with a Micasense RedEdge Dual MX multispectral camera was flown to take 1.2 million pixel reflectance photographs with ten spectral bands ranging from the blue to the near infrared (NIR): 444, 475, 531, 560, 650, 668, 705, 717, 740 and 840 nm. To ensure consistent lighting conditions across flight paths, the drone’s trajectory was aligned to maintain a solar azimuth angle of 90 degrees. An overlap of 70% and 80% (side and front respectively) between each image was set for each flight. A downwelling light sensor (DLS2) was used to acquire irradiance data concomitantly with the camera measurements. Raw data were calibrated in reflectance using a calibration panel reflective at ~50% provided by the manufacturer. Across all sites, flights were made at two different altitudes : 12 m or/and 120 m. (<a href=\"#tbl-flights\" class=\"quarto-xref\">Table 1</a>)."
   ],
   "id": "12e37872-3258-472b-b906-6287ac69757d"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df<-data.frame(\n",
    "  Country = c(\n",
    "    rep(\"France\",5),\n",
    "    rep(\"Portugal\",4)\n",
    "  ),\n",
    "  Site = c(\n",
    "    rep(\"Gulf of Morbihan\",3),\n",
    "    rep(\"Bourgneuf Bay\",2),\n",
    "    rep(\"Aveiro Lagoon\",4)\n",
    "  ),\n",
    "  Name = c(\n",
    "    \"Arz Island\",\n",
    "    \"Duer\",\n",
    "    \"Duer\",\n",
    "    \"Barbâtre\",\n",
    "    \"L'Epine\",\n",
    "    \"Marinha Lanzarote\",\n",
    "    \"Mataducos\",\n",
    "    \"Gafanha\",\n",
    "    \"Gafanha\"\n",
    "  ),\n",
    "  Altitude = c(\n",
    "    \"12m\",\n",
    "    \"12m\",\n",
    "    \"120m\",\n",
    "    \"120m\",\n",
    "    \"120m\",\n",
    "    \"120m\",\n",
    "    \"120m\",\n",
    "    \"120m\",\n",
    "    \"12m\"\n",
    "  ),\n",
    "  Utility = c(\n",
    "    \"Training\",\n",
    "    \"Training\",\n",
    "    \"Validation\",\n",
    "    \"Validation\",\n",
    "    \"Validation\",\n",
    "    \"Validation\",\n",
    "    \"Validation\",\n",
    "    \"Validation\",\n",
    "    \"Training\"\n",
    "  )\n",
    ")\n",
    "\n",
    "brdr1 <- fp_border_default(color = \"black\", width = 1.5)\n",
    "brdr2 <- fp_border_default(color = \"grey\", width = 1.5)\n",
    "brdr3 <- fp_border_default(color = \"grey40\", width = 1.5)\n",
    "\n",
    " i = 16.5 # width of the side borders in the word_document output (in centimeters)\n",
    " w = i*0.3937 # width of the side borders in the word_document output (in inches)\n",
    " \n",
    "flx1<-flextable(df)  %>%\n",
    "  flextable::width(width = (w/(ncol(df)))) %>% \n",
    "  merge_v(j = 1) %>%\n",
    "  merge_v(j = 2) %>%\n",
    "    border_remove() %>%\n",
    "    hline_top(border=brdr1) %>%\n",
    "    hline(i=5, border=brdr1) %>%\n",
    "    hline(i=1,j = 3:5, border=brdr2) %>%\n",
    "    hline(i=2, border=brdr2) %>%\n",
    "    hline(i=3, j = 2:5, border=brdr3) %>%\n",
    "    hline(i=4, j = 3:5, border=brdr2) %>%\n",
    "    hline(i=6, border=brdr2) %>%\n",
    "    hline(i=6, j = 1:2, border=brdr1) %>%\n",
    "    hline(i=7, border=brdr2) %>%\n",
    "    hline(i=8, border=brdr2) %>%\n",
    "    hline(i=9, border=brdr1) %>%\n",
    "    vline(j=2, border = brdr1) %>%\n",
    "    flextable::align(align = \"center\",part = \"all\") %>%\n",
    "    flextable::align(align = \"center\",part = \"header\") %>%\n",
    "  # set_caption(caption = \"Table 1 List of drone Flight, summarising the date, the altitude and the purpose of each flight.\") %>%\n",
    "  set_table_properties(layout = \"autofit\") \n",
    "\n",
    "\n",
    "\n",
    "save_as_image(flx1, \"Figures/table_flights.png\", res = 300)"
   ],
   "id": "a7f82ec8-89db-407d-bacb-f1c33774dc2c"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "knitr::include_graphics(\"Figures/table_flights.png\")"
   ],
   "id": "f8782530-1412-497e-bf5d-780649757583"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2.2 Ground Control Points\n",
    "\n",
    "Before each flight, targets used as Ground Control Points (GCPs) were distributed over the study site and georeferenced with a Trimble © Geo XH 6000 differential GPS (dGPS). GCPs were used to correct georeferencing imprecision of orthomosaics with an horizontal and vertical accuracy of 10cm. dGPS was also used to georeference quadrats of 0.25 m² assessing the presence or absence of five key taxonomic classes of intertidal vegetation : Bacillariophyceae (Unicellular benthic diatoms forming biofilms at the sediment surface during low tide), Phaeophyceae (brown macroalgae), Magnoliopsida (dwarf seagrass), Chlorophyceae (green macroalgae) and Rhodophyceae (red macroalgae) (<a href=\"#fig-vegetation\" class=\"quarto-xref\">Figure 2</a>). Pictures of each quadrat were uploaded online to the Global Biodiversity Information Facility (GBIF) platform ([Davies et al., 2023b](#ref-BedeGbif)), a public open portal to store and share biodiversity data. Each photograph was also processed to estimate the percent cover of each type of vegetation using an image processing software (imageJ). For all quadrats, the hyperspectral reflectance signatures of each vegetation class was measured using an ASD FieldSpec HandHeld 2 spectroradiometer, which acquires reflectance between 325 and 1075 nm, with 1 nm of spectral resolution. Hyperspectral signatures from quadrats serve for dual purposes: they validate the radiometric calibration of drone data and contribute to error reduction in photo interpretation. The workflow developed in this study is presented in <a href=\"#fig-workflow\" class=\"quarto-xref\">Figure 3</a>."
   ],
   "id": "e79d7e31-a908-4e20-bb49-445b630791a1"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "micasense<-list.files(pattern = \"Wavelength_Micasense_DualMX.csv\",recursive = T) %>% \n",
    "  read.delim(sep = \";\")\n",
    "\n",
    "imglist<-\n",
    "  data.frame(path = list.files(pattern = \".tif\", recursive = T, full.names = T)) %>% \n",
    "  mutate(folder = case_when(str_detect(path, \"Data\") ~ \"Input\",\n",
    "                            TRUE ~ \"Output\"),\n",
    "         alt = case_when(str_detect(path, \"high\") ~ \"high\",\n",
    "                         str_detect(path, \"High\") ~ \"high\",\n",
    "                         TRUE ~ \"low\"),\n",
    "         site = case_when(str_detect(path, \"Gafanha\") ~ \"Gafanha\",\n",
    "                          str_detect(path, \"Arz\") ~ \"Arz\",\n",
    "                          TRUE ~ \"Duer\"),\n",
    "         type = case_when(str_detect(path, \"pred\") ~ \"Pred\",\n",
    "                          str_detect(path, \"NDVI\") ~ \"NDVI\",\n",
    "                          str_detect(path, \"Pred\") ~ \"Pred\",\n",
    "                          TRUE ~ \"ortho\"),\n",
    "         resampled = case_when(str_detect(path, \"resampled\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         correction = case_when(str_detect(path, \"corrected_per_bands\") ~ \"Bands\",\n",
    "                          str_detect(path, \"corrected.tif\") ~ \"lm_all\",\n",
    "                          TRUE ~ \"NA\")) %>% \n",
    "  dplyr::filter(alt != \"NA\",\n",
    "                !str_detect(path,\".xml\"),\n",
    "                !str_detect(path,\".enp\"))\n",
    "\n",
    "gaf_low_multi<- imglist %>% \n",
    "  filter(site == \"Gafanha\",\n",
    "         alt == \"low\",\n",
    "         type == \"ortho\",\n",
    "         resampled == \"FALSE\",\n",
    "         folder == \"Input\") %>%\n",
    "  slice(1) %>% \n",
    "  pull(path) %>%\n",
    "  rast() \n",
    " names(gaf_low_multi)<-c(paste0(\"B\",c(1:10)))\n",
    "\n",
    " Gafanha_low_pred<-\"./Data/image/prediction/NewPreds/AveiroLowPrediction_nnNewStanRawMetrics_Preds_NoXantho_7.tif\" %>% \n",
    "     rast()\n",
    " names(Gafanha_low_pred)<-\"pred\"\n",
    "\n",
    "\n",
    "# arz_multi<- imglist %>% \n",
    "#   filter(site == \"Arz\", \n",
    "#          type == \"ortho\") %>% \n",
    "#   slice(1) %>% \n",
    "#   pull(path) %>% \n",
    "#   rast()\n",
    "#  names(arz_multi)<-c(paste0(\"B\",c(1:10)))\n",
    "# \n",
    "# arz_pred<- imglist %>% \n",
    "#   filter(site == \"Arz\", \n",
    "#          type == \"Pred\",\n",
    "#          str_detect(path, \"7by7\")) %>% \n",
    "#   slice(1) %>% \n",
    "#   pull(path) %>% \n",
    "#   rast()\n",
    "#  names(arz_pred)<-\"pred\"\n",
    " \n",
    " \n",
    " \n",
    "gaf<-rast(list(gaf_low_multi,Gafanha_low_pred)) %>% \n",
    "  as.data.frame(xy = F) %>% \n",
    "  dtplyr::lazy_dt() %>% \n",
    "  dplyr::filter(B1 != 65535,\n",
    "                pred != 0) %>%\n",
    "  as_tibble() %>% \n",
    "  rename(Class = \"pred\") %>%\n",
    "  mutate(pxl_ID = c(1:nrow(.))) %>% \n",
    "  pivot_longer(-c(Class,pxl_ID), names_to = \"band\",values_to = \"value\") %>% \n",
    "  left_join(micasense, by = c(\"band\" = \"Band_ID\")) %>% \n",
    "  group_by(Class, Wavelength) %>% \n",
    "  dplyr::summarise(mean = mean(value),\n",
    "                   sd = sd(value),\n",
    "                   max = max(value),\n",
    "                   min = min(value),\n",
    "                   median = median(value)) %>%  \n",
    "  mutate(Class = case_when(Class == 1 ~ \"Bare Sediment\",\n",
    "                              Class == 2 ~ \"D - Chloropyceae\",\n",
    "                              Class == 3 ~ \"Bare Sediment\",\n",
    "                              Class == 4 ~ \"A - MagnoliopsidaL\",\n",
    "                              Class == 5 ~ \"E - Microphytobenthos\",\n",
    "                              Class == 6 ~ \"A - Magnoliopsida\",\n",
    "                              Class == 7 ~ \"B - Phaeophyceae\",\n",
    "                              Class == 8 ~ \"C - Rhodophyceae\",\n",
    "                              Class == 9 ~ \"Sun Glint\",\n",
    "                              Class == 10 ~ \"Water\",\n",
    "                              Class == 11 ~ \"Xanthophyceae\",\n",
    "                              TRUE ~ \"NA\"))\n",
    "\n",
    "spectra<-gaf %>% \n",
    "  dplyr::filter(Class != \"Bare Sediment\",\n",
    "                Class != \"Sun Glint\",\n",
    "                Class != \"Water\",\n",
    "                Class != \"Xanthophyceae\",\n",
    "                Class != \"A - MagnoliopsidaL\") %>% \n",
    "  ggplot(aes(x=Wavelength, y = mean/10000, color = Class))+\n",
    "  geom_line(linewidth = 0.5)+\n",
    "  scale_color_manual(values=c(\"#389318\", \"#873e23\", \"#b3002d\", \"#b3ff1a\", \"#DAA520\"))+\n",
    "  ylab(\"Reflectance\")+\n",
    "  xlab(\"Wavelength (nm)\")+\n",
    "  theme_Bede()+\n",
    "  theme(axis.text.x = element_text(size = 5),\n",
    "        axis.text.y = element_text(size = 5),\n",
    "        axis.title.x = element_text(size = 7),\n",
    "        axis.title.y = element_text(size = 7),\n",
    "        legend.position = c(0.2, 0.7),\n",
    "        plot.background = element_rect(fill = rgb(1,1,1, alpha=0), colour = NA),\n",
    "        legend.title=element_blank(),\n",
    "        legend.text = element_text(size = 5),\n",
    "        legend.key.size = unit(0.25, \"cm\"),\n",
    "        legend.text.align = 0,\n",
    "        panel.border = element_blank(), \n",
    "        panel.background = element_blank(),\n",
    "        axis.line = element_line(color=\"black\"))\n",
    "\n",
    "ggsave(\"Figures/Spectral_shapes.png\",spectra, bg = \"transparent\",  height = 2, width = 3, dpi = 2000)\n",
    "\n",
    "sp <- image_ggplot(magick::image_read(\"Figures/Spectral_shapes.png\"), interpolate = T)\n",
    "\n",
    "# sp<-ggplot()+\n",
    "#   geom_spatraster_rgb(data = rast(\"Figures/Spectral_shapes.png\"),maxcell = 4009008)+\n",
    "#   coord_equal()+\n",
    "#     theme_void()+\n",
    "#   theme(axis.text = element_blank(), \n",
    "#         axis.title = element_blank(), \n",
    "#         axis.ticks = element_blank(),\n",
    "#         panel.background = element_rect(fill='transparent'),\n",
    "#         plot.background = element_rect(fill='transparent', color=NA),\n",
    "#         axis.ticks.length = unit(0, \"pt\"),\n",
    "#         panel.grid.major=element_blank(), \n",
    "#         panel.grid.minor=element_blank(), \n",
    "#         plot.margin = margin(0, 0, 0, 0, \"pt\"))\n",
    "# \n",
    "# \n",
    "# Magno <- magick::image_read(\"Data/Pictures/Figure2/Magnoliopsida_cropped2.png\")%>%\n",
    "#   # magick::image_rotate(90)%>%\n",
    "#   grid::rasterGrob()\n",
    "# \n",
    "# Brown <- magick::image_read(\"Data/Pictures/Figure2/Phaeophyceae_cropped.png\")%>%\n",
    "#   grid::rasterGrob()\n",
    "# \n",
    "# Red <- magick::image_read(\"Data/Pictures/Figure2/Rhodophyceae_cropped.png\")%>% \n",
    "#   grid::rasterGrob()\n",
    "# \n",
    "# green <- magick::image_read(\"Data/Pictures/Figure2/Chlorophyceae_cropped.png\")%>% \n",
    "#   grid::rasterGrob()\n",
    "# \n",
    "# MPB <- magick::image_read(\"Data/Pictures/Figure2/Bacillariophyceae_cropped.png\")%>% \n",
    "#   grid::rasterGrob()\n",
    "# \n",
    "# Lines<-data.frame(x=c(10,20,10),\n",
    "#                   y=c(-0.1,-0.1,10),\n",
    "#                   xend=c(10,20,30),\n",
    "#                   yend=c(20,20,10))\n",
    "# labels<-data.frame(name = c(\"A\",\"B\",\"C\",\"D\",\"E\"),\n",
    "#                    x= c(1,11,21,11,21),\n",
    "#                    y=c(19,19,19,1,1))\n",
    "# \n",
    "# a<-ggplot(expand_grid(x = 0:30,y=0:20),aes(x = x, y = y) ) +\n",
    "#   geom_point(alpha=0) +\n",
    "#   coord_fixed()+\n",
    "#   ggplot2::annotation_custom(Magno, 0,10, 0, 20) +\n",
    "#   ggplot2::annotation_custom(green, 10,20, 10, 20)+\n",
    "#   ggplot2::annotation_custom(Red, 10,20, 0, 10)+\n",
    "#   ggplot2::annotation_custom(Brown, 20,30, 10, 20)+\n",
    "#   ggplot2::annotation_custom(MPB, 20,30, 0,10)+\n",
    "#   geom_segment(data=Lines,aes(x=x,y=y,xend=xend,yend=yend),\n",
    "#                colour=\"white\",linewidth=2)+\n",
    "#   ggplot2::annotation_custom(ggplotGrob(spectra), 5,25, 5,15)+\n",
    "#   geom_label(data=labels,aes(x=x,y=y,label=name), size = 10)+\n",
    "#   theme_void()+\n",
    "#   theme(axis.title.x = element_blank(),\n",
    "#         axis.title.y = element_blank(),\n",
    "#         axis.text.x = element_blank(),\n",
    "#         axis.text.y = element_blank(),\n",
    "#         axis.ticks.x = element_blank(),\n",
    "#         axis.ticks.y = element_blank(),\n",
    "#         plot.margin = grid::unit(c(0.01,0.01,0.01,0.01), \"mm\"),\n",
    "#         # rect = element_blank(),\n",
    "#         panel.margin = grid::unit(c(0.01, 0.01, 0.01, 0.01), \"mm\"))\n",
    "# \n",
    "# Lines1<-data.frame(x=c(0,10,20,30),\n",
    "#                   y=c(10,0,0,0),\n",
    "#                   xend=c(40.1,10,20,30),\n",
    "#                   yend=c(10,10,10,10))\n",
    "# \n",
    "# labels1<-data.frame(name = c(\"A\",\"B\",\"C\",\"D\",\"E\"),\n",
    "#                    x= c(1,1,11,21,31),\n",
    "#                    y=c(29,9,9,9,9))\n",
    "# \n",
    "# ggplot(expand_grid(x = 0:40,y=0:30),aes(x = x, y = y) ) +\n",
    "#   geom_point(alpha=0) +\n",
    "#   coord_fixed()+\n",
    "#   ggplot2::annotation_custom(Magno, 0,10, 10, 30) +\n",
    "#   ggplot2::annotation_custom(Brown, 0,10, 0, 10)+\n",
    "#   ggplot2::annotation_custom(Red, 10,20, 0, 10)+\n",
    "#   ggplot2::annotation_custom(green, 20,30, 0, 10)+\n",
    "#   ggplot2::annotation_custom(MPB, 30,40, 0,10)+\n",
    "#   geom_segment(data=Lines1,aes(x=x,y=y,xend=xend,yend=yend),\n",
    "#                colour=\"white\",linewidth=2)+\n",
    "#   ggplot2::annotation_custom(ggplotGrob(spectra), 10,40, 10,30)+\n",
    "#   geom_label(data=labels1,aes(x=x,y=y,label=name), size = 10)+\n",
    "#   theme_void()+\n",
    "#   theme(axis.title.x = element_blank(),\n",
    "#         axis.title.y = element_blank(),\n",
    "#         axis.text.x = element_blank(),\n",
    "#         axis.text.y = element_blank(),\n",
    "#         axis.ticks.x = element_blank(),\n",
    "#         axis.ticks.y = element_blank(),\n",
    "#         plot.margin = grid::unit(c(0.01,0.01,0.01,0.01), \"mm\"),\n",
    "#         # rect = element_blank(),\n",
    "#         panel.margin = grid::unit(c(0.01, 0.01, 0.01, 0.01), \"mm\"),\n",
    "#         legend.text.align = 0)\n",
    "\n",
    "Magno<-ggplot()+\n",
    "  geom_spatraster_rgb(data = rast(\"Data/Pictures/Figure2/Magnoliopsida_cropped2.png\"))+\n",
    "  coord_equal()+\n",
    "  geom_label(aes(x = 50,y= 950, label = \"A\"))+\n",
    "    theme_void()+\n",
    "  theme(axis.text = element_blank(), \n",
    "        axis.title = element_blank(), \n",
    "        axis.ticks = element_blank(), \n",
    "        axis.ticks.length = unit(0, \"pt\"),\n",
    "        panel.grid.major=element_blank(), \n",
    "        panel.grid.minor=element_blank(), \n",
    "        plot.margin = margin(0, 0, 0, 0, \"pt\"))\n",
    "\n",
    "Brown<-ggplot()+\n",
    "  geom_spatraster_rgb(data = rast(\"Data/Pictures/Figure2/Pheophyceae_cropped.png\"))+\n",
    "  coord_equal()+\n",
    "  geom_label(aes(x = 50,y=450, label = \"B\"))+\n",
    "    theme_void()+\n",
    "  theme(axis.text = element_blank(), \n",
    "        axis.title = element_blank(), \n",
    "        axis.ticks = element_blank(), \n",
    "        axis.ticks.length = unit(0, \"pt\"),\n",
    "        panel.grid.major=element_blank(),\n",
    "        panel.grid.minor=element_blank(),\n",
    "        plot.margin = margin(0, 0, 0, 0, \"pt\"))\n",
    "\n",
    "Red <- ggplot()+\n",
    "  geom_spatraster_rgb(data = rast(\"Data/Pictures/Figure2/Rodophyceae_cropped.png\"))+\n",
    "  coord_equal()+\n",
    "  geom_label(aes(x = 50,y=450, label = \"C\"))+\n",
    "    theme_void()+\n",
    "  theme(axis.text = element_blank(), \n",
    "        axis.title = element_blank(), \n",
    "        axis.ticks = element_blank(), \n",
    "        axis.ticks.length = unit(0, \"pt\"),\n",
    "        panel.grid.major=element_blank(), \n",
    "        panel.grid.minor=element_blank(), \n",
    "        plot.margin = margin(0, 0, 0, 0, \"pt\"))\n",
    "\n",
    "\n",
    "green <- ggplot()+\n",
    "  geom_spatraster_rgb(data = rast(\"Data/Pictures/Figure2/Chlorophyceae_cropped.png\"))+\n",
    "  coord_equal()+\n",
    "  geom_label(aes(x = 50,y=450, label = \"D\"))+\n",
    "    theme_void()+\n",
    "  theme(axis.text = element_blank(), \n",
    "        axis.title = element_blank(), \n",
    "        axis.ticks = element_blank(), \n",
    "        axis.ticks.length = unit(0, \"pt\"),\n",
    "        panel.grid.major=element_blank(), \n",
    "        panel.grid.minor=element_blank(), \n",
    "        plot.margin = margin(0, 0, 0, 0, \"pt\"))\n",
    "  \n",
    "  \n",
    "\n",
    "MPB <- ggplot()+\n",
    "  geom_spatraster_rgb(data = rast(\"Data/Pictures/Figure2/Bacillariophyceae_cropped.png\"))+\n",
    "  coord_equal()+\n",
    "  geom_label(aes(x = 50,y=450, label = \"E\"))+\n",
    "  theme_void()+\n",
    "  theme(axis.text = element_blank(), \n",
    "        axis.title = element_blank(), \n",
    "        axis.ticks = element_blank(), \n",
    "        axis.ticks.length = unit(0, \"pt\"),\n",
    "        panel.grid.major=element_blank(), \n",
    "        panel.grid.minor=element_blank(), \n",
    "        plot.margin = margin(0, 0, 0, 0, \"pt\"))\n",
    "\n",
    "layout <- c(\n",
    "  patchwork::area(t = 10, l = 0, b = 30, r = 10),\n",
    "    patchwork::area(t = 0, l = 0, b = 10, r = 10),\n",
    "    patchwork::area(t = 0, l = 10, b = 10, r = 20),\n",
    "    patchwork::area(t = 0, l = 20, b = 10, r = 30),\n",
    "    patchwork::area(t = 0, l = 30, b = 10, r = 40),\n",
    "    patchwork::area(t = 10, l = 10, b = 30, r = 40)\n",
    ")\n",
    "\n",
    "layout1 <- c(\n",
    "  patchwork::area(t = 1, l = 1, b = 20, r = 10),\n",
    "    patchwork::area(t = 20, l = 1, b = 29, r = 10),\n",
    "    patchwork::area(t = 20, l = 9, b = 29, r = 18),\n",
    "    patchwork::area(t = 20, l = 17, b = 29, r = 26),\n",
    "    patchwork::area(t = 20, l = 25, b = 29, r = 34),\n",
    "    patchwork::area(t = 1, l = 9, b = 20, r = 34)\n",
    ")\n",
    "\n",
    "# plot(layout1)\n",
    "\n",
    "plot<-Magno+\n",
    "  Brown+\n",
    "  Red+\n",
    "  green+\n",
    "  MPB+\n",
    "  sp+\n",
    "  patchwork::plot_layout(design=layout1)\n",
    "\n",
    "ggsave(\"Figures/Spectral_shapes_total.png\",plot, height = 7, width = 10, dpi = 800)"
   ],
   "id": "f2dfe455-a3dc-4ed4-8a68-8b903a8b5ea0"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"Figures/Low_res/Spectral_shapes_total.png\")"
   ],
   "id": "cell-fig-vegetation"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3 Drone Processing\n",
    "\n",
    "A structure-from-motion photogrammetry software (Agisoft Metashape) was used to process images to obtain multispectral orthomosaics of each flight. The workflow for orthomosaicking was the same for every flight. First, tying key points were detected inside of each image and between overlapping images in order to obtain a sparse point cloud. This cloud was cleaned using reprojection accuracy metric in order to remove noisy points. A dense point cloud was then produced using a structure from motion algorithm. A surface interpolation of this dense point cloud was made to obtain a digital surface model (DSM), used to reconstruct the multispectral ortho-image. Low altitude drone flights produced ortho-images with a very high spatial resolution (8 mm per pixel), making it efficient to visually distinguish between the various types of vegetation. High altitude fights on the other hand allowed to cover large areas and produced images with a pixel size of 80 mm (<a href=\"#tbl-flights\" class=\"quarto-xref\">Table 1</a>).\n",
    "\n",
    "## 2.4 General Workflow"
   ],
   "id": "bb0c5049-1ef5-40bf-b4fa-147baffdc629"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "layout<-\"Data/Figure2_Workflow\" %>%\n",
    "  list.files(\"V4.txt\",full.names = T, recursive = T) %>%\n",
    "  read_delim(delim = \"\\t\", escape_double = FALSE,\n",
    "    trim_ws = TRUE) %>%\n",
    "  dplyr::filter(Evaluate == T)\n",
    "\n",
    "\n",
    "layout$Text<- gsub(\"alaligne\",\"\\n\", layout$Text)\n",
    "\n",
    "boxes<- layout %>% \n",
    "  dplyr::filter(Type == \"Box\",\n",
    "                shape == \"square\")\n",
    "\n",
    "\n",
    "text<- layout %>% \n",
    "  dplyr::filter(Type == \"Box\")\n",
    "\n",
    "diamonds<- layout %>% \n",
    "  dplyr::filter(Type == \"Box\",\n",
    "                shape == \"diamond\")\n",
    "diff = 0.1\n",
    "for (i in 1:nrow(diamonds)) {\n",
    "  diam_a<-diamonds[i,]\n",
    "  \n",
    "df<-data.frame(  \n",
    "  x = c(diam_a$xmin - diff, diam_a$xmin + diff, diam_a$xmax + diff, diam_a$xmax - diff),\n",
    "  y = c(diam_a$ymin, diam_a$ymax, diam_a$ymax, diam_a$ymin),\n",
    "  Type = rep(diam_a$Type,4),\n",
    "  Text = rep(diam_a$Text,4),\n",
    "  Color = rep(diam_a$color,4),\n",
    "  Evaluate = rep(diam_a$Evaluate,4),\n",
    "  Resolution =  rep(diam_a$Resolution,4),\n",
    "  Evaluate_bigbox = rep(diam_a$Evaluate_bigbox,4),\n",
    "  alpha = rep(diam_a$alpha,4),\n",
    "  shape =  rep(diam_a$shape,4)\n",
    ")\n",
    "  if(i == 1){\n",
    "    diamond_list<-list(df)\n",
    "  }else{\n",
    "    diamond_list[[i]]<-df\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "Arrows<- layout %>% \n",
    "  dplyr::filter(Type == \"Arrow\",\n",
    "                Evaluate == T)\n",
    "\n",
    "Group_box<-layout %>% \n",
    "  dplyr::filter(Type == \"Box\",\n",
    "                Evaluate_bigbox == T) %>% \n",
    "  group_by(Group) %>% \n",
    "  dplyr::summarise(xmin = min(xmin)-0.1,\n",
    "                   xmax = max(xmax)+0.1,\n",
    "                   ymin = min(ymin)-0.1,\n",
    "                   ymax = max(ymax)+0.1,\n",
    "                   alpha = 0.1,\n",
    "                   color = color)\n",
    "\n",
    "resolution_box<-layout %>% \n",
    "  dplyr::filter(Type == \"Box\",\n",
    "                Resolution != \"NA\") %>% \n",
    "  group_by(Resolution) %>% \n",
    "  dplyr::summarise(xmin = min(xmin)-0.1,\n",
    "                   xmax = max(xmax)+0.1,\n",
    "                   ymin = min(ymin)-0.1,\n",
    "                   ymax = max(ymax)+0.1,\n",
    "                   alpha = 0.1,\n",
    "                   color = color)\n",
    "\n",
    "legend = data.frame(color = unique(layout$color)) %>% \n",
    "  dplyr::filter(!is.na(color)) %>% \n",
    "  mutate(label = case_when(color == \"#EA6155\" ~ \"Input Data\",\n",
    "                           color == \"#DC9614\" ~ \"Pre-processing\",\n",
    "                           color == \"#52A4DA\" ~ \"Model Building\",\n",
    "                           color == \"#79DA52\" ~ \"Validation\",\n",
    "                           color == \"#DA9652\" ~ \"Output Data\",\n",
    "                           color == \"darkcyan\" ~ \"Final Product\",\n",
    "                           color == \"#DA5273\"  ~ \"Prediction\",\n",
    "                           TRUE ~ \"NA\")) %>% \n",
    "  arrange\n",
    "\n",
    "segment<-layout %>% \n",
    "  dplyr::filter(Type == \"Segment\")"
   ],
   "id": "41e53c72-0bc6-4dee-b712-712e70637258"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "building <- F\n",
    "\n",
    "if(building == T){\n",
    "  \n",
    "text_size = 6 #### 6 to plot it fullscreen in R,  1.8 for exporting with ggsave\n",
    "line_size =1 ####arrow_size 0.3 to export  1 to for R plotting\n",
    "\n",
    "legend_size = 1 #### 0.3 to export, 1 for R ploting\n",
    "legend_text = 10 #### 4 to export, 10 for R ploting\n",
    "}else{\n",
    "  \n",
    "text_size = 1.85 #### 6 to plot it fullscreen in R,  1.8 for exporting with ggsave\n",
    "line_size =0.3 ####arrow_size 0.3 to export  1 to for R plotting\n",
    "\n",
    "legend_size = 0.3 #### 0.3 to export, 1 for R ploting\n",
    "legend_text = 5 #### 5 to export, 10 for R ploting\n",
    "  \n",
    "}\n",
    "\n",
    "\n",
    "a<-ggplot()+\n",
    "  \n",
    "  statebins:::geom_rrect(data = Group_box,\n",
    "            mapping = aes(xmin = xmin,\n",
    "                          xmax= xmax,\n",
    "                          ymin = ymin,\n",
    "                          ymax = ymax),\n",
    "            alpha = 0.05,\n",
    "            size = 0.2,\n",
    "            show.legend = F,\n",
    "            fill = Group_box$color,\n",
    "            color = \"black\")+\n",
    "  \n",
    "    statebins:::geom_rrect(data = resolution_box,\n",
    "            mapping = aes(xmin = xmin,\n",
    "                          xmax= xmax,\n",
    "                          ymin = ymin,\n",
    "                          ymax = ymax,\n",
    "                          alpha = rep(c(\"Input Data\", \"Pre-processing\", \"Model Building\", \"Prediction\", \"Validation\", \"Output Data\", \"Final Product\"),2)),\n",
    "            fill = \"NA\",\n",
    "            linetype = 2,\n",
    "            size = 0.4,\n",
    "            color = \"black\")+\n",
    "  \n",
    "  scale_alpha_manual(name = \"\", \n",
    "                     values = c(1,1,1,1,1,1,1),\n",
    "                     labels = c(\"Input Data\", \"Pre-Processing\", \"Model Building\", \"Prediction\", \"Validation\", \"Output Data\", \"Final Product\"),\n",
    "                     guide = guide_legend(ncol = 1, override.aes = list(fill = c(\"#f94144\", \"#f3722c\", \"#f8961e\", \"#f9c74f\", \"#90be6d\", \"#43aa8b\", \"#577590\"),\n",
    "                                                                        shape = c(22),\n",
    "                                                                        linetype = c(1),\n",
    "                                                                        alpha = c(0.5)\n",
    "                                                                        )\n",
    "                                          )\n",
    "                     )+\n",
    "  \n",
    "  statebins:::geom_rrect(data = boxes,\n",
    "            mapping = aes(xmin = xmin,\n",
    "                          xmax= xmax,\n",
    "                          ymin = ymin,\n",
    "                          ymax = ymax\n",
    "                          ),\n",
    "            size = 0.2,\n",
    "            alpha = as.numeric(boxes$alpha),\n",
    "            fill = boxes$color,\n",
    "            color = \"black\")\n",
    "  \n",
    "  for(i in 1:length(diamond_list)){\n",
    "    poly_i<-diamond_list[[i]]\n",
    "    a<-a+\n",
    "      geom_polygon(data = poly_i, \n",
    "                   mapping = aes(x = x, \n",
    "                                 y = y\n",
    "                                 ),\n",
    "                  size = 0.2,\n",
    "                   alpha = poly_i$alpha,\n",
    "                   fill = poly_i$Color,\n",
    "                   color = \"black\")\n",
    "    \n",
    "  }\n",
    "\n",
    "if(building == T){\n",
    "    a<-a+\n",
    "    geom_text(data = text,\n",
    "            mapping = aes(x = (xmax+xmin)/2, y = (ymax + ymin)/2, label = (ID)), size = text_size)\n",
    "  \n",
    "}else{\n",
    "    a<-a+\n",
    "    geom_text(data = text,\n",
    "            mapping = aes(x = (xmax+xmin)/2, y = (ymax + ymin)/2, label = (Text)), size = text_size)\n",
    "}\n",
    "\n",
    "\n",
    "  a<-a+\n",
    "    geom_segment(data = Arrows, aes( x = xmin, xend = xmax, y = ymin, yend = ymax),\n",
    "    size=line_size, linejoin = \"mitre\", lineend = \"butt\",\n",
    "    arrow = arrow(length = unit(1, \"mm\"), type= \"closed\")\n",
    "    )+\n",
    "    geom_segment(data = segment, aes( x = xmin, xend = xmax, y = ymin, yend = ymax),\n",
    "    size=line_size, linejoin = \"mitre\", lineend = \"square\"\n",
    "    )+\n",
    "  \n",
    "  geom_text(aes(x = -5.4, y = -2, label = \"Flight height: 12 m \\n Pixel size : 8 mm\", angle = 90), size = text_size+1)+\n",
    "  \n",
    "  geom_text(aes(x = -5.4, y = -6.8, label = \"Flight height: 120 m \\n Pixel size : 80 mm\", angle = 90), size = text_size+1)+\n",
    "  \n",
    "  geom_text(aes(x = 4, y = 1.75, label = \"Pre Processing\"), size = text_size)\n",
    "    \n",
    "  if(building == F){\n",
    "  a<- a+ theme_void()\n",
    "  }\n",
    "  \n",
    "  a<-a+\n",
    "  theme(legend.position = c(0.88,0.3),\n",
    "        legend.text = element_text(size = legend_text),\n",
    "        legend.key.size = unit(legend_size, 'cm')\n",
    "        )+\n",
    "  scale_x_continuous(breaks = scales::breaks_width(0.5))+\n",
    "  scale_y_continuous(breaks = scales::breaks_width(0.5))\n",
    "  \n",
    "if(building == T){\n",
    "a\n",
    "}else{\n",
    "  ggsave(\"Figures/Figure3_workflow.png\",a, width = 1920, height = 1129, units = \"px\")\n",
    "}"
   ],
   "id": "97410b72-e420-415f-845c-5c6092a12993"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Figure3_workflow.png\")"
   ],
   "id": "cell-fig-workflow"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The spectral similarities of the reflectance signatures between intertidal green macrophytes (Magnoliopsida and Chlorophyceae) make their discrimination challenging using simple classification algorithms (<a href=\"#fig-vegetation\" class=\"quarto-xref\">Figure 2</a> F). To overcome this challenge, a deep learning classification method was trained, validated, and applied to each drone flight (<a href=\"#fig-workflow\" class=\"quarto-xref\">Figure 3</a>).\n",
    "\n",
    "### 2.4.1 Neural Network model building"
   ],
   "id": "ca7bb6fe-180d-4ae0-975b-a373f67757a8"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_training<-\"Data/training/TrainandValid_NewStanRawMetrics.csv\" %>% \n",
    "  read_delim(delim = \";\", escape_double = FALSE, trim_ws = TRUE) %>% \n",
    "  dplyr::filter(True_Class != \"Xanthophyceae\",\n",
    "                True_Class != \"Deep_Sediment\",\n",
    "                True_Class != \"Water\",\n",
    "                True_Class != \"SunGlint\",\n",
    "                True_Class != \"Bare_Sediment\") %>% \n",
    "  mutate(True_Class = case_when(True_Class == \"Clorophyta\" ~ \"Chlorophyceae\",\n",
    "                               True_Class == \"Rhodphyta\" ~ \"Rhodophyceae\",\n",
    "                               True_Class == \"MPB\" ~ \"Bacillariophyceae\",\n",
    "                               True_Class == \"Magnoliosida\" ~ \"Magnoliopsida\",\n",
    "                               True_Class == \"Phaeophyta\" ~ \"Phaeophyceae\",\n",
    "                               True_Class == \"Low_SPC\" ~ \"Magnoliopsida\",\n",
    "                               TRUE ~ \"NA\"\n",
    "                               )) %>% \n",
    "  group_by(True_Class) %>% \n",
    "  summarise(n = n()) %>% \n",
    "  rename(True = \"True_Class\", \n",
    "         Training_Pixels = \"n\")\n",
    "\n",
    "write.csv(df_training,\"Data/training/TrainandValid_NewStanRawMetrics_Processed.csv\",row.names = F)"
   ],
   "id": "fe35a913-4fd5-4b35-90be-9f76e1421f8b"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_training<-read.csv(\"Data/training/TrainandValid_NewStanRawMetrics_Processed.csv\")"
   ],
   "id": "c4228def-5dc1-42d1-9775-22c29ed7ea5d"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_validation<-\"Data/accuracy_assesment_low&high_flights.csv\" %>% \n",
    "  read_delim(delim = \",\", escape_double = FALSE, trim_ws = TRUE) %>% \n",
    "  dplyr::filter(True != \"Sediment\",\n",
    "                True != \"Sun Glint\",\n",
    "                True != \"Water\",\n",
    "                True != \"Xanthophyceae\") %>% \n",
    "  mutate(True = case_when(True == \"Green Algae\" ~ \"Chlorophyceae\",\n",
    "                          True == \"Microphytobenthos\" ~ \"Bacillariophyceae\",\n",
    "                          True == \"Rodophyceae\" ~ \"Rhodophyceae\",\n",
    "                          True == \"Pheophyceae\" ~ \"Phaeophyceae\",\n",
    "                          TRUE ~ True)) %>% \n",
    "  group_by(True) %>% \n",
    "  summarise(n = n()) \n",
    "\n",
    "write.csv(df_validation,\"Data/training/TrainandValid_NewStanRawMetrics_Processed_valid.csv\",row.names = F)"
   ],
   "id": "d9ef34ab-1123-4700-b282-eb33ccb3950c"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_validation<-read.csv(\"Data/training/TrainandValid_NewStanRawMetrics_Processed_valid.csv\")\n",
    "\n",
    " i = 16.5 # width of the side borders in the word_document output (in centimeters)\n",
    " w = i*0.3937 # width of the side borders in the word_document output (in inches)\n",
    " \n",
    "df_flx2<-df_training %>% \n",
    "  left_join(df_validation, by = \"True\") %>% \n",
    "  rename(`Validation Pixels` = \"n\",\n",
    "         Class = \"True\",\n",
    "         `Training Pixels` = \"Training_Pixels\")\n",
    "flx2 <- df_flx2 %>% \n",
    "  flextable() %>% \n",
    "  flextable::width(width = (w/(ncol(df_flx2)))) %>% \n",
    "  # border_remove() %>%\n",
    "  flextable::align(align = \"center\",part = \"all\") %>%\n",
    "  flextable::align(align = \"center\",part = \"header\") %>% \n",
    "  set_table_properties(layout = \"autofit\") \n",
    "\n",
    "save_as_image(flx2, \"Figures/table_validation_px.png\", res = 300)"
   ],
   "id": "eb2dfed1-ee80-4101-b363-d3bf09def9f0"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "knitr::include_graphics(\"Figures/table_validation_px.png\")"
   ],
   "id": "56a3c53e-8802-43b7-8cd1-4fd12dbdd430"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A dataset containing photo-interpreted drone reflectance pixels was built to train a Neural Network model with 2 hidden layers. The training pixels were categorized into seven different classes, representing the various habitats encountered at the different study sites: Sediment, Water, Chlorophyceae, Magnoliopsida, Bacillariophyceae, Phaeophyceae and Rhodophyceae. Only low-altitude flights (<a href=\"#tbl-flights\" class=\"quarto-xref\">Table 1</a>) were used for training because their 8 mm spatial resolution allowied to avoid spectral sub-pixel mixing and to accurately identify vegetation classes. More than 418,000 pixels at 8 mm resolution from the 3 training flights were used to train the model (<a href=\"#tbl-validationPX\" class=\"quarto-xref\">Table 2</a>). Twenty one variables were used by the model as predictors: the ten raw spectral bands of the Micasense RedEdge Dual MX multispectral camera (ranging from 444 nm to 840 nm), the same ten spectral bands standardized using a min/max transformation (<a href=\"#eq-std\" class=\"quarto-xref\">Equation 1</a> ; Cao et al. ([2017](#ref-Cao2017))) and the Normalized difference vegetation index (NDVI, <a href=\"#eq-ndvi\" class=\"quarto-xref\">Equation 2</a>). Standardisation of spectral bands is commonly used to eliminate the scaling differences between spectra and to limit the effect of biomass on the spectra shape ([Douay et al., 2022](#ref-Douay2022) ; [Davies et al., 2023a](#ref-Davies2023)).\n",
    "\n",
    "<span id=\"eq-std\">$$\n",
    "R_{i}^{*}(\\lambda) = \\frac{R_{i}(\\lambda) - min(R_{i})}{max(R_{i})- min(R_{i})}\n",
    " \\qquad(1)$$</span>\n",
    "\n",
    "where $R_{i}(\\lambda)$ is the reflectance at the wavelength $(\\lambda)$ of each individual spectra $(i)$, $min(R_{i})$, and $max(R_{i})$ are the minimum and maximum value of the spectra $(i)$\n",
    "\n",
    "<span id=\"eq-ndvi\">$$\n",
    "NDVI = \\frac{R(840nm)-R(668nm)}{R(840nm)+R(668nm)}\n",
    " \\qquad(2)$$</span>\n",
    "\n",
    "where $R(840nm)$ is the reflectance at 840 nm and $R(668nm)$ is the reflectance at 668 nm.\n",
    "\n",
    "### 2.4.2 Validation"
   ],
   "id": "d64fa9e6-1a2d-495c-b6c7-e3f12b5bd8d3"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df<-\"Output/Validation_data\" %>% \n",
    "  list.files(\"ResultsNewNN_NoXanthowithNDVIWithOldandBoat.csv\",recursive = T, full.names = T) %>% \n",
    "  read.csv() %>% \n",
    "  as_tibble() %>% \n",
    "  rename(Truth_ID = \"Truth\") %>% \n",
    "  mutate(Truth =  case_when(Truth_ID == 1 ~ \"Bare_Sediment\",\n",
    "                          Truth_ID == 2 ~ \"Chlorophyta\",\n",
    "                          Truth_ID == 3 ~ \"Bare_Sediment\",\n",
    "                          Truth_ID == 4 ~ \"Magnoliosida\",\n",
    "                          Truth_ID == 5 ~ \"Bare_Sediment\",\n",
    "                          Truth_ID == 6 ~ \"Magnoliosida\",\n",
    "                          Truth_ID == 7 ~ \"Phaeophyta\",\n",
    "                          Truth_ID == 8 ~ \"Rhodphyta\",\n",
    "                          Truth_ID == 9 ~ \"SunGlint\",\n",
    "                          Truth_ID == 10 ~ \"Water\",\n",
    "                          Truth_ID == 11 ~ \"Xanthophyceae\",\n",
    "                          Truth_ID == 12 ~ \"Spartina\",\n",
    "                          Truth_ID == 41 ~ \"Bleached seagrasses\",\n",
    "                          TRUE ~ \"NA\"),\n",
    "         Pred_Class = case_when(Pred_Class == \"Clorophyta\" ~ \"Chlorophyta\",\n",
    "                                Pred_Class == \"Rhodphyta\" ~ \"Rhodphyta\",\n",
    "                                Pred_Class == \"Deep_Sediment\" ~ \"Bare_Sediment\",\n",
    "                                Pred_Class == \"MPB\" ~ \"Bare_Sediment\",\n",
    "                                (Pred_Class == \"Low_SPC\" & NDVI > 0.2) ~ \"Magnoliosida\",\n",
    "                                (Pred_Class == \"Low_SPC\" & NDVI <= 0.2 )~ \"Bare_Sediment\",\n",
    "                          TRUE ~ Pred_Class)) %>% \n",
    "  dplyr::filter(Truth != \"Water\",\n",
    "                Truth  != \"Spartina\",\n",
    "                Truth != \"SunGlint\",\n",
    "                Truth != \"Bleached seagrasses\",\n",
    "                Truth != \"Xanthophyceae\",\n",
    "                # Pred_Class != \"Low_SPC\",\n",
    "                Pred_Class != \"Xanthophyceae\",\n",
    "                 Pred_Class != \"Water\",\n",
    "                 Pred_Class != \"SunGlint\",\n",
    "                !Flight == \"Handfisher_1_old\",\n",
    "                !Flight == \"Handfisher_2_old\",\n",
    "                !Flight == \"Handfisher_2\") %>%\n",
    "  dplyr::select(-NDVI)\n",
    "\n",
    "\n",
    "\n",
    "count_validation_pixel<-nrow(df)\n",
    "write.csv(count_validation_pixel,\"Output/count_validation_pixel.csv\", row.names = F)"
   ],
   "id": "b4b00bb7-88ee-49d2-9890-ca37f0eeddc3"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_validation_pixel<-read.csv(\"Output/count_validation_pixel.csv\") %>% \n",
    "  pull(x)"
   ],
   "id": "e7277650-83d5-4d74-88ca-677242530407"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The workflow of this study revolves around two distinct flight heights (12 m and 120 m, <a href=\"#fig-workflow\" class=\"quarto-xref\">Figure 3</a>) where ensuring consistency between reflectances at both heights is crucial. This comparison was conducted at sites where low and high altitude flights overlapped. The low altitude flights were resampled to the same spatial resolution and grid as the high flights using a median resampling method. Reflectance values were then extracted, and a scatterplot was generated. Subsequently, the slope of the linear model and the coefficient of determination (R²) were computed.\n",
    "\n",
    "The model was applied to all flights at both 12 and 120 m of altitude. In situ information on georeferenced class type and percent cover collected during each flight was used to assess the model accuracy. These images were used to construct a validation dataset indicating the presence or absence of each class. Additionally to the quadrat-based validation dataset, polygons of each class were photo interpreted in order to increase the number of pixels of the validation dataset. A confusion matrix, along with precision metrics such as global accuracy, sensitivity, specificity, and Kappa coefficient, was generated for each sites. All validation matrices were then aggregated to create a unique matrix . Altogether, a total of 536,000 pixels was used to validate the model, thus providing a geographically robust validation dataset.\n",
    "\n",
    "## 2.5 Variable Importance Plots (VIP)\n",
    "\n",
    "Variable Importance Plots (VIP) serve as a method to identify which predictors are important for predicting a specific class. This is achieved by repeatedly predicting the same dataset while shuffling one predictor at a time. The benchmark score obtained after each iteration is then compared to the benchmark score obtained without shuffling any variables. The greater the difference between these two benchmark values, the more important the variable is for the model ([Wei et al., 2015](#ref-WEI2015399)).\n",
    "\n",
    "## 2.6 Impact of the percent cover on the prediction\n",
    "\n",
    "The key aspect of the workflow adopted in the present study is the mapping of specific areas at two different altitudes (12m and 120m), resulting in two distinct resolutions for the same area (8mm and 80mm). The predictions made on the high-resolution flight can be used to estimate the percent cover of each vegetation class for every pixel of the lower resolution flight. Consequently, for each pixel of the high-altitude flights, percent covers of each vegetation class were obtained, and a kernel density plot was generated. This plot provides a visual representation of the behavior of the model in different vegetation percent cover scenarios.\n",
    "\n",
    "# 3. Results\n",
    "\n",
    "## 3.1 Classification\n",
    "\n",
    "The nine drone flights were utilized to produce a total of nine prediction maps, each of them associated with a probability map that indicates the probability of the selected class for every pixel. The low-altitude flight conducted in Gafanha, Portugal, represents the site with the highest complexity (<a href=\"#fig-GafLow\" class=\"quarto-xref\">Figure 4</a>). Among the five vegetation classes on which the model was trained, four were present on this site, with Chlorophyceae and Rhodophyceae mixed with the seagrass meadow. There was also Bacillariophyceae forming biofilms on parts of sediment surface. Although the seagrass bed was solely composed of *Nanozostera noltei*, various colors of this specie could be observed from dark green (corresponding to healthy leaves) to whitish/brown (when leaves were discolored having lost their pigmentation). Regardless of color, Magnoliopsida pixels were accurately predicted by the model."
   ],
   "id": "52d1dd6a-ca19-47f6-bf62-d395f6e2da75"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imglist<-\n",
    "  data.frame(path = list.files(pattern = \".tif\", recursive = T, full.names = T)) %>% \n",
    "  mutate(folder = case_when(str_detect(path, \"Data\") ~ \"Input\",\n",
    "                            TRUE ~ \"Output\"),\n",
    "         alt = case_when(str_detect(path, \"high\") ~ \"high\",\n",
    "                         TRUE ~ \"low\"),\n",
    "         site = case_when(str_detect(path, \"Gafanha\") ~ \"Gafanha\",\n",
    "                          TRUE ~ \"Duer\"),\n",
    "         type = case_when(str_detect(path, \"pred\") ~ \"Pred\",\n",
    "                          str_detect(path, \"NDVI\") ~ \"NDVI\",\n",
    "                          TRUE ~ \"ortho\"),\n",
    "         resampled = case_when(str_detect(path, \"resampled\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         new = case_when(str_detect(path, \"no_Xantho\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         col = case_when(str_detect(path, \"colored\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         correction = case_when(str_detect(path, \"corrected_per_bands\") ~ \"Bands\",\n",
    "                          str_detect(path, \"corrected.tif\") ~ \"lm_all\",\n",
    "                          TRUE ~ \"NA\")) %>% \n",
    "  dplyr::filter(alt != \"NA\",\n",
    "                !str_detect(path,\".aux.xml\"))\n",
    "\n",
    "shp_roi<-\"Data/reflectance_calibration_low_&_high_flights/\" %>%\n",
    "  list.files(pattern = \".shp\", recursive = T, full.names = T) %>%\n",
    "  read_sf() %>%\n",
    "  mutate(ID = case_when(name == \"Microphytobenthos\" ~ 1,\n",
    "                        name == \"ZOS\" ~ 2,\n",
    "                        name == \"BL\" ~ 3,\n",
    "                        name == \"SED\" ~ 4,\n",
    "                        name == \"GRE\" ~ 5,\n",
    "                        name == \"RED\" ~ 6,\n",
    "                        name == \"Xanthophyceae\" ~ 7)) %>% \n",
    "  as(\"Spatial\") %>% \n",
    "  vect()\n",
    "\n",
    "\n",
    "zoom_extent<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"zoom_extent_map_Gafanha_comparison\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "background_GafHigh<-\"Data/shp/Aveiro\" %>% \n",
    "  list.files(\".shp\", full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path, \"background_Gafanha_high\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "###### Preview ###########\n",
    "\n",
    "filelist<-list.files(\"Data/shp/\", pattern = \".shp\", recursive = T,full.names = T)\n",
    "mask_aveiro <- filelist[str_detect(filelist,\"mask_land_intertidal_Aveiro\")] %>% \n",
    "  read_sf()\n",
    "\n",
    "Flight_aveiro<-data.frame(\n",
    "  Name=c(\n",
    "  \"Gafanha\"\n",
    "  ) ,\n",
    "  Description = c(\"12m and 120m Flights\"),\n",
    "  Long=c(-8.742485) ,\n",
    "  Lat=c(40.597599) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(crs(mask_aveiro))  %>% \n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)\n",
    "\n",
    "\n",
    "p1_background<- read_sf(\"Data/shp/Aveiro/Background_p1_aveiro.shp\")\n",
    "\n",
    "p1<-ggplot() +\n",
    "  scale_fill_manual(\"\", labels = c(\"Intertidal area\", \"Land area\"),\n",
    "                    values = c(\"#7DC27D\", \"#BDBDBD\"))+\n",
    "  geom_sf(data = p1_background, fill = \"white\",show.legend = F, linewidth=0.05,alpha=0.93,colour=\"transparent\")+\n",
    "  geom_sf(data = mask_aveiro, mapping = aes(fill = Type),show.legend = F, linewidth=0.05,alpha=0.93,colour=\"grey30\")+\n",
    "\n",
    "      coord_sf(xlim=c(519749,535233) ,\n",
    "          ylim=c(4493310,4508794) )+\n",
    "  ggforce::geom_mark_ellipse(data=Flight_aveiro,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = Name,\n",
    "                     group = Name),\n",
    "                     # description=Description),\n",
    "                 size=0.3,\n",
    "                 fill = \"goldenrod\",\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 show.legend=F,\n",
    "                 label.fontsize = 18,\n",
    "                 label.hjust = 0.5,\n",
    "                 con.size = 2,\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") ,\n",
    "  label.fill = \"grey90\",\n",
    "  label.buffer = unit(5, \"mm\")) +\n",
    "    # theme_void()+\n",
    "  theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  # geom_label(aes(x = 521000, y = 4507794, label = \"C\"), size = 20)+\n",
    "  scale_x_continuous(breaks = seq(-8.75, -8.6, by = 0.1))+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\") ,\n",
    "        legend.position = \"none\",\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20))+\n",
    "  theme_void()\n",
    "######### RGB #########\n",
    " Gafanha_low_ortho<-imglist %>% \n",
    "  filter(site == \"Gafanha\",\n",
    "         alt == \"low\",\n",
    "         type == \"ortho\",\n",
    "         resampled == \"TRUE\",\n",
    "         col == \"TRUE\") %>%\n",
    "  pull(path) %>% \n",
    "  rast() \n",
    " names(Gafanha_low_ortho)<-c(\"B1\", \"B3\", \"B5\",\"alpha\")\n",
    "\n",
    " values(Gafanha_low_ortho)[values(Gafanha_low_ortho$alpha) == 0] = NA\n",
    "  \n",
    " Gafanha_low_ortho_zoomed<-Gafanha_low_ortho %>% \n",
    "   crop(zoom_extent)\n",
    " \n",
    " \n",
    " map_extent_full<- st_as_sf(as.polygons(ext(Gafanha_low_ortho)))\n",
    " sf::st_crs(map_extent_full)<-st_crs(Gafanha_low_ortho)\n",
    " \n",
    "\n",
    "  map_extent_zommed<- st_as_sf(as.polygons(ext(Gafanha_low_ortho_zoomed)))\n",
    " sf::st_crs(map_extent_zommed)<-st_crs(Gafanha_low_ortho_zoomed)\n",
    " \n",
    " \n",
    " rgb_Gafanha_low_zoomed<-ggplot()+\n",
    "   tidyterra::geom_spatraster_rgb(data = Gafanha_low_ortho_zoomed, \n",
    "                                  r = 3,\n",
    "                                  g = 2,\n",
    "                                  b = 1)+\n",
    "   theme_Bede_Map()+\n",
    "   geom_sf(data = zoom_extent, fill = \"transparent\", colour = \"black\",linetype = \"dashed\", linewidth = 1)+\n",
    "   coord_sf(crs = 4326)+\n",
    "     theme_void()\n",
    "\n",
    "\n",
    "\n",
    " scale_labels<-data.frame(x = c(-8.74368, -8.743565,-8.74342),\n",
    "                          y = c(rep(40.59792,3)),\n",
    "                          text = c(\"0\",\"10\",\"20 m\"))\n",
    " \n",
    " \n",
    " \n",
    "  rgb_Gafanha_low<- ggplot()+\n",
    "  scale_fill_manual(\"\", \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  geom_sf(data = background_GafHigh,\n",
    "          fill = \"white\",\n",
    "          show.legend = F,\n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"white\")+\n",
    "  geom_sf(data = background_GafHigh,\n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"grey30\")+\n",
    "   tidyterra::geom_spatraster_rgb(data = Gafanha_low_ortho,\n",
    "         r = 1,\n",
    "         g = 2,\n",
    "         b = 3,\n",
    "         maxcell = 4010533\n",
    "         )+\n",
    "    ggspatial::annotation_scale(location = \"tl\",\n",
    "                                text_cex=2)+\n",
    "      annotation_custom(ggplotGrob(p1),\n",
    "                        xmin = -8.7430,\n",
    "                        xmax = -8.742702,\n",
    "                        ymin = 40.5970515, \n",
    "                        ymax = 40.5974) +\n",
    "   theme_Bede_Map()+\n",
    "   geom_sf(data = zoom_extent, \n",
    "           fill = \"transparent\",\n",
    "           colour = \"black\",\n",
    "           linetype = \"dashed\", \n",
    "           linewidth = 1)+\n",
    "   coord_sf(crs = 4326)+\n",
    "  annotation_custom(ggplotGrob(rgb_Gafanha_low_zoomed),\n",
    "                    xmin = -8.7430,\n",
    "                    xmax=-8.7427,\n",
    "                    ymin = 40.5977) +\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.7432439,-8.743,-8.7432439,-8.743),\n",
    "                              y=c(40.5976275,40.5979,40.5975373,40.59771),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", linewidth = 1)+\n",
    "  theme(axis.title = element_blank(),\n",
    "        axis.text.y = element_text(size = 20),\n",
    "        axis.text.x = element_text(size = 20, angle = 20, hjust = 1))+\n",
    "    scale_x_continuous(limits = c(-8.7437, -8.74275), breaks = -8.743)+\n",
    "    scale_y_continuous(limits = c(40.59715, 40.5979), breaks = 40.5975)\n",
    "\n",
    " \n",
    " \n",
    "  \n",
    "  \n",
    "\n",
    "ggsave(\"Figures/Gaf_Low_RGB.png\",rgb_Gafanha_low , width = 1920*4, height = 1129*4, units = \"px\")\n",
    "\n",
    "zoom_extent<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"zoom_extent_map_Gafanha_comparison\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "####### Zoomed map ########\n",
    "\n",
    "\n",
    "Gafanha_low_pred_zoomed<-imglist %>% \n",
    "filter(site == \"Gafanha\",\n",
    "       alt == \"low\",\n",
    "       type == \"Pred\",\n",
    "       resampled == \"FALSE\",\n",
    "       folder == \"Input\",\n",
    "       new == \"TRUE\") %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Gafanha_low_pred_zoomed)<-\"layer\"\n",
    "\n",
    "Gafanha_low_pred_zoomed<-Gafanha_low_pred_zoomed %>% \n",
    "  crop(zoom_extent)\n",
    "Gafanha_low_pred_zoomed<-as.factor(Gafanha_low_pred_zoomed)\n",
    "\n",
    "Gafanha_low_pred_map_zoomed<-ggplot()+\n",
    "  tidyterra::geom_spatraster(data = Gafanha_low_pred_zoomed, mapping = aes(fill = layer))+\n",
    "  geom_sf(data = zoom_extent, fill = \"transparent\", colour = \"black\",linetype = \"dashed\", linewidth = 1)+\n",
    "  coord_sf(crs = 4326)+\n",
    "  scale_fill_manual(values=c(\"#70543e\", \"#b3ff1a\",\"#70543e\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", NA, \"#42c9bc\"), na.value = NA)+\n",
    "  theme_void()+\n",
    "  # coord_equal()+\n",
    "  theme(axis.text.x = element_blank(),\n",
    "      axis.text.y = element_blank(),\n",
    "      axis.title = element_blank())+\n",
    "  theme(legend.position = \"none\")\n",
    " # Gafanha_low_pred_map_zoomed\n",
    "\n",
    "####  Overall map\n",
    "Gafanha_low_pred<-imglist %>% \n",
    "filter(site == \"Gafanha\",\n",
    "       alt == \"low\",\n",
    "       type == \"Pred\",\n",
    "       # resampled == \"TRUE\",\n",
    "       new == \"TRUE\") %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Gafanha_low_pred)<-\"layer\"\n",
    "\n",
    "\n",
    "values(Gafanha_low_pred)[values(Gafanha_low_pred) == 0] = NA\n",
    "\n",
    "Gafanha_low_pred<-Gafanha_low_pred %>% \n",
    "  as.factor()\n",
    "\n",
    "####### Pred ######\n",
    "\n",
    "\n",
    "Gafanha_low_pred_map <-\n",
    "  ggplot()+\n",
    "  scale_fill_manual(\"\", \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  geom_sf(data = background_GafHigh,\n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"grey30\")+\n",
    "  # geom_sf(data = background_GafHigh,\n",
    "  #         fill = \"white\",\n",
    "  #         show.legend = F, \n",
    "  #         linewidth=0.05,\n",
    "  #         alpha=1,\n",
    "  #         colour=\"white\")+\n",
    "   ggnewscale::new_scale(\"fill\")+\n",
    "    tidyterra::geom_spatraster(data = Gafanha_low_pred, \n",
    "                               mapping = aes(fill = layer),\n",
    "                               maxcell = 500832*8)+\n",
    "  geom_sf(data = zoom_extent, \n",
    "          fill = \"transparent\", \n",
    "          colour = \"black\",\n",
    "          linetype = \"dashed\", \n",
    "          linewidth = 1)+\n",
    "  coord_sf(crs = 4326)+\n",
    "  annotation_custom(ggplotGrob(Gafanha_low_pred_map_zoomed), \n",
    "                    xmin = -8.7430,\n",
    "                    xmax=-8.7427, \n",
    "                    ymin = 40.5977) +\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.7432439,-8.743,-8.7432439,-8.743),\n",
    "                              y=c(40.5976275,40.5979,40.5975373,40.59771),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", linewidth = 1)+\n",
    "  scale_fill_manual(breaks = c(1,2,5,6,7,8,10),\n",
    "                    values=c(\"#70543e\", \"#b3ff1a\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", \"#42c9bc\",NA,NA,NA,NA),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Chlorophyceae\",\n",
    "                               \"Bacillariophyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\",\n",
    "                               \"Water\"))+\n",
    "  theme_Bede_Map()+\n",
    "  # coord_equal()+\n",
    "   theme(axis.text.x = element_text(size = 20, angle = 20,hjust = 1),\n",
    "         # axis.text.y = element_text(size = 20),\n",
    "         axis.text.y = element_blank(),\n",
    "         axis.title = element_blank(),\n",
    "         legend.background = element_blank(),\n",
    "         legend.position = c(0.98,0.3),\n",
    "         legend.text = element_text(size = 20),\n",
    "         legend.key.size = unit(1,\"cm\"))+\n",
    "    scale_x_continuous(limits = c(-8.7437, -8.74275), breaks = -8.743)+\n",
    "    scale_y_continuous(limits = c(40.59715, 40.5979), breaks = 40.5975)\n",
    "\n",
    "\n",
    "ggsave(\"Figures/Gaf_Low_Pred.png\",Gafanha_low_pred_map , width = 1920*4, height = 1129*4, units = \"px\")\n",
    "\n",
    "################# PLOT AND SAVING ###########\n",
    "\n",
    "plot_Gaf_Low<-rgb_Gafanha_low+\n",
    "  Gafanha_low_pred_map\n",
    " \n",
    "\n",
    "\n",
    "\n",
    "ggsave(\"Figures/Maps Pred/FigX-Gaf_Low_Pred.png\",plot_Gaf_Low , width = 1920*4, height = 1009*4, units = \"px\")"
   ],
   "id": "e132a85f-e060-49de-8ac1-bd7ca3dc1ad9"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Maps Pred/FigX-Gaf_Low_Pred.png\")"
   ],
   "id": "cell-fig-GafLow"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imglist<-\n",
    "  data.frame(\"Data\",\n",
    "             path = list.files(pattern = \".tif\", recursive = T, full.names = T)) %>% \n",
    "  mutate(folder = case_when(str_detect(path, \"Data\") ~ \"Input\",\n",
    "                            TRUE ~ \"Output\"),\n",
    "         alt = case_when(str_detect(path, \"high\") ~ \"high\",\n",
    "                         str_detect(path, \"High\") ~ \"high\",\n",
    "                         TRUE ~ \"low\"),\n",
    "         site = case_when(str_detect(path, \"Gafanha\") ~ \"Gafanha\",\n",
    "                          str_detect(path,\"AveiroHigh\")~\"Gafanha\",\n",
    "                          TRUE ~ \"Duer\"),\n",
    "         type = case_when(str_detect(path, \"pred\") ~ \"Pred\",\n",
    "                          str_detect(path, \"NDVI\") ~ \"NDVI\",\n",
    "                          TRUE ~ \"ortho\"),\n",
    "         resampled = case_when(str_detect(path, \"resampled\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         new = case_when(str_detect(path, \"no_Xantho\") ~ \"TRUE\",\n",
    "                         str_detect(path, \"NewPreds\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         col = case_when(str_detect(path, \"colored\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         correction = case_when(str_detect(path, \"corrected_per_bands\") ~ \"Bands\",\n",
    "                          str_detect(path, \"corrected.tif\") ~ \"lm_all\",\n",
    "                          TRUE ~ \"NA\")) %>% \n",
    "  dplyr::filter(alt != \"NA\",\n",
    "                !str_detect(path,\".aux.xml\"))\n",
    "\n",
    "########## ZOOM #############\n",
    "\n",
    "zoom_extent<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"zoom\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "mask_low<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"mask_Gafanha_low\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "Gafanha_low_ortho_zoomed<-imglist %>% \n",
    "  filter(site == \"Gafanha\",\n",
    "         alt == \"low\",\n",
    "         type == \"ortho\",\n",
    "         resampled == \"TRUE\",\n",
    "         col == \"TRUE\") %>%\n",
    "  pull(path) %>% \n",
    "  rast() %>% \n",
    "  crop(zoom_extent)\n",
    "\n",
    "names(Gafanha_low_ortho_zoomed)<-c(\"B1\", \"B3\", \"B5\",\"alpha\")\n",
    "\n",
    " Gafanha_low_pred_zoomed<-imglist %>% \n",
    "  filter(site == \"Gafanha\",\n",
    "         alt == \"low\",\n",
    "         type == \"Pred\",\n",
    "         # resampled == \"TRUE\",\n",
    "         new == \"TRUE\") %>% \n",
    "  pull(path) %>% \n",
    "  rast() %>% \n",
    "  crop(zoom_extent) %>% \n",
    "  mask(mask_low, inverse = T) %>% \n",
    "   as.factor()\n",
    "   \n",
    "names(Gafanha_low_pred_zoomed)<-\"layer\"\n",
    "\n",
    "zoomed_low_map<-ggplot()+\n",
    "  tidyterra::geom_spatraster_rgb(data = Gafanha_low_ortho_zoomed,\n",
    "                                 r = 1,\n",
    "                                 g = 2,\n",
    "                                 b = 3)+\n",
    "  tidyterra::geom_spatraster(data = Gafanha_low_pred_zoomed, mapping = aes(fill = layer), show.legend = F)+\n",
    "  \n",
    "  scale_fill_manual(breaks = c(1,2,5,6,7,8,10),\n",
    "                    values=c(\"#70543e\", \"#b3ff1a\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", \"#42c9bc\",NA,NA), \n",
    "                    na.value = NA,\n",
    "                    name = \"\", \n",
    "                    labels = c(\"Sediment\", \n",
    "                               \"Chlorophyceae\", \n",
    "                               \"Bacillariophyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\",\n",
    "                               \"Water\"))+\n",
    "  geom_sf(data = zoom_extent, fill = \"transparent\", colour = \"black\",linetype = \"dashed\", linewidth = 2)+\n",
    "  theme(legend.position = \"none\")+\n",
    "  theme_void()\n",
    "\n",
    "\n",
    "\n",
    "#############\n",
    "\n",
    " Gafanha_low_ortho<-imglist %>% \n",
    "  filter(site == \"Gafanha\",\n",
    "         alt == \"low\",\n",
    "         type == \"ortho\",\n",
    "         resampled == \"TRUE\",\n",
    "         col == \"TRUE\") %>%\n",
    "  pull(path) %>% \n",
    "  rast() \n",
    " names(Gafanha_low_ortho)<-c(\"B1\", \"B3\", \"B5\",\"alpha\")\n",
    "values(Gafanha_low_ortho)[values(Gafanha_low_ortho$alpha) == 0] = NA\n",
    "\n",
    " \n",
    " Gafanha_low_pred<-imglist %>% \n",
    "  filter(site == \"Gafanha\",\n",
    "         alt == \"low\",\n",
    "         type == \"Pred\",\n",
    "         # resampled == \"TRUE\",\n",
    "         new == \"TRUE\") %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Gafanha_low_pred)<-\"layer\"\n",
    "\n",
    "Gafanha_low_pred_masked<-Gafanha_low_pred %>% \n",
    "  # crop(mask_low) %>% \n",
    "  mask(mask_low, inverse = T)\n",
    "\n",
    "values(Gafanha_low_pred_masked)[values(Gafanha_low_pred_masked) == 0] = NA\n",
    "\n",
    "Gafanha_low_pred_masked<-Gafanha_low_pred_masked %>% \n",
    "  as.factor()\n",
    "\n",
    "rgb_Gafanha_low <- ggplot()+\n",
    "  tidyterra::geom_spatraster_rgb(data = Gafanha_low_ortho,\n",
    "                                 r = 1,\n",
    "                                 g = 2,\n",
    "                                 b = 3)+\n",
    "  tidyterra::geom_spatraster(data = Gafanha_low_pred_masked, mapping = aes(fill = layer))+\n",
    "  scale_fill_manual(breaks = c(1,2,5,6,7,8,10),\n",
    "                    values=c(\"#70543e\", \"#b3ff1a\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", \"#42c9bc\",NA,NA), \n",
    "                    na.value = NA,\n",
    "                    name = \"\", \n",
    "                    labels = c(\"Sediment\", \n",
    "                               \"Chlorophyceae\", \n",
    "                               \"Bacillariophyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\",\n",
    "                               \"Water\"))+\n",
    "  ggspatial::annotation_scale(location = \"bl\",text_cex=2)+\n",
    "  geom_sf(data = zoom_extent, fill = \"transparent\", colour = \"black\",linetype = \"dashed\", linewidth = 1)+\n",
    "\n",
    "  geom_segment(aes(x = -8.7434247325, y=40.5977530006, xend = -8.7430558162, yend = 40.5972152706), linetype =\"dashed\" ,linewidth=2)+\n",
    "  annotation_custom(ggplotGrob(zoomed_low_map), xmin = -8.743725, xmax = -8.7434057328, ymin = 40.5977200450, ymax = 40.59798) +\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.74336194,-8.74371,-8.743243794,-8.7434057328), y=c(40.59762751,40.5977400850,40.59762725,40.5977200450),grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", linewidth = 1)+\n",
    "  theme_Bede_Map()+\n",
    "  coord_sf(crs = 4326)+\n",
    "  theme(axis.title = element_blank(),\n",
    "        axis.text.y = element_text(size = 20),\n",
    "        axis.text.x = element_text(size = 20, angle = 20,hjust = 1),\n",
    "        legend.position=c(0.3, 0.3))+\n",
    "  # ggtitle(\"a \")+\n",
    "  scale_x_continuous(breaks=seq(-8.7437, -8.7430, 0.0002))\n",
    "rgb_Gafanha_low\n",
    "\n",
    "ggsave(\"Figures/Gaf_Low_RGB_Pred_V2.png\",rgb_Gafanha_low , width = 890*4, height = 1120*4, units = \"px\")"
   ],
   "id": "3efbbcef-1f19-488a-bfea-40662ee93f60"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The high-altitude flight over Gafanha covered a total area of approwimately 1 km² (<a href=\"#fig-GafHigh\" class=\"quarto-xref\">Figure 5</a>). A channel contouring a small island was masked in the prediction map. Most of the intertidal area was classified as Magnoliopsida by the model, including seagrass patches with discolored leaves. Only a few pixels were classified as Chlorophyceae at this spatial scale. Furthermore, the area that was classified as Bacillariophyceae in the low-altitude flight remained mostly classified as such in the high-altitude flight, though some pixels were classified as Magnoliopsida. Patches of Rhodophyceae were correctly classified. In the northern part of the site and near the land eadges, patches of the schorre angiosperm *Spartina sp.* were misclassified, either as Magnoliopsida or as Phaeophyceae."
   ],
   "id": "9f0eac56-acc7-49d5-a383-a8d7ccea3db7"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imglist<-\n",
    "  data.frame(path = list.files(pattern = \".tif\", recursive = T, full.names = T)) %>% \n",
    "  mutate(folder = case_when(str_detect(path, \"Data\") ~ \"Input\",\n",
    "                            TRUE ~ \"Output\"),\n",
    "         alt = case_when(str_detect(path, \"high\") ~ \"high\",\n",
    "                         str_detect(path, \"High\") ~ \"high\",\n",
    "                         TRUE ~ \"low\"),\n",
    "         site = case_when(str_detect(path, \"Gafanha\") ~ \"Gafanha\",\n",
    "                          str_detect(path, \"AveiroHigh\") ~ \"Gafanha\",\n",
    "                          TRUE ~ \"Duer\"),\n",
    "         type = case_when(str_detect(path, \"pred\") ~ \"Pred\",\n",
    "                          str_detect(path, \"Preds\") ~ \"Pred\",\n",
    "                          str_detect(path, \"NDVI\") ~ \"NDVI\",\n",
    "                          TRUE ~ \"ortho\"),\n",
    "         resampled = case_when(str_detect(path, \"resampled\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         new = case_when(str_detect(path, \"no_Xantho\") ~ \"TRUE\",\n",
    "                         str_detect(path, \"NewPreds\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         col = case_when(str_detect(path, \"colored\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         correction = case_when(str_detect(path, \"corrected_per_bands\") ~ \"Bands\",\n",
    "                          str_detect(path, \"corrected.tif\") ~ \"lm_all\",\n",
    "                          TRUE ~ \"NA\")) %>% \n",
    "  dplyr::filter(alt != \"NA\",\n",
    "                !str_detect(path,\".aux.xml\"))\n",
    "\n",
    "#### Preview ####\n",
    "\n",
    "filelist<-list.files(\"Data/shp/\", pattern = \".shp\", recursive = T,full.names = T)\n",
    "mask_aveiro <- filelist[str_detect(filelist,\"mask_land_intertidal_Aveiro\")] %>% \n",
    "  read_sf()\n",
    "\n",
    "Flight_aveiro<-data.frame(\n",
    "  Name=c(\n",
    "  \"Gafanha\"\n",
    "  ) ,\n",
    "  Description = c(\"12m and 120m Flights\"),\n",
    "  Long=c(-8.742485) ,\n",
    "  Lat=c(40.597599) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(crs(mask_aveiro))  %>% \n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)\n",
    "\n",
    "p1_background<- read_sf(\"Data/shp/Aveiro/Background_p1_aveiro.shp\")\n",
    "\n",
    "p1<-ggplot() +\n",
    "  scale_fill_manual(\"\", labels = c(\"Intertidal area\", \"Land area\"),\n",
    "                    values = c(\"#7DC27D\", \"#BDBDBD\"))+\n",
    "  geom_sf(data = p1_background, fill = \"white\",show.legend = F, linewidth=0.05,alpha=0.93,colour=\"transparent\")+\n",
    "  geom_sf(data = mask_aveiro, mapping = aes(fill = Type),show.legend = F, linewidth=0.05,alpha=0.93,colour=\"grey30\")+\n",
    "      coord_sf(xlim=c(519749,535233) ,\n",
    "          ylim=c(4493310,4508794) )+\n",
    "  ggforce::geom_mark_ellipse(data=Flight_aveiro,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = Name,\n",
    "                     group = Name),\n",
    "                     # description=Description),\n",
    "                 size=0.3,\n",
    "                 fill = \"goldenrod\",\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 show.legend=F,\n",
    "                 label.fontsize = 20,\n",
    "                 label.hjust = 0.5,\n",
    "                 con.size = 2,\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") ,\n",
    "  label.fill = \"grey90\",\n",
    "  label.buffer = unit(5, \"mm\")) +\n",
    "    # theme_void()+\n",
    "  theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  # geom_label(aes(x = 521000, y = 4507794, label = \"C\"), size = 20)+\n",
    "  scale_x_continuous(breaks = seq(-8.75, -8.6, by = 0.1))+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\") ,\n",
    "        legend.position = \"none\",\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20))+\n",
    "  theme_void()\n",
    "\n",
    "#####\n",
    "\n",
    "land_mask_GafHigh<-\"Data/shp/Aveiro\" %>% \n",
    "  list.files(\".shp\", full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path, \"mask_Gafanha_high1\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "background_GafHigh<-\"Data/shp/Aveiro\" %>% \n",
    "  list.files(\".shp\", full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path, \"background_Gafanha_high\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "GafHigh_ortho<-imglist %>% \n",
    "  dplyr::filter(site == \"Gafanha\",\n",
    "                type == \"ortho\",\n",
    "                alt==\"high\",\n",
    "                col == \"TRUE\") %>%\n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(GafHigh_ortho)<-c(\"B1\",\"B3\", \"B5\", \"alpha\")\n",
    "\n",
    "# values(GafHigh_ortho)[values(GafHigh_ortho$B1) == 0] = NA\n",
    "\n",
    "GafHigh_ortho_masked<-GafHigh_ortho %>% \n",
    "  mask(land_mask_GafHigh, inverse = F) %>% \n",
    "  crop(land_mask_GafHigh)\n",
    "\n",
    "GafHigh_ortho_masked<-na.omit(GafHigh_ortho_masked, field=NA, geom=FALSE)\n",
    "\n",
    "\n",
    "###### RGB ########\n",
    "zoom_gaf_high_sf<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"zoom_extent_map_Gafanha_High_comparison\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "Gaflow_extent<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"GafLow_extent.shp\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "rgb_Gafanha_High_zoomed<-GafHigh_ortho_masked %>% \n",
    "  crop(zoom_gaf_high_sf)\n",
    "\n",
    "rgb_Gafanha_High_zoomed<-\n",
    "  ggplot()+\n",
    "  tidyterra::geom_spatraster_rgb(data = rgb_Gafanha_High_zoomed,\n",
    "         r = 1,\n",
    "         g = 2,\n",
    "         b = 3)+\n",
    "     geom_sf(data = zoom_gaf_high_sf, fill = \"transparent\", colour = \"black\",linetype = \"dashed\", linewidth = 1)+\n",
    "  theme_void()\n",
    "\n",
    "  \n",
    "  \n",
    "\n",
    "rgb_Gafanha_High <- ggplot()+\n",
    "  scale_fill_manual(\"\",  \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  geom_sf(data = background_GafHigh,\n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.93,\n",
    "          colour=\"grey30\")+\n",
    "  geom_sf(data = background_GafHigh,\n",
    "          fill = \"white\",\n",
    "          show.legend = F,\n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"white\")+\n",
    "   tidyterra::geom_spatraster_rgb(data = GafHigh_ortho_masked,\n",
    "         r = 1,\n",
    "         g = 2,\n",
    "         b = 3,\n",
    "         maxcell = 5007120)+\n",
    "    ggspatial::annotation_scale(location = \"bl\",\n",
    "                                text_cex=2)+\n",
    "    annotation_custom(ggplotGrob(p1), \n",
    "                      xmin = -8.744, \n",
    "                      xmax = -8.7412,\n",
    "                      ymin = 40.5923,\n",
    "                      ymax = 40.595) +\n",
    "    geom_sf(data = zoom_gaf_high_sf,\n",
    "            fill = \"transparent\", \n",
    "            colour = \"black\",\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 1)+\n",
    "    geom_sf(data = Gaflow_extent,\n",
    "            fill = \"transparent\", \n",
    "            colour = \"red\",\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 2)+\n",
    "    geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.74363778,-8.74757,-8.74351982,-8.7453), \n",
    "                              y=c(40.5983781677,40.5995,40.598377893,40.5995),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 1)+\n",
    "  annotation_custom(ggplotGrob(rgb_Gafanha_High_zoomed),\n",
    "                    xmin = -8.7479, \n",
    "                    xmax = -8.7452, \n",
    "                    ymin = 40.5979,\n",
    "                    ymax = 40.603) +\n",
    "   theme_Bede_Map()+\n",
    "   coord_sf(crs = 4326)+\n",
    "   # xlim(c(-8.748, -8.741))+\n",
    "  # ylim(c(40.593,40.601))+\n",
    "  theme(axis.title = element_blank(),\n",
    "         axis.text.x = element_text(size = 20, angle = 20, hjust = 1),\n",
    "         axis.text.y = element_text(size = 20))+\n",
    "  scale_x_continuous(limits = c(-8.747576, -8.741465),breaks =  -8.745)+\n",
    "  scale_y_continuous(limits = c(40.59297,40.60109),breaks =  40.595)\n",
    "# rgb_Gafanha_High\n",
    "\n",
    "ggsave(\"Figures/Gaf_High_RGB.png\",rgb_Gafanha_High , width = 712*4, height = 1120*4, units = \"px\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "######### Pred #######\n",
    "Gafanha_high_pred<-imglist %>% \n",
    "filter(site == \"Gafanha\",\n",
    "       alt == \"high\",\n",
    "       type == \"Pred\",\n",
    "       new == \"TRUE\",\n",
    "       !str_detect(path, \"Probs\")) %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Gafanha_high_pred)<-\"layer\"\n",
    "\n",
    "GafHigh_pred_masked<-Gafanha_high_pred %>% \n",
    "  mask(land_mask_GafHigh, inverse = F) %>% \n",
    "  crop(land_mask_GafHigh)\n",
    "\n",
    "GafHigh_pred_masked<-na.omit(GafHigh_pred_masked, field=NA, geom=FALSE) \n",
    "\n",
    "values(GafHigh_pred_masked)[values(GafHigh_pred_masked) == 0] = NA\n",
    "values(GafHigh_pred_masked)[values(GafHigh_pred_masked) == 3] = 1\n",
    "\n",
    "GafHigh_pred_masked<-GafHigh_pred_masked %>% \n",
    "  as.factor()\n",
    "\n",
    "Gafanha_High_zoomed_pred<-GafHigh_pred_masked %>% \n",
    "  crop(zoom_gaf_high_sf)\n",
    "\n",
    "Gafanha_High_zoomed_pred<-\n",
    "  ggplot()+\n",
    "   tidyterra::geom_spatraster(data = Gafanha_High_zoomed_pred, mapping = aes(fill = layer),\n",
    "                            show.legend = F,\n",
    "                             maxcell = 5007120)+\n",
    "  scale_fill_manual(breaks = c(1,2,5,6,7,8,10),\n",
    "                    values=c(\"#70543e\", \"#b3ff1a\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", \"#42c9bc\",NA,NA),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Chlorophyceae\",\n",
    "                               \"Bacillariophyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\",\n",
    "                               \"Water\"))+\n",
    "     geom_sf(data = zoom_gaf_high_sf, fill = \"transparent\", colour = \"black\",linetype = \"dashed\", linewidth = 1)+\n",
    "  theme_void()\n",
    "\n",
    "Gafanha_high_pred_map<-ggplot()+\n",
    "  geom_sf(data = background_GafHigh, \n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F,\n",
    "          linewidth=0.05,\n",
    "          alpha=0.93,\n",
    "          colour=\"grey30\")+\n",
    "  geom_sf(data = background_GafHigh,\n",
    "          fill = \"white\",\n",
    "          show.legend = F,\n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"white\")+\n",
    "    scale_fill_manual(\"\", \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  ggnewscale::new_scale(\"fill\")+\n",
    "  tidyterra::geom_spatraster(data = GafHigh_pred_masked, \n",
    "                             # show.legend = F,\n",
    "                             mapping = aes(fill = layer),\n",
    "                             maxcell = 5010420)+\n",
    "  scale_fill_manual(breaks = c(1,2,5,6,7,8,10),\n",
    "                    values=c(\"#70543e\", \"#b3ff1a\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", \"#42c9bc\",NA,NA),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Chlorophyceae\",\n",
    "                               \"Bacillariophyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\",\n",
    "                               \"Water\"))+\n",
    "       geom_sf(data = zoom_gaf_high_sf, \n",
    "               fill = \"transparent\", \n",
    "               colour = \"black\",\n",
    "               linetype = \"dashed\", \n",
    "               linewidth = 1)+\n",
    "    geom_sf(data = Gaflow_extent,\n",
    "            fill = \"transparent\", \n",
    "            colour = \"red\",\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 2)+\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.74363778,-8.74757,-8.74351982,-8.7453), \n",
    "                              y=c(40.5983781677,40.5995,40.598377893,40.5995),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", linewidth = 1)+\n",
    "  annotation_custom(ggplotGrob(Gafanha_High_zoomed_pred), \n",
    "                    xmin = -8.7479, \n",
    "                    xmax = -8.7452, \n",
    "                    ymin = 40.5979,\n",
    "                    ymax = 40.603) +\n",
    "  coord_sf(crs = 4326)+\n",
    "  theme_Bede_Map()+\n",
    "   theme(axis.text.x = element_text(size = 20, angle = 20,hjust = 1),\n",
    "         axis.title = element_blank(),\n",
    "        # axis.text.y = element_text(size = 20),\n",
    "        axis.text.y = element_blank(),\n",
    "         legend.background = element_blank(),\n",
    "         legend.position = c(0.96,0.23),\n",
    "         legend.text = element_text(size = 23),\n",
    "         legend.key.size = unit(1.2,\"cm\"))+\n",
    "  scale_x_continuous(limits = c(-8.747576, -8.741465),breaks =  -8.745)+\n",
    "  scale_y_continuous(limits = c(40.59297,40.60109),breaks =  40.595)\n",
    "# Gafanha_high_pred_map\n",
    "\n",
    "ggsave(\"Figures/Gaf_High_Pred.png\",Gafanha_high_pred_map , width = 712*4, height = 1120*4, units = \"px\")\n",
    "\n",
    "################# PLOT AND SAVING ###########\n",
    "plot_GafHigh<-rgb_Gafanha_High+\n",
    "  Gafanha_high_pred_map\n",
    "\n",
    "ggsave(\"Figures/Maps Pred/FigX-Gaf_High_Pred.png\",plot_GafHigh , width = 2105*4, height = 1363*4, units = \"px\")\n",
    "\n",
    "\n",
    "#########"
   ],
   "id": "7d791760-4ac4-4873-b6fd-c2340d6505aa"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Maps Pred/FigX-Gaf_High_Pred.png\")"
   ],
   "id": "cell-fig-GafHigh"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Amoung the high altitude flights, the one acquired over the inner lagoon of the Ria de Aveiro covered the largest area with approximately 1.5 km² (<a href=\"#fig-Boat\" class=\"quarto-xref\">Figure 6</a>). This site was dominated by seagrass and the red macroalgae *Gracilaria vermicullophyla*. The classification provided consistent results, with a patchy seagrass meadow mixed with red macroalgae on the eastern part of the site. As shown in the zoom (<a href=\"#fig-Boat\" class=\"quarto-xref\">Figure 6</a>), the edges of the meadow were colonised by green macroalgae"
   ],
   "id": "1e9b7d6f-efca-4a40-9dc4-5bb63365a1df"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imglist<-\n",
    "  data.frame(path = list.files(pattern = \".tif\", recursive = T, full.names = T)) %>% \n",
    "  mutate(folder = case_when(str_detect(path, \"Data\") ~ \"Input\",\n",
    "                            TRUE ~ \"Output\"),\n",
    "         alt = case_when(str_detect(path, \"high\") ~ \"high\",\n",
    "                         str_detect(path, \"High\") ~ \"high\",\n",
    "                         TRUE ~ \"low\"),\n",
    "         site = case_when(str_detect(path, \"Gafanha\") ~ \"Gafanha\",\n",
    "                          str_detect(path, \"AveiroHigh\") ~ \"Gafanha\",\n",
    "                          TRUE ~ \"Duer\"),\n",
    "         type = case_when(str_detect(path, \"pred\") ~ \"Pred\",\n",
    "                          str_detect(path, \"Preds\") ~ \"Pred\",\n",
    "                          str_detect(path, \"NDVI\") ~ \"NDVI\",\n",
    "                          TRUE ~ \"ortho\"),\n",
    "         resampled = case_when(str_detect(path, \"resampled\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         new = case_when(str_detect(path, \"no_Xantho\") ~ \"TRUE\",\n",
    "                         str_detect(path, \"NewPreds\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         col = case_when(str_detect(path, \"colored\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         correction = case_when(str_detect(path, \"corrected_per_bands\") ~ \"Bands\",\n",
    "                          str_detect(path, \"corrected.tif\") ~ \"lm_all\",\n",
    "                          TRUE ~ \"NA\")) %>% \n",
    "  dplyr::filter(alt != \"NA\",\n",
    "                !str_detect(path,\".aux.xml\"))\n",
    "########### PREVIEW #############\n",
    "filelist<-list.files(\"Data/shp/\", pattern = \".shp\", recursive = T,full.names = T)\n",
    "mask_aveiro <- filelist[str_detect(filelist,\"mask_land_intertidal_Aveiro\")] %>% \n",
    "  read_sf()\n",
    "\n",
    "Flight_aveiro<-data.frame(\n",
    "  Name=c(\n",
    "  \"Marinha Lanzarote\"\n",
    "  ) ,\n",
    "  Description = c(\n",
    "                  \"120m Flight\"),\n",
    "  Long=c( \n",
    "         -8.690677) ,\n",
    "  Lat=c(\n",
    "        40.665120) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(crs(mask_aveiro))  %>% \n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)\n",
    "\n",
    "\n",
    "p1_background<- read_sf(\"Data/shp/Aveiro/Background_p1_aveiro.shp\")\n",
    "\n",
    "p1<-ggplot() +\n",
    "  scale_fill_manual(\"\", labels = c(\"Intertidal area\", \"Land area\"),\n",
    "                    values = c(\"#7DC27D\", \"#BDBDBD\"))+\n",
    "  geom_sf(data = p1_background, fill = \"white\",show.legend = F, linewidth=0.05,alpha=0.93,colour=\"transparent\")+\n",
    "  geom_sf(data = mask_aveiro, mapping = aes(fill = Type),linewidth=0.05,alpha=0.93,show.legend = F,colour=\"grey30\")+\n",
    "      coord_sf(xlim=c(519749,535233) ,\n",
    "          ylim=c(4493310,4508794) )+\n",
    "  ggforce::geom_mark_ellipse(data=Flight_aveiro,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = Name,\n",
    "                     group = Name),\n",
    "                     # description=Description),\n",
    "                 size=0.3,\n",
    "                 fill = \"goldenrod\",\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 show.legend=F,\n",
    "                 label.fontsize = 10,\n",
    "                 label.hjust = 0.5,\n",
    "                 con.size = 2,\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") ,\n",
    "  label.fill = \"grey90\",\n",
    "  label.buffer = unit(5, \"mm\")) +\n",
    "    # theme_void()+\n",
    "  theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  # geom_label(aes(x = 521000, y = 4507794, label = \"C\"), size = 20)+\n",
    "  scale_x_continuous(breaks = seq(-8.75, -8.6, by = 0.1))+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\") ,\n",
    "        legend.position = \"none\",\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20))+\n",
    "  theme_void()\n",
    "##### ZOOM and masking #####\n",
    "\n",
    "Boat_rgb<-imglist %>% \n",
    "  filter(str_detect(path,\"Boat_high_RGB_merged.tif\")) %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Boat_rgb)<-c(\"B1\",\"B3\",\"B5\")\n",
    "values(Boat_rgb)[values(Boat_rgb) == 0] = NA\n",
    "\n",
    "mask_boat<-\"Data/shp/Aveiro\" %>% \n",
    "  list.files(\".shp\", full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path, \"boat\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "background_Boat<-\"Data/shp/Aveiro\" %>% \n",
    "  list.files(\".shp\", full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path, \"background_Boat\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "zoom_boat_high_sf<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"zoom_extent_map_Boat_comparison\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "zoom_boat_high_sf_2<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"zoom_extent_map_Boat_2_comparison\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "rgb_Boat_zoomed<-Boat_rgb %>% \n",
    "  crop(zoom_boat_high_sf)\n",
    "\n",
    "rgb_Boat_zoomed_2<-Boat_rgb %>% \n",
    "  crop(zoom_boat_high_sf_2)\n",
    "\n",
    "Boat_rgb_mask<-Boat_rgb %>% \n",
    "  mask(mask_boat, inverse = T)\n",
    "\n",
    " \n",
    "rgb_Boat_zoomed_map<- ggplot()+\n",
    "  geom_spatraster_rgb(data = rgb_Boat_zoomed,\n",
    "                      r = 3,\n",
    "                      g = 2,\n",
    "                      b = 1,\n",
    "                      max_col_value = 4000)+\n",
    "  geom_sf(data = zoom_boat_high_sf, fill = \"transparent\", linetype = \"dashed\",colour = \"black\", linewidth = 1)+\n",
    "\n",
    "  theme_void()\n",
    "\n",
    "rgb_Boat_zoomed_map_2\n",
    "ggplot()+\n",
    "  geom_spatraster_rgb(data = rgb_Boat_zoomed_2,\n",
    "                      r = 3,\n",
    "                      g = 2,\n",
    "                      b = 1,\n",
    "                      interpolate = F,\n",
    "                      max_col_value = 4000)+\n",
    "  geom_sf(data = zoom_boat_high_sf_2, fill = \"transparent\", linetype = \"dashed\",colour = \"black\", linewidth = 1)+\n",
    "\n",
    "  theme_void()\n",
    "\n",
    "####### RGB ############\n",
    "\n",
    "RGB_Boat_map<-\n",
    "ggplot()+\n",
    "  scale_fill_manual(\"\",  \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  geom_sf(data = background_Boat,\n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.93,\n",
    "          colour=\"transparent\")+\n",
    "  geom_sf(data = background_Boat,\n",
    "          fill = \"white\",\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"transparent\")+\n",
    "  tidyterra::geom_spatraster_rgb(data = Boat_rgb_mask,\n",
    "                      r = 3,\n",
    "                      g = 2,\n",
    "                      b = 1,\n",
    "                      # maxcell = 5000,\n",
    "                      maxcell = 500832*8,\n",
    "                      max_col_value = 5000)+\n",
    "  geom_sf(data = zoom_boat_high_sf,\n",
    "          fill = \"transparent\",\n",
    "          colour = \"black\", \n",
    "          linewidth = 1)+\n",
    "  geom_sf(data = zoom_boat_high_sf_2,\n",
    "          fill = \"transparent\",\n",
    "          colour = \"black\", \n",
    "          linewidth = 1)+\n",
    "  annotation_custom(ggplotGrob(rgb_Boat_zoomed_map),\n",
    "                    xmin = -8.6975,\n",
    "                    xmax = -8.69,\n",
    "                    ymin = 40.6666,\n",
    "                    ymax = 40.6711) +\n",
    "  annotation_custom(ggplotGrob(rgb_Boat_zoomed_map_2),\n",
    "                    xmin = -8.6975+0.007,\n",
    "                    xmax = -8.69+0.007,\n",
    "                    ymin = 40.6666,\n",
    "                    ymax = 40.6711) +\n",
    "  annotation_custom(ggplotGrob(p1),\n",
    "                    xmin = -8.6795,\n",
    "                    xmax = -8.6736,\n",
    "                    ymin = 40.6585,\n",
    "                    ymax = 40.6625) +\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.69452127,-8.6965,-8.69440039,-8.69105), \n",
    "                              y=c(40.66244282,40.66673,40.66244282,40.66673),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 1)+\n",
    "    geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.6819877,-8.69645+0.007,-8.6818696,-8.6911+0.007), \n",
    "                              y=c(40.6652874,40.66675,40.6652874,40.66673),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 1)+\n",
    "  ggspatial::annotation_scale(location = \"bl\",\n",
    "                              text_cex=2)+\n",
    "  scale_x_continuous(limits = c(-8.6965,-8.675),\n",
    "                     breaks =  -8.685)+\n",
    "  scale_y_continuous(limits = c(40.659,40.6704),\n",
    "                     breaks =  40.664)+\n",
    "  theme_Bede_Map()+\n",
    "  theme(axis.title = element_blank(),\n",
    "        # axis.text.x = element_text(size = 20, angle = 20 , hjust = 1),\n",
    "        axis.text.x = element_blank(),\n",
    "        axis.text.y = element_text(size = 20))+\n",
    "   coord_sf(crs = 4326)\n",
    "\n",
    "\n",
    "####### Pred ######## \n",
    "Boat1_pred<-imglist %>% \n",
    "  filter(str_detect(path,\"AveiroBoat1_highPrediction_nnNewStanRawMetrics_Preds_NoXantho_7.tif\")) %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Boat1_pred)<-\"layer\"\n",
    "\n",
    "Boat2_pred<-imglist %>% \n",
    "  filter(str_detect(path,\"AveiroBoat2_highPrediction_nnNewStanRawMetrics_Preds_NoXantho_7.tif\")) %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Boat2_pred)<-\"layer\"\n",
    "\n",
    "values(Boat1_pred)[values(Boat1_pred) == 0] = NA\n",
    "values(Boat2_pred)[values(Boat2_pred) == 0] = NA\n",
    "\n",
    "\n",
    "B1_overextent<-ext(Boat1_pred)\n",
    "B1_overextent[2]<-ext(Boat2_pred)[2]\n",
    "B1_overextent[4]<-ext(Boat2_pred)[4]\n",
    "B1_overextent <- rast(B1_overextent)\n",
    "res(B1_overextent)<-res(Boat1_pred)\n",
    "# terra::writeRaster(B1$B1,\"B1.tif\")\n",
    "# terra::writeRaster(B2$B2,\"B2.tif\")\n",
    "\n",
    "B2_resampled <- terra::resample(Boat2_pred,B1_overextent, method = \"near\",threads = T)\n",
    "# terra::writeRaster(B2_resampled$B1,\"B2_resampled.tif\", overwrite = T)\n",
    "\n",
    "Boat_pred<-merge(Boat1_pred,B2_resampled)\n",
    "\n",
    "Boat_pred_aggregate<-terra::aggregate(Boat_pred,fact = 6, fun =\"modal\") \n",
    "\n",
    "masked_pred<-Boat_pred_aggregate %>% \n",
    "  mask(mask_boat,inverse=TRUE)%>% \n",
    "  as.factor()\n",
    "\n",
    "Boat_pred_zoom<-Boat1_pred %>% \n",
    "  crop(zoom_boat_high_sf) %>% \n",
    "  as.factor()\n",
    "\n",
    "Boat_pred_zoom_2<-Boat2_pred %>% \n",
    "  crop(zoom_boat_high_sf_2) %>% \n",
    "  as.factor()\n",
    "\n",
    "Boat_Pred_zoomed_map<- ggplot()+\n",
    "  tidyterra::geom_spatraster(data = Boat_pred_zoom, \n",
    "                             mapping = aes(fill = layer), \n",
    "                             show.legend = F,\n",
    "                             maxcell = 5010420)+\n",
    "  scale_fill_manual(values=c(\"#70543e\",\"#b3ff1a\",\"#389318\",\"#873e23\"),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Chlorphyceae\",\n",
    "                               \"Magnoliopsida\"))+\n",
    "  geom_sf(data = zoom_boat_high_sf, \n",
    "          fill = \"transparent\", \n",
    "          linetype = \"dashed\",\n",
    "          colour = \"black\", \n",
    "          linewidth = 1)+\n",
    "  theme_void()\n",
    "\n",
    "Boat_Pred_zoomed_map_2<- ggplot()+\n",
    "  tidyterra::geom_spatraster(data = Boat_pred_zoom_2, \n",
    "                             mapping = aes(fill = layer), \n",
    "                             show.legend = F,\n",
    "                             maxcell = 5010420)+\n",
    "  scale_fill_manual(values=c(\"#70543e\",\"#389318\",\"#873e23\",\"#b3002d\"),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\"))+\n",
    "  geom_sf(data = zoom_boat_high_sf_2, \n",
    "          fill = \"transparent\", \n",
    "          linetype = \"dashed\",\n",
    "          colour = \"black\", \n",
    "          linewidth = 1)+\n",
    "  theme_void()\n",
    "\n",
    "\n",
    "Boat_pred_map<-ggplot()+\n",
    "  geom_sf(data = background_Boat, \n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.93,\n",
    "          colour=\"transparent\")+\n",
    "  geom_sf(data = background_Boat,\n",
    "          fill = \"white\",\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"transparent\")+\n",
    "    scale_fill_manual(\"\", \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  ggnewscale::new_scale(\"fill\")+\n",
    "  tidyterra::geom_spatraster(data = masked_pred, \n",
    "                             mapping = aes(fill = layer),\n",
    "                             # show.legend = F,\n",
    "                             maxcell = 5010420,\n",
    "                             )+\n",
    "  scale_fill_manual(breaks = c(1,2,5,6,7,8,10),\n",
    "                    values=c(\"#70543e\", \"#b3ff1a\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", \"#42c9bc\",NA,NA,NA,NA),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Chlorophyceae\",\n",
    "                               \"Bacillariophyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\",\n",
    "                               \"Water\"))+\n",
    "  ggnewscale::new_scale(\"fill\")+\n",
    "  geom_sf(data = zoom_boat_high_sf, \n",
    "               fill = \"transparent\", \n",
    "               colour = \"black\",\n",
    "               linetype = \"dashed\", \n",
    "               linewidth = 1)+\n",
    "  geom_sf(data = zoom_boat_high_sf_2, \n",
    "               fill = \"transparent\", \n",
    "               colour = \"black\",\n",
    "               linetype = \"dashed\", \n",
    "               linewidth = 1)+\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.69452127,-8.6965,-8.69440039,-8.69105), \n",
    "                              y=c(40.66244282,40.66673,40.66244282,40.66673),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 1)+\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.6819877,-8.69645+0.007,-8.6818696,-8.6911+0.007), \n",
    "                              y=c(40.6652874,40.66675,40.6652874,40.66673),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 1)+\n",
    "  annotation_custom(ggplotGrob(Boat_Pred_zoomed_map),\n",
    "                    xmin = -8.6975,\n",
    "                    xmax = -8.69,\n",
    "                    ymin = 40.6666,\n",
    "                    ymax = 40.6711) +\n",
    "  annotation_custom(ggplotGrob(Boat_Pred_zoomed_map_2),\n",
    "                    xmin = -8.6975+0.007,\n",
    "                    xmax = -8.69+0.007,\n",
    "                    ymin = 40.6666,\n",
    "                    ymax = 40.6711) +\n",
    "  coord_sf(crs = 4326)+\n",
    "  theme_Bede_Map()+\n",
    "   theme(\n",
    "     # axis.text.x = element_text(size = 20, angle = 20,hjust = 1),\n",
    "     axis.text.x = element_blank(),\n",
    "         axis.title = element_blank(),\n",
    "        axis.text.y = element_text(size = 20),\n",
    "         legend.background = element_blank(),\n",
    "         legend.position = c(0.98,0.38),\n",
    "         legend.text = element_text(size = 18),\n",
    "         legend.key.size = unit(0.8,\"cm\"))+\n",
    "  scale_x_continuous(limits = c(-8.6965,-8.675),\n",
    "                     breaks =  -8.685)+\n",
    "  scale_y_continuous(limits = c(40.659,40.6704),\n",
    "                     breaks =  40.664)\n",
    "################# PLOT AND SAVING ###########\n",
    "plot_Boat<-RGB_Boat_map/\n",
    "  Boat_pred_map\n",
    "\n",
    "ggsave(\"Figures/Maps Pred/FigX-Boat_Pred.png\",plot_Boat , width = 1000*4, height = 1120*4, units = \"px\")"
   ],
   "id": "aa38320f-d51f-40a4-8d48-231fcb79805d"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Maps Pred/FigX-Boat_Pred.png\")"
   ],
   "id": "cell-fig-Boat"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The flight over L’Epine in Noirmoutier Island, France (<a href=\"#fig-Dike\" class=\"quarto-xref\">Figure 7</a>) was conducted near a dike, which crossed the northern part of the site from west to east. Alongside this dike, brown algae for the order Fucales (*Fucus spp.*, *Ascophyllum nodosum*) were attached to sparse rocks and stranded green algae (*Ulva spp.*) could be observed Despite the high mixture between Chlorophyceae and Magnoliopsida these two classes were correctly discriminated by the classifier."
   ],
   "id": "486669dd-5fb8-4710-84f9-0ec56409b150"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imglist<-\n",
    "  data.frame(path = list.files(pattern = \".tif\", recursive = T, full.names = T)) %>% \n",
    "  mutate(folder = case_when(str_detect(path, \"Data\") ~ \"Input\",\n",
    "                            TRUE ~ \"Output\"),\n",
    "         alt = case_when(str_detect(path, \"high\") ~ \"high\",\n",
    "                         str_detect(path, \"High\") ~ \"high\",\n",
    "                         TRUE ~ \"low\"),\n",
    "         site = case_when(str_detect(path, \"Gafanha\") ~ \"Gafanha\",\n",
    "                          str_detect(path, \"AveiroHigh\") ~ \"Gafanha\",\n",
    "                          TRUE ~ \"Duer\"),\n",
    "         type = case_when(str_detect(path, \"pred\") ~ \"Pred\",\n",
    "                          str_detect(path, \"Preds\") ~ \"Pred\",\n",
    "                          str_detect(path, \"NDVI\") ~ \"NDVI\",\n",
    "                          TRUE ~ \"ortho\"),\n",
    "         resampled = case_when(str_detect(path, \"resampled\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         new = case_when(str_detect(path, \"no_Xantho\") ~ \"TRUE\",\n",
    "                         str_detect(path, \"NewPreds\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         col = case_when(str_detect(path, \"colored\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         correction = case_when(str_detect(path, \"corrected_per_bands\") ~ \"Bands\",\n",
    "                          str_detect(path, \"corrected.tif\") ~ \"lm_all\",\n",
    "                          TRUE ~ \"NA\")) %>% \n",
    "  dplyr::filter(alt != \"NA\",\n",
    "                !str_detect(path,\".aux.xml\"))\n",
    "\n",
    "##### Preview #####\n",
    "\n",
    "filelist<-list.files(\"Data/shp/\", pattern = \".shp\", recursive = T,full.names = T)\n",
    "mask_BB <- filelist[str_detect(filelist,\"mask_land_intertidal_BB\")] %>% \n",
    "  read_sf()\n",
    "\n",
    "Flight_BB<-data.frame(\n",
    "  Name=c(\n",
    "  \"L'Epine\"\n",
    "  ) ,\n",
    "  Description = c(\n",
    "                  \"120m Flight\"),\n",
    "  Long=c(\n",
    "         -2.228239) ,\n",
    "  Lat=c( \n",
    "        46.989515) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(crs(mask_BB))  %>% \n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)\n",
    "\n",
    "p1 <- ggplot() +\n",
    "  scale_fill_manual(\"\", labels = c(\"Intertidal area\", \"Land mask\"),\n",
    "                    values = c(\"#7DC27D\", \"#BDBDBD\"))+\n",
    "  geom_sf(data = mask_BB, mapping = aes(fill = Type), linewidth=0.05,alpha=0.93, show.legend = F,colour=\"grey30\")+\n",
    "      coord_sf(xlim=c(550609,579539) ,\n",
    "          ylim=c(5189447,5218377) )+\n",
    "  ggforce::geom_mark_ellipse(data=Flight_BB,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = Name,\n",
    "                     group = Name),\n",
    "                     # description=Description),\n",
    "                 size=0.3,\n",
    "                 fill = \"goldenrod\",\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 show.legend=F,\n",
    "                 con.size = 2,\n",
    "                 label.fontsize = 10,\n",
    "                 label.hjust = 0.5,\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") ,\n",
    "  label.fill = \"grey90\",\n",
    "  label.buffer = unit(5, \"mm\")) +\n",
    "  # theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  # geom_label(aes(x = 553460, y = 5216442, label = \"B\"), size = 20)+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\") ,\n",
    "        legend.position = c(0.95,0.2),\n",
    "        legend.key.size =  unit(1, \"cm\"),\n",
    "        legend.text = element_text(size = 20),\n",
    "        legend.title=element_blank(),\n",
    "        legend.background = element_rect(linewidth = 0.5, \n",
    "                                         fill = \"white\",\n",
    "                                         linetype=\"solid\", \n",
    "                                  colour =\"black\"),\n",
    "        # panel.background = element_rect(fill = \"#FFFFFF90\"),\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_blank(),\n",
    "        axis.text.y = element_blank(),\n",
    "        axis.ticks = element_blank())+\n",
    "  theme_void()\n",
    "\n",
    "##### RGB ######\n",
    "\n",
    "Dike_rgb<-imglist %>% \n",
    "  filter(str_detect(path,\"NorthDike_high_multi_4326_colored.tif\")) %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Dike_rgb)<-c(\"B1\",\"B3\",\"B5\",\"alpha\")\n",
    "values(Dike_rgb)[values(Dike_rgb$alpha) == 0] = NA\n",
    "\n",
    "mask_Dike<-\"Data/shp/Bourgneuf\" %>% \n",
    "  list.files(\".shp\", full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path, \"mask_NorthDike\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "background_Dike<-\"Data/shp/Bourgneuf\" %>% \n",
    "  list.files(\".shp\", full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path, \"background\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "zoom_Dike_high_sf<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"zoom_extent_map_Dike_comparison\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "Dike_rgb_masked<-Dike_rgb %>% \n",
    "  crop(mask_Dike) %>%\n",
    "  mask(mask_Dike)\n",
    "\n",
    "\n",
    "zoom_rgb_dike<-Dike_rgb_masked %>% \n",
    "  crop(zoom_Dike_high_sf)\n",
    "\n",
    "zoom_rgb_dike_map<-ggplot()+\n",
    "  geom_spatraster_rgb(data = zoom_rgb_dike,\n",
    "                      r = 3,\n",
    "                      g = 2,\n",
    "                      b = 1)+\n",
    "  geom_sf(data = zoom_Dike_high_sf, fill = \"transparent\", linetype = \"dashed\",colour = \"black\", linewidth = 1)+\n",
    "  coord_sf(crs = 32630)+\n",
    "  theme_void()\n",
    "\n",
    "\n",
    "RGB_Dike_map <- ggplot()+\n",
    "  scale_fill_manual(\"\",  \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  geom_sf(data = background_Dike,\n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.93,\n",
    "          colour=\"transparent\")+\n",
    "  geom_sf(data = background_Dike,\n",
    "          fill = \"white\",\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"transparent\")+\n",
    "  tidyterra::geom_spatraster_rgb(data = Dike_rgb_masked,\n",
    "                      r = 3,\n",
    "                      g = 2,\n",
    "                      b = 1,\n",
    "                      # maxcell = 500,\n",
    "                      maxcell = 500832*8\n",
    "                      )+\n",
    "  geom_sf(data = zoom_Dike_high_sf, \n",
    "          fill = \"transparent\", \n",
    "          linetype = \"dashed\",\n",
    "          colour = \"black\", \n",
    "          linewidth = 1)+\n",
    "  annotation_custom(ggplotGrob(zoom_rgb_dike_map), \n",
    "                    xmin = -2.226, \n",
    "                    ymin = 46.9888,\n",
    "                    ymax = 46.9893) +\n",
    "  annotation_custom(ggplotGrob(p1),\n",
    "                    xmin = -2.226,\n",
    "                    ymin = 46.9875) +\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-2.226834,-2.225905,-2.226834,-2.225905),\n",
    "                              y=c(46.9886439,46.9892105,46.9885538,46.988805),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", linewidth = 1)+\n",
    "  ggspatial::annotation_scale(location = \"tl\",text_cex=2)+\n",
    "  scale_x_continuous(limits = c(-2.2293,-2.2253),\n",
    "                     breaks =  -2.228)+\n",
    "  scale_y_continuous(limits = c(46.9882,46.98935),\n",
    "                     breaks =  46.9890)+\n",
    "  theme_Bede_Map()+\n",
    "  theme(\n",
    "    # axis.text.x =element_text(size = 20, angle = 20, hjust = 1),\n",
    "    axis.text.x =element_blank(),\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.y = element_text(size = 20))+\n",
    "   coord_sf(crs = 4326)\n",
    "\n",
    "\n",
    "##### Pred ######\n",
    "\n",
    "Dike_Pred<-imglist %>% \n",
    "  filter(str_detect(path,\"Noirmoutier_highPrediction_nnNewStanRawMetrics_Preds_NoXantho_7_4326.tif\")) %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Dike_Pred)<-c(\"layer\")\n",
    "values(Dike_Pred)[values(Dike_Pred) == 0] = NA\n",
    "\n",
    "Dike_Pred_masked<-Dike_Pred %>% \n",
    "  crop(mask_Dike) %>% \n",
    "  mask(mask_Dike) %>% \n",
    "  as.factor()\n",
    "\n",
    "zoom_Pred_dike<-Dike_Pred_masked %>% \n",
    "  crop(zoom_Dike_high_sf)\n",
    "\n",
    "zoom_Pred_dike_map<-ggplot()+\n",
    "  tidyterra::geom_spatraster(data = zoom_Pred_dike, mapping = aes(fill = layer), show.legend = F)+\n",
    "  scale_fill_manual(values=c(\"#70543e\",\"#b3ff1a\",\"#389318\",\"#873e23\",\"#42c9bc\"),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Chlorphyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Water\"))+\n",
    "  geom_sf(data = zoom_Dike_high_sf, fill = \"transparent\", linetype = \"dashed\",colour = \"black\", linewidth = 1)+\n",
    "  theme_void()\n",
    "\n",
    "\n",
    "Dike_pred_map<-ggplot()+\n",
    "  geom_sf(data = background_Dike, \n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.93,\n",
    "          colour=\"grey30\")+\n",
    "  geom_sf(data = background_Dike,\n",
    "          fill = \"white\",\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"white\")+\n",
    "    scale_fill_manual(\"\",  \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  ggnewscale::new_scale(\"fill\")+\n",
    "  tidyterra::geom_spatraster(data = Dike_Pred_masked, \n",
    "                             mapping = aes(fill = layer),\n",
    "                             maxcell = 5010420,\n",
    "                             )+\n",
    "  scale_fill_manual(breaks = c(1,2,5,6,7,8,10),\n",
    "                    values=c(\"#70543e\", \"#b3ff1a\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", \"#42c9bc\",NA,NA,NA,NA),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Chlorophyceae\",\n",
    "                               \"Bacillariophyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\",\n",
    "                               \"Water\"))+\n",
    "  ggnewscale::new_scale(\"fill\")+\n",
    "  geom_sf(data = zoom_Dike_high_sf, \n",
    "          fill = \"transparent\", \n",
    "          colour = \"black\",\n",
    "          linetype = \"dashed\", \n",
    "          linewidth = 1)+\n",
    "  annotation_custom(ggplotGrob(zoom_Pred_dike_map), \n",
    "                    xmin = -2.226, \n",
    "                    ymin = 46.9888,\n",
    "                    ymax = 46.9893) +\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-2.226834,-2.225905,-2.226834,-2.225905),\n",
    "                              y=c(46.9886439,46.9892105,46.9885538,46.988805),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", linewidth = 1)+\n",
    "  # annotation_custom(ggplotGrob(Boat_Pred_zoomed_map), xmin = -8.697, xmax = -8.692, ymin = 40.667,ymax = 40.6705) +\n",
    "  coord_sf(crs = 4326)+\n",
    "  theme_Bede_Map()+\n",
    "   theme(axis.text.x = element_text(size = 20, angle = 20,hjust = 1),\n",
    "         axis.title = element_blank(),\n",
    "        axis.text.y = element_text(size = 20),\n",
    "         legend.background = element_blank(),\n",
    "         legend.position = c(1,0.50),\n",
    "         legend.text = element_text(size = 18),\n",
    "         legend.key.size = unit(0.8,\"cm\"))+\n",
    "  scale_x_continuous(limits = c(-2.2293,-2.2253),\n",
    "                     breaks =  -2.228)+\n",
    "  scale_y_continuous(limits = c(46.9882,46.98935),\n",
    "                     breaks =  46.9890)\n",
    "\n",
    "################# PLOT AND SAVING ###########\n",
    "\n",
    "plot_Dike<-RGB_Dike_map/\n",
    "  Dike_pred_map\n",
    "  \n",
    "ggsave(\"Figures/Maps Pred/FigX-Dike_Pred.png\",plot_Dike , width = 1200*4, height = 900*4, units = \"px\")"
   ],
   "id": "54030b61-01e3-44e9-8b7e-7aa20f3d29ae"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Maps Pred/FigX-Dike_Pred.png\")"
   ],
   "id": "cell-fig-Dike"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.2 Validation\n",
    "\n",
    "### 3.2.1 Reflectance comparison between the two different altitudes"
   ],
   "id": "1124bdfc-95b7-4356-aea4-1cdcc86ffd83"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# high_alt<- \"Data/image/micasense/Gafanha_high_multi.tif\" %>% \n",
    "#   rast()\n",
    "# \n",
    "# # low_alt<- \"Data/image/micasense/Gafanha_Low_Multi_rsp.tif\" %>% \n",
    "# #   rast() \n",
    "# \n",
    "# low_alt<- \"Data/image/micasense/Gafanha_low_multi.tif\" %>% \n",
    "#   rast() %>% \n",
    "#   resample(high_alt, method = \"med\")\n",
    "# \n",
    "# \n",
    "# low_alt_pred<- \"Data/image/prediction/NewPreds/Gafanha_Low_Pred_rsp.tif\" %>% \n",
    "#   rast() \n",
    "# \n",
    "# mask<-\"Data/shp/Mask_Alt_Compare_All.shp\" %>% \n",
    "#   vect()\n",
    "#   \n",
    "# high_alt_masked<-crop(high_alt,mask)\n",
    "# high_alt_masked<-mask(high_alt_masked,mask)\n",
    "# \n",
    "# low_alt_masked<-crop(low_alt,mask)\n",
    "# low_alt_masked<-mask(low_alt_masked,mask)\n",
    "# \n",
    "# low_alt_pred_masked<-crop(low_alt_pred,mask)\n",
    "# low_alt_pred_masked<-mask(low_alt_pred_masked,mask)\n",
    "# \n",
    "# \n",
    "# stk<-rast(list(c(high_alt_masked,low_alt_masked,low_alt_pred_masked)))%>% \n",
    "#   as.data.frame()\n",
    "# \n",
    "# names(stk)<-c(\"B1_high\",\"B2_high\",\"B3_high\",\"B4_high\",\"B5_high\",\"B6_high\",\"B7_high\",\"B8_high\",\"B9_high\",\"B10_high\",\n",
    "#               \"B1_low\",\"B2_low\",\"B3_low\",\"B4_low\",\"B5_low\",\"B6_low\",\"B7_low\",\"B8_low\",\"B9_low\",\"B10_low\",\n",
    "#               \"pred\")\n",
    "# \n",
    "# stk_long<-stk %>% \n",
    "#   mutate(px_ID = c(1:nrow(stk))) %>% \n",
    "#   pivot_longer(-c(px_ID,pred),values_to = \"Value\",names_to = \"Bands\") %>% \n",
    "#   mutate(alt = case_when(str_detect(Bands, \"low\") ~ \"Low\",\n",
    "#                          str_detect(Bands, \"high\") ~ \"High\",\n",
    "#                          TRUE ~ \"NA\"),\n",
    "#          Bands = gsub(\"_high\",\"\",Bands),\n",
    "#          Bands = gsub(\"_low\",\"\",Bands))\n",
    "# \n",
    "# \n",
    "# stk_summarized<-stk_long %>% \n",
    "#   # group_by(px_ID,alt) %>% \n",
    "#   # reframe(Value = mean(Value),\n",
    "#   #         pred = unique(pred)) %>% \n",
    "#   pivot_wider(names_from = \"alt\", values_from = \"Value\")\n",
    "# \n",
    "# df<-stk_summarized %>% \n",
    "#   ungroup() %>% \n",
    "#   group_by(Bands) %>% \n",
    "#   filter(Low<10000,\n",
    "#          High < 10000,\n",
    "#          Low < quantile(Low,probs = 0.99),\n",
    "#          High < quantile(High,probs = 0.99),\n",
    "#          Low > quantile(Low,probs = 0.01),\n",
    "#          High > quantile(High,probs = 0.01),\n",
    "#          !pred %in% c(7,9,3,10)) %>%\n",
    "#   ungroup()\n",
    " \n",
    "\n",
    "RAW <- read.csv(\"Data/CompareReflectance/Compare_ref_raw.csv\")%>%  \n",
    "  mutate(Low = Low /10000,\n",
    "         High = High/10000)\n",
    "\n",
    "RMSE_raw <- sqrt(mean((RAW$Low - RAW$High)^2))\n",
    "lm_raw <- lm(RAW$High ~ RAW$Low)\n",
    "\n",
    "STD <- read.csv(\"Data/CompareReflectance/Compare_ref_std.csv\")\n",
    "\n",
    "RMSE_STD <- sqrt(mean((STD$Low - STD$High)^2))\n",
    "lm_STD <- lm(STD$High ~ STD$Low)\n",
    "\n",
    "plot_RAW <- RAW %>%  \n",
    "  ggplot(aes(x=Low, y=High))+\n",
    "  # geom_point(aes(color = as.factor(pred)))\n",
    "  ggpointdensity::geom_pointdensity(adjust = .03)+\n",
    "  viridis::scale_color_viridis(limits = c(0,6000),\n",
    "                               breaks = c(0,2000,4000,6000),\n",
    "                               labels = c(\"1\",\"2000\",\"4000\",\"6000\"))+\n",
    "  # geom_smooth(method = \"loess\")+\n",
    "  geom_abline(intercept = 0, slope = 1, linewidth = 1, linetype = \"dashed\", color = \"black\" )+\n",
    "  # geom_smooth(method = \"lm\", color = \"red\", linetype = \"dashed\",linewidth = 1)+\n",
    "  theme_Bede()+\n",
    "  ylim(c(0.05,0.6))+\n",
    "  xlim(c(0.05,0.6))+\n",
    "  ylab(\"Reflectance of the 120m flight\")+\n",
    "  xlab(\"Reflectance of the 12m flight\")+\n",
    "  labs(color = \"Pixel density\")+\n",
    "  theme(\n",
    "    legend.position = \"top\",\n",
    "        legend.key.width = unit(3,\"line\"))\n",
    "\n",
    "plot_STD <- STD %>%  \n",
    "    ggplot(aes(x=Low, y=High))+\n",
    "  # geom_point(aes(color = as.factor(pred)))\n",
    "  ggpointdensity::geom_pointdensity(adjust = .03)+\n",
    "  viridis::scale_color_viridis(limits = c(1,2000),\n",
    "                               breaks = c(1,1000,2000),\n",
    "                               labels = c(\"1\",\"1000\",\"2000\"))+\n",
    "  # scale_color_gradient(na.value = 1)+\n",
    "  # geom_smooth(method = \"loess\")+\n",
    "  geom_abline(intercept = 0, slope = 1, linewidth = 1, linetype = \"dashed\", color = \"black\" )+\n",
    "  # geom_smooth(method = \"lm\", color = \"red\", linetype = \"dashed\",linewidth = 1)+\n",
    "  theme_Bede()+\n",
    "  ylim(c(0.05,1))+\n",
    "  xlim(c(0.05,1))+\n",
    "  ylab(\"Standardised reflectance of the 120m flight\")+\n",
    "  xlab(\"Standardised reflectance of the 12m flight\")+\n",
    "  labs(color = \"Pixel density\")+\n",
    "  theme(\n",
    "    legend.position = \"top\",\n",
    "        legend.key.width = unit(3,\"line\"))\n",
    "\n",
    "ggsave(\"./Figures/High_res/Compare_reflectance.png\",plot_RAW, width= 472*4, height=402*4, units = \"px\")\n",
    "ggsave(\"./Figures/High_res/Compare_reflectance_std.png\",plot_STD, width= 472*4, height=402*4, units = \"px\")"
   ],
   "id": "6ea9acb8-75ff-4d74-b22d-48c9110339df"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RAW <- ggplot()+\n",
    "  geom_spatraster_rgb(data = rast(\"./Figures/High_res/Compare_reflectance.png\"))+\n",
    "  coord_equal()+\n",
    "  geom_label(aes(x = 300,y= 1200, label = \"A\") ,size = 10)+\n",
    "    theme_void()+\n",
    "  theme(axis.text = element_blank(), \n",
    "        axis.title = element_blank(), \n",
    "        axis.ticks = element_blank(), \n",
    "        axis.ticks.length = unit(0, \"pt\"),\n",
    "        panel.grid.major=element_blank(), \n",
    "        panel.grid.minor=element_blank(), \n",
    "        plot.margin = margin(0, 0, 0, 0, \"pt\"))\n",
    "\n",
    "STD <- ggplot()+\n",
    "  geom_spatraster_rgb(data = rast(\"./Figures/High_res/Compare_reflectance_std.png\"))+\n",
    "  coord_equal()+\n",
    "  geom_label(aes(x = 300,y= 1200, label = \"B\") ,size = 10)+\n",
    "    theme_void()+\n",
    "  theme(axis.text = element_blank(), \n",
    "        axis.title = element_blank(), \n",
    "        axis.ticks = element_blank(), \n",
    "        axis.ticks.length = unit(0, \"pt\"),\n",
    "        panel.grid.major=element_blank(), \n",
    "        panel.grid.minor=element_blank(), \n",
    "        plot.margin = margin(0, 0, 0, 0, \"pt\"))\n",
    "\n",
    "\n",
    "plot <- RAW+STD\n",
    "\n",
    "ggsave(\"./Figures/High_res/Compare_reflectance_both.png\",plot,width= 1920*4, height=1129*4, units = \"px\")"
   ],
   "id": "40cf23c0-29db-4483-9f51-52e05a58fe55"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/High_res/Compare_reflectance_both.png\")"
   ],
   "id": "cell-fig-CompareRef"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this study, a key innovation lies in the utilization of drone flights at two different altitudes (12m and 120m) for constructing the neural network model. The lower altitude flights (8mm spatial resolution) enabled precise selection of pure pixels to train the neural network model. This methodology implies a consistency between the reflectance of both altitudes. <a href=\"#fig-CompareRef\" class=\"quarto-xref\">Figure 8</a> depicts the relationship between reflectance from a low and a high-altitude flights conducted over the same area. Overall there was a good agreement between the two altitudes. There was however a slight underestimation for raw reflectance values in the high-altitude flight, particularly for higher reflectance values (<a href=\"#fig-CompareRef\" class=\"quarto-xref\">Figure 8</a> A). Since both flights were conducted over vegetation areas, the highest reflectance values correspond to the infrared part of the spectrum. This different is not present anymore when reflectance values of both flights have been standardized (<a href=\"#eq-std\" class=\"quarto-xref\">Equation 1</a> ; <a href=\"#fig-CompareRef\" class=\"quarto-xref\">Figure 8</a> B).\n",
    "\n",
    "### 3.2.2 Neural network classification validation"
   ],
   "id": "85313f2f-810e-4be6-8d9c-21eadcd858db"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: Duplicated aesthetics after name standardisation: size"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    }
   ],
   "source": [
    "df<-\"Output/Validation_data\" %>% \n",
    "  list.files(\"ResultsNewNN_NoXanthowithNDVIWithOldandBoat.csv\",recursive = T, full.names = T) %>% \n",
    "  read.csv() %>% \n",
    "  as_tibble() %>% \n",
    "  rename(Truth_ID = \"Truth\") %>% \n",
    "  mutate(Truth =  case_when(Truth_ID == 1 ~ \"Sediment\",\n",
    "                          Truth_ID == 2 ~ \"Chlorophyceae\",\n",
    "                          Truth_ID == 3 ~ \"Sediment\",\n",
    "                          Truth_ID == 4 ~ \"Magnoliopsida\",\n",
    "                          Truth_ID == 5 ~ \"Bacillariophyceae\",\n",
    "                          Truth_ID == 6 ~ \"Magnoliopsida\",\n",
    "                          Truth_ID == 7 ~ \"Phaeophyceae\",\n",
    "                          Truth_ID == 8 ~ \"Rhodophyceae\",\n",
    "                          Truth_ID == 9 ~ \"SunGlint\",\n",
    "                          Truth_ID == 10 ~ \"Water\",\n",
    "                          Truth_ID == 11 ~ \"Xanthophyceae\",\n",
    "                          Truth_ID == 12 ~ \"Spartina\",\n",
    "                          Truth_ID == 41 ~ \"Bleached seagrasses\",\n",
    "                          TRUE ~ \"NA\"),\n",
    "         Pred_Class = case_when(Pred_Class == \"Clorophyta\" ~ \"Chlorophyceae\",\n",
    "                                Pred_Class == \"Phaeophyta\" ~ \"Phaeophyceae\",\n",
    "                                Pred_Class == \"Rhodphyta\" ~ \"Rhodophyceae\",\n",
    "                                Pred_Class == \"Deep_Sediment\" ~ \"Sediment\",\n",
    "                                Pred_Class == \"MPB\" ~ \"Bacillariophyceae\",\n",
    "                                Pred_Class == \"Bare_Sediment\" ~ \"Sediment\",\n",
    "                                 Pred_Class == \"Magnoliosida\" ~ \"Magnoliopsida\",\n",
    "                                (Pred_Class == \"Low_SPC\" & NDVI > 0.2) ~ \"Magnoliopsida\",\n",
    "                                (Pred_Class == \"Low_SPC\" & NDVI <= 0.2 )~ \"Sediment\",\n",
    "                          TRUE ~ Pred_Class)) %>% \n",
    "  dplyr::filter(Truth != \"Water\",\n",
    "                Truth  != \"Spartina\",\n",
    "                Truth != \"SunGlint\",\n",
    "                Truth != \"Bleached seagrasses\",\n",
    "                Truth != \"Xanthophyceae\",\n",
    "                # Pred_Class != \"Low_SPC\",\n",
    "                Pred_Class != \"Xanthophyceae\",\n",
    "                 Pred_Class != \"Water\",\n",
    "                 Pred_Class != \"SunGlint\",\n",
    "                !Flight == \"Handfisher_1_old\",\n",
    "                !Flight == \"Handfisher_2_old\",\n",
    "                !Flight == \"Handfisher_2\") %>%\n",
    "  dplyr::select(-NDVI)\n",
    "\n",
    "\n",
    "a<-as.character(unique(df$Truth))\n",
    "b<-as.character(unique(df$Pred_Class))\n",
    "ab<-setdiff(a,b)\n",
    "ba<-setdiff(b,a)\n",
    "\n",
    "if (!(length(ab) == 0 & length(ba) == 0)) {\n",
    "  \n",
    "\n",
    "lines_to_add<-data.frame(Pred_Class = c(ab,ba),\n",
    "                         Pred_ID = NA,\n",
    "                         Flight = NA,\n",
    "                         Truth_ID = NA,\n",
    "                         Truth = c(ab,ba))\n",
    "\n",
    "df<-df %>% \n",
    "  rbind(lines_to_add)\n",
    "  rm(lines_to_add)\n",
    "\n",
    "  \n",
    "}\n",
    "\n",
    "\n",
    "  df<- df %>%\n",
    "  mutate(Pred_Class = as_factor(Pred_Class),\n",
    "         Truth = as_factor(Truth),\n",
    "         Truth = fct_relevel(Truth, c(\"Sediment\",\"Bacillariophyceae\",\"Chlorophyceae\",\"Magnoliopsida\",\"Rhodophyceae\",\"Phaeophyceae\")),\n",
    "         Pred_Class = fct_relevel(Pred_Class, c(\"Sediment\",\"Bacillariophyceae\",\"Chlorophyceae\",\"Magnoliopsida\",\"Rhodophyceae\",\"Phaeophyceae\")),\n",
    "         Flight = case_when(str_detect(Flight, \"Boat\") ~ \"Boat\",\n",
    "                            TRUE ~ Flight))\n",
    "  \n",
    "####### GafanahaLow #######\n",
    "df_GafanhaLow<-df %>% \n",
    "  filter(Flight == \"GafanhaLow\")\n",
    "\n",
    "validation_metrics_GafanhaLow<-confusionMatrix(df_GafanhaLow$Pred_Class,df_GafanhaLow$Truth)\n",
    "\n",
    "conf_mat_GafanhaLow<-validation_metrics_GafanhaLow[[2]] %>% \n",
    "  as.numeric() %>%\n",
    "  base::matrix(ncol = 6) %>%\n",
    "  as.data.frame()\n",
    "conf_mat_GafanhaLow[6,]<-NA\n",
    "conf_mat_GafanhaLow[,6]<-NA\n",
    "\n",
    "for(i in 1:(nrow(conf_mat_GafanhaLow))){\n",
    "  if(i == 1 ){\n",
    "    df_sum<-data.frame(row = NA,\n",
    "                       col = NA)\n",
    "  }\n",
    "  df_sum[i,1] <- sum(conf_mat_GafanhaLow[i,],na.rm = T)\n",
    "  df_sum[i,2] <- sum(conf_mat_GafanhaLow[,i],na.rm = T)\n",
    "}\n",
    "conf_mat_std_GafanhaLow<-conf_mat_GafanhaLow\n",
    "\n",
    "for(row in 1:nrow(conf_mat_std_GafanhaLow)){\n",
    "  for(col in 1:ncol(conf_mat_std_GafanhaLow)){\n",
    "    conf_mat_std_GafanhaLow[row,col]<-conf_mat_std_GafanhaLow[row,col]/((df_sum$row[row] + df_sum$col[col])-conf_mat_std_GafanhaLow[row,col])\n",
    "    }\n",
    "}\n",
    "\n",
    "confmat_rast_GafanhaLow<-conf_mat_std_GafanhaLow %>% \n",
    "  as.matrix() %>% \n",
    "  as.array() %>% \n",
    "  rast()\n",
    "\n",
    "confmat_rast_GafanhaLow<-subst(confmat_rast_GafanhaLow, NA, -99)\n",
    "names(confmat_rast_GafanhaLow)<-\"layer\"\n",
    "ext(confmat_rast_GafanhaLow)<-c(18.1,21.9,8.2,11.8)\n",
    "confmat_rast_GafanhaLow_sf<-as.polygons(confmat_rast_GafanhaLow,aggregate=F) %>% \n",
    "  sf::st_as_sf()\n",
    "confmat_rast_GafanhaLow_centroid<-st_centroid(confmat_rast_GafanhaLow_sf)"
   ],
   "id": "8b73b39e-bf76-4dc9-b055-cff9dc436de8"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Validation/ConfusionMatrixGlobal.png\")"
   ],
   "id": "cell-fig-Validation"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A total of 536,000 pixels were used to validate the Neural Network classifier. The sites with the lowest and highest number of validation data were Kerdrean (5557 pixels) and Marinha Lanzarote (159713 pixels), respectively. Model global accuracy was 94.26% with a Kappa coefficient of 0.92 (<a href=\"#fig-Validation\" class=\"quarto-xref\">Figure 9</a>). The lowest performing site was Gafanha High (global accuracy of 75.45%) whereas Mataducos was the site with the most accurate prediction (global accuracy of 98.05%). Overall, the classes Phaeophyceae, Magnoliopsida, Sediment and Rhodophyceae were correctly classified with a balanced accuracy of 1, 0.96, 0.96 and 0.91 respectively. Bacillariophyceae was the least performing class (accuracy of 0.72 ) mainly due to the confusion with Magnoliopsida and Sediment.\n",
    "\n",
    "## 3.3 Variable importance"
   ],
   "id": "e7841a94-50aa-44f0-9ebe-6f496b609c98"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################# SPECTRA ########### \n",
    "SP1<-read.csv(\"Output/Spectra_VIP.csv\") %>% \n",
    "  dplyr::select(-X) %>% \n",
    "  mutate(type = \"Min/Max Standardised Spectra\") %>% \n",
    "  dplyr::filter(Class != \"Bare Sediment\",\n",
    "                Class != \"Sun Glint\",\n",
    "                Class != \"Water\",\n",
    "                Class != \"Xanthophyceae\",\n",
    "                Class != \"A - MagnoliopsidaL\") %>% \n",
    "  as_tibble()%>% \n",
    "  rename(value = \"STD\")\n",
    "\n",
    "SP2<-read.csv(\"Output/Spectra_VIP2.csv\") %>% \n",
    "  dplyr::select(-X) %>% \n",
    "  mutate(type = \"Raw Spectra\",\n",
    "         mean = mean / 10000)%>% \n",
    "  dplyr::filter(Class != \"Bare Sediment\",\n",
    "                Class != \"Sun Glint\",\n",
    "                Class != \"Water\",\n",
    "                Class != \"Xanthophyceae\",\n",
    "                Class != \"A - MagnoliopsidaL\") %>% \n",
    "  as_tibble() %>% \n",
    "  dplyr::reframe(Class,Wavelength, mean, type) %>% \n",
    "  rename(value = \"mean\")\n",
    "\n",
    "SP<-SP1 %>% \n",
    "  rbind(SP2)\n",
    "\n",
    "\n",
    "SP_plot<-ggplot(SP, aes( x = Wavelength, y = value, color = Class))+\n",
    "  geom_line(linewidth = 1.5, show.legend = F)+\n",
    "  facet_wrap(~ type, scale = \"free\",ncol =1) +\n",
    "  scale_color_manual(values=c(\"#389318\", \"#873e23\", \"#b3002d\", \"#b3ff1a\", \"#DAA520\"))+\n",
    "  ylab(\"Reflectance\")+\n",
    "  xlab(\"Wavelength (nm)\")+\n",
    "  theme_Bede()+\n",
    "  theme(axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20),\n",
    "        axis.title.x = element_text(size = 25),\n",
    "        axis.title.y = element_text(size = 25),\n",
    "        strip.text = element_text(size=25),\n",
    "        legend.position = c(0.2, 0.7),\n",
    "        plot.background = element_rect(fill = rgb(1,1,1, alpha=0), colour = NA),\n",
    "        legend.title=element_blank(),\n",
    "        legend.text = element_text(size = 5),\n",
    "        legend.key.size = unit(0.25, \"cm\"),\n",
    "        legend.text.align = 0,\n",
    "        panel.border = element_blank(), \n",
    "        panel.background = element_blank(),\n",
    "        axis.line = element_line(color=\"black\"))\n",
    "\n",
    "############## VIP ##############\n",
    "\n",
    "filelist<-\"Data\" %>% \n",
    "  list.files(pattern = \"NN_VIP_byClass_NoXanth.csv\", recursive = T,full.names = T)\n",
    "\n",
    "VIP<-filelist[which(str_detect(filelist,\"VIP\"))] %>% \n",
    "  read.csv() %>% \n",
    "  mutate(Magnoliopsida = (Magnoliosida + Low_SPC)/2) %>% \n",
    "  dplyr::select(-c(Magnoliosida ,Low_SPC))\n",
    "\n",
    "VIP1<-VIP %>%\n",
    "  dplyr::select(-c(Water,\n",
    "                   SunGlint,\n",
    "                   Deep_Sediment,\n",
    "                   Bare_Sediment)) %>%\n",
    "  # dplyr::filter(!str_detect(X,\"NDVI\")) %>% \n",
    "  mutate(isSTD = case_when(str_detect(X,\"Stan\") ~ \"Standardised\", \n",
    "                         TRUE ~ \"RAW\")) %>%\n",
    "  pivot_longer(-c(X,isSTD), names_to = \"Class\",values_to = \"value\") %>% \n",
    "  dplyr::filter(X != \"NDVI\",\n",
    "                X != \"NDVI_Stan\") %>% \n",
    "  group_by(Class) %>%\n",
    "  mutate(STD = (value-min(value))/(max(value)-min(value))) %>% \n",
    "  ungroup() %>% \n",
    "  mutate(Class = case_when(Class == \"Magnoliosida\" ~ \"Magnoliopsida\",\n",
    "                           Class == \"Low_SPC\" ~ \"Magnoliopsida\",\n",
    "                           Class == \"Clorophyta\" ~ \"Chlorophyceae\",\n",
    "                           Class == \"Rhodphyta\" ~ \"Rhodophyceae\",\n",
    "                           Class == \"MPB\" ~ \"Bacillariophyceae\",\n",
    "                           Class == \"Phaeophyta\" ~ \"Phaeophyceae\",\n",
    "                           TRUE ~ Class),\n",
    "         Variable = str_replace_all(str_replace_all(str_replace_all(X,\"Reflectance\",\"\"),\"Stan\",\"\"),\"_\",\"\")) %>% \n",
    "  dplyr::filter(Class != \"Total\") \n",
    "\n",
    "VIP_summarised<-VIP1 %>% \n",
    "  group_by(Variable, Class) %>% \n",
    "  dplyr::summarise(sumVIP = sum(STD)) %>% \n",
    "  ungroup() %>% \n",
    "  group_by(Class) %>% \n",
    "  mutate(STD = (sumVIP-min(sumVIP))/(max(sumVIP)-min(sumVIP)))\n",
    "\n",
    "VIP_subPLOT<-VIP_summarised %>% \n",
    "  ggplot(aes(x = Variable, y = STD, fill = Class)) +\n",
    "  geom_col(color = \"black\",position =  \"dodge\") +\n",
    "  coord_polar()+\n",
    "  scale_y_continuous(limits = c(0,1), breaks = c(0.25,0.5,0.75,1))+\n",
    "  facet_wrap(~Class)+\n",
    "  scale_fill_manual(\"Class\", labels = c(\"Bacillariophyceae\",\"Chlorophyceae\", \"Magnoliopsida\", \"Phaeophyceae\",\"Rhodophyceae\"),\n",
    "                    values = c(\"#DAA520\",\"#b3ff1a\", \"#389318\",\"#873e23\",\"#b3002d\"))+\n",
    "  theme_Bede()+\n",
    "  theme(panel.grid.major.y = element_line(linetype = \"dotted\",color = c(\"grey9\",\"grey9\",\"grey9\",\"grey9\",NA)))+\n",
    "  ylab(\"Importance\")+\n",
    "  theme(legend.position=\"top\",\n",
    "        strip.text = element_text(size = 25),\n",
    "        axis.text.y = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.ticks.y =element_blank(),\n",
    "        axis.title.x =element_blank(),\n",
    "        axis.title.y =element_text(size = 20),\n",
    "        legend.key.size = unit(1, 'cm'),\n",
    "        legend.title = element_text(size=20),\n",
    "        legend.text = element_text(size=20))+\n",
    "  guides(fill = guide_legend(nrow = 1))\n",
    "\n",
    "VIP_PLOT<-VIP_subPLOT+\n",
    "  # geom_text(aes(label = \"Standardised Reflectance Spectra\"), color = \"black\", size = 8)+\n",
    "  inset_element(SP_plot,left = 0.6341, top = 0.47, bottom = -0.01, right = 0.90,  align_to = \"full\")\n",
    "VIP_PLOT\n",
    "ggsave(\"Figures/VIP/Fig_VIP.png\", VIP_PLOT, width = 1920*4, height = 1129*4, units = \"px\")\n",
    "\n",
    "\n",
    "################# VIP NDVI ######################\n",
    "\n",
    "VIP_NDVI<-\n",
    "  VIP %>%\n",
    "  dplyr::select(-c(Water,\n",
    "                   SunGlint,\n",
    "                   Deep_Sediment,\n",
    "                   Bare_Sediment)) %>%\n",
    "  # dplyr::filter(!str_detect(X,\"NDVI\")) %>% \n",
    "  mutate(isSTD = case_when(str_detect(X,\"Stan\") ~ \"Standardised\", \n",
    "                         TRUE ~ \"RAW\")) %>%\n",
    "  pivot_longer(-c(X,isSTD), names_to = \"Class\",values_to = \"value\") %>% \n",
    "  dplyr::filter(X == \"NDVI\") %>% \n",
    "  group_by(Class) %>%\n",
    "  mutate(STD = (value-min(value))/(max(value)-min(value))) %>% \n",
    "  ungroup() %>% \n",
    "  mutate(Class = case_when(Class == \"Magnoliosida\" ~ \"Magnoliopsida\",\n",
    "                           Class == \"Low_SPC\" ~ \"Magnoliopsida\",\n",
    "                           Class == \"Clorophyta\" ~ \"Chlorophyceae\",\n",
    "                           Class == \"Rhodphyta\" ~ \"Rhodophyceae\",\n",
    "                           Class == \"MPB\" ~ \"Bacillariophyceae\",\n",
    "                           Class == \"Phaeophyta\" ~ \"Phaeophyceae\",\n",
    "                           TRUE ~ Class),\n",
    "         Variable = str_replace_all(str_replace_all(str_replace_all(X,\"Reflectance\",\"\"),\"Stan\",\"\"),\"_\",\"\")) %>% \n",
    "  dplyr::filter(Class != \"Total\") %>% \n",
    "  group_by(Class) %>% \n",
    "  summarise(sum = sum(value)) %>% \n",
    "  mutate(STD = (sum-min(sum))/(max(sum)-min(sum))) %>% \n",
    "  ggplot(aes(x = Class, y = STD, fill = Class)) +\n",
    "  geom_col(color = \"black\",position =  \"dodge\") +\n",
    "  # coord_polar()+\n",
    "  scale_y_continuous(limits = c(0,1), breaks = c(0.25,0.5,0.75,1))+\n",
    "  # facet_wrap(~Class)+\n",
    "  scale_fill_manual(\"Class\", labels = c(\"Bacillariophyceae\",\"Chlorophyceae\", \"Magnoliopsida\", \"Phaeophyceae\",\"Rhodophyceae\"),\n",
    "                    values = c(\"#DAA520\",\"#b3ff1a\", \"#389318\",\"#873e23\",\"#b3002d\"))+  theme_Bede()+\n",
    "  theme(panel.grid.major.y = element_line(linetype = \"dotted\",color = c(\"grey9\",\"grey9\",\"grey9\",\"grey9\",NA)))+\n",
    "  ylab(\"Importance\")+\n",
    "  theme(legend.position=\"top\",\n",
    "        strip.text = element_text(size = 25),\n",
    "        axis.text.y = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.ticks.y =element_blank(),\n",
    "        axis.title.x =element_blank(),\n",
    "        axis.title.y =element_text(size = 20),\n",
    "        legend.key.size = unit(1, 'cm'),\n",
    "        legend.title = element_text(size=20),\n",
    "        legend.text = element_text(size=20))+\n",
    "  guides(fill = guide_legend(nrow = 1))"
   ],
   "id": "962d8e28-7971-4977-a396-051a7d4ef853"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"Figures/Low_res/VIP/Fig_VIP.png\")"
   ],
   "id": "cell-fig-VIP"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The computation of the variable importance made it possible to identify which bands were the most useful for class prediction (<a href=\"#fig-VIP\" class=\"quarto-xref\">Figure 10</a>). The spectral bands at 444, 717 and 842 nm of the Micasense camera did not provide as important information to discriminate any of the vegetation classes. The band at 531 nm was the only important predictor for the classifier to accurately predict Chlorophyceae. In fact, at this wavelength, the Chlorophyceae spectra showed the highest reflectance among of all vegetation classes. The bands at 531 and 740 nm were the most important predictors for Phaeophyceae, corresponding to the lowest reflectance among all classes. Bands at 475 and 560 nm were the most important predictors for Bacillariophyceae and Rhodophyceae, respectively. Four predictors, ranging from the green (560 nm) to the RedEdge (705 nm) bands were important to accurately predict Magnoliopsida.\n",
    "\n",
    "## 3.4 Effect of the flight height on the prediction"
   ],
   "id": "19ac9c15-5f66-4d80-be5e-bbaf07c26fe0"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if(length(list.files(\"./Output\",pattern = \"output_80_filtered_Upscaling.csv\")) == 0 ){\n",
    " imglist<-\"Output/High_Flights_correction/Low_flight_spatial_resampling/\" %>%\n",
    "   list.files(\".tif\",full.names = T) %>%\n",
    "   as.data.frame() %>%\n",
    "   rename(path = \".\") %>%\n",
    "   mutate(size = case_when(str_detect(path,\"_8_7x7.tif\") ~ 8,\n",
    "                           str_detect(path,\"16.tif\") ~ 16,\n",
    "                           str_detect(path,\"32.tif\") ~ 32,\n",
    "                           str_detect(path,\"48.tif\") ~ 48,\n",
    "                           str_detect(path,\"64.tif\") ~ 64,\n",
    "                           str_detect(path,\"80.tif\") ~ 80),\n",
    "          type = case_when(str_detect(path, \"ortho\") ~ \"ortho\",\n",
    "                           str_detect(path, \"pred\") ~ \"Pred\",\n",
    "                           TRUE ~ \"Prob\"))\n",
    "\n",
    "pred_8<-imglist %>%\n",
    "   filter(size == 8,\n",
    "          type == \"Pred\") %>%\n",
    "   pull(path) %>%\n",
    "   terra::rast()\n",
    " names(pred_8)<-\"Pred_8\"\n",
    "\n",
    " pred_80<-imglist %>%\n",
    "   filter(size == 80,\n",
    "          type == \"Pred\") %>%\n",
    "   pull(path) %>%\n",
    "   terra::rast()\n",
    " names(pred_80)<-\"Pred_80\"\n",
    " \n",
    "prob_80<-imglist %>%\n",
    "   filter(size == 80,\n",
    "          type == \"Prob\") %>%\n",
    "   pull(path) %>%\n",
    "   terra::rast()\n",
    " names(prob_80)<-\"Prob_80\"\n",
    "\n",
    "\n",
    " model_80<-imglist %>%\n",
    "   filter(size == 80,\n",
    "          type == \"Pred\") %>%\n",
    "   pull(path) %>%\n",
    "   terra::rast()\n",
    " values(model_80)<-c(1:ncell(model_80))\n",
    " names(model_80)<-\"ID_80\"\n",
    "\n",
    "\n",
    " model_80_rs_to_8<-terra::resample(model_80,pred_8,method = \"near\")\n",
    " names(model_80_rs_to_8)<-\"ID_80\"\n",
    "\n",
    " df_80_8<-terra::rast(list(model_80_rs_to_8,pred_8)) %>%\n",
    "   as.data.frame(xy=F) %>%\n",
    "   filter(!Pred_8 == 0)\n",
    "\n",
    " df_80<-terra::rast(list(model_80,pred_80,prob_80)) %>%\n",
    "   as.data.frame(xy=F) %>%\n",
    "   filter(!Pred_80 == 0)\n",
    "\n",
    "#############\n",
    "output_80<-df_80 %>% \n",
    "  left_join(df_80_8, by = \"ID_80\") %>% \n",
    "  filter(!is.na(Pred_8)) %>% \n",
    "  mutate(Class_80 = case_when(Pred_80 == 1 ~ \"Bare Sediment\",\n",
    "                              Pred_80 == 2 ~ \"Green Algae\",\n",
    "                              Pred_80 == 3 ~ \"Bare Sediment\",\n",
    "                              Pred_80 == 4 ~ \"Magnoliopsida\",\n",
    "                              Pred_80 == 5 ~ \"Microphytobenthos\",\n",
    "                              Pred_80 == 6 ~ \"Magnoliopsida\",\n",
    "                              Pred_80 == 7 ~ \"Phaeophyceae\",\n",
    "                              Pred_80 == 8 ~ \"Rhodophyceae\",\n",
    "                              Pred_80 == 9 ~ \"Sun Glint\",\n",
    "                              Pred_80 == 10 ~ \"Water\",\n",
    "                              Pred_80 == 11 ~ \"Xanthophyceae\",\n",
    "                              TRUE ~ as.character(Pred_80)),\n",
    "         Class_8 = case_when(Pred_8 == 1 ~ \"Bare Sediment\",\n",
    "                              Pred_8 == 2 ~ \"Green Algae\",\n",
    "                              Pred_8 == 3 ~ \"Bare Sediment\",\n",
    "                              Pred_8 == 4 ~ \"Magnoliopsida\",\n",
    "                              Pred_8 == 5 ~ \"Microphytobenthos\",\n",
    "                              Pred_8 == 6 ~ \"Magnoliopsida\",\n",
    "                              Pred_8 == 7 ~ \"Phaeophyceae\",\n",
    "                              Pred_8 == 8 ~ \"Rhodophyceae\",\n",
    "                              Pred_8 == 9 ~ \"Sun Glint\",\n",
    "                              Pred_8 == 10 ~ \"Water\",\n",
    "                              Pred_8 == 11 ~ \"Xanthophyceae\",\n",
    "                              TRUE ~ as.character(Pred_8))\n",
    "         ) %>% \n",
    "   dplyr::filter(Class_80 != \"Water\",\n",
    "                 Class_80 != \"Phaeophyceae\",\n",
    "                 Class_80 != \"Sun Glint\",\n",
    "                 Class_8 != \"Water\",\n",
    "                 Class_8 != \"Phaeophyceae\",\n",
    "                 Class_8 != \"Sun Glint\") %>% \n",
    "           group_by(ID_80, Prob_80) %>% \n",
    "           summarise(prop_Sed = sum((Pred_8 == 1)|(Pred_8 == 3))/n(),\n",
    "                     prop_Green = sum(Pred_8 == 2)/n(),\n",
    "                     prop_Magno = sum((Pred_8 == 4)|(Pred_8 == 6))/n(),\n",
    "                     prop_MPB = sum(Pred_8 == 5)/n(),\n",
    "                     prop_Rodo = sum(Pred_8 == 8)/n(),\n",
    "                     prop_Xantho = sum(Pred_8 == 11)/n(),\n",
    "                     True = unique(Class_80)) %>%\n",
    "  pivot_longer(-c(ID_80,True,Prob_80), names_to = \"Class\", values_to = \"Proportion\") %>% \n",
    "  mutate(Class = case_when(Class == \"prop_Sed\" ~ \"Bare Sediment\",\n",
    "                           Class == \"prop_Green\" ~ \"Green Algae\",\n",
    "                           Class == \"prop_Magno\" ~ \"Magnoliopsida\",\n",
    "                           Class == \"prop_MPB\" ~ \"Microphytobenthos\",\n",
    "                           Class == \"prop_Pheo\" ~ \"Phaeophyceae\",\n",
    "                           Class == \"prop_Rodo\" ~ \"Rhodophyceae\",\n",
    "                           Class == \"prop_Glint\" ~ \"Sun Glint\",\n",
    "                           Class == \"prop_Water\" ~ \"Water\",\n",
    "                           Class == \"prop_Xantho\" ~ \"Xanthophyceae\"),\n",
    "         same = case_when(True == Class ~ 1,\n",
    "                          TRUE ~ 0)) %>% \n",
    "  dplyr::filter(!Proportion == 0)\n",
    " \n",
    "  list<-output_80 %>% \n",
    "  dplyr::filter(Proportion == 1,\n",
    "                Class == True) %>% \n",
    "  pull(ID_80)\n",
    "\n",
    "output_80_filtered<-output_80 %>% \n",
    "  dplyr::filter(!ID_80 %in% list )\n",
    "\n",
    "\n",
    "\n",
    "output_80_filtered_same<-output_80_filtered %>% \n",
    "  dplyr::filter(same == 1)\n",
    "\n",
    "write.csv(output_80_filtered, \"./Output/output_80_filtered_Upscaling.csv\")\n",
    "write.csv(output_80_filtered_same, \"./Output/output_80_filtered_same_Upscaling.csv\")\n",
    "\n",
    "}else{\n",
    "  \n",
    "  output_80_filtered<-read.csv(\"./Output/output_80_filtered_Upscaling.csv\") %>% \n",
    "    dplyr::mutate(Proportion = Proportion *100) %>% \n",
    "    dplyr::select(-X) %>% \n",
    "    as_tibble()\n",
    "  output_80_filtered_same<-read.csv(\"./Output/output_80_filtered_same_Upscaling.csv\")%>% \n",
    "    dplyr::mutate(Proportion = Proportion *100) %>% \n",
    "    dplyr::select(-X) %>% \n",
    "    as_tibble()\n",
    "}\n",
    "\n",
    "################# LM ####################\n",
    "lm_sediment<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Bare Sediment\") \n",
    "\n",
    " lm_sediment<- lm(lm_sediment$Prob_80 ~ lm_sediment$Proportion)\n",
    " \n",
    " lm_green<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Green Algae\") \n",
    "  lm_green<- lm(lm_green$Prob_80 ~ lm_green$Proportion)\n",
    " \n",
    " lm_magno<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Magnoliopsida\") \n",
    "   lm_magno<- lm(lm_magno$Prob_80 ~ lm_magno$Proportion)\n",
    "\n",
    " \n",
    " lm_MPB<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Microphytobenthos\")\n",
    "    lm_MPB<- lm(lm_MPB$Prob_80 ~ lm_MPB$Proportion)\n",
    "    \n",
    "lm_Rodo<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Rodophyceae\")\n",
    "    lm_Rodo<- lm(lm_Rodo$Prob_80 ~ lm_Rodo$Proportion)\n",
    " \n",
    " ########### ALPHA ###############\n",
    "alpha_sed<-data.frame(prop = seq(0,1,  0.01)) %>% \n",
    "  mutate(alpha = lm_sediment$coefficients[[2]]*prop+lm_sediment$coefficients[[1]],\n",
    "         ID = c(1:nrow(.)),\n",
    "         xmin = seq(0,1,  0.01),\n",
    "         xmax = seq(0.01,1.01,0.01),\n",
    "         ymin = rep(0,nrow(.)),\n",
    "         ymax = rep(1,nrow(.))) %>% \n",
    "  pivot_longer(c(xmin,xmax),names_to = \"name_x\", values_to = \"x\") %>% \n",
    "  pivot_longer(c(ymax,ymin), names_to = \"name_y\",values_to = \"y\") %>% \n",
    "  mutate(y = rep(c(0,1,1,0),nrow(.)/4)) %>% \n",
    "  dplyr::select(-c(name_x,name_y)) %>% \n",
    "  mutate(alpha = case_when(alpha > 1 ~ 1,\n",
    "                           TRUE ~ alpha),\n",
    "         alpha = 1-alpha)%>% \n",
    "  mutate(alpha = (alpha - min(alpha))/ (max(alpha) - min(alpha)))\n",
    "\n",
    "alpha_green<-data.frame(prop = seq(0,1,  0.01)) %>% \n",
    "  mutate(alpha = lm_green$coefficients[[2]]*prop+lm_green$coefficients[[1]],\n",
    "         ID = c(1:nrow(.)),\n",
    "         xmin = seq(0,1,  0.01),\n",
    "         xmax = seq(0.01,1.01,0.01),\n",
    "         ymin = rep(0,nrow(.)),\n",
    "         ymax = rep(1,nrow(.))) %>% \n",
    "  pivot_longer(c(xmin,xmax),names_to = \"name_x\", values_to = \"x\") %>% \n",
    "  pivot_longer(c(ymax,ymin), names_to = \"name_y\",values_to = \"y\") %>% \n",
    "  mutate(y = rep(c(0,1,1,0),nrow(.)/4)) %>% \n",
    "  dplyr::select(-c(name_x,name_y)) %>% \n",
    "  mutate(alpha = case_when(alpha > 1 ~ 1,\n",
    "                           TRUE ~ alpha),\n",
    "         alpha = 1-alpha)%>% \n",
    "  mutate(alpha = (alpha - min(alpha))/ (max(alpha) - min(alpha)))\n",
    "\n",
    "alpha_magno<-data.frame(prop = seq(0,1,  0.01)) %>% \n",
    "  mutate(alpha = lm_magno$coefficients[[2]]*prop+lm_magno$coefficients[[1]],\n",
    "         ID = c(1:nrow(.)),\n",
    "         xmin = seq(0,1,  0.01),\n",
    "         xmax = seq(0.01,1.01,0.01),\n",
    "         ymin = rep(0,nrow(.)),\n",
    "         ymax = rep(1,nrow(.))) %>% \n",
    "  pivot_longer(c(xmin,xmax),names_to = \"name_x\", values_to = \"x\") %>% \n",
    "  pivot_longer(c(ymax,ymin), names_to = \"name_y\",values_to = \"y\") %>% \n",
    "  mutate(y = rep(c(0,1,1,0),nrow(.)/4)) %>% \n",
    "  dplyr::select(-c(name_x,name_y)) %>% \n",
    "  mutate(alpha = case_when(alpha > 1 ~ 1,\n",
    "                           TRUE ~ alpha),\n",
    "         alpha = 1-alpha)%>% \n",
    "  mutate(alpha = (alpha - min(alpha))/ (max(alpha) - min(alpha)))\n",
    "\n",
    "alpha_MPB<-data.frame(prop = seq(0,1,  0.01)) %>% \n",
    "  mutate(alpha = lm_MPB$coefficients[[2]]*prop+lm_MPB$coefficients[[1]],\n",
    "         ID = c(1:nrow(.)),\n",
    "         xmin = seq(0,1,  0.01),\n",
    "         xmax = seq(0.01,1.01,0.01),\n",
    "         ymin = rep(0,nrow(.)),\n",
    "         ymax = rep(1,nrow(.))) %>% \n",
    "  pivot_longer(c(xmin,xmax),names_to = \"name_x\", values_to = \"x\") %>% \n",
    "  pivot_longer(c(ymax,ymin), names_to = \"name_y\",values_to = \"y\") %>% \n",
    "  mutate(y = rep(c(0,1,1,0),nrow(.)/4)) %>% \n",
    "  dplyr::select(-c(name_x,name_y)) %>% \n",
    "  mutate(alpha = case_when(alpha > 1 ~ 1,\n",
    "                           TRUE ~ alpha),\n",
    "         alpha = 1-alpha)%>% \n",
    "  mutate(alpha = (alpha - min(alpha))/ (max(alpha) - min(alpha)))\n",
    "\n",
    "alpha_Rodo<-data.frame(prop = seq(0,1,  0.01)) %>% \n",
    "  mutate(alpha = lm_Rodo$coefficients[[2]]*prop+lm_Rodo$coefficients[[1]],\n",
    "         ID = c(1:nrow(.)),\n",
    "         xmin = seq(0,1,  0.01),\n",
    "         xmax = seq(0.01,1.01,0.01),\n",
    "         ymin = rep(0,nrow(.)),\n",
    "         ymax = rep(1,nrow(.))) %>% \n",
    "  pivot_longer(c(xmin,xmax),names_to = \"name_x\", values_to = \"x\") %>% \n",
    "  pivot_longer(c(ymax,ymin), names_to = \"name_y\",values_to = \"y\") %>% \n",
    "  mutate(y = rep(c(0,1,1,0),nrow(.)/4)) %>% \n",
    "  dplyr::select(-c(name_x,name_y)) %>% \n",
    "  mutate(alpha = case_when(alpha > 1 ~ 1,\n",
    "                           TRUE ~ alpha),\n",
    "         alpha = 1-alpha)%>% \n",
    "  mutate(alpha = (alpha - min(alpha))/ (max(alpha) - min(alpha)))\n",
    "\n",
    "# alpha_xantho<-data.frame(prop = seq(0,1,  0.01)) %>% \n",
    "#   mutate(alpha = lm_xantho$coefficients[[2]]*prop+lm_xantho$coefficients[[1]],\n",
    "#          ID = c(1:nrow(.)),\n",
    "#          xmin = seq(0,1,  0.01),\n",
    "#          xmax = seq(0.01,1.01,0.01),\n",
    "#          ymin = rep(0,nrow(.)),\n",
    "#          ymax = rep(1,nrow(.))) %>% \n",
    "#   pivot_longer(c(xmin,xmax),names_to = \"name_x\", values_to = \"x\") %>% \n",
    "#   pivot_longer(c(ymax,ymin), names_to = \"name_y\",values_to = \"y\") %>% \n",
    "#   mutate(y = rep(c(0,1,1,0),nrow(.)/4)) %>% \n",
    "#   dplyr::select(-c(name_x,name_y)) %>% \n",
    "#   mutate(alpha = case_when(alpha > 1 ~ 1,\n",
    "#                            TRUE ~ alpha),\n",
    "#          alpha = 1-alpha) %>% \n",
    "#   mutate(alpha = (alpha - min(alpha))/ (max(alpha) - min(alpha)))\n",
    "\n",
    "#################\n",
    "\n",
    "sediment<-output_80_filtered %>% \n",
    "  dplyr::filter(True == \"Bare Sediment\") %>% \n",
    "  mutate(Class = case_when(Class != True ~ \"Other\",\n",
    "                            TRUE ~ Class)) %>% \n",
    "  group_by(ID_80,Class) %>% \n",
    "  reframe(Prob_80 = unique(Prob_80),\n",
    "          True = unique(True),\n",
    "          Class = unique(Class),\n",
    "          Proportion = sum(Proportion)) %>% \n",
    "  mutate(Class = as.factor(Class),\n",
    "         Class = forcats::fct_relevel(Class,\"Bare Sediment\", after = Inf)) %>% \n",
    "  ggplot()+\n",
    "  geom_density(mapping = aes(x = Proportion , y = after_stat(ndensity) , fill = Class),alpha = 0.8,show.legend = F,position = \"fill\",adjust = 3)+\n",
    "  # facet_wrap(~ True)+\n",
    "  scale_fill_manual(values=c(\"gray85\",\"#70543e\"))+\n",
    "  # scale_alpha_manual(values = c(0.5,1))+\n",
    "  theme_Bede()+\n",
    "  # geom_polygon(data = alpha_sed, mapping = aes(x=x, y = y, alpha = alpha, group = ID),alpha = alpha_sed$alpha,color=NA, fill = \"white\")+\n",
    "  geom_vline(xintercept = (lm_sediment$coefficients[[2]]*85+lm_sediment$coefficients[[1]])*100, linetype = \"dashed\",linewidth = 2)+\n",
    "  ggtitle(\"Bare Sediment\")+\n",
    "  # annotate(geom=\"text\",x = lm_sediment$coefficients[[2]]*0.85+lm_sediment$coefficients[[1]]+0.05, y = 0.85, label = \"p = 0.85\", angle = - 90, size = 7)+\n",
    "  theme(axis.text.x = element_text(size = 15),\n",
    "        axis.text.y = element_text(size = 15)\n",
    "        # axis.title.y = element_blank(),\n",
    "        # axis.title.x = element_text(size = 20)\n",
    "        )\n",
    "\n",
    "green<-output_80_filtered %>% \n",
    "  dplyr::filter(True == \"Green Algae\") %>% \n",
    "  mutate(Class = case_when(Class != True ~ \"Other\",\n",
    "                            TRUE ~ Class)) %>% \n",
    "  group_by(ID_80,Class) %>% \n",
    "  reframe(Prob_80 = unique(Prob_80),\n",
    "          True = unique(True),\n",
    "          Class = unique(Class),\n",
    "          Proportion = sum(Proportion)) %>%\n",
    "  mutate(Class = as.factor(Class),\n",
    "         Class = forcats::fct_relevel(Class,\"Green Algae\", after = Inf)) %>% \n",
    "  ggplot()+\n",
    "  geom_density(mapping = aes(x = Proportion , y = after_stat(ndensity) , fill = Class),alpha = 0.8,show.legend = F,position = \"fill\",adjust = 3)+\n",
    "  # facet_wrap(~ True)+\n",
    "  scale_fill_manual(values=c(\"gray85\", \"#b3ff1a\"))+\n",
    "  # scale_alpha_manual(values = c(0.5,1))+\n",
    "  theme_Bede()+\n",
    "  # geom_polygon(data = alpha_green, mapping = aes(x=x, y = y, alpha = alpha, group = ID),alpha = alpha_green$alpha,color=NA, fill = \"white\")+\n",
    "  geom_vline(xintercept = (lm_green$coefficients[[2]]*85+lm_green$coefficients[[1]])*100, linetype = \"dashed\",linewidth = 2)+\n",
    "  ggtitle(\"Green Algae\")+\n",
    "  # annotate(geom=\"text\",x = lm_green$coefficients[[2]]*0.85+lm_green$coefficients[[1]]+0.05, y = 0.85, label = \"p = 0.85\", angle = - 90, size = 7)+\n",
    "  theme(axis.text.x = element_text(size = 15),\n",
    "        axis.text.y = element_text(size = 15)\n",
    "        # axis.title.y = element_blank(),\n",
    "        # axis.title.x = element_text(size = 20)\n",
    "        )\n",
    "\n",
    "magno<-output_80_filtered %>% \n",
    "  dplyr::filter(True == \"Magnoliopsida\") %>% \n",
    "  mutate(Class = case_when(Class != True ~ \"Other\",\n",
    "                            TRUE ~ Class)) %>% \n",
    "  group_by(ID_80,Class) %>% \n",
    "  reframe(Prob_80 = unique(Prob_80),\n",
    "          True = unique(True),\n",
    "          Class = unique(Class),\n",
    "          Proportion = sum(Proportion)) %>%\n",
    "  mutate(Class = as.factor(Class),\n",
    "         Class = forcats::fct_relevel(Class,\"Magnoliopsida\", after = Inf)) %>%\n",
    "  ggplot()+\n",
    "  geom_density(mapping = aes(x = Proportion , y = after_stat(ndensity) , fill = Class),alpha = 0.8,show.legend = F, position = \"fill\",adjust = 3)+\n",
    "  # facet_wrap(~ True)+\n",
    "  scale_fill_manual(values=c(\"gray85\",\"#389350\"))+\n",
    "  # scale_alpha_manual(values = c(0.5,1))+\n",
    "  theme_Bede()+\n",
    "  # geom_polygon(data = alpha_magno, mapping = aes(x=x, y = y, alpha = alpha, group = ID),alpha = alpha_magno$alpha,color=NA, fill = \"white\")+\n",
    "  geom_vline(xintercept = (lm_magno$coefficients[[2]]*85+lm_magno$coefficients[[1]])*100, linetype = \"dashed\",linewidth = 2)+\n",
    "  ggtitle(\"Magnoliopsida\")+\n",
    "  # annotate(geom=\"text\",x = lm_magno$coefficients[[2]]*0.85+lm_magno$coefficients[[1]]+0.05, y = 0.85, label = \"p = 0.85\", angle = - 90, size = 7)+\n",
    "  theme(axis.text.x = element_text(size = 15),\n",
    "        axis.text.y = element_text(size = 15)\n",
    "        # axis.title.y = element_blank(),\n",
    "        # axis.title.x = element_text(size = 20)\n",
    "        )\n",
    "\n",
    "\n",
    "MPB<-output_80_filtered %>% \n",
    "  dplyr::filter(True == \"Microphytobenthos\") %>% \n",
    "  mutate(Class = case_when(Class != True ~ \"Other\",\n",
    "                            TRUE ~ Class)) %>% \n",
    "  group_by(ID_80,Class) %>% \n",
    "  reframe(Prob_80 = unique(Prob_80),\n",
    "          True = unique(True),\n",
    "          Class = unique(Class),\n",
    "          Proportion = sum(Proportion)) %>%\n",
    "  mutate(Class = as.factor(Class),\n",
    "         Class = forcats::fct_relevel(Class,\"Microphytobenthos\", after = Inf)) %>%\n",
    "  ggplot()+\n",
    "  geom_density(mapping = aes(x = Proportion , y = after_stat(ndensity) , fill = Class),alpha = 0.8,show.legend = F,position = \"fill\",adjust = 3)+\n",
    "  # facet_wrap(~ True)+\n",
    "  scale_fill_manual(values=c(\"gray85\",\"#DAA520\"))+\n",
    "  # scale_alpha_manual(values = c(0.5,1))+\n",
    "  theme_Bede()+\n",
    "  # geom_polygon(data = alpha_MPB, mapping = aes(x=x, y = y, alpha = alpha, group = ID),alpha = alpha_MPB$alpha,color=NA, fill = \"white\")+\n",
    "  geom_vline(xintercept = (lm_MPB$coefficients[[2]]*85+lm_MPB$coefficients[[1]])*100, linetype = \"dashed\",linewidth = 2)+\n",
    "  ggtitle(\"Microphytobenthos\")+\n",
    "  # annotate(geom=\"text\",x = lm_MPB$coefficients[[2]]*0.85+lm_MPB$coefficients[[1]]+0.05, y = 0.85, label = \"p = 0.85\", angle = - 90, size = 7)+\n",
    "  theme(axis.text.x = element_text(size = 15),\n",
    "        axis.text.y = element_text(size = 15)\n",
    "        # axis.title.y = element_blank(),\n",
    "        # axis.title.x = element_text(size = 20)\n",
    "        )\n",
    "\n",
    "rodo<-output_80_filtered %>% \n",
    "  dplyr::filter(True == \"Rodophyceae\") %>% \n",
    "  mutate(Class = case_when(Class != True ~ \"Others Classes\",\n",
    "                           Class == \"Rodophyceae\" ~ \"Rhodophyceae\",\n",
    "                            TRUE ~ Class),\n",
    "         True = case_when(True == \"Rodophyceae\" ~ \"Rhodophyceae\",\n",
    "                          TRUE ~ True)) %>% \n",
    "  group_by(ID_80,Class) %>% \n",
    "  reframe(Prob_80 = unique(Prob_80),\n",
    "          True = unique(True),\n",
    "          Class = unique(Class),\n",
    "          Proportion = sum(Proportion)) %>%\n",
    "  mutate(Class = factor(Class, c(\"Others Classes\",\"Rhodophyceae\",\"Bacillariophyceae\", \"Magnoliopsida\",\"Chlorophyceae\",\"Phaeophyceae\"))) %>% \n",
    "         # Class = forcats::fct_relevel(Class,\"Rodophyceae\", after = Inf)) %>%\n",
    "  ggplot()+\n",
    "  geom_density(mapping = aes(x = Proportion , y = after_stat(ndensity) , fill = Class),alpha = 0.8,position = \"fill\",adjust = 3)+\n",
    "  # facet_wrap(~ True)+\n",
    "  scale_fill_manual(\"Class\",\n",
    "                    labels = c(\"Others Classes\",\"Rhodophyceae\",\"Bacillariophyceae\", \"Magnoliopsida\",\"Chlorophyceae\",\"Phaeophyceae\"),\n",
    "                    values=c(\"gray85\",\"#b3002d\",\"#DAA520\",\"#389350\",\"#b3ff1a\",\"#70543e\"),\n",
    "                    drop = F)+\n",
    "  # scale_alpha_manual(values = c(0.5,1))+\n",
    "  theme_Bede()+\n",
    "  # geom_polygon(data = alpha_Rodo, mapping = aes(x=x, y = y, alpha = alpha, group = ID),alpha = alpha_Rodo$alpha,color=NA, fill = \"white\")+\n",
    "  geom_vline(xintercept = (lm_Rodo$coefficients[[2]]*85+lm_Rodo$coefficients[[1]])*100, linetype = \"dashed\",linewidth = 2)+\n",
    "  ggtitle(\"Rhodophyceae\")+\n",
    "  # annotate(geom=\"text\",x = lm_Rodo$coefficients[[2]]*0.85+lm_Rodo$coefficients[[1]]+0.05, y = 0.85, label = \"p = 0.85\", angle = - 90, size = 7)+\n",
    "  theme(axis.text.x = element_text(size = 15),\n",
    "        axis.text.y = element_text(size = 15),\n",
    "        # axis.title.y = element_blank(),\n",
    "        # axis.title.x = element_text(size = 20),\n",
    "        legend.text = element_text(size = 15),\n",
    "        legend.key.size = unit(1,\"cm\"),\n",
    "        legend.title = element_blank())\n",
    "\n",
    "# xantho<-output_80_filtered %>% \n",
    "#   dplyr::filter(True == \"Xanthophyceae\") %>% \n",
    "#   ggplot()+\n",
    "#   geom_density(mapping = aes(x = Proportion , y = after_stat(ndensity) , fill = Class),position = \"fill\",adjust = 3)+\n",
    "#   # facet_wrap(~ True)+\n",
    "#   scale_fill_manual(values=c(\"#70543e\", \"#b3ff1a\", \"#389350\", \"#DAA520\",\"#b3002d\", \"#cccc00\"))+\n",
    "#   # scale_alpha_manual(values = c(0.5,1))+\n",
    "#   theme_Bede()+\n",
    "#   geom_polygon(data = alpha_xantho, mapping = aes(x=x, y = y, alpha = alpha, group = ID),alpha = alpha_xantho$alpha,color=NA, fill = \"white\")+\n",
    "#   geom_vline(xintercept = lm_xantho$coefficients[[2]]*0.85+lm_xantho$coefficients[[1]])+\n",
    "#   ggtitle(\"Xanthophyceae\")+\n",
    "#   annotate(geom=\"text\",x = lm_xantho$coefficients[[2]]*0.85+lm_xantho$coefficients[[1]]+0.05, y = 0.85, label = \"p = 0.85\", angle = - 90, size = 7)+\n",
    "#   theme(axis.text.x = element_text(size = 15, angle = 45, hjust = 0.7),\n",
    "#         axis.text.y = element_text(size = 15),\n",
    "#         axis.title.y = element_blank(),\n",
    "#         axis.title.x = element_text(size = 20))\n",
    "\n",
    "plot<-(sediment + green) /\n",
    "(MPB + rodo)+\n",
    "magno+\n",
    "  patchwork::plot_layout(guides = \"collect\") &\n",
    "  theme(legend.position = \"bottom\",\n",
    "        plot.title = element_text(size=25, hjust = 0.55),\n",
    "        axis.text.x = element_text(size = 15, hjust = 0.8),\n",
    "        axis.title = element_text(size=20),\n",
    "        axis.text.y = element_text(size = 15),\n",
    "        panel.grid=element_blank())&\n",
    "  scale_x_continuous(breaks = c(25,50,75,100),limits= c(0,100.01),expand = c(0, 0)) &\n",
    "  scale_y_continuous(expand = c(0, 0)) &\n",
    "  xlab(\"Percent Cover (%)\") & \n",
    "  ylab(\"Density\")\n",
    "  \n",
    "ggsave(\"Figures/Upscaling/density_vs_Proportion.png\",plot , width = 833*4, height = 1118*4, units = \"px\")"
   ],
   "id": "8d32d919-e26c-40fe-8f09-021eb5ad7ff3"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_80_filtered_same<-read.csv(\"./Output/output_80_filtered_same_Upscaling.csv\")%>% \n",
    "    dplyr::mutate(Proportion = Proportion *100) %>% \n",
    "    dplyr::select(-X) %>% \n",
    "    as_tibble()\n",
    "\n",
    "################# LM ####################\n",
    "lm_sediment<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Bare Sediment\") \n",
    "\n",
    " lm_sediment<- lm(lm_sediment$Prob_80 ~ lm_sediment$Proportion)\n",
    " \n",
    " lm_green<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Green Algae\") \n",
    "  lm_green<- lm(lm_green$Prob_80 ~ lm_green$Proportion)\n",
    " \n",
    " lm_magno<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Magnoliopsida\") \n",
    "   lm_magno<- lm(lm_magno$Prob_80 ~ lm_magno$Proportion)\n",
    "\n",
    " \n",
    " lm_MPB<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Microphytobenthos\")\n",
    "    lm_MPB<- lm(lm_MPB$Prob_80 ~ lm_MPB$Proportion)\n",
    "    \n",
    "lm_Rodo<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Rodophyceae\")\n",
    "    lm_Rodo<- lm(lm_Rodo$Prob_80 ~ lm_Rodo$Proportion)"
   ],
   "id": "940adc15-2385-48a5-95cc-9665f207d5af"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"Figures/Low_res/Upscaling/density_vs_Proportion.png\")"
   ],
   "id": "cell-fig-upscaling"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the very high resolution low altitude flight (8 mm pixels), we determined the minimal percent cover required to correctly classify a given class within the corresponding high altitude flight (<a href=\"#fig-upscaling\" class=\"quarto-xref\">Figure 11</a>). When the percent cover of a given class was 100 %, coarser high-flight pixels are well classified for all the classes except for Bare Sediment, which was only well classified 80% of the time. This phenomenon is likely attributed to the time gap between the two flights, allowing for microphytobenthos migration to surface during low tide, consequently altering the model’s classification from bare sediment to Bacillariophyceae. A percent cover of at least 80% was sufficient to have all the pixels of high altitude flights correctly classified, with the exception of Magnoliopsida that required a higher percent cover (\\>90 %) to be well classified. Concerning the probability of each class, the highest percent cover was needed to confidently predict Bacillariophyceae. To predict Chlorophyceae with a confidence of 0.85, a percent cover of 93 % was needed, 90 % for Magnoliopsida, 92 % for Rhodophyceae and 97 % for Bacillariophyceae.\n",
    "\n",
    "# 4. Discussion\n",
    "\n",
    "## 4.1 Vegetation Discrimination"
   ],
   "id": "6f7e2945-1040-4b93-9835-48e7985ee13a"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pigment<-\n",
    "  data.frame(\n",
    "    Classes=c(\"Magnoliopsida\",\n",
    "           \"Chlorophyceae\",\n",
    "           \"Bacillariophyceae\",\n",
    "           \"Phaeophyceae\",\n",
    "           \"Rhodophyceae\"),\n",
    "    chlb=c(1,1,0,0,0),\n",
    "    chlc=c(0,0,1,1,0),\n",
    "    fuco=c(0,0,1,1,0),\n",
    "    zea=c(1,1,0,1,0),\n",
    "    diad=c(0,0,1,0,0),\n",
    "    lutein=c(1,1,0,0,0),\n",
    "    Neox=c(1,1,0,0,0),\n",
    "    PE=c(0,0,0,0,1),\n",
    "    PC=c(0,0,0,0,1))%>%\n",
    "  select(-Classes) %>% \n",
    "  as.matrix() %>% \n",
    "  as.array() %>% \n",
    "  rast()\n",
    "\n",
    "ext(df_pigment)<-c(0,9,0,5)\n",
    "\n",
    "\n",
    "y_axis_text_pigment<-data.frame(x = rep(-0.05,5),\n",
    "                                y = c(0.5,1.5,2.5,3.5,4.5), \n",
    "                                lab = c(\"Rhodophyceae\",\n",
    "                                    \"Phaeophyceae\",\n",
    "                                    \"Bacillariophy.\",\n",
    "                                    \"Chlorophyceae\",\n",
    "                                    \"Magnoliopsida\")\n",
    "                                )\n",
    "x_axis_text_pigment<-data.frame(y = rep(5.5,9),\n",
    "                                x = c(0.5,1.5,2.5,3.5,4.5,5.5,6.5,7.5,8.5),\n",
    "                                lab = c(\"Chlb\",\"Chlc\",\"Fuco\",\"Zea\",\"Diad\",\"Lut\",\"Neo\",\"PE\",\"PC\")\n",
    "                                )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "min_x_line = -1.5\n",
    "max_y_line = 6\n",
    "\n",
    "line_grid<-data.frame(xmin=NA,\n",
    "                      xmax=NA,\n",
    "                      ymin=NA,\n",
    "                      ymax=NA,\n",
    "                      group=NA)\n",
    "\n",
    "for(i in 0:max_y_line){\n",
    "  line_grid[i+1,1]=min_x_line\n",
    "  line_grid[i+1,2]=9\n",
    "  line_grid[i+1,3]=i\n",
    "  line_grid[i+1,4]=i\n",
    "  line_grid[i+1,5]=i+1\n",
    "}\n",
    "nrowgrid=nrow(line_grid)\n",
    "for(i in 0:9){\n",
    "  line_grid[i+1+nrowgrid,1]=i\n",
    "  line_grid[i+1+nrowgrid,2]=i\n",
    "  line_grid[i+1+nrowgrid,3]=0\n",
    "  line_grid[i+1+nrowgrid,4]=max_y_line\n",
    "  line_grid[i+1+nrowgrid,5]=i+1+nrowgrid\n",
    "  \n",
    "}\n",
    "\n",
    " line_grid[nrow(line_grid)+1,1] = min_x_line\n",
    " line_grid[nrow(line_grid),2] = min_x_line\n",
    " line_grid[nrow(line_grid),3] = 0\n",
    " line_grid[nrow(line_grid),4] = max_y_line\n",
    " line_grid[nrow(line_grid),5] = nrow(line_grid)\n",
    "\n",
    " \n",
    " line_grid<- line_grid %>% \n",
    "   rbind(c(-1.5,9,-0.6,-0.6,19))%>% \n",
    "   rbind(c(-1.5,-1.5,-0.6,0,20))%>% \n",
    "   rbind(c(0,0,-0.6,0,21))%>% \n",
    "   rbind(c(1,1,-0.6,0,22))%>% \n",
    "   rbind(c(2,2,-0.6,0,23))%>% \n",
    "   rbind(c(3,3,-0.6,0,24))%>% \n",
    "   rbind(c(4,4,-0.6,0,25))%>% \n",
    "   rbind(c(5,5,-0.6,0,26))%>% \n",
    "   rbind(c(6,6,-0.6,0,27))%>% \n",
    "   rbind(c(7,7,-0.6,0,28))%>% \n",
    "   rbind(c(8,8,-0.6,0,29))%>% \n",
    "   rbind(c(9,9,-0.6,0,30))\n",
    " \n",
    " \n",
    " Absorption_wv<- data.frame(x = c(0.5,1.5,2.5,3.5,4.5,5.5,6.5,7.5,8.5),\n",
    "                            y = rep(-0.3,9),\n",
    "                            lab = c(\"650\",\"636\",\"550\",\"489\",\"496\",\"490\",\"450\",\"566\",\"615\"))\n",
    " \n",
    " \n",
    "rast_pigment<-ggplot()+\n",
    "  geom_spatraster(data = (df_pigment),\n",
    "                  # show.legend = F,\n",
    "                  na.rm = TRUE\n",
    "                  )+\n",
    "  coord_fixed(ratio = 0.7)+\n",
    "  scale_fill_gradientn(\n",
    "    # colours = c(\"black\",\"firebrick4\",\"goldenrod\",\"yellow\",\"chartreuse3\"),\n",
    "    # colours = c(\"khaki1\",\"goldenrod\",\"firebrick3\",\"chartreuse4\"),\n",
    "    colours = c(\"firebrick4\",\"orange\",\"#007d00\"),\n",
    "    # colours = c(\"#7d0000\",\"#00b300\",\"#00b300\",\"#00b300\",\"#007d00\"),\n",
    "    na.value = \"white\",\n",
    "    labels = c(\"Absence\",\"no se\",\"Presence\"),\n",
    "    breaks = c(0,0.5,1))+\n",
    "  geom_text(\n",
    "    data = y_axis_text_pigment,\n",
    "    aes(x = x, y = y, label = lab),\n",
    "    # fill = \"#ffffff99\",\n",
    "    hjust = 1,\n",
    "    size = 6,\n",
    "    # angle = 20\n",
    "    )+\n",
    "  geom_text(\n",
    "    data = data.frame(x = -0.05, y = -0.3, lab = \"Absorption (nm)\"),\n",
    "    aes(x = x, y = y, label = lab),\n",
    "    # fill = \"#ffffff99\",\n",
    "    hjust = 1,\n",
    "    size = 5,\n",
    "    # angle = 20\n",
    "    )+\n",
    "  geom_text(\n",
    "    data = Absorption_wv,\n",
    "    aes(x = x, y = y, label = lab),\n",
    "    # fill = \"#ffffff99\",\n",
    "    hjust = 0.5,\n",
    "    size = 6,\n",
    "    # angle = 20\n",
    "    )+\n",
    "  geom_text(\n",
    "    data = x_axis_text_pigment,\n",
    "    aes(x = x, y = y, label = lab),\n",
    "    # fill = \"#ffffff99\",\n",
    "    hjust = 0.5,\n",
    "    size = 7,\n",
    "    # angle = 20\n",
    "    )+\n",
    "  geom_segment(data = line_grid, \n",
    "               aes(x = xmin, xend = xmax, y = ymin, yend = ymax, group = group),\n",
    "               linewidth = 1)+\n",
    "  xlim(c(-2,9))+\n",
    "  ylim(c(-1,6.1))+\n",
    "  theme_Bede()+\n",
    "  theme(\n",
    "  # axis.title.x = element_text(size = 35),\n",
    "        # axis.title.y = element_text(size = 35),\n",
    "        axis.title.y = element_blank(),\n",
    "        axis.title.x = element_blank(),\n",
    "\n",
    "        # axis.text.x = element_text(size = 30),\n",
    "        axis.text.x = element_blank(),\n",
    "        # axis.text.y = element_text(size = 25),\n",
    "         axis.text.y = element_blank(),\n",
    "\n",
    "        axis.ticks.x = element_blank(),\n",
    "        axis.ticks.y = element_blank(),\n",
    "        panel.border = element_blank(),\n",
    "        axis.line=element_blank(),\n",
    "        # legend.position = \"top\",\n",
    "        legend.position = \"none\",\n",
    "        legend.direction=\"horizontal\",\n",
    "        legend.key.width = unit(2,\"cm\"),\n",
    "        legend.text = element_text(size = 20),\n",
    "        legend.title = element_text(size = 20, vjust = 2),\n",
    "        legend.title.align = 0.5,\n",
    "        legend.justification = \"center\")\n",
    "\n",
    "# rast_pigment\n",
    "\n",
    "ggsave(\"./Figures/Disc_Pigment_Table.png\",rast_pigment, width= 1066*4, height=516*4, units = \"px\")"
   ],
   "id": "e91c9d74-238a-4078-8105-b5083a24085e"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Disc_Pigment_Table.png\")"
   ],
   "id": "cell-fig-Pigm"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The primary objective of this study was to develop a method for the accurate classification of macrophytes on intertidal mudflats, specifically focusing on distinguishing between Chlorophyceae (green macroalgae) and marine Magnoliopsida (seagrasses) using multispectral drone data. The ability to differentiate between various types of vegetation, as demonstrated in this study, plays a critical role in ecological monitoring and coastal management. By distinguishing between seagrasses and macroalgae, our approach facilitates targeted conservation strategies, enabling more effective preservation and restoration efforts in coastal ecosystems. The discrimination of seagrasses from green macroalgae presents significant challenges ([Oiry and Barillé, 2021](#ref-oiry2021using) ; [Bannari et al., 2022](#ref-bannari2022) ; [Veettil et al., 2020](#ref-veettil2020opportunities)), primarily due to the similarities on their spectal signatures in the visible region of the eletromagnetic spectra due to shared pigment , <a href=\"#fig-Pigm\" class=\"quarto-xref\">Figure 12</a>, but also by the frequent spatial mixing of these macrophytes. Besides chlorophyll-a that is common to all vegetation types, these two green macrophytes have in common other photosynthetic pigments such as chlorophyll-b, and accessory carotenoid pigments such as zeaxantin, lutein and neoxanthin (<a href=\"#fig-Pigm\" class=\"quarto-xref\">Figure 12</a>). This spectral overlap is further complicated by the intermingling of seagrasses and green macroalgae in the same spatial locations, posing a significant challenge for remote sensing aimed at accurately mapping and monitoring coastal ecosystems. Our study addresses these complexities by using high-resolution drone imagery with 10 spectral bands. Davies et al. ([2023a](#ref-Davies2023)) have shown that having at least eight spectral bands ranging between 500 nm to 850 nm including a green band at 530 nm and another one in the RedEdge region at 730 nm was crucial to accurately discriminate amoung green macrophytes.\n",
    "\n",
    "``` r\n",
    "knitr::include_graphics(\"./Figures/Low_res/Validation/ConfusionMatrixGreen.png\")\n",
    "```\n",
    "\n",
    "<figure id=\"fig-ValidationGreen\">\n",
    "<img src=\"attachment:./Figures/Low_res/Validation/ConfusionMatrixGreen.png\" />\n",
    "<figcaption>Figure 13: Sample of <a href=\"#fig-Validation\" class=\"quarto-xref\">Figure 9</a> focusing on green macrophytes. The labels inside the matrix indicate the number of pixels.</figcaption>\n",
    "</figure>\n",
    "\n",
    "Meeting these two criteria, the Micasense RedEdge-MX DUAL camera used in this study, enabled the classifier to achieve 97% of accuracy between these two classes (<a href=\"#fig-ValidationGreen\" class=\"quarto-xref\">Figure 13</a>). Even if the pigment composition of green macrophytes is similar, differences in the spectral shape can still be observed (<a href=\"#fig-vegetation\" class=\"quarto-xref\">Figure 2</a>). Several factors can explain these differences such has different concentration and/or pigment ratios ([Bargain et al., 2013](#ref-bargain2013seasonal)), different cellular organisation and the disposition of the plant over the sediment surface ([Beach et al., 1997](#ref-beach1997vivo) ; [Kirk, 1994](#ref-kirk1994light) ; [Hedley et al., 2018](#ref-hedley2018influence)).\n",
    "\n",
    "The VIP analysis of the Neural Network model (<a href=\"#fig-VIP\" class=\"quarto-xref\">Figure 10</a>) identified that the band at 531 nm was the most important one for accurately identifying Chlorophyceae. In fact, at this wavelength, Chlorophyceae exhibited the highest reflectance (raw and standardised) among all other classes, highlighting the difference in accessory pigments (i.e. carotenoid) between seagrasses and green macroalgae ([Repolho et al., 2017](#ref-repolho2017seagrass)).\n",
    "\n",
    "Concerning Phaeophyceae, the thick cell walls of these macroalgae ([Charrier et al., 2021](#ref-charrier2021growth)) make it more reflective in the infrared part of the spectra ([Slaton et al., 2001](#ref-Slaton2001)) whereas the presence of Fucoxanthin and Zeaxanthin result in a low reflectance in the visible region (<a href=\"#fig-Pigm\" class=\"quarto-xref\">Figure 12</a> ; <a href=\"#fig-VIP\" class=\"quarto-xref\">Figure 10</a>). These two key features have been identified by the Neural Network as the two principal predictors to accurately identify brown algae (<a href=\"#fig-VIP\" class=\"quarto-xref\">Figure 10</a>). Similarly, the presence of phycoerythrin and phycocyanin in Rhodophyceae contributes to the lowest reflectance among all classes in the spectral range of 560 to 615 nm (<a href=\"#fig-VIP\" class=\"quarto-xref\">Figure 10</a>). Indeed the band at 560 nm has been identified as important for identifying this class likely due to phycoerythrin absorption at this wavelength (**REF**). Regarding Bacillariophyceae, the VIP analysis (<a href=\"#fig-VIP\" class=\"quarto-xref\">Figure 10</a>) indicated that 475 nm was the most important predictor for this class. This is not a usual diagnostic wavelength for Bacillariophyceae which is mainly identified by the presence of fucoxanthin and chlorophyll c (<a href=\"#fig-Pigm\" class=\"quarto-xref\">Figure 12</a>), but it correspond to the absorption of diadinoxanthin an accessory pigment characteristic of this class ([Méléder et al., 2003](#ref-meleder2003spectrometric)). Furthermore, it is the vegetation with the lowest concentration of chlorphyll-a, pigment absorbing light both in the blue and the red. The transparency of Bacillariophyceae makes the reflectance of the sediment part of the overall reflectance of Bacillariophyceae, further explaining the high reflectance in the blue.\n",
    "\n",
    "## 4.2 Spectral Spatial Temporal Resolution impact on the prediction\n",
    "\n",
    "While comparing the reflectance of both altitudes (12 m and 120 m), it was observed <a href=\"#fig-CompareRef\" class=\"quarto-xref\">Figure 8</a> that there is an underestimation of the infrared part of the spectra in the high-altitude dataset. Such disparity in infrared reflectance may stem from temporal differences between the flights, possibly resulting in a slightly drier intertidal area and consequently higher infrared reflectance. This disparity poses an issue for the methodology followed in the present study, relying solely on one flight height for training. To address this issue, we employed min/max standardized reflectance spectra as predictors for the model (<a href=\"#eq-std\" class=\"quarto-xref\">Equation 1</a> ; ([Cao et al., 2017](#ref-Cao2017))). This approach allowed us to eliminate the slight reflectance difference between the flights (<a href=\"#fig-CompareRef\" class=\"quarto-xref\">Figure 8</a> B) and to focus on the shape of the spectra in the visible part of the electromagnetic spectra, where different pigmentation are associated to taxonomic dignostic features. This was a key feature in building a model that could reliably predict vegetation across geographical sites and seasons. It enabled consistent prediction of vegetation classes across variations in biomass and variability in light conditions ([Fyfe, 2003](#ref-fyfe2003spatial) ; [Costa et al., 2021](#ref-COSTA2021107018) ; [Piaser et al., 2023](#ref-piaser2023impact)).\n",
    "\n",
    "The <a href=\"#fig-upscaling\" class=\"quarto-xref\">Figure 11</a> demonstrates that at least 90 % of seagrass cover is necessary for confident prediction of its presence. This highlights a limitation of the methodology used to construct the training dataset for the model. The dataset used to train our algorithm was composed exclusively of pure pixels, which has resulted in the model’s reduced confidence when faced with lower percentages of seagrass cover. Also, intertidal seagrasses exhibit marked phenology, with varying pigment composition throughout the year([Bargain et al., 2013](#ref-bargain2013seasonal) ; [Légaré et al., 2022](#ref-legare2022remote)). Since the training dataset has been made using well developed seagrass meadows, this model may be less accurate outside of the seasonal seagrass peak of biomass. Further investigation is required to evaluate the accuracy of the method along different periods of the year.\n",
    "\n",
    "## 4.3 Application of the model\n",
    "\n",
    "Climate change, global warming, alien and invasive species development, coastal erosion, and sealevel rise are expected to continue impacting coastal ecosystems in the future ([Holon et al., 2018](#ref-holon2018predictive) ; [Marquet et al., 2024](#ref-marquet2024global); [Schibalski et al., 2022](#ref-SCHIBALSKI2022101414)). Therefore the demand for meaningful and efficient monitoring methods for these habitats has never been greater ([Muller-Karger et al., 2018](#ref-muller2018satellite) ; [Villalobos Perna et al., 2023](#ref-villalobos2023remote) ; [Oiry and Barillé, 2021](#ref-oiry2021using)). Our findings, particularly the improved discrimination of seagrass from other intertidal vegetation classes, underscore the potential of drone-based remote sensing to support diverse applications, from biodiversity conservation to climate change adaptation strategies. Employing traditional sampling methods to monitor these ecosystems is time and resource-intensive, and the findings are often difficult to up-scale. Earth observation methods can bridge this gap and meet the needs for monitoring coastal ecosystems. ([Papathanasopoulou et al., 2019](#ref-papathanasopoulou2019satellite)). The retrieval of Essential Biodiversity Variables (EBVs) and Essential Ocean Variables (EOVs) through satellite observations is increasingly common, enabling comprehensive monitoring of entire ecosystems over extended time periods ([Ratnarajah et al., 2023](#ref-ratnarajah2023monitoring) ; [Zoffoli et al., 2020](#ref-ZOFFOLI2020112020)). The significance of monitoring the coastal environment, particularly in light of the Water Framework Directive (WFD ; ([European-Commission, 2000](#ref-WFD2000))), cannot be overstated in environmental management and conservation efforts. The WFD mandates the achievement and maintenance of “good ecological status” for all European waters, which necessitates a comprehensive understanding and monitoring of aquatic ecosystems, including coastal habitats like seagrass beds ([Foden and Brazier, 2007](#ref-foden2007angiosperms) ; [Nordlund et al., 2024](#ref-nordlund2024one) ; [Zoffoli et al., 2021](#ref-Zoffoli2021)).\n",
    "\n",
    "Effective and efficient monitoring tools are essential for identifying the impacts of human activities and natural changes on these ecosystems. Drones, with their capability for high-resolution, multispectral imagery and, flight on-demand, provide a novel and powerful means to rapidly and accurately acquire ground truth data (i.e. output of the model used in the present study). These data are critical for calibrating and validating satellite remote sensing applications, thereby enhancing our capacity to monitor vast coastal areas systematically under the WFD. A perspective of this work could be applying this algorithm to satellite imagery (e.g. Sentinel-2) to evaluate if the discrimination between seagrass and green macroalgae is still operational at a coarser spatial resolution. The integration of drone technology facilitates a scalable approach to environmental surveillance, offering significant advancements in the spatial and temporal resolution of data collection. This, in turn, supports the directive’s objectives by enabling more informed and timely management decisions for the conservation and restoration of aquatic ecosystems.\n",
    "\n",
    "# 5. Conclusion\n",
    "\n",
    "The utilization of very high-resolution drone-based remote sensing coupled with machine learning techniques has proven to be an effective method for the discrimination of seagrasses from green macroalgae with a multispectral resolution sensor. Standardized reflectance was incorporated in the Neural Network model allowing for a better discrimination of spectral features related to pigment absorption in the visible. There was a striking difference for the variable of importance to discriminate Magnoliopsida from Chlorophyceae. The latter was essentially identified with the 451 nm spectral band while more spectral bands were needed to identify the former, notably 650, 560, 668, and 705 nm. As the spectral bands of the Micasense RedEdge Dual MX are very similar to those of Sentinel-2, we suggest that multispectral satellite data have the potential to perform this discrimination between green macrophytes. A sentinel-2 algorithm could be developed, using the output of this current workflow as training and validation data. The findings underscore the importance of adopting advanced remote sensing tools in ecological studies and environmental monitoring, providing a foundation for future research and policy implementation aimed at ecosystem conservation and restoration.\n",
    "\n",
    "Adade, R., Aibinu, A.M., Ekumah, B., Asaana, J., 2021. Unmanned aerial vehicle (UAV) applications in coastal zone management—a review. Environmental Monitoring and Assessment 193, 1–12.\n",
    "\n",
    "Angnuureng, D.B., Brempong, K., Jayson-Quashigah, P., Dada, O., Akuoko, S., Frimpomaa, J., Mattah, P., Almar, R., 2022. Satellite, drone and video camera multi-platform monitoring of coastal erosion at an engineered pocket beach: A showcase for coastal management at elmina bay, ghana (west africa). Regional Studies in Marine Science 53, 102437.\n",
    "\n",
    "Bannari, A., Ali, T.S., Abahussain, A., 2022. The capabilities of sentinel-MSI (2A/2B) and landsat-OLI (8/9) in seagrass and algae species differentiation using spectral reflectance. Ocean Science 18, 361–388.\n",
    "\n",
    "Bargain, A., Robin, M., Méléder, V., Rosa, P., Le Menn, E., Harin, N., Barillé, L., 2013. Seasonal spectral variation of zostera noltii and its influence on pigment-based vegetation indices. Journal of experimental marine biology and ecology 446, 86–94.\n",
    "\n",
    "Beach, K., Borgeas, H., Nishimura, N., Smith, C., 1997. In vivo absorbance spectra and the ecophysiology of reef macroalgae. Coral Reefs 16, 21–28.\n",
    "\n",
    "Brunier, G., Oiry, S., Gruet, Y., Dubois, S.F., Barillé, L., 2022. Topographic analysis of intertidal polychaete reefs (sabellaria alveolata) at a very high spatial resolution. Remote Sensing 2022, Vol. 14, Page 307 14, 307. <https://doi.org/10.3390/RS14020307>\n",
    "\n",
    "Cao, J., Thorson, J.T., Richards, R.A., Chen, Y., 2017. Spatiotemporal index standardization improves the stock assessment of northern shrimp in the gulf of maine. Canadian Journal of Fisheries and Aquatic Sciences 74, 1781–1793. <https://doi.org/10.1139/cjfas-2016-0137>\n",
    "\n",
    "Casella, E., Drechsel, J., Winter, C., Benninghoff, M., Rovere, A., 2020. Accuracy of sand beach topography surveying by drones and photogrammetry. Geo-Marine Letters 40, 255–268.\n",
    "\n",
    "Charrier, B., Boscq, S., Nelson, B.J., Läubli, N.F., 2021. Growth and labelling of cell wall components of the brown alga ectocarpus in microfluidic chips. Frontiers in Marine Science 8, 745654.\n",
    "\n",
    "Chefaoui, R.M., Duarte, C.M., Serrão, E.A., 2018. Dramatic loss of seagrass habitat under projected climate change in the mediterranean sea. Global change biology 24, 4919–4928.\n",
    "\n",
    "Coffer, M.M., Graybill, D.D., Whitman, P.J., Schaeffer, B.A., Salls, W.B., Zimmerman, R.C., Hill, V., Lebrasse, M.C., Li, J., Keith, D.J., others, 2023. Providing a framework for seagrass mapping in united states coastal ecosystems using high spatial resolution satellite imagery. Journal of Environmental Management 337, 117669.\n",
    "\n",
    "Collin, A., Dubois, S., James, D., Houet, T., 2019. Improving intertidal reef mapping using UAV surface, red edge, and near-infrared data. Drones 3, 67.\n",
    "\n",
    "Costa, V., Serôdio, J., Lillebø, A.I., Sousa, A.I., 2021. Use of hyperspectral reflectance to non-destructively estimate seagrass zostera noltei biomass. Ecological Indicators 121, 107018. https://doi.org/<https://doi.org/10.1016/j.ecolind.2020.107018>\n",
    "\n",
    "Davies, B.F.R., Gernez, P., Geraud, A., Oiry, Simon, Rosa, P., Zoffoli, M.L., Barillé, L., 2023a. Multi- and hyperspectral classification of soft-bottom intertidal vegetation using a spectral library for coastal biodiversity remote sensing. Remote Sensing of Environment 290, 113554. <https://doi.org/10.1016/j.rse.2023.113554>\n",
    "\n",
    "Davies, B.F.R., Sousa, A.I., Figueira, R., Oiry, S., Gernez, P., Barillé, L., 2023b. Benthic intertidal vegetation from the tagus estuary and aveiro lagoon. <https://doi.org/10.15468/n4ak6x>\n",
    "\n",
    "Devlin, M., Brodie, J., 2023. Nutrients and eutrophication, in: Marine Pollution–Monitoring, Management and Mitigation. Springer, pp. 75–100.\n",
    "\n",
    "Douay, F., Verpoorter, C., Duong, G., Spilmont, N., Gevaert, F., 2022. New hyperspectral procedure to discriminate intertidal macroalgae. Remote Sensing 14. <https://doi.org/10.3390/rs14020346>\n",
    "\n",
    "Duffy, J.E., Benedetti-Cecchi, L., Trinanes, J., Muller-Karger, F.E., Ambo-Rappe, R., Boström, C., Buschmann, A.H., Byrnes, J., Coles, R.G., Creed, J., others, 2019. Toward a coordinated global observing system for seagrasses and marine macroalgae. Frontiers in Marine Science 6, 317.\n",
    "\n",
    "European-Commission, 2000. <span class=\"nocase\">“DIRECTIVE 2000/60/EC OF THE EUROPEAN PARLIAMENT AND OF THE COUNCIL of 23 October 2000 establishing a framework for Community action in the field of water policy” or, in short, the EU Water Framework Directive</span>. Official Journal of the European Communities L 327, 1–72.\n",
    "\n",
    "Fairley, I., Williamson, B.J., McIlvenny, J., King, N., Masters, I., Lewis, M., Neill, S., Glasby, D., Coles, D., Powell, B., others, 2022. Drone-based large-scale particle image velocimetry applied to tidal stream energy resource assessment. Renewable Energy 196, 839–855.\n",
    "\n",
    "Foden, J., Brazier, D., 2007. Angiosperms (seagrass) within the EU water framework directive: A UK perspective. Marine Pollution Bulletin 55, 181–195.\n",
    "\n",
    "Fyfe, S., 2003. Spatial and temporal variation in spectral reflectance: Are seagrass species spectrally distinct? Limnology and Oceanography 48, 464–479.\n",
    "\n",
    "Gardner, R.C., Finlayson, C., 2018. Global wetland outlook: State of the world’s wetlands and their services to people. Stetson Law.\n",
    "\n",
    "Hedley, J.D., Mirhakak, M., Wentworth, A., Dierssen, H.M., 2018. Influence of three-dimensional coral structures on hyperspectral benthic reflectance and water-leaving reflectance. Applied Sciences 8, 2688.\n",
    "\n",
    "Holon, F., Marre, G., Parravicini, V., Mouquet, N., Bockel, T., Descamp, P., Tribot, A.-S., Boissery, P., Deter, J., 2018. A predictive model based on multiple coastal anthropogenic pressures explains the degradation status of a marine ecosystem: Implications for management and conservation. Biological Conservation 222, 125–135.\n",
    "\n",
    "Jankowska, E., Michel, L.N., Lepoint, G., Włodarska-Kowalczuk, M., 2019. Stabilizing effects of seagrass meadows on coastal water benthic food webs. Journal of Experimental Marine Biology and Ecology 510, 54–63.\n",
    "\n",
    "Joyce, K.E., Fickas, K.C., Kalamandeen, M., 2023. The unique value proposition for using drones to map coastal ecosystems. Cambridge Prisms: Coastal Futures 1, e6.\n",
    "\n",
    "Kirk, J.T., 1994. Light and photosynthesis in aquatic ecosystems. Cambridge university press.\n",
    "\n",
    "Légaré, B., Bélanger, S., Singh, R.K., Bernatchez, P., Cusson, M., 2022. Remote sensing of coastal vegetation phenology in a cold temperate intertidal system: Implications for classification of coastal habitats. Remote Sensing 14, 3000.\n",
    "\n",
    "Lin, H., Sun, T., Zhou, Y., Gu, R., Zhang, X., Yang, W., 2018. Which genes in a typical intertidal seagrass (zostera japonica) indicate copper-, lead-, and cadmium pollution? Frontiers in Plant Science 9, 1545.\n",
    "\n",
    "Los Santos, C.B. de, Krause-Jensen, D., Alcoverro, T., Marbà, N., Duarte, C.M., Van Katwijk, M.M., Pérez, M., Romero, J., Sánchez-Lizaso, J.L., Roca, G., others, 2019. Recent trend reversal for declining european seagrass meadows. Nature communications 10, 3356.\n",
    "\n",
    "Marquet, P.A., Buschmann, A.H., Corcoran, D., Dı́az, P.A., Fuentes-Castillo, T., Garreaud, R., Pliscoff, P., Salazar, A., 2024. Global change and acceleration of anthropic pressures on patagonian ecosystems, in: Conservation in Chilean Patagonia: Assessing the State of Knowledge, Opportunities, and Challenges. Springer International Publishing Cham, pp. 33–65.\n",
    "\n",
    "Méléder, V., Barillé, L., Launeau, P., Carrère, V., Rincé, Y., 2003. Spectrometric constraint in analysis of benthic diatom biomass using monospecific cultures. Remote Sensing of Environment 88, 386–400.\n",
    "\n",
    "Miloslavich, P., Bax, N.J., Simmons, S.E., Klein, E., Appeltans, W., Aburto-Oropeza, O., Garcia, M.A., Batten, S.D., Benedetti-Cecchi, L., Checkley, D.M., Chiba, S., Duffy, J.E., Dunn, D.C., Fischer, A., Gunn, J., Kudela, R., Marsac, F., Muller-Karger, F.E., Obura, D., Shin, Y.J., 2018. Essential ocean variables for global sustained observations of biodiversity and ecosystem changes. Global Change Biology 24, 2416–2433. <https://doi.org/10.1111/GCB.14108>\n",
    "\n",
    "Muller-Karger, F.E., Hestir, E., Ade, C., Turpie, K., Roberts, D.A., Siegel, D., Miller, R.J., Humm, D., Izenberg, N., Keller, M., others, 2018. Satellite sensor requirements for monitoring essential biodiversity variables of coastal ecosystems. Ecological applications 28, 749–760.\n",
    "\n",
    "Nguyen, H.M., Ralph, P.J., Marı́n-Guirao, L., Pernice, M., Procaccini, G., 2021. Seagrasses in an era of ocean warming: A review. Biological Reviews 96, 2009–2030.\n",
    "\n",
    "Nijland, W., Reshitnyk, L., Rubidge, E., 2019. Satellite remote sensing of canopy-forming kelp on a complex coastline: A novel procedure using the landsat image archive. Remote Sensing of Environment 220, 41–50.\n",
    "\n",
    "Nordlund, L.M., Unsworth, R.K., Wallner-Hahn, S., Ratnarajah, L., Beca-Carretero, P., Boikova, E., Bull, J.C., Chefaoui, R.M., Santos, C.B. de los, Gagnon, K., others, 2024. One hundred priority questions for advancing seagrass conservation in europe. Plants, People, Planet.\n",
    "\n",
    "Oh, J., Kim, D., Lee, H., 2017. Use of a drone for mapping and time series image acquisition of tidal zones. Journal of the Korean Institute of Intelligent Systems 27, 119–125.\n",
    "\n",
    "Oiry, S., Barillé, L., 2021. Using sentinel-2 satellite imagery to develop microphytobenthos-based water quality indices in estuaries. Ecological Indicators 121, 107184.\n",
    "\n",
    "Orth, R.J., Carruthers, T.J., Dennison, W.C., Duarte, C.M., Fourqurean, J.W., Heck, K.L., Hughes, A.R., Kendrick, G.A., Kenworthy, W.J., Olyarnik, S., others, 2006. A global crisis for seagrass ecosystems. Bioscience 56, 987–996.\n",
    "\n",
    "Papathanasopoulou, E., Simis, S., Alikas, K., Ansper, A., Anttila, J., Barillé, A., Barillé, L., Brando, V., Bresciani, M., Bučas, M., others, 2019. Satellite-assisted monitoring of water quality to support the implementation of the water framework directive. EOMORES white paper.\n",
    "\n",
    "Pereira, H.M., Ferrier, S., Walters, M., Geller, G.N., Jongman, R.H., Scholes, R.J., Bruford, M.W., Brummitt, N., Butchart, S.H., Cardoso, A., others, 2013. Essential biodiversity variables. Science 339, 277–278.\n",
    "\n",
    "Piaser, E., Berton, A., Bolpagni, R., Caccia, M., Castellani, M.B., Coppi, A., Dalla Vecchia, A., Gallivanone, F., Sona, G., Villa, P., 2023. Impact of radiometric variability on ultra-high resolution hyperspectral imagery over aquatic vegetation: Preliminary results. IEEE Journal of Selected Topics in Applied Earth Observations and Remote Sensing.\n",
    "\n",
    "Ralph, P., Polk, S., Moore, K., Orth, R., Smith Jr, W., 2002. Operation of the xanthophyll cycle in the seagrass zostera marina in response to variable irradiance. Journal of Experimental Marine Biology and Ecology 271, 189–207.\n",
    "\n",
    "Rasheed, M.A., Unsworth, R.K., 2011. Long-term climate-associated dynamics of a tropical seagrass meadow: Implications for the future. Marine Ecology Progress Series 422, 93–103.\n",
    "\n",
    "Ratnarajah, L., Abu-Alhaija, R., Atkinson, A., Batten, S., Bax, N.J., Bernard, K.S., Canonico, G., Cornils, A., Everett, J.D., Grigoratou, M., others, 2023. Monitoring and modelling marine zooplankton in a changing climate. Nature Communications 14, 564.\n",
    "\n",
    "Repolho, T., Duarte, B., Dionı́sio, G., Paula, J.R., Lopes, A.R., Rosa, I.C., Grilo, T.F., Caçador, I., Calado, R., Rosa, R., 2017. Seagrass ecophysiological performance under ocean warming and acidification. Scientific Reports 7, 41443.\n",
    "\n",
    "Roca, M., Dunbar, M.B., Román, A., Caballero, I., Zoffoli, M.L., Gernez, P., Navarro, G., 2022. Monitoring the marine invasive alien species rugulopteryx okamurae using unmanned aerial vehicles and satellites. Frontiers in Marine Science 9. <https://doi.org/10.3389/fmars.2022.1004012>\n",
    "\n",
    "Román, A., Tovar-Sánchez, A., Olivé, I., Navarro, G., 2021. Using a UAV-mounted multispectral camera for the monitoring of marine macrophytes. Frontiers in Marine Science 1225.\n",
    "\n",
    "Rossiter, T., Furey, T., McCarthy, T., Stengel, D.B., 2020. UAV-mounted hyperspectral mapping of intertidal macroalgae. Estuarine, Coastal and Shelf Science 242, 106789.\n",
    "\n",
    "Schibalski, A., Kleyer, M., Maier, M., Schröder, B., 2022. Spatiotemporally explicit prediction of future ecosystem service provisioning in response to climate change, sea level rise, and adaptation strategies. Ecosystem Services 54, 101414. https://doi.org/<https://doi.org/10.1016/j.ecoser.2022.101414>\n",
    "\n",
    "Slaton, M.R., Raymond Hunt Jr., E., Smith, W.K., 2001. Estimating near-infrared leaf reflectance from leaf structural characteristics. American Journal of Botany 88, 278–284. https://doi.org/<https://doi.org/10.2307/2657019>\n",
    "\n",
    "Soissons, L.M., Haanstra, E.P., Van Katwijk, M.M., Asmus, R., Auby, I., Barillé, L., Brun, F.G., Cardoso, P.G., Desroy, N., Fournier, J., others, 2018. Latitudinal patterns in european seagrass carbon reserves: Influence of seasonal fluctuations versus short-term stress and disturbance events. Frontiers in Plant Science 9, 88.\n",
    "\n",
    "Tallam, K., Nguyen, N., Ventura, J., Fricker, A., Calhoun, S., O’Leary, J., Fitzgibbons, M., Robbins, I., Walter, R.K., 2023. Application of deep learning for classification of intertidal eelgrass from drone-acquired imagery. Remote Sensing 15, 2321.\n",
    "\n",
    "Traganos, D., Reinartz, P., 2018. Mapping mediterranean seagrasses with sentinel-2 imagery. Marine Pollution Bulletin 134, 197–209. <https://doi.org/10.1016/j.marpolbul.2017.06.075>\n",
    "\n",
    "Tuya, F., Hernandez-Zerpa, H., Espino, F., Haroun, R., 2013. Drastic decadal decline of the seagrass cymodocea nodosa at gran canaria (eastern atlantic): Interactions with the green algae caulerpa prolifera. Aquatic Botany 105, 1–6.\n",
    "\n",
    "Unsworth, R.K., Cullen-Unsworth, L.C., Jones, B.L., Lilley, R.J., 2022. The planetary role of seagrass conservation. Science 377, 609–613.\n",
    "\n",
    "Veettil, B.K., Ward, R.D., Lima, M.D.A.C., Stankovic, M., Hoai, P.N., Quang, N.X., 2020. Opportunities for seagrass research derived from remote sensing: A review of current methods. Ecological Indicators 117, 106560.\n",
    "\n",
    "Villalobos Perna, P., Di Febbraro, M., Carranza, M.L., Marzialetti, F., Innangi, M., 2023. Remote sensing and invasive plants in coastal ecosystems: What we know so far and future prospects. Land 12, 341.\n",
    "\n",
    "Wang, Z., Fang, Z., Liang, J., Song, X., 2022. Assessment of global habitat suitability and risk of ocean green tides. Harmful Algae 119, 102324.\n",
    "\n",
    "Wei, P., Lu, Z., Song, J., 2015. Variable importance analysis: A comprehensive review. Reliability Engineering and System Safety 142, 399–432. https://doi.org/<https://doi.org/10.1016/j.ress.2015.05.018>\n",
    "\n",
    "Xu, S., Xu, S., Zhou, Y., Yue, S., Zhang, X., Gu, R., Zhang, Y., Qiao, Y., Liu, M., 2021. Long-term changes in the unique and largest seagrass meadows in the bohai sea (china) using satellite (1974–2019) and sonar data: Implication for conservation and restoration. Remote Sensing 13, 856.\n",
    "\n",
    "Zoffoli, M.L., Gernez, P., Godet, L., Peters, S., Oiry, S., Barillé, L., 2021. Decadal increase in the ecological status of a north-atlantic intertidal seagrass meadow observed with multi-mission satellite time-series. Ecological Indicators 130, 108033. <https://doi.org/10.1016/j.ecolind.2021.108033>\n",
    "\n",
    "Zoffoli, M.L., Gernez, P., Oiry, S., Godet, L., Dalloyau, S., Davies, B.F.R., Barillé, L., 2022. Remote sensing in seagrass ecology: Coupled dynamics between migratory herbivorous birds and intertidal meadows observed by satellite during four decades. Remote Sensing in Ecology and Conservation. <https://doi.org/10.1002/rse2.319>\n",
    "\n",
    "Zoffoli, M.L., Gernez, P., Rosa, P., Le Bris, A., Brando, V.E., Barillé, A.-L., Harin, N., Peters, S., Poser, K., Spaias, L., Peralta, G., Barillé, L., 2020. Sentinel-2 remote sensing of zostera noltei-dominated intertidal seagrass meadows. Remote Sensing of Environment 251, 112020. https://doi.org/<https://doi.org/10.1016/j.rse.2020.112020>"
   ],
   "attachments": {
    "./Figures/Low_res/Validation/ConfusionMatrixGreen.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAA3YAAAJ4CAYAAAAp0HFJAAAABGdBTUEAALGPC/xhBQAACklpQ0NQ\nc1JHQiBJRUM2MTk2Ni0yLjEAAEiJnVN3WJP3Fj7f92UPVkLY8LGXbIEAIiOsCMgQWaIQkgBhhBAS\nQMWFiApWFBURnEhVxILVCkidiOKgKLhnQYqIWotVXDjuH9yntX167+3t+9f7vOec5/zOec8PgBES\nJpHmomoAOVKFPDrYH49PSMTJvYACFUjgBCAQ5svCZwXFAADwA3l4fnSwP/wBr28AAgBw1S4kEsfh\n/4O6UCZXACCRAOAiEucLAZBSAMguVMgUAMgYALBTs2QKAJQAAGx5fEIiAKoNAOz0ST4FANipk9wX\nANiiHKkIAI0BAJkoRyQCQLsAYFWBUiwCwMIAoKxAIi4EwK4BgFm2MkcCgL0FAHaOWJAPQGAAgJlC\nLMwAIDgCAEMeE80DIEwDoDDSv+CpX3CFuEgBAMDLlc2XS9IzFLiV0Bp38vDg4iHiwmyxQmEXKRBm\nCeQinJebIxNI5wNMzgwAABr50cH+OD+Q5+bk4eZm52zv9MWi/mvwbyI+IfHf/ryMAgQAEE7P79pf\n5eXWA3DHAbB1v2upWwDaVgBo3/ldM9sJoFoK0Hr5i3k4/EAenqFQyDwdHAoLC+0lYqG9MOOLPv8z\n4W/gi372/EAe/tt68ABxmkCZrcCjg/1xYW52rlKO58sEQjFu9+cj/seFf/2OKdHiNLFcLBWK8ViJ\nuFAiTcd5uVKRRCHJleIS6X8y8R+W/QmTdw0ArIZPwE62B7XLbMB+7gECiw5Y0nYAQH7zLYwaC5EA\nEGc0Mnn3AACTv/mPQCsBAM2XpOMAALzoGFyolBdMxggAAESggSqwQQcMwRSswA6cwR28wBcCYQZE\nQAwkwDwQQgbkgBwKoRiWQRlUwDrYBLWwAxqgEZrhELTBMTgN5+ASXIHrcBcGYBiewhi8hgkEQcgI\nE2EhOogRYo7YIs4IF5mOBCJhSDSSgKQg6YgUUSLFyHKkAqlCapFdSCPyLXIUOY1cQPqQ28ggMor8\nirxHMZSBslED1AJ1QLmoHxqKxqBz0XQ0D12AlqJr0Rq0Hj2AtqKn0UvodXQAfYqOY4DRMQ5mjNlh\nXIyHRWCJWBomxxZj5Vg1Vo81Yx1YN3YVG8CeYe8IJAKLgBPsCF6EEMJsgpCQR1hMWEOoJewjtBK6\nCFcJg4Qxwicik6hPtCV6EvnEeGI6sZBYRqwm7iEeIZ4lXicOE1+TSCQOyZLkTgohJZAySQtJa0jb\nSC2kU6Q+0hBpnEwm65Btyd7kCLKArCCXkbeQD5BPkvvJw+S3FDrFiOJMCaIkUqSUEko1ZT/lBKWf\nMkKZoKpRzame1AiqiDqfWkltoHZQL1OHqRM0dZolzZsWQ8ukLaPV0JppZ2n3aC/pdLoJ3YMeRZfQ\nl9Jr6Afp5+mD9HcMDYYNg8dIYigZaxl7GacYtxkvmUymBdOXmchUMNcyG5lnmA+Yb1VYKvYqfBWR\nyhKVOpVWlX6V56pUVXNVP9V5qgtUq1UPq15WfaZGVbNQ46kJ1Bar1akdVbupNq7OUndSj1DPUV+j\nvl/9gvpjDbKGhUaghkijVGO3xhmNIRbGMmXxWELWclYD6yxrmE1iW7L57Ex2Bfsbdi97TFNDc6pm\nrGaRZp3mcc0BDsax4PA52ZxKziHODc57LQMtPy2x1mqtZq1+rTfaetq+2mLtcu0W7eva73VwnUCd\nLJ31Om0693UJuja6UbqFutt1z+o+02PreekJ9cr1Dund0Uf1bfSj9Rfq79bv0R83MDQINpAZbDE4\nY/DMkGPoa5hpuNHwhOGoEctoupHEaKPRSaMnuCbuh2fjNXgXPmasbxxirDTeZdxrPGFiaTLbpMSk\nxeS+Kc2Ua5pmutG003TMzMgs3KzYrMnsjjnVnGueYb7ZvNv8jYWlRZzFSos2i8eW2pZ8ywWWTZb3\nrJhWPlZ5VvVW16xJ1lzrLOtt1ldsUBtXmwybOpvLtqitm63Edptt3xTiFI8p0in1U27aMez87Ars\nmuwG7Tn2YfYl9m32zx3MHBId1jt0O3xydHXMdmxwvOuk4TTDqcSpw+lXZxtnoXOd8zUXpkuQyxKX\ndpcXU22niqdun3rLleUa7rrStdP1o5u7m9yt2W3U3cw9xX2r+00umxvJXcM970H08PdY4nHM452n\nm6fC85DnL152Xlle+70eT7OcJp7WMG3I28Rb4L3Le2A6Pj1l+s7pAz7GPgKfep+Hvqa+It89viN+\n1n6Zfgf8nvs7+sv9j/i/4XnyFvFOBWABwQHlAb2BGoGzA2sDHwSZBKUHNQWNBbsGLww+FUIMCQ1Z\nH3KTb8AX8hv5YzPcZyya0RXKCJ0VWhv6MMwmTB7WEY6GzwjfEH5vpvlM6cy2CIjgR2yIuB9pGZkX\n+X0UKSoyqi7qUbRTdHF09yzWrORZ+2e9jvGPqYy5O9tqtnJ2Z6xqbFJsY+ybuIC4qriBeIf4RfGX\nEnQTJAntieTE2MQ9ieNzAudsmjOc5JpUlnRjruXcorkX5unOy553PFk1WZB8OIWYEpeyP+WDIEJQ\nLxhP5aduTR0T8oSbhU9FvqKNolGxt7hKPJLmnVaV9jjdO31D+miGT0Z1xjMJT1IreZEZkrkj801W\nRNberM/ZcdktOZSclJyjUg1plrQr1zC3KLdPZisrkw3keeZtyhuTh8r35CP5c/PbFWyFTNGjtFKu\nUA4WTC+oK3hbGFt4uEi9SFrUM99m/ur5IwuCFny9kLBQuLCz2Lh4WfHgIr9FuxYji1MXdy4xXVK6\nZHhp8NJ9y2jLspb9UOJYUlXyannc8o5Sg9KlpUMrglc0lamUycturvRauWMVYZVkVe9ql9VbVn8q\nF5VfrHCsqK74sEa45uJXTl/VfPV5bdra3kq3yu3rSOuk626s91m/r0q9akHV0IbwDa0b8Y3lG19t\nSt50oXpq9Y7NtM3KzQM1YTXtW8y2rNvyoTaj9nqdf13LVv2tq7e+2Sba1r/dd3vzDoMdFTve75Ts\nvLUreFdrvUV99W7S7oLdjxpiG7q/5n7duEd3T8Wej3ulewf2Re/ranRvbNyvv7+yCW1SNo0eSDpw\n5ZuAb9qb7Zp3tXBaKg7CQeXBJ9+mfHvjUOihzsPcw83fmX+39QjrSHkr0jq/dawto22gPaG97+iM\no50dXh1Hvrf/fu8x42N1xzWPV56gnSg98fnkgpPjp2Snnp1OPz3Umdx590z8mWtdUV29Z0PPnj8X\ndO5Mt1/3yfPe549d8Lxw9CL3Ytslt0utPa49R35w/eFIr1tv62X3y+1XPK509E3rO9Hv03/6asDV\nc9f41y5dn3m978bsG7duJt0cuCW69fh29u0XdwruTNxdeo94r/y+2v3qB/oP6n+0/rFlwG3g+GDA\nYM/DWQ/vDgmHnv6U/9OH4dJHzEfVI0YjjY+dHx8bDRq98mTOk+GnsqcTz8p+Vv9563Or59/94vtL\nz1j82PAL+YvPv655qfNy76uprzrHI8cfvM55PfGm/K3O233vuO+638e9H5ko/ED+UPPR+mPHp9BP\n9z7nfP78L/eE8/stRzjPAAAAIGNIUk0AAHomAACAhAAA+gAAAIDoAAB1MAAA6mAAADqYAAAXcJy6\nUTwAAAAJcEhZcwAALiMAAC4jAXilP3YAAPtASURBVHic7N13fBzV1cbx38yuVl22LPfeMZhiMJ3Q\nW4DQWyiBvPSYTkINBEiBEBIgBEiogUBCCb33FsCAccE2bhj3XiSrS6vdnfePO1u1K61WktWebz6b\nmZ25M3MlsNmz595zLcdxEBERERERka7L7ugOiIiIiIiISOsosBMREREREeniFNiJiIiIiIh0cQrs\nREREREREujgFdiIiIiIiIl2cAjsREREREZEuToGdiIiIiIhIF6fATkREREREpItTYCciIiIiItLF\nKbATERERERHp4rwd3YG2dubBB3d0F1rPccDjgZwcsy8iHcLr8eAA737xBes2bcrkFv8HPAbUArkx\nxyuBb4GbgQ9beM+TgHuBQcBmYAxQnknnuqtzzjiakcMGUVlV09FdERHAcf+vKN8my2MR0kebVsnL\nzWZTaQV/uv95GgKBju5Oh3H0GbkRZexERFLwBwL4vF6O3n9/+vfpk8ktchO2YYXAj4APgO1bcD8f\n0aDua+B5oCGTjomIiEj3osBORCQF27Korq3FsiyOPegg+rU8uAu6248BC8h2twOARe65MxOuGQyc\nA5wI5Cec2wkT1AH8BLgICKelCoGfAmcB/ROuywIGutt84HRgUsz5AcDPgZOBohTXZrvvDwDOB4aT\n3PbABcCxQF6S8zZwuNtmmxT3GA2cjfk9bJeijYiIiMTodkMxRUTakm3bVFZXU5ifz/EHHcTLH37I\nhtLSlt4mHHz53e0GYCYwHugX0+5y4J6Y93WYIOhTTED1Ucy5DcDfgMuAw4A3AU/M+fOBR9z9Pd17\nPAfUAz9z73UQJoB6POa6IHAk8K77/kDgHeB+TKD2fzFtjwNeiXn/MiagC6t1+/8/9/0ITJZyTEyb\nvwNTYt7/CbiaeG9gAlkRERFJQRk7EZFmhIM7gOMOOiiTYZmjgBMw2bSTgT8Cp7rnwoHRIZigrgL4\nMSbgyiEaYE0HPnf3Q5jA8DX3/TuYoO5CTCBWBjwM7O2e97nbUzBZPYCnMZm9xzHB3rHAae593gGG\nJFx7MTAS2AX4r3vs5pif8QH3HtMwmcXfYoagvu7+HLjXjcEMJ50IfAb8ArjKPX8QJqibAfRx2ywA\njorpd5NGjxjCwP4l1NX5m28sIiLSjShjJyKShtjM3XEtz9xtC7yQ5PjvgFfd/XAm7GpMYAVmWOPV\nmCGJj2GCm5WYTN5emIAsnN16AHjI3b8Y+A8mUPuCaKYQ4Ayigdnf3O11Mf3YBrgFOB64j+hw0i2Y\nrFkN8AdMgDoq5r4/c7fnAbOBOZiiLrluf3cCdnOPXx7zM38PnAvcBRS4x3MxQ07nAbtjvoRsshJK\nbk4Oe+46kQP22YWQ41BVVYNt67tLERHpORTYiYikqRXB3RpMYPYjzJBKMBm6D2LajHa3vwauBwJE\n59jtGO6Cu80iOuxyvLs9HjiUaIYNooVZwsFZOfBWzPmJ7vYq4FJM8bpwdi38zHDZsVVEg6vwNvys\n0ZigbD0mGAtfd1fMs3Z2t8MxWbhCzFBNMJnALExw+Rnm9/Sde242JpB8jiacftJh7DZpW1auXk9t\nvR+PgjoREelhFNiJiLRAhsHdLOAmd/9fmOzW45gM1rqEtkswQykHYIKcWvd6iAZZscLH1mCyX6Mw\nwzSXA+E1GqyYNlVJrl2GmbM3GJPhqyQaoIVVxOynqq+dh8m2VbrvszFBZSDmWeVu/0YAKzBz7LLc\na8uBfYF9gF9i5g7uCDyLCfYuS/rQ3ByKCvNZu34z/oaAgjoREemRFNiJiLRQBsFddsz++Zi5b2OA\nJzFZNjBVMvcEngEedI8Nw/w9vcp9b0VvE8nYzXe3XxMtQjIQU91yhfs+/Hd94tIICzHz2l4mml0b\nisnEJT4zNlqK7QeYYHQDZs7eZEwVUID3MQFZf3cfTNB3mrvfGxPArsYEnEdjfj/3Y+YkgqnW+U+i\nmc44Xq+Hs356JH379Ka0rALbTuyaiIhIz6CvNUVEMpBmQZVkUUY90QDsEKJZqIfd7T+AW4EbMIHZ\nEsw8s1T3exKoxhQheYToHLeFmGItsdclXv+ou/0LcDtmvt4K4Adgvyaujd0PD90M9/9JTDD2NCao\nWweUYoK3jzDDPz/CzBd8BzMsMzzXz4cJ7t52+/RjzNIMEA0MI0r69GLKOScyZsQQSsvKFdSJiEiP\npsBORCRDaQR3qf6OfReTlQK4GzNP7jNMsOMHfoOZV9aACc7C1TDDWTor5t6bMJUw12KKkDyMWULh\nd0SLqXhjros1HbNkQQ2mgMqfMEMmLyAaSHkStrH3i92/EfgzJuP3T/dn+Q5T5CU8x+9koks3PI0J\nWF/GBKVgCsyc7PbnKsx8wEOBpzBDMwHIz8vlkP1248qLTmNQ/76s3bBZhVJERKTHsxwn2ZSNruvM\ngw/u6C60nuOAxwM5OWZfRDq1UChEQZ5Zi3vmggXM/f576vx+MEMid8XMYZub5fWSl5NDRXU1juMU\nATtgqj/OAha7t/NhlhQAU/bf7/F4sIBAMJiNGa7ZAHyV5fUGPbYdfhaY+WjFmDlssXPiernXbQBm\nZnm99CsuZtOWLfgbGsAEZ7u42xmYKpZhxcAemPl5s91jOZiArQ74CrP8Qmz7SZhAcwFAYUEeHtvD\nlorw1DtGYYaiznPvS1FhPg0NAWrr6sEEkTu6/Z6LO1cwPy+XvXfbgT1324HiXoVsKt1Cfb1fQZ1I\nJ+e4/1eUb5PlsQjpo02r5OVms6m0gj/d/zwNgVRTnru/7hbDtAUFdp2RAjuRLicUCuHLyqJ3YSEV\n1dV8u2gRi5YtAyAYDNK3uJjdtt+efsXFLF6xglkLFlBbX4/HtrFtm21GjWJLRQUr163Dl5UFQEND\nAwX5+eyx4454PR6+mj2bsgoTr/UqKGDPnXaiIC+Pr+fMYdX69Xg9HizLIhAM4svKYucJE8j2+Zgx\nfz6VJpikb3Exu06cyJABA1izfj3fzJtHaWU52e4z6/0NFPcq5IAf7UJdXT2ffDGL2to6HMdh+23H\nsN02I/nfl7NZs24jAAP7l7Df3pMor6hm2ox5bCmvxOfLIhRyCASDbDd+JAftN5msrCz+N3UWM2Yv\nxMLC47EJhUKEQg47bj+WA/bZhS3llbz9wVesWbeRnGxf5HcHFrvstA377LEjxb2KKC0rp7bOr6GX\nIl2EAru2pcDO6G4xTFtQYNcZhQO73FwFdiJdjOM4ZPt85OXmUl1bSzAYxAHyc3NxHIfaujoK8vNp\naGigrr4ey7KwbZvC/Hz8DQ3U1NZGMlChUIjcnBw8to3j3rum1qwQkJebiwUEQyGyvF6qamsj4yzD\nQWZ2djYW4HefhWWR7/69UlVTQ35ODo7PR15vL/nZNsGgQzAYIi83h+wcH+BQW1tPnRtEFeTn4fNl\nUVtXT3W16Ud+fi4+nxfbsqmr91NdXYvHY+M4DrZlUVhUQE1NLcFgiKLCfKqqaggEg1iWheM4eDw2\nhQX5VFbVkJ2dhce2qaisjv4OHIcsr5fCgjwqKqsV0Il0QY4DWFBcYJPltQiFmr1EmqDAzuhuMUxb\nUFXMzsj8i9rH8XqGgeVXcCfStdQ7DvU1NSaDZtvYQLXfTzAUwrJtttTW4rVtPD6fCdiAzdXV2JaF\n1z0G4LEsaoNBgu5QS49t48k2BTZrAgGTzbIs7GAQjzf617kHaHAc6mpr3eDJY65zHKr8fkKhEJbX\nS6XfjxVyrOyglzq/44RCDpZls6WyjsCWarAssjwmoxgIOGworSIYCuH12Hg9ZspdaXkNwaD5lOb1\nePB4bAIN4Z/AoWr9lsjPU1mzhSyvJ3IOgECIiuot5khVPR6Pjddj4wSjf+/5GwKUV5VFf8FB/Z0o\n0pW4gZ0PWGFblDn6bqZVbAssi+TltKRH636Bnbfr/0iO14NdV39h1qw5txEKguaPiHR5XpL/hRv+\n73J2kmNNXZfqeJiH+GonYVmxb2pq2VIymCMO3pPdRkBZZTDJFSIirZPlsagPOHw4u+bqjRXBP2dn\nKSJpDccOUVMf0qguaaTrR0GJukMQ5M0Cp872bNgAwQbwdL9/TCLSCVRXEbAKGDowm23GQmVpzx3S\nIyLtJzvLotbv8PGcWk+t38Gj4dSt0hCEQCgy7kEkovtFDN3h2wsz+Dzo5ORA0Gvm24mItLVAADs7\ni5q6IP5KKK/WxBcRaXu+LIs6vwMOAa9tobiudWzLrHejX6Mk6gbpLRERERERkZ5NgZ2IiIiIiEgX\np8BORERERESki1NgJyIiIiIi0sUpsBMREREREeniFNiJiIiIiIh0cQrsREREREREujgFdiIiIiIi\nIl2cAjsREREREZEuToGdiIiIiIhIF6fATkREREREpItTYCciIiIiItLFKbATERERERHp4hTYiYiI\niIiIdHEK7ERERERERLo4BXYiIiIiIiJdnAI7ERERERGRLk6BnYiIiIiISBenwE5ERERERKSLU2An\nIiIiIiLSxSmwExERERER6eIU2ImIiIiIiHRxCuxERERERES6OAV2IiIiIiIiXZwCOxERERERkS7O\n29EdaGvFhYVb9XmWZVFTX09tfT3WVn2yiIiIiIiI0e0Cu89nz95qzwo5Do7jMGboUAb37UtlTQ2B\nQADLUognIiIiIiJbT7cL7GYuWrTVnzl78WJ2Hj+eH+20EzV1ddTU1WHbGuUqIiIiIiJbh6KPNjJz\n0SL+++GHhByHooICQqFQR3dJRERERER6CAV2bWjd5s288umnBAIBCvPzFdyJiIiIiMhWocCujZVW\nVPDqZ5/R0NBAjs/X0d0REREREZEeQIFdO9hYVsa3ixfTq7BQhVRERERERKTdKbBrJ98tWcKajRvp\nXViILytLwzJFRERERKTdKLBrJ/UNDbzw0Ue8/MknAOTm5OA4Tgf3SkREREREuiMFdu0oGAqxasMG\npi9YQEFuroZlioiIiIhIu1BgtxXMXryYxatWUZSfr6ydiIiIiIi0OQV2W8k38+cTCoXwervdmvAi\nIiIiItLBFNhtJRvKypj1/fcU5eUpayciIiIiIm1Kgd1WNHvxYrZUVZGTnd3RXRERERERkW5Egd1W\nVFtfz4Lly8lXhUwREREREWlDCuy2srlLllBWWamsnYiIiIiItBkFdltZnbJ2IiIiIiLSxhTYdYC5\nP/xAWUWFsnYiIiIiItImFNh1gDq/X1k7ERERERFpMwrsOsjcJUsoVdZORERERETagAK7DlLn97NQ\nWTsREREREWkDCuw60NwlSzTXTkREREREWk2BXdsZDvwMuBF4CNjBPb4TsF+yC+r8fhasWKGsnYiI\niIiItIoCu9YbDLwGLAf+BfwOOB/Yxj3/E+AT4CnAm3jx3B9+oLSigmyfb+v0VkREREREuh0Fdq2z\nK7AUE7wBVAF17n44BTfE3Z4BfJp4gzq/n5Xr15Or4ZgiIiIiIpIhBXaZywdeAHzAAuBooBB43j1v\nudtrgD+7+3sBv0680fqysnbtqIiIiIiIdG8K7DL3C8y8ulnAROB197jH3YYDuyrgamCK+/7/Em+0\nbvNmqmtr8Xo8iadERERERESapcAuc4e52+uBUBrt/w6sAsYAE2JPOI6DSqeIiIiIiEimFNhlbg93\nO6sF18x3t0PbtisiIiIiItKTKbDLXIO7bVTpsgnh0pf+Nu6LiIiIiIj0YArsMveJuz06zfbjgP0x\nwzbntEuPRERERESkR1Jgl7l33e2dRNesA1JOl/t7zHVxZTALcnPxeb2EtEi5iIiIiIhkQIFd5h4E\nZmCWPZgF/AYoAHLc8+GqmKcC3wMHu+//mHijQSUl5GZnEwylU4NFREREREQkngK71jkWWIkJ5m4F\nKoET3HMPAfXAM8BY99jlRIdwRgwsKaG+oSESCYqIiIiIiLSEArvWWYUZhnkvjQui9CJaLGU2sK/b\nLs6owYMZPnAgNbW17dlPERERERHpxlpS0VGSq8Vk4m4E9sSsUTcQk61bCnwNLEx1sce28Xo8WsdO\nREREREQypsCu7VQC77mvtC1etYqFy5czcvBgqpW1ExERERGRDGgo5tY3GsiNPbBq40ayPJ4O6o6I\niIiIiHR1yti13q7A8UCe+z5xVKWN+T3nAIdiqmWO3mq9ExERERGRbk+BXevcDlzXwmtWAoF26IuI\niIiIiPRQGoqZuQOJBnWVwMdEi6SsxMy1+wbYEnPNo5isXRyPrX8MIiIiIiKSOUUUmTvN3c4DhmMC\nvcPdY6uBw4DdgGLgFve4lyQVMnsVFBByVBdTREREREQyo6GYmdvZ3f6aaFZuOSZbtyfQByh1j9+K\nWRbhDuAr4O+xNyqrrMRW1k5ERERERDKkaCJz493tlwnHw+93Tzj+ZyAI3JB4o7WbNlFdW6shmSIi\nIiIikhFFEpkLpTi+xN0OSdL+PWAoMCL2RCAYJOQ4WJbVtj0UEREREZEeQYFd5ua72z4Jx79zt4kZ\nO4B17nZcu/RIRERERER6JAV2mfvK3Z6ScHyNu90ryTU7uFstdyAiIiIiIm1GgV3mnne3N2OKopS4\n7z8HNmGCuF/FtL8cmAzUAN9upT6KiIiIiEgPoMAuc58DT7j71wC/c/frgIfd/TsxyxssBO5xj70G\nlMXeqLiwkByfj2Ao1bQ9ERERERGR1BTYtc7Pgdvc/RUxx28A3nX3xxOtoLkYmJJ4k4ElJfi8Xq1l\nJyIiIiIiGdE6dq33a8y6dInptsOBnwKnAjnA25hMXk1so+ysLEYPHkxNXR2qiSkiIiIiIplQYNc2\nVqU4/oz7SmnHcePo27s3m8vLtdyBiIiIiIhkREMxO9igkhJCmlsnIiIiIiKtoMCubfQCLsZUypwK\nbAQOdc8dC/wDyE924bylS3G0OLmIiIiIiLSCArvWOxOzvMF9wInAnkBfoMg9vzdwIbAas9xBnB9W\nrWJdaSm52dlbp7ciIiIiItLtKLBrnSuBJzFzFSuApzCVL2NVuNtewCfAyNiTDrBm40ZyfD5UE1NE\nRERERDKhwC5z2wJ3ufv3YwK3nwFfu8fCv9s/ALsBGzDDMX+beKN1paX4AwFsDccUEREREZEMKLDL\n3Dnu9p/AJTHHk0Vn32CGaQIcB+TFniytqKDO78dj6x+HiIiIiIi0nCKJzIWLo9yXZvvPgOlAIbB9\nu/RIRERERER6JAV2mRvlbte04JpSd1vUZCsREREREZEWUGCXufXutn8Lrhnqbje3cV9ERERERKQH\nU2CXuU/c7TlNtoo6AlNwZTMwt116JCIiIiIiPZICu8z9x91eDlwWczzZqgWTgWdjrmuIPdm3d2/y\nsrMJhkJt3kkREREREen+FNhl7iPgEXf/r5jCKOcCo91jO2AKrLyIqYpZiBm+eWvijQb26YPH4yHk\naCU7ERERERFpOW9Hd6CLOx+zvMG5wC5EAz2AmxLargcOJmF+nce2GdK/P3X19UnXSRAREREREWmO\nMnatdx6wL/AasCnhXANmPt31wEjgu8SLJ4wcSf/iYmr9/nbupoiIiIiIdFfK2LWNz9yXjQngBmCC\nusXAlqYuHDt0KDiOeVnK2YmIiIiISMspY9c2dgV+DISAJcBUzLw6C7gDOCbVhcvWrsVSQCciIiIi\nIq2gwK71HgGmAbckOTcBuAZ4xW0zKLHBvKVL2bRlCznZ2e3ZRxERERER6cYU2LXOq5jCKQAFSc5X\nAYvc/V0xwV3v2AYNgQCrNmwg1+drrz6KiIiIiEg3p8Auc2cCR7v7VwPbJ2kzB9jGPbcEGIIZmhln\nXWkpIcfRkEwREREREcmIArvMhYO664E/N9P2O+AEd/9nQH7syQ1lZVTX1eGx9Y9DRERERERaTpFE\n5g5wt/9Js/23mMqZuZg170RERERERNqEArvM5bjblixAV+tuVSlFRERERETajAK7zE1ztzu14Jod\n3e2SNu6LiIiIiIj0YArsMveBu/1Dmu2vwSxcPouEwC47KwuPbeM4Ttv1TkREREREegwFdpn7J1AB\nTAbmAgemaDcMeJhoNcx7ExsM6NOH/JwcgqFQe/RTRERERES6OW9Hd6ALWwf8HHgRmAh8CJQBy4FN\nmHXthgKDiQbQD2MCwjgD+vQhEAy2f49FREREpEtzQg7Z2T48Xg8NAX1+lCgFdq3zErA7JlibCBS7\nr0R1wC+BBxJPlPTqxfCBA6mtr2/PfoqIiIhIF+c4Dvn5uSxbuZa6upbU75OeQIFd603DLEA+CbME\nwlCgBKgH1gMzgXeIVsSMs9O4ceTl5FBWUaEFykVEREQkJY9tEwqF+PyrOR3dFemEFNi1nVnuK22+\nrCxGDBxIvV/fuIiIiIhIao7jUNynF9/NX8KiH1Z2dHekE1LxlMxdCdwH9Mn0Bv6GBmYsXEh2Vpay\ndSIiIiKSUiRb97WydZKcArvMjQcuxhRKeR04Achq6U2+/f571peWUpiXp+UORERERCQi25dFbk42\n2b4sBg3oy/xFy5Wtk5QU2GXuA8y8OQs4CngB2Aw8BOzVkhtNnTuXkOPg8XjavJMiIiIi0nUEgyHy\ncrPpW9KLmro6tlRUUVtfz4bNpXzy+cyO7p50Yppjl7nn3dd+wFnAqUAhcL77WgE8DjxBwoLkiVZv\n3MjSNWsYP3w4ZRUV7dlnEREREemkgqEQfUt6UVlVy0v/fY85837Asiy8XhvHgbo6VVGX1JSxa71P\ngfMwlTCPB14FQsBw4DfAD8A3wAUkXwoBgLlLlhAIBJS1ExEREemBgsEQ/fsWU15Rzb0P/ZcZsxfS\nEAjgb2igprae2rp6NGlHmqLAru34gZeBYzFB3gWYRcsDwGTgQWCD2+ZoEn73azdt4ofVqynKz9dc\nOxEREZEeJBgM0b9fMVvKK7nvkRcor6jq6C5JF6TArn1sAR4GDgb6AY+5x72YwO9VoBS4Hegfvmju\nkiXU+/14lbUTERER6RFCIYfi3gVsLi3nvkdeVFAnGVNg1372A+7FZO3Oijk+A1gI9AKuwyxifg7A\nus2bWblhA7nZ2Vu5qyIiIiKytTlATo4Px4F/Pfs2W8orO7pL0oUpsGtbk4A7gdXAJ8ClwM6Y4ZiP\nAz/CDMuc4O5/6F73KGa5BNZs2kSWVzVtRERERLozx3HwejwM7N+H9z6exuq1Gzu6S9LFKYJovbHA\n6cCZwLiEcx8CjwAvAolljD7HDNV8EzgCU0nzRc2vExEREeneQiGH3FwfvYsK+OzL2Xz02YyO7pJ0\nAwrsMnco8AAmsIu1CDOn7klgTRr3eR8T2A1v096JiIiISKcTCoXIz8ulqCifx59+k1lzvu/oLkk3\nocAucycQDerKgGcwQyqnt/A+Pnf7Vhv1S1LQuOPuwXFfIj2Nx2NhYXV0NzqlYMhRRWnpMny+LHJz\nfPzrmbcU1EmbUmCXuSBmEfLLgdfc95l4GLgfqAQoyM0lFAq1SQfFLCiYY9tkh0KUVVbiDwQIhYL6\ncNRFBXEozssnv7CQikAAB/RPUrq1YMghN9tL//4FVJSVU1VVQygUwrL0bz6Y4WxZXoshQ/pTVR1i\n0+YqvF59jSedV8hx6FvSm9ff+ZwZsxd1dHekm1Fgl7lfApdhYofW2Bze6VVQwIQRI6ipr2vlLQVM\nVqfQ66WipoaXKivw5ORQUNyb/NxcgsFM43DpKJZlYVkWKzZsYNj6dRzcrz/ljr6ll+4r5DgUFeSQ\nmw1X//45PptTRX5hL4oK8vQFYJhlU1fvx6pbyx9/eRg7bT+OZSs34fUouJPOqSAvl3UbNvPZl7M7\nuivSDSmwy1xiMZRW2370aArz8thUU4NWsms9D1BXU8trlRWcc8cfufzCiwg2NJCXl0cgEOjo7kkL\nWZZFbm4uK9eu5cwzz+S9Tz/lkH792eI4ytpJt+SxLYp7ZXHmVU+zJLAHf33yj2R5IEdfTkV4vV58\nvmwe+9dznHLNDbxyt4chgwdSXqkvSKXzCTkOvYry+XTqLGrr2vxjpIgCuzYwEDgeU/zEhxkZliqF\nUAisBH6XeKJXQQHjhw+nsqZGQ2zaSJ5t81ZlOUdPmcKd11zb0d2RNtJr2DBeffllDp4wgZUbNtKn\nb18CytpJN9S3OI//vPgZ60Lb8MUHz3Z0dzq1399wATVV5dz24L386+6zFNhJp1SQl8va9crWSftR\nYNc6Pwf+2cJrkgZ2e22/PbnZ2WyprASfL8ll0lK241AJ7Lvffh3dFWljw4uLmTB5F1a89Tb9LUuB\nnXRL2TlZzP1+PRN3PKWju9IlHHLw/jw89UHqKqvx2BbBkP5ekM7Btm3AoXevAj75YqayddJuFNhl\nbg+iQZ0DLAZWYebcpRrc3wdYkHhw3LBhjB8+nNLycizLUsW/NhL+rO+xNdeiO8r2Zbd6gqtI52bK\nA2Vn68u+dPh8WWB7CASCYJkP0iIdxXEcvF4vvXsV0NAQIMvrYe36TcrWSbtSYJe5c9xtKWYduq8z\nvdGAPn0IhUKEHEfDMNuS1fS42FRqa2uZMWMGCxcuZN26dfh8PkaOHMlOO+3EuHGJa9C37L7Tpk1j\nzpw5lJaWMnDgQCZNmsTkyZPdb/My9/333zNz5kyWLl0KwOjRo9ljjz0YPjyz5RFDoRAzZ85k1qxZ\nrFmzhuLiYiZOnMgee+xBXl5eq/o6b948vvnmG1asWEF+fj7bbrste+65J717925xHxWyS3dnW7S4\nUEpdXR3ffPMN3333HRs2bKB3795ss8027L777i3+cxZr7dq1fPnll3z//fcEg0HGjh3bqr9nmvLV\nV1/x0ksvYVkWp5xyCjvvvHOz1zghM9/Wsq3oN3siHSAYDNG7VwFZWV6+/OY7vvh6DqGQQzAYpK7e\n39Hdk25MgV3mwuP7rqQVQV2vggLGDh2quXWdwIYNG7jpppt46qmnqKmpSdpm/Pjx3HzzzZx++ulp\n39dxHG699VbuvPPOpPft27cvt956K1OmTGlxn1944QVuvPFGFixolAgGYJ999uFvf/tbWh+Kwh54\n4AFuvfVWNmzY0OhcdnY2Z599NrfddhslJSUt6usbb7zBlVdeyfffN16zx7IszjrrLO666y769OnT\novuKiOH3+/nNb37DfffdR3V1daPztm1z+umnc+eddzJw4MC07ztr1iyuuuoqPvroo6Tnd999d+64\n4w4OOOCATLsep7y8nJNOOolVq1YBMGDAgBb9HSbSkYLBEAP6F1NaVsHD/3qNDZvKOrpL0oPoC+/M\nDXa377fmJpPGjaMwL48GVWnsUK+99hoDBgzgoYceShnUASxatIgzzjiDww47LK3KmuXl5ey4447c\neuutKe+7adMmLr74Yg4++OAW9fnYY4/lpJNOShnUAXz++efssssu3Hfffc3er6amhr322ouLL744\naVAHUF9fz0MPPUT//v15++230+7rVVddxU9+8pOkQR2Y4PeJJ55gyJAhTJ8+Pe37ioixYsUKRo0a\nxR133JE0qAOT+XvqqacYNGgQ7777blr3vf3229l5551TBnUAX3/9NQceeCDnnHNOyjYtcckll0SC\nOoCsrKw2ua9IezNBXR/KtlRx3yMvKKiTrU4Zu8zNBn4EDADWZHqTwrw8Qhoy0qHeeustjjnmmLhj\ne++9Nz/5yU8YOHAgfr+f6dOn8/TTT1NVVQXAe++9x6GHHtrkhx3HcTj44IOZO3du5Nhee+3FKaec\nQp8+fVi6dClPPPFEZPjkhx9+yDHHHMOrr77abJ8PPvhgPvzww8j7YcOGceaZZzJ+/Hjq6up44403\neP311yPnL730Uvr168epp56a8p6HHnooX375ZeT92LFjOeOMMxg5ciR1dXW89957vPjii4D5gHjE\nEUfw8ccfs//++zfZ15tvvpm777478n7QoEGcffbZTJgwgerqal544YXIz1JXV8f+++/Pt99+y5gx\nY5r9PYgIVFVVsf/++7NmTfQ/Rdtttx2nn346w4YNo7y8nJdeeinu76vDDz+cr776it133z3lfX/7\n299y8803xx3bY489OPnkk+nXrx/r1q3j6aefZtasWQD885//pLS0lJdffjnjn+Vf//oXTz31VMbX\ni3SUYDDEgH7FlG2p4G8PPU95ZfIvWETakwK7zL2MCexOAWZmepNvFy9m2IAB2LathZY7QH19PRde\neGHkvWVZvPDCCxx//PGN2t57770cddRRkSDk448/5ve//z033nhj0ntfe+21cdmnu+66iyuvvDKu\nzc0338y5557LY489BpjM4UMPPcQFF1yQss/XXnttXFB37rnn8sgjj8S1ueiii/joo4844ogjqK83\n1beuu+46jjvuOLKzsxvd85ZbbuGLL76IvL/44osbZfkuuugipk2bxkEHHRQJcC+88MJmM4a//e1v\nI+8PO+ww3nnnnbg2U6ZM4bnnnosEndXV1VxyySW89dZbKe8rIlFXX301y5Yti7z/5S9/yZ///Oe4\nNpdeein/+c9/OOOMMyLHpkyZwjfffJP0nu+//36joO6JJ57grLPOijt2zTXX8Ne//pUrrrgCgFde\neYVbbrmFW265pcU/x4oVK7j44otbfJ1IRwuFQvQpLmJzWQX3P/qigjrpMBqKmbm/AO8B1wHHNNM2\npZXr1/P9ypXk5+S0Wcckff/4xz9YuXJl5P2LL76YNKgDyMnJ4YMPPmDy5MmRY/fffz+1tbWN2i5Z\nsiQuS3XVVVc1CurCHn30UU444YTI+zvuuCPl4sMLFizgT3/6U+T9Oeec0yioCzvwwAMjASPAsmXL\n+Pe//92oXWVlJQ888EDk/bHHHpty6OZuu+3Gf/7zn8j7hQsX8swzzyRtC/D73/8+sj9+/HjeeOON\npO1OOeUUnnjiicj7t99+u0VDPUV6qtLS0rg/kyeccEKjoC7s9NNPj/v7Y/r06Sm/QLn11lvj3r/4\n4ouNgrqwyy+/nLvuuivy/ve//z1LlixJ+2cIO//88yNfGol0FY7jkJ+XS0NDgEeefI3yCv07LB1H\ngV16TgOuAaa4r8sxlTCnuudfAd4FbgIuAS6KaRv7+hXQ6L+MazZtwuvxtO9PIEm99NJLkf0f//jH\nHHfccc1ec/3110f2161bx8cff9yozdNPPx2Zg1dUVMS11za9QPof/vCHyP6SJUt49tnkixH/7W9/\ni+yPHDmSv//9703e9/TTT2ennXaKvH/llVcatXnyySfZuHFj5P0NN9zQ5D2PPvpodtxxx8j79957\nL2m7OXPmxAVn119/PV5v6kECZ511FgcddFDkfaqAVUSiZs2aRUVFReT9pZde2mT7Cy64gOLi4sj7\nzz77rFGb6dOnxx0/99xzU37hFXbllVeyyy67ABAMBpv9uynRHXfcEZn3d8wxx6hYinQpvXoV8N7H\n0zSnTjqcArv03ALcAdzvvu4B3gR+E9PmUOC3wN+Av8e0jX3dCfwe6RRWrFjBJ598Enl/6KGHpnXd\nAQccQFFRUeT9/PnzG7WJnd927LHH0r9//ybvOWHCBI488sjI+2Tz7Px+f9zcleuuuw5fGovZn3XW\nWUyePJmzzjqLI444otH5tWvXRva33XbbJufchE2aNCmyv3DhwqRt3nzzzch+nz59mv1gCHDaaadF\n9t977z02b97c7DUiPdnixYsj+7ZtM3r06Cbb9+rVix122CHyPnYOcNi0adPi3scOV29K7J/fVNn5\nZKZNm8Z1110HwJgxY/jHP/6BR192ShcQCoXo36+YufN/4NOpszq6OyKaY5emFcBwYG2Sc4kLDDUV\nLPdz7xV/gWXh03/EtrqGhgZ22WUXysrKWLp0KRMnTkzrupKSEoqLiyPfkjc0NMSdX79+PbNnRxcg\nPfDAA9O67wEHHBAJhr766qtG57/66qtIcYS8vLy44ZtNueqqq7jqqqtSnv/d737HlVdeydSpUyko\nKEjrnqWlpc22+frr6Coge+21F7169Wr2mv322y+yX1FRwfTp0znssMPS6pNITxSbfQuFQikrYsaq\nrKyM7Ofn5zc6HzvvbujQoZFMXHP23HPPyP78+fNZsGABEyZMaPa6c889N7J///33M2jQoJSVeUU6\nk9zcbCoqqnn1rcaZb5GOoMAuPemlcjIUtCzK6+vJsm0CjtPiBbUlM2PGjIkUN9m8eTM5ac5znDdv\nHsuXL4+8LywsjDs/Y8aMuKUN0vlgA2YOWtiyZcuYPXt23JDH//3vf5H9XXbZhX79+qV133T06dOH\no446Kq22lZWVcdUz995770ZtHMeJ+9Z/u+22S+vew4cPZ/DgwZEAdurUqQrsRJoQO9Qa4IMPPmDb\nbbdN2X7hwoXMnBmt97XXXns1alNeXh7ZHzt2bNrZs8Qvb5YuXdrs339XXHEFc+bMAcyc4cMPPzyt\nZ4l0tFDIobhXIe99PI31GzUEUzoHDcXsYP369aN44EBmrl7NhpoacjwebMtScLeVlZSUJP3mOpnE\nIUaJgc2mTZsi+/n5+XEBW1MSPwDF3geICyYTP8ytXr2aJ554gmuuuYZLLrmEe+65J+7DW1u6+eab\n4/qWLHNYVlYW9+Ew3cAuJyeHsWPHRt7rW3uRpo0fP56f/OQnkfc33nhjXEGoRLFFnPLz85P++bUs\nK7LfmjXkvvvuuybPv/nmm/z1r38FzJzh8BzidNYIFekIwVCInGwf/fsWM3LEQKpqavl06rcd3S2R\nCGXsOlBJSQnbb789DYEAFrC8rAwch0ElJfixQOFdp1NRURFXcW7nnXeOm28GZu5eWHZ2dlrz4IBG\n7VavXh33/ttvo//xGDZsGAArV65kypQpcXP6Yk2cOJGHH3446bfyLbVq1SqmTJnCa6+9Fjl2+eWX\nxw2/CluzZk1cQYfc3Ny0nxP7e4hdl0tEkrvrrrv46KOPqK6upry8nF122YUHHniAE088Eds239/O\nmTOHCy+8kKlTp8ZdN2TIkEb3i50THDuHrzmxXz4BcSMXElVUVMTN3XvwwQfJy8tL+1kiW1swGKJv\nn15UVFbz7EsfEAgG2VJeSYWWNpBORIFd2ygETgX2x8zF6wNUAKuAL4F/A41SD8OGDcOyLPz19WR5\nPIS8XpaXl4Mvi0EeD07Mt6bSOVx22WVxWaRf/epXjdqsX78+st+/f/9GQzVTKSwspKCgIFLuO7ZS\nJRBXBnzs2LGsXr2a4cOHN3nP7777jr333ps777wzaV+b89vf/pZFixaxdOnSuHXuwKydFVs6PVZi\ntnHAgAFpP7NPnz4p7yMijY0bN47vvvuOyy67jNdff51NmzZxyimn0Lt3b0pKSqiurmbdunVx7e+8\n806OPfbYpPeLXdJl6dKlzJgxI615drHra0LTgd2UKVNYtWoVYP4e1ZBr6cyCQVMkpWxLBfc98qKW\nNJBOS4Fd6/0CUwkz1SSEU4C7MNUwb4o9EQgEIkNOHMfJ9VrW9ng8zrIt5Ys8Pt+gIbaNvx07Li1z\n2223xa21dtRRR3H66ac3ahdbvCDd4Z3htrGBXWzGC+I/JL333ntxSyicddZZHHnkkfTu3ZvS0lJe\nfvllnnvuucj5q6++moKCAi666KK0+7Ny5UpuvfVWQqHE+kBmOFeqoA5otLZfS34PsRVHE38HIpLc\niBEjeOqppzjssMMic2C3bNnCli1b4tplZ2fz6KOPsu+++6a81z777BP3/p577uFf//pXk88vKyuL\n+/sRoL6+Pmnbxx9/PLKm5s4778ydd97Z5L1FOlIo5FDcu5BNm7fwj8dfVlAnnZrm2LXOHcADmKCu\nFvgQeAi43T3+BhAuH3gj8ELsxXV1dYRCIRoaGggEAhMc+Npr29Mcxykvr6v7hWVZIeXsOoe//e1v\n/PrXv468HzZsGE8++WTStrGBUEvmp3i93rj2sR+K6urqIsGSZVk8/PDD/PDDDwwdOpQFCxbwxBNP\ncOqpp3L44Ydz2mmn8eyzz/LFF1/EZQsvueQSFi1alHZ/Nm/enDSoA/P7mDhxYtI1/IBG17Xk9xA7\nFDPVB0MRiXfDDTdQVFQUV9goJyeHESNGxP09UF9fz3777cfpp59OXV1d0nuNHz+en/70p5H3Tz75\nJPfcc0+Tzz/77LPTqpa7ZMkSpkyZEnn/8MMPN3uNSEfKyc7C3xDgiWffYku5gjrp3BTYZe4gzKLl\nAI8BJcDBwIXADcDFwE/c4+HUygnAZeEbbNiwAY/HQ25uLllZWcuB84Oh0LlYHFLg873oOI7++XQC\nd955J5ddFvnHRkFBAe+9915cmfFYsYUHUgVGyYRCobj24bkxYBb8DQaDgKk4GQqFyMvL49NPP2Wb\nbbZJer+99torbkHyYDAYt8B5c3w+H7/+9a956aWXePfdd3n00Uc55JBDAJNtnjdvHgceeCCPPvpo\no2uthGHELfk9hH9OiP8diEhyxxxzDLfffnvk/e67786HH37Ipk2bWLp0KRs2bGDmzJmccsopkTZP\nP/00kyZNSpkV/8Mf/hA3N/bKK6/k1FNPZenSpXHtpk6dyuTJk3nttdfo378/Q4cOjZxLNr/4vPPO\ni3xJdfvtt8cN+xTpTEKhEB6PzaiRQ5g5exFr12ldVen89Kkpc2e5278D52Iydqn8CTjH3b84fLCi\nooLVq1dTVFSEbdulIcd5pD4QeGxoUdEHAwsKZjWEQiqf0sF+9atfcc0110TeFxQU8OWXX6YMpoC4\n0uAtyTjV19fHfYMeWzo8MZsHcNNNNzFq1Kgm73nggQdyxhlnRN6/9NJL+P3pDfDdbrvt+P3vf89x\nxx3HoYceyjnnnMN7773HSy+9FNfuvPPOY9asWXHHEsujt+T3EDuMM52170R6sj/84Q9xBY0uueQS\nvvrqKw488EDy8/OxLIucnBwmTZrEs88+GzfSYOHChfz85z9Pet/Ro0fz/PPPxx177rnnGD16NKNG\njWL33Xdn4MCB7L333syYMQOAZ599Nm7dzsRiKH/4wx/46KOPALNuZ3hRcpHOJhRyyM/LITc3h7c/\n+JKPP2+fKtMibU2BXebCM73vSLP9P4FFwHhgUvjgsmXLKCsrIy8/n4ZQiMGFhQwv6kUgFCLkKKzr\nSCeeeCJ/+ctfIu/79+/P7Nmzm13IPHZ9ubKysrRLd1dUVFBWFl0Lp6SkJLKfnZ3d6EPSySefnNZ9\njz/++Mj+6tWrW70MwnHHHcczzzwTd+zuu++Oe5+YzYz9uZoTWzAltpCKiMSrqqqKLBcA5ouc5rLy\nZ555Jrfeemvk/UsvvcSnn36atO2RRx7JN998w6BBg+KOL1u2jGnTpkUKRfXt25dp06ZxwAEHMH/+\n/Ei77OzsyP6XX37JjTfeCJj1Kl9++eWUfbQsKy5b35KquiKt5TgOuTk+CvLzeP6Vj3jmxffZUl7Z\n0d0SSYuKp2Qu/Mm1JWX7lmMCu77hA4FAgK++/pqhw4ax87hx9M/NpT4YxHEcvGiGXUfYsmULBx54\nYFwWauLEiXz22Wf07t272etHjBgR2S8tLaWysjLlsM1Yfr8/bshi4oep2MBu7NixjBkzptl7QuM1\n5JYvX84ee+yR1rWpnHrqqTz44IORb9/ff/996uvrIx/khgwZQnZ2diRTt3lz+kNYYrN7AwcObFU/\nRbqzTz/9NK56buwadU259NJLufvuuyOFVV555RX222+/pG0nT57MqlWreOyxx/jvf//LvHnz2LBh\nA/369WO77bbjpJNO4v/+7/8iIwoSFzcPix3Ovt9++7FmzRrefvvtRtn9vLw8Kisr46oAf/HFF0yY\nMIFVq1bR0NDA4YcfTt++fRFpDcdxsC2bXkX5+HxZkf/+ej0efL4sHv3368yck/68dJHOQIFd5j7F\nZO22Bb5J85rw+L35cUcdh1UrVrDD4MHk9elDTU1NyhKb0r7mzp3L/vvvH1cE4JBDDuHdd99tNG8s\nldhgpK6uLu3ALrYcOTTOVg0ePJjZs2cD8ZUjm1NSUkJeXl6kqmbsHLbW2H///SOB3Zo1a1i+fHlk\nMfb+/ftTXFwc+ZlaUt0ydrkIZexEUpszZ05kPz8/P+makskUFxdz4IEHRoZVz5s3r8n2tm1z3nnn\ncd555xEIBKisrKSgoKDR8PDvv/+e77//PvJ+3LhxgMnYxxZu+ve//81TTz2VVl8BHn300bi5vO++\n+y6HHnpo2teLJAoGQ/Qqyic3J5sZsxeyqbSCbJ/59zkn28fm0nK+mbWgg3sp0nIK7DL3CCaw+xuQ\nzurPF2HWuHsGWJ2swazFixk+cGCjbzBl6/j000854IADcGKGwP7iF7/ggQceaNF9wsENmIzs7Nmz\nm11vDogEbWA+eE2YMCHu/M4778zbb78NmMXCq6ur01pGoLy8PG6phMQPY3V1dZSXlzNv3jzGjRsX\nV/ygKbGLGAM0NDTEPWObbbaJBHbTp09P655r1qxhwYLof0x33nnntK4T6Yli/8wNGjQobhh4c2Lb\nxmbHmuP1elN+URWeaxe+f6pRBU4rpxm09nrp2YLBECUlvaiv9/Pov1/nuwVLm79IpItQYJe5/wK/\nBP4CLMAUUPk8Sbs84NeYSpmbgEvd443SP2s3bXIWLF/OhHHjqK7UeO6t6cMPP+Tggw+OO3bHHXfE\nFU5J17bbbsvo0aNZsmQJYAK2n/zkJ81eFzv3bZtttmk07DM2yNmwYQNz585Na0hlbKAE8UMza2tr\nmTx5cmRezPXXX89tt93W7D0hPrMGjefB7LzzznzyySeAyYSmY/bs2XFzErfffvu0rhPpiWKrTi5b\ntoyVK1cybNiwtK6NXd8u8QuihoYG1q1bx4YNG9iwYQP77rsvBQUFzd7zzTffjOzvvvvukWuys7M5\n8MADWbVqVbMFkbxeL36/ny+++CIyLHvcuHFMmDAh8iVV4jB1kXQFgyH69e1NdXUt9z78PKVlWitV\nuhcFdpk7AzgCqMcMsfwMk4n71j1mAQOAyUD4v75VwMuYzF1DzL1GY5ZE+NPmigq8ythtVTNnzoyU\n8Q976qmn4qpJttRBBx0UCezeeOMNbrjhhibbB4NB3n333cj7H//4x43a7L///nFDKp999tm0Arv3\n338/sj9u3Li4wC43Nzfug9abb76ZdmAXu4bdDjvswOjRo+POH3bYYZG1r6ZNm8bs2bPZcccdm7xn\n7AfDSZMmNdtepCfbddddI/uBQIDPPvuM0047rdnramtr4wqmxI4yAPPF0dixYyMVdN96662kfycl\nXvPiiy9G3h933HGR/by8vEbVdJsSCAQYM2YMK1asAOCqq67ioosuSvt6kWTCQV1VdS1/U1An3ZSq\nYmbuR8AhQHbMsSHAkcDxwHGYIZqxC/mMBPYBhmGCufALQJOJOkBtbS0nn3xy3NCel156qVVBHRC3\nXtQXX3zB66+/3mT7Bx98kGXLliW9Pqx///4cc8wxkfcPPPBAJHhMZcmSJTz44IOR9yeccEKjNkce\neWRk/9tvv21U8TKZF154IZKNg+SB6I9//OO44gnNVetbvnw5jzzySOT9qaee2mw/RHqyvffem8GD\nB0fex1bIbMoDDzwQN6f3qKOOijs/ZMiQuKHgzz33XLP3vPXWWyNDOktKStr0z6+GXkprBYMh+iuo\nkx5AgV3mngFuBi5pg9ctwCsAHtvG0ep1W821117LDz/8EHn/zDPPxH3TnKlDDz00rpDB2WefnTII\n++KLL7j44sjyhpxyyilsu+22KfsbVl9fz/HHHx+37lus+vp6Tj311MhwpqysLM4777xG7S688MK4\nrN1FF13E4sWLU/5sX3zxRVxWoLCwkKuuuqpRO8uy4p73yCOP8I9//CPlfU877bTIz1JUVMS5556b\nsq2IQE5ODlOmTIm8/+qrr5L+GY/1+uuv86tf/Sryfu+99+bwww9v1C72S6B//vOfkUJJydx9991x\nc5Gvv/56CgsL0/oZ0hFbLVikpeKCuocU1En3pqGYmfvEfbUZ27YZPXgw9f6G5htLq82bNy8ui5Sb\nm8u3337LtGnTqKysjFtHKZV169Zx7rnnJp1Dd9ttt3HQQQcBZtmDSZMmcdddd/HTn/6UgoICysrK\neOihh7j++usj1+Tn5/OnP/0p5fMmTZrE7373O2666SaASGGW22+/neOPP56SkhJqamp4++23ueKK\nK1i5cmXk2j/+8Y9xGbSw/v37c+edd3LBBRcAptjKDjvswG233cYZZ5xBv379CAQCLF26lPvvv597\n77037vqHH3445bIE1157LU899VRkjt0vfvELZsyYwU033cSwYcNoaGjg448/ZsqUKXHB5B//+McW\nFYIQ6al+/etf8/rrr/Pll18CpoLk119/zW9/+1v23XdfevfuTSAQYPHixdxzzz1xWXHLslIWh7ri\niit48MEHWbt2LQCHH344t912Gz//+c8pKSnB7/czb948br31Vl555ZXIdYceeii//OUv2/EnFklf\nMBiiX7/eVFXVcu9Dz1O6RUGddG8K7DqRbUeMYGj//myurVEqdStILLddW1vL7bff3uL7bL/99kkD\nuwMPPJB77703sn5TZWUl559/Ppdeeil9+vRhw4YNjRYvf/755+PWwUvmxhtvpKKigjvvvBMwC3qf\nf/75XHTRRQwcOJCysrK4KphgsnDJsmph559/PuvWreM3v/kNYCplXnXVVVx11VUMHjyYurq6uCUg\nwh577LFmh1w9//zz7LHHHpH1rR5++OFIMFhTU9NoKYSLL76YX/ziF03eU0Si3nrrLfbdd9/IFyhz\n5szh+OOPx+PxMGjQICorK+PWl4u9bqeddkp6z169evHkk09G5h83NDRw9dVXc8011zB48GAqKysb\n/dndc889eeONN9r4pxPJjMnUFVNVVaOgTnoMxQ9t5yfA3cCbmDXu3gb+DlwINLvKsmVZTBw9mjq/\nnyQFM6UdfP55siKmbevSSy/l5Zdfjqs6V1dXx5o1a+KCuoEDB/Lll182W6Ag7E9/+hPPP/983BDK\nYDDI6tWr44I6j8fDAw88wN///vdm73nTTTfx+uuvU1JSEnd8zZo1jYK6HXfckdmzZ/N///d/zd53\nm222Yd68eY3W2Fq3bl2jD4Z33HEH9913X7P3FJGo3r17M2fOHC6//PK448FgkFWrVjUK6vbee2+W\nLFmSdAhmrIMPPphvvvkmrtKm4zisXr260Z/dq6++mqlTpzZaTiUT9fX1cX+PxS7rIJKO6PDLGv76\n0H8V1EmPoYxd6+0DPIspnJLKPzDz6G5N1WC7UaMY0KcPm8vLwedL1UwykCpMPuqooxg2bFirFsFe\nt24de++9d5Ntjj32WDZu3MiTTz7Ju+++y6JFiygtLaVfv35ss802HH300Zx22mlpDf2MdeKJJ3LM\nMcfw7LPP8vbbb7Nw4ULWrl1LcXEx48eP54ADDuDss89u0WLmRx11FBs2bOC///0vb775JgsXLmTN\nmjVkZ2czatQoJk6cyEknncQ+++zTor4OHjyYqVOn8uGHH/LMM88wd+5cVq5cSW5uLmPGjGHvvffm\nvPPOUxlzkVgOWLaFJ82/G+655x6uu+46nnrqKb788kuWLFnCxo0byc/PZ9SoUWy//faceuqpcdU0\nmzN58mSWL1/Oc889x6uvvsrChQtZt24dRUVFjB07lr333puf//znKYdjZyI3N5cpU6awfPly/H4/\nkydPbrN7S/cXydS5QV3ZFi0fJT2HArvWOQa36IlrOTALKAXygW2BiZjM6C3AUOD8xJv0692bPSdO\npKq2FsuyVDqlDYVIXVEtkzXqMpWbm8sFF1wQmcfWVrKysjjzzDM588wz2+yetm1z6qmntktVyoMO\nOigy77C1LNvWnxXp3mwLvz/YonnXAwcOjCuO0hYsy2q3vxOSsW2bW29N+T1oSpZlvsZzHDTwpYeK\nzdTd+/DzCuqkx1Fgl7n+wBPu/jzgbOCbJO1KMEM0fwacB3wE/Ce2wYhBg+hdWMiajRtbnLWRplj4\ngUAw2NEdkXYQ1D9X6e6yPGwsraS0YXVH96RLCAaDhEIhbMtyozvpSUKhEH2KC6moquG+R15Q9Uvp\nkRRFZO58oDcmQ7cDyYM6gM3AWUB41eeLExssW7uW0ooKsryKs9uUBUXAgvnzO7on0sbKGhpYNG8e\nA4GgPsBJNxWs9jN54mDqqjd3dFe6hBmzZuOzasgryicU0t8LPU1uTjZ+f4CH//WqgjrpsRTYZS48\nnuzXmBF/zbkFqAD2BgbHnvA3NBAMhSLDSKRt1DgO+2bn8OQdd/Dwq692dHekjQSB0889l9olSxjb\nrz8NCuykm9qwuZoLTj+I4MavOfn/riNcbimgV+QV9syrX3Dfn2/mktN3p742qGHaPUwo5FDcu4hp\nM+ezdr2+CJGeSymizIVnn89Is30D8BVwKLAdsKY9OiVRAcehX+/e/Kiigl8deyyP7bADhb16E1CF\ntS7Jsiwsr4fFi3+AdWs5pqQvVQrqpBtrCIQIOF6evP1Yptz6OAfv8y4lffvhhALNX9wDWBaEHIu6\nugY2rZzLg1fvxn77T2L5kg14PPreuifJzc2mdEsF//tydkd3RaRDKbDLXDg6aEmaLdw2nQyftJIF\nlAUCDCkq4mdZWSyaM4c6wNPRHZOMOJg/dLt5PAzv259KCwKhkGokSLdl2xbrN1bSu3chLz54LrO/\nXcTGzRVkZeV0dNc6BQvwNwSxbR977PozsnMLWL5kA7aCuh4lFHIo7lXA+598w5ZyFUuRnk2BXeY+\nAU4AjgIeSaP9EOAQd//b9uqUxLOBmkAAKyuLUX37KQjo4izMtyJbnJApBd/RHRJpZ16vTWVlHdXV\ndQwdOpSRI2zVBYkRnsJQXtVAYMsWZep6IJOtq+QzZetEFNi1wvuYwO4vwIfAkmba3+9u38MUVIko\nys8nOyuLuvr6Nu+kGA5maKaISFcTLvJYU6chmE3RPPWeJzZbV6ZsnYiKp7TC3zHLHBRhMnBXAyOI\nTyIUA0djKmce6x77Y+KNBpWUkOPzEQxphKaIiIhIOsJz6z77Stk6EVBg11onA1VAAfAnYBmwGpiP\nyeCtBV4FdnLbX4nJ7sUZWFJCnd+vbxtFRERE0hDO1k3/dpEWIhdxKbBrnXnAGOC5mGODgAnAKCA7\npt2BwD2JNxg+YABD+vWjVsMwRURERNISnlv3ubJ1IhGaY9d6G4BTgUuAA4DxRH+va4H/YTJ4Se04\ndiyWZRHSOnYiIiIizQpn69775Btl60RiKLDL3NHAKcDdmLXsNgL/bckNxg4dyqjBgymtqFBQJyIi\nIpKG3NxsSssq+ELZOpE4GoqZuZ8DZwL3ZXqDhkCAoDJ1IiIiImmJzK2bvZBSZetE4iiwy9zu7vbj\nTG+wfN06Fq9aRa/8fByV4hcRERFpUm5uNpvLKvjiqzkd3RWRTkeBXebC8+YGtuYmc3/4gWAohMfW\nPwoRERGRVMLZuhnfKlsnkoyiicydBvwA/B9wVqY3WbNpEwtXrCA/N7fNOiYiIiLS3eTm+szcuq/n\ndnRXRDolFU/J3BbgVsxC5U8AvwK+ANZg1rbzJLkmH1Mp88HYg+tLS9lu5Mh27KqIiIhI12WydYW8\n+/E0SrdUdHR3RDolBXaZuw+4KOb9Du6rOStJCOxEREREJLVwJcypytaJpKTALnNrgCBmrl0QCJe2\nbKrEZV9gQTv3S0RERKTbCDlmbp2ydSJNU2CXud+5r1bLzc5GNTFFREREGsvN8SlbJ5IGFU/pYDk+\nH+OHD6e2vr6juyIiIiLSqYTn1n0za6GydSLNUMaubU0AJmKKpFQDy4B5QG2qC3YaN44BxcWsLyvD\n1kLlIiIiIhG5udmUbqlg6jfK1ok0R4Fd27gS+CUwJMm5WuBV4FpgeeLJiupq/IGAgjoRERGRGOF1\n6979aBqlZcrWiTRHQzFbJxeYBtxFfFAXhMi0uVzgVEz27tDEG8xftoy1mzaRm53dvj0VERER6UJM\nJcxKZetE0qTArnVeAHZ1958GfoIZjjkUGAscgCmwssVt8wowKvEmazdtIsfnUwEVEREREaLZum9m\nLVC2TiRNGoqZuZOBI9z9I4G3krRZAnwC/AX4DNgeuAk4J7bR2s2bI8MxHUfhnYiIiPRsuTk+1q7f\nzBfT5nR0V0S6DAV2mTvW3d5I8qAuVjlwJjALOAWYAtSFT5ZVVlLn9+OxbQLBYDt0VURERKTrKCjI\nY878JZRtqWzJZRbwILAtUJrkvI35/LURWAW8C3zTyq52dwOBGiA2bToYeBToBZyHKRQonYACu8zt\n726fSbP9t8BUYC9gEvBlO/RJREREpFvIYBSTF/gxMCzN9n8APgSOJz5wEeNy4FZgd+J/P8WY3zPA\nSBTYdRoK7DJX5G6rW3BNlbstaOO+iIiIiPR0AfcFMBd4DeiNyeSFhQAfpqDdCOAgTOZuz63Wy67h\nCOAeTDHAxKC3BpP17Ats2rrdkqYosMvcAsw3GGOBdWleM8HdrmqXHomIiIgImGkyNzTT5r/AScAe\nmGkyD7R3p7qQcIyQLG26FOi/FfsiaVJVzMx96G5/mWb7czBDAxZhgsIIr8ejwikiIiIibSc3jTYX\nYOoggAnwJKo+Zl8fULsIBXaZe8LdHgc8S9N/gVyMmWQK8EjiyYElJeTn5hIMhdq0gyIiIiKSUhlm\nGCbAOMwQzbATgTuBQe77uzFrF/+d6FJXsXYB/gR8AHwNPAmcAWSlePZYt/0p7vuRmKGP04H/Yeb/\njUnjZ+gLXAa8iCkE8zrwa8ww01QOB+4lOpLsZve5/wAOwyzfdZp7zgIuBH4F7Owe6w1c5T5ncIpn\n2Jhg+XHgK+BjzO+wqSGvR7j92tF9vwtwH/A58BHm97VtE9f3eBqKmbkFwLXAHZg/lEcD72MqX1Zi\n/iCPwfwBGepe8ynmL4k4/YuLCSmoExEREdnaytytD/PZze++vxLYB5iJKZh3gXt8V2AAcIL73gv8\nm2iAFrYbpiL63ZjPiF8lnN8VuBqzJvIGTOAS60eYoaRXA39O0fdzSZIwAI4Cfo+p3P6HFNedjCl6\ncgpwi3t8F0xQl0V0qKWFKaAC5nPvTEyw+xf32DRgTcL9dwZeJfr5N2x/4ArM3MeTiP6uw87D/F6/\nBnYC/pVw/gDM7+OXwF1Jfq4eTxm71vkTcKm7n4v5g3uTe/wPmOGX4X+pXwIOTrxBUX4+wwcMoLa+\nPvGUiIiIiLSvcAZpCfEF8Rrc7W8wQd1MTBYuhMmOhX1ANKh7AFN/YQfgbEzA0w9TCX3vhOeG738o\nZnpPDfAzTEA1EfO5EUxCYEqSfl9BNKj7GrOm8kRM8PO2e/z3JEkoxDz7cszn1kXAP90+PAIcAjzl\ntnEwGb2rgffcYwGiahPuPRmYgfn8uxHzu9sRE8je7bY5GpOVTNWvWzBB3WzgVEyQ/AtgvXv+L5gq\n85JAGbvWuw8z+fZcTGWlkYAHEzSvx3yT8S8af1MDwA5jxlDSqxfrNm/GthVni4iIiLSBuuabcD7R\nYX+fJpwLzyvbBvgbZrgjmGGJ4cDmBmA/d/9k4PmY6+diPh9+A2wHPIwJvBIVYQKqHTDBJZiA6ARM\nkHUuJlnwX/c4mOGI4SDpP5ghn7E+wYwouwYzhPIdzKiyRBPcPp/svg8HkHWYIPZMzO/hQaJBFcRX\nGU30D3e7BBPkbYk5N93txxuYAPgPmOGcicZgfl8XxBz7xv255mI+Y5+GWUZMYiiSaBvrgdsw33CM\nBUZhxjbvjplflzSoA1hXWkp9QwMeBXUiIiIibWUwkIMJgrZzX9tiArV9MIHDQ27bcqKBUqJSTNYu\nrBaTWbIwI7PAzCN7nsZqiY7s2g6TfUrmeqJBXaybMEFWb6Jz3oh57uaY+ye6luj6cr9I0QbiA6s6\nogFx7HzDpgK5WD8mOv/wcuKDurA3iVYfvYDkS4BVkTzgm080kzkyzT71KIomOtgPq1axeNUq8nLT\nKd4kIiIiImk4HRNYzQO+c1/zMDUSPsPM5wo7hdRLV31O8gBlMtHiJs800Y8PgTnu/o9StHkhxfG1\nmEAIzBDLsPB9XsEEnqmE+7UHyYu4zMAMw2wr+7rbpZgiLqn82932xfweE00jmp1MFF43L6/FvesB\nNBQzPYdixjwnjiNuqRzMJN23Yg+u2rCBcUMT55eKiIiISDsIYoKPd4A/0vT6wqkCvu3dbT3RwC2V\nOZihlsmCmDnA6iauXepuR7lbX8yzZzTz3NnudghmGOishPPraVuT3G3icxItwBQaLMT8Tj5JOL+5\niWvD8yCVnEpCgV167gPGt9G9VgLD2+heIiIiItLYM8BFmCF74aGENmYY5WqaznTFShVk5Ljb8iba\nhK10t8myTDXNXBsOOsPLLlhEh0k2F5jFBqzZSc431++WCv9OUmXbwkoxAXMhyZcLq05yLCz8z1Jr\n6yWhwC491QnbTOXT/B9gEREREWmdlZig69tW3idVZigUc7657FE44MkkGAnPQYv9/Bh+tqeZa5MF\nc7HaOusVTPO+Nk33ranr053v1yMpsEvPAZixya1dbM4mvkQsAD6vV187iIiIiLSdtipekCqQKHe3\nfTHDJOelaAfRRbWXJTnXu5nnj3S3P7jbIGYYYw5mUfWmhBcgD5J8uGdbB0kV7nZsM+1GEh29tizJ\neQVvGVJgl56K5ptkJsfnY7tRo7SOnYiIiEjXMS1mfw9SB3a5RJdU+C7J+W0wRVh+SHIOovPW5rrb\ngPvsI2m8Nl6i3dztDzQ9j7CtfAGciFnovITUQz13idlP9juRDGniYQcLOQ7BUEhfTYiIiIh0HUsw\n1TXBrG2XyjnAQHf/7RRtzk5x/MdEC668HHP8XXd7BCaoTGYYcJa7/04T/UvFSbHflA/cbRFm/b1U\nwssvzKX5QivSAgrs2k425tuJEqBPildfElLu/oYG5i9bRm52c8OgRURERKQT+ZO73QN4Isn5ozEF\n+MCsv/ZZkjZg1qs7IeHYBKLLArxGfOXIvxNd9+4FohUzw4rca/Iw69L9JeVPkFp46pBNdGhlc3HD\nt0R/D3dglpFI9A/goJg20oY0FLN1soAbMYtG9sUEdxapv9nIw6wXsk3swUAwmLy1iIiIiKTLJjo/\nq7WDoeyEbTKvAbcCN2OyYwdjAq0tmPoM+7nt5gI/S3GPakzw9QLwMfA+0cXMPZg5aP+XcI0fEzRN\nxSxlsBh4EbN0whjgZKJzDE8BlidcH1slNJUFmODO697bh1nQ/dqEdokFXM7DLMcwGXgWs1D5B5gi\nMCcSnVt3J/BUwrXp/M7Dz9NgtyQU2LXOB0QXY0yXViIXERERaXsO0S/XW1uXLnx9c4XzbgFmAg9g\ngqzLYs6FgLuA64hWjEy0DjgWk8k6gPiFyJ/DBEqVSa6bjsmkPQQcDpzkvsI+wwwRTTb3L52fbTkm\no/ZrzFrOYAJOiC8EmPhzBYDd3Wt/hZkHGDsXcDVwKSaDmSiUsE1G2ZAmKLDL3GVEg7qPMd86lGL+\nsKT6piG8QLmIiIiIpOD3N9C3T6+WXhbCfDYroPm11JrzU6CY9BbxfgV4FVMUZDLmS/wFwOdAVTPX\nDsAEUfsCe7mvzZjkQXMFT1Zg5uENwwRPQzCB4uc0ztLFuhz4Pc3/jm7ErAd4EGYh9vAcuuWY4Z+e\nFH0MAVcDv3P7NQGzXMPXND2n7hJMBrSpfv0WeBB9nk5KgV3mDnO3T5E6vZ6WEQMH4g80WgVBRERE\npEeqqKxmwvgR7L7Ldnw9o6mVBBpJVtY/E+tJL6gLczBZtOkZPCv8eXyq+2qplZhhj+na5L7SMZdo\nRc6wIMmXKUhUgSkYk6poTCb9KkNBXUoqnpK5cBWiB1pzk9GDBzN6yBBq6uraoEsiIiIi3YPfH2CP\nyds131BEAAV2rRFOsS1tzU22HzMGAMfREuUiIiIiAJZlsaW8ktEjBrP7LgruRNKhwC5z4XHGuzXZ\nqhkby8qwLRX2EREREUlU7/d316xdlrvNdl8irabALnPhdUl+05qbfLt4MaUVFWT7fG3QJREREZHu\nwWTtqrpr1m45ppDI8zRfYEUkLQrsMvclpsTtrsBHwLZk8Pusqatj4YoV5CqwExEREWmk3t/QHbN2\nU4GdgdMxa9mJtJqqYqbnNUy51nXuexvz7UodpgrSAZh1QsowFZQCJA/y+mIWjzwk9mBNXR2WhmOK\niIiIxEmca9fCCpkiPYoCu/SMB/q4r6YUu6/m7iUiIiIiaar3+9lzsgI7kaYosEvPH4ERQGUr75NP\nNOsnIiIiIs0Iz7UbNWIwu++yLV/PmN/RXcrEQGAfIA+zqPdMYEs7Pi8XKARqaf3n19bKd1/VaNhp\nu1Jgl55/tufN+/XuTTAUas9HiIiIiHRp9X4/e+w6sSsGdr8Bbk04diLwYjs+83rgJuBd4PB2fE46\n/gJcCDwJnNXBfenWVDylgw0sKWGbESOo1gLlIiIiIklFKmQO73IVMk8iGtTVAt8DFcCSdn5uobsd\n0M7PSUe+u83r0F70AArsMlNI8/PtdgPOp5ms6E5jx+LzegkGg23VNxEREZFuqd7vZ89dJ3Z0N1ri\nAHe7BVNBfTymHsPsdn7u58A3wDPt/Jx0hD/kanhaO1Ng13I/xXzTcl8z7S4AHsJUyfxZqkYBDcEU\nERERaZbJ2lUzavigrpS1G+xu38CsXQcmwGnvD4DPY5IMf2zn50gnosCuZZ4Bnnb3d2ym7Q7utg/w\nL+DZZI2+XbSImro6vF5NdxQRERFpmtMZsnbFwDAgnUWIwwHcxjZ89mCgpetkWbSstoYFeJppMwBT\nFKatFGGKFea24T17FAV26XsIONXdLwMeaab9ycDFRKtgnpLsmk3l5Sxft4687Oy26qeIiIhItxSt\nkLnVs3Y+4BpgPlAKrADKgdeBHyVp/6Db9gD3/U+BDzBZuxPSeF4/4CPgPff9TsBc99mrMSPC7gR6\npbj+XLftn933DnAV5nPp18B+Ka47AfgOWBrT91gDgL8Ca917rQU2APdg1mtuKQu4AliA+X0uw1TO\n/AbzOVoLPbeAArv0/BgzXw7gJUwW7p5mrlkJPAAMwqTfwfwhOzKx4aYtW8jyenEcp006KyIiItKd\n1ddv1azdNpiA4w5gAqYISimQDRwF/A+4O+GaSW7bEvf9QOAgYDgwNo1nFmACq0nAjcAMYCLm8+Vy\nTOD3K2Bxivtth8nu7R9z7HWgP2aI5v1JrhmKqQS/HSbD+EHC+UMwgdxl7s9TiQnG+gGXA2uAvdP4\n2WJNxfzutgHq3OcGgcmYaU9TUbySNv2i0nOBu51Ket+yJDoa8+0HwJTEk2WVldT7/Xhs/eMQERER\naUp81m7b9n5cMfAh5ov6UuAYTHXHEszQwfCSWFcQv6TBSZjs1vvu+xcwwdqOwMNpPDcANGCyYL/D\nBFQ7YgLDkZjgbL17/uUU14MJlsLmYZZZANge+G3CNQ+7P9NmTMAaayfM0gkWpvDLTm7b3sAYzGfk\nLMzPOzKNnw/359oDMxJuP8wQzP7ufa5w2+yBCaglDYokmucl+u3DXRnewwFud/f3Jlr2FYAtVVXU\nNTRgK7ATERERSUt9vZ+9dtu+vR9zPWZOWxCT/Xot5lwVcA7Rgnq/IZo9W4kZoljmvl+HGWI4J+ZY\numrdZ8+JOfYN0VFgE4FL07zXS5hhomDWuZvk7v8SM0INzM+0IeG6OzBB3feYoDK2qucSzOfbOZjg\n7LY0+3K8u/0tJusZ66/Ao+7+Tmner8dTJNG8bYmuATKtFfcJp7OLMX8ARURERCQDkazd8EHsMbnd\n5trZROsr3IWZ45bMLZghiQCnJZwLFyBJp9BKKn8FfkhyfAZm0W+A41pwv0sxQzjBDOccSTQBcRfw\nakL77Ykucn4r4E9x3zvd7fGkt35eeP7cmBTnrwNGYwJYzbVLgwK75hW420pa/g1LrM2YNDqY1LWI\niIiItIJt2wwZ1K+9br8rZugjmEImqWwGPnX3d2/D52e528S5brHC/dqJhBFhTWggOs3oDGC6+6xp\nmMxdop3dbR3wThp9yYm5pimfuNtLMEM4TyG+AMsmTBGXAGb0mzRDNfabF/4GphATkFVkeJ98opWL\n6lvbKREREZGezLZtamrrmL9oefONMzPU3TpEayWk8h2mpkJ7jMpKlSkEUyETzJy/0cQP12zKR8AN\nmGGTfTCfb49P0Tb8M3mAt939xORQ0D3vYLJrI9Low/XAYZiM3cHuy8EMM30V+C+wMI37iEuBXfMW\nYMYZ98cMy1yV4X0mYCbbOpjxySIiIiKSIcuCUMihrr7dvi8Pr0VVD9Q003ZLe3UCEzSlEjuarKUj\nwr5JuM/6FO36u1svsAvpDYtMtQxDrHLMMM+rMJXjR7v33s19/Q4z9+4UosuHSRMU2DUvBHyBGbt8\nEtH1RFrqFHc7l4R/ObO8XjyWRTAUanyViIiIiDTiOGauXbavNdPXmhSeS+aj+UWz+7jbhvbqTAr9\nY/Y3t+A6L/FLHowA/gWcnqRtpbtdAhyLSVLkJLSxMJ+ZA5hRakvS7EcdJmt4G2bJg8MwRVwOwfze\n98V89t4hzfv1aJpjl55n3e0FmH/hWmpv4Ep3//nEkwNLSsjPzVVgJyIiIpImx3HIyfYxasSg9nrE\nGndr0/wQy3AFl/nt0I9RTZwb4m43YNbaS9c9wDhMwuFy99hpwNlJ2s5zt30x1T7nYQq3xL6mAzMx\nQ0G/pHFVzXQsBP6GWWohG3jMPb49cHIG9+txFNil5xnM+hxgJo0e24JrDyG6hkkp8PfEBkV5eYQU\n1ImIiIi0mOO0W12NaUTnsO3fRLtiYB93/9s2fH44+9fUs8MJh2nEr1nXlKOBi939e4B7iVbXvI/o\n3MKw8M/Ui8br28XaA/N5+SVMMNaUMZgM4bukXrD9EqIBYjpVNns8BXbp+zlm/REwC0F+jElXjyNa\ntQjMNwzjMd96vI9JH4fT9z8DNsbeNMvrZcTAgdQ1bO3MvYiIiEjX5m8I0Kd3uxUbD2AKeIBZMHtc\ninY3YII7gOfaoR8Xk7zi5f5EFxx/Js179SK6jt3bRNeKuwqTgCigcRLiS+Brd/83pJ7KdR2wJ2YR\n9+bmJNqYz8WHkjobN4joUNO1KdpIDAV26VuEmcgZDsz2B/7tHl+FKbKy0N1fCPwHU90HTEB4EPBm\n4k2H9OtHcVER9f5US4KIiIiISDJ1dfUMGzKAvNzEKV9t5o+Ysvs5mPL8sdkzC7PG3K/c9/fSfPXM\nTIzALKcwJObY4cAb7v4HwFNp3usBTMAUIH5R802Y4A7gJ8CUhOt+424nYLKD28Scy8Zk345z319H\n83PsvgcecfdvAy5KOD+EaJD8A/BiM/cTFNi11HxMevo24r+J6I/5F3w8jdff+D0mfZx0/ZOJo0cD\n7TqMQERERKRbqqv3079vb3bYbnR7PWI9ZrhjBSYg+hhTpGQpprjKZW67x4jOVYsVjjg9Sc41J3zN\nR5g5fKswQ0M3YLJt+Zg5bSc0cW12zLFriRZHuYboIuVhTxBdnPx+4oPYd4D/c/cnYRIa64HlmATG\nz9xzDxBdqDwsXN0mK+H4xcBsd//vmM/WP2B+vlXAZMxw1DPQOnZpUWDXcn7g15jqR8dgSrE+BryF\n+cPwd8y6HPthAr6biA7hjDN84EBGDBhAZU0NlpVO5VgRERERCXMcqG9o4Ed77IjH024fa2cCwzDZ\nu7WYz4AjMcHTh5h6Cuc2cS1kth5b+Af6JWaI4zpgMNAPU9jlGszC5MnWWP4Bs0zCtJhjR2A+k/4T\nuDvFM68CPgNqiQ7zDHscMx/uSczyDv2JLuD+P/f+F9NYeB2+xDX2/JhF4G/ABJm5mCUP+mGGhd6P\n+b1/laKvksDqbpmirhQg/WSffRg2YADlVVVx/XaysrCrq6/zzZ57O8EAeDL5kkdEpBlVVWwaOJoL\nbzuUQ8bDhi1NLZUkIpIZX5ZFnd/hqQ8rf7WhPPiXvOy2/axmWTZ5uTYf/m82H/5vepveO4XBmLlo\nKzEBUHsYBqxw9w8jutzWMMwQ0BXJLmpCOFvWkqIOWZjgMFmFPxszii0b83tIt3BLU3oBA4EqokVr\nUupuMUxbUMaugwwfOJDhytaJiIiIZMyybCortxAMenjs/kd48cl/s/MO7b7k2RpMjYX2CuqaspKW\nB3VgArqWVuprIHlQh3t8BWauXFsEdWAWLF9IGkGdJKfAroPsOGYMDmiZAxEREZEM1dfXkp2dw+GH\nnILHm8ehB+7H9P/9j9uuvrqjuyay1Smw6wDDBwxgmLJ1IiIiIhmzLIuq6gq2m7AzQwaPpLR0I4uX\nLGVTRQUH7bkn+wETLYsiWx93pWdItQ6FtKMdx5p1GEOhkAI7ERERkQzU19fRu1cJE8btRHV1BZZl\nYds2fr+fNZs2UQMMyspimM/HukCApX4/5V1npFTsZ/R2W8tBuhcFdlvZ8IEDla0TERERaQXLsqiu\nrmT3Xfenb79BVFWV4/VGP9ZamPr41aEQvlCIQV4vA71e1gcC/OD3U9H5A7wNwPlAETCjg/siXYQC\nu61sxzFjAGXrRERERDJVX19Hr959mDB+EnW1qb8stywLB6gKhbCBAV4vA7zerpDBqya6gLdIWjTo\neCtStk5ERESkdSzLorqmkm3G7ki/voOoqU26XHD8NUQzeLWhEAO9XvbKy2NSTg69NAdPugll7LYi\nZetEREREWqe+vo7eRX2YMH4nauta9mV5bIBngxmi6fGYDF5DQ2fO4Ik0S19RbCXbDB9usnXV1Qrq\nRERERDIQztaNH7cD/foOojaNbF3S+0BkiGat4zAwK4vJts2BQ4ey5+67t2mfRbYWBXZbgce2mTxh\nAsFgkJDjdHR3RERERLqk+vo6evUyc+tMtq5194vN4JUGAgwvLOS1f/+bF/7zHybvtFNbdFlkq1Fg\ntxXsPnEifYqKqKqtVbZOREREJAOJc+tMtq5tPldZ7mtzVRWr167luCOP5KPXX+epBx9k0vbbt8kz\nRNpbTwjsSoCdgEnAju5rJ2AXIL+9Hz6kXz92GjeOChVMEREREclYNFvX8rl16bJtm2AwyPc//MDy\nVas45cQTeeeFFzj/zDPb/Fkiba27B3bXAJuAWcBM4Fv3NQuYDuzcng+3LYt9dtwRx3EIBALt+SgR\nERGRbis+Wzcw47l16bJtm1AwyIKFC1m/cSPXX3kl//jLXxjYv3+7PlekNbpzYLcjcIe7X48J8NZj\nFnwswwypbmjPDuy/yy70691bBVNEREREWiE+W7f1prZ4PB5qamtZ8P33HLjPPrzy1FNcedFF9Csp\n2SrPF2mJ7hzYhbNxfwZygH7AQGAA0Afzs3/VXg/fduRIth89mgoFdSIiIiIZSz63bus+37Islixf\njt/v5/ILL+TVf/+bKy68kMEDB+LpPOvgHQec09GdkI7Tndexm+VuZ2ztBw/u14+DJk+msqaGoNas\nExEREclYdN26Se02ty4dHo+H2ro65i9aREF+Ppecdx6nn3wyc+fPZ+ny5eTm5FDcqxdPv/gin06d\nurW7tzvwErAZeGxrP1w6h+4c2H0L/Am4DfgaWE206BHuthZo05UoC/LyOHCXXfAHAvgbGhTUiYiI\niGQonK3bffIB9Os7kNKyjR3+2Soc4P2wbBlZXi87bbcd++y+O8FgkIL8fPbbay/Ou+IKpn7zzdbq\n0kBMUAewcWs9VDqfTpM7bgfZmEqYI4HFQDXmW4xNQDlQBezZlg8syM3l+P33p1dBAVWqgikiIiLS\nKpG5ddt0bLYuGcuyCASDbC4rY/nKlaxas4a5CxZQ5/fz2L33stduu22NbkzAjE4bvDUeJp1bdw7s\nCoF9Y97bQK778sYcaxM5Ph/H7b8/hbm5lFVUYHee8dYiIiIiXU7c3LqS9q+E2Ra8Hg/rN26kpraW\nR+65h/332ac9HzcFmA8MwhQFlB6uO0cfpcB2wDhMwZQsTBGVIcBY93ib5chHDhpESVERW6qqFNSJ\niIiItFJ9fR29ivowYZv2W7euPXg9HtZv2kRFZSXPPfooPz3++LZ+xHBgEXC/+/4+4A9t/RDperpz\nBBIClmGGYW4gmqVbA/zgHq9rq4eNGTqU2vr6LvOXjoiIiEhnFcnWjduRfiVbvxJma3k9HsorKli7\nbh3nnnkmebm5bXn7H2ESFHXA+cClwLq2fIB0Td05sAv7OSawq8X8AfgeOKytH+L1eJQDFxEREWkD\nkWzd+EnUdaFsXSzbtlm3YQPjRo/m9JNOastbVwEPAyXAI+6x/LZ8gHRN3T2wuxb4J2YNO8d9jQXe\nAU5uywfNXbIE27bpen/tiIiIiHQecdm6vgOp6WLZuliWZbFyzRouPucc9t1rr7a67avABUBNW91Q\nuofuvNxBf+CPmCGZxwLvYwLZUzHre9wN/LetHuZvaMBxHLAscJS7ExEREclEbW0NgwYNZ699DiUr\nN4dedkmT7R3HwZeVhcfjwXEcepeUkF9UZL7Rb8FnshAmy1bUpw9FvXuT5fPh9XqxfT4CmGUOMhEI\nBikoLua/Tz3FWRdfzLtvv53RfUSa050Du4Pd7dnA6zHH/4lZAuE3mOUOvty63RIRERGRVIKhAL0L\n+rBl7QZWVpU3OwzTCYWK8wsLH87y+XqFQqHqQCDg3bhmTRWwwbIsG9IbUOWFkL++3rt2+fIBNVVV\n3tqqqpAvO3tg6YYN769YvPjGLJ8v45/p+2CQQf37c9UppzBz6lQ2lpdnfC+RVLpzYBf+2RqSnAsf\n6+5DUUVERES6lL6DBzP/m2k8/evb8eXmpjO/Lt+y7ePdIA4sCycYpDA/v0Vz8woLCqguL+c3p58e\nuc6yLIKBQJW/DQrkBYNB+hcUMAStIi7tozsHdh+628eA1cBn7vvjgd9hlkP4ogP6JSIiIiIpWJZF\nKBgkCATr/ekEVLWO49QABWCGX3q8XrJzclo0FNOyLEJOiPraWpyQExPcEciyW19Fwev1UFVXS/+s\nLPJsm5pQqNX3FInVnQO71cDtwPXA/zAVhGwgzz1/c1s+zOf1mr8ANL9OREREJGMhx8Hj8ZBnefDl\n5Zr6BU2zkpYmd5yWFbVzHDyWRV5uXuIZq60+39mA17Lw4m+T+4nE6s6BHcANwHpMEZUC91glcBXR\n8rBtYrtRowiFQlryQERERKSVLMDjvtK7oB3rkjsWaU7Ta/o2QJZlkWPZOCpoKe2guwd2AH8FHsAU\nTGkAlkPbxl8e2yY/N7dF6X4RERERaczBASxs939pBVXtud6U1XZxY4HHZn5tLZWhYNvcUCRGTyke\n0oBZmHwZbRzUAQRDIWYvXow3wzK4IiIiIhJlYT6kdoaXZYFjt+4VsqEgy8PqgJ9va5Wtk/bRHTN2\nM4DNwM+AR4ExmOGXsbxAEXAiMKstHlpZU0PIMRNtlbkTERERyZC7LLDHtrHt9NJl7ZmwC/epNbIs\ni1AwxKyqrrvYunR+3TGw2xlTRTYbmAQMbqJtQRPnRERERGQrMwMxwYOFh7aZ39Z6mUd2DlBo23xV\nXU2lKmFKO+qOgd2PMUMv1wL7AwMw2bnwYOYsTEZvC6ZypoiIiIh0Mh7LvDpFXNcKObZNeSDAwvr6\n9nxMeD5Qbns+RDq37hjYvROzv9h9JbKBYlBJIhEREZHOxnKzdXanyNi1LluXb1nMqq8n2L5Tdd4G\nCoE57fkQ6dy6Y2AXVgJsAp4Ffppw7gbMIuXbAIu2cr9EREREJIXwUEzbMq+OLooJZLxOca5tURYI\nsKB9s3UAM92X9GDdMbD7DbAnZrgl7v7dQC/M3xWVwBHuuf4osBMRERHpVFq8jl07r3dgZVBI3gEK\nLS/T6ysJqLCebAXdMbBbRTRwAxgBXJGk3Q/A1K3RIRERERFJj4MphJnVWebYOZkFdnm2TVkwyHf1\nte3QKZHGumNg9xiwHvgRcB1mOYM/YoqogCmiYgOvEi2o0molRUV4bFtLHYiIiIi0klnHrrPMsQsv\nmt6S9pBnW0ytqVG2Traa7hjYAbwBfIrJyn0JzG3Ph3k9HrYfMwZ/Q0N7PkZERESk2wuHQS1a7qA9\nYz+75ffPs21Kgw3MrVedPtl6Wp5X7joqgUeA44BjYo4/D/y5LR8UCAZZu2kTXk/6I8FFREREJLnw\nHDu7C74soMC2mV1bQ4OydbIVddeMXdjzwInALZihl9nA3u4xB7i6rR60eNUqxg4dimVZGo4pIiIi\nkjHHlCvxWNi2ZSbcNaPjB2tG5ds2mxoa+LZGc+tk6+rOGbudMAHcUuDf7rF6zNIHlcCvgN5t9bCQ\n09LR1yIiIiKSKLzcgacFr47O0iVm62bV1ipbJ1tdd87Y7eJuryB+kfJPgd8CdwLbA59t3W6JiIiI\nSEpOuCqmhWWll7FrzxDKzPJL7wnhbN3MGs2tk62vO2fs1rrbXZKc293dlm2lvoiIiIhIGiILlHeS\nV7rDPB1Mtm66snXSQbpzxu49zJp2NwO7Af8DcjBr3O0OfAF812G9ExEREZGUItPrOsMEujTitALb\nZn1DA98oWycdpDsHdkHgbEyAd6T7CisFzm/Lh/Xt1Uvr2ImIiIi0Ujhj540sddDRkV3zn+0coJdt\n835VlbJ10mG6c2AH8CEwCDgNMyQziMnUPQNUtdVDLMti7JCh+Ov8hAKh1t/Qdgg2BB1/ZTU0BMGr\nZRREpO05DXXUVdVR0icX31APg3O0FqeItD3bZxGoc8jJqcFb7eDzNT0TyMr2EvTaBEg/Y9euoZQD\nNBGsOUChbbOuoYFpytZJB+rugR3ABuCv7fkAx3HYUltF/4H9CNS0/lslx+PB9uR78rYfgxUK4djd\neSqkiHQUu74ep6g/n325ns3f+6ms6+geiUh35LXN99Sr14bsyjqHumZin2C9TU5lA0Uem1CaGbt2\nDewsC+zUX7I7QC9vFm+Xl+JXtk46UHcM7E4B6oD3gYOAXoA/oY0N+IC3gE1t8dAla9ew6+SJWNm0\nejhmMATZvXzWiDET8FgQ0t8RItIOcr2wPpDDiy8uoGb1Onx5OR3dJRHphkLBEB6vh37jhlpZOdmE\nAsEm29dn++m/uZa+WR78njS/3G63gMoBy8ayvaQKHws8HtYGAnydXrZuW+BnwK6Y9ZVnYpblmtaK\nTg7ALOcVdF/JlADvAl+nOF8MnAUc4N5vMfCS+5IuojsGds8CGzEFUx4EBjfRdl/aaLkDywmRHWrA\nHwq0OrCrC9kUhhoCk0Ob8VkODU5Hjy0Xke6oTyjAdHsQlG0kv3QpWeXd8T8JItLRnGAAy5uFb1Tf\noNfyEXICTbYPOQ14nSAey6xRl9YzWt/NFCz3/5NPtXGAItvizaoK6p1mp+PcCPwu4dh+wOXAP4Bf\nZNjJo4B70miXQ/LA7iDgHeLjgr0wAehM9/yWDPsmW1F3/K/4B5hlDGqBj4GRNJ5P5wUKaeN/SUPu\nq7V/uTgWBLGsmpAHv+UQUGAnIu0gK+RQh4UnOwsbC09ubkd3SUS6IScYxPJ6wbbNsMbm1qVz16+z\nnfTX5bLSWOsuY5aVcjCo17KoDgZZUNvsWPbLiAZ1jwJ/x3xsPAe4BLgIM+Lsygx6OMHdVgGrgawk\nbUZjigcmmoT57AymDsXNmCXD9gf+COyMGQW3awb9kq2sOwZ2h8Tsn7G1HtqnV5GqYoqIiIi0Urgq\npof0M3Ydpci2mVlby9qGJotPDQTucPd/Q3zW7lJgDmaU2RWYkWdftrAbE93tI7Q8MLzT3X4IHBxz\n/DvgE2AGMBm4DhPoSSfWHatyZAO5QIG7zUnxyqUNf/5AIIDjdHxBXhEREZGuLBzYZdkWWZaFN42X\npz1fmExIslehbVMVbHrOICYrlwOsIRrgxXoIszxXuG1LZGMqv0PLpxftSTQhkjhEFExw95sM+yUd\noDsGdquBGqAC2AwsB9Zh0tOVmD9U69w2+7bVQyuqawg5oeaHF4iIiIhIkyzMh9TO8krVR7/jsNqf\nWKOvkfBayi/TuKBf2OsxbZMNpUxlKCYjCLCwBdcB/NjdLsBMX0rmLXc7DjMfUDqx7hjYhYeXhqe8\n9cdUxgx/4VLsvq/FBH8iIiIi0ok4LXx1BAtocBzWBZosBtMb2MHdn91Eu/C5IcCYFnRjN3e7DJjr\n7g8AdsJ83m1KONM3t4k284H17v7EJtpJJ9AdA7ttMd9eeIGP3GOHYsq89sFU+AH4F6bSj4iIiIhI\nexgCFLn7TQVQq2L2t2vB/ce522mYDNxazMi0WZgigR8Be6S4NhxANtWvBszoN4DtW9Av6QDdsXjK\nWnc7AfgJZk2O92POPwWMB27CVP5Zj4iIiIh0CpZtEwoEqNtSjt/nS2eai+M4TmQtOycUwuPzkV1Y\n2KKidpZlEQqFqK+owAmFTKVNxwGPx7GyshqtlWcDftvGaXqOXVHMflML3W3EVHUvxow2S1e4WuWJ\nwMmY0WpLMCPVRmDWpfsSU1TlnoRr89PoF0Q/Kw9sspV0uO4Y2IXt6W6TrbgbPjYOBXYiIiIinUag\ntpbcPn0YeeCB1JHGUEvHse2sLI/l8YDj4M3OpnrjRjbNn4/l9aa9FEKwoQFvdjYj9tsPb3Y2oUAA\n2+OhobbWW1dZiWXHD3SzLYtePh++mTOhrCzVbWMDu6YWuqt3X5D+5/McooGdDdwA3B5zfgymyuZk\n4G5gKfCKey6faGDXXPWX8FoOnb1IaY/XnQO7ae72IWAUpoyrDzgOOB8ox6SpRURERKSTCNTVkzOg\nP0O2GU+1399sYOc4ju3Lz7c9WVk4oRA5vXqxfs4c1s+ejdfjSbuwXSgQwMrPZ8iuu5JdVESgrg6P\nz0dNaam9ZcUKPFnxNU1sy6I4N5eshQubCuxiP2s3FUDFLoWc7lQpL/AMprLln4B/J5z/AbPQ+FzM\naLXfEQ3svEQDtaZXjI8GpN1xCle30p0Du+8w31pcH/OKdQmNFy7PWLbP176LY4qIiIj0ALnFxZQt\nWcJnL76Q7lDMgOM4DThONpgF0b05OWT36oUTaipJFi8rN5dAXR1f/PnPhGKGYlper9/2+ZIOxSy0\nbaorK5u6bezK5U1Vu2xJoBVWBfyymTYNmCUWHsUUcdkTMzTTT7RCp6+Ze4TjhWbXdZCO1Z0DOzAp\n6c8xBVOGYwoYzQPuBb5tywftOnEcoWBIC5SLiIiItEIoFMTOyiKnd29sny+dqpdJI7+WBHUAjuNg\nWRbZvXs3ur+TJLi0Aa/Hg1Vd3dRtyxMuSSXPfUF0SGZb+SJmfyImsKvFLAM2oJl+QXTIZl2TraTD\ndffADuAN99Vu+vQqZGj/ftT5G9rzMSIiIiI9hmVbHTIaKukzkxyzwm2b7mNsOq93E+2GAgXu/vfN\ndLGlYjOAxTH71UmOJTPC3bZ0nTzZynrCWNnjgP9hxi1fhplo+keaTzunrbS8ki9nzyc/L1mdFhER\nERHpoRZiqlQCTGqi3YCY/WVp3nsUcBpwPE0na2KHgC6N2Q/Xo2iqXwUxfVvSRDvpBLp7YPdb4CXg\nR+57GzMk81rggbZ80PK166PjsUVERESkWwsBObbN+Ozs5pqG103etYk2+8e0XZFmF44C/gO8SHQR\n9GT2idmPXSR9urvdGUj1Q+yBWQca4Os0+yUdpDsHdiMxa9WtwKS3V2L+xV2EWbDxXKBfB/VNRERE\nRLo4G8izm/04/ZK7PRWTZUuUi6kHAfBaCx7/acz+z1K2gsvd7ZvED/N8HTOibQDw0xTXnuNu/wcs\naEHfpAN058DuQHc7BVjt7g9xt39wt9ts1R6JiIiISLdRHQoxwtfs7J5/Y4r3WcDzNM6OPY1JQpQB\nf01y/S8wQdlvif/sPht4yt2/Ejg7ybWvAdu7+79POLcK+Ie7/wiwe8L5XwGnu/t3Jrm3dDLdObDb\n6G7Dae8Q0bKu4ZT05q3aIxERERHpNmpDIYZkZTEoq6mVDIBo5msXYB0moLobM6LsWPfcuUBpkmtP\nBI7ALNWVn3DuQmCOu/848BXR5Q3KgJ/E3HtqkntfhlkizOte+xJmvbuviQZz99CyTKJ0kO5cFfN1\nYDFwC+YbiDxgNPAW8GNMSnl+R3VORERERLq2esdhSFYWhxUV8cTmJvMFXwE7YrJzEzEBWdga4Azg\n4xTXhpcZKMMsUxCrBpgM3IUJ/HYnPvM2Dzif+CUPYgUw8+gexQwVPc59gUmK/AoTgEoX0J0DO4AT\nMH9IjnTf98MMv1wN/F9bPqh/n954bFvr2ImIiIj0EDawORBg97w83quoYE1Dk0tfzcEMi5yICaZs\nTLGUmZggKpWfAYWYoC7Z4uUNwKXAjZgFyEcDFcA3pLdEQTVmjt1VmIKDfTE1Kb7ELIIuXUR3D+zm\nACWYNPSeQNA99irRYZmtVlSQx547bUdtXZvdUkRERES6gDrHoa/Xy2FFRTzedNYu7Dv3la4y99Wc\ncuCdFtw30RrguVZcLx2sOwd22wNPuK97MUMz28XQAf3oV9yLtRs3YzdfGUlEREREugkb2BgIsHd+\nPu82n7UTaTfdOQqZiJmgegtmccV2M++H5SxesZqigsT5rCIiIiLSnTlAocfDCr9fQZ10qO4c2D2L\nGXJZDBzS3g+bPu97vB4PWp9cREREpOewgCzgnYqKllyWgyns1x7yMWvjSQ/TnYdiQnTC50vAWsx6\nHQHMz12Eqf7zbZs8qKaWkBPC/PFWARURERGRnqCPx8Psujqm1dSk0/wcTEXMcOXKOZhpQ38leWGU\ndE0ErgAOwhRPAVP9/W3gL0TXdE7mOuBQYH2K83nAFkyBlspW9FHaWXcO7PoCx8e8H+S+YhVuve6I\niIiISHfjtSy+qEqreOQ7wGEJx3YA/oyp1r4PpgBKS/3SvUeibd3XlcAxpF6L7v+A8c08oxa4GgV2\nnVp3DuzKgaMx3zIEE87ZmKz5vLZ6mIqmiIiIiPQsObbNhkCAeXV1zTW9HxPUBYCfA/92jx8LPInJ\nuP2XxoFfc35KNKh7C7gc+B7zWXdvzPp04zHTk3bBLK0QawzRoO5VzAi3xCGi4Yxdsz+kdKzuHNg1\nAB9srYdNHDsSy7K0jp2IiIhID1Fo28yoqaE8mJhDiLMrMMXdPxNTByLsFeBwzALih2Iya6+2oAs3\nuNt3iK7bDGZdvM8wy30txKzlfD1wSsL1o9xtHWaYaFrrNUjn1F3TTHsCLwBfYb4FaS693CojBw9g\nh3Ejqayubc/HiIiIiEgn4rUs6kJNrS0OwNnudibxQV3Y1JjjP2vB43+EGcoJZmmvZMow2UIwGbzE\npM5e7nY6Cuq6vO4Y2B2N+QNyAmZi6pmYIZc7ttcD9540Ea/HQ7Dpb2tEREREpJuwgHrHYXF9fXNN\n93G3TS0e/oa7/TGmons6bGAxJmCc1US7de52gPuKNc7dzkjzmdKJdcfA7gp3OxO4FpN+9gC/ao+H\neT0e+pf0pt7fgKW1DkRERER6BAsIOg6VTWfshgCT3P05TbT7wd0WACPT7MKnmMBsF2BNE+3CI9fW\nYObQxdrD3X7ubncDLsIMy5yQZj+kk+iOgd1kd3sm8CfgXPf9nu3xsEAwyPTvFpGfm6P5dSIiIiIS\nqwQTA0I0eEtmJdFif9u14fMLMJ+JwYxoi41CBwFj3f3JmKIrXwN/xxRdmQ9MQwFel9Edi6f0crfL\n3W34D9FA2mmRuenzvmfyxPHk5eZQW9dsOl5EREREUrA9XoL19VRu3ESDxwPNj4hynFDICX/BHnIc\nfDk55PXta750T/OLd8u2CQUCVG/cSCgYjBTFs7OyHG9O4y/wbSDLsgg1NDR1214x+/4m2m1xXyWY\ntZbbyt1Af3f/rwnnxhBN8lyNqSj/OCarNwk4AlP4ZT5mfbyP2rBf0g66Y2CXqN3TaDV19cyY/z3H\nHrgPK9au15BMERERkQzVlpVSPGY0+x12KNUNDc1+kHMcx5uVl+fzZGXhhEJkFxWxcd485r/0Et7s\nbKw0l6RqqK0lu6iIfa6+muzCQgL19Xh8PmpKS31bVq7Ek5UV1962LIpzc8l/801Yk3IkZEHMflML\nkPsxFd2h7T6f3wyc5+7fjsnYxdo/Zv9+4JKE8zsCH2KCzacxmcTSNuqbtIPuGNj5AR9wFWYs8Uj3\nuA8z564Osx5HDvAQsKotHvrp9DmMHjqYbUYOZd2mUgV3IiIiIhnwZGXRUFND6eLF1DpO89/QO47j\n8flCtseD4zhk5eZSuXo1dnrZvgjbtnGCQcp++AFvbi6hhgZsrxd/VVWoprTU3C+2vWWR5fMRbLp4\nSiYfCNviQ+QtmMAO4BmiyyLEetF91hbgviTnZwMnY4K7AZjA77dt0DdpJ90xsKvFBHGJ/+JlY76t\niPUubRTY4cDjL7/Dz487XMGdiIiISIZ8BQVUrVjB/BnT8ft86QRnQcdxgpGhmMEgWbm55BYX4zS/\nFEGEJzubYEMDC155xQzFtG1wHCyvN2BnZzca0mkDFbZNXVVVU7etidlv6nN3Vsz5Jsd2puF+ouvm\nPQ2cnqLdfOD3zdzrI+B/wL7uSzqx7hjYvYz5VqGpP2VeIJ92WK/j8Vfe4f+OPZzxI4eydlMptoI7\nERERkbSFgkE8Ph95/frizfKlM6cm6YetlgR14faWZZHXt2+j+ztJPs/ZQLbHg11XB4GUoyy3xOx7\nUjUCConOratMq8ON2ZhlE37svr8XuDzDe8VagAnqtsEEoK0NPKWddMfA7ucd+nQHHn/lXc4+9jC2\nUXAnIiIi0pOVxewPbqLdcMyIMzCZtJbqDXwBbOu+/yVwVwb3SSZcrdND96yo323oH047cByHJ155\nh0XLVjGobx8tgyAiIiLSzTiYD9I5TX+BvxwzVw1MMZJURrjbBszSBy0xCLNGXjioO5Hmg7osTO2J\nV4FjmmkbXjD9e0Dl3zsxBXbtxHHg8ZffZeGyVQzs24eQgjsRERGRbsMBcmyb8dnZzTX9yt3+qIk2\nR7rbT4ENLehGoXvNUEwBwb0wRVGa04BZ6/lo4LIm2vUGDnX3p7WgX9IBFNi1IweTuVvoZu4U3ImI\niIh0H0HHIbv55RRecLeHY9aDSzQBONvd/3cLu/AoZpHx9Zis35ctuPY/7vZg4MAUbW4A+rj7T7aw\nb7KVdcc5dp2K48ATr7zL2cccysRxI1mzYTMhd3KuiIiIiHRdNaEQo7OzsYEmSrW8A7yFWfD7ZeBY\noot9b+ueB5gO/DPJ9fcBFwNvAz8hOuftYMxyBAA/AAdg1pqDxus452AqdN6DWYgc4C/ABZihnC8C\npwDvxVxzD9HiK78jOqRUOikFdluB4zg89foHHHfQPuy83Vi2VFThb2hQcCciIiLShdU6DoOystgm\nJ4f5dXVNNT0D+AYYjVkXbg1m6ORI93wZ0SAt0QR3uyemqnuF+z52COXe7qs5zxMN7CoxweZUzJDL\nd4GNbl9GYJYKA3gA+E0a95YOpsBuKwkEgzz/3qcsXbOOI360G8FQiFALy/CKiIiISOcRcBzyLItD\nCwubC+zKgInArcCFRCtk1mLWmruG1MtwzcYEbZ8A1Qn3XAmsTqOrBZiAcFPC8W8x8/Nuxqx31899\ngamy+TtMplC6AAV2W9n07xYxeuggdho/io1l5craiYiIiHRRFrAxEGDnvDx2yM1lTm1tU83rMJUo\nrwNGYT6HLwFSLoLnusp9Jfp5izucXClmyOXlmGUXCjHVPJtceV06HxVP6QBfz1lAQyCIx9PUOpUi\nIiIi0tkFMR+oDywsTPcSBxPQLaL5oG5rWwF8h4K6LkmBXQdYvmY985esoE9Rgda4ExEREenCIlm7\n3Fx2yM3t6O5ID6bAroN8NWcBfmXtRERERLq8IGBbVkuydiJtToFdBzFZu+XK2omIiIh0cRawsaFB\nWTvpUArsOtDXsxcqayciIiLSDYQXlzusqKhD+yE9lwK7DrR87Xrm/7CCPkWFytqJiIiIdGEWsCUY\nZEJ2NoOysjq6O9IDKbDrYF/PmY8/EFDWTkRERKQLCwH9vF6mVleztqGho7sjPZACuw62fO0GZi1Y\nTFG+xmOLiIiIdFXZlkVlKMT7lZUd3RXpoRTYbV35wP5AXuzB1Rs241XGTkRERKRLCmfrplVXs8Lv\n7+juSA/l7egOdAMnAycAuZgFJxMny9mY33MOcDCwDhi0NTsoIiIiIu1H2TrpDBTYtc6/gJ+18Jq6\n9uiIiIiIiGx94WzdR5WVytZJh1Jgl7kTiAZ1S4G3gW2BA4CFwPtAX2AnYILb7k7gxsQb5fh8jdJ8\nIiIiItL5hbN176afrRsO7Ab4gBmYz41tZXtgR0wi4UtgTRrX5GCmCQVTnLcw8WsljUemSSeiwC5z\nx7rbDzFDLAEGAmuBauCSmLY/BZ4GdgYafZXTuzBfyx2IiIiIdDEOUOzx8FFlJSubz9aNAp4E9kk4\nPg84C5jeiq4cCDyOCRpjvQn8HNjYxLUvAkdgArdkCoENwHbA5lb0UdqZiqdkbhd3+8eYY+uARe65\nwTHHn8H8oToEuC7xRms2bMZ8GSIiIiIiXYUH8DsOX9bUNNd0AvAdJqirBP6DCcRKMQHTN8DhGXbj\nVEyiYTiwAngQeBkIAEcCc4n/XJr4I+zg7hemeIXPKW7o5JSxy1z4G5E5CcdnAeOBXYFXY44/AdwL\nXEt8MMiKtRuorK7G6/UQCKTKgouIiIhIZ5Lv8bDU72dhXbMlFB7HFNqbiQnuat3jFvAu5sv/x4Gx\nmJFf6RoGPOruP4FJJISNBP4HDAUewQR5iXZyz4MZbTYbKElok4UZ2lnRgn5JB1Bgl7lU+fYF7nZo\nknMfA8dgAr9F4YMhx8FxUNJOREREpAspsG2W1Nc31+xkYA93/wyiQR2Y0ZzHA4sxU3qmYGoypOsy\nzHJaC4D/Szi3DJPN+xwz1PIAzGfRWOPc7WrgH6SeZyddgFKqmZvrbhOXLghPgN0tyTXl7jZx/LOI\niIiIdDEhoDYUaq7ZKe72VWB+kvNVmErrYILAdFmYYn4AT5G8sMkXwCdN3Htnd/sdCuq6PAV2mZvq\nbs9NOL7c3e6f5Jpd3W1tknMiIiIiEhYiukJwJ3zZDtQHQyyqazZjN8ndftlEm4/d7W6knwCYAIxu\nwb33TXJuR3f7WZrPlE5MgV3mnna3FwOvEF3S4HNM6nsUcH9M+79glkPYgpmHJyIiIiLJWJiyHnYn\neVlO0pfjvpowHjNvDuDbJtqtjdkfnbJVvHAhv3qSZwLDVrvb4UCvhHPbudvPMct0/RX4CHgLuBkz\nT0+6CM2xy9wcTBGU6zDz5lZjxkWDCejudN+f7R7Ld7f/IWFS7KB+fSjIy6G8qtmKSiIiIiLdmwOW\nBR4feLLSWjgtvonjYNk2lseT0eNDgUD8AcvCshsXQrABn9fGTnIuRmHM/vom2q3CDMkswASCH6fR\n1fC9t9D0cgbL3G0vTKAWDjC3BUZgfn/3EK2OGfZj4BbMZ9070uiPdDAFdq1zPWay6uPEFEMB/gyM\nAS4iGtCB+Qbk0sSbDBvYD6/XSyjkYKmAioiIiPRgoWAIj89LXv9eeGxvOoGdZVnRT1CWZREKBgk2\n+GnZBysHsPH68mLuDE7IwQmEGhW5sy3I8XqxvU0GkLEZsqYm41VjpuoUANlpdrh3zH2b+jVVxexn\nxexv624tTFD3T0yFzXpMcPlrzILnf8QsYH5zmv2SDqLArvWeAJ6l8bDWX2D+gJwK5ABvYBaJjDOg\nTzG7bDuOLZVVCupERESkx6uvqKBo1FB22eNIKuv9OE7ToZ3jOF5fXr7Xk+XDCYXILurFxvlzWPDq\nc3h9OVh2ejOPAvV1ZBf2YuJJPyO7sIhAfR0eXzY1pRuztqxYiicrK669bVn0Kcgjd+OLUJFyhYLc\nmP2Gph7vvsAMQk1H+N7+mGuTiX1u7Gf/ETH7B2ESEGHfYNZhfhk4FvgN5rPs12n2TTqAAru2kWrx\nkq9p5g+A1+vB6/Xgb2jqz7qIiIhIz+DLK6Bm4zoWv7WIWux0MnYBy7ICWCbTZdk2Qb+fnF6FWC34\n1tzjy8e2Q/zw7gtg2+CYMaFOMBhwGgKNsn+2ZRHI9uKvLE9xR9O32EuaaGcRzQk2W2Yz4d7h2Yip\nrot9bmybe4HnMQHiIpK7FLNUgg+zRp4Cu05MgV0HW71hE1/Nns8+O29PWUVlR3dHREREpGPZNqFg\ngIb6MgJWWkMxnbi0nuNgeTz48rJMcNZCDdVb4m4NtuNYXhJHO9qWZfrnNBmHxX64ayoTl4MJnuD/\n27vv8Laq+4/j76thySvOcvZw9iATAgESIGxKmWmhbChlFkpLaWkLlPZXWqCsMstu2XtDKaFAGQUy\nIIEkZO/EWY5jx9uypPv742hbsmXZTuLk83oePVe6uvfqSpaOz/eec76n6da3ZMd20XRg5425H3vs\nALC+mddYj2mpO5VoshbZTSmwS89lmCyXFa08Th6wEbgvdmXpjkqcTQ+8FREREdkr2EEbp8uJNycH\n23amNcauLV/f5UmveuywLLK9LhzOJl8+tjmvqbFz3YGuofubmtgu2bHdmKAwVUDYM+Z+U0lWUgkn\nfSnCtO5p2q7dlAK79FyNSVfbFtaTENiJiIiIiGHbNpYT3NlOGoJpBXa0cWyX7KwarXFYFi6vq7ms\nmBuJZrscQer55vrF3F+S5kmtCS27YaZIWJhiu0Gh5XKiLXRWzNKi6cnJw105K1FQt1tTYJeeJZjM\nQ+ErFg5MxqBizGSPnUPrlwPfhbZzYX5I+xHNiPQ58CrmyooG1YmIiIgkss04OZfXjSuQ1hi79pdk\nrJ5lQU6eh75dcpi3OWWnrjJgPnAwjacTiBV+bjPRgK058zGBVjZmPrpUgd2Y0HJpzDoX8D4wDfgz\n8PsmXmdIkv1lN6TALj0np1h/FnAiJnPQOaT+wv8auA0TAP6dhKAuy+3C4XSaK1RKjSkiIiJ7MTto\n43BYZHldNPh3j8DOSnEWBXluRvXM553Fm5va/T+YwO4U4Fcptvl+zLb1aZ5WKSaT5fHAdOClJNtk\nY+ajAzPpeFgD0a6hPyJ1YDcaODJ0/8M0z0t2kfTyv0oyBwHPAp8C+9P0VYzbgfOAfYDXE5/csq2M\nmto6nBlOpCkiIiKyx7BtLIeF0+PG6XGldXO1883hTX6rc1p065zd3Dt6MrQcgpkbLtE5RIOvfyR5\nfggwFVOPTGwBCG//o5hjxLoFKMRMYv5cwnOPhJbDgGuTn3pkmw0x92U3pRa7zP0uYdmcp4E/YlLG\n9icmC1FZRRV19T4cDotgugluRURERPZAtg1WqMXO15CqrWz3UGdbTBzYjS45WZTV+FJtthr4E2Yu\nuD8DA4C7MS1zPwZuCG33CPBxkv0fAI7FDAEajJm3LuxVTNbK74eWvwNeBnJDx/1RaLtrMcFdrCeA\nH4b2/SswEjMFQiWma+itmHGBAJdgJlGX3ZgCu8wdFlquasE+KzE/yBE0n15WREREZK9jkqeYxCQu\nR7K0Jcm081CWFNMm1Ns2w/oWcOzYPrwwa01TR/gD0An4BSZIuiTh+ZeAS1PsG+7SFZ4SITGCnI7p\nZnkEJkD7a8Lz1wGPpjj2CZhA8IeYIPPHCc/XAicBH6TYX3YjCuwyV4b5gfbGDHRNR3jw6fZ2OSMR\nERGRDs62weFw4PY6TGC3q5vsLNvckrBt8LnhwiOH8drX6/D5m+x6dTVmGM9PMMN4nMA3oXVNBU4P\nA4uBRUBNkud9mHFwJ2DyP4zFZOL8EvgnsKCpkwJOwwwxOgszV10BZsLy9zE9ztRS10EosMvcDMzV\nll9jfgjNuQrTWrccmBv7hNfjxuV00hBoKtOsiIiIyF7AtkNZMV24LDvNwG7nT3cQtr3Gz37De3LG\nlCE89cny5g70VejWEq+Ebs15J3TLxJehm3RgCuwy909MYHcmpmn8MmBrku08mP7U4SxI9ydu0L9X\nD/Jzsykp25H4lIiIiMhexQ4lT3F7XbhIL7Br/0a91IGjbUOl3+bCY0fx/OcraWi61U6k3Siwy9xM\nTGrYm4BTMSls52PmsavE9IEeBkzCBH4AL2AGpcbpWpBPQFlTRERERMw8dk4HzuwsnATTa7Fr58jO\nbuYFSmrrOWCfXpxzxHD++X6684uLtC0Fdq3zZ0zXykeBfGB86JbM7zDZheL06t6VfUcPo6IqWZdp\nERERkb2LHTQtdi6vE1fQ2vVdMW2wmjkJ24YqX4BLTxjLsx8tx+fX8BrZ+RTYtd6LwBuYaQymAaMw\nk0EGMeltP8PMXVeWbOeDJ4ymU24OG6u34XBoWkERERHZu5nkKRZujxtXILDrA7s0j7+1toHJE/tx\n/nGjePSdhe18PiKNKbBrG/WY4O6Nlu64saQUfyCAZbV3gSQiIiLSAYQnKPe6cQYcaQV27Z46JUVW\nzFgOGyp8DVx++gSenLEYX4Na7WTnUmC3i838djGjBg+gT2E3qmvrdvXpiIiIiOxSdigrptvjxtWQ\nZlfMdr8+nt4LbK30MXHiQH5y0jgefHVeO5+TSDwFdukZD/QENmDmEAEYFLq1ZHCcGzMXSNx0B4tW\nrGVwv94K7ERERERCE5S7s124/enNY7crs2LGsm3w1fu54pxJ/OOt+dSr1U52IgV26XkJGA58DBwe\nWnctZoqDlloPDIhd4Q8G2/9Ck4iIiEgHYNuY5CnZbpwNu8EE5QAtGDKztaqWfSYXccnpE7nv2ZZO\nWSeSOQV2IiIiIrLbsG0bh8OBy+vGVZ9mYNfewV8LrsDbtg01Pq688CAeeWmeWu1kp1Fgl57jga7E\nT0D+J+ARTOKUdLmA2sSVDsvaCV0IRERERDqAUFdMl9eNsy69Ccp3t4rUtrJahh84iEvP3Z97/zFz\nV5+O7CUU2KVnZegWa1Po1mqjBg+g3tfQFocSERER6dDsoOmKideFy5PmBOXtyaJFXTEh1GpX38DP\nf3oojzzzFXU+f/ucm0gMBXa72L6jhrLP0IEUby3FoSkPREREZG8XyopJdhbO2lCQtMu1vI5WVVrL\n4P2GculFB3PP3z9th3MSiafArm30B0qJz5A5BLgeGIHpwvkw8F7ijtsrKqmsrsXtdBIIBnfGuYqI\niIjstuzQPHZ4XLg9QYK7QVrMjA5v29BQxzVXTOPhxz6nzqexdtK+FNi1Tn/gdWA/YCSwNLR+HPA1\n8Z/vKcD9wM9iD7CmeAtLV69n9NCB7KisbvcTFhEREdmt2WA5LfC6cXqDWMFmwyoLcLTjGVmZpi8P\nVFTTf/RQLr3sUO65979te1YiCRTYZc4NfAgMCz0eRDSwu53oZ/s6kA0cB1yJmcPun7EH2rBlGxNG\nDcVmJ8yvKSIiIrIbswNBHFlZgBdnViCdiM2inatQdsZHtwEfv/nl0TzyyGfU1mmsnbQfBXaZuxQT\n1NUDVwDvh9aPBI4J3b8JuDF0/2rgLkxwFxfYrdu8larqGlxOB4GAumOKiIjI3ivL46R0Ww1gQ44L\ny9ds3Wg7cCjgAdo4crK94NxiWS4y7+/po/fAgYzcZyDzvk7MxSfSdhTYZW5aaPkz4PGY9ceFlkHg\nHzHr/wb8BtgX04VzffiJmtp6GvwBLCVPERERkb1c98JsvvikmM8/XM+UI4dAcEdzWSn9QDvOBG7H\n3DLZ3QKrDgK+tjwpkUbasz/ynu7g0PJfCeuPDi0/BtYkPDc/tBzRPqckIiIi0rE5nRbBILzy4nKg\nATzOXX1KrWNBqwJDkTQpsMtcbmgZ2z+gG6YrAMAnO/d0RERERDq+YNCmV59cZn1ezGcfrAVnHrt+\nMjuR3Z8Cu8ytCC0LY9YdBuSF7n+YsL0T0w0TYFU7npeIiIhIh+Z0WliWg9de2kNa7UR2AgV2mQvn\nrL00Zl34/hLg84Ttr8W06C0jIbBzu5w4HA610IuIiIhgWu1698ll1v82qtVOJE0K7DL3dGh5BfAB\nJpALZ8N8Pma7KcBHwM2hx7EJVQAY0LsHnfJy8Ac0caWIiIgIhFrtHA5ee2kF0ABZarUTaYoCu8x9\niwnqAI4kmkzlI+BPMdudChweuv828NfEA/Xr2V3THIiIiIjEiG+1WwcutdqJNEWBXev8HRiLmcrg\nSeA8TJAX61ugAvgxcFLiAboU5DN0YF+qa2vb+VRFREREOhbTamfx2svLUKudSNMU2LXeQuCXwAVE\nu2fGehooAJ5ItvOUCaPp0imPOl9De52fiIiISIcUabX7LDTWTq12IikpsGt7BcAgTKKUZhX17UW9\nrwFNTS4iIiLSWHis3esvhzJkqtVOJCkFdm3jQExXzKVAMSbrZTGwGngVODbVjnMWLsOT5d4Z5ygi\nIiLS4YRb7WZ+prF2Ik1RYNd6LwFfYsbXDSc6cbkHKAKmA+8BnwKdE3eeu2g5G7eWkpPt3RnnKiIi\nItLhqNVOpHkK7FrnHeC00P15wP9hkqScggn0riM6UfkhwEwgLoJr8PtZtX4TuV6PprETERERSSLa\nalesVjuRFFy7+gQ6sPOA74funw88lWK7W4DJwAxgBHAT8OvYDTZs2UbQtjXOTkRERCQFp9PCshy8\n/vIyDjlqgGm1a9B0USJharHL3Cmh5c9JHdSFzcLMZwdwEaabZsSmku00NPhxOdWtQERERCSZYNCm\nd1+NtRNJRYFd5sIJUZoL6sL+i+mK2Rk4IPaJbeU7+GbpSrp0yiOoAkpEREQkqUir3Uua104kkQK7\nzPkTlumoDC09iU98Pm8RldW1eLPcaLCdiIiISGPxrXaa104klgK7zP07tDwsze2zgCmh+3MSn9y+\no4Jvl62ic75a7URERERSiWTIfEkZMkViKbDL3M2h5UvAgDS2vwfIAX4L7Ei2wewFSyivrNK8diIi\nIiIpxLbafapWO5EIBXaZmw9cgAnWVgG/B4YRn2m0ADgCM+XBZcBy4FmgHzAOGBO6TQJ6lpZXMG/x\nSvJzsnfSWxARERHpeJxOCxwO3lSrnUiEArvM3Qo8BtQBTuBPwDKgHNgKlALbMEHdEaF9+mKCu/XA\nt8CC0G0OcDVARVU1Dof+LCIiIiKpBIM2ffrm8sX/NvLph2q1EwEFdq1RgGmd8yaszwUKga40nicw\nJ8n2IiIiItJC4bF2b7ywDLAhJ0vBnezVNEF55m4Fngdq2+BYOcBKgJxsL0Fbk22KiIiINCUYtOnb\nN5eZX2zmqQfmcN4Vk8FrQ10DWNauPj2RnU6BXebWhm5tJi8nm/EjBlNdW4dG2YmIiIg0zeGwKOyR\nzV13fIPTaXH2ZQeAt1rBneyV1BWz7bmBTpjpDVpk8riR9OjWmdo6X9uflYiIiMgeJhi08XpdDBiY\nz+1/ncezD88GRy543eqWKXsdtdi1jSLgCmAqMBjIA2qAdcAiTJfNd5s6QG6OlwkjhlBRVYNDV5h2\nOQvNEy8iItIRBIM22dmh4O7WeWCjljvZKymwa71fAbcnWZ8DdAf2Bc4B/gucgAn4Gjlo/Gi6de7E\n5tIysJSyd1cI2jYOy6JHvgeP00FDIIil8K7dWYDT5aS8poEdtQ04HPoHLLsPG3BZFh7Atm2CQY2B\nlsacDgd+y6LetlEJtmskBnc2Fudctr+CO9mrKLBrnRuAm0L3yzGtcguA7ZjsmEOBozHz2x2Omfrg\noMSDjB4ygMMPGM+28grTWqdYYqcL2ja5nizysyxe+HA2MxaVk53tpEcnNw0B/UHakz9gQ4OPs6aM\nYP99hrG2tAKn/gHLbsAG8h1OttTVMNvnIwh4MPPbqFSQMBsz79Fgh4OxuXlUYy4CyM4XCe6K8rnj\nr3PBsjnnUrXcyd5DgV3mxhAN6p4ELgRSXcq9APgncCDwaxJa+Drl5eJwOAgEAlgqdHYJl8NBntvm\n+uc/p3LkUVzy0Lk4bBuX06kr9O3Isiw8Hi/rijfyx9v+yCXl33DiwePYsL1KXZJll8u2HKypKGdF\nbi43/ulP7DN2LIGGBhwOhyruEuH1eqmorOTmm//Cl3PncWCnAqrUcrfLxLbc3RHqlnmOumXKXkKB\nXebOCy1fwQRuTXkCCABPYQLAuMBu+dpitpdX4na78PsDbXyako4e+R4e/tfn1Iw9nv+988quPp29\n0pRDD+Xnpx7O+IGbyelUgM+vgFp2HQvw1dexDHj8xRc58/vf39WnJLu5k046if3HjGHZsmUU5Xei\nQcH/LpPYLdOy4Gy13MleQFkxM3dEaHlXmts/jZkeYSQwPPaJYNDW1d9dyAIsO8DyMj8/vuCCXX06\ne60jJo5kzMFH8NnCleR7dM1Jdi2HZbHeV8+YMWM5Q0GdpKGT280PzzqLlZhxmbJrhYO7gUX53H7L\nPJ55eI6yZcoeT4Fd5oaFlqtbsM+y0HJAG5+LtIYFDf4AQYeT/Py8XX02e7X8gs7U+vw4HSqaZNey\ngHogr6CTutRJ2nJzc8yYDAUOu4W4MXe3zOXZhzQVguzZdFk8c2WY+eo6AZvT3KdTaFndLmckGbMw\nVzl2t5bTHTt2UFlZSadOnejUqVPzO6TBtm22bdtGQ0MDhYWFuN3uNjmu3+9n69atZGVl0b1794yO\nEQyYoG53+zvI3skBBNuge3xDQwNbt27Fsix69uyJ07n7ZT4Olwk9evRot/Pz+XyUlJQQDAbp3Lkz\n+fn5GR0nGAxGjlNYWIjL1TZVmerqasrKyujSpQu5ubkZHUNF1+4ntuXutlvnYQPnXDYZvFXqlil7\nHAV2mfsEM87uLOCPaWy/HzAZE9R9236nJe1t4cKFXHrppeTl5aVdoVi3bh0XX3wxV111VbPbvvHG\nGzz77LN89913bNiwIRLY9evXj7Fjx3LhhRdyzDHHtPi8n3/+eZ599lkWLVrEhg0b8Pv99OnTh6FD\nh3LyySdzySWXZFSZeeaZZ3jqqadYtGgRGzduJCsri/79+zNx4kR++tOfMm3atBYfU2RP8tBDD3HV\nVVfhdrtZs2YNffr0Sbntpk2buOiii6itrSU7Ozuj19u6dSuTJ0/m/vvvb3K7//znPzzyyCN88803\nrFu3Dr/fT79+/Rg5ciSnnXYa5513HllZWRmdQzK/+93vuOsuM3rhhhtu4Kabbmpmj6hgMMjTTz/N\nCy+8wOLFiykuLiYYDNK3b1+GDRvG9OnTueiii/B4PC06p/Xr13Pffffx0UcfsWrVqkhgN2TIEI44\n4giuuOIKBgxQJ5uOLhi08Ya7Zd46D1BwJ3smBXaZew0T2P0BmB96nEoh8Gzo/qskzGXXu7AruTle\nKqqSTnEnu5kZM2bwxRdftHi/2bNnN/n8ggULOPXUU1m5cmWj5yoqKli0aBGLFi3ixRdfZL/99uPN\nN9+kb9++zb7uihUrOP7441m+fHmj54qLiykuLuaTTz7hhhtu4JVXXuF73/teWu9n/fr1HHXUUSxb\ntixufX19PStWrGDFihW8/PLLnHzyybzyyittdlVdpCNZuHAhV199NWBa7gKBplsAKysreffdd1v9\nuvX19SmfKysrY/r06Xz88ceNnlu3bh3r1q3j/fff5ze/+Q0vvfQSRx55ZKvP57333osEdWBa+NO1\ncOFCjj/+eNavX9/oufXr17N+/Xo++ugjbrjhBt58800OPfTQtI578803c/311zdaX1ZWxldffcVX\nX33Fbbfdxm9/+1tuueWWtM9Xdk/hlruiolC2TJQtU/Y8GsiSuTdDNzDB2tvA6cB4YAgmScoxmOQq\nq4ERmNa6PyQeqH+vQtwuF8Gg+nB0BCtWrGjzY37wwQeMGzcuLqjLyspizJgxHHrooQwbNgxHzLiz\nr7/+mhEjRrBkyZImj7to0SLGjRsXF9R5vV723XdfpkyZQteuXSPra2pqOP7443n22WeTHSrO6tWr\nGTduXFxQ17lzZw4++GAmTJgQ173zzTffZOrUqepeKXudkpISjj/++Egw53Q6437HybTVlDepXmfb\ntm2MHz8+LqjzeDyMHz+eKVOm0LNnz8j67du3c9RRR/HUU0+16lyqq6u55JJL0jq/RHPmzGH8+PFx\nQV12djaTJk3i4IMPpnPnzpH15eXlHHbYYbz11lvNHvecc85pFNR16tSJAw88kEmTJpGTkxNZf+ut\nt3LUUUeldb6ye4u03A0yLXfPPKgxd7Jn0SX01jkN0yXzIOCE0C2VAHAcsCZ2pcvpZHC/3tTU1uli\nUQcxc+bMyP0TTjiB008/nc2bmx5mWVpayuGHH570uU2bNvHDH/4wbt2vf/1rfv/738eNQdmwYQPX\nXHMNL730EmAqS9OnT2fRokUpX/ecc86htrY28vjSSy/lrrvuiqu0PPjgg1xxxRWRwOu8885j0qRJ\njBgxIuVxTz/9dMrLyyOPb7/9dn71q19FHm/bto3LLruMV199FYBZs2Zx2WWX8fDDD6c8psieZOPG\njUyePJkNGza0aL/CwkJuu+02fD5fs90gbdvG6XQydOhQ/vznP/PVV19FnrvmmmuS7nP22WfHBUk/\n+clP+Nvf/hZX1rz11ltcdNFFlJSUAHD++eczceJExo4d26L3EnbppZcmbW1rTiAQ4Oyzz46bS/Tq\nq6/m1ltvjfts7rzzzrjy58wzz2Tp0qX069cv6XFvvPHGuAtYTqeTO++8k6uuuioSWNfW1nLjjTdy\nxx13APDhhx9y7rnn8vTTT7f4fcjuJRi08XpDLXd/nYdtwbnqlil7CAV2rdMAHAxcAlyNaaVLtA14\nEfg/oCTxyQPGjqRvj26UlO3Q5OQdwI4dO1i7dm3k8c9+9rOMxrvFuvnmm9mxY0fk8SuvvMIPfvCD\nRtv169ePF198kREjRkTGpixevJi//vWv/OY3v2m0/VNPPcW8efMij//0pz/x+9//vtF2l19+OcOG\nDePoo48GzFiWW265hSeeeCLp+T744INxFciXXnqJ0047LW6b7t2788orr/CLX/yCe+65B4BHHnmE\niy++mEmTJqX6KET2CP/5z3846aSTqKura/G+nTt35te//nWL9lmwYEHcb/3666/n3HPPbbTdCy+8\nwPvvvx95/H//93/ceOONjbY76aST+PLLLxk/fjzV1SbX11/+8hdeeOGFFp0XwBNPPJFWL4BkHnro\nobjeBn/729/4xS9+0Wi7a665hiFDhnDqqacCpvfBX//6V+67775G2y5ZsiRubJ/H42HmzJlMmDAh\nbrvs7Gxuv/12Jk6cyNlnnw2Y8cQnnHACP/rRjzJ6P7L7iG25u+PWeVioW6bsGdQVs208AowC+gNH\nYlrmjgPGAT2AK0kS1AH07NYZl8ulbmodxJIlSygrKwNMl6lRo0a16njV1dWRFjiAiy66KGlQF+tP\nf/oTBx10UORx7P6x3nvvvcj9MWPGJA3qwo466iguvvjiyOMPPvgg5RiY2MrS2Wef3Sioi3X33XfH\nVZgeeeSRlNuKdHR+v59LLrmEY445JqOgLhPr169n6tSpke6eRx55JH/+85+TbvvPf/4zcn/MmDFJ\ng7qwIUOGxAWY7733XqTsS9eqVau4/PLLAejVq1eTSWOSiS3DDjzwwKRBXdgpp5zCGWecEXkcG8DG\nSuw18OSTTzYK6mKdddZZ/PSnP408jh0nKB1bXMvdrfN4Otwt06NumdJxKbBrWxuAj4AZodsCoMnS\n4fN537GjsgpPVtuknJf29c0330TuT5w4kf79+7fqeJ999hlbt26NPD7//PPT2i82mJo7dy5r1qxp\ntE3s+LsDDzyw2WMeccQRkfvFxcXMnz+/0Tb//e9/Wbx4ceTxj3/842aPe9FFF0Xuv/baa1RUVDS7\nj0hH88ADD1BQUMCjjz4aWTdx4kR++ctftuvrXnTRRZHfVHZ2No8//njS7TZu3Mgnn3wSeXzmmWc2\ne+zYjLY7duyIa6lPx09+8pNIgHvnnXe2OEPu0qVLI/cPPvjgZrePTfKybNmyRomdAP79739H7h9w\nwAFptb6Fg1MwSbBiu+NLxxabLfOOv4aCO6fG3EnHpcBuF9u6vZy5i1bQKcM5c2Tniq0ojByZrOdt\ny8QGX927d2/yynGsiRMnxj0Od5eKFZsVL52MlInz5MWOzQuLrRgOGDCAww47rNnjhrt4ghlrGHsM\nkT3B3LlzufLKK6mpiWY2PuKII5g7d26TLdqtdf/998e1TN19990MHDgw6bb5+fk88sgjXHPNNUyd\nOjWtICs2MQm0LJPlzTffHEnQ8qMf/Yizzjorrht7c/x+Pz6fL/I4nbn1mivD1qxZE9e1c/r06Wmd\ny5gxYxg3blzk8QcffJDWftIxhIO7okH53PnXeTz9kFrupOPSGLv0/B34HvAcEE6jdTlwGZD+fyro\nAiwDfhK7snRHBV6PC7fLia8h/X+csvPFXqndb7/9Wn28008/nbFjx7Jhwwbq6+vTnrcqsYKUbJLx\nkSNHRhKrfPfdd80e89tvo9Mr5uXlMXz48EbbzJo1K3J/1KhRaQWMQ4cOZeDAgZFznjNnDieeeGKz\n+4l0RL179+bBBx/k5JNPBkxypPawefNmrrvuusjjww8/vFHmyVj5+fmcd955LXqNxFb7dKZXAfMb\nD2ec7Ny5c6T7Y1NTMCRyuVwMHz48Um6kU4bF9qjo0aMHw4YNi3t+w4YNcYlYmkoQlWjw4MGRzyO2\nJVH2DOFumQNjpkI4NzLmzgcacicdhAK79BwJFGESpYSNi7m1RKPLqVt21PDht6sY2qMTnXO9bKtp\nyPhEpf00NDTEBVTHHnssYCb+/vDDD1myZAmVlZUMGDCAcePGcfrppzN+/Pgmj9mnT58WjzsBk6Et\nrF+/fhQVFTXaZtq0abz2mple8bPPPmPGjBmRc05UW1vLgw8+GHk8ZcoUCgsLG20Xm+Ev3cDW4XAw\nduzYyGeXrHuUSEdWXV1N586dueWWW7jssst2ymveeOONVFZWRh7ffvvtbf4ajz32WOT+vvvuG9dq\n1ZTY7tcPPfQQBQUFGb3+YYcdxn/+8x8A3n33Xb744ouUXTK3b98eN4b3kEMOicv+C8QFdZB+oJpo\nwYIFGe0nu7dIy11RPnfGBnceG+oDCu6kQ1Bgl54XgMOAd2LWfYZJllLcguMUYOa0i8jKyqJz126s\n3lxKSdkOpo4aSJ7XS7Chcdc62bUWLFhAcbH5cxcWFvLGG29wwgknNBrfNn/+fN555x1uvvlmjjvu\nOJ566qmkQVKm1q5dG5ed7rDDDkuaFv3KK6/k4Ycfjlzp/uEPf8iMGTMaVYxqa2s55phj4tKR/+53\nv2t0vOrqarZs2RJ5nNjtqSldunSJ3A+nUBfZUxxwwAFs3rwZj8ezU15v9uzZcWP5LrvssjbpQRDr\nrLPO4tNPP408bipxSaxf/vKXkZatCy+8sFUZJK+++moee+yxSBl70kkn8cEHHzTqsl5eXs7RRx9N\naWlpZF2yMiyxnCwuLmb//fdP61xiu3Vu27aNYDCY9lx80nGEg7sBjVruaqE+qOBOdnsK7NLTaFJx\nTLfM51pzUIfDwahRo8jLy6Ouro7qeh+fLVrDAaMG0y8vi6BtqxDZjcSOhystLeXGG2+MXAF2OBy4\n3e5GXY3ee+89Bg0axKeffsq+++7bJufx61//moaGaKtuqgQmlmXx/vvvM23aNJYvX05VVRVTpkzh\n2GOPZerUqTidTlavXs2LL74Yl9Dk+eefTzp2rqSkJC4oa0mwGnvlPDZZjMieYGcFdGG33XZb3Gvf\ncMMNrTpeeXl5pIyYMWMGjz32WFyQdO211yadPiHRu+++y9/+9jcAioqKuP/++1t1Xjk5OXzwwQdM\nmzaNDRs2UFpaysSJEznxxBOZPHkyDoeDZcuW8dJLL8WNb3z77beTBrqjR4+ma9eubN++HTDdz085\n5ZRmz6O6ujpuOon6+npqa2vJ1dj4PVIwaJMdSqhiWu5szr3sIMgGMyVxJtzmZuligLQvBXa70MiR\nI+nSpQsVFRVYlkWe10NVXT3/W7yWY0f1wtPdhR3QmLvdxcKFCyP3wwHd4YcfzrXXXsvYsWPxer1U\nVlYyc+ZM/u///i8SCFZXV3PkkUeycOHCjLv+hN155528/PLLkcc/+tGP4jLBJerTpw/Lli3j73//\nO1dccQUAM2bMYMaMGY22Pfnkk3n99ddTzqeYmM2ya9euaZ93t27dIvd37NhBQ0ND0nGBItK0L7/8\nkldffTXy+PLLL291uXLvvffyhz80vn6Zm5vLjBkzmDJlSrPHKC8vj+uC+eijj6Y9ZrgpQ4YMYf36\n9dx+++1ce+21gAnc3n777UbbnnnmmTz3XOrrrZ06dWLs2LGRBE7PP/980ved6J133om7IOXz+aiv\nr1dgtwcLB3cDivL52+3fEgxanHTqEEq3ZTaNSVaWi5zsLGxV6aSdKbDbherq6iItL7Ztjwe+yfN6\n2FxR61u+sTRwUs8cuyGgNrvdRWI2x4ceeohLL700bl23bt0oKirijDPO4JJLLol0lyovL+eKK67g\njTfeyPj1n3rqKX71q19FHvfv35+HHnqo2f1efvll/vWvfzW73cqVK3n88cc599xzk7ZAhOfJCmtJ\nK4XX643cr6mpwefzKbATyUDsuDe3290m0ymsXLky6fqsrCw++OADevXqxZAhQ5o8xqWXXhpJFPOr\nX/2Ko446qtXnBWDbNs8991xkrF1TlixZwlNPPcXZZ5+dMovm6aefHinLly5dyvXXX89f/vKXlMes\nqalp1K3Ttu1G4/VkzxNtuevEPx9bxHNPL8143nLbhrycbPJcPYB1bXqeIrEU2KXnHkySlNLmNmxG\nZ2AFJpsmO3bsoEePHvj9fhwOR5XL5frCsiw/llXmdDpGAK3Ppy9t5pJLLuHggw9m1qxZnH766Y2C\nukSPPPII69evj0yy++abbzJz5sy05pRL9OSTT3LBBRdEHluWxVtvvdUoHXmsYDDIkUceGUk5DiYY\nPO+88xg1ahROp5O1a9fy8ssv8/XXX7Nw4UIuvvhibrrpJj799NOUadNjzyFdsdsGg0FspZAWabF1\n69bx9NNPRx6feeaZrZ5LM6xPnz4ceuihVFZWMmfOHLZu3UpZWRl//OMf+eMf/8jdd9/Nz3/+86T7\nPvroo7z00ksAjBs3rs0SudTU1HD44Ycze/bsyLrBgwdz7rnnMnz4cBwOB6tWreLFF19k/vz5zJs3\nj/PPP5+bb76ZTz/9lB49ejQ65uWXX859990X6VFx8803U1NTE+lCGmvz5s0ce+yxrF69OlKG2baN\nw+FIa/oF6fiCQRu326JbNy9+v51xYAcQsBsYOXgQ28pLWbJ6dfM7iGRAgV16jgMa537PTOQ427dv\np7y8nOzsbAKBwErLsqZU1foo7JTL8D7dfuvzV97SRq8pbSDdycNj3XjjjZHADuD1119vcWB37733\nNqpQffrpp83OeXf66afHBXW/+MUvklZefvOb3/Dcc89x9tlnA6byeOSRR/Ldd9/FtcolJgqIHefX\nnNj5qLKzs9VaJ5KBp59+Ou5319zFpXQEAgEefvjhuFZ1gIcffpgrr7wyMnfdL37xC7p27dporN2K\nFSu48sorI49jk7q01vTp0+OCuhtuuIGbbrqp0XbXXXcdjz76aGS6h6VLl3L00UfHTeESZlkWTzzx\nRFw5fPfdd/Pyyy/zwx/+kGHDhtHQ0MCXX37Ja6+9ht/vZ5999uHUU0/lz3/+M2BaMhM/L9lz2TY4\nHBZZWa3rQGXbNnX19UwcNZoNW7ZQFTMuVKStKLBLz8vAaKCslceJy4pp2zbbt29n4MCB2LZNZa2P\nghwPk0b0J9cDPr+fLI2z7dAOOugghg8fHknx39I02ddccw133XVX5LHH4+GTTz5h8uTJTe739ttv\nx43Dueqqq5IGdWFnnXUWeXl5kbm3Vq5cyR/+8AduvfXWyDaJ42WqqqrSfh9lZdGfTteuXXd6sgmR\nPcGLL74YuX/IIYekTP3fEk6nM2nr06WXXsqECRM46KCDIi3sv/3tb5k+fXrc2LKLL744cuHmtttu\n44ADDkj5WrEt981llHzuuefixgJff/31SYO62PPIzc2NXKCaP38+f/nLXyLz6cWaPHky//vf/zjq\nqKOoqzNjpoqLi7nnnnsabZufn8/s2bP58ssv4wI7lWHSUpZlUVNXS6e8PI6begjvfPIxdS2Y21Ek\nHQob0nMDMB0zsXhrbj8Efh17YF+Dn6o6H2VVtXTJ9TB1VH+8bhfV9T4crWnzl91G7Fx2y5cvT3u/\nk046KS6oKywsZP78+c0GdWCSAoT16dOHm2++Oa3Xi53A+KmnnopraevRo0dc18+WTLy8Y8eOyP3Y\nRCoikp5Zs2bFXRj6wQ9+0O6vOXnyZP70pz9FHm/cuDFuvO5f/vKXSK+AQw89lF//+teJh4gT2wW7\nuTFqsVO6DB8+PO48UjnrrLM49dRTI4+feOKJlNtOmTKF4uJiLr744qSBrdfr5aqrrmLDhg3k5OTE\nzWHavXt3TXUgGXE4HFRUVVGQl8eJ0w7HqwsE0sZUMu1i+dkeBvbowqCeXThgWF88Cur2OLHdDsOJ\nQ5pSWlrK6NGj47K+jRo1iiVLljB8eHo9gmfOnBm5f9xxx6WdvW369OmR+5s2bWLOnDmRx507d6Z3\n796Rx+Er3emITUXes2fPtPcTESM2oLIsi5NOOmmnvO6ZZ54Z9/izzz6L3H/22Wcj9zds2MD+++/P\ngAEDGDx4cNxtyJAhDB06NC4wfeCBBxgxYgRFRUVkZWXxzjvRaWJra2vjumAef/zxaQdSsWXYihUr\nInPqJdO1a1ceeeQRysrKeP/997nnnnu47777+Oijj9i2bRv33HNPZL7O2G6do0aNSutcRJJxOByU\nVVSQn5ur4E7anLpitq3emC6buUANptvlKiBlpoh9RxRxyIRhbCvfQSAQpLrOh2VpUPaeJDxGBUyr\nV7LJxMOWLFnClClTIvMsgQnM3n333bSTldTW1sZ1k2xuLF6scePGxT2OPQ8wQdnixYsBIst0LFq0\nKHK/X79+ae8nIsaHH34YuX/44YczaNCgnfK6/fv3p3///qxfvx6Iv0gTW7atWrWKVatWpX3cHTt2\nxLXkb9u2LXK/uro64zIscdvEMiyZ/Px8jj76aI4++uiU28ROdzNgwIC0z0ckGYfDQXlFBQX5+Zxy\nxJF8PHs2m0u3xf2fV5IxyYQCu7ZxPqaL5T5JntsCvAb8EWg0M/PmbWXU+hqobwhg27b5Ueu3vNt5\n4403uO+++7Asi5KSEt544420K1axV6mHDRuWcru5c+cyZcqUuJawK664osWT/FqWFffPIbby1FKJ\nV8kPPPDASNer2GCtKZs3b2bNmjWRx/vvv3/G5yOyNyouLo5rwcpkKoH//Oc/PPbYY2zcuJHs7Gze\neuuttBKA+P3+uIQtLle02tCSzLjNib3g5XA44o5dXl6e9nESK8Nt0WVyw4YNcS2VU6dObfUxRRwO\nBzsqK8nLyeHoKVMoKSvD6XCQ7fGypXQbn3391a4+RemAFNi1jgN4H0g9QzT0BC7HjLE7DJgZ++Sc\n75YxYeQQuncpoLo2s4kvpf1t27aNjz76KPJ41qxZaQV233zzTVzLVqrEAt999x1Tp06NC+puueUW\nfvvb37b4XL1eL/369YtMqPvll1+mve/cuXPjHiemCz/ooIPitl2+fHmTwSrARx99RG1tLWAqggrs\nRFpm5syZca1jhxxySIuPsXjx4siUBGB+v+kkX1m2bBmbN2+OPB46dGjk/muvvUZFRUWTvRDAJH1y\nuVxMnz49Uh6ef/753HHHHaxbt476+npGjozO7tO1a1f69u3L0qVLAVOGpZpqIdHXX38d97iwsDDu\n8d///neeeuopVq9ezbnnnssdd9zR7DFfe+21SHBbUFCgwE7ajMPhoLq2FqfTSWHnzpHpNEYUFbFy\n/To2bm3UHiDSJAV2rfMi0aDuPeAZYDlQDuQAA4GjgYsAD/BvYBQQ+S8ZDAZZXbyZAb17UFVT26ZX\nQKXtHHPMMTgcjsiA/2effZYzzjij2f3uu+++uMcnnHBCo20qKio44YQTIsEPwOOPP86FF16Y8fke\ndthhkSDt3XffZfHixWmNC4lNWDBgwAAmTZoU9/wRRxxBYWEhJSUlgEm/3lxSg6eeeipy/9BDD212\nomMRiRfb6l9YWMi+++7b4mMktvI988wzaQV2Tz75ZMrj7LNPsk4qqeXn50fuDxo0iO7du9O9e/ek\n2x5yyCGRwO6NN95g/fr1ac3ZF1uGjR49ulG5t3TpUmbNmhU57u23397s/92HHnoocv+0006jS5cu\nzZ6HSLosyyIYDFIbypBp2zZOp5N9hgxVYCctpuQpmTsRk+US4Czge8CzwGxgGfAN8CZwJTAUM96u\nM/D7xAOt31yCr8GvLFu7sQEDBnDKKadEHr/zzjtxAUsyTz31FP/4xz8ijy+++OK4q9Jhl112WVxX\nxeeff75VQR0QSfkddvXVVze7zxNPPMErr7wSeXzOOec0qvDk5eVxzjnnRB7fdNNNfPPNNymP+eST\nT8alLL/sssuaPQ8RiRfbDXPo0KHk5OS0+BijR4/m2GOPjTx+8MEH4+a5TObtt9/m7rvvjjw+5phj\nmDhxYotfOyw2E2ZzSaRiy5n6+vq0yrD77rsvrrxJnHMPiMv8u3Llymbn3fvZz34W1+viF7/4RbPn\nIdIalmVRVVNDv5696KtkY9JCiiQyFw7qbgaeb2pDYAMQbt45G4jrt7J5WxnVtXU4nfpz7M5+//v4\nmPz888/ngQceSLrtX/7yl7gJzXv16hU3J1zYhx9+GDc1wb777huZHPzLL79k1qxZTd6+/PJLvvji\nC6qrq+OOu99++0Um6wWYMWMGU6ZMiUvZHevPf/4zP/7xjyOPBwwYwHXXXZd029/97ncUFBREHh92\n2GFx40/CHn74YS644ILI42nTpqXVyiki8datWxe535JEIon++Mc/xj0+7rjjePPNN5Nue9ddd8Vl\n3nQ6ndx5550Zv3ZLHXbYYXEZOV999VWOOuqouG6hsa677jquuuqqyOORI0dy7bXXNtpuv/32i0uS\ncumll/L666832s7n83H22WfHjXG+4YYbWtxKKZKJYDAIFoxWDxdpIXXFzNyhoWXTzTZRs4GvgEnA\neGBO05vL7mbChAk8+uijXHzxxZF1V155Jffccw/HHnssXbt2pbi4mPfee4/i4uLINi6XizfffJOu\nXbs2OmZsFx8wCUn69u0bl6wgHXPmzGnUbfLhhx9m8eLFkaDriy++oKioiEMOOYT999+f7Oxs1q1b\nxwcffBA3J53D4eDVV19NOUVCYWEhDz30UKTSVVFRwaGHHsq0adOYOnUqPp+Pf//733Hdx3Jzc3n8\n8cdb9J5ExEw7Eu6SCI3HvbbEgQceyN///nd++tOfAqYl7JRTTmHChAlMmzaN/Px8NmzYwAcffBDJ\nghn26quvMmbMmIxfOxNPPPEES5cujXQr//DDD+nbty/Tpk1j3333xePxsHr1aj744IPImGKA7Oxs\nXnvttZS9YO6++24mTJgQKWenT5/OpEmTmDZtGh6Ph2XLlvH+++/HJZ6aPn16kxOki7SlSKtdD9Nq\nV7xly64+JekgFNhlLlxLb0nKwbLQsqDJrWS3ddFFF1FQUMDpp58eWbd8+fKUE48PGDCA9957L+n4\nto0bNza6Wt6SueFipUqL/Omnn3LOOefEzTf12WefJW1hAygqKuLf//530i6jsc444wwCgUBcd6mP\nP/44adeurl278tFHHzF48OA03onIniX2txkIBJqdmDtRRUVFXOKU5n6bzbn88svp0qVLXGvYN998\nk7JLdY8ePXjzzTc58MADW/W6QNwFq0Ag0Oz2WVlZzJo1i9NOO4033ngDMC0ZH330UVwyq1gjR47k\nvffeY+DAgSmPO3r0aD7++GOOOuqoyNjmr776iq++Sp6F8Kc//WnK3hki7SW21U6BnaRLff8ytzK0\nLGrBPuE29U1NbiW7tdNOO42SkhKuueaauAm7Y+2zzz7cf//9rFmzJmXSkvLychwOB16vly5dumR0\ny87Oxu12x02CnuiZZ57hs88+48QTT0zZCjdhwgQefvhhVq9enXbF8eyzz2bVqlWcc845Scf8dO/e\nnd/85jesXbuW8ePHp3VMkd1OK/NZeWImH+7evXvcdAHpqKqqwuVyRX7jqcqcljjjjDPYsmULP//5\nz+mZYgzP2LFjuffeeykuLm6ToA6gW7dukft5eXlp7eNyuXj99df58MMPOe6448jOzm60jWVZTJo0\niSeeeILFixc3GdSFHXzwwRQXF/Pzn/88aW8Kr9fLSSedxOzZsxXUyS4RbrXr26Mnw9L4TouAWuxa\n40NgInAVCVMYpHAmMBhYA3wX+4TTYeYd02SUu5bT4YA0s5J2796dO+64g9tvv52VK1eybNkyKioq\nKCwsZPTo0WlVvoYPH05JSUmrM6Hatp0yYAubOnUqU6dOpba2lqVLl7J69WoaGhro1asXo0aNapQS\nPF2DBg3i6aef5tFHH2XBggWsXr0al8vFiBEjGDlyJE6ns4VHtNBEjrI7cTpa+h2Od+yxx1JRUQGY\nilpzv9VE48ePp7S0FIfDgW3bGSVOSaZHjx7cfffd3HXXXSxfvpxly5ZRU1NDz549GTVqVMqArzXe\neecd/H4/tm2nNYderCOOOIIjjjiCmpoaFi9ezNq1a/H7/fTp04dRo0bFBY3p6tKlC3fffTe33347\nS5YsYfny5fj9foqKihgzZkxGn7USW0tbCgaD1Pt8HLLvJBr8ftbEDPMQSUaBXeaeBn6FCdgqgUub\n2PYM4LnQ/X8mPjmgT0865eVQWl7R5icpzbOwCNoWpVW1BGK6PKW1r2UxdOjQuLmd0uVyueLSf+8M\n2dnZTJgwoVUJGJLxer3sv//+rZ6jznI48AdtTfshu5wVuhWXbMUm84a71v7OXS4XnTp1ynj/5jgc\nDkaMGMGIESPa7TXCkrW2tVROTg777bcf++23XxuckeF2uxk7dixjx45t9bF0fVbakmVZ+BoasCyL\nIyYfyH9nzWR1cTGerCxsms8uK3sfBXaZmw/8GbgBuASTJfMtYAEm0HMDg4DvY+auA5gLNBp93a9n\nd7XW7UK2beNxO8l2BNmsOWN2qdKSLYzN9eALtGwckkhbCwLZQE11DVX19eTHdKkUSWXL5s0m7bVl\nKcqTNmFZFvU+H0HbZsrEfZkwchSerCyqamr49/8+S2u8quw9NMaudX4P/CF0vytwAXAn8AjwAKZF\nLxzUfQxMI6GfWafcHIYO6EN1bWZJM6T1bMBnW3x/dC8eve9uttUqqNgV7nr6DZZ8+i5HTxpNZV3L\nsoKKtLWAbTMgJ5eNG4u57b77dvXpSAcwd/kynn/kEcYAPgV10oYsy6KhoQFfQwPZXi9ej4esJsbW\ny95LLXat9ydMN8vLMVMgDMQEzBZQAnyN6bb5XrKdDxw/il7du7J+c4kZ4yW7REllPSdOncC6t7/k\n4LHD+d4PzsDpdBLwK8BoTy6Xix1VNZRs2cTyj97kuiOH4s3No2JHDQ51x5RdLOB0sr/Hy+2//jXz\nPv2U/Q86CF99fYszW8qey+VyUVtfx47S7bz2yit0raqiT0FnKgOB1ubdEYljWRZB2zZdM7Hwq6VO\nklBg1zZWANeE7jsBD+AHmu38vHV7OT5fg4K63cCGshquOPFg9luwjI9fvRWny4XToX/N7a263k+f\nHBe//NG+5HUqYFN5NS597rIb8Nk2eV4v0yyL2W+/zQdvv61/mtKIH3Mld7zLTa9OCupEZNfR/6i2\nFwBq0t34myUrGT1kIEP696GyOu3dpI1ZFgRtm7WllQwfNIBJo4doiMRO4nRY+PxBSqvr2bJDQZ3s\nPiygLhDAcrnYP78TTstSzlZpxML8r6i3g1QGFdSJyK6jwK5t9AT6YBKmNJWv3Q1UA9/Erly6ZgOj\nBw+gsj3PUJplYbo6VPv8VPtalh1T2oa6X8ruJvyNrLdtXemRZqkEE5FdSYFd6wzAJEo5tgX7rA/t\nJyIiIiIi0iYU2GWuAPgC6LurT0RERERERPZuCuwy9zOiQd39wBOYZCkBUk8j4QJqE1dmuVwatyEi\nIiIiIhlTYJe5I0PLOzHz1WXE6XQwZlgRtXX1bXNWIiIiIiKy11GO/cyNCy2fbM1BJu0zgkF9e2mC\nchERERERyZgCu8yVhZblrTnIPkMGEtBktyIiIiIi0goK7DL3bmh5ZJNbNWPxqnWanFxERERERFpF\nEUXm7g4tbwU8mR5kzsKlrNm4hdxsb5uclIiIiIiI7H0U2GVuFfB9zOTka4CzgTFAEdAPM1dd4m0Q\nCdMj+AMBFixfTbY349hQRERERET2csqKmbk/ABdhxtr1Ap5Jc79GE5T7GvxYbXtuIiIiIiKyF1Fg\nl7lemJa5VnO7nDgcDmwbLEV4IiIiIiLSQgrsMncPMAOoasE+WUBF4sq6+gYsywJsUNudiIiIiIi0\nkAK7zC0J3Vpt/tKVHDxhFF0LOlFZU9sWhxQRERERkb2IkqfsBoK2zVcLl5Hj9YBt7+rTERERERGR\nDkYtdpnrEro1pLl9uK9lDbAD8Mc++dV3y9hvzHC6F3SipKquTU9URERERET2bArsMncLcCkmsAti\nAjcX8a2gQaIBnAU4gbrQbR3wNnAHUBG0beYsXMYpRxwMCuxERERERKQF1BWz9SzMBOVZNP48HaH1\nWYA79DgH6ApMAH4PrACGAny7ZCUbt5aS49Vk5SIiIiIikj4Fdpm7DPg50VbPJ4BjMUFa/9DyKOB+\nTMsdwJ3AROB8TItfBVAIvArQ4PdTvHUbniz3znkHIiIiIiKyR1Bgl7lewF2h+4cAPwbeB1YCG0LL\nD4GfAeND250PLAOeAq4DRgNbgXHAuQCbS7bjcjp3zjsQEREREZE9ggK7zF2HGTP3c+B/zWy7ELgY\n6A78OWZ9ceg4AIe39QmKiIiIiMjeQYFd5k4KLV9Jc/t/hZanJqz/NLQ8qNVnJCIiIiIieyUFdpnr\nEloGm9wq9X6JsltxLiIiIiIishdTYJe52aHl0WluH+5qOS9hfc/Qcn2rz0hERERERPZKCuwyNyO0\nfAAY0cy2RcC9ofvvJzx3YWg5s21OS0RERERE9jYK7DJ3PybzZT7wLXArMAnTAtcN6A1MxiRLWRBa\ntw64L7T/QExw+OPQ45cAHE4HYO+UNyAiIiIiInsGV/ObSAp1mG6YnwL9gN+EbkEggPlsrZjtN4W2\nrwo9vhA4JnT/TmAOwKC+vfA1+BVxi4iIiIhI2hQ/tM5qYAhwIyZwA/OZuokGdaXA/wGDMHPYhXUD\nSjAB3q8Aivr2ZHC/3lTV1rX/mYuIiIiIyB5DLXat5wNuCt3GAaOALEyr3QrgKxpnzrSAq4ErY1ce\nOG4Uniw3gao6/WVERERERCRtCh/a1vzQrTk20BC7YlC/3owcNICyikosy0qxm4iIiIiISGPqirnz\ndQc8iSsnjx2B0+kgEGjptHgiIiIiIrK3U4td6w0BvgfkhR7HprS0MMGzC/ACR2ImKB8be4BB/Xox\nclD/aGudkmKKiIiIiEgLKLBrnauBu1q4z3qgPnZF7+7d8HqyKKuoUjdMERERERFpMXXFzNxkokFd\nA7AU2Bh6vBUz1m4d8WPpXgHOTTyQr6EBvz+goE5ERERERDKiwC5z54SWG4AxwEjg0NC65cB4zCTk\nvYCHQ+s3A58kHqhH184Egup/KSIiIiIimVFXzMztF1r+huj8dCuBLcAUIB+oBLYDlwHloW1nAc/E\nHqi8shqnUzG2iIiIiIhkRtFE5vYJLT9OWP+/0HJywvo/JCwjVhdvoqq6FqfT2XZnJyIiIiIiew0F\ndq2XOD/B2tCyb8L6euBDYCjQJ/aJuvoG/IEAGmInIiIiIiKZUGCXueWhZV7C+gWh5QFJ9gkHfaPb\n5YxERERERGSvpMAuc7NDy1MS1m8OLZMFdsNDS2VKERERERGRNqPALnOvh5a3A1cB7tDjL4BqYBJw\nfsz2ZwNTMdMfzN9J5ygiIiIiInsBBXaZ+w/wWuj+PcDdofsVwOOh+08AnwOfAU+H1s0ASmIP1LUg\nD68ni6CmPBARERERkQwosGud04AnQ/eLY9ZfC8wL3T8Y01JnYaY++FniQQb07kmO10MgEGjHUxUR\nERERkT2V5rFrnSBwAXAXpvtlWD1mjN2VwOmAB/gAuIOE1rpsTxajBg+gsroGS2kxRUREREQkAwrs\n2kayMXN+TPfMu5vacXhRP4r69mTD5hIcDjWgioiIiIhIyymS2MUWLl/DsjXF5OXm7OpTERERERGR\nDkotdukZBXQCfK08jhvTZfO78IpAMMjCFav5fp/J1NTWtfLwIiIiIiKyN1Jgl543iM5B11rrgQGx\nKxr8ATS6TkREREREMqWumC0XCN1ERERERER2C2qxa7ly4J/Av4FlQA7QD8gF7NAtFU9o/zhqrRMR\nERERkdZQYJee64GrgEOAbsCvgMOBR4CXMAFexkYM6k99Q0Nrz1FERERERPZS6oqZnleAQ4GewOXA\nHGA/4GGgDNgC3AKMbemBx40YzD5DBlJRXdt2ZysiIiIiInsVBXYtsxV4CDP5eB9MK963QA/gt5j5\n7NYDfyDNZCtup9PMX2c31YNTREREREQkNQV2mdsE3AdMwGS5/BWwGDPe7o/AUmAF8BugKNVBvl60\nnMWr1pGveexERERERCRDCuzaxnrgTmA0MBj4HSaoGwLcCqwGvgJ+DvRO3HnxqrV4stw77WRFRERE\nRGTPosCu7a3GBHPDgDzgROBzzJi8u0P34wSCtjJjioiIiIhIxhTYta9qoBSTYCVMmUhFRERERKRN\nKchoe/2AHwBnAAcmPLcA02onIiIiIiLSZhTYtY0BRIO5AxKe+xZ4FngZWJNs5z6F3fAHgu15fiIi\nIiIisgdTYJe5IuCHwI+ASQnPzQeewwRzq5o6yKC+vdh39FAqqmva4xxFRERERGQvoMCuZQZjgrkz\ngIkJzy0AngdeAlame8DehV3Jz8lm+45KnA4NeRQRERERkZZTYJeen2DmqRuZsP5/mGDuZaAkkwPP\nXbyCfUcPIy8nm7p6X+vOUkRERERE9koK7NJzLTA8dH8N8DQmmFsQs40FNDcZnQUEgYbwirp6H98u\nXcUxB+2nwE5ERERERDKiwC49sZlN/MB44HDMZOPZCc83pQfwDTA5dmVNXT2WJrITEREREZEMKbBL\nT+zgt6GhW6Z6t/JcRERERERE4iiwS899mMQpFa08Th5Q3PrTERERERERiVJgl5772/PgA3v3oMEf\naM+XEBERERGRPZjy6+9ig/v1ZtzwwVRqHjsREREREcmQArtdbPK4kTidDgLBdPOviIiIiIiIxFNg\ntwsN7tebEYP6U7ajEktpMUVEREREJEMK7HahyeNG4nSotU5ERERERFpHgd0uotY6ERERERFpKwrs\ndpHJ40aptU5ERERERNqEArtdwLTW9VNrnYiIiIiItAkFdrvA5HGjcFhqrRMRERERkbahwG4nGzV4\nACOK+lFeodY6ERERERFpGwrsdqJBfXtx2rGHUufzqbVORERERETajAK7naSoTy/OP/kYgkGbmto6\ntdaJiIiIiEibUWC3ExT16ckFpxyDPxCgoqoah0Mfu4iIiIiItB1FGO2sqE9PLjj1WAV1IiIiIiLS\nbly7+gT2ZEV9e3HBycfg9yuoExERERGR9qNIo51Egjq11ImIiIiISDtTtNEOivr2VFAnIiIiIiI7\njbpitjHTUnc0fr+fiuoaBXUiIiIiItLuFHW0objulwrqRERERERkJ1GLXRsZ1NfMU9cQCFBZpaBO\nRERERER2HkUfbcAEdUfjDwSo1Jg6ERERERHZyRSBtFK4pc4fCCpRioiIiIiI7BKKQlohGtQp+6WI\niIiIiOw6ikQypKBORERERER2F0qekoHImDp/gIpqBXUiIiIiIrJrKbBroWj2Sz+VmtJARERERER2\nA4pKWiDa/dKvKQ1ERERERGS3ocgkTYP69uL8U8Jj6hTUiYiIiIjI7mOP64qZl5PdpserqqllSP8+\nnHPCkWZMnRKliIiIiIjIbmaPC+x+Mv17bXq8QDBIp7wc6n0+aurqFdSJiIiIiMhuZ48L7DxZbfuW\nLMuivt6HPxBQUCciIiIiIrulPS6w8zX42+W4lmW1y3FFRERERERaS01QIiIiIiIiHZwCOxERERER\nkQ5OgZ2IiIiIiEgHp8BORERERESkg1NgJyIiIiIi0sEpsBMREREREengFNiJiIiIiIh0cArsRERE\nREREOjgFdiIiIiIiIh2cAjsREREREZEOToGdiIiIiIhIB6fATkREREREpINTYCciIiIiItLBKbAT\nERERERHp4BTYiYiIiIiIdHAK7ERERERERDo4BXYiIiIiIiIdnAI7ERERERGRDk6BnYiIiIiISAen\nwE5ERERERKSDc+3qE5DUHNihm0g6rNDS3qVnIR2HAxsLG4I2YEMwuKtPSUT2RIEAOBz69yTSzva4\nwC64BxQagSDYDrIdFjgsNatK82wbsrMcuBwOu9bnt/zBIJZlNb+j7NWcFjiwsG0bG7AV2EkzbMBh\nWbidTjsYDFoNwSAqaaQ5djAI5v+Sx7Is/X9qJcuhz1CS2+MCuxxXx4/ssglgYX0xs7rgXQs227rG\nJc3I82ZVrC+umLqtomb/EX27P5vldJTX+hq8KvilKVkO295uO1xdivpC11y/w+3WF0ZSsgG3w1EX\nCAa7bqupOTM3K2tWN4/n84ZAoGBXn5vs5mzbwuHoFbSDs31VleZqpGTMEQzgr6/T5yiN7HGBXZaj\n43/JHZZNQ9AxY2ODZwagq6HSpE45HpZuqeDzpduur65v2H9VtfOaA4f33RKwPNT5GnRVT1JyWOAP\nWmTn5xLM8eq7IimFgjosC9aX7+i7PRg8s8bles3bqdNt+Z4s6v2BXX2KsjuzLLBtGiorCNb7sBzq\ni9QaQQuCfv+uPg3ZDe1xgd2e0BXTtk3XqGzHntnFxeFw4Ha78LhdeN1ugqErTpZlEQgGqfM10NDg\nx9cQH5TYmCDX68nCEfqn4Gvw09CQXvBiA94sN06nEwC/P0C9z9ehK7Odsj2sLy1n7vL15Dod7m6d\nvWwp3W7NWx7goOF98XosbMuB22V+6v5AgPr6jv2epe04Lai1nAQa/ATr63E495zKlm3bOJ1OspzO\nSJeHcBnT0EyFKNzdMMsV/RdpAb5AgEAgsNf9fsJBnW1ZrCsvZ0ddnZXjcuGvr3ev3bqVfp060SUn\nhyAQDAbxdcAKZ+zf3AKwLBr8fvxp/L1t28bpiJazlmXhDwRo8Pv3uu9KcywsHA6nCfQkY5bl0HdL\nktrjAjvZPQVtm9xsL3nZXmrq6qmoqmZtWQWbSssilSd/IEC2x0Ofwi4U5OXSq1sXqmrrqKypBaAg\nN4cst5tNJaX4/H5s26ZbQSd6duvC9h2V+Px+HEkKuqBtk+v1kp/jZXNpGdW1dQB0ys2hZ7cu7Kiq\nprbel3Tf3ZUN5GS5WF9awRdLN+CwHHjdTmcgECDP43at2VpGMBjkyAnDqKyqYUv5dgDycrLp2a0L\nFdU11NTVd6j3LJIO27ZxOBx0ysvD5/NRVlERuRAUCAbxZmXRpaCAuvp66urr4y8e2TaWZZGfk4Nt\n22zfsSOyLhgM0qVTJ7Jycqisria4l4xjtW2bbK8Xt9vNurIyqv1+srOy3IFAAKdlOf3BIFtra3E4\nHAR8PlwuF107d6ahoYGauroO8xnlZmfjsCy2l5cTtG2Ctk3n/HzycnOprK5OGtCHvxud8vIIBAJs\n37EDMMFtXk4OXQoKqKqpSfvio4hIaymwk3YVDqoK8nPYvK2cD2d9w+rizZSWV+L3B7AThg9amKAl\nN9vL6CEDmDx2JAN7FwKwduNWPpv3HfOXrQYbbGy6dMrjkH3HMHHkELoW5LN9RyUNgQBW5LU9FOTn\nsqW0nP9+9S1zFi7H19AAQLbXw0HjRjJ57Ej6FHalrKIqEuAl+ydst6Ave2JlsbltWvoaXreT8up6\nZi0rxrLA43ZgORzP5eXkLLZtuyQvLw87y8ubn3/L4sWLqarzmf2ysjhg7HAOHDeKPoVdKa+optZn\n3nNL3l9LZfpeW/MZpapItWb/9vyMpHViAzpsm9XFxSxYtozSHTtwxgR2LqeT4QMHMnrwYLoWFFBT\nV0dtvbnIkZebi9vppHjrVhYsX07x1q2R8iAQDNK1oIBxw4ZR1LcvlmXFBXht+V3N5HuW6TFSnbdt\n23g9HnKzsyktK+OL+fPZUVdHj27dsD2ezXl5eec0NDR85cnKoiEQ4P0vv6SuthaHZTG4f3/GDBlC\n9y5dqK2vp7auDiwr6YWklr7XTN5nsvcY/rvlZGfjcbvZVFLCghUrWLdpU+T/UEFeHmOHDWNwv364\nXK64AM+yLPJycnA6nazftIn5y5ezpbQUh2URtG2yPR72GTKE4UVF5BcUUFldHWnBa23ZJCKSirWn\nFSS/v/DUXX0KrWYBARvqgo4O2xXTtm1ysk1QtbmkjC/nL2bOwmU0tHAchsOymLrvPthBm/99813K\nccLhAG/CiMFkuV04HBZZbhebt5Uxc/4SZi9YlrJ7ULYni4PGj+KAMcPp1rkTwWCQel8DgdA//nA3\nm6wsN7X1Pmpq6ggEg5hOJTZ5udnkZHvAJtKSCBaBQACHw0GW20VVTR01NXW43S7y87Jxu1zU1fvi\nPi+3y4XL5aSqppbaOp85jg1ZWS4K8nMj5xIIBsn3ZrFqSzmffLcWJ0ECgQAej4e8vDwA8vPzWbFi\nBQsXLkz6nj1Zbg4cO5LJ40bQKS8HbHC5nFRUVVNb58PpcNApPwdPlptg0MbhsKiuqaOqxlTQIhMr\nBG08Hjed8nJCFRaoq2+goqoGbMjJ8ZCb7TVdmvymQmS6PJkgs7beR2V1beTKd+jDICfbS16u1/wd\nAvF/B4/HTXVNHdU19XEDUE3XOwcFeblkZbmor2+I60Zl2zYul5Mst4uKqhpq63wJAbg5r075pmW4\n0d/H7cLldLKjspr6PWTsotOCWtvJxq1VHbYrZiSgy80laNus2biRBcuXU1pe3uR+LqeTEUVFjBw0\niG6dO+MPBCjeupWFoYCuKd0KChg7bBgD+/Qhy+Uy3Q8DgbhWvPBv2mlZVNbW4o/plhfuVp7j9ZLl\nduNraIi7xOV2ubCDQRyOUHcr26a6ro56n49sj4dsrzcSBPj9flwuFw0Jr+9xu/EHAlTX1BCM+X2F\nf0fZXi9ulwtfQwPB0PlEyjqXCyyL7RUVLFm1iqVr1xIImLLbsizy8/MZP348+fn5lJSUsHDhQmpr\na+M+I8uyGNq/P6NDAZ4DUz4mdm3McrsJ+P3U1NdHgibbtvFkZZHj9UIoY2swGMRyOCLdaGO7Ttb7\nfNTU10fLpdB3ItvjMZ+v328+g9j3GBoGsHnbNr5bsYK1mzal/Ht3zs9nzNChDOrbl2yPB38ggD8Q\nYPO2bcxfvpzN27al3Dfb62WfIUMYNmAAeTk52MEgDaH94/4moW7DNXV11HXwIQJNCfp8JkPmHvr+\ndpYst5uqmhpmzJ4dqo/snfa0GKYtKLDbDXXkwM4OdbnslJfL5m3b+fLbJcxZuJSGQJMBXRcgDygB\n6lrz+oVdCsjLyaZ/z+5kuV38b94i6ny+5nfEdFOcPGYEOd4sVqzfRGVNHW6XkwZ/gNxsDwN6FdK3\nRzeG9O8TuoIfwJuVxeJV61hdvAWAnt06Y9s2sxcuC7UOuCjq04MJIwYxoHdPyiurWL6umA2bt1Fc\nUkogYIImX0MDXQvyKerTk/49u9O3ZyHBYBC328X2HRV8u3QVPp8fh9NBp9xs6ut9LFm3CTw59OnV\ni0CoUpednU1OTg5r1qxh0aJFzb7nTrk5dCnIJz8nmwG9ChlR1Jf+vXtQW1fPt0tXsXlbGVluF/5A\ngKH9+zB8UD9qauvZUVFFdraHLp3y2bxtO/OXrcYfCBAIBOnToyvjRwwBYE3xZjZs2caajVspr6wm\ny+0iEAzidrnoU9iVIf16MWrwAOp8DVRUVtO5Uy5ZbjerizezYt1GNmwtpSr0d/D7A2R7PfTr0Y2h\nA3ozuH8f/P4AZRWVWJZF986dqG/wM2/xCqpqatlWXkFJWQVZbtMxwdfgpyA/lwE9uzNsYF8G9O5B\nRVU1ldW1eD1uCvJNd6olq9ezasNmNm0rI2jbOCyL+oYGuhd0YmDvQoYX9aNn9y7U1NRRUV0TF+h2\nNB09sLNtmyy3G6/Hw9qNG5mfRkCXyO1yMWboUCpralixbl2L9u3ZrRtFffoQDAbZXFpKXX19ZAxv\ng99P5/x8enTtyoDevcnPyaEi1GqTn5OD2+Vi3ebNbN62jdLycvyhMtLldNKre3eyvV5q6+oiAc2g\nPn3o3rkzxSUlbNiyBcuycDocFOTlUVVTw8Zt26ivr490O+3aqRM9u3dnUJ8+AFRUV2NZFgW5ufj8\nfopLSthWVsa28nL8fj8OhwN/IEB+Tg69CwupqKpi6Zo1Kccjer1e+vbtS0lJCRUVFSk/I8uyGFFU\nRPfOndm0bRvllZUmcA0Fm907d6Z7ly70LSzEk5WFr6EBb1YWW8vKWLd5M4SCnmyPh8qaGraUlmKH\nukpmuVx069yZ3oWF9O/Zk7r6emrr6+mUm0sgGGRjSQml5eWUlJVR39CAM/Qec71e+vbowZbt21m5\nfn3af+/unTszfOBAdlRXs7GkhPKKirQrlnk5OQwfOBBPVhYl27dTFvocwAxDyPF6KezShT6FhfTq\n1i3SkrynBXgK7NqGAjtjT4th2oICu91QRwzswgFdQV4um7aV8eX8RcxeuAx/6ha6I4CfAJOAwZhu\nweXAEuAt4H6gMsW+04BHgXnA+UBtiu2SGQs8AZSG9k19mTaF/r0KOfrAieR4PXzy9QIWLF8T93z4\ninOsHK+H/ccMZ9X6zazfUtLsa0wYMZhD9h3Dus1b+WDmN5Fxgcnk5+fTp08f+vTpQ5cuXdi6dSv/\n+9//YjfpCzwJFAAXAN+lOlZutpepE0ezfvM2Fq1qXMkdP3wQh+w7hkH9elFaVsHMBUv44tvFca1b\nAOOGDyLL7eKr75Y3+15HFPXj8P3HUdSnF9+tXMusBUtYuWFzpIUgGYfDwaA+Pdl/zDAmjBhCMGgz\nb+lKPvlqAVu3lzf7mi6XkwPGDOegcaPo17M7ZRVVzPluGfOWrGTztrIm9832ZDFq8AAOHDeSQX17\nUV1ba1oosTpcXaUjB3a2bZOVlUVedjYfz5nDitQV9HuAKUAFEAAuAtam+TK/B04D6oFZwJWZnGtO\nqNVmRFER2R4PG0JdPTds2ZL2MfJycujdvTtrN22KdCeH5OVNrMIuXRg7bBiD+vYlEAiwYv16FqxY\nwY7KVMVrk/bFlJ8bgPOA2KaqfwJHA5cDb7f0wHk5OUwcMYLCrl1Zvm4di1atarIMSFTUpw/jhw+n\nW+fOrNm4kW+XLWtxkB9yAvAA8BFwIS2bbug04O/Aa8Clmby4w+GIdBXuVlBAdV1do7GgHZkCu7ah\nwM7Y02KYtqDAbjfUkQK7YNAmN8dL57xcNm3bzsz5S5i1YGnk6nMSw4CXgfHNHLoBOBX4V5LnzsME\nKtXAAGB7C075aOD90P2RwNIW7Ltb6969O6NHj2bt2rWsXRtXbx0FhJvvjgY+aM3ruF1Oxg4bxNpN\nWyktT32lviWy3C4KuxSwqWR7JEtqukYW9cMfCLBifYtjdHK8HibtM5wV64rZWNKSr5Exfvggpu67\nD0W9e1JTX8+Oyhqg49RZOmpgZ9s2Hreb3Jwc/jtnTnOtLguBfWIe/xL4WxovUwCsArqGHi/FlBkZ\n61pQQEF+Pqs3bGjNYTLSv1cvGvz+JrsNpuEEokHbIGBN6L4L81n1B64Hbm7Ni2Qqy+0mPzc304Au\n7GrgLkzQOhCoacG+vwFuxVygHNWak3BYFsOLiiJjQWv3kBY8BXZtQ4GdsafFMG1ByVMkI8Ggjdvt\npHf3zmwtK+e1Dz83LXRNX2GdBvw35vFLwLPAasAH9ABOAq4CsoB3gDOBFxKOE+4b5AvdWiJ2+4aU\nW3VA27Zt49NPP032VOwfpVVdXQEa/AHmLl7R2sPE8TX4Kd5amtG+S9ZkXkmuqavn068XZLz/t8tW\n8+2y1dEAr09PqmvrQy14qru0h0hQl52dTlAHjVv0TyK9wO57RIO6ZMdpse07dkQyJ+5s6zdvbovD\nhMtPm/jy048J8voD6fdtbGO+hobWBnUQbYVcTOZ/81bXNoO2zZLVq1m2dm1cC96e2kVTRNqGAjtp\nsUAwSLeCfBwOB5/NXcj7X86ltr7Z+GogJlADqAKOBGYnbLMU+Ax4BNPtqQum28+XpN91SmSnSxbg\n1dTWs6OqukOPwdvdhJNq5Ia6X7ZkfBSwAlMOTQOGhh435Qeh5Q5M65007dBdfQJt5OnQbbcQDAYj\nAd6IgQMZpQBPRJrQMfrdyG4h3OTdv2ch5ZXVPPDCO7z58cx0gjqA24Dc0P3jaBzUxVoOnBO678F0\nbRHZ7X27bDUPvPAOz7zzEVu3l9O7sCsFuWZONHUYaZ3wmLrc5sfUpfIW0YtL329m20Lgh5guda+2\n9IVE2lowGGTx6tW88d//8tm8edTW19Otc2c8Ho+6o4lIhFrsJC3BoI0ny023zvnMXriU1//7JT5f\n2j0ZJwCnh+7fB3yexj7vAh9iWvZSdYexMa1/DuB3wDFAPqYy9jLweronmKA/ZhzfIUA/TNeizzDj\n+opT7PMD4HDgdkzr4r3AYcBc4DEav+dJwFmhZQGwDJiB6Zqa7P0OxiRumIe5mlwE/ArT+lAbOr/H\naSIxSkj42BeGzrlf6HzfwyQ/SHztK4ERmErxf5o47gGYYHwpJvFArE6Y93oMpqVkK6YV9p+YcTmp\ndMf8HY7AvN+NwCeh/TY2sR+YMVU/wHz3BmK6iq0CZgLPYJLnNGUccG7ofXUJva+3geeIdgVOKdyC\nN274IA6ZuA9FfXtSUVVDVU1tJGuhpC/SUuf1ptv9MpmNmPF2pwInY5KqpBIO/J4jve54hwEnYr53\nvTGJn5ZjxrO+SHx36ESjgR8DBwPZmIte94fO9WTM7+YxzG8/bBrmN/U05rc/GrgMmBx6fh6mvPqy\nmfPeB/O7PRDoBqzElLvPYJJZtcRFwBhML4tvkjyfC5wNHIsZa10GzMGUefOSbA/m93chJhnJ+5j3\ndxWwH6Yc+QDz2TRVHowm+lsuxJTnXwPPY7pbJpqAKXcWYcrUZFHTacCPMOPo1mHK+3+TXvf+EZiL\nBvtiyiYb87nPxnzuTc61EduCF9tFs7KmBp8mQhfZ6yl5ym5od0ueEgwGycvJJjfby4wv5/Lf2d+2\n9BC/BW4J3T8IU7lOR3dMNrrE9G1nYSoDi4D/Ax4kfixM2IeYZCGxX/LDgI9D94fQOLC4DvhLE+d0\nNXB3itc6ApN1cjImM1zYu0Qrik5MheK0FMevAE4hfiwimMroa5hkMn8jdQKUPwF/SFg3nGiSmPMw\nFaVpSfbdBEwl/jN5FZiOCTxHpHhNgDcxY5dexVRawqZjxki6U+x3FSbYT3QmplKdyqWYLrvJPIWp\nyKUSwFSWP0rx/IOYSnIymzAV0xYNzBs3fBDfmzKJHl0L2LK9PDKn367WEZKnxHa/zDCom4O5gHID\n8BDRMVRDMRXqZP4FHI8JCr4H3IkJViYmbFeIqdDv18Trr8JcJEoWfNyIKcOSORdTpvwYcxHnzpjn\n7gV+FlofvuiQzL3Az1M89xCpMzc2AGdgypxYx2AuQNmYC2DhC10OzFjpAcC1mAtcsb4PvAJ4U7ze\nE5j3mej3mDLtr5jgL3G8ddjZJC8vbsZc9EvlNkzCk1g/x5TxmzEJYmLHJRdiyvqxSY71S0z5/Rgm\nYBydZJuHgUuaOB8wn9W7zWwT4XA4GFFUxL6jRuFyOqkMTWuxu1LylLah5CnGnhbDtIXd77+47FaC\nwSD5uTlkez08/vqMTII6iF5F3oi5UpqubaSe8gDM1dIXMYlWfoO5AnoEpuIBprXvry14vVuJBnUz\nMJWxYZhgMBxo/Q1TQUwUvlJ7PSao+xozJYMPU6EJ+w/RoO6B0DmPwLRorsa0bn0EHJXi+NMwQV0t\nJsDtgglQnw09fyOmspfKU6Fj/B0TZB+AqeCBaWn4Z8L2D4eWwzGfZzL9MUEdMccCE9S9ignqvsO0\naAwLve57oW3uxVQgY11ItJI2BxNIDce0KrwRc17JUs8/RjSouw3YH5OBdRKmgujDBNePkzzYfJ5o\nUPfP0P7DMYH1csxnNJfklbaU5i9bzb3Pvcl/Zs6jc34enfPzWpz9c28UmyglgzF1iXpgWmrDrUPH\np9iub+i5ckwFvUcTx3wfE9SVAVdgWnvGYaZXeDK0zWAat2ID/JpoUPcmppwcirn4UoppjTsz9Hxi\nf/dwefAzTFC3EhPcjMOUL8tCz19F/IWWsJeJBnVPYcqB4Zjf6ALMb+PVmNdvTpBoq2RiFsnpmC6w\nXkwQeiqmzNsP83sFc0HsrSTHDb/PizFB3UpMGd8ZE2SHs0U9S+O/5zlEg7rbQ9vvgylH3gytvxYz\n7U6scIt8LebCYpiDaFBXiwkAh2E+85cxmTSbuih4L9Gg7m6iZdN+mP8p4Zbhf2DmdU1LMBhk8apV\nvPHhh5SWl9Otc+dmp8AQkT2XAjtJKRzUeT1Z/PON91m+rrnebyn1Cy2/o20zUVqY+ZSKMJX4eZgA\n7Diiwd0PSO97fgjRK7d/Cx3jf5gEC59iKhPhoOUmTKAQK/xfdBimG9UkzD/xfKID8a/FdNcEE8yE\nu1Yuw1QMhgJfhZ5/hOSBRy6msjEOE4SUY1oEzsFUHMLn17+J93o4phI6ExM4XR7aB0wChHEx276P\nCWggmkwi0cmh5QqiLYkuzN8ETGVoDKZytyL0ut8j2srwF6J/o95Eu8i9gKlwhs9hFqZSGG79vSvh\nfU4lWkm7EvP3/AqYjwm0/4CpQIL5zhyb8D7OIhpkXoQJML8KvfYbmIrv7NB7uzvZB9GU2nof733+\nNQ++9C9Kd1RQ2MXk41AFLLnEoC6DMXWJskPLcABxSortTggtnwkt81Ns90tMIAfm+/x34FtMYPQF\n5rsWzr55ItAnZt9BRL/H94XOZTYmcHkaE/gUk7qFK2wg5rs5FHMxZAHmQtIBQHiSvLMT9rmAaLB3\nJWY+zzmY7/k7mN9/uPx8gKYD2+bkYz4XMOXoyND5LsNcILmY6G/2ROCnKY7TFRNkj8SU8TswLaiH\nEe0iflvCPheHljdhyt5vML08ZmE+7/B7vCDN93I90Za6wzHl7QrMZ346JjjrmWLf/TBBOKFzuZpo\n2TQXUwaGg+iepL7okFJ1XR3vfPop85cto0unTjgdDo3rFdkLKbCTpOKDuv+0JqjzYCrrEK1otKXr\nSD5e6qnQsj/QK43jhCsXy4FrUmzzC6Jdj1J1YaogvmuVj+gV4AtDy2cxLY2Jgpgr7GAqfokVsrDr\nSJ7R74+YQM9L6ivtfyfaFTVWbBa4AQnPhc/1JJIHm+GAL7ZlcjqmJRFMy0Qyd4WWhUTnB7sEc7W6\nmtRdyP6A+Tu4ie++5cMEkW8QbcFM9A6mUgjmqn+sK2O2eTzF/teHlkdjKs8ttn5zCfc8+yaLVq6j\nT49uuFxOBXcJ4rJffvVVWwR1QKRneziwOwLzO0v0o9AynDQlVZ+xDZhxcPdggoVkwsdwYlrjw84K\nrVuPqegnKsUEjum4Lsm6HUS7LSb+ni8KLd8neUsiRMuhLqQf+CRzNtFgJ9Xk7v8g+jdpqpvi1SQf\n3/r70HIf4pPidAstU022+QfMZ5yqvE8ULlNT/b3/ghnfnUwQE0i+S7QlN9G/MN0/oXHZlLZZCxbw\nzZIldCkowKmWO5G9jgI7aaRxS12qfCEt1h6d6j9LsT580m7SS1U+JbR8jdRzENUTDV4OSrHNTKJj\neGKNJzpGLVlQF/Yl0Va7KSm2SZUUpozohO7TUmzzcYr1W4gGPF0Snns+tOyLacmMNTLmtV6OWR9e\nN4vUiRFmYoLuzkQnUD8s5jVTJRFoiDmno2PWz8Z0YT2V1IkfsokGp56Y9f2J/k3/kWJfMC2SC0P3\nU3VNTcsz//qIN//7Jd07F5DldqsCFhLXUtd2QV2suUSTeyS2jBRhWmM2k/q3EvYSpgXnF01skxtz\nPyvm/rTQ8j+knlvy35jfdFOWkjwBCEQveOXErOuL6QIITZdDy4iO80pVDqUjPAXCJzQ9LjX8ex6P\naX1MtJJoC1uiWUT/nofFrA8HWbdjLi7FBtbh/f5G0xmaw0YQnXC8qfFvr6RYPw9Tdn6f1OVaDtGy\nKSvFNmmZ8913zFu82AR3arkT2asoK6bEsW2bbK8Ht8vFE2+2qqUurB5TSepL43+srVVN6quxsZns\nmgsoBxGtTDQ3BjA8yHAfTDeoxPn1Us0CHB6TFaT5xBvzMV05kyVkWIYZi5fKmtCyKMXz21Os92Eq\nmAU0/rwWYa7uH4NpnYtN1BBurfsCU2EOC3e/bapWXkd8pdZN9O8wAtNFtA/xgbaFGasZTmIR270t\n0QjM32gIpovswUTHe5Jw3OEx908I7ZeYkCeIycIa/h4PbOK10/Lp1wvxB4KcesTBlO6owOfbu7Pa\nhac0yPF626r7ZSpvYbpRnkx8q9WJoWWqFt9UsjC/8X6Y79JwTIA4PMm2Lkz5AU2XN5WYsuCwJrZJ\ndhEprDq0dMasG0k0aGhuwPS3mMB3IuYicCYZGsLduuc2uVX0YgmYv0tij4Tm9l8b2m9IzLrbiZZP\nt2HGW8/GBIhv0bLx3hNCywDRRFTJhMc2Nle3GoYpo8Nl04GY8qnNzPnOJEieOGoU23fsIBgM7tVl\ni8jeQoGdxHE6HXQtyOeZf/2XZWvbrKVuPSZIaVHCiTTUET+4PVZsZaa5C5axLTclzWwb+6F4kjyf\nKo1+eKxMRRPbhG1I2CdWc6nXw/v2wgQgiYFvqv0dRFvwk31eL2ICu5Mw42bCSW3C3daeT9g+3P21\nJVcGCokGaoeEbs3phRkDFHsV/HpMF6tkmVK3Ys4/O2F9v5j7F5KepsYxpu2LbxZhAUcfOBG/P7DX\nttzZtk2W202u18sn7dNSF+ttTKKho4m/QBOeluWNNI/TFRNAnEvjbspBTEt4uCti+A+bQ/S33dzv\nuaqZ5xMTlcQK1+Jjv1CxZUpTQSFEyzoP5r2lKmubEg4imytXN2MC0VySl3vVSdbFCp9r7IWeWZj/\nOw9jLpRZmAs7kzF/+03AHUS7hDcl3OpZQdOtqOHvUaq61a8wXW8LkzxXgimX0k6c0hwFdyJ7HwV2\nEhEM2vTs1pnP5i7kmyWpsoBn5GvMYPUhmCvYy5rcOuoYzFQJb2DGgCX+Q3XQNt07Y69EO1NuZcRW\nOpLVwFPtH84Y52jha7RUONFDDckrYpl2v34TkxSmCyZRxEuYZCXhZAKJXZDC7zcxgGqKTfQzfRnT\nFa2pxA3ZmO9EbPD6Wei8wHTH/BSTtGdJ6Hg1mHmnEs8r/B2wMdlRN5A6aQaYCn2603Y06/NvFjFq\ncH/GDx/E6o1bwLb2qmzgkaAuO5tPvv66vYM6MN2d52Fao47HTHExBPPdWYdJnNScgZiWpPAFhA2Y\n79tSzHfuVcw4zMTxWA1Ex4qlmgaENJ9vrixJFDufXmvLunSEf1fNvVY4eMz0tcK/1cQAcC6m6+lg\nTGvssZjuobmYsd93YlrLTqdp4feRTvmdyvtEu45XYMqqhZiy6X3MRYA1tGFgB6HgzrKYOGIEZRUV\nBBTciezRFNgJYCpWnfNzWbuphHc+SWfIQYt8QDTr4vGkH9j9ENOd6XBMRau58SaZqsRUtFyY7nsf\nNrFteJzFNswV33SFx691wlQymupaFH6NZF0uOzfzOkWh5Uoyu8KeSimmonoOZgzbS0SzYb5M4y6o\nsS2HTemPCbZKMZ/nWkwFexGNp15ozjVEg7pfEs1IGCuf5BWz8Nx9FiaITZUMo918NPtbPFlu+vXo\nTlllNX6/f6+ogEWCupwcPmnf7peJ3sIEdidhArvw+NGm5k+MdTsmqAtgAoZk5UaywKwW03JcSPS3\nnoxF8vFmrbEj5v5QGs/jGSt8bmtpPN1CS1+vqTkww+cSbt1L7N4O0UQoqYS7RS9P8fwqTNKTezB/\nk6MwF3DC00McReq5QSHa86AA03U/1f+iwaFlYvbnnxIN6m4g+bQILtqpTjZn4UKwbSaOGkXZjh0K\n7kT2YEqeIgBYlkWW28XHc+a3x2SXM4kOUP8l8QkFUhlMNBvbDJofY9Eam4iOt2hunEM48cAymu8m\nFesrolfpJzexnZvoeI7vkjw/kOj4nGTCY89aNIF2ml4KLQ/HXM0Pd5V8Ocm24QrWGFKXMzmYz2Ub\n0WkGwokgmhpXBCa4/APxCUzCKdxnkDyoAzOeJZxMJ/a8lhLtFtfUa7sx2f2uIfkkxRlbtWEzD7zw\nDu998TX5Odk4HHt+8RzXUtf+3S8ThceKHo35Pocr3m+ksW9PouPxbib1xaDYwC32D/pJaDmV1PYl\nGii0lXlEuy2nSgAV+/oQTWyUiXCrdlNlHkTL1TqSZ5bcn+Rd38H8nseH7ofHDXbFjK/7JY3H4TZg\nWu+PJtrC19wwgblEu70mTnUTK3wegYT14bLpM1LPdTeU6IWwNv/xN0qospd2+RbZ0+35NQdJS0F+\nLsvWFvPdymQXS9vEzaFlf5IHAomeI3q1+852OaN44Qmzzyb11eXxRAfjN5UZLZliopOcp5oqAcz4\nrvDYrX+n2ObHKdYfRbSy9kZLTi5N/8Jcue6J6SK7P+aK/DtJtg2//iDM1AfJfI9oV8twQoLwftNo\nPEl7WAFm8vc/Ep0bCqB7aNnUeJxzY+7HXh3fRjSj6OWk7kL6fcy8Y3fQxskOwj75agHfLltF14L8\nPbry1aj75bp1O/sUvsZU2J2YVpQjMC3d6bTWFhDtqtjU9+38mPux37dwq+AkouNUE/0qjfNoqTqi\nLVMXkfoi23SiF4laWtbFCpdhw0g9fYuL6DQHH5B87F8vUk/hcj7m7xEgWhb1xXQPv5PUY3VriCZv\nam5+1VKin8PFKbbphJnGIpmWlk2ZdvdsUiS40zx3InssBXaCZVk4HQ4WLF/Tni/zJtGJar+HafE6\nmcZj5I7DVK7CV3jvIjoBbXu6h+gA/3do3IVwANG5ljYSnQy8Jf4aWo4neVrsE4hOgv420UAw0TXA\neQnrhhNNYPJvmu5WlKkg0WyBv8KUHy+QPAHEbKKf1z+JtkKGFRH9PrxIdEqEx4lmyHuF5JlBXyTa\nNev2mPXhCvmJJL8C/1viU9MnjmW8I+bc/kXjlOOjiU6FsAQziXy7+Oq75fh8DTj30Fa7Ri11Oz+o\nCwt/R3+H6aabbjbMZUQvRpxD8v+lzxPfIhf7ffucaHD3PNG55cCMsXqeaCs2tO1UMeGJvPthyrrE\n7n8HE52c/QuiLfWZeA8zfgzM/G1HJNnmLUzgB9EyMpmHaNzKeDym7Aa4m2gX8AVEp0e4m+RZgm/B\nlCP1xGf6TSX8uU0Cnkjy/CukHhMcbrk8juRl2s+Jn4+wJWOTW2TOd98xd/FiunfpgkfTrIjscTTG\nTnBYFrX1PjaWNJessdWuwFyJvBTzj/wNzFXThZirraOJn3PuAZJPHhuu5MRmckwUe8UztlLkSHG/\nDDPW4mNMl5j1mPnilmK6Pp4c2j6AuZpdSbzwsZq60vohZj6lcBruEsy8eaWYAf3h+aIWkfrqdm3o\ntZ/EtCy9j2lh/GHotdcR30oA8e8/1fnFJgVoKpp4LvQewlf6m6r0/RjT1XIQJnB7BzPf1AjMZ+jE\njCNMbMH8ESYwLAjt/0HocSHmfYbn2bsGU0EOux/zubkxQd5TmAr4IEwluSemtbgAk5gnsUvrLMx3\n9AFMd9MyzLjCdZixOOGudwFM60G71YjWbNzC2k1bGdC7B1U1zSVO7Fhs28btcpHj9e6sRCnh/3PJ\nvtdvY1p+w8+9mWSb8O8i8f/l/ZjW2zGY7/GTmO/MvpjvsBt4DPO7HhW6xX5fL8CMJ52MaYG+FzOW\nLVwGvoX5rYyg8Xct/JtuqrxJ9f/9W8xv85+YlvFSTDlUHDqXcEt5MY1bE8Ofk0XjzzN8Tomvexbm\nd1yEKQM/xgSMPTHlQPj3/FNSJ60pxbzXLzAXrmZjgrxjQs//l8YtnNdiulv2woyxextzQcaLCQjD\n4xd/QzQgjH0fTqJlPsAczIWhuzFl7KGY8iEXUx50xvx9pyT5DB4g2tL3GSYZ2CLMBcMfYVoYX8d8\nZ06g6e72rfbVokWUV1YyeexYcrKzqa6p2Su6fovsDRTYyc52Geaf4R8xV4ZzMJnjYn2FaV1JNW4l\nPAjQT+q5lfwx92O3CaRYD2bcyzBMJWsaJtCL9Ram8pFsHojwcf1Jnot1B2Zuqgcx42cuiXkugOk6\ndH0TxynDjCu7C9PyeWDMcy+EjpcYdKZ6/4nbNDSzDZjA7N+h114CfNTEttsxAdF9mErsCaFb2Euh\n892RsN8iTIX375iK31HEd8vcgAkGE7uIzcRU5F7AXIn/acxzNqZSdg/mu3UMpmJmEV9p/jvmQsND\nmIp4bPcoMN/Jn5A8wUObqqypxe1qlx5Zu5TT4cDldPLZ3Lk7q6UunPgj2W9qLiZICGevTDaWN7xf\nYne9+zHfn3sxFfTfxzxXgrmY8BEmWByF+d0+FrNNA+b3+yvM93kopqVmK6aL8UuYiwrQuFU8GHOM\nVMLPJSsjnyDa6jyW6HjmsIdD55U4jjh8rECS46Y6p1JMoHI3JriZRnSCdjDl4WXAl0nOM2wOplUr\nXPZ8L+a5W4hv7Yo97jhMALs/JknOSTHPb8RcyHkjxftI9v/lHkwQ/xjmglFsMHk6prfJ1zRuYf0W\nU968hAk0L0l4/reY1sqfYcrIQzABaKrJ61ttxfr1bCwp4bBJk+jSqRP1vkzz44jI7sTa05rhf3/h\nqbv6FFrNAgI21AUdbdr/JhWnw0FWlpt/vD6DjSWp5q9uFz0wlYoizMD4DZiK1YYm9gFT+emB6UKz\nheQtJ1mYK8LhuaTClTNPaN/E9Yn6Y64I98S03n1B/FxpibphrrSX0jhQSWUfTIUjB9MyOJPUYzBO\nwFxxbsBUDLZjWgYOwqT1/4jUWTpdoffhwLznZP/BrdA2Xsz7bGp+rDcxFaQ/YRKYpKM7JpAfGDr+\nZ6Q3x10hpvI7EPO5fkPziWFcmIr6WMx341tMZSv8t87G/L28mM8s1Wc+HNPtqismmJ9N8qC+Xfzw\n6KlMGDGYsoqW5OhpGacFtbaTjVurCNbX43C2/1V7t8tFnc/Hu599hq+huaFNbaIQ072xDPNbSdQF\n8zfeTvJsh51D21SRfD62fEwrzVDMRZVZxCcA6US06/A6Ul846Y5pJdoS87gYU5ZNx7TohBWEjrmD\n1PNi5mHeezVNl13DMb+XAkxg8iWpyzAPppzwh84z9r30xJRlW0n9m+qCKbMGYz7rWTSejDzWbzGB\n2zdEx/wdhunOvh5T7qVT3g4M7d8ndG7zMOVIsv8duZjPrZbo3yKRM/Q+JmK+E+GxgR7MNAo1JP/M\nnZgyZULo8XzMhczwD8GL+bvmYMqm9vvxh4wfMYL9Ro1iR1W7vxQAQZ8POxhkr5rPpR1kud1U1dQw\nY/bs9kh412HsaTFMW1CLnexKW2l6aoFUamm+xcSH+cefqD7F+kTr09wurJTmJx5P9B3JM182J/y7\nnUt62UL9NB+Q2DSesiCZ/kSveqebFh5MpeetZrdqrIT0xr/E8mMC8S9SPF9L8xcPwHTjTHdqDtl9\nldD0BNllND2VSjnJA8KwSqLJl5KpIH6uRTBdig/GjOkLZ+RNTBrSl+g4z8SLIDtoPqCpIr3AoCXf\n83qirYiJUgVBscrILBlLbBTwCdGsoulaS/qt7NU0PyF6ANNtNLHraD1mLrqm9ptF6gQ9dezEi0ci\nsudRp2oRaYlwkoc3iSaPEJGWGQxcjRl3mmpC6nBX4uXsgnkVRUSk41FgJyJN8WASvRwH/Bm4MbS+\n3TJCiuwFwklaemG6WA4mmgyqJ2aMW3gc1l07/exERKRDUldMkY7BHbNMNVFve6gHbiJ+ouUnaN3c\nViJ7u6WYJDyPYxIDrUyx3U1Ep0DZG4XLvZ1Z5omIdFhqsRPpGNZikho8T/rJWdpKeE68CkxFM9UE\n6SKSvn8AIzGtc7EJRNZhpuqYRLSFfG+1AJMlN5PxuSIie509LiumiIiIiIjI3kYtdiIiIiIiIh2c\nAjsREREREZEOToGdiIiIiIhIB6fATkREREREpINTYCciIiIiItLBKbATERERERHp4P4f2OsS04mx\nYCQAAAAASUVORK5CYII=\n"
    }
   },
   "id": "8dc17c87-7845-484a-a4e2-bd2070ddcc3e"
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
