{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PREPRINT – Discriminating Seagrasses From Green Macroalgae in European\n",
    "\n",
    "Intertidal areas using high resolution multispectral drone imagery – PREPRINT.\n",
    "\n",
    "Simon Oiry [![](data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==)](https://orcid.org/0000-0001-7161-5246) (Institut des Substances et Organismes de la Mer, ISOMer, Nantes Université, UR 2160, F-44000 Nantes, France)  \n",
    "Bede Ffinian Rowe Davies (Institut des Substances et Organismes de la Mer, ISOMer, Nantes Université, UR 2160, F-44000 Nantes, France)  \n",
    "Ana I. Sousa (CESAM – Centre for Environmental and Marine Studies, Department of Biology, University of Aveiro, Aveiro, Portugal)  \n",
    "Philippe Rosa (Institut des Substances et Organismes de la Mer, ISOMer, Nantes Université, UR 2160, F-44000 Nantes, France)  \n",
    "Maria Laura Zoffoli (Consiglio Nazionale delle Ricerche, Istituto di Scienze Marine (CNR-ISMAR), 00133 Rome, Italy)  \n",
    "Guillaume Brunier (BRGM French Geological Survey, Cayenne 97300, French Guiana)  \n",
    "Pierre Gernez (Institut des Substances et Organismes de la Mer, ISOMer, Nantes Université, UR 2160, F-44000 Nantes, France)  \n",
    "Laurent Barillé (Institut des Substances et Organismes de la Mer, ISOMer, Nantes Université, UR 2160, F-44000 Nantes, France)  \n",
    "March 8, 2024\n",
    "\n",
    "Coastal areas host seagrass meadows, which offer crucial ecosystem services including erosion control and carbon sequestration. However, these areas are increasingly impacted by human activities, leading to seagrass decline and habitat fragmentation. In situ surveys, traditionally performed to monitor these ecosystems face limitations on temporal and spatial coverage, particularly in intertidal zones, prompting the use of satellite data within monitoring programs. Yet, satellite remote sensing struggles with spatial and spectral resolution, making it difficult to discriminate seagrass from other macrophytes in highly heterogenous meadows. To address these challenges and reduce the issue of mlixed pixel, drone images at a very high spatial resolution offer a promising solution. This study focuses on using drone acquisitions with a ten spectral bands sensor mirroring those of Sentinel-2, for mapping intertidal macrophytes and effectively discriminating between seagrass and green macroalgae. Nine drone flights were conducted at two different altitudes (12m and 120m) across diverse European habitats in France and Portugal. Low altitude flights were used to train a Deep Learning classifier based on Neural Networks to discrimintate among five taxonomic classes of intertidal vegetation: Magnoliopsida (Seagrass), Chlorophyceae (Green macroalgae), Phaeophyceae (Brown algae), Rhodophyceae (Red macroalgae) and Bacillariophyceae (Diatoms). Drone mapping demonstrated an overall accuracy of 94% across all the sites and images, covering a total area of 467 000 m². The model exhibited an accuracy of 96.4% in identifying seagrass. This work showed that although they share the same pigmentary composition, seagrass and green algae can be discriminated using ten bands multispectral sensor. The algorithm should now be adapted for Sentinel-2 to see if this discrimination is still operational at a coarser spatial resolution."
   ],
   "id": "082d9369-a8da-4a9b-a178-653b69823717"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: le package 'ggplot2' a été compilé avec la version R 4.3.2"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: le package 'stringr' a été compilé avec la version R 4.3.2"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──\n",
      "✔ dplyr     1.1.2     ✔ readr     2.1.4\n",
      "✔ forcats   1.0.0     ✔ stringr   1.5.1\n",
      "✔ ggplot2   3.4.4     ✔ tibble    3.2.1\n",
      "✔ lubridate 1.9.2     ✔ tidyr     1.3.0\n",
      "✔ purrr     1.0.1     \n",
      "── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "✖ purrr::compose() masks flextable::compose()\n",
      "✖ dplyr::filter()  masks stats::filter()\n",
      "✖ dplyr::lag()     masks stats::lag()\n",
      "ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attachement du package : 'tidyterra'\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:stats':\n",
      "\n",
      "    filter"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "── Attaching packages ────────────────────────────────────── tidymodels 1.1.0 ──\n",
      "✔ broom        1.0.5     ✔ rsample      1.1.1\n",
      "✔ dials        1.2.0     ✔ tune         1.1.1\n",
      "✔ infer        1.0.4     ✔ workflows    1.1.3\n",
      "✔ modeldata    1.1.0     ✔ workflowsets 1.0.1\n",
      "✔ parsnip      1.1.0     ✔ yardstick    1.2.0\n",
      "✔ recipes      1.0.6     \n",
      "── Conflicts ───────────────────────────────────────── tidymodels_conflicts() ──\n",
      "✖ purrr::compose()    masks flextable::compose()\n",
      "✖ scales::discard()   masks purrr::discard()\n",
      "✖ tidyterra::filter() masks dplyr::filter(), stats::filter()\n",
      "✖ recipes::fixed()    masks stringr::fixed()\n",
      "✖ dplyr::lag()        masks stats::lag()\n",
      "✖ yardstick::spec()   masks readr::spec()\n",
      "✖ recipes::step()     masks stats::step()\n",
      "• Use suppressPackageStartupMessages() to eliminate package startup messages"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Le chargement a nécessité le package : cowplot\n",
      "\n",
      "Attachement du package : 'cowplot'\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:lubridate':\n",
      "\n",
      "    stamp\n",
      "\n",
      "Le chargement a nécessité le package : patchwork\n",
      "\n",
      "Attachement du package : 'patchwork'\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:cowplot':\n",
      "\n",
      "    align_plots\n",
      "\n",
      "Le chargement a nécessité le package : magrittr\n",
      "\n",
      "Attachement du package : 'magrittr'\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:purrr':\n",
      "\n",
      "    set_names\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:tidyr':\n",
      "\n",
      "    extract"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Linking to GEOS 3.11.2, GDAL 3.6.2, PROJ 9.2.0; sf_use_s2() is TRUE"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,\n",
      "which was just loaded, will retire in October 2023.\n",
      "Please refer to R-spatial evolution reports for details, especially\n",
      "https://r-spatial.org/r/2023/05/15/evolution4.html.\n",
      "It may be desirable to make the sf package available;\n",
      "package maintainers should consider adding sf to Suggests:.\n",
      "The sp package is now running under evolution status 2\n",
      "     (status 2 uses the sf package in place of rgdal)\n",
      "Support for Spatial objects (`sp`) will be deprecated in {rnaturalearth} and will be removed in a future release of the package. Please use `sf` objects with {rnaturalearth}. For example: `ne_download(returnclass = 'sf')`"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attachement du package : 'rnaturalearthdata'\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:rnaturalearth':\n",
      "\n",
      "    countries110"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "terra 1.7.39\n",
      "\n",
      "Attachement du package : 'terra'\n",
      "\n",
      "Les objets suivants sont masqués depuis 'package:magrittr':\n",
      "\n",
      "    extract, inset\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:patchwork':\n",
      "\n",
      "    area\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:scales':\n",
      "\n",
      "    rescale\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:tidyr':\n",
      "\n",
      "    extract\n",
      "\n",
      "Les objets suivants sont masqués depuis 'package:flextable':\n",
      "\n",
      "    align, colorize, rotate, width"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Linking to ImageMagick 6.9.12.3\n",
      "Enabled features: cairo, freetype, fftw, ghostscript, heic, lcms, pango, raw, rsvg, webp\n",
      "Disabled features: fontconfig, x11"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "\n",
      "Attachement du package : 'grid'\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:terra':\n",
      "\n",
      "    depth"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Le chargement a nécessité le package : abind"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Le chargement a nécessité le package : lattice\n",
      "\n",
      "Attachement du package : 'caret'\n",
      "\n",
      "Les objets suivants sont masqués depuis 'package:yardstick':\n",
      "\n",
      "    precision, recall, sensitivity, specificity\n",
      "\n",
      "L'objet suivant est masqué depuis 'package:purrr':\n",
      "\n",
      "    lift"
     ]
    }
   ],
   "source": [
    "library(flextable)\n",
    "library(tidyverse)"
   ],
   "id": "b1907a3a-e8b5-4a9f-8d1d-de9719009978"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_comma<-scales::label_comma(accuracy = NULL, big.mark = \",\",decimal.mark = \".\")"
   ],
   "id": "48eb7b96-2288-4f0d-abb6-79d5d8cb715d"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Introduction\n",
    "\n",
    "Coastal areas are vital hotspots for marine biodiversity, with intertidal seagrass meadows playing a crucial role at the interface between land and ocean ([Unsworth et al., 2022](#ref-unsworth2022)). Seagrass meadows provide a myriad of ecosystem services to humanity, including carbon sequestration, oxygen production, protection against sea-level rise and coastline erosion, and limitation of eutrophication. They serve as vital habitats for a diverse array of marine and terrestrial species, providing living, breeding, and feeding grounds ([Gardner and Finlayson, 2018](#ref-gardner2018) ; [Zoffoli et al., 2022](#ref-Zoffoli2022) ; [Jankowska et al., 2019](#ref-jankowska2019)). Due to the concentration of human activities in coastal zones, seagrass meadows are directly exposed to and impacted by anthropogenic pressures. Global regression and fragmentation are currently observed due to climate change, diseases, urbanization, land reclamation, dredging, competition with alien species, and reduction in water quality ([Nguyen et al., 2021](#ref-nguyen2021) ; [Soissons et al., 2018](#ref-soissons2018) ; [Orth et al., 2006](#ref-orth2006) ; [Lin et al., 2018](#ref-lin2018) ; [Duffy et al., 2019](#ref-duffy2019)). While improvements in water quality have been recently reported in Europe, allowing an overall recovery of seagrass ecosystems at local scale, many coastal waters worldwide are still subjected to strong eutrophication processes ([Los Santos et al., 2019](#ref-deSantos2019) ; [Zoffoli et al., 2021](#ref-Zoffoli2021)). Coastal eutrophication has been ass ociated to anomalous accumulation of green macroalgae, the so-called green tides (Devlin and Brodie ([2023](#ref-devlin2023nutrients))). Green tides produce shade and suffocation over seagrass individuals, thus threatening the health of seagrass ecosystems ([Wang et al., 2022](#ref-wang2022)).\n",
    "\n",
    "The importance of seagrass meadows and the variety of ecosystem services they provide have led to the enhancement of both global and regional programs monitoring Essential Oceanic Variable (EOVs) such as seagrass composition ([Miloslavich et al., 2018](#ref-Miloslavich2018)), as well as Essential Biodiversity Variable (EBVs) such as seagrass taxonomic diversity, species distribution, population abundance, and phenology ([Pereira et al., 2013](#ref-Pereira2013)). Seagrass monitoring is all the more important as it makes possible to identify which pressure should be addressed in priority to facilitate effective mitigation actions. Traditionally, indicators of seagrass status have been quantified using *in situ* measurements. The acquisition of field data in intertidal zones is however notoriously challenging. Intertidal seagrass meadows are only partially exposed during low tide and can be situated in difficult-to-reach mudflats, potentially leading to inaccurate and limited estimations with conventional sampling techniques ([Nijland et al., 2019](#ref-nijland2019)). Satellite observations have been proven effective in complementing in situ sampling, allowing for the near real-time and consistent retrieval of seagrass EOVs and EBVs over extensive meadows. ([Zoffoli et al., 2021](#ref-Zoffoli2021) ; [Xu et al., 2021](#ref-xu2021) ; [Traganos and Reinartz, 2018](#ref-Traganos2018) ; [Coffer et al., 2023](#ref-coffer2023))\n",
    "\n",
    "While satellite remote sensing provides temporally consistent observations over large spatial scales, it is limited by several constraints. Satellite missions with a high temporal resolution (e.g. daily MODIS observation) are limited by a coarse spatial resolution (\\>100m). Mission with a high spatial resolution such as Sentinel-2 (10m) or Landsat8/9 (30m) can be limited by low spectral resolution. The limited number of spectral bands is indeed a challenge to accurately discriminate seagrass from others co-existing macrophytes. In particular Chlorophyceae (green algae) and marine Magnoliopsida (seagrass) share the same pigment composition ([Ralph et al., 2002](#ref-ralph2002) ; [Douay et al., 2022](#ref-Douay2022)). As a result, their respective spectral signatures can be considered similar by a non-expert observer ([Davies et al., 2023a](#ref-Davies2023) ; [Bannari et al., 2022](#ref-bannari2022)). Recently, using advanced machine-learning algorithms trained with a large hyperspectral library of more then 300 field reflectance spectra, Davies et al. ([2023a](#ref-Davies2023)) demonstrated that it was possible to discriminate Magnoliopsida from Chlorophyceae using radiometric data acquired at Sentinel-2 ’s spectral resolution. The application of this approach to satellite remote sensing remains however to be validated. Moreover patches of green algae can develop at small spatial scales that are not observable using non-commercial satellite imagery ([Tuya et al., 2013](#ref-tuya2013)), especially during the initial stage of the event.\n",
    "\n",
    "Drones can potentially fill the gaps left by satellite remote sensing and in situ measurements, due to their ability to provide spatially-explicit observations at very high spatial resolutions (pixel size from cm to mm) and still capturing data at multi-spectral resolution ([Fairley et al., 2022](#ref-fairley2022drone) ; [Oh et al., 2017](#ref-oh2017use)). The versatility of drones allows for their application across a diverse thematic range , from coastal zone management ([Adade et al., 2021](#ref-adade2021) ; [Casella et al., 2020](#ref-casella2020) ; [Angnuureng et al., 2022](#ref-angnuureng2022)) to mapping species distribution ([Joyce et al., 2023](#ref-joyce2023) ; [Tallam et al., 2023](#ref-tallam2023) ; [Roca et al., 2022](#ref-Roca2022) ; [Román et al., 2021](#ref-Roman2021) ; [Brunier et al., 2022](#ref-Brunier2022Topographic)). However, when applied to coastal habitat mapping, most study cases were limited to a single flight, restricting the generalizability of their application over wider geographical scales ([Román et al., 2021](#ref-Roman2021) ; [Collin et al., 2019](#ref-collin2019improving) ; [Rossiter et al., 2020](#ref-rossiter2020uav) ; [Brunier et al., 2022](#ref-Brunier2022Topographic)). The present study aimed at analyzing the potential of multispectral drone remote sensing to map intertidal macrophytes, with a particular focus on discriminating Magnoliopsida and Chlorophyceae. Ten drone flights were performed over soft-bottom intertidal areas along two European countries (France and Portugal), covering a wide range of habitats, from monospecific seagrass meadows to meadows mixed with green, or red macroalgae. A deep learning algorithm was trained and validated for macrophyte discrimination, emphasizing applicability across diverse sites without a loss of prediction accuracy.\n",
    "\n",
    "# 2. Material & Methods\n",
    "\n",
    "## 2.1 Study sites\n",
    "\n",
    "Seven study sites distributed between France and Portugal were selected for their relatively extensive intertidal seagrass beds. Two sites are located in Gulf of Morbihan (<a href=\"#fig-map\" class=\"quarto-xref\">Figure 1</a> A), France (47.5791°N, 2.8018°W). This gulf covers an area of 115 km² and is only connected to the sea through a 900m wide channel. A total of 53 small islands are scattered across the gulf leading to 250 km of shorelines. Patchy seagrass meadows can be found on a lot of these islands. Within the Gulf of Morbihan we have chosen two sites : one is located in one of these island (Arz island); the other one is located in the southern part of the gulf (Duer). Two others sites are located in Bourgneuf Bay, France (46.9849°N, 2.1488°W). This bay is 340 km² semi-enclosed macrotidal bay, protected from waves by Noirmoutier Island. Bougneuf bay hosts a large intertidal seagrass meadow of about 6 km². Within this meadow, the sites observed by drones (L’Epine and Barbatre, <a href=\"#fig-map\" class=\"quarto-xref\">Figure 1</a> B) contain a monospecific bed of *Zostera noltei* with very little mixing with other macrophytes. Three sites have been surveyed in the Ria de Aveiro Lagoon in Portugal (40.6887°N, 8.6810°W). The extent of this lagoon is about 80 km² with many narrow channels, large salt marshes and many mudflats that uncovers at low tide. It is connected to the open sea through a single channel, with a tidal lag between the North and the South of the lagoon. The southernmost site (Gafanha) is a mudflat located in an affluent of the lagoon whereas the two other sites are situated in the middle of the lagoon and are only accessible by boat (<a href=\"#fig-map\" class=\"quarto-xref\">Figure 1</a> C). These Portuguese sites are characterized by a more diverse intertidal vegetation, where patches of seagrass could intermingle with red, brown and green macroalgae."
   ],
   "id": "0c389b1d-229a-44f0-b494-e04032209576"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "world_map <- sovereignty10 %>% \n",
    "  st_as_sf() %>% \n",
    "  dplyr::filter(SOVEREIGNT%in%c(\"Spain\",\"France\",\"Portugal\",\n",
    "                                \"Italy\",\"Andorra\",\n",
    "                                \"United Kingdom\",\n",
    "                                \"Switzerland\",\"Belgium\",\n",
    "                                \"Germany\",\"Luxembourg\") ) \n",
    "\n",
    "bbox_europe <- st_bbox(c(xmin = -20, ymin = 34,\n",
    "                         xmax = 20, ymax = 55) ,\n",
    "                       crs = st_crs(world_map) ) \n",
    "\n",
    "world_map<-st_make_valid(world_map) \n",
    "\n",
    "european_union_map_cropped <- st_crop(world_map, bbox_europe)  %>% \n",
    "  st_transform(\"+proj=laea +lat_0=52 +lon_0=10 +x_0=4321000 +y_0=3210000 +ellps=GRS80 +units=m +no_defs \")  \n",
    "\n",
    "\n",
    "UnitedKingdom<-sovereignty10 %>% \n",
    "  st_as_sf() %>% \n",
    "  dplyr::filter(SOVEREIGNT%in%c(\"United Kingdom\") ) %>% \n",
    "  st_cast(\"POLYGON\") \n",
    "\n",
    "bbox_UK <- st_bbox(c(xmin = -20, ymin = 45,\n",
    "                         xmax = 20, ymax = 55) ,\n",
    "                       crs = st_crs(UnitedKingdom) ) \n",
    "\n",
    "UnitedKingdom<-st_make_valid(UnitedKingdom) \n",
    "\n",
    "UK_map_cropped <- st_crop(UnitedKingdom, bbox_UK)  %>% \n",
    "  st_transform(\"+proj=laea +lat_0=52 +lon_0=10 +x_0=4321000 +y_0=3210000 +ellps=GRS80 +units=m +no_defs \") \n",
    "\n",
    "Europe_sf<-european_union_map_cropped %>% \n",
    "  dplyr::bind_rows(UK_map_cropped) \n",
    "\n",
    "Projects<-data.frame(\n",
    "  Name=c(\n",
    "  \"Portugal - Aveiro Lagoon Flights\\nLower-Resolution: 4\\nHigher-Resolution: 1\",\n",
    "  \"France - Morbihan Gulf Flights\\nLower-Resolution: 3\\nHigher-Resolution: 3\",\n",
    "  \"France - Bourgneuf Bay Flights\\nLower-Resolution: 4\\nHigher-Resolution: 0\"\n",
    "  ) ,\n",
    "  Long=c(-8.745730,\n",
    "         -2.840592, \n",
    "         -2.111868) ,\n",
    "  Lat=c(40.617610, \n",
    "        47.578736,\n",
    "        47.024719) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(\"+proj=laea +lat_0=52 +lon_0=10 +x_0=4321000 +y_0=3210000 +ellps=GRS80 +units=m +no_defs \")  \n",
    "\n",
    "Projects_df<-Projects %>%\n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)  %>% \n",
    "  separate(Name, into=c(\"Country\",\"Bay\") ,sep=\" - \")  %>% \n",
    "  dplyr::mutate(Country=as.factor(Country) ,\n",
    "                Bay=as.factor(Bay) ) \n",
    "\n",
    "Miniworld_map <- sovereignty10 %>% \n",
    "  st_as_sf()\n",
    "\n",
    "sf_use_s2(FALSE)\n",
    "\n",
    "bbox_EU <- st_bbox(c(xmin = -30, ymin = 20,\n",
    "                         xmax = 50, ymax = 70) ,\n",
    "                       crs = st_crs(Miniworld_map) ) \n",
    "\n",
    "  MiniEU_map<-st_crop(Miniworld_map, bbox_EU)  %>% \n",
    "  st_transform(\"+proj=laea +lat_0=52 +lon_0=10 +x_0=4321000 +y_0=3210000 +ellps=GRS80 +units=m +no_defs \")  \n",
    "\n",
    "  df_P1<-Projects_df %>% \n",
    "    mutate(ID = c(1:nrow(.))) %>% \n",
    "    mutate(Site = case_when(ID == 1 ~ \"Aveiro Lagoon\",\n",
    "                            ID == 2 ~ \"Morbihan Gulf\",\n",
    "                            ID == 3 ~ \"Bourgneuf Bay\",\n",
    "                            TRUE ~ \"NA\"),\n",
    "           ID = case_when(ID == 1 ~ \"C\",\n",
    "                          ID == 2 ~ \"A\",\n",
    "                          TRUE ~ \"B\")) %>% \n",
    "    dplyr::select(c(lon,lat,ID,Site))\n",
    "    \n",
    "    scaleFUN <- function(x) paste0(sprintf(\"%.2f\", x),\"°N\")\n",
    "  \n",
    "\n",
    "p1  <-\n",
    "  ggplot(MiniEU_map) +\n",
    "  geom_sf(linewidth=0.5,alpha=0.93,\n",
    "          fill=\"#CFCFCF\",colour=\"grey30\")+\n",
    "    ggforce::geom_mark_ellipse(data=df_P1,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = ID,\n",
    "                     description=Site) ,\n",
    "                 linewidth=0.3,\n",
    "                 fill=\"goldenrod\",\n",
    "                 show.legend=F,\n",
    "                 label.hjust = 0.5,\n",
    "                 con.size = 2,\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 label.fontsize = c(25,20),\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") , \n",
    "  label.buffer = unit(5, \"mm\") ,\n",
    "  label.fill = \"grey90\")+\n",
    "  coord_sf(xlim=c(2600000,4100000) ,\n",
    "          ylim=c(1600000,3100000))+\n",
    "  theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  scale_y_continuous(labels=scaleFUN)+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\"),\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20) )\n",
    "\n",
    "\n",
    "\n",
    "ggsave(\"Figures/Figure1/Map_Drone_Sites.png\",p1,width= 10, height=10)  "
   ],
   "id": "5ce25a23-cf76-406b-aa6c-d36fac0a81ad"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filelist<-list.files(\"Data/shp/\", pattern = \".shp\", recursive = T,full.names = T)\n",
    "mask_aveiro <- filelist[str_detect(filelist,\"mask_land_intertidal_Aveiro\")] %>% \n",
    "  read_sf()\n",
    "\n",
    "Flight_aveiro<-data.frame(\n",
    "  Name=c(\n",
    "  \"Gafanha\",\n",
    "  \"Mataducos\",\n",
    "  \"Marinha Lanzarote\"\n",
    "  ) ,\n",
    "  Description = c(\"12m and 120m Flights\",\n",
    "                  \"120m Flight\",\n",
    "                  \"120m Flight\"),\n",
    "  Long=c(-8.742485,\n",
    "         -8.646980, \n",
    "         -8.690677) ,\n",
    "  Lat=c(40.597599, \n",
    "        40.666634,\n",
    "        40.665120) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(crs(mask_aveiro))  %>% \n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)\n",
    "\n",
    "aveiro_plot <-\n",
    "ggplot() +\n",
    "  scale_fill_manual(\"\", labels = c(\"Intertidal area\", \"Land area\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\"))+\n",
    "  geom_sf(data = mask_aveiro, mapping = aes(fill = Type),linewidth=0.05,alpha=0.93,colour=\"grey30\")+\n",
    "      coord_sf(xlim=c(519749,535233) ,\n",
    "          ylim=c(4493310,4508794) )+\n",
    "  ggforce::geom_mark_ellipse(data=Flight_aveiro,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = Name,\n",
    "                     group = Name),\n",
    "                     # description=Description),\n",
    "                 size=0.3,\n",
    "                 fill = \"goldenrod\",\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 show.legend=F,\n",
    "                 label.fontsize = 25,\n",
    "                 label.hjust = 0.5,\n",
    "                 con.size = 2,\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") ,\n",
    "  label.fill = \"grey90\",\n",
    "  label.buffer = unit(5, \"mm\")) +\n",
    "    # theme_void()+\n",
    "  theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  geom_label(aes(x = 521000, y = 4507794, label = \"C\"), size = 20)+\n",
    "  scale_x_continuous(breaks = seq(-8.70, -8.60, by = 0.10))+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\") ,\n",
    "        legend.position = \"none\",\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20))\n",
    "\n",
    "\n",
    "ggsave(\"Figures/Figure1/Map_Aveiro.png\",aveiro_plot,width= 10, height=10)  "
   ],
   "id": "1f51a812-72d8-40ab-b8da-463b8fbfd34b"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filelist<-list.files(\"Data/shp/\", pattern = \".shp\", recursive = T,full.names = T)\n",
    "mask_BB <- filelist[str_detect(filelist,\"mask_land_intertidal_BB\")] %>% \n",
    "  read_sf()\n",
    "\n",
    "Flight_BB<-data.frame(\n",
    "  Name=c(\n",
    "  \"Barbâtre\",\n",
    "  \"L'Epine\"\n",
    "  ) ,\n",
    "  Description = c(\"120m Flight\",\n",
    "                  \"120m Flight\"),\n",
    "  Long=c(-2.166313,\n",
    "         -2.228239) ,\n",
    "  Lat=c(46.956420, \n",
    "        46.989515) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(crs(mask_BB))  %>% \n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)\n",
    "\n",
    "BB_plot <- ggplot() +\n",
    "  scale_fill_manual(\"\", labels = c(\"Intertidal area\", \"Land area\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\"))+\n",
    "  geom_sf(data = mask_BB, mapping = aes(fill = Type),linewidth=0.05,alpha=0.93,colour=\"grey30\")+\n",
    "      coord_sf(xlim=c(550609,579539) ,\n",
    "          ylim=c(5189447,5218377) )+\n",
    "  ggforce::geom_mark_ellipse(data=Flight_BB,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = Name,\n",
    "                     group = Name),\n",
    "                     # description=Description),\n",
    "                 size=0.3,\n",
    "                 fill = \"goldenrod\",\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 show.legend=F,\n",
    "                 con.size = 2,\n",
    "                 label.fontsize = 25,\n",
    "                 label.hjust = 0.5,\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") ,\n",
    "  label.fill = \"grey90\",\n",
    "  label.buffer = unit(5, \"mm\")) +\n",
    "  theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  geom_label(aes(x = 553460, y = 5216442, label = \"B\"), size = 20)+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\") ,\n",
    "        legend.position = c(0.95,0.2),\n",
    "        legend.key.size =  unit(1, \"cm\"),\n",
    "        legend.text = element_text(size = 20),\n",
    "        legend.title=element_blank(),\n",
    "        legend.background = element_rect(linewidth = 0.5, \n",
    "                                         fill = \"white\",\n",
    "                                         linetype=\"solid\", \n",
    "                                  colour =\"black\"),\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20))\n",
    "\n",
    "\n",
    "ggsave(\"Figures/Figure1/Map_BB.png\",width= 10, height=10)  "
   ],
   "id": "a44f1b65-1e78-4da4-978f-f1c0beb5ed3e"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filelist<-list.files(\"Data/shp/\", pattern = \".shp\", recursive = T,full.names = T)\n",
    "mask_Morb <- filelist[str_detect(filelist,\"mask_land_intertidal_Morbihan\")] %>% \n",
    "  read_sf()\n",
    "\n",
    "Flight_Morbihan<-data.frame(\n",
    "  Name=c(\n",
    "  \"Arz\",\n",
    "  \"Duer\"\n",
    "  ) ,\n",
    "  Description = c(\"12m Flight\",\n",
    "                  \"12m and 120m Flight\"),\n",
    "  Long=c(-2.795961,\n",
    "         -2.746186) ,\n",
    "  Lat=c(47.603260, \n",
    "        47.543877) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(crs(mask_Morb))  %>% \n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)\n",
    "\n",
    "Morbihan_plot<-ggplot() +\n",
    "  scale_fill_manual(\"\", labels = c(\"Intertidal area\", \"Land mask\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\"))+\n",
    "  geom_sf(data = mask_Morb, mapping = aes(fill = Type),linewidth=0.05,alpha=0.93,colour=\"grey30\")+\n",
    "      coord_sf(xlim=c(503019,526177) ,\n",
    "          ylim=c(5257824,5280982))+\n",
    "  ggforce::geom_mark_ellipse(data=Flight_Morbihan,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = Name,\n",
    "                     group = Name),\n",
    "                     # description=Description),\n",
    "                 size=0.3,\n",
    "                 fill = \"goldenrod\",\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 show.legend=F,\n",
    "                 con.size = 2,\n",
    "                 label.hjust = 0.5,\n",
    "                 label.fontsize = 25,\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") ,\n",
    "  label.fill = \"grey90\",\n",
    "  label.buffer = unit(5, \"mm\")) +\n",
    "  theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  geom_label(aes(x = 505000, y = 5280982-1981, label = \"A\"), size = 20)+\n",
    "  scale_x_continuous(breaks = seq(-3, -2.6, by = 0.1))+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\"),\n",
    "        legend.position = \"none\",\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20))\n",
    "\n",
    "\n",
    "  ggsave(\"Figures/Figure1/Map_Morbihan.png\",width= 10, height=10)  "
   ],
   "id": "91f7899c-6839-4114-9cca-b37f16a0394a"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "layout <- c(\n",
    "  patchwork::area(t = 1, l = 1, b = 10, r = 10),\n",
    "  patchwork::area(t = 1, l = 11, b = 10, r = 20),\n",
    "  patchwork::area(t = 11, l = 1, b = 20, r = 10),\n",
    "  patchwork::area(t = 11, l = 11, b = 20, r = 20)\n",
    ")\n",
    "# plot(layout)\n",
    "\n",
    "Tot <- magick::image_ggplot(magick::image_read(\"Figures/Figure1/Map_Drone_Sites.png\"), interpolate = T)\n",
    "BB <- magick::image_ggplot(magick::image_read(\"Figures/Figure1/Map_BB.png\"), interpolate = T)\n",
    "Mor <- magick::image_ggplot(magick::image_read(\"Figures/Figure1/Map_Morbihan.png\"), interpolate = T)\n",
    "Av <- magick::image_ggplot(magick::image_read(\"Figures/Figure1/Map_Aveiro.png\"), interpolate = T)\n",
    "\n",
    "\n",
    "a<-Tot+Mor+BB+Av+patchwork::plot_layout(design=layout)\n",
    "\n",
    "ggsave(\"./Figures/Fig1_Map_Drone_Sites.png\",a, width= 10, height=10, dpi = 800)"
   ],
   "id": "efc1e7e4-613a-4772-83e7-9bd0e6997e72"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Fig1_Map_Drone_Sites.png\")"
   ],
   "id": "cell-fig-map"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Drone Data\n",
    "\n",
    "At each location, a DJI Matrice 200 quadcopter drone equipped with a Micasense RedEdge Dual MX multispectral camera was flown to take 1.2 million pixel reflectance photographs in ten spectral bands, from 444 nm (blue) to 840 nm (near infrared, NIR). An angle of 90° was maintained between the sun and the drone’s route to guarantee uniform lighting conditions between flight lines. A side overlap of 70% and a front overlap of 80% between each image was set for each flight. A downwelling light sensor (DLS2) was used to acquire irradiance data concomitantly with the camera measurements. Raw data were calibrated in reflectance using a calibration panel reflective at ~50% provided by the manufacturer. A structure-from-motion photogrammetry software (Agisoft Metashape) was used to process images to obtain multispectral orthomosaics of each flight. The workflow for orthomosaicking was the same for every flight. First, tying key points were detected inside of each image and between overlapping images in order to obtain a sparse point cloud. This cloud was cleaned using reprojection accuracy metric in order to remove noisy points. A dense point cloud was then produced using a structure from motion algorithm. A surface interpolation of this dense point cloud was made to obtain a digital surface model (DSM), used to reconstruct the multispectral ortho-image. Across all sites, flights were made at two different altitudes : 12 m or/and 120 m. Low altitude drone flights produce ortho-images with a very high spatial resolution (8 mm per pixel), making it efficient to visually distinguish between the various types of vegetation. High altitude fights on the other hand allow to cover large areas and produced images with a pixel size of 80 mm (<a href=\"#tbl-flights\" class=\"quarto-xref\">Table 1</a>)."
   ],
   "id": "e73dc981-9c5e-41db-a2ae-8aef76c76711"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df<-data.frame(\n",
    "  Country = c(\n",
    "    rep(\"France\",5),\n",
    "    rep(\"Portugal\",4)\n",
    "  ),\n",
    "  Site = c(\n",
    "    rep(\"Gulf of Morbihan\",3),\n",
    "    rep(\"Bourgneuf Bay\",2),\n",
    "    rep(\"Aveiro Lagoon\",4)\n",
    "  ),\n",
    "  Name = c(\n",
    "    \"Arz Island\",\n",
    "    \"Duer\",\n",
    "    \"Duer\",\n",
    "    \"Barbâtre\",\n",
    "    \"L'Epine\",\n",
    "    \"Inner Lagoon\",\n",
    "    \"Mataducos\",\n",
    "    \"Gafanha\",\n",
    "    \"Gafanha\"\n",
    "  ),\n",
    "  Altitude = c(\n",
    "    \"12m\",\n",
    "    \"12m\",\n",
    "    \"120m\",\n",
    "    \"120m\",\n",
    "    \"120m\",\n",
    "    \"120m\",\n",
    "    \"120m\",\n",
    "    \"120m\",\n",
    "    \"12m\"\n",
    "  ),\n",
    "  Utility = c(\n",
    "    \"Training\",\n",
    "    \"Training\",\n",
    "    \"Validation\",\n",
    "    \"Validation\",\n",
    "    \"Validation\",\n",
    "    \"Validation\",\n",
    "    \"Validation\",\n",
    "    \"Validation\",\n",
    "    \"Training\"\n",
    "  )\n",
    ")\n",
    "\n",
    "brdr1 <- fp_border_default(color = \"black\", width = 1.5)\n",
    "brdr2 <- fp_border_default(color = \"grey\", width = 1.5)\n",
    "brdr3 <- fp_border_default(color = \"grey40\", width = 1.5)\n",
    "\n",
    " i = 16.5 # width of the side borders in the word_document output (in centimeters)\n",
    " w = i*0.3937 # width of the side borders in the word_document output (in inches)\n",
    " \n",
    "flx1<-flextable(df)  %>%\n",
    "  flextable::width(width = (w/(ncol(df)))) %>% \n",
    "  merge_v(j = 1) %>%\n",
    "  merge_v(j = 2) %>%\n",
    "    border_remove() %>%\n",
    "    hline_top(border=brdr1) %>%\n",
    "    hline(i=5, border=brdr1) %>%\n",
    "    hline(i=1,j = 3:5, border=brdr2) %>%\n",
    "    hline(i=2, border=brdr2) %>%\n",
    "    hline(i=3, j = 2:5, border=brdr3) %>%\n",
    "    hline(i=4, j = 3:5, border=brdr2) %>%\n",
    "    hline(i=6, border=brdr2) %>%\n",
    "    hline(i=6, j = 1:2, border=brdr1) %>%\n",
    "    hline(i=7, border=brdr2) %>%\n",
    "    hline(i=8, border=brdr2) %>%\n",
    "    hline(i=9, border=brdr1) %>%\n",
    "    vline(j=2, border = brdr1) %>%\n",
    "    flextable::align(align = \"center\",part = \"all\") %>%\n",
    "    flextable::align(align = \"center\",part = \"header\") %>%\n",
    "  # set_caption(caption = \"Table 1 List of drone Flight, summarising the date, the altitude and the purpose of each flight.\") %>%\n",
    "  set_table_properties(layout = \"autofit\") \n",
    "\n",
    "\n",
    "\n",
    "save_as_image(flx1, \"Figures/table_flights.png\", res = 300)"
   ],
   "id": "1038b8e0-d3e7-4bc1-8c97-a984ec029b8f"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "knitr::include_graphics(\"Figures/table_flights.png\")"
   ],
   "id": "ff27ee24-c0f5-4e29-ba5a-b0f7450bd686"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3 Field sampling\n",
    "\n",
    "Before each flight, targets used as Ground Control Points (GCPs) were distributed over the study site and georeferenced with a Trimble © Geo XH 6000 differential GPS (dGPS). GCPs were used to correct georeferencing imprecision of orthomosaics with an horizontal and vertical accuracy of 10cm. dGPS was also used to georeference quadrats of 0.25 m² assessing the presence or absence of five key taxonomic classes of intertidal vegetation : Bacillariophyceae (Unicellular benthic diatoms forming large biofilms at the sediment surface during low tide), Phaeophyceae (brown macroalgae), Magnoliopsida (seagrass), Chlorophyceae (green macroalgae) and Rhodophyceae (red macroalgae) (<a href=\"#fig-vegetation\" class=\"quarto-xref\">Figure 2</a>). Pictures of each quadrat were uploaded online on the Global Biodiversity Information Facility (GBIF) platform ([Davies et al., 2023b](#ref-BedeGbif)), a public open portal to store and share biodiversity data. Each photograph was also processed to estimate the percent cover of each type of vegetation using an image processing software (imageJ). For all quadrats, the hyperspectral reflectance signatures of each vegetation class was measured using an ASD FieldSpec HandHeld 2 spectroradiometer, which acquires reflectance between 325 and 1075 nm, with 1 nm of spectral resolution. Hyperspectral signatures from quadrats serve dual purposes: they validate the radiometric calibration of drone data and contribute to error reduction in photo interpretation. The workflow developped in this study is presented in <a href=\"#fig-workflow\" class=\"quarto-xref\">Figure 3</a>."
   ],
   "id": "3325dab2-30ee-4684-9dbd-7e19dd9a51bf"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "micasense<-list.files(pattern = \"Wavelength_Micasense_DualMX.csv\",recursive = T) %>% \n",
    "  read.delim(sep = \";\")\n",
    "\n",
    "imglist<-\n",
    "  data.frame(path = list.files(pattern = \".tif\", recursive = T, full.names = T)) %>% \n",
    "  mutate(folder = case_when(str_detect(path, \"Data\") ~ \"Input\",\n",
    "                            TRUE ~ \"Output\"),\n",
    "         alt = case_when(str_detect(path, \"high\") ~ \"high\",\n",
    "                         str_detect(path, \"High\") ~ \"high\",\n",
    "                         TRUE ~ \"low\"),\n",
    "         site = case_when(str_detect(path, \"Gafanha\") ~ \"Gafanha\",\n",
    "                          str_detect(path, \"Arz\") ~ \"Arz\",\n",
    "                          TRUE ~ \"Duer\"),\n",
    "         type = case_when(str_detect(path, \"pred\") ~ \"Pred\",\n",
    "                          str_detect(path, \"NDVI\") ~ \"NDVI\",\n",
    "                          str_detect(path, \"Pred\") ~ \"Pred\",\n",
    "                          TRUE ~ \"ortho\"),\n",
    "         resampled = case_when(str_detect(path, \"resampled\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         correction = case_when(str_detect(path, \"corrected_per_bands\") ~ \"Bands\",\n",
    "                          str_detect(path, \"corrected.tif\") ~ \"lm_all\",\n",
    "                          TRUE ~ \"NA\")) %>% \n",
    "  dplyr::filter(alt != \"NA\",\n",
    "                !str_detect(path,\".xml\"),\n",
    "                !str_detect(path,\".enp\"))\n",
    "\n",
    "gaf_low_multi<- imglist %>% \n",
    "  filter(site == \"Gafanha\",\n",
    "         alt == \"low\",\n",
    "         type == \"ortho\",\n",
    "         resampled == \"FALSE\",\n",
    "         folder == \"Input\") %>%\n",
    "  slice(1) %>% \n",
    "  pull(path) %>%\n",
    "  rast() \n",
    " names(gaf_low_multi)<-c(paste0(\"B\",c(1:10)))\n",
    "\n",
    " Gafanha_low_pred<-\"./Data/image/prediction/NewPreds/AveiroLowPrediction_nnNewStanRawMetrics_Preds_NoXantho_7.tif\" %>% \n",
    "     rast()\n",
    " names(Gafanha_low_pred)<-\"pred\"\n",
    "\n",
    "\n",
    "# arz_multi<- imglist %>% \n",
    "#   filter(site == \"Arz\", \n",
    "#          type == \"ortho\") %>% \n",
    "#   slice(1) %>% \n",
    "#   pull(path) %>% \n",
    "#   rast()\n",
    "#  names(arz_multi)<-c(paste0(\"B\",c(1:10)))\n",
    "# \n",
    "# arz_pred<- imglist %>% \n",
    "#   filter(site == \"Arz\", \n",
    "#          type == \"Pred\",\n",
    "#          str_detect(path, \"7by7\")) %>% \n",
    "#   slice(1) %>% \n",
    "#   pull(path) %>% \n",
    "#   rast()\n",
    "#  names(arz_pred)<-\"pred\"\n",
    " \n",
    " \n",
    " \n",
    "gaf<-rast(list(gaf_low_multi,Gafanha_low_pred)) %>% \n",
    "  as.data.frame(xy = F) %>% \n",
    "  dtplyr::lazy_dt() %>% \n",
    "  dplyr::filter(B1 != 65535,\n",
    "                pred != 0) %>%\n",
    "  as_tibble() %>% \n",
    "  rename(Class = \"pred\") %>%\n",
    "  mutate(pxl_ID = c(1:nrow(.))) %>% \n",
    "  pivot_longer(-c(Class,pxl_ID), names_to = \"band\",values_to = \"value\") %>% \n",
    "  left_join(micasense, by = c(\"band\" = \"Band_ID\")) %>% \n",
    "  group_by(Class, Wavelength) %>% \n",
    "  dplyr::summarise(mean = mean(value),\n",
    "                   sd = sd(value),\n",
    "                   max = max(value),\n",
    "                   min = min(value),\n",
    "                   median = median(value)) %>%  \n",
    "  mutate(Class = case_when(Class == 1 ~ \"Bare Sediment\",\n",
    "                              Class == 2 ~ \"D - Chloropyceae\",\n",
    "                              Class == 3 ~ \"Bare Sediment\",\n",
    "                              Class == 4 ~ \"A - MagnoliopsidaL\",\n",
    "                              Class == 5 ~ \"E - Microphytobenthos\",\n",
    "                              Class == 6 ~ \"A - Magnoliopsida\",\n",
    "                              Class == 7 ~ \"B - Phaeophyceae\",\n",
    "                              Class == 8 ~ \"C - Rhodophyceae\",\n",
    "                              Class == 9 ~ \"Sun Glint\",\n",
    "                              Class == 10 ~ \"Water\",\n",
    "                              Class == 11 ~ \"Xanthophyceae\",\n",
    "                              TRUE ~ \"NA\"))\n",
    "\n",
    "spectra<-gaf %>% \n",
    "  dplyr::filter(Class != \"Bare Sediment\",\n",
    "                Class != \"Sun Glint\",\n",
    "                Class != \"Water\",\n",
    "                Class != \"Xanthophyceae\",\n",
    "                Class != \"A - MagnoliopsidaL\") %>% \n",
    "  ggplot(aes(x=Wavelength, y = mean/10000, color = Class))+\n",
    "  geom_line(linewidth = 0.5)+\n",
    "  scale_color_manual(values=c(\"#389318\", \"#873e23\", \"#b3002d\", \"#b3ff1a\", \"#DAA520\"))+\n",
    "  ylab(\"Reflectance\")+\n",
    "  xlab(\"Wavelength (nm)\")+\n",
    "  theme_Bede()+\n",
    "  theme(axis.text.x = element_text(size = 5),\n",
    "        axis.text.y = element_text(size = 5),\n",
    "        axis.title.x = element_text(size = 7),\n",
    "        axis.title.y = element_text(size = 7),\n",
    "        legend.position = c(0.2, 0.7),\n",
    "        plot.background = element_rect(fill = rgb(1,1,1, alpha=0), colour = NA),\n",
    "        legend.title=element_blank(),\n",
    "        legend.text = element_text(size = 5),\n",
    "        legend.key.size = unit(0.25, \"cm\"),\n",
    "        legend.text.align = 0,\n",
    "        panel.border = element_blank(), \n",
    "        panel.background = element_blank(),\n",
    "        axis.line = element_line(color=\"black\"))\n",
    "\n",
    "ggsave(\"Figures/Spectral_shapes.png\",spectra, bg = \"transparent\",  height = 2, width = 3, dpi = 2000)\n",
    "\n",
    "sp <- image_ggplot(magick::image_read(\"Figures/Spectral_shapes.png\"), interpolate = T)\n",
    "\n",
    "# sp<-ggplot()+\n",
    "#   geom_spatraster_rgb(data = rast(\"Figures/Spectral_shapes.png\"),maxcell = 4009008)+\n",
    "#   coord_equal()+\n",
    "#     theme_void()+\n",
    "#   theme(axis.text = element_blank(), \n",
    "#         axis.title = element_blank(), \n",
    "#         axis.ticks = element_blank(),\n",
    "#         panel.background = element_rect(fill='transparent'),\n",
    "#         plot.background = element_rect(fill='transparent', color=NA),\n",
    "#         axis.ticks.length = unit(0, \"pt\"),\n",
    "#         panel.grid.major=element_blank(), \n",
    "#         panel.grid.minor=element_blank(), \n",
    "#         plot.margin = margin(0, 0, 0, 0, \"pt\"))\n",
    "# \n",
    "# \n",
    "# Magno <- magick::image_read(\"Data/Pictures/Figure2/Magnoliopsida_cropped2.png\")%>%\n",
    "#   # magick::image_rotate(90)%>%\n",
    "#   grid::rasterGrob()\n",
    "# \n",
    "# Brown <- magick::image_read(\"Data/Pictures/Figure2/Phaeophyceae_cropped.png\")%>%\n",
    "#   grid::rasterGrob()\n",
    "# \n",
    "# Red <- magick::image_read(\"Data/Pictures/Figure2/Rhodophyceae_cropped.png\")%>% \n",
    "#   grid::rasterGrob()\n",
    "# \n",
    "# green <- magick::image_read(\"Data/Pictures/Figure2/Chlorophyceae_cropped.png\")%>% \n",
    "#   grid::rasterGrob()\n",
    "# \n",
    "# MPB <- magick::image_read(\"Data/Pictures/Figure2/Bacillariophyceae_cropped.png\")%>% \n",
    "#   grid::rasterGrob()\n",
    "# \n",
    "# Lines<-data.frame(x=c(10,20,10),\n",
    "#                   y=c(-0.1,-0.1,10),\n",
    "#                   xend=c(10,20,30),\n",
    "#                   yend=c(20,20,10))\n",
    "# labels<-data.frame(name = c(\"A\",\"B\",\"C\",\"D\",\"E\"),\n",
    "#                    x= c(1,11,21,11,21),\n",
    "#                    y=c(19,19,19,1,1))\n",
    "# \n",
    "# a<-ggplot(expand_grid(x = 0:30,y=0:20),aes(x = x, y = y) ) +\n",
    "#   geom_point(alpha=0) +\n",
    "#   coord_fixed()+\n",
    "#   ggplot2::annotation_custom(Magno, 0,10, 0, 20) +\n",
    "#   ggplot2::annotation_custom(green, 10,20, 10, 20)+\n",
    "#   ggplot2::annotation_custom(Red, 10,20, 0, 10)+\n",
    "#   ggplot2::annotation_custom(Brown, 20,30, 10, 20)+\n",
    "#   ggplot2::annotation_custom(MPB, 20,30, 0,10)+\n",
    "#   geom_segment(data=Lines,aes(x=x,y=y,xend=xend,yend=yend),\n",
    "#                colour=\"white\",linewidth=2)+\n",
    "#   ggplot2::annotation_custom(ggplotGrob(spectra), 5,25, 5,15)+\n",
    "#   geom_label(data=labels,aes(x=x,y=y,label=name), size = 10)+\n",
    "#   theme_void()+\n",
    "#   theme(axis.title.x = element_blank(),\n",
    "#         axis.title.y = element_blank(),\n",
    "#         axis.text.x = element_blank(),\n",
    "#         axis.text.y = element_blank(),\n",
    "#         axis.ticks.x = element_blank(),\n",
    "#         axis.ticks.y = element_blank(),\n",
    "#         plot.margin = grid::unit(c(0.01,0.01,0.01,0.01), \"mm\"),\n",
    "#         # rect = element_blank(),\n",
    "#         panel.margin = grid::unit(c(0.01, 0.01, 0.01, 0.01), \"mm\"))\n",
    "# \n",
    "# Lines1<-data.frame(x=c(0,10,20,30),\n",
    "#                   y=c(10,0,0,0),\n",
    "#                   xend=c(40.1,10,20,30),\n",
    "#                   yend=c(10,10,10,10))\n",
    "# \n",
    "# labels1<-data.frame(name = c(\"A\",\"B\",\"C\",\"D\",\"E\"),\n",
    "#                    x= c(1,1,11,21,31),\n",
    "#                    y=c(29,9,9,9,9))\n",
    "# \n",
    "# ggplot(expand_grid(x = 0:40,y=0:30),aes(x = x, y = y) ) +\n",
    "#   geom_point(alpha=0) +\n",
    "#   coord_fixed()+\n",
    "#   ggplot2::annotation_custom(Magno, 0,10, 10, 30) +\n",
    "#   ggplot2::annotation_custom(Brown, 0,10, 0, 10)+\n",
    "#   ggplot2::annotation_custom(Red, 10,20, 0, 10)+\n",
    "#   ggplot2::annotation_custom(green, 20,30, 0, 10)+\n",
    "#   ggplot2::annotation_custom(MPB, 30,40, 0,10)+\n",
    "#   geom_segment(data=Lines1,aes(x=x,y=y,xend=xend,yend=yend),\n",
    "#                colour=\"white\",linewidth=2)+\n",
    "#   ggplot2::annotation_custom(ggplotGrob(spectra), 10,40, 10,30)+\n",
    "#   geom_label(data=labels1,aes(x=x,y=y,label=name), size = 10)+\n",
    "#   theme_void()+\n",
    "#   theme(axis.title.x = element_blank(),\n",
    "#         axis.title.y = element_blank(),\n",
    "#         axis.text.x = element_blank(),\n",
    "#         axis.text.y = element_blank(),\n",
    "#         axis.ticks.x = element_blank(),\n",
    "#         axis.ticks.y = element_blank(),\n",
    "#         plot.margin = grid::unit(c(0.01,0.01,0.01,0.01), \"mm\"),\n",
    "#         # rect = element_blank(),\n",
    "#         panel.margin = grid::unit(c(0.01, 0.01, 0.01, 0.01), \"mm\"),\n",
    "#         legend.text.align = 0)\n",
    "\n",
    "Magno<-ggplot()+\n",
    "  geom_spatraster_rgb(data = rast(\"Data/Pictures/Figure2/Magnoliopsida_cropped2.png\"))+\n",
    "  coord_equal()+\n",
    "  geom_label(aes(x = 50,y= 950, label = \"A\"))+\n",
    "    theme_void()+\n",
    "  theme(axis.text = element_blank(), \n",
    "        axis.title = element_blank(), \n",
    "        axis.ticks = element_blank(), \n",
    "        axis.ticks.length = unit(0, \"pt\"),\n",
    "        panel.grid.major=element_blank(), \n",
    "        panel.grid.minor=element_blank(), \n",
    "        plot.margin = margin(0, 0, 0, 0, \"pt\"))\n",
    "\n",
    "Brown<-ggplot()+\n",
    "  geom_spatraster_rgb(data = rast(\"Data/Pictures/Figure2/Pheophyceae_cropped.png\"))+\n",
    "  coord_equal()+\n",
    "  geom_label(aes(x = 50,y=450, label = \"B\"))+\n",
    "    theme_void()+\n",
    "  theme(axis.text = element_blank(), \n",
    "        axis.title = element_blank(), \n",
    "        axis.ticks = element_blank(), \n",
    "        axis.ticks.length = unit(0, \"pt\"),\n",
    "        panel.grid.major=element_blank(),\n",
    "        panel.grid.minor=element_blank(),\n",
    "        plot.margin = margin(0, 0, 0, 0, \"pt\"))\n",
    "\n",
    "Red <- ggplot()+\n",
    "  geom_spatraster_rgb(data = rast(\"Data/Pictures/Figure2/Rodophyceae_cropped.png\"))+\n",
    "  coord_equal()+\n",
    "  geom_label(aes(x = 50,y=450, label = \"C\"))+\n",
    "    theme_void()+\n",
    "  theme(axis.text = element_blank(), \n",
    "        axis.title = element_blank(), \n",
    "        axis.ticks = element_blank(), \n",
    "        axis.ticks.length = unit(0, \"pt\"),\n",
    "        panel.grid.major=element_blank(), \n",
    "        panel.grid.minor=element_blank(), \n",
    "        plot.margin = margin(0, 0, 0, 0, \"pt\"))\n",
    "\n",
    "\n",
    "green <- ggplot()+\n",
    "  geom_spatraster_rgb(data = rast(\"Data/Pictures/Figure2/Chlorophyceae_cropped.png\"))+\n",
    "  coord_equal()+\n",
    "  geom_label(aes(x = 50,y=450, label = \"D\"))+\n",
    "    theme_void()+\n",
    "  theme(axis.text = element_blank(), \n",
    "        axis.title = element_blank(), \n",
    "        axis.ticks = element_blank(), \n",
    "        axis.ticks.length = unit(0, \"pt\"),\n",
    "        panel.grid.major=element_blank(), \n",
    "        panel.grid.minor=element_blank(), \n",
    "        plot.margin = margin(0, 0, 0, 0, \"pt\"))\n",
    "  \n",
    "  \n",
    "\n",
    "MPB <- ggplot()+\n",
    "  geom_spatraster_rgb(data = rast(\"Data/Pictures/Figure2/Bacillariophyceae_cropped.png\"))+\n",
    "  coord_equal()+\n",
    "  geom_label(aes(x = 50,y=450, label = \"E\"))+\n",
    "  theme_void()+\n",
    "  theme(axis.text = element_blank(), \n",
    "        axis.title = element_blank(), \n",
    "        axis.ticks = element_blank(), \n",
    "        axis.ticks.length = unit(0, \"pt\"),\n",
    "        panel.grid.major=element_blank(), \n",
    "        panel.grid.minor=element_blank(), \n",
    "        plot.margin = margin(0, 0, 0, 0, \"pt\"))\n",
    "\n",
    "layout <- c(\n",
    "  patchwork::area(t = 10, l = 0, b = 30, r = 10),\n",
    "    patchwork::area(t = 0, l = 0, b = 10, r = 10),\n",
    "    patchwork::area(t = 0, l = 10, b = 10, r = 20),\n",
    "    patchwork::area(t = 0, l = 20, b = 10, r = 30),\n",
    "    patchwork::area(t = 0, l = 30, b = 10, r = 40),\n",
    "    patchwork::area(t = 10, l = 10, b = 30, r = 40)\n",
    ")\n",
    "\n",
    "layout1 <- c(\n",
    "  patchwork::area(t = 1, l = 1, b = 20, r = 10),\n",
    "    patchwork::area(t = 20, l = 1, b = 29, r = 10),\n",
    "    patchwork::area(t = 20, l = 9, b = 29, r = 18),\n",
    "    patchwork::area(t = 20, l = 17, b = 29, r = 26),\n",
    "    patchwork::area(t = 20, l = 25, b = 29, r = 34),\n",
    "    patchwork::area(t = 1, l = 9, b = 20, r = 34)\n",
    ")\n",
    "\n",
    "# plot(layout1)\n",
    "\n",
    "plot<-Magno+\n",
    "  Brown+\n",
    "  Red+\n",
    "  green+\n",
    "  MPB+\n",
    "  sp+\n",
    "  patchwork::plot_layout(design=layout1)\n",
    "\n",
    "ggsave(\"Figures/Spectral_shapes_total.png\",plot, height = 7, width = 10, dpi = 800)"
   ],
   "id": "b684c518-5cb5-43aa-87ff-b5dceb365491"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"Figures/Low_res/Spectral_shapes_total.png\")"
   ],
   "id": "cell-fig-vegetation"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.4 Intertidal vegetation mapping"
   ],
   "id": "f39c810a-e7f2-4c2b-a736-1e21186525df"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "layout<-\"Data/Figure2_Workflow\" %>%\n",
    "  list.files(\"V4.txt\",full.names = T, recursive = T) %>%\n",
    "  read_delim(delim = \"\\t\", escape_double = FALSE,\n",
    "    trim_ws = TRUE) %>%\n",
    "  dplyr::filter(Evaluate == T)\n",
    "\n",
    "\n",
    "layout$Text<- gsub(\"alaligne\",\"\\n\", layout$Text)\n",
    "\n",
    "boxes<- layout %>% \n",
    "  dplyr::filter(Type == \"Box\",\n",
    "                shape == \"square\")\n",
    "\n",
    "\n",
    "text<- layout %>% \n",
    "  dplyr::filter(Type == \"Box\")\n",
    "\n",
    "diamonds<- layout %>% \n",
    "  dplyr::filter(Type == \"Box\",\n",
    "                shape == \"diamond\")\n",
    "diff = 0.1\n",
    "for (i in 1:nrow(diamonds)) {\n",
    "  diam_a<-diamonds[i,]\n",
    "  \n",
    "df<-data.frame(  \n",
    "  x = c(diam_a$xmin - diff, diam_a$xmin + diff, diam_a$xmax + diff, diam_a$xmax - diff),\n",
    "  y = c(diam_a$ymin, diam_a$ymax, diam_a$ymax, diam_a$ymin),\n",
    "  Type = rep(diam_a$Type,4),\n",
    "  Text = rep(diam_a$Text,4),\n",
    "  Color = rep(diam_a$color,4),\n",
    "  Evaluate = rep(diam_a$Evaluate,4),\n",
    "  Resolution =  rep(diam_a$Resolution,4),\n",
    "  Evaluate_bigbox = rep(diam_a$Evaluate_bigbox,4),\n",
    "  alpha = rep(diam_a$alpha,4),\n",
    "  shape =  rep(diam_a$shape,4)\n",
    ")\n",
    "  if(i == 1){\n",
    "    diamond_list<-list(df)\n",
    "  }else{\n",
    "    diamond_list[[i]]<-df\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "Arrows<- layout %>% \n",
    "  dplyr::filter(Type == \"Arrow\",\n",
    "                Evaluate == T)\n",
    "\n",
    "Group_box<-layout %>% \n",
    "  dplyr::filter(Type == \"Box\",\n",
    "                Evaluate_bigbox == T) %>% \n",
    "  group_by(Group) %>% \n",
    "  dplyr::summarise(xmin = min(xmin)-0.1,\n",
    "                   xmax = max(xmax)+0.1,\n",
    "                   ymin = min(ymin)-0.1,\n",
    "                   ymax = max(ymax)+0.1,\n",
    "                   alpha = 0.1,\n",
    "                   color = color)\n",
    "\n",
    "resolution_box<-layout %>% \n",
    "  dplyr::filter(Type == \"Box\",\n",
    "                Resolution != \"NA\") %>% \n",
    "  group_by(Resolution) %>% \n",
    "  dplyr::summarise(xmin = min(xmin)-0.1,\n",
    "                   xmax = max(xmax)+0.1,\n",
    "                   ymin = min(ymin)-0.1,\n",
    "                   ymax = max(ymax)+0.1,\n",
    "                   alpha = 0.1,\n",
    "                   color = color)\n",
    "\n",
    "legend = data.frame(color = unique(layout$color)) %>% \n",
    "  dplyr::filter(!is.na(color)) %>% \n",
    "  mutate(label = case_when(color == \"#EA6155\" ~ \"Input Data\",\n",
    "                           color == \"#DC9614\" ~ \"Pre-processing\",\n",
    "                           color == \"#52A4DA\" ~ \"Model Building\",\n",
    "                           color == \"#79DA52\" ~ \"Validation\",\n",
    "                           color == \"#DA9652\" ~ \"Output Data\",\n",
    "                           color == \"darkcyan\" ~ \"Final Product\",\n",
    "                           color == \"#DA5273\"  ~ \"Prediction\",\n",
    "                           TRUE ~ \"NA\")) %>% \n",
    "  arrange\n",
    "\n",
    "segment<-layout %>% \n",
    "  dplyr::filter(Type == \"Segment\")"
   ],
   "id": "f307fd7c-07b9-4deb-8302-f935ee942e5a"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "building <- F\n",
    "\n",
    "if(building == T){\n",
    "  \n",
    "text_size = 6 #### 6 to plot it fullscreen in R,  1.8 for exporting with ggsave\n",
    "line_size =1 ####arrow_size 0.3 to export  1 to for R plotting\n",
    "\n",
    "legend_size = 1 #### 0.3 to export, 1 for R ploting\n",
    "legend_text = 10 #### 4 to export, 10 for R ploting\n",
    "}else{\n",
    "  \n",
    "text_size = 1.85 #### 6 to plot it fullscreen in R,  1.8 for exporting with ggsave\n",
    "line_size =0.3 ####arrow_size 0.3 to export  1 to for R plotting\n",
    "\n",
    "legend_size = 0.3 #### 0.3 to export, 1 for R ploting\n",
    "legend_text = 5 #### 5 to export, 10 for R ploting\n",
    "  \n",
    "}\n",
    "\n",
    "\n",
    "a<-ggplot()+\n",
    "  \n",
    "  statebins:::geom_rrect(data = Group_box,\n",
    "            mapping = aes(xmin = xmin,\n",
    "                          xmax= xmax,\n",
    "                          ymin = ymin,\n",
    "                          ymax = ymax),\n",
    "            alpha = 0.05,\n",
    "            size = 0.2,\n",
    "            show.legend = F,\n",
    "            fill = Group_box$color,\n",
    "            color = \"black\")+\n",
    "  \n",
    "    statebins:::geom_rrect(data = resolution_box,\n",
    "            mapping = aes(xmin = xmin,\n",
    "                          xmax= xmax,\n",
    "                          ymin = ymin,\n",
    "                          ymax = ymax,\n",
    "                          alpha = rep(c(\"Input Data\", \"Pre-processing\", \"Model Building\", \"Prediction\", \"Validation\", \"Output Data\", \"Final Product\"),2)),\n",
    "            fill = \"NA\",\n",
    "            linetype = 2,\n",
    "            size = 0.4,\n",
    "            color = \"black\")+\n",
    "  \n",
    "  scale_alpha_manual(name = \"\", \n",
    "                     values = c(1,1,1,1,1,1,1),\n",
    "                     labels = c(\"Input Data\", \"Pre-Processing\", \"Model Building\", \"Prediction\", \"Validation\", \"Output Data\", \"Final Product\"),\n",
    "                     guide = guide_legend(ncol = 1, override.aes = list(fill = c(\"#f94144\", \"#f3722c\", \"#f8961e\", \"#f9c74f\", \"#90be6d\", \"#43aa8b\", \"#577590\"),\n",
    "                                                                        shape = c(22),\n",
    "                                                                        linetype = c(1),\n",
    "                                                                        alpha = c(0.5)\n",
    "                                                                        )\n",
    "                                          )\n",
    "                     )+\n",
    "  \n",
    "  statebins:::geom_rrect(data = boxes,\n",
    "            mapping = aes(xmin = xmin,\n",
    "                          xmax= xmax,\n",
    "                          ymin = ymin,\n",
    "                          ymax = ymax\n",
    "                          ),\n",
    "            size = 0.2,\n",
    "            alpha = as.numeric(boxes$alpha),\n",
    "            fill = boxes$color,\n",
    "            color = \"black\")\n",
    "  \n",
    "  for(i in 1:length(diamond_list)){\n",
    "    poly_i<-diamond_list[[i]]\n",
    "    a<-a+\n",
    "      geom_polygon(data = poly_i, \n",
    "                   mapping = aes(x = x, \n",
    "                                 y = y\n",
    "                                 ),\n",
    "                  size = 0.2,\n",
    "                   alpha = poly_i$alpha,\n",
    "                   fill = poly_i$Color,\n",
    "                   color = \"black\")\n",
    "    \n",
    "  }\n",
    "\n",
    "if(building == T){\n",
    "    a<-a+\n",
    "    geom_text(data = text,\n",
    "            mapping = aes(x = (xmax+xmin)/2, y = (ymax + ymin)/2, label = (ID)), size = text_size)\n",
    "  \n",
    "}else{\n",
    "    a<-a+\n",
    "    geom_text(data = text,\n",
    "            mapping = aes(x = (xmax+xmin)/2, y = (ymax + ymin)/2, label = (Text)), size = text_size)\n",
    "}\n",
    "\n",
    "\n",
    "  a<-a+\n",
    "    geom_segment(data = Arrows, aes( x = xmin, xend = xmax, y = ymin, yend = ymax),\n",
    "    size=line_size, linejoin = \"mitre\", lineend = \"butt\",\n",
    "    arrow = arrow(length = unit(1, \"mm\"), type= \"closed\")\n",
    "    )+\n",
    "    geom_segment(data = segment, aes( x = xmin, xend = xmax, y = ymin, yend = ymax),\n",
    "    size=line_size, linejoin = \"mitre\", lineend = \"square\"\n",
    "    )+\n",
    "  \n",
    "  geom_text(aes(x = -5.4, y = -2, label = \"Flight height: 12 m \\n Pixel size : 8 mm\", angle = 90), size = text_size+1)+\n",
    "  \n",
    "  geom_text(aes(x = -5.4, y = -6.8, label = \"Flight height: 120 m \\n Pixel size : 80 mm\", angle = 90), size = text_size+1)+\n",
    "  \n",
    "  geom_text(aes(x = 4, y = 1.75, label = \"Pre Processing\"), size = text_size)\n",
    "    \n",
    "  if(building == F){\n",
    "  a<- a+ theme_void()\n",
    "  }\n",
    "  \n",
    "  a<-a+\n",
    "  theme(legend.position = c(0.88,0.3),\n",
    "        legend.text = element_text(size = legend_text),\n",
    "        legend.key.size = unit(legend_size, 'cm')\n",
    "        )+\n",
    "  scale_x_continuous(breaks = scales::breaks_width(0.5))+\n",
    "  scale_y_continuous(breaks = scales::breaks_width(0.5))\n",
    "  \n",
    "if(building == T){\n",
    "a\n",
    "}else{\n",
    "  ggsave(\"Figures/Figure3_workflow.png\",a, width = 1920, height = 1129, units = \"px\")\n",
    "}"
   ],
   "id": "57681a17-340a-4a52-8cb1-94c2fbfae580"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Figure3_workflow.png\")"
   ],
   "id": "cell-fig-workflow"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The spectral similarities of the reflectance signatures between intertidal green macrophytes (Magnoliopsida and Chlorophyceae) make their discrimination challenging using simple classification algorithms (<a href=\"#fig-vegetation\" class=\"quarto-xref\">Figure 2</a> F). To overcome this challenge, a deep learning classification method was developed, trained, validated, and applied to each drone flight (<a href=\"#fig-workflow\" class=\"quarto-xref\">Figure 3</a>).\n",
    "\n",
    "### 2.4.1 Neural Network model building"
   ],
   "id": "77339cea-9be3-4165-bdc8-f9252b73ba81"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_training<-\"Data/training/TrainandValid_NewStanRawMetrics.csv\" %>% \n",
    "  read_delim(delim = \";\", escape_double = FALSE, trim_ws = TRUE) %>% \n",
    "  dplyr::filter(True_Class != \"Xanthophyceae\",\n",
    "                True_Class != \"Deep_Sediment\",\n",
    "                True_Class != \"Water\",\n",
    "                True_Class != \"SunGlint\",\n",
    "                True_Class != \"Bare_Sediment\") %>% \n",
    "  mutate(True_Class = case_when(True_Class == \"Clorophyta\" ~ \"Chlorophyceae\",\n",
    "                               True_Class == \"Rhodphyta\" ~ \"Rhodophyceae\",\n",
    "                               True_Class == \"MPB\" ~ \"Bacillariophyceae\",\n",
    "                               True_Class == \"Magnoliosida\" ~ \"Magnoliopsida\",\n",
    "                               True_Class == \"Phaeophyta\" ~ \"Phaeophyceae\",\n",
    "                               True_Class == \"Low_SPC\" ~ \"Magnoliopsida\",\n",
    "                               TRUE ~ \"NA\"\n",
    "                               )) %>% \n",
    "  group_by(True_Class) %>% \n",
    "  summarise(n = n()) %>% \n",
    "  rename(True = \"True_Class\", \n",
    "         Training_Pixels = \"n\")\n",
    "\n",
    "write.csv(df_training,\"Data/training/TrainandValid_NewStanRawMetrics_Processed.csv\",row.names = F)"
   ],
   "id": "33e833a4-fd94-4a2e-97e0-8c0d3bb5f058"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_training<-read.csv(\"Data/training/TrainandValid_NewStanRawMetrics_Processed.csv\")"
   ],
   "id": "7b54a67f-dba8-47fd-bd19-e3a7c582e5ef"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_validation<-\"Data/accuracy_assesment_low&high_flights.csv\" %>% \n",
    "  read_delim(delim = \",\", escape_double = FALSE, trim_ws = TRUE) %>% \n",
    "  dplyr::filter(True != \"Sediment\",\n",
    "                True != \"Sun Glint\",\n",
    "                True != \"Water\",\n",
    "                True != \"Xanthophyceae\") %>% \n",
    "  mutate(True = case_when(True == \"Green Algae\" ~ \"Chlorophyceae\",\n",
    "                          True == \"Microphytobenthos\" ~ \"Bacillariophyceae\",\n",
    "                          True == \"Rodophyceae\" ~ \"Rhodophyceae\",\n",
    "                          True == \"Pheophyceae\" ~ \"Phaeophyceae\",\n",
    "                          TRUE ~ True)) %>% \n",
    "  group_by(True) %>% \n",
    "  summarise(n = n()) \n",
    "\n",
    "write.csv(df_validation,\"Data/training/TrainandValid_NewStanRawMetrics_Processed_valid.csv\",row.names = F)"
   ],
   "id": "80f8cf4b-aa9f-45b0-87f7-4cb573d918f8"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_validation<-read.csv(\"Data/training/TrainandValid_NewStanRawMetrics_Processed_valid.csv\")\n",
    "\n",
    " i = 16.5 # width of the side borders in the word_document output (in centimeters)\n",
    " w = i*0.3937 # width of the side borders in the word_document output (in inches)\n",
    " \n",
    "df_flx2<-df_training %>% \n",
    "  left_join(df_validation, by = \"True\") %>% \n",
    "  rename(`Validation Pixels` = \"n\",\n",
    "         Class = \"True\",\n",
    "         `Training Pixels` = \"Training_Pixels\")\n",
    "flx2 <- df_flx2 %>% \n",
    "  flextable() %>% \n",
    "  flextable::width(width = (w/(ncol(df_flx2)))) %>% \n",
    "  # border_remove() %>%\n",
    "  flextable::align(align = \"center\",part = \"all\") %>%\n",
    "  flextable::align(align = \"center\",part = \"header\") %>% \n",
    "  set_table_properties(layout = \"autofit\") \n",
    "\n",
    "save_as_image(flx2, \"Figures/table_validation_px.png\", res = 300)"
   ],
   "id": "e5c783c0-697b-4c6f-b52e-ae8b8845874b"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "knitr::include_graphics(\"Figures/table_validation_px.png\")"
   ],
   "id": "58b959d9-14fd-468f-a71f-30f16729c815"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A dataset containing photo-interpreted drone reflectance pixels was built to train a Neural Network model with 2 hidden layers. The training pixels were categorized into seven different classes, representing the various habitats encountered at the different study sites: Sediment, Water, Chlorophyceae, Magnoliopsida, Bacillariophyceae, Phaeophyceae and Rhodophyceae. Only low-altitude flights (<a href=\"#tbl-flights\" class=\"quarto-xref\">Table 1</a>) were used for training because their 8 mm spatial resolution allowied to avoid spectral sub-pixel mixing and to accurately differentiate vegetation classes. More than 418,000 pixels at 8 mm resolution from the 3 training flights were used to train the model (<a href=\"#tbl-validationPX\" class=\"quarto-xref\">Table 2</a>). Twenty one variables were used by the model as predictors: the ten raw spectral bands of the Micasense RedEdge Dual MX multispectral camera (ranging from 444 nm to 840 nm), the same ten spectral bands standardized using a min/max transformation (<a href=\"#eq-std\" class=\"quarto-xref\">Equation 1</a> ; Cao et al. ([2017](#ref-Cao2017))) and the Normalized difference vegetation index (NDVI, <a href=\"#eq-ndvi\" class=\"quarto-xref\">Equation 2</a>). Standardisation of spectral bands is used to eliminate the scaling differences between spectra and to limit the effect of biomass on the shape of the spectra ([Douay et al., 2022](#ref-Douay2022) ; [Davies et al., 2023a](#ref-Davies2023)).\n",
    "\n",
    "<span id=\"eq-std\">$$\n",
    "R_{i}^{*}(\\lambda) = \\frac{R_{i}(\\lambda) - min(R_{i})}{max(R_{i})- min(R_{i})}\n",
    " \\qquad(1)$$</span>\n",
    "\n",
    "where $R_{i}(\\lambda)$ is the reflectance at the wavelength $(\\lambda)$ of each individual spectra $(i)$, $min(R_{i})$, and $max(R_{i})$ are the minimum and maximum value of the spectra $(i)$\n",
    "\n",
    "<span id=\"eq-ndvi\">$$\n",
    "NDVI = \\frac{R(840nm)-R(668nm)}{R(840nm)+R(668nm)}\n",
    " \\qquad(2)$$</span>\n",
    "\n",
    "where $R(840nm)$ is the reflectance at 840 nm and $R(668nm)$ is the reflectance at 668 nm.\n",
    "\n",
    "### 2.4.2 Validation"
   ],
   "id": "359b9c4c-849e-4d93-a1b6-d32b8902c7cd"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df<-\"Output/Validation_data\" %>% \n",
    "  list.files(\"ResultsNewNN_NoXanthowithNDVIWithOldandBoat.csv\",recursive = T, full.names = T) %>% \n",
    "  read.csv() %>% \n",
    "  as_tibble() %>% \n",
    "  rename(Truth_ID = \"Truth\") %>% \n",
    "  mutate(Truth =  case_when(Truth_ID == 1 ~ \"Bare_Sediment\",\n",
    "                          Truth_ID == 2 ~ \"Chlorophyta\",\n",
    "                          Truth_ID == 3 ~ \"Bare_Sediment\",\n",
    "                          Truth_ID == 4 ~ \"Magnoliosida\",\n",
    "                          Truth_ID == 5 ~ \"Bare_Sediment\",\n",
    "                          Truth_ID == 6 ~ \"Magnoliosida\",\n",
    "                          Truth_ID == 7 ~ \"Phaeophyta\",\n",
    "                          Truth_ID == 8 ~ \"Rhodphyta\",\n",
    "                          Truth_ID == 9 ~ \"SunGlint\",\n",
    "                          Truth_ID == 10 ~ \"Water\",\n",
    "                          Truth_ID == 11 ~ \"Xanthophyceae\",\n",
    "                          Truth_ID == 12 ~ \"Spartina\",\n",
    "                          Truth_ID == 41 ~ \"Bleached seagrasses\",\n",
    "                          TRUE ~ \"NA\"),\n",
    "         Pred_Class = case_when(Pred_Class == \"Clorophyta\" ~ \"Chlorophyta\",\n",
    "                                Pred_Class == \"Rhodphyta\" ~ \"Rhodphyta\",\n",
    "                                Pred_Class == \"Deep_Sediment\" ~ \"Bare_Sediment\",\n",
    "                                Pred_Class == \"MPB\" ~ \"Bare_Sediment\",\n",
    "                                (Pred_Class == \"Low_SPC\" & NDVI > 0.2) ~ \"Magnoliosida\",\n",
    "                                (Pred_Class == \"Low_SPC\" & NDVI <= 0.2 )~ \"Bare_Sediment\",\n",
    "                          TRUE ~ Pred_Class)) %>% \n",
    "  dplyr::filter(Truth != \"Water\",\n",
    "                Truth  != \"Spartina\",\n",
    "                Truth != \"SunGlint\",\n",
    "                Truth != \"Bleached seagrasses\",\n",
    "                Truth != \"Xanthophyceae\",\n",
    "                # Pred_Class != \"Low_SPC\",\n",
    "                Pred_Class != \"Xanthophyceae\",\n",
    "                 Pred_Class != \"Water\",\n",
    "                 Pred_Class != \"SunGlint\",\n",
    "                !Flight == \"Handfisher_1_old\",\n",
    "                !Flight == \"Handfisher_2_old\",\n",
    "                !Flight == \"Handfisher_2\") %>%\n",
    "  dplyr::select(-NDVI)\n",
    "\n",
    "\n",
    "\n",
    "count_validation_pixel<-nrow(df)\n",
    "write.csv(count_validation_pixel,\"Output/count_validation_pixel.csv\", row.names = F)"
   ],
   "id": "82aadc17-a287-4e7a-9de7-10ea18aab01e"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_validation_pixel<-read.csv(\"Output/count_validation_pixel.csv\") %>% \n",
    "  pull(x)"
   ],
   "id": "a8f45764-47ea-4a70-9c88-7412d2c7eb67"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The workflow of this study revolves around two distinct flight heights (12 m and 120 m, <a href=\"#fig-workflow\" class=\"quarto-xref\">Figure 3</a>) where ensuring consistency between reflectances at both heights is crucial. This comparison was conducted at sites where low and high altitude flights overlapped. The low altitude flights were resampled to the same spatial resolution and grid as the high flights using a median resampling method. Reflectance values were then extracted, and a scatterplot was generated. Subsequently, the slope of the linear model and the coefficient of determination (R²) were computed.\n",
    "\n",
    "The model was applied to all flights at both 12 and 120 m of altitude. In situ information on georeferenced class type and percent cover collected during each flight was used to assess the model accuracy. These images were used to construct a validation dataset indicating the presence or absence of each class. Additionally to the quadrat-based validation dataset, polygons of each class were photo interpreted in order to increase the number of pixels of the validation dataset. A confusion matrix, along with precision metrics such as global accuracy, sensitivity, specificity, and Kappa coefficient, was generated for each sites. All validation matrices were then merged to create a unique matrix . Altogether, a total of 536,000 pixels was used to validate the model, thus providing a geographically robust validation dataset.\n",
    "\n",
    "# 3. Results\n",
    "\n",
    "## 3.1 Classification\n",
    "\n",
    "A total of 9 prediction maps corresponding to the 9 drone flights were obtained. Each prediction map is associated with a probability map, indicating the probability of the selected class for every pixel. The low-altitude flight conducted in Gafanaha, Portugal, represents the site with the highest complexity (<a href=\"#fig-GafLow\" class=\"quarto-xref\">Figure 4</a>). Among the 5 vegetation classes on which the model was trained, 4 were present on this site. On this site, there is a mixture of Chlorophyceae and Rhodophyceae over the seagrass meadow. This is also where Bacillariophyceae is most abundant. Although the seagrass bed is solely composed of *Zostera noltei*, various colors can be observed: dark green (indicating healthy beds with 100% coverage) and whitish/brown (indicating beds where leaves are dying). Regardless of the color, the meadow is predicted as Magnoliopsida by the model."
   ],
   "id": "82c567d7-e858-43cb-9635-76bbfb9b110a"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imglist<-\n",
    "  data.frame(path = list.files(pattern = \".tif\", recursive = T, full.names = T)) %>% \n",
    "  mutate(folder = case_when(str_detect(path, \"Data\") ~ \"Input\",\n",
    "                            TRUE ~ \"Output\"),\n",
    "         alt = case_when(str_detect(path, \"high\") ~ \"high\",\n",
    "                         TRUE ~ \"low\"),\n",
    "         site = case_when(str_detect(path, \"Gafanha\") ~ \"Gafanha\",\n",
    "                          TRUE ~ \"Duer\"),\n",
    "         type = case_when(str_detect(path, \"pred\") ~ \"Pred\",\n",
    "                          str_detect(path, \"NDVI\") ~ \"NDVI\",\n",
    "                          TRUE ~ \"ortho\"),\n",
    "         resampled = case_when(str_detect(path, \"resampled\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         new = case_when(str_detect(path, \"no_Xantho\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         col = case_when(str_detect(path, \"colored\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         correction = case_when(str_detect(path, \"corrected_per_bands\") ~ \"Bands\",\n",
    "                          str_detect(path, \"corrected.tif\") ~ \"lm_all\",\n",
    "                          TRUE ~ \"NA\")) %>% \n",
    "  dplyr::filter(alt != \"NA\",\n",
    "                !str_detect(path,\".aux.xml\"))\n",
    "\n",
    "shp_roi<-\"Data/reflectance_calibration_low_&_high_flights/\" %>%\n",
    "  list.files(pattern = \".shp\", recursive = T, full.names = T) %>%\n",
    "  read_sf() %>%\n",
    "  mutate(ID = case_when(name == \"Microphytobenthos\" ~ 1,\n",
    "                        name == \"ZOS\" ~ 2,\n",
    "                        name == \"BL\" ~ 3,\n",
    "                        name == \"SED\" ~ 4,\n",
    "                        name == \"GRE\" ~ 5,\n",
    "                        name == \"RED\" ~ 6,\n",
    "                        name == \"Xanthophyceae\" ~ 7)) %>% \n",
    "  as(\"Spatial\") %>% \n",
    "  vect()\n",
    "\n",
    "\n",
    "zoom_extent<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"zoom_extent_map_Gafanha_comparison\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "background_GafHigh<-\"Data/shp/Aveiro\" %>% \n",
    "  list.files(\".shp\", full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path, \"background_Gafanha_high\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "###### Preview ###########\n",
    "\n",
    "filelist<-list.files(\"Data/shp/\", pattern = \".shp\", recursive = T,full.names = T)\n",
    "mask_aveiro <- filelist[str_detect(filelist,\"mask_land_intertidal_Aveiro\")] %>% \n",
    "  read_sf()\n",
    "\n",
    "Flight_aveiro<-data.frame(\n",
    "  Name=c(\n",
    "  \"Gafanha\"\n",
    "  ) ,\n",
    "  Description = c(\"12m and 120m Flights\"),\n",
    "  Long=c(-8.742485) ,\n",
    "  Lat=c(40.597599) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(crs(mask_aveiro))  %>% \n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)\n",
    "\n",
    "\n",
    "p1_background<- read_sf(\"Data/shp/Aveiro/Background_p1_aveiro.shp\")\n",
    "\n",
    "p1<-ggplot() +\n",
    "  scale_fill_manual(\"\", labels = c(\"Intertidal area\", \"Land area\"),\n",
    "                    values = c(\"#7DC27D\", \"#BDBDBD\"))+\n",
    "  geom_sf(data = p1_background, fill = \"white\",show.legend = F, linewidth=0.05,alpha=0.93,colour=\"transparent\")+\n",
    "  geom_sf(data = mask_aveiro, mapping = aes(fill = Type),show.legend = F, linewidth=0.05,alpha=0.93,colour=\"grey30\")+\n",
    "\n",
    "      coord_sf(xlim=c(519749,535233) ,\n",
    "          ylim=c(4493310,4508794) )+\n",
    "  ggforce::geom_mark_ellipse(data=Flight_aveiro,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = Name,\n",
    "                     group = Name),\n",
    "                     # description=Description),\n",
    "                 size=0.3,\n",
    "                 fill = \"goldenrod\",\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 show.legend=F,\n",
    "                 label.fontsize = 18,\n",
    "                 label.hjust = 0.5,\n",
    "                 con.size = 2,\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") ,\n",
    "  label.fill = \"grey90\",\n",
    "  label.buffer = unit(5, \"mm\")) +\n",
    "    # theme_void()+\n",
    "  theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  # geom_label(aes(x = 521000, y = 4507794, label = \"C\"), size = 20)+\n",
    "  scale_x_continuous(breaks = seq(-8.75, -8.6, by = 0.1))+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\") ,\n",
    "        legend.position = \"none\",\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20))+\n",
    "  theme_void()\n",
    "######### RGB #########\n",
    " Gafanha_low_ortho<-imglist %>% \n",
    "  filter(site == \"Gafanha\",\n",
    "         alt == \"low\",\n",
    "         type == \"ortho\",\n",
    "         resampled == \"TRUE\",\n",
    "         col == \"TRUE\") %>%\n",
    "  pull(path) %>% \n",
    "  rast() \n",
    " names(Gafanha_low_ortho)<-c(\"B1\", \"B3\", \"B5\",\"alpha\")\n",
    "\n",
    " values(Gafanha_low_ortho)[values(Gafanha_low_ortho$alpha) == 0] = NA\n",
    "  \n",
    " Gafanha_low_ortho_zoomed<-Gafanha_low_ortho %>% \n",
    "   crop(zoom_extent)\n",
    " \n",
    " \n",
    " map_extent_full<- st_as_sf(as.polygons(ext(Gafanha_low_ortho)))\n",
    " sf::st_crs(map_extent_full)<-st_crs(Gafanha_low_ortho)\n",
    " \n",
    "\n",
    "  map_extent_zommed<- st_as_sf(as.polygons(ext(Gafanha_low_ortho_zoomed)))\n",
    " sf::st_crs(map_extent_zommed)<-st_crs(Gafanha_low_ortho_zoomed)\n",
    " \n",
    " \n",
    " rgb_Gafanha_low_zoomed<-ggplot()+\n",
    "   tidyterra::geom_spatraster_rgb(data = Gafanha_low_ortho_zoomed, \n",
    "                                  r = 3,\n",
    "                                  g = 2,\n",
    "                                  b = 1)+\n",
    "   theme_Bede_Map()+\n",
    "   geom_sf(data = zoom_extent, fill = \"transparent\", colour = \"black\",linetype = \"dashed\", linewidth = 1)+\n",
    "   coord_sf(crs = 4326)+\n",
    "     theme_void()\n",
    "\n",
    "\n",
    "\n",
    " scale_labels<-data.frame(x = c(-8.74368, -8.743565,-8.74342),\n",
    "                          y = c(rep(40.59792,3)),\n",
    "                          text = c(\"0\",\"10\",\"20 m\"))\n",
    " \n",
    " \n",
    " \n",
    "  rgb_Gafanha_low<- ggplot()+\n",
    "  scale_fill_manual(\"\", \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  geom_sf(data = background_GafHigh,\n",
    "          fill = \"white\",\n",
    "          show.legend = F,\n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"white\")+\n",
    "  geom_sf(data = background_GafHigh,\n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"grey30\")+\n",
    "   tidyterra::geom_spatraster_rgb(data = Gafanha_low_ortho,\n",
    "         r = 1,\n",
    "         g = 2,\n",
    "         b = 3,\n",
    "         maxcell = 4010533\n",
    "         )+\n",
    "    ggspatial::annotation_scale(location = \"tl\",\n",
    "                                text_cex=2)+\n",
    "      annotation_custom(ggplotGrob(p1),\n",
    "                        xmin = -8.7430,\n",
    "                        xmax = -8.742702,\n",
    "                        ymin = 40.5970515, \n",
    "                        ymax = 40.5974) +\n",
    "   theme_Bede_Map()+\n",
    "   geom_sf(data = zoom_extent, \n",
    "           fill = \"transparent\",\n",
    "           colour = \"black\",\n",
    "           linetype = \"dashed\", \n",
    "           linewidth = 1)+\n",
    "   coord_sf(crs = 4326)+\n",
    "  annotation_custom(ggplotGrob(rgb_Gafanha_low_zoomed),\n",
    "                    xmin = -8.7430,\n",
    "                    xmax=-8.7427,\n",
    "                    ymin = 40.5977) +\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.7432439,-8.743,-8.7432439,-8.743),\n",
    "                              y=c(40.5976275,40.5979,40.5975373,40.59771),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", linewidth = 1)+\n",
    "  theme(axis.title = element_blank(),\n",
    "        axis.text.y = element_text(size = 20),\n",
    "        axis.text.x = element_text(size = 20, angle = 20, hjust = 1))+\n",
    "    scale_x_continuous(limits = c(-8.7437, -8.74275), breaks = -8.743)+\n",
    "    scale_y_continuous(limits = c(40.59715, 40.5979), breaks = 40.5975)\n",
    "\n",
    " \n",
    " \n",
    "  \n",
    "  \n",
    "\n",
    "ggsave(\"Figures/Gaf_Low_RGB.png\",rgb_Gafanha_low , width = 1920*4, height = 1129*4, units = \"px\")\n",
    "\n",
    "zoom_extent<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"zoom_extent_map_Gafanha_comparison\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "####### Zoomed map ########\n",
    "\n",
    "\n",
    "Gafanha_low_pred_zoomed<-imglist %>% \n",
    "filter(site == \"Gafanha\",\n",
    "       alt == \"low\",\n",
    "       type == \"Pred\",\n",
    "       resampled == \"FALSE\",\n",
    "       folder == \"Input\",\n",
    "       new == \"TRUE\") %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Gafanha_low_pred_zoomed)<-\"layer\"\n",
    "\n",
    "Gafanha_low_pred_zoomed<-Gafanha_low_pred_zoomed %>% \n",
    "  crop(zoom_extent)\n",
    "Gafanha_low_pred_zoomed<-as.factor(Gafanha_low_pred_zoomed)\n",
    "\n",
    "Gafanha_low_pred_map_zoomed<-ggplot()+\n",
    "  tidyterra::geom_spatraster(data = Gafanha_low_pred_zoomed, mapping = aes(fill = layer))+\n",
    "  geom_sf(data = zoom_extent, fill = \"transparent\", colour = \"black\",linetype = \"dashed\", linewidth = 1)+\n",
    "  coord_sf(crs = 4326)+\n",
    "  scale_fill_manual(values=c(\"#70543e\", \"#b3ff1a\",\"#70543e\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", NA, \"#42c9bc\"), na.value = NA)+\n",
    "  theme_void()+\n",
    "  # coord_equal()+\n",
    "  theme(axis.text.x = element_blank(),\n",
    "      axis.text.y = element_blank(),\n",
    "      axis.title = element_blank())+\n",
    "  theme(legend.position = \"none\")\n",
    " # Gafanha_low_pred_map_zoomed\n",
    "\n",
    "####  Overall map\n",
    "Gafanha_low_pred<-imglist %>% \n",
    "filter(site == \"Gafanha\",\n",
    "       alt == \"low\",\n",
    "       type == \"Pred\",\n",
    "       # resampled == \"TRUE\",\n",
    "       new == \"TRUE\") %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Gafanha_low_pred)<-\"layer\"\n",
    "\n",
    "\n",
    "values(Gafanha_low_pred)[values(Gafanha_low_pred) == 0] = NA\n",
    "\n",
    "Gafanha_low_pred<-Gafanha_low_pred %>% \n",
    "  as.factor()\n",
    "\n",
    "####### Pred ######\n",
    "\n",
    "\n",
    "Gafanha_low_pred_map <-\n",
    "  ggplot()+\n",
    "  scale_fill_manual(\"\", \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  geom_sf(data = background_GafHigh,\n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"grey30\")+\n",
    "  # geom_sf(data = background_GafHigh,\n",
    "  #         fill = \"white\",\n",
    "  #         show.legend = F, \n",
    "  #         linewidth=0.05,\n",
    "  #         alpha=1,\n",
    "  #         colour=\"white\")+\n",
    "   ggnewscale::new_scale(\"fill\")+\n",
    "    tidyterra::geom_spatraster(data = Gafanha_low_pred, \n",
    "                               mapping = aes(fill = layer),\n",
    "                               maxcell = 500832*8)+\n",
    "  geom_sf(data = zoom_extent, \n",
    "          fill = \"transparent\", \n",
    "          colour = \"black\",\n",
    "          linetype = \"dashed\", \n",
    "          linewidth = 1)+\n",
    "  coord_sf(crs = 4326)+\n",
    "  annotation_custom(ggplotGrob(Gafanha_low_pred_map_zoomed), \n",
    "                    xmin = -8.7430,\n",
    "                    xmax=-8.7427, \n",
    "                    ymin = 40.5977) +\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.7432439,-8.743,-8.7432439,-8.743),\n",
    "                              y=c(40.5976275,40.5979,40.5975373,40.59771),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", linewidth = 1)+\n",
    "  scale_fill_manual(breaks = c(1,2,5,6,7,8,10),\n",
    "                    values=c(\"#70543e\", \"#b3ff1a\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", \"#42c9bc\",NA,NA,NA,NA),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Chlorophyceae\",\n",
    "                               \"Bacillariophyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\",\n",
    "                               \"Water\"))+\n",
    "  theme_Bede_Map()+\n",
    "  # coord_equal()+\n",
    "   theme(axis.text.x = element_text(size = 20, angle = 20,hjust = 1),\n",
    "         # axis.text.y = element_text(size = 20),\n",
    "         axis.text.y = element_blank(),\n",
    "         axis.title = element_blank(),\n",
    "         legend.background = element_blank(),\n",
    "         legend.position = c(0.98,0.3),\n",
    "         legend.text = element_text(size = 20),\n",
    "         legend.key.size = unit(1,\"cm\"))+\n",
    "    scale_x_continuous(limits = c(-8.7437, -8.74275), breaks = -8.743)+\n",
    "    scale_y_continuous(limits = c(40.59715, 40.5979), breaks = 40.5975)\n",
    "\n",
    "\n",
    "ggsave(\"Figures/Gaf_Low_Pred.png\",Gafanha_low_pred_map , width = 1920*4, height = 1129*4, units = \"px\")\n",
    "\n",
    "################# PLOT AND SAVING ###########\n",
    "\n",
    "plot_Gaf_Low<-rgb_Gafanha_low+\n",
    "  Gafanha_low_pred_map\n",
    " \n",
    "\n",
    "\n",
    "\n",
    "ggsave(\"Figures/Maps Pred/FigX-Gaf_Low_Pred.png\",plot_Gaf_Low , width = 1920*4, height = 1009*4, units = \"px\")"
   ],
   "id": "44b852b0-b011-4cb0-9607-f23eb12fc650"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Maps Pred/FigX-Gaf_Low_Pred.png\")"
   ],
   "id": "cell-fig-GafLow"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imglist<-\n",
    "  data.frame(\"Data\",\n",
    "             path = list.files(pattern = \".tif\", recursive = T, full.names = T)) %>% \n",
    "  mutate(folder = case_when(str_detect(path, \"Data\") ~ \"Input\",\n",
    "                            TRUE ~ \"Output\"),\n",
    "         alt = case_when(str_detect(path, \"high\") ~ \"high\",\n",
    "                         str_detect(path, \"High\") ~ \"high\",\n",
    "                         TRUE ~ \"low\"),\n",
    "         site = case_when(str_detect(path, \"Gafanha\") ~ \"Gafanha\",\n",
    "                          str_detect(path,\"AveiroHigh\")~\"Gafanha\",\n",
    "                          TRUE ~ \"Duer\"),\n",
    "         type = case_when(str_detect(path, \"pred\") ~ \"Pred\",\n",
    "                          str_detect(path, \"NDVI\") ~ \"NDVI\",\n",
    "                          TRUE ~ \"ortho\"),\n",
    "         resampled = case_when(str_detect(path, \"resampled\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         new = case_when(str_detect(path, \"no_Xantho\") ~ \"TRUE\",\n",
    "                         str_detect(path, \"NewPreds\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         col = case_when(str_detect(path, \"colored\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         correction = case_when(str_detect(path, \"corrected_per_bands\") ~ \"Bands\",\n",
    "                          str_detect(path, \"corrected.tif\") ~ \"lm_all\",\n",
    "                          TRUE ~ \"NA\")) %>% \n",
    "  dplyr::filter(alt != \"NA\",\n",
    "                !str_detect(path,\".aux.xml\"))\n",
    "\n",
    "########## ZOOM #############\n",
    "\n",
    "zoom_extent<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"zoom\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "mask_low<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"mask_Gafanha_low\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "Gafanha_low_ortho_zoomed<-imglist %>% \n",
    "  filter(site == \"Gafanha\",\n",
    "         alt == \"low\",\n",
    "         type == \"ortho\",\n",
    "         resampled == \"TRUE\",\n",
    "         col == \"TRUE\") %>%\n",
    "  pull(path) %>% \n",
    "  rast() %>% \n",
    "  crop(zoom_extent)\n",
    "\n",
    "names(Gafanha_low_ortho_zoomed)<-c(\"B1\", \"B3\", \"B5\",\"alpha\")\n",
    "\n",
    " Gafanha_low_pred_zoomed<-imglist %>% \n",
    "  filter(site == \"Gafanha\",\n",
    "         alt == \"low\",\n",
    "         type == \"Pred\",\n",
    "         # resampled == \"TRUE\",\n",
    "         new == \"TRUE\") %>% \n",
    "  pull(path) %>% \n",
    "  rast() %>% \n",
    "  crop(zoom_extent) %>% \n",
    "  mask(mask_low, inverse = T) %>% \n",
    "   as.factor()\n",
    "   \n",
    "names(Gafanha_low_pred_zoomed)<-\"layer\"\n",
    "\n",
    "zoomed_low_map<-ggplot()+\n",
    "  tidyterra::geom_spatraster_rgb(data = Gafanha_low_ortho_zoomed,\n",
    "                                 r = 1,\n",
    "                                 g = 2,\n",
    "                                 b = 3)+\n",
    "  tidyterra::geom_spatraster(data = Gafanha_low_pred_zoomed, mapping = aes(fill = layer), show.legend = F)+\n",
    "  \n",
    "  scale_fill_manual(breaks = c(1,2,5,6,7,8,10),\n",
    "                    values=c(\"#70543e\", \"#b3ff1a\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", \"#42c9bc\",NA,NA), \n",
    "                    na.value = NA,\n",
    "                    name = \"\", \n",
    "                    labels = c(\"Sediment\", \n",
    "                               \"Chlorophyceae\", \n",
    "                               \"Bacillariophyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\",\n",
    "                               \"Water\"))+\n",
    "  geom_sf(data = zoom_extent, fill = \"transparent\", colour = \"black\",linetype = \"dashed\", linewidth = 2)+\n",
    "  theme(legend.position = \"none\")+\n",
    "  theme_void()\n",
    "\n",
    "\n",
    "\n",
    "#############\n",
    "\n",
    " Gafanha_low_ortho<-imglist %>% \n",
    "  filter(site == \"Gafanha\",\n",
    "         alt == \"low\",\n",
    "         type == \"ortho\",\n",
    "         resampled == \"TRUE\",\n",
    "         col == \"TRUE\") %>%\n",
    "  pull(path) %>% \n",
    "  rast() \n",
    " names(Gafanha_low_ortho)<-c(\"B1\", \"B3\", \"B5\",\"alpha\")\n",
    "values(Gafanha_low_ortho)[values(Gafanha_low_ortho$alpha) == 0] = NA\n",
    "\n",
    " \n",
    " Gafanha_low_pred<-imglist %>% \n",
    "  filter(site == \"Gafanha\",\n",
    "         alt == \"low\",\n",
    "         type == \"Pred\",\n",
    "         # resampled == \"TRUE\",\n",
    "         new == \"TRUE\") %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Gafanha_low_pred)<-\"layer\"\n",
    "\n",
    "Gafanha_low_pred_masked<-Gafanha_low_pred %>% \n",
    "  # crop(mask_low) %>% \n",
    "  mask(mask_low, inverse = T)\n",
    "\n",
    "values(Gafanha_low_pred_masked)[values(Gafanha_low_pred_masked) == 0] = NA\n",
    "\n",
    "Gafanha_low_pred_masked<-Gafanha_low_pred_masked %>% \n",
    "  as.factor()\n",
    "\n",
    "rgb_Gafanha_low <- ggplot()+\n",
    "  tidyterra::geom_spatraster_rgb(data = Gafanha_low_ortho,\n",
    "                                 r = 1,\n",
    "                                 g = 2,\n",
    "                                 b = 3)+\n",
    "  tidyterra::geom_spatraster(data = Gafanha_low_pred_masked, mapping = aes(fill = layer))+\n",
    "  scale_fill_manual(breaks = c(1,2,5,6,7,8,10),\n",
    "                    values=c(\"#70543e\", \"#b3ff1a\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", \"#42c9bc\",NA,NA), \n",
    "                    na.value = NA,\n",
    "                    name = \"\", \n",
    "                    labels = c(\"Sediment\", \n",
    "                               \"Chlorophyceae\", \n",
    "                               \"Bacillariophyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\",\n",
    "                               \"Water\"))+\n",
    "  ggspatial::annotation_scale(location = \"bl\",text_cex=2)+\n",
    "  geom_sf(data = zoom_extent, fill = \"transparent\", colour = \"black\",linetype = \"dashed\", linewidth = 1)+\n",
    "\n",
    "  geom_segment(aes(x = -8.7434247325, y=40.5977530006, xend = -8.7430558162, yend = 40.5972152706), linetype =\"dashed\" ,linewidth=2)+\n",
    "  annotation_custom(ggplotGrob(zoomed_low_map), xmin = -8.743725, xmax = -8.7434057328, ymin = 40.5977200450, ymax = 40.59798) +\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.74336194,-8.74371,-8.743243794,-8.7434057328), y=c(40.59762751,40.5977400850,40.59762725,40.5977200450),grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", linewidth = 1)+\n",
    "  theme_Bede_Map()+\n",
    "  coord_sf(crs = 4326)+\n",
    "  theme(axis.title = element_blank(),\n",
    "        axis.text.y = element_text(size = 20),\n",
    "        axis.text.x = element_text(size = 20, angle = 20,hjust = 1),\n",
    "        legend.position=c(0.3, 0.3))+\n",
    "  # ggtitle(\"a \")+\n",
    "  scale_x_continuous(breaks=seq(-8.7437, -8.7430, 0.0002))\n",
    "rgb_Gafanha_low\n",
    "\n",
    "ggsave(\"Figures/Gaf_Low_RGB_Pred_V2.png\",rgb_Gafanha_low , width = 890*4, height = 1120*4, units = \"px\")"
   ],
   "id": "243e6936-2b49-48ad-85e2-b25f3849a4a1"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The high-altitude flight over Gafanha covered nearly 1 km² in total (<a href=\"#fig-GafHigh\" class=\"quarto-xref\">Figure 5</a>). A channel delineating a small island was masked in the prediction map due to sunglint and misclassification caused by the turbid shallow water. Most of the intertidal area has been classified as Magnoliopsida by the model, even though there is discoloration of seagrass blades. Only a few pixels have been classified as Chlorophyceae at this scale. Furthermore, the area that was classified as Bacillariophyceae in the low-altitude flight remains mostly classified as such in the high-altitude flight, though some pixels were classified as Magnoliopsida. Patches of Rhodophyceae were correctly classified. In the nothern part of the scene, near the land limit, patches of *Spartina sp.* were misclassified, either as Magnoliopsida or as Phaeophyceae."
   ],
   "id": "6ecec472-ef56-4948-9cb8-8323a8dcb29a"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imglist<-\n",
    "  data.frame(path = list.files(pattern = \".tif\", recursive = T, full.names = T)) %>% \n",
    "  mutate(folder = case_when(str_detect(path, \"Data\") ~ \"Input\",\n",
    "                            TRUE ~ \"Output\"),\n",
    "         alt = case_when(str_detect(path, \"high\") ~ \"high\",\n",
    "                         str_detect(path, \"High\") ~ \"high\",\n",
    "                         TRUE ~ \"low\"),\n",
    "         site = case_when(str_detect(path, \"Gafanha\") ~ \"Gafanha\",\n",
    "                          str_detect(path, \"AveiroHigh\") ~ \"Gafanha\",\n",
    "                          TRUE ~ \"Duer\"),\n",
    "         type = case_when(str_detect(path, \"pred\") ~ \"Pred\",\n",
    "                          str_detect(path, \"Preds\") ~ \"Pred\",\n",
    "                          str_detect(path, \"NDVI\") ~ \"NDVI\",\n",
    "                          TRUE ~ \"ortho\"),\n",
    "         resampled = case_when(str_detect(path, \"resampled\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         new = case_when(str_detect(path, \"no_Xantho\") ~ \"TRUE\",\n",
    "                         str_detect(path, \"NewPreds\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         col = case_when(str_detect(path, \"colored\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         correction = case_when(str_detect(path, \"corrected_per_bands\") ~ \"Bands\",\n",
    "                          str_detect(path, \"corrected.tif\") ~ \"lm_all\",\n",
    "                          TRUE ~ \"NA\")) %>% \n",
    "  dplyr::filter(alt != \"NA\",\n",
    "                !str_detect(path,\".aux.xml\"))\n",
    "\n",
    "#### Preview ####\n",
    "\n",
    "filelist<-list.files(\"Data/shp/\", pattern = \".shp\", recursive = T,full.names = T)\n",
    "mask_aveiro <- filelist[str_detect(filelist,\"mask_land_intertidal_Aveiro\")] %>% \n",
    "  read_sf()\n",
    "\n",
    "Flight_aveiro<-data.frame(\n",
    "  Name=c(\n",
    "  \"Gafanha\"\n",
    "  ) ,\n",
    "  Description = c(\"12m and 120m Flights\"),\n",
    "  Long=c(-8.742485) ,\n",
    "  Lat=c(40.597599) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(crs(mask_aveiro))  %>% \n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)\n",
    "\n",
    "p1_background<- read_sf(\"Data/shp/Aveiro/Background_p1_aveiro.shp\")\n",
    "\n",
    "p1<-ggplot() +\n",
    "  scale_fill_manual(\"\", labels = c(\"Intertidal area\", \"Land area\"),\n",
    "                    values = c(\"#7DC27D\", \"#BDBDBD\"))+\n",
    "  geom_sf(data = p1_background, fill = \"white\",show.legend = F, linewidth=0.05,alpha=0.93,colour=\"transparent\")+\n",
    "  geom_sf(data = mask_aveiro, mapping = aes(fill = Type),show.legend = F, linewidth=0.05,alpha=0.93,colour=\"grey30\")+\n",
    "      coord_sf(xlim=c(519749,535233) ,\n",
    "          ylim=c(4493310,4508794) )+\n",
    "  ggforce::geom_mark_ellipse(data=Flight_aveiro,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = Name,\n",
    "                     group = Name),\n",
    "                     # description=Description),\n",
    "                 size=0.3,\n",
    "                 fill = \"goldenrod\",\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 show.legend=F,\n",
    "                 label.fontsize = 20,\n",
    "                 label.hjust = 0.5,\n",
    "                 con.size = 2,\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") ,\n",
    "  label.fill = \"grey90\",\n",
    "  label.buffer = unit(5, \"mm\")) +\n",
    "    # theme_void()+\n",
    "  theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  # geom_label(aes(x = 521000, y = 4507794, label = \"C\"), size = 20)+\n",
    "  scale_x_continuous(breaks = seq(-8.75, -8.6, by = 0.1))+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\") ,\n",
    "        legend.position = \"none\",\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20))+\n",
    "  theme_void()\n",
    "\n",
    "#####\n",
    "\n",
    "land_mask_GafHigh<-\"Data/shp/Aveiro\" %>% \n",
    "  list.files(\".shp\", full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path, \"mask_Gafanha_high1\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "background_GafHigh<-\"Data/shp/Aveiro\" %>% \n",
    "  list.files(\".shp\", full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path, \"background_Gafanha_high\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "GafHigh_ortho<-imglist %>% \n",
    "  dplyr::filter(site == \"Gafanha\",\n",
    "                type == \"ortho\",\n",
    "                alt==\"high\",\n",
    "                col == \"TRUE\") %>%\n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(GafHigh_ortho)<-c(\"B1\",\"B3\", \"B5\", \"alpha\")\n",
    "\n",
    "# values(GafHigh_ortho)[values(GafHigh_ortho$B1) == 0] = NA\n",
    "\n",
    "GafHigh_ortho_masked<-GafHigh_ortho %>% \n",
    "  mask(land_mask_GafHigh, inverse = F) %>% \n",
    "  crop(land_mask_GafHigh)\n",
    "\n",
    "GafHigh_ortho_masked<-na.omit(GafHigh_ortho_masked, field=NA, geom=FALSE)\n",
    "\n",
    "\n",
    "###### RGB ########\n",
    "zoom_gaf_high_sf<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"zoom_extent_map_Gafanha_High_comparison\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "Gaflow_extent<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"GafLow_extent.shp\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "rgb_Gafanha_High_zoomed<-GafHigh_ortho_masked %>% \n",
    "  crop(zoom_gaf_high_sf)\n",
    "\n",
    "rgb_Gafanha_High_zoomed<-\n",
    "  ggplot()+\n",
    "  tidyterra::geom_spatraster_rgb(data = rgb_Gafanha_High_zoomed,\n",
    "         r = 1,\n",
    "         g = 2,\n",
    "         b = 3)+\n",
    "     geom_sf(data = zoom_gaf_high_sf, fill = \"transparent\", colour = \"black\",linetype = \"dashed\", linewidth = 1)+\n",
    "  theme_void()\n",
    "\n",
    "  \n",
    "  \n",
    "\n",
    "rgb_Gafanha_High <- ggplot()+\n",
    "  scale_fill_manual(\"\",  \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  geom_sf(data = background_GafHigh,\n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.93,\n",
    "          colour=\"grey30\")+\n",
    "  geom_sf(data = background_GafHigh,\n",
    "          fill = \"white\",\n",
    "          show.legend = F,\n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"white\")+\n",
    "   tidyterra::geom_spatraster_rgb(data = GafHigh_ortho_masked,\n",
    "         r = 1,\n",
    "         g = 2,\n",
    "         b = 3,\n",
    "         maxcell = 5007120)+\n",
    "    ggspatial::annotation_scale(location = \"bl\",\n",
    "                                text_cex=2)+\n",
    "    annotation_custom(ggplotGrob(p1), \n",
    "                      xmin = -8.744, \n",
    "                      xmax = -8.7412,\n",
    "                      ymin = 40.5923,\n",
    "                      ymax = 40.595) +\n",
    "    geom_sf(data = zoom_gaf_high_sf,\n",
    "            fill = \"transparent\", \n",
    "            colour = \"black\",\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 1)+\n",
    "    geom_sf(data = Gaflow_extent,\n",
    "            fill = \"transparent\", \n",
    "            colour = \"red\",\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 2)+\n",
    "    geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.74363778,-8.74757,-8.74351982,-8.7453), \n",
    "                              y=c(40.5983781677,40.5995,40.598377893,40.5995),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 1)+\n",
    "  annotation_custom(ggplotGrob(rgb_Gafanha_High_zoomed),\n",
    "                    xmin = -8.7479, \n",
    "                    xmax = -8.7452, \n",
    "                    ymin = 40.5979,\n",
    "                    ymax = 40.603) +\n",
    "   theme_Bede_Map()+\n",
    "   coord_sf(crs = 4326)+\n",
    "   # xlim(c(-8.748, -8.741))+\n",
    "  # ylim(c(40.593,40.601))+\n",
    "  theme(axis.title = element_blank(),\n",
    "         axis.text.x = element_text(size = 20, angle = 20, hjust = 1),\n",
    "         axis.text.y = element_text(size = 20))+\n",
    "  scale_x_continuous(limits = c(-8.747576, -8.741465),breaks =  -8.745)+\n",
    "  scale_y_continuous(limits = c(40.59297,40.60109),breaks =  40.595)\n",
    "# rgb_Gafanha_High\n",
    "\n",
    "ggsave(\"Figures/Gaf_High_RGB.png\",rgb_Gafanha_High , width = 712*4, height = 1120*4, units = \"px\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "######### Pred #######\n",
    "Gafanha_high_pred<-imglist %>% \n",
    "filter(site == \"Gafanha\",\n",
    "       alt == \"high\",\n",
    "       type == \"Pred\",\n",
    "       new == \"TRUE\",\n",
    "       !str_detect(path, \"Probs\")) %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Gafanha_high_pred)<-\"layer\"\n",
    "\n",
    "GafHigh_pred_masked<-Gafanha_high_pred %>% \n",
    "  mask(land_mask_GafHigh, inverse = F) %>% \n",
    "  crop(land_mask_GafHigh)\n",
    "\n",
    "GafHigh_pred_masked<-na.omit(GafHigh_pred_masked, field=NA, geom=FALSE) \n",
    "\n",
    "values(GafHigh_pred_masked)[values(GafHigh_pred_masked) == 0] = NA\n",
    "values(GafHigh_pred_masked)[values(GafHigh_pred_masked) == 3] = 1\n",
    "\n",
    "GafHigh_pred_masked<-GafHigh_pred_masked %>% \n",
    "  as.factor()\n",
    "\n",
    "Gafanha_High_zoomed_pred<-GafHigh_pred_masked %>% \n",
    "  crop(zoom_gaf_high_sf)\n",
    "\n",
    "Gafanha_High_zoomed_pred<-\n",
    "  ggplot()+\n",
    "   tidyterra::geom_spatraster(data = Gafanha_High_zoomed_pred, mapping = aes(fill = layer),\n",
    "                            show.legend = F,\n",
    "                             maxcell = 5007120)+\n",
    "  scale_fill_manual(breaks = c(1,2,5,6,7,8,10),\n",
    "                    values=c(\"#70543e\", \"#b3ff1a\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", \"#42c9bc\",NA,NA),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Chlorophyceae\",\n",
    "                               \"Bacillariophyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\",\n",
    "                               \"Water\"))+\n",
    "     geom_sf(data = zoom_gaf_high_sf, fill = \"transparent\", colour = \"black\",linetype = \"dashed\", linewidth = 1)+\n",
    "  theme_void()\n",
    "\n",
    "Gafanha_high_pred_map<-ggplot()+\n",
    "  geom_sf(data = background_GafHigh, \n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F,\n",
    "          linewidth=0.05,\n",
    "          alpha=0.93,\n",
    "          colour=\"grey30\")+\n",
    "  geom_sf(data = background_GafHigh,\n",
    "          fill = \"white\",\n",
    "          show.legend = F,\n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"white\")+\n",
    "    scale_fill_manual(\"\", \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  ggnewscale::new_scale(\"fill\")+\n",
    "  tidyterra::geom_spatraster(data = GafHigh_pred_masked, \n",
    "                             # show.legend = F,\n",
    "                             mapping = aes(fill = layer),\n",
    "                             maxcell = 5010420)+\n",
    "  scale_fill_manual(breaks = c(1,2,5,6,7,8,10),\n",
    "                    values=c(\"#70543e\", \"#b3ff1a\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", \"#42c9bc\",NA,NA),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Chlorophyceae\",\n",
    "                               \"Bacillariophyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\",\n",
    "                               \"Water\"))+\n",
    "       geom_sf(data = zoom_gaf_high_sf, \n",
    "               fill = \"transparent\", \n",
    "               colour = \"black\",\n",
    "               linetype = \"dashed\", \n",
    "               linewidth = 1)+\n",
    "    geom_sf(data = Gaflow_extent,\n",
    "            fill = \"transparent\", \n",
    "            colour = \"red\",\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 2)+\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.74363778,-8.74757,-8.74351982,-8.7453), \n",
    "                              y=c(40.5983781677,40.5995,40.598377893,40.5995),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", linewidth = 1)+\n",
    "  annotation_custom(ggplotGrob(Gafanha_High_zoomed_pred), \n",
    "                    xmin = -8.7479, \n",
    "                    xmax = -8.7452, \n",
    "                    ymin = 40.5979,\n",
    "                    ymax = 40.603) +\n",
    "  coord_sf(crs = 4326)+\n",
    "  theme_Bede_Map()+\n",
    "   theme(axis.text.x = element_text(size = 20, angle = 20,hjust = 1),\n",
    "         axis.title = element_blank(),\n",
    "        # axis.text.y = element_text(size = 20),\n",
    "        axis.text.y = element_blank(),\n",
    "         legend.background = element_blank(),\n",
    "         legend.position = c(0.96,0.23),\n",
    "         legend.text = element_text(size = 23),\n",
    "         legend.key.size = unit(1.2,\"cm\"))+\n",
    "  scale_x_continuous(limits = c(-8.747576, -8.741465),breaks =  -8.745)+\n",
    "  scale_y_continuous(limits = c(40.59297,40.60109),breaks =  40.595)\n",
    "# Gafanha_high_pred_map\n",
    "\n",
    "ggsave(\"Figures/Gaf_High_Pred.png\",Gafanha_high_pred_map , width = 712*4, height = 1120*4, units = \"px\")\n",
    "\n",
    "################# PLOT AND SAVING ###########\n",
    "plot_GafHigh<-rgb_Gafanha_High+\n",
    "  Gafanha_high_pred_map\n",
    "\n",
    "ggsave(\"Figures/Maps Pred/FigX-Gaf_High_Pred.png\",plot_GafHigh , width = 2105*4, height = 1363*4, units = \"px\")\n",
    "\n",
    "\n",
    "#########"
   ],
   "id": "8554455d-8e51-4d99-b308-1ef36133b15c"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Maps Pred/FigX-Gaf_High_Pred.png\")"
   ],
   "id": "cell-fig-GafHigh"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The high altitude flight acquired over the inner lagoon of the Ria de Aveiro is the largest of all flights, covering almost 1.5 km² (<a href=\"#fig-Boat\" class=\"quarto-xref\">Figure 6</a>). On this site, only seagrass and red algae where seen on the field. The classification provided consistent results, with a patchy Magnoliopsida meadow mixed with Rhodophyceae on the eastern part of the scene. As shown in the zoom (<a href=\"#fig-Boat\" class=\"quarto-xref\">Figure 6</a>), the edges of the meadow can be colonised by Chlorophyceae."
   ],
   "id": "169b51bb-4c69-4db9-a75f-ccac609d1b1a"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imglist<-\n",
    "  data.frame(path = list.files(pattern = \".tif\", recursive = T, full.names = T)) %>% \n",
    "  mutate(folder = case_when(str_detect(path, \"Data\") ~ \"Input\",\n",
    "                            TRUE ~ \"Output\"),\n",
    "         alt = case_when(str_detect(path, \"high\") ~ \"high\",\n",
    "                         str_detect(path, \"High\") ~ \"high\",\n",
    "                         TRUE ~ \"low\"),\n",
    "         site = case_when(str_detect(path, \"Gafanha\") ~ \"Gafanha\",\n",
    "                          str_detect(path, \"AveiroHigh\") ~ \"Gafanha\",\n",
    "                          TRUE ~ \"Duer\"),\n",
    "         type = case_when(str_detect(path, \"pred\") ~ \"Pred\",\n",
    "                          str_detect(path, \"Preds\") ~ \"Pred\",\n",
    "                          str_detect(path, \"NDVI\") ~ \"NDVI\",\n",
    "                          TRUE ~ \"ortho\"),\n",
    "         resampled = case_when(str_detect(path, \"resampled\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         new = case_when(str_detect(path, \"no_Xantho\") ~ \"TRUE\",\n",
    "                         str_detect(path, \"NewPreds\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         col = case_when(str_detect(path, \"colored\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         correction = case_when(str_detect(path, \"corrected_per_bands\") ~ \"Bands\",\n",
    "                          str_detect(path, \"corrected.tif\") ~ \"lm_all\",\n",
    "                          TRUE ~ \"NA\")) %>% \n",
    "  dplyr::filter(alt != \"NA\",\n",
    "                !str_detect(path,\".aux.xml\"))\n",
    "########### PREVIEW #############\n",
    "filelist<-list.files(\"Data/shp/\", pattern = \".shp\", recursive = T,full.names = T)\n",
    "mask_aveiro <- filelist[str_detect(filelist,\"mask_land_intertidal_Aveiro\")] %>% \n",
    "  read_sf()\n",
    "\n",
    "Flight_aveiro<-data.frame(\n",
    "  Name=c(\n",
    "  \"Marinha Lanzarote\"\n",
    "  ) ,\n",
    "  Description = c(\n",
    "                  \"120m Flight\"),\n",
    "  Long=c( \n",
    "         -8.690677) ,\n",
    "  Lat=c(\n",
    "        40.665120) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(crs(mask_aveiro))  %>% \n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)\n",
    "\n",
    "\n",
    "p1_background<- read_sf(\"Data/shp/Aveiro/Background_p1_aveiro.shp\")\n",
    "\n",
    "p1<-ggplot() +\n",
    "  scale_fill_manual(\"\", labels = c(\"Intertidal area\", \"Land area\"),\n",
    "                    values = c(\"#7DC27D\", \"#BDBDBD\"))+\n",
    "  geom_sf(data = p1_background, fill = \"white\",show.legend = F, linewidth=0.05,alpha=0.93,colour=\"transparent\")+\n",
    "  geom_sf(data = mask_aveiro, mapping = aes(fill = Type),linewidth=0.05,alpha=0.93,show.legend = F,colour=\"grey30\")+\n",
    "      coord_sf(xlim=c(519749,535233) ,\n",
    "          ylim=c(4493310,4508794) )+\n",
    "  ggforce::geom_mark_ellipse(data=Flight_aveiro,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = Name,\n",
    "                     group = Name),\n",
    "                     # description=Description),\n",
    "                 size=0.3,\n",
    "                 fill = \"goldenrod\",\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 show.legend=F,\n",
    "                 label.fontsize = 10,\n",
    "                 label.hjust = 0.5,\n",
    "                 con.size = 2,\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") ,\n",
    "  label.fill = \"grey90\",\n",
    "  label.buffer = unit(5, \"mm\")) +\n",
    "    # theme_void()+\n",
    "  theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  # geom_label(aes(x = 521000, y = 4507794, label = \"C\"), size = 20)+\n",
    "  scale_x_continuous(breaks = seq(-8.75, -8.6, by = 0.1))+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\") ,\n",
    "        legend.position = \"none\",\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20))+\n",
    "  theme_void()\n",
    "##### ZOOM and masking #####\n",
    "\n",
    "Boat_rgb<-imglist %>% \n",
    "  filter(str_detect(path,\"Boat_high_RGB_merged.tif\")) %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Boat_rgb)<-c(\"B1\",\"B3\",\"B5\")\n",
    "values(Boat_rgb)[values(Boat_rgb) == 0] = NA\n",
    "\n",
    "mask_boat<-\"Data/shp/Aveiro\" %>% \n",
    "  list.files(\".shp\", full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path, \"boat\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "background_Boat<-\"Data/shp/Aveiro\" %>% \n",
    "  list.files(\".shp\", full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path, \"background_Boat\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "zoom_boat_high_sf<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"zoom_extent_map_Boat_comparison\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "zoom_boat_high_sf_2<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"zoom_extent_map_Boat_2_comparison\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "rgb_Boat_zoomed<-Boat_rgb %>% \n",
    "  crop(zoom_boat_high_sf)\n",
    "\n",
    "rgb_Boat_zoomed_2<-Boat_rgb %>% \n",
    "  crop(zoom_boat_high_sf_2)\n",
    "\n",
    "Boat_rgb_mask<-Boat_rgb %>% \n",
    "  mask(mask_boat, inverse = T)\n",
    "\n",
    " \n",
    "rgb_Boat_zoomed_map<- ggplot()+\n",
    "  geom_spatraster_rgb(data = rgb_Boat_zoomed,\n",
    "                      r = 3,\n",
    "                      g = 2,\n",
    "                      b = 1,\n",
    "                      max_col_value = 4000)+\n",
    "  geom_sf(data = zoom_boat_high_sf, fill = \"transparent\", linetype = \"dashed\",colour = \"black\", linewidth = 1)+\n",
    "\n",
    "  theme_void()\n",
    "\n",
    "rgb_Boat_zoomed_map_2\n",
    "ggplot()+\n",
    "  geom_spatraster_rgb(data = rgb_Boat_zoomed_2,\n",
    "                      r = 3,\n",
    "                      g = 2,\n",
    "                      b = 1,\n",
    "                      interpolate = F,\n",
    "                      max_col_value = 4000)+\n",
    "  geom_sf(data = zoom_boat_high_sf_2, fill = \"transparent\", linetype = \"dashed\",colour = \"black\", linewidth = 1)+\n",
    "\n",
    "  theme_void()\n",
    "\n",
    "####### RGB ############\n",
    "\n",
    "RGB_Boat_map<-\n",
    "ggplot()+\n",
    "  scale_fill_manual(\"\",  \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  geom_sf(data = background_Boat,\n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.93,\n",
    "          colour=\"transparent\")+\n",
    "  geom_sf(data = background_Boat,\n",
    "          fill = \"white\",\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"transparent\")+\n",
    "  tidyterra::geom_spatraster_rgb(data = Boat_rgb_mask,\n",
    "                      r = 3,\n",
    "                      g = 2,\n",
    "                      b = 1,\n",
    "                      # maxcell = 5000,\n",
    "                      maxcell = 500832*8,\n",
    "                      max_col_value = 5000)+\n",
    "  geom_sf(data = zoom_boat_high_sf,\n",
    "          fill = \"transparent\",\n",
    "          colour = \"black\", \n",
    "          linewidth = 1)+\n",
    "  geom_sf(data = zoom_boat_high_sf_2,\n",
    "          fill = \"transparent\",\n",
    "          colour = \"black\", \n",
    "          linewidth = 1)+\n",
    "  annotation_custom(ggplotGrob(rgb_Boat_zoomed_map),\n",
    "                    xmin = -8.6975,\n",
    "                    xmax = -8.69,\n",
    "                    ymin = 40.6666,\n",
    "                    ymax = 40.6711) +\n",
    "  annotation_custom(ggplotGrob(rgb_Boat_zoomed_map_2),\n",
    "                    xmin = -8.6975+0.007,\n",
    "                    xmax = -8.69+0.007,\n",
    "                    ymin = 40.6666,\n",
    "                    ymax = 40.6711) +\n",
    "  annotation_custom(ggplotGrob(p1),\n",
    "                    xmin = -8.6795,\n",
    "                    xmax = -8.6736,\n",
    "                    ymin = 40.6585,\n",
    "                    ymax = 40.6625) +\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.69452127,-8.6965,-8.69440039,-8.69105), \n",
    "                              y=c(40.66244282,40.66673,40.66244282,40.66673),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 1)+\n",
    "    geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.6819877,-8.69645+0.007,-8.6818696,-8.6911+0.007), \n",
    "                              y=c(40.6652874,40.66675,40.6652874,40.66673),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 1)+\n",
    "  ggspatial::annotation_scale(location = \"bl\",\n",
    "                              text_cex=2)+\n",
    "  scale_x_continuous(limits = c(-8.6965,-8.675),\n",
    "                     breaks =  -8.685)+\n",
    "  scale_y_continuous(limits = c(40.659,40.6704),\n",
    "                     breaks =  40.664)+\n",
    "  theme_Bede_Map()+\n",
    "  theme(axis.title = element_blank(),\n",
    "        # axis.text.x = element_text(size = 20, angle = 20 , hjust = 1),\n",
    "        axis.text.x = element_blank(),\n",
    "        axis.text.y = element_text(size = 20))+\n",
    "   coord_sf(crs = 4326)\n",
    "\n",
    "\n",
    "####### Pred ######## \n",
    "Boat1_pred<-imglist %>% \n",
    "  filter(str_detect(path,\"AveiroBoat1_highPrediction_nnNewStanRawMetrics_Preds_NoXantho_7.tif\")) %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Boat1_pred)<-\"layer\"\n",
    "\n",
    "Boat2_pred<-imglist %>% \n",
    "  filter(str_detect(path,\"AveiroBoat2_highPrediction_nnNewStanRawMetrics_Preds_NoXantho_7.tif\")) %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Boat2_pred)<-\"layer\"\n",
    "\n",
    "values(Boat1_pred)[values(Boat1_pred) == 0] = NA\n",
    "values(Boat2_pred)[values(Boat2_pred) == 0] = NA\n",
    "\n",
    "\n",
    "B1_overextent<-ext(Boat1_pred)\n",
    "B1_overextent[2]<-ext(Boat2_pred)[2]\n",
    "B1_overextent[4]<-ext(Boat2_pred)[4]\n",
    "B1_overextent <- rast(B1_overextent)\n",
    "res(B1_overextent)<-res(Boat1_pred)\n",
    "# terra::writeRaster(B1$B1,\"B1.tif\")\n",
    "# terra::writeRaster(B2$B2,\"B2.tif\")\n",
    "\n",
    "B2_resampled <- terra::resample(Boat2_pred,B1_overextent, method = \"near\",threads = T)\n",
    "# terra::writeRaster(B2_resampled$B1,\"B2_resampled.tif\", overwrite = T)\n",
    "\n",
    "Boat_pred<-merge(Boat1_pred,B2_resampled)\n",
    "\n",
    "Boat_pred_aggregate<-terra::aggregate(Boat_pred,fact = 6, fun =\"modal\") \n",
    "\n",
    "masked_pred<-Boat_pred_aggregate %>% \n",
    "  mask(mask_boat,inverse=TRUE)%>% \n",
    "  as.factor()\n",
    "\n",
    "Boat_pred_zoom<-Boat1_pred %>% \n",
    "  crop(zoom_boat_high_sf) %>% \n",
    "  as.factor()\n",
    "\n",
    "Boat_pred_zoom_2<-Boat2_pred %>% \n",
    "  crop(zoom_boat_high_sf_2) %>% \n",
    "  as.factor()\n",
    "\n",
    "Boat_Pred_zoomed_map<- ggplot()+\n",
    "  tidyterra::geom_spatraster(data = Boat_pred_zoom, \n",
    "                             mapping = aes(fill = layer), \n",
    "                             show.legend = F,\n",
    "                             maxcell = 5010420)+\n",
    "  scale_fill_manual(values=c(\"#70543e\",\"#b3ff1a\",\"#389318\",\"#873e23\"),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Chlorphyceae\",\n",
    "                               \"Magnoliopsida\"))+\n",
    "  geom_sf(data = zoom_boat_high_sf, \n",
    "          fill = \"transparent\", \n",
    "          linetype = \"dashed\",\n",
    "          colour = \"black\", \n",
    "          linewidth = 1)+\n",
    "  theme_void()\n",
    "\n",
    "Boat_Pred_zoomed_map_2<- ggplot()+\n",
    "  tidyterra::geom_spatraster(data = Boat_pred_zoom_2, \n",
    "                             mapping = aes(fill = layer), \n",
    "                             show.legend = F,\n",
    "                             maxcell = 5010420)+\n",
    "  scale_fill_manual(values=c(\"#70543e\",\"#389318\",\"#873e23\",\"#b3002d\"),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\"))+\n",
    "  geom_sf(data = zoom_boat_high_sf_2, \n",
    "          fill = \"transparent\", \n",
    "          linetype = \"dashed\",\n",
    "          colour = \"black\", \n",
    "          linewidth = 1)+\n",
    "  theme_void()\n",
    "\n",
    "\n",
    "Boat_pred_map<-ggplot()+\n",
    "  geom_sf(data = background_Boat, \n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.93,\n",
    "          colour=\"transparent\")+\n",
    "  geom_sf(data = background_Boat,\n",
    "          fill = \"white\",\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"transparent\")+\n",
    "    scale_fill_manual(\"\", \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  ggnewscale::new_scale(\"fill\")+\n",
    "  tidyterra::geom_spatraster(data = masked_pred, \n",
    "                             mapping = aes(fill = layer),\n",
    "                             # show.legend = F,\n",
    "                             maxcell = 5010420,\n",
    "                             )+\n",
    "  scale_fill_manual(breaks = c(1,2,5,6,7,8,10),\n",
    "                    values=c(\"#70543e\", \"#b3ff1a\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", \"#42c9bc\",NA,NA,NA,NA),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Chlorophyceae\",\n",
    "                               \"Bacillariophyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\",\n",
    "                               \"Water\"))+\n",
    "  ggnewscale::new_scale(\"fill\")+\n",
    "  geom_sf(data = zoom_boat_high_sf, \n",
    "               fill = \"transparent\", \n",
    "               colour = \"black\",\n",
    "               linetype = \"dashed\", \n",
    "               linewidth = 1)+\n",
    "  geom_sf(data = zoom_boat_high_sf_2, \n",
    "               fill = \"transparent\", \n",
    "               colour = \"black\",\n",
    "               linetype = \"dashed\", \n",
    "               linewidth = 1)+\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.69452127,-8.6965,-8.69440039,-8.69105), \n",
    "                              y=c(40.66244282,40.66673,40.66244282,40.66673),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 1)+\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-8.6819877,-8.69645+0.007,-8.6818696,-8.6911+0.007), \n",
    "                              y=c(40.6652874,40.66675,40.6652874,40.66673),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", \n",
    "            linewidth = 1)+\n",
    "  annotation_custom(ggplotGrob(Boat_Pred_zoomed_map),\n",
    "                    xmin = -8.6975,\n",
    "                    xmax = -8.69,\n",
    "                    ymin = 40.6666,\n",
    "                    ymax = 40.6711) +\n",
    "  annotation_custom(ggplotGrob(Boat_Pred_zoomed_map_2),\n",
    "                    xmin = -8.6975+0.007,\n",
    "                    xmax = -8.69+0.007,\n",
    "                    ymin = 40.6666,\n",
    "                    ymax = 40.6711) +\n",
    "  coord_sf(crs = 4326)+\n",
    "  theme_Bede_Map()+\n",
    "   theme(\n",
    "     # axis.text.x = element_text(size = 20, angle = 20,hjust = 1),\n",
    "     axis.text.x = element_blank(),\n",
    "         axis.title = element_blank(),\n",
    "        axis.text.y = element_text(size = 20),\n",
    "         legend.background = element_blank(),\n",
    "         legend.position = c(0.98,0.38),\n",
    "         legend.text = element_text(size = 18),\n",
    "         legend.key.size = unit(0.8,\"cm\"))+\n",
    "  scale_x_continuous(limits = c(-8.6965,-8.675),\n",
    "                     breaks =  -8.685)+\n",
    "  scale_y_continuous(limits = c(40.659,40.6704),\n",
    "                     breaks =  40.664)\n",
    "################# PLOT AND SAVING ###########\n",
    "plot_Boat<-RGB_Boat_map/\n",
    "  Boat_pred_map\n",
    "\n",
    "ggsave(\"Figures/Maps Pred/FigX-Boat_Pred.png\",plot_Boat , width = 1000*4, height = 1120*4, units = \"px\")"
   ],
   "id": "b1669f57-8a3d-452d-bb1e-7e9182336abe"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Maps Pred/FigX-Boat_Pred.png\")"
   ],
   "id": "cell-fig-Boat"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The flight over L’Epine in Noirmoutier Island, France (<a href=\"#fig-Dike\" class=\"quarto-xref\">Figure 7</a>) was conducted near a dike crossing the northern part of the scene from west to east. Alongside this dike, brown algae attached to rocks and stranded green algae could be found. Despite the high mixture between Chlorophyceae and Magnoliopsida these two classes were correctly discriminated by the classifier."
   ],
   "id": "97879d0a-c034-46cd-9356-70d480268970"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imglist<-\n",
    "  data.frame(path = list.files(pattern = \".tif\", recursive = T, full.names = T)) %>% \n",
    "  mutate(folder = case_when(str_detect(path, \"Data\") ~ \"Input\",\n",
    "                            TRUE ~ \"Output\"),\n",
    "         alt = case_when(str_detect(path, \"high\") ~ \"high\",\n",
    "                         str_detect(path, \"High\") ~ \"high\",\n",
    "                         TRUE ~ \"low\"),\n",
    "         site = case_when(str_detect(path, \"Gafanha\") ~ \"Gafanha\",\n",
    "                          str_detect(path, \"AveiroHigh\") ~ \"Gafanha\",\n",
    "                          TRUE ~ \"Duer\"),\n",
    "         type = case_when(str_detect(path, \"pred\") ~ \"Pred\",\n",
    "                          str_detect(path, \"Preds\") ~ \"Pred\",\n",
    "                          str_detect(path, \"NDVI\") ~ \"NDVI\",\n",
    "                          TRUE ~ \"ortho\"),\n",
    "         resampled = case_when(str_detect(path, \"resampled\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         new = case_when(str_detect(path, \"no_Xantho\") ~ \"TRUE\",\n",
    "                         str_detect(path, \"NewPreds\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         col = case_when(str_detect(path, \"colored\") ~ \"TRUE\",\n",
    "                          TRUE ~ \"FALSE\"),\n",
    "         correction = case_when(str_detect(path, \"corrected_per_bands\") ~ \"Bands\",\n",
    "                          str_detect(path, \"corrected.tif\") ~ \"lm_all\",\n",
    "                          TRUE ~ \"NA\")) %>% \n",
    "  dplyr::filter(alt != \"NA\",\n",
    "                !str_detect(path,\".aux.xml\"))\n",
    "\n",
    "##### Preview #####\n",
    "\n",
    "filelist<-list.files(\"Data/shp/\", pattern = \".shp\", recursive = T,full.names = T)\n",
    "mask_BB <- filelist[str_detect(filelist,\"mask_land_intertidal_BB\")] %>% \n",
    "  read_sf()\n",
    "\n",
    "Flight_BB<-data.frame(\n",
    "  Name=c(\n",
    "  \"L'Epine\"\n",
    "  ) ,\n",
    "  Description = c(\n",
    "                  \"120m Flight\"),\n",
    "  Long=c(\n",
    "         -2.228239) ,\n",
    "  Lat=c( \n",
    "        46.989515) \n",
    "  )  %>% \n",
    "  st_as_sf(coords=c(\"Long\",\"Lat\") )  %>% \n",
    "  st_set_crs(\"EPSG:4326\")  %>% \n",
    "  st_transform(crs(mask_BB))  %>% \n",
    "  dplyr::mutate(lon = sf::st_coordinates(.) [,1],\n",
    "                lat = sf::st_coordinates(.) [,2]) %>% \n",
    "  sf::st_set_geometry(NULL)\n",
    "\n",
    "p1 <- ggplot() +\n",
    "  scale_fill_manual(\"\", labels = c(\"Intertidal area\", \"Land mask\"),\n",
    "                    values = c(\"#7DC27D\", \"#BDBDBD\"))+\n",
    "  geom_sf(data = mask_BB, mapping = aes(fill = Type), linewidth=0.05,alpha=0.93, show.legend = F,colour=\"grey30\")+\n",
    "      coord_sf(xlim=c(550609,579539) ,\n",
    "          ylim=c(5189447,5218377) )+\n",
    "  ggforce::geom_mark_ellipse(data=Flight_BB,\n",
    "                 aes(x=lon,\n",
    "                     y=lat,\n",
    "                     label = Name,\n",
    "                     group = Name),\n",
    "                     # description=Description),\n",
    "                 size=0.3,\n",
    "                 fill = \"goldenrod\",\n",
    "                 con.colour = \"goldenrod4\",\n",
    "                 show.legend=F,\n",
    "                 con.size = 2,\n",
    "                 label.fontsize = 10,\n",
    "                 label.hjust = 0.5,\n",
    "                 alpha=0.8,\n",
    "  expand = unit(2, \"mm\") , \n",
    "  radius = unit(2, \"mm\") ,\n",
    "  label.fill = \"grey90\",\n",
    "  label.buffer = unit(5, \"mm\")) +\n",
    "  # theme_Bede_Map()+\n",
    "  labs(x=\"Longitude\",\n",
    "       y=\"Latitude\")+\n",
    "  # geom_label(aes(x = 553460, y = 5216442, label = \"B\"), size = 20)+\n",
    "  theme(plot.margin = unit(c(0.1,0.1,0.1,0.1), \"cm\") ,\n",
    "        legend.position = c(0.95,0.2),\n",
    "        legend.key.size =  unit(1, \"cm\"),\n",
    "        legend.text = element_text(size = 20),\n",
    "        legend.title=element_blank(),\n",
    "        legend.background = element_rect(linewidth = 0.5, \n",
    "                                         fill = \"white\",\n",
    "                                         linetype=\"solid\", \n",
    "                                  colour =\"black\"),\n",
    "        # panel.background = element_rect(fill = \"#FFFFFF90\"),\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.x = element_blank(),\n",
    "        axis.text.y = element_blank(),\n",
    "        axis.ticks = element_blank())+\n",
    "  theme_void()\n",
    "\n",
    "##### RGB ######\n",
    "\n",
    "Dike_rgb<-imglist %>% \n",
    "  filter(str_detect(path,\"NorthDike_high_multi_4326_colored.tif\")) %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Dike_rgb)<-c(\"B1\",\"B3\",\"B5\",\"alpha\")\n",
    "values(Dike_rgb)[values(Dike_rgb$alpha) == 0] = NA\n",
    "\n",
    "mask_Dike<-\"Data/shp/Bourgneuf\" %>% \n",
    "  list.files(\".shp\", full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path, \"mask_NorthDike\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "background_Dike<-\"Data/shp/Bourgneuf\" %>% \n",
    "  list.files(\".shp\", full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path, \"background\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "zoom_Dike_high_sf<-\"Data/shp\" %>% \n",
    "  list.files(\".shp\",recursive = T,full.names = T) %>% \n",
    "  as.data.frame() %>% \n",
    "  rename(path = \".\") %>% \n",
    "  dplyr::filter(str_detect(path,\"zoom_extent_map_Dike_comparison\")) %>% \n",
    "  pull(path) %>% \n",
    "  read_sf()\n",
    "\n",
    "Dike_rgb_masked<-Dike_rgb %>% \n",
    "  crop(mask_Dike) %>%\n",
    "  mask(mask_Dike)\n",
    "\n",
    "\n",
    "zoom_rgb_dike<-Dike_rgb_masked %>% \n",
    "  crop(zoom_Dike_high_sf)\n",
    "\n",
    "zoom_rgb_dike_map<-ggplot()+\n",
    "  geom_spatraster_rgb(data = zoom_rgb_dike,\n",
    "                      r = 3,\n",
    "                      g = 2,\n",
    "                      b = 1)+\n",
    "  geom_sf(data = zoom_Dike_high_sf, fill = \"transparent\", linetype = \"dashed\",colour = \"black\", linewidth = 1)+\n",
    "  coord_sf(crs = 32630)+\n",
    "  theme_void()\n",
    "\n",
    "\n",
    "RGB_Dike_map <- ggplot()+\n",
    "  scale_fill_manual(\"\",  \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  geom_sf(data = background_Dike,\n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.93,\n",
    "          colour=\"transparent\")+\n",
    "  geom_sf(data = background_Dike,\n",
    "          fill = \"white\",\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"transparent\")+\n",
    "  tidyterra::geom_spatraster_rgb(data = Dike_rgb_masked,\n",
    "                      r = 3,\n",
    "                      g = 2,\n",
    "                      b = 1,\n",
    "                      # maxcell = 500,\n",
    "                      maxcell = 500832*8\n",
    "                      )+\n",
    "  geom_sf(data = zoom_Dike_high_sf, \n",
    "          fill = \"transparent\", \n",
    "          linetype = \"dashed\",\n",
    "          colour = \"black\", \n",
    "          linewidth = 1)+\n",
    "  annotation_custom(ggplotGrob(zoom_rgb_dike_map), \n",
    "                    xmin = -2.226, \n",
    "                    ymin = 46.9888,\n",
    "                    ymax = 46.9893) +\n",
    "  annotation_custom(ggplotGrob(p1),\n",
    "                    xmin = -2.226,\n",
    "                    ymin = 46.9875) +\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-2.226834,-2.225905,-2.226834,-2.225905),\n",
    "                              y=c(46.9886439,46.9892105,46.9885538,46.988805),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", linewidth = 1)+\n",
    "  ggspatial::annotation_scale(location = \"tl\",text_cex=2)+\n",
    "  scale_x_continuous(limits = c(-2.2293,-2.2253),\n",
    "                     breaks =  -2.228)+\n",
    "  scale_y_continuous(limits = c(46.9882,46.98935),\n",
    "                     breaks =  46.9890)+\n",
    "  theme_Bede_Map()+\n",
    "  theme(\n",
    "    # axis.text.x =element_text(size = 20, angle = 20, hjust = 1),\n",
    "    axis.text.x =element_blank(),\n",
    "        axis.title = element_blank(),\n",
    "        axis.text.y = element_text(size = 20))+\n",
    "   coord_sf(crs = 4326)\n",
    "\n",
    "\n",
    "##### Pred ######\n",
    "\n",
    "Dike_Pred<-imglist %>% \n",
    "  filter(str_detect(path,\"Noirmoutier_highPrediction_nnNewStanRawMetrics_Preds_NoXantho_7_4326.tif\")) %>% \n",
    "  pull(path) %>% \n",
    "  rast()\n",
    "names(Dike_Pred)<-c(\"layer\")\n",
    "values(Dike_Pred)[values(Dike_Pred) == 0] = NA\n",
    "\n",
    "Dike_Pred_masked<-Dike_Pred %>% \n",
    "  crop(mask_Dike) %>% \n",
    "  mask(mask_Dike) %>% \n",
    "  as.factor()\n",
    "\n",
    "zoom_Pred_dike<-Dike_Pred_masked %>% \n",
    "  crop(zoom_Dike_high_sf)\n",
    "\n",
    "zoom_Pred_dike_map<-ggplot()+\n",
    "  tidyterra::geom_spatraster(data = zoom_Pred_dike, mapping = aes(fill = layer), show.legend = F)+\n",
    "  scale_fill_manual(values=c(\"#70543e\",\"#b3ff1a\",\"#389318\",\"#873e23\",\"#42c9bc\"),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Chlorphyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Water\"))+\n",
    "  geom_sf(data = zoom_Dike_high_sf, fill = \"transparent\", linetype = \"dashed\",colour = \"black\", linewidth = 1)+\n",
    "  theme_void()\n",
    "\n",
    "\n",
    "Dike_pred_map<-ggplot()+\n",
    "  geom_sf(data = background_Dike, \n",
    "          mapping = aes(fill = Type),\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.93,\n",
    "          colour=\"grey30\")+\n",
    "  geom_sf(data = background_Dike,\n",
    "          fill = \"white\",\n",
    "          show.legend = F, \n",
    "          linewidth=0.05,\n",
    "          alpha=0.4,\n",
    "          colour=\"white\")+\n",
    "    scale_fill_manual(\"\",  \n",
    "                    labels = c(\"Intertidal area\", \"Land area\", \"water\"),\n",
    "                    values = c(\"#7DC27D\", \"#CFCFCF\",\"#42c9bc\"))+\n",
    "  ggnewscale::new_scale(\"fill\")+\n",
    "  tidyterra::geom_spatraster(data = Dike_Pred_masked, \n",
    "                             mapping = aes(fill = layer),\n",
    "                             maxcell = 5010420,\n",
    "                             )+\n",
    "  scale_fill_manual(breaks = c(1,2,5,6,7,8,10),\n",
    "                    values=c(\"#70543e\", \"#b3ff1a\", \"#DAA520\", \"#389318\",\"#873e23\",\"#b3002d\", \"#42c9bc\",NA,NA,NA,NA),\n",
    "                    na.value = NA,\n",
    "                    name = \"\",\n",
    "                    labels = c(\"Sediment\",\n",
    "                               \"Chlorophyceae\",\n",
    "                               \"Bacillariophyceae\",\n",
    "                               \"Magnoliopsida\",\n",
    "                               \"Phaeophyceae\",\n",
    "                               \"Rhodophyceae\",\n",
    "                               \"Water\"))+\n",
    "  ggnewscale::new_scale(\"fill\")+\n",
    "  geom_sf(data = zoom_Dike_high_sf, \n",
    "          fill = \"transparent\", \n",
    "          colour = \"black\",\n",
    "          linetype = \"dashed\", \n",
    "          linewidth = 1)+\n",
    "  annotation_custom(ggplotGrob(zoom_Pred_dike_map), \n",
    "                    xmin = -2.226, \n",
    "                    ymin = 46.9888,\n",
    "                    ymax = 46.9893) +\n",
    "  geom_path(aes(x,y,group=grp),\n",
    "            data = data.frame(x = c(-2.226834,-2.225905,-2.226834,-2.225905),\n",
    "                              y=c(46.9886439,46.9892105,46.9885538,46.988805),\n",
    "                              grp=c(1,1,2,2)),\n",
    "            linetype = \"dashed\", linewidth = 1)+\n",
    "  # annotation_custom(ggplotGrob(Boat_Pred_zoomed_map), xmin = -8.697, xmax = -8.692, ymin = 40.667,ymax = 40.6705) +\n",
    "  coord_sf(crs = 4326)+\n",
    "  theme_Bede_Map()+\n",
    "   theme(axis.text.x = element_text(size = 20, angle = 20,hjust = 1),\n",
    "         axis.title = element_blank(),\n",
    "        axis.text.y = element_text(size = 20),\n",
    "         legend.background = element_blank(),\n",
    "         legend.position = c(1,0.50),\n",
    "         legend.text = element_text(size = 18),\n",
    "         legend.key.size = unit(0.8,\"cm\"))+\n",
    "  scale_x_continuous(limits = c(-2.2293,-2.2253),\n",
    "                     breaks =  -2.228)+\n",
    "  scale_y_continuous(limits = c(46.9882,46.98935),\n",
    "                     breaks =  46.9890)\n",
    "\n",
    "################# PLOT AND SAVING ###########\n",
    "\n",
    "plot_Dike<-RGB_Dike_map/\n",
    "  Dike_pred_map\n",
    "  \n",
    "ggsave(\"Figures/Maps Pred/FigX-Dike_Pred.png\",plot_Dike , width = 1200*4, height = 900*4, units = \"px\")"
   ],
   "id": "237f8e72-280d-4c0d-8edb-c58e79f1f2f8"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Maps Pred/FigX-Dike_Pred.png\")"
   ],
   "id": "cell-fig-Dike"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.2 Validation"
   ],
   "id": "e8a504b6-fedd-4969-8772-36785ce424d2"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: Duplicated aesthetics after name standardisation: size"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: st_centroid assumes attributes are constant over geometries"
     ]
    }
   ],
   "source": [
    "df<-\"Output/Validation_data\" %>% \n",
    "  list.files(\"ResultsNewNN_NoXanthowithNDVIWithOldandBoat.csv\",recursive = T, full.names = T) %>% \n",
    "  read.csv() %>% \n",
    "  as_tibble() %>% \n",
    "  rename(Truth_ID = \"Truth\") %>% \n",
    "  mutate(Truth =  case_when(Truth_ID == 1 ~ \"Sediment\",\n",
    "                          Truth_ID == 2 ~ \"Chlorophyceae\",\n",
    "                          Truth_ID == 3 ~ \"Sediment\",\n",
    "                          Truth_ID == 4 ~ \"Magnoliopsida\",\n",
    "                          Truth_ID == 5 ~ \"Bacillariophyceae\",\n",
    "                          Truth_ID == 6 ~ \"Magnoliopsida\",\n",
    "                          Truth_ID == 7 ~ \"Phaeophyceae\",\n",
    "                          Truth_ID == 8 ~ \"Rhodophyceae\",\n",
    "                          Truth_ID == 9 ~ \"SunGlint\",\n",
    "                          Truth_ID == 10 ~ \"Water\",\n",
    "                          Truth_ID == 11 ~ \"Xanthophyceae\",\n",
    "                          Truth_ID == 12 ~ \"Spartina\",\n",
    "                          Truth_ID == 41 ~ \"Bleached seagrasses\",\n",
    "                          TRUE ~ \"NA\"),\n",
    "         Pred_Class = case_when(Pred_Class == \"Clorophyta\" ~ \"Chlorophyceae\",\n",
    "                                Pred_Class == \"Phaeophyta\" ~ \"Phaeophyceae\",\n",
    "                                Pred_Class == \"Rhodphyta\" ~ \"Rhodophyceae\",\n",
    "                                Pred_Class == \"Deep_Sediment\" ~ \"Sediment\",\n",
    "                                Pred_Class == \"MPB\" ~ \"Bacillariophyceae\",\n",
    "                                Pred_Class == \"Bare_Sediment\" ~ \"Sediment\",\n",
    "                                 Pred_Class == \"Magnoliosida\" ~ \"Magnoliopsida\",\n",
    "                                (Pred_Class == \"Low_SPC\" & NDVI > 0.2) ~ \"Magnoliopsida\",\n",
    "                                (Pred_Class == \"Low_SPC\" & NDVI <= 0.2 )~ \"Sediment\",\n",
    "                          TRUE ~ Pred_Class)) %>% \n",
    "  dplyr::filter(Truth != \"Water\",\n",
    "                Truth  != \"Spartina\",\n",
    "                Truth != \"SunGlint\",\n",
    "                Truth != \"Bleached seagrasses\",\n",
    "                Truth != \"Xanthophyceae\",\n",
    "                # Pred_Class != \"Low_SPC\",\n",
    "                Pred_Class != \"Xanthophyceae\",\n",
    "                 Pred_Class != \"Water\",\n",
    "                 Pred_Class != \"SunGlint\",\n",
    "                !Flight == \"Handfisher_1_old\",\n",
    "                !Flight == \"Handfisher_2_old\",\n",
    "                !Flight == \"Handfisher_2\") %>%\n",
    "  dplyr::select(-NDVI)\n",
    "\n",
    "\n",
    "a<-as.character(unique(df$Truth))\n",
    "b<-as.character(unique(df$Pred_Class))\n",
    "ab<-setdiff(a,b)\n",
    "ba<-setdiff(b,a)\n",
    "\n",
    "if (!(length(ab) == 0 & length(ba) == 0)) {\n",
    "  \n",
    "\n",
    "lines_to_add<-data.frame(Pred_Class = c(ab,ba),\n",
    "                         Pred_ID = NA,\n",
    "                         Flight = NA,\n",
    "                         Truth_ID = NA,\n",
    "                         Truth = c(ab,ba))\n",
    "\n",
    "df<-df %>% \n",
    "  rbind(lines_to_add)\n",
    "  rm(lines_to_add)\n",
    "\n",
    "  \n",
    "}\n",
    "\n",
    "\n",
    "  df<- df %>%\n",
    "  mutate(Pred_Class = as_factor(Pred_Class),\n",
    "         Truth = as_factor(Truth),\n",
    "         Truth = fct_relevel(Truth, c(\"Sediment\",\"Bacillariophyceae\",\"Chlorophyceae\",\"Magnoliopsida\",\"Rhodophyceae\",\"Phaeophyceae\")),\n",
    "         Pred_Class = fct_relevel(Pred_Class, c(\"Sediment\",\"Bacillariophyceae\",\"Chlorophyceae\",\"Magnoliopsida\",\"Rhodophyceae\",\"Phaeophyceae\")),\n",
    "         Flight = case_when(str_detect(Flight, \"Boat\") ~ \"Boat\",\n",
    "                            TRUE ~ Flight))\n",
    "  \n",
    "####### GafanahaLow #######\n",
    "df_GafanhaLow<-df %>% \n",
    "  filter(Flight == \"GafanhaLow\")\n",
    "\n",
    "validation_metrics_GafanhaLow<-confusionMatrix(df_GafanhaLow$Pred_Class,df_GafanhaLow$Truth)\n",
    "\n",
    "conf_mat_GafanhaLow<-validation_metrics_GafanhaLow[[2]] %>% \n",
    "  as.numeric() %>%\n",
    "  base::matrix(ncol = 6) %>%\n",
    "  as.data.frame()\n",
    "conf_mat_GafanhaLow[6,]<-NA\n",
    "conf_mat_GafanhaLow[,6]<-NA\n",
    "\n",
    "for(i in 1:(nrow(conf_mat_GafanhaLow))){\n",
    "  if(i == 1 ){\n",
    "    df_sum<-data.frame(row = NA,\n",
    "                       col = NA)\n",
    "  }\n",
    "  df_sum[i,1] <- sum(conf_mat_GafanhaLow[i,],na.rm = T)\n",
    "  df_sum[i,2] <- sum(conf_mat_GafanhaLow[,i],na.rm = T)\n",
    "}\n",
    "conf_mat_std_GafanhaLow<-conf_mat_GafanhaLow\n",
    "\n",
    "for(row in 1:nrow(conf_mat_std_GafanhaLow)){\n",
    "  for(col in 1:ncol(conf_mat_std_GafanhaLow)){\n",
    "    conf_mat_std_GafanhaLow[row,col]<-conf_mat_std_GafanhaLow[row,col]/((df_sum$row[row] + df_sum$col[col])-conf_mat_std_GafanhaLow[row,col])\n",
    "    }\n",
    "}\n",
    "\n",
    "confmat_rast_GafanhaLow<-conf_mat_std_GafanhaLow %>% \n",
    "  as.matrix() %>% \n",
    "  as.array() %>% \n",
    "  rast()\n",
    "\n",
    "confmat_rast_GafanhaLow<-subst(confmat_rast_GafanhaLow, NA, -99)\n",
    "names(confmat_rast_GafanhaLow)<-\"layer\"\n",
    "ext(confmat_rast_GafanhaLow)<-c(18.1,21.9,8.2,11.8)\n",
    "confmat_rast_GafanhaLow_sf<-as.polygons(confmat_rast_GafanhaLow,aggregate=F) %>% \n",
    "  sf::st_as_sf()\n",
    "confmat_rast_GafanhaLow_centroid<-st_centroid(confmat_rast_GafanhaLow_sf)"
   ],
   "id": "6fe57dd6-130c-4674-b99a-6caad5f7d632"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"./Figures/Low_res/Validation/ConfusionMatrixGlobal.png\")"
   ],
   "id": "cell-fig-Validation"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A total of 536,000 pixels was used to validate the Neural Network classifier. The sites with the lowest and highest number of validation data were Kerdrean (5557 pixels) and Marinha Lanzarote (159713 pixels), respectively. Model global accuracy was 94.26% with a Kappa coefficient of 0.92 (<a href=\"#fig-Validation\" class=\"quarto-xref\">Figure 8</a>). The least performing site was Gafanha High (global accuracy of 75.45%) whereas Mataducos was the site with the most accurate prediction (global accuracy of 98.05%). Overall, the classes Phaeophyceae, Magnoliopsida, Sediment and Rhodophyceae were correctly classified with a balanced accuracy of 1, 0.96, 0.96 and 0.91 respectively. Bacillariophyceae was the least performing class (accuracy of 0.72 ) mainly due to the confusion between Magnoliopsida and Sediment.\n",
    "\n",
    "## 3.3 Variable importance"
   ],
   "id": "2c8f7fbe-6655-4e5d-a85b-cbc85fa2feb4"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################# SPECTRA ########### \n",
    "SP1<-read.csv(\"Output/Spectra_VIP.csv\") %>% \n",
    "  dplyr::select(-X) %>% \n",
    "  mutate(type = \"Min/Max Standardised Spectra\") %>% \n",
    "  dplyr::filter(Class != \"Bare Sediment\",\n",
    "                Class != \"Sun Glint\",\n",
    "                Class != \"Water\",\n",
    "                Class != \"Xanthophyceae\",\n",
    "                Class != \"A - MagnoliopsidaL\") %>% \n",
    "  as_tibble()%>% \n",
    "  rename(value = \"STD\")\n",
    "\n",
    "SP2<-read.csv(\"Output/Spectra_VIP2.csv\") %>% \n",
    "  dplyr::select(-X) %>% \n",
    "  mutate(type = \"Raw Spectra\",\n",
    "         mean = mean / 10000)%>% \n",
    "  dplyr::filter(Class != \"Bare Sediment\",\n",
    "                Class != \"Sun Glint\",\n",
    "                Class != \"Water\",\n",
    "                Class != \"Xanthophyceae\",\n",
    "                Class != \"A - MagnoliopsidaL\") %>% \n",
    "  as_tibble() %>% \n",
    "  dplyr::reframe(Class,Wavelength, mean, type) %>% \n",
    "  rename(value = \"mean\")\n",
    "\n",
    "SP<-SP1 %>% \n",
    "  rbind(SP2)\n",
    "\n",
    "\n",
    "SP_plot<-ggplot(SP, aes( x = Wavelength, y = value, color = Class))+\n",
    "  geom_line(linewidth = 1.5, show.legend = F)+\n",
    "  facet_wrap(~ type, scale = \"free\",ncol =1) +\n",
    "  scale_color_manual(values=c(\"#389318\", \"#873e23\", \"#b3002d\", \"#b3ff1a\", \"#DAA520\"))+\n",
    "  ylab(\"Reflectance\")+\n",
    "  xlab(\"Wavelength (nm)\")+\n",
    "  theme_Bede()+\n",
    "  theme(axis.text.x = element_text(size = 20),\n",
    "        axis.text.y = element_text(size = 20),\n",
    "        axis.title.x = element_text(size = 25),\n",
    "        axis.title.y = element_text(size = 25),\n",
    "        strip.text = element_text(size=25),\n",
    "        legend.position = c(0.2, 0.7),\n",
    "        plot.background = element_rect(fill = rgb(1,1,1, alpha=0), colour = NA),\n",
    "        legend.title=element_blank(),\n",
    "        legend.text = element_text(size = 5),\n",
    "        legend.key.size = unit(0.25, \"cm\"),\n",
    "        legend.text.align = 0,\n",
    "        panel.border = element_blank(), \n",
    "        panel.background = element_blank(),\n",
    "        axis.line = element_line(color=\"black\"))\n",
    "\n",
    "############## VIP ##############\n",
    "\n",
    "filelist<-\"Data\" %>% \n",
    "  list.files(pattern = \"NN_VIP_byClass_NoXanth.csv\", recursive = T,full.names = T)\n",
    "\n",
    "VIP<-filelist[which(str_detect(filelist,\"VIP\"))] %>% \n",
    "  read.csv() %>% \n",
    "  mutate(Magnoliopsida = (Magnoliosida + Low_SPC)/2) %>% \n",
    "  dplyr::select(-c(Magnoliosida ,Low_SPC))\n",
    "\n",
    "VIP1<-VIP %>%\n",
    "  dplyr::select(-c(Water,\n",
    "                   SunGlint,\n",
    "                   Deep_Sediment,\n",
    "                   Bare_Sediment)) %>%\n",
    "  # dplyr::filter(!str_detect(X,\"NDVI\")) %>% \n",
    "  mutate(isSTD = case_when(str_detect(X,\"Stan\") ~ \"Standardised\", \n",
    "                         TRUE ~ \"RAW\")) %>%\n",
    "  pivot_longer(-c(X,isSTD), names_to = \"Class\",values_to = \"value\") %>% \n",
    "  dplyr::filter(X != \"NDVI\",\n",
    "                X != \"NDVI_Stan\") %>% \n",
    "  group_by(Class) %>%\n",
    "  mutate(STD = (value-min(value))/(max(value)-min(value))) %>% \n",
    "  ungroup() %>% \n",
    "  mutate(Class = case_when(Class == \"Magnoliosida\" ~ \"Magnoliopsida\",\n",
    "                           Class == \"Low_SPC\" ~ \"Magnoliopsida\",\n",
    "                           Class == \"Clorophyta\" ~ \"Chlorophyceae\",\n",
    "                           Class == \"Rhodphyta\" ~ \"Rhodophyceae\",\n",
    "                           Class == \"MPB\" ~ \"Bacillariophyceae\",\n",
    "                           Class == \"Phaeophyta\" ~ \"Phaeophyceae\",\n",
    "                           TRUE ~ Class),\n",
    "         Variable = str_replace_all(str_replace_all(str_replace_all(X,\"Reflectance\",\"\"),\"Stan\",\"\"),\"_\",\"\")) %>% \n",
    "  dplyr::filter(Class != \"Total\") \n",
    "\n",
    "VIP_summarised<-VIP1 %>% \n",
    "  group_by(Variable, Class) %>% \n",
    "  dplyr::summarise(sumVIP = sum(STD)) %>% \n",
    "  ungroup() %>% \n",
    "  group_by(Class) %>% \n",
    "  mutate(STD = (sumVIP-min(sumVIP))/(max(sumVIP)-min(sumVIP)))\n",
    "\n",
    "VIP_subPLOT<-VIP_summarised %>% \n",
    "  ggplot(aes(x = Variable, y = STD, fill = Class)) +\n",
    "  geom_col(color = \"black\",position =  \"dodge\") +\n",
    "  coord_polar()+\n",
    "  scale_y_continuous(limits = c(0,1), breaks = c(0.25,0.5,0.75,1))+\n",
    "  facet_wrap(~Class)+\n",
    "  scale_fill_manual(\"Class\", labels = c(\"Bacillariophyceae\",\"Chlorophyceae\", \"Magnoliopsida\", \"Phaeophyceae\",\"Rhodophyceae\"),\n",
    "                    values = c(\"#DAA520\",\"#b3ff1a\", \"#389318\",\"#873e23\",\"#b3002d\"))+\n",
    "  theme_Bede()+\n",
    "  theme(panel.grid.major.y = element_line(linetype = \"dotted\",color = c(\"grey9\",\"grey9\",\"grey9\",\"grey9\",NA)))+\n",
    "  ylab(\"Importance\")+\n",
    "  theme(legend.position=\"top\",\n",
    "        strip.text = element_text(size = 25),\n",
    "        axis.text.y = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.ticks.y =element_blank(),\n",
    "        axis.title.x =element_blank(),\n",
    "        axis.title.y =element_text(size = 20),\n",
    "        legend.key.size = unit(1, 'cm'),\n",
    "        legend.title = element_text(size=20),\n",
    "        legend.text = element_text(size=20))+\n",
    "  guides(fill = guide_legend(nrow = 1))\n",
    "\n",
    "VIP_PLOT<-VIP_subPLOT+\n",
    "  # geom_text(aes(label = \"Standardised Reflectance Spectra\"), color = \"black\", size = 8)+\n",
    "  inset_element(SP_plot,left = 0.6341, top = 0.47, bottom = -0.01, right = 0.90,  align_to = \"full\")\n",
    "VIP_PLOT\n",
    "ggsave(\"Figures/VIP/Fig_VIP.png\", VIP_PLOT, width = 1920*4, height = 1129*4, units = \"px\")\n",
    "\n",
    "\n",
    "################# VIP NDVI ######################\n",
    "\n",
    "VIP_NDVI<-\n",
    "  VIP %>%\n",
    "  dplyr::select(-c(Water,\n",
    "                   SunGlint,\n",
    "                   Deep_Sediment,\n",
    "                   Bare_Sediment)) %>%\n",
    "  # dplyr::filter(!str_detect(X,\"NDVI\")) %>% \n",
    "  mutate(isSTD = case_when(str_detect(X,\"Stan\") ~ \"Standardised\", \n",
    "                         TRUE ~ \"RAW\")) %>%\n",
    "  pivot_longer(-c(X,isSTD), names_to = \"Class\",values_to = \"value\") %>% \n",
    "  dplyr::filter(X == \"NDVI\") %>% \n",
    "  group_by(Class) %>%\n",
    "  mutate(STD = (value-min(value))/(max(value)-min(value))) %>% \n",
    "  ungroup() %>% \n",
    "  mutate(Class = case_when(Class == \"Magnoliosida\" ~ \"Magnoliopsida\",\n",
    "                           Class == \"Low_SPC\" ~ \"Magnoliopsida\",\n",
    "                           Class == \"Clorophyta\" ~ \"Chlorophyceae\",\n",
    "                           Class == \"Rhodphyta\" ~ \"Rhodophyceae\",\n",
    "                           Class == \"MPB\" ~ \"Bacillariophyceae\",\n",
    "                           Class == \"Phaeophyta\" ~ \"Phaeophyceae\",\n",
    "                           TRUE ~ Class),\n",
    "         Variable = str_replace_all(str_replace_all(str_replace_all(X,\"Reflectance\",\"\"),\"Stan\",\"\"),\"_\",\"\")) %>% \n",
    "  dplyr::filter(Class != \"Total\") %>% \n",
    "  group_by(Class) %>% \n",
    "  summarise(sum = sum(value)) %>% \n",
    "  mutate(STD = (sum-min(sum))/(max(sum)-min(sum))) %>% \n",
    "  ggplot(aes(x = Class, y = STD, fill = Class)) +\n",
    "  geom_col(color = \"black\",position =  \"dodge\") +\n",
    "  # coord_polar()+\n",
    "  scale_y_continuous(limits = c(0,1), breaks = c(0.25,0.5,0.75,1))+\n",
    "  # facet_wrap(~Class)+\n",
    "  scale_fill_manual(\"Class\", labels = c(\"Bacillariophyceae\",\"Chlorophyceae\", \"Magnoliopsida\", \"Phaeophyceae\",\"Rhodophyceae\"),\n",
    "                    values = c(\"#DAA520\",\"#b3ff1a\", \"#389318\",\"#873e23\",\"#b3002d\"))+  theme_Bede()+\n",
    "  theme(panel.grid.major.y = element_line(linetype = \"dotted\",color = c(\"grey9\",\"grey9\",\"grey9\",\"grey9\",NA)))+\n",
    "  ylab(\"Importance\")+\n",
    "  theme(legend.position=\"top\",\n",
    "        strip.text = element_text(size = 25),\n",
    "        axis.text.y = element_blank(),\n",
    "        axis.text.x = element_text(size = 20),\n",
    "        axis.ticks.y =element_blank(),\n",
    "        axis.title.x =element_blank(),\n",
    "        axis.title.y =element_text(size = 20),\n",
    "        legend.key.size = unit(1, 'cm'),\n",
    "        legend.title = element_text(size=20),\n",
    "        legend.text = element_text(size=20))+\n",
    "  guides(fill = guide_legend(nrow = 1))"
   ],
   "id": "ae8d77a3-7368-49ce-9123-aad4ac4f37c2"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"Figures/Low_res/VIP/Fig_VIP.png\")"
   ],
   "id": "cell-fig-VIP"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The computation of the variable importance makes it possible to identify which wavelengths are the most useful for class prediction (<a href=\"#fig-VIP\" class=\"quarto-xref\">Figure 9</a>). The spectral bands at 444 , 717 and 842 nm of the Micasense camera are important for none of the vegetation classes. The band at 531 nm is the only important predictor for the classifier to accurately predict chlorophyceae. In fact, at this wavelength, the Chlorophyceae spectra has the maximum reflectance among of all vegetation classes. The bands at 531 and 740 nm are the most important predictors for Phaeophyceae, corresponding to the minimum reflectance among all classes. Bands at 475 and 560 nm are the most important predictors for Bacillariophyceae and Rhodophyceae, respectively. Four predictors, ranging from the green (560 nm) to the RedEdge (705 nm) bands are important to accurately predict magnoliopsida.\n",
    "\n",
    "## 3.4 Effect of the flight height on the prediction"
   ],
   "id": "32714d57-218b-4b04-8e9c-57b3ef6bdbe5"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if(length(list.files(\"./Output\",pattern = \"output_80_filtered_Upscaling.csv\")) == 0 ){\n",
    " imglist<-\"Output/High_Flights_correction/Low_flight_spatial_resampling/\" %>%\n",
    "   list.files(\".tif\",full.names = T) %>%\n",
    "   as.data.frame() %>%\n",
    "   rename(path = \".\") %>%\n",
    "   mutate(size = case_when(str_detect(path,\"_8_7x7.tif\") ~ 8,\n",
    "                           str_detect(path,\"16.tif\") ~ 16,\n",
    "                           str_detect(path,\"32.tif\") ~ 32,\n",
    "                           str_detect(path,\"48.tif\") ~ 48,\n",
    "                           str_detect(path,\"64.tif\") ~ 64,\n",
    "                           str_detect(path,\"80.tif\") ~ 80),\n",
    "          type = case_when(str_detect(path, \"ortho\") ~ \"ortho\",\n",
    "                           str_detect(path, \"pred\") ~ \"Pred\",\n",
    "                           TRUE ~ \"Prob\"))\n",
    "\n",
    "pred_8<-imglist %>%\n",
    "   filter(size == 8,\n",
    "          type == \"Pred\") %>%\n",
    "   pull(path) %>%\n",
    "   terra::rast()\n",
    " names(pred_8)<-\"Pred_8\"\n",
    "\n",
    " pred_80<-imglist %>%\n",
    "   filter(size == 80,\n",
    "          type == \"Pred\") %>%\n",
    "   pull(path) %>%\n",
    "   terra::rast()\n",
    " names(pred_80)<-\"Pred_80\"\n",
    " \n",
    "prob_80<-imglist %>%\n",
    "   filter(size == 80,\n",
    "          type == \"Prob\") %>%\n",
    "   pull(path) %>%\n",
    "   terra::rast()\n",
    " names(prob_80)<-\"Prob_80\"\n",
    "\n",
    "\n",
    " model_80<-imglist %>%\n",
    "   filter(size == 80,\n",
    "          type == \"Pred\") %>%\n",
    "   pull(path) %>%\n",
    "   terra::rast()\n",
    " values(model_80)<-c(1:ncell(model_80))\n",
    " names(model_80)<-\"ID_80\"\n",
    "\n",
    "\n",
    " model_80_rs_to_8<-terra::resample(model_80,pred_8,method = \"near\")\n",
    " names(model_80_rs_to_8)<-\"ID_80\"\n",
    "\n",
    " df_80_8<-terra::rast(list(model_80_rs_to_8,pred_8)) %>%\n",
    "   as.data.frame(xy=F) %>%\n",
    "   filter(!Pred_8 == 0)\n",
    "\n",
    " df_80<-terra::rast(list(model_80,pred_80,prob_80)) %>%\n",
    "   as.data.frame(xy=F) %>%\n",
    "   filter(!Pred_80 == 0)\n",
    "\n",
    "#############\n",
    "output_80<-df_80 %>% \n",
    "  left_join(df_80_8, by = \"ID_80\") %>% \n",
    "  filter(!is.na(Pred_8)) %>% \n",
    "  mutate(Class_80 = case_when(Pred_80 == 1 ~ \"Bare Sediment\",\n",
    "                              Pred_80 == 2 ~ \"Green Algae\",\n",
    "                              Pred_80 == 3 ~ \"Bare Sediment\",\n",
    "                              Pred_80 == 4 ~ \"Magnoliopsida\",\n",
    "                              Pred_80 == 5 ~ \"Microphytobenthos\",\n",
    "                              Pred_80 == 6 ~ \"Magnoliopsida\",\n",
    "                              Pred_80 == 7 ~ \"Phaeophyceae\",\n",
    "                              Pred_80 == 8 ~ \"Rhodophyceae\",\n",
    "                              Pred_80 == 9 ~ \"Sun Glint\",\n",
    "                              Pred_80 == 10 ~ \"Water\",\n",
    "                              Pred_80 == 11 ~ \"Xanthophyceae\",\n",
    "                              TRUE ~ as.character(Pred_80)),\n",
    "         Class_8 = case_when(Pred_8 == 1 ~ \"Bare Sediment\",\n",
    "                              Pred_8 == 2 ~ \"Green Algae\",\n",
    "                              Pred_8 == 3 ~ \"Bare Sediment\",\n",
    "                              Pred_8 == 4 ~ \"Magnoliopsida\",\n",
    "                              Pred_8 == 5 ~ \"Microphytobenthos\",\n",
    "                              Pred_8 == 6 ~ \"Magnoliopsida\",\n",
    "                              Pred_8 == 7 ~ \"Phaeophyceae\",\n",
    "                              Pred_8 == 8 ~ \"Rhodophyceae\",\n",
    "                              Pred_8 == 9 ~ \"Sun Glint\",\n",
    "                              Pred_8 == 10 ~ \"Water\",\n",
    "                              Pred_8 == 11 ~ \"Xanthophyceae\",\n",
    "                              TRUE ~ as.character(Pred_8))\n",
    "         ) %>% \n",
    "   dplyr::filter(Class_80 != \"Water\",\n",
    "                 Class_80 != \"Phaeophyceae\",\n",
    "                 Class_80 != \"Sun Glint\",\n",
    "                 Class_8 != \"Water\",\n",
    "                 Class_8 != \"Phaeophyceae\",\n",
    "                 Class_8 != \"Sun Glint\") %>% \n",
    "           group_by(ID_80, Prob_80) %>% \n",
    "           summarise(prop_Sed = sum((Pred_8 == 1)|(Pred_8 == 3))/n(),\n",
    "                     prop_Green = sum(Pred_8 == 2)/n(),\n",
    "                     prop_Magno = sum((Pred_8 == 4)|(Pred_8 == 6))/n(),\n",
    "                     prop_MPB = sum(Pred_8 == 5)/n(),\n",
    "                     prop_Rodo = sum(Pred_8 == 8)/n(),\n",
    "                     prop_Xantho = sum(Pred_8 == 11)/n(),\n",
    "                     True = unique(Class_80)) %>%\n",
    "  pivot_longer(-c(ID_80,True,Prob_80), names_to = \"Class\", values_to = \"Proportion\") %>% \n",
    "  mutate(Class = case_when(Class == \"prop_Sed\" ~ \"Bare Sediment\",\n",
    "                           Class == \"prop_Green\" ~ \"Green Algae\",\n",
    "                           Class == \"prop_Magno\" ~ \"Magnoliopsida\",\n",
    "                           Class == \"prop_MPB\" ~ \"Microphytobenthos\",\n",
    "                           Class == \"prop_Pheo\" ~ \"Phaeophyceae\",\n",
    "                           Class == \"prop_Rodo\" ~ \"Rhodophyceae\",\n",
    "                           Class == \"prop_Glint\" ~ \"Sun Glint\",\n",
    "                           Class == \"prop_Water\" ~ \"Water\",\n",
    "                           Class == \"prop_Xantho\" ~ \"Xanthophyceae\"),\n",
    "         same = case_when(True == Class ~ 1,\n",
    "                          TRUE ~ 0)) %>% \n",
    "  dplyr::filter(!Proportion == 0)\n",
    " \n",
    "  list<-output_80 %>% \n",
    "  dplyr::filter(Proportion == 1,\n",
    "                Class == True) %>% \n",
    "  pull(ID_80)\n",
    "\n",
    "output_80_filtered<-output_80 %>% \n",
    "  dplyr::filter(!ID_80 %in% list )\n",
    "\n",
    "\n",
    "\n",
    "output_80_filtered_same<-output_80_filtered %>% \n",
    "  dplyr::filter(same == 1)\n",
    "\n",
    "write.csv(output_80_filtered, \"./Output/output_80_filtered_Upscaling.csv\")\n",
    "write.csv(output_80_filtered_same, \"./Output/output_80_filtered_same_Upscaling.csv\")\n",
    "\n",
    "}else{\n",
    "  \n",
    "  output_80_filtered<-read.csv(\"./Output/output_80_filtered_Upscaling.csv\") %>% \n",
    "    dplyr::mutate(Proportion = Proportion *100) %>% \n",
    "    dplyr::select(-X) %>% \n",
    "    as_tibble()\n",
    "  output_80_filtered_same<-read.csv(\"./Output/output_80_filtered_same_Upscaling.csv\")%>% \n",
    "    dplyr::mutate(Proportion = Proportion *100) %>% \n",
    "    dplyr::select(-X) %>% \n",
    "    as_tibble()\n",
    "}\n",
    "\n",
    "################# LM ####################\n",
    "lm_sediment<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Bare Sediment\") \n",
    "\n",
    " lm_sediment<- lm(lm_sediment$Prob_80 ~ lm_sediment$Proportion)\n",
    " \n",
    " lm_green<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Green Algae\") \n",
    "  lm_green<- lm(lm_green$Prob_80 ~ lm_green$Proportion)\n",
    " \n",
    " lm_magno<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Magnoliopsida\") \n",
    "   lm_magno<- lm(lm_magno$Prob_80 ~ lm_magno$Proportion)\n",
    "\n",
    " \n",
    " lm_MPB<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Microphytobenthos\")\n",
    "    lm_MPB<- lm(lm_MPB$Prob_80 ~ lm_MPB$Proportion)\n",
    "    \n",
    "lm_Rodo<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Rodophyceae\")\n",
    "    lm_Rodo<- lm(lm_Rodo$Prob_80 ~ lm_Rodo$Proportion)\n",
    " \n",
    " ########### ALPHA ###############\n",
    "alpha_sed<-data.frame(prop = seq(0,1,  0.01)) %>% \n",
    "  mutate(alpha = lm_sediment$coefficients[[2]]*prop+lm_sediment$coefficients[[1]],\n",
    "         ID = c(1:nrow(.)),\n",
    "         xmin = seq(0,1,  0.01),\n",
    "         xmax = seq(0.01,1.01,0.01),\n",
    "         ymin = rep(0,nrow(.)),\n",
    "         ymax = rep(1,nrow(.))) %>% \n",
    "  pivot_longer(c(xmin,xmax),names_to = \"name_x\", values_to = \"x\") %>% \n",
    "  pivot_longer(c(ymax,ymin), names_to = \"name_y\",values_to = \"y\") %>% \n",
    "  mutate(y = rep(c(0,1,1,0),nrow(.)/4)) %>% \n",
    "  dplyr::select(-c(name_x,name_y)) %>% \n",
    "  mutate(alpha = case_when(alpha > 1 ~ 1,\n",
    "                           TRUE ~ alpha),\n",
    "         alpha = 1-alpha)%>% \n",
    "  mutate(alpha = (alpha - min(alpha))/ (max(alpha) - min(alpha)))\n",
    "\n",
    "alpha_green<-data.frame(prop = seq(0,1,  0.01)) %>% \n",
    "  mutate(alpha = lm_green$coefficients[[2]]*prop+lm_green$coefficients[[1]],\n",
    "         ID = c(1:nrow(.)),\n",
    "         xmin = seq(0,1,  0.01),\n",
    "         xmax = seq(0.01,1.01,0.01),\n",
    "         ymin = rep(0,nrow(.)),\n",
    "         ymax = rep(1,nrow(.))) %>% \n",
    "  pivot_longer(c(xmin,xmax),names_to = \"name_x\", values_to = \"x\") %>% \n",
    "  pivot_longer(c(ymax,ymin), names_to = \"name_y\",values_to = \"y\") %>% \n",
    "  mutate(y = rep(c(0,1,1,0),nrow(.)/4)) %>% \n",
    "  dplyr::select(-c(name_x,name_y)) %>% \n",
    "  mutate(alpha = case_when(alpha > 1 ~ 1,\n",
    "                           TRUE ~ alpha),\n",
    "         alpha = 1-alpha)%>% \n",
    "  mutate(alpha = (alpha - min(alpha))/ (max(alpha) - min(alpha)))\n",
    "\n",
    "alpha_magno<-data.frame(prop = seq(0,1,  0.01)) %>% \n",
    "  mutate(alpha = lm_magno$coefficients[[2]]*prop+lm_magno$coefficients[[1]],\n",
    "         ID = c(1:nrow(.)),\n",
    "         xmin = seq(0,1,  0.01),\n",
    "         xmax = seq(0.01,1.01,0.01),\n",
    "         ymin = rep(0,nrow(.)),\n",
    "         ymax = rep(1,nrow(.))) %>% \n",
    "  pivot_longer(c(xmin,xmax),names_to = \"name_x\", values_to = \"x\") %>% \n",
    "  pivot_longer(c(ymax,ymin), names_to = \"name_y\",values_to = \"y\") %>% \n",
    "  mutate(y = rep(c(0,1,1,0),nrow(.)/4)) %>% \n",
    "  dplyr::select(-c(name_x,name_y)) %>% \n",
    "  mutate(alpha = case_when(alpha > 1 ~ 1,\n",
    "                           TRUE ~ alpha),\n",
    "         alpha = 1-alpha)%>% \n",
    "  mutate(alpha = (alpha - min(alpha))/ (max(alpha) - min(alpha)))\n",
    "\n",
    "alpha_MPB<-data.frame(prop = seq(0,1,  0.01)) %>% \n",
    "  mutate(alpha = lm_MPB$coefficients[[2]]*prop+lm_MPB$coefficients[[1]],\n",
    "         ID = c(1:nrow(.)),\n",
    "         xmin = seq(0,1,  0.01),\n",
    "         xmax = seq(0.01,1.01,0.01),\n",
    "         ymin = rep(0,nrow(.)),\n",
    "         ymax = rep(1,nrow(.))) %>% \n",
    "  pivot_longer(c(xmin,xmax),names_to = \"name_x\", values_to = \"x\") %>% \n",
    "  pivot_longer(c(ymax,ymin), names_to = \"name_y\",values_to = \"y\") %>% \n",
    "  mutate(y = rep(c(0,1,1,0),nrow(.)/4)) %>% \n",
    "  dplyr::select(-c(name_x,name_y)) %>% \n",
    "  mutate(alpha = case_when(alpha > 1 ~ 1,\n",
    "                           TRUE ~ alpha),\n",
    "         alpha = 1-alpha)%>% \n",
    "  mutate(alpha = (alpha - min(alpha))/ (max(alpha) - min(alpha)))\n",
    "\n",
    "alpha_Rodo<-data.frame(prop = seq(0,1,  0.01)) %>% \n",
    "  mutate(alpha = lm_Rodo$coefficients[[2]]*prop+lm_Rodo$coefficients[[1]],\n",
    "         ID = c(1:nrow(.)),\n",
    "         xmin = seq(0,1,  0.01),\n",
    "         xmax = seq(0.01,1.01,0.01),\n",
    "         ymin = rep(0,nrow(.)),\n",
    "         ymax = rep(1,nrow(.))) %>% \n",
    "  pivot_longer(c(xmin,xmax),names_to = \"name_x\", values_to = \"x\") %>% \n",
    "  pivot_longer(c(ymax,ymin), names_to = \"name_y\",values_to = \"y\") %>% \n",
    "  mutate(y = rep(c(0,1,1,0),nrow(.)/4)) %>% \n",
    "  dplyr::select(-c(name_x,name_y)) %>% \n",
    "  mutate(alpha = case_when(alpha > 1 ~ 1,\n",
    "                           TRUE ~ alpha),\n",
    "         alpha = 1-alpha)%>% \n",
    "  mutate(alpha = (alpha - min(alpha))/ (max(alpha) - min(alpha)))\n",
    "\n",
    "# alpha_xantho<-data.frame(prop = seq(0,1,  0.01)) %>% \n",
    "#   mutate(alpha = lm_xantho$coefficients[[2]]*prop+lm_xantho$coefficients[[1]],\n",
    "#          ID = c(1:nrow(.)),\n",
    "#          xmin = seq(0,1,  0.01),\n",
    "#          xmax = seq(0.01,1.01,0.01),\n",
    "#          ymin = rep(0,nrow(.)),\n",
    "#          ymax = rep(1,nrow(.))) %>% \n",
    "#   pivot_longer(c(xmin,xmax),names_to = \"name_x\", values_to = \"x\") %>% \n",
    "#   pivot_longer(c(ymax,ymin), names_to = \"name_y\",values_to = \"y\") %>% \n",
    "#   mutate(y = rep(c(0,1,1,0),nrow(.)/4)) %>% \n",
    "#   dplyr::select(-c(name_x,name_y)) %>% \n",
    "#   mutate(alpha = case_when(alpha > 1 ~ 1,\n",
    "#                            TRUE ~ alpha),\n",
    "#          alpha = 1-alpha) %>% \n",
    "#   mutate(alpha = (alpha - min(alpha))/ (max(alpha) - min(alpha)))\n",
    "\n",
    "#################\n",
    "\n",
    "sediment<-output_80_filtered %>% \n",
    "  dplyr::filter(True == \"Bare Sediment\") %>% \n",
    "  mutate(Class = case_when(Class != True ~ \"Other\",\n",
    "                            TRUE ~ Class)) %>% \n",
    "  group_by(ID_80,Class) %>% \n",
    "  reframe(Prob_80 = unique(Prob_80),\n",
    "          True = unique(True),\n",
    "          Class = unique(Class),\n",
    "          Proportion = sum(Proportion)) %>% \n",
    "  mutate(Class = as.factor(Class),\n",
    "         Class = forcats::fct_relevel(Class,\"Bare Sediment\", after = Inf)) %>% \n",
    "  ggplot()+\n",
    "  geom_density(mapping = aes(x = Proportion , y = after_stat(ndensity) , fill = Class),alpha = 0.8,show.legend = F,position = \"fill\",adjust = 3)+\n",
    "  # facet_wrap(~ True)+\n",
    "  scale_fill_manual(values=c(\"gray85\",\"#70543e\"))+\n",
    "  # scale_alpha_manual(values = c(0.5,1))+\n",
    "  theme_Bede()+\n",
    "  # geom_polygon(data = alpha_sed, mapping = aes(x=x, y = y, alpha = alpha, group = ID),alpha = alpha_sed$alpha,color=NA, fill = \"white\")+\n",
    "  geom_vline(xintercept = (lm_sediment$coefficients[[2]]*85+lm_sediment$coefficients[[1]])*100, linetype = \"dashed\",linewidth = 2)+\n",
    "  ggtitle(\"Bare Sediment\")+\n",
    "  # annotate(geom=\"text\",x = lm_sediment$coefficients[[2]]*0.85+lm_sediment$coefficients[[1]]+0.05, y = 0.85, label = \"p = 0.85\", angle = - 90, size = 7)+\n",
    "  theme(axis.text.x = element_text(size = 15),\n",
    "        axis.text.y = element_text(size = 15)\n",
    "        # axis.title.y = element_blank(),\n",
    "        # axis.title.x = element_text(size = 20)\n",
    "        )\n",
    "\n",
    "green<-output_80_filtered %>% \n",
    "  dplyr::filter(True == \"Green Algae\") %>% \n",
    "  mutate(Class = case_when(Class != True ~ \"Other\",\n",
    "                            TRUE ~ Class)) %>% \n",
    "  group_by(ID_80,Class) %>% \n",
    "  reframe(Prob_80 = unique(Prob_80),\n",
    "          True = unique(True),\n",
    "          Class = unique(Class),\n",
    "          Proportion = sum(Proportion)) %>%\n",
    "  mutate(Class = as.factor(Class),\n",
    "         Class = forcats::fct_relevel(Class,\"Green Algae\", after = Inf)) %>% \n",
    "  ggplot()+\n",
    "  geom_density(mapping = aes(x = Proportion , y = after_stat(ndensity) , fill = Class),alpha = 0.8,show.legend = F,position = \"fill\",adjust = 3)+\n",
    "  # facet_wrap(~ True)+\n",
    "  scale_fill_manual(values=c(\"gray85\", \"#b3ff1a\"))+\n",
    "  # scale_alpha_manual(values = c(0.5,1))+\n",
    "  theme_Bede()+\n",
    "  # geom_polygon(data = alpha_green, mapping = aes(x=x, y = y, alpha = alpha, group = ID),alpha = alpha_green$alpha,color=NA, fill = \"white\")+\n",
    "  geom_vline(xintercept = (lm_green$coefficients[[2]]*85+lm_green$coefficients[[1]])*100, linetype = \"dashed\",linewidth = 2)+\n",
    "  ggtitle(\"Green Algae\")+\n",
    "  # annotate(geom=\"text\",x = lm_green$coefficients[[2]]*0.85+lm_green$coefficients[[1]]+0.05, y = 0.85, label = \"p = 0.85\", angle = - 90, size = 7)+\n",
    "  theme(axis.text.x = element_text(size = 15),\n",
    "        axis.text.y = element_text(size = 15)\n",
    "        # axis.title.y = element_blank(),\n",
    "        # axis.title.x = element_text(size = 20)\n",
    "        )\n",
    "\n",
    "magno<-output_80_filtered %>% \n",
    "  dplyr::filter(True == \"Magnoliopsida\") %>% \n",
    "  mutate(Class = case_when(Class != True ~ \"Other\",\n",
    "                            TRUE ~ Class)) %>% \n",
    "  group_by(ID_80,Class) %>% \n",
    "  reframe(Prob_80 = unique(Prob_80),\n",
    "          True = unique(True),\n",
    "          Class = unique(Class),\n",
    "          Proportion = sum(Proportion)) %>%\n",
    "  mutate(Class = as.factor(Class),\n",
    "         Class = forcats::fct_relevel(Class,\"Magnoliopsida\", after = Inf)) %>%\n",
    "  ggplot()+\n",
    "  geom_density(mapping = aes(x = Proportion , y = after_stat(ndensity) , fill = Class),alpha = 0.8,show.legend = F, position = \"fill\",adjust = 3)+\n",
    "  # facet_wrap(~ True)+\n",
    "  scale_fill_manual(values=c(\"gray85\",\"#389350\"))+\n",
    "  # scale_alpha_manual(values = c(0.5,1))+\n",
    "  theme_Bede()+\n",
    "  # geom_polygon(data = alpha_magno, mapping = aes(x=x, y = y, alpha = alpha, group = ID),alpha = alpha_magno$alpha,color=NA, fill = \"white\")+\n",
    "  geom_vline(xintercept = (lm_magno$coefficients[[2]]*85+lm_magno$coefficients[[1]])*100, linetype = \"dashed\",linewidth = 2)+\n",
    "  ggtitle(\"Magnoliopsida\")+\n",
    "  # annotate(geom=\"text\",x = lm_magno$coefficients[[2]]*0.85+lm_magno$coefficients[[1]]+0.05, y = 0.85, label = \"p = 0.85\", angle = - 90, size = 7)+\n",
    "  theme(axis.text.x = element_text(size = 15),\n",
    "        axis.text.y = element_text(size = 15)\n",
    "        # axis.title.y = element_blank(),\n",
    "        # axis.title.x = element_text(size = 20)\n",
    "        )\n",
    "\n",
    "\n",
    "MPB<-output_80_filtered %>% \n",
    "  dplyr::filter(True == \"Microphytobenthos\") %>% \n",
    "  mutate(Class = case_when(Class != True ~ \"Other\",\n",
    "                            TRUE ~ Class)) %>% \n",
    "  group_by(ID_80,Class) %>% \n",
    "  reframe(Prob_80 = unique(Prob_80),\n",
    "          True = unique(True),\n",
    "          Class = unique(Class),\n",
    "          Proportion = sum(Proportion)) %>%\n",
    "  mutate(Class = as.factor(Class),\n",
    "         Class = forcats::fct_relevel(Class,\"Microphytobenthos\", after = Inf)) %>%\n",
    "  ggplot()+\n",
    "  geom_density(mapping = aes(x = Proportion , y = after_stat(ndensity) , fill = Class),alpha = 0.8,show.legend = F,position = \"fill\",adjust = 3)+\n",
    "  # facet_wrap(~ True)+\n",
    "  scale_fill_manual(values=c(\"gray85\",\"#DAA520\"))+\n",
    "  # scale_alpha_manual(values = c(0.5,1))+\n",
    "  theme_Bede()+\n",
    "  # geom_polygon(data = alpha_MPB, mapping = aes(x=x, y = y, alpha = alpha, group = ID),alpha = alpha_MPB$alpha,color=NA, fill = \"white\")+\n",
    "  geom_vline(xintercept = (lm_MPB$coefficients[[2]]*85+lm_MPB$coefficients[[1]])*100, linetype = \"dashed\",linewidth = 2)+\n",
    "  ggtitle(\"Microphytobenthos\")+\n",
    "  # annotate(geom=\"text\",x = lm_MPB$coefficients[[2]]*0.85+lm_MPB$coefficients[[1]]+0.05, y = 0.85, label = \"p = 0.85\", angle = - 90, size = 7)+\n",
    "  theme(axis.text.x = element_text(size = 15),\n",
    "        axis.text.y = element_text(size = 15)\n",
    "        # axis.title.y = element_blank(),\n",
    "        # axis.title.x = element_text(size = 20)\n",
    "        )\n",
    "\n",
    "rodo<-output_80_filtered %>% \n",
    "  dplyr::filter(True == \"Rodophyceae\") %>% \n",
    "  mutate(Class = case_when(Class != True ~ \"Others Classes\",\n",
    "                           Class == \"Rodophyceae\" ~ \"Rhodophyceae\",\n",
    "                            TRUE ~ Class),\n",
    "         True = case_when(True == \"Rodophyceae\" ~ \"Rhodophyceae\",\n",
    "                          TRUE ~ True)) %>% \n",
    "  group_by(ID_80,Class) %>% \n",
    "  reframe(Prob_80 = unique(Prob_80),\n",
    "          True = unique(True),\n",
    "          Class = unique(Class),\n",
    "          Proportion = sum(Proportion)) %>%\n",
    "  mutate(Class = factor(Class, c(\"Others Classes\",\"Rhodophyceae\",\"Bacillariophyceae\", \"Magnoliopsida\",\"Chlorophyceae\",\"Phaeophyceae\"))) %>% \n",
    "         # Class = forcats::fct_relevel(Class,\"Rodophyceae\", after = Inf)) %>%\n",
    "  ggplot()+\n",
    "  geom_density(mapping = aes(x = Proportion , y = after_stat(ndensity) , fill = Class),alpha = 0.8,position = \"fill\",adjust = 3)+\n",
    "  # facet_wrap(~ True)+\n",
    "  scale_fill_manual(\"Class\",\n",
    "                    labels = c(\"Others Classes\",\"Rhodophyceae\",\"Bacillariophyceae\", \"Magnoliopsida\",\"Chlorophyceae\",\"Phaeophyceae\"),\n",
    "                    values=c(\"gray85\",\"#b3002d\",\"#DAA520\",\"#389350\",\"#b3ff1a\",\"#70543e\"),\n",
    "                    drop = F)+\n",
    "  # scale_alpha_manual(values = c(0.5,1))+\n",
    "  theme_Bede()+\n",
    "  # geom_polygon(data = alpha_Rodo, mapping = aes(x=x, y = y, alpha = alpha, group = ID),alpha = alpha_Rodo$alpha,color=NA, fill = \"white\")+\n",
    "  geom_vline(xintercept = (lm_Rodo$coefficients[[2]]*85+lm_Rodo$coefficients[[1]])*100, linetype = \"dashed\",linewidth = 2)+\n",
    "  ggtitle(\"Rhodophyceae\")+\n",
    "  # annotate(geom=\"text\",x = lm_Rodo$coefficients[[2]]*0.85+lm_Rodo$coefficients[[1]]+0.05, y = 0.85, label = \"p = 0.85\", angle = - 90, size = 7)+\n",
    "  theme(axis.text.x = element_text(size = 15),\n",
    "        axis.text.y = element_text(size = 15),\n",
    "        # axis.title.y = element_blank(),\n",
    "        # axis.title.x = element_text(size = 20),\n",
    "        legend.text = element_text(size = 15),\n",
    "        legend.key.size = unit(1,\"cm\"),\n",
    "        legend.title = element_blank())\n",
    "\n",
    "# xantho<-output_80_filtered %>% \n",
    "#   dplyr::filter(True == \"Xanthophyceae\") %>% \n",
    "#   ggplot()+\n",
    "#   geom_density(mapping = aes(x = Proportion , y = after_stat(ndensity) , fill = Class),position = \"fill\",adjust = 3)+\n",
    "#   # facet_wrap(~ True)+\n",
    "#   scale_fill_manual(values=c(\"#70543e\", \"#b3ff1a\", \"#389350\", \"#DAA520\",\"#b3002d\", \"#cccc00\"))+\n",
    "#   # scale_alpha_manual(values = c(0.5,1))+\n",
    "#   theme_Bede()+\n",
    "#   geom_polygon(data = alpha_xantho, mapping = aes(x=x, y = y, alpha = alpha, group = ID),alpha = alpha_xantho$alpha,color=NA, fill = \"white\")+\n",
    "#   geom_vline(xintercept = lm_xantho$coefficients[[2]]*0.85+lm_xantho$coefficients[[1]])+\n",
    "#   ggtitle(\"Xanthophyceae\")+\n",
    "#   annotate(geom=\"text\",x = lm_xantho$coefficients[[2]]*0.85+lm_xantho$coefficients[[1]]+0.05, y = 0.85, label = \"p = 0.85\", angle = - 90, size = 7)+\n",
    "#   theme(axis.text.x = element_text(size = 15, angle = 45, hjust = 0.7),\n",
    "#         axis.text.y = element_text(size = 15),\n",
    "#         axis.title.y = element_blank(),\n",
    "#         axis.title.x = element_text(size = 20))\n",
    "\n",
    "plot<-(sediment + green) /\n",
    "(MPB + rodo)+\n",
    "magno+\n",
    "  patchwork::plot_layout(guides = \"collect\") &\n",
    "  theme(legend.position = \"bottom\",\n",
    "        plot.title = element_text(size=25, hjust = 0.55),\n",
    "        axis.text.x = element_text(size = 15, hjust = 0.8),\n",
    "        axis.title = element_text(size=20),\n",
    "        axis.text.y = element_text(size = 15),\n",
    "        panel.grid=element_blank())&\n",
    "  scale_x_continuous(breaks = c(25,50,75,100),limits= c(0,100.01),expand = c(0, 0)) &\n",
    "  scale_y_continuous(expand = c(0, 0)) &\n",
    "  xlab(\"Percent Cover (%)\") & \n",
    "  ylab(\"Density\")\n",
    "  \n",
    "ggsave(\"Figures/Upscaling/density_vs_Proportion.png\",plot , width = 833*4, height = 1118*4, units = \"px\")"
   ],
   "id": "db7cee0f-fce7-4d66-89e6-3bd9b97e6672"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_80_filtered_same<-read.csv(\"./Output/output_80_filtered_same_Upscaling.csv\")%>% \n",
    "    dplyr::mutate(Proportion = Proportion *100) %>% \n",
    "    dplyr::select(-X) %>% \n",
    "    as_tibble()\n",
    "\n",
    "################# LM ####################\n",
    "lm_sediment<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Bare Sediment\") \n",
    "\n",
    " lm_sediment<- lm(lm_sediment$Prob_80 ~ lm_sediment$Proportion)\n",
    " \n",
    " lm_green<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Green Algae\") \n",
    "  lm_green<- lm(lm_green$Prob_80 ~ lm_green$Proportion)\n",
    " \n",
    " lm_magno<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Magnoliopsida\") \n",
    "   lm_magno<- lm(lm_magno$Prob_80 ~ lm_magno$Proportion)\n",
    "\n",
    " \n",
    " lm_MPB<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Microphytobenthos\")\n",
    "    lm_MPB<- lm(lm_MPB$Prob_80 ~ lm_MPB$Proportion)\n",
    "    \n",
    "lm_Rodo<-output_80_filtered_same %>%\n",
    "  dplyr::filter(Class == \"Rodophyceae\")\n",
    "    lm_Rodo<- lm(lm_Rodo$Prob_80 ~ lm_Rodo$Proportion)"
   ],
   "id": "14697a4e-841a-4a76-b52b-fb425dc4227f"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "knitr::include_graphics(\"Figures/Low_res/Upscaling/density_vs_Proportion.png\")"
   ],
   "id": "cell-fig-upscaling"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the very high precision low altitude flight, we determined the minimal percent cover required to classify correctly classify a given class using the high altitude flight (<a href=\"#fig-upscaling\" class=\"quarto-xref\">Figure 10</a>). When the percent cover of the class is 100 %, big pixels are well classified for all the classes excepted for Bare Sediment, where it’s well classified 80% of the time. A vegetative percent cover of at least 80% is need to have all the big pixels well classified, at the exception of Magnoliopsida that needs an higher percent cover (\\>90 %) to be well classified. Concerning the probability of each class, a really high Percent cover is needed to confidently predict Bacillariophyceae. To predict Chlorophyceae with a confidence of 0.85, a percent cover of 93 % is needed, 90 % for magnoliopsida, 92 % for Rhodophyceae and 97 % for Bacillariophyceae.\n",
    "\n",
    "# 4. Discussion\n",
    "\n",
    "## 4.1 Vegetation Discrimination"
   ],
   "id": "4a3f8101-020b-4caf-82a4-826ab655f909"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pigment<-\n",
    "  data.frame(\n",
    "    Classes=c(\"Magnoliopsida\",\n",
    "           \"Chlorophyceae\",\n",
    "           \"Bacillariophyceae\",\n",
    "           \"Phaeophyceae\",\n",
    "           \"Rhodophyceae\"),\n",
    "    chlb=c(1,1,0,0,0),\n",
    "    chlc=c(0,0,1,1,0),\n",
    "    fuco=c(0,0,1,1,0),\n",
    "    zea=c(1,1,0,1,0),\n",
    "    diad=c(0,0,1,0,0),\n",
    "    lutein=c(1,1,0,0,0),\n",
    "    Neox=c(1,1,0,0,0),\n",
    "    PE=c(0,0,0,0,1),\n",
    "    PC=c(0,0,0,0,1))%>%\n",
    "  select(-Classes) %>% \n",
    "  as.matrix() %>% \n",
    "  as.array() %>% \n",
    "  rast()\n",
    "\n",
    "ext(df_pigment)<-c(0,9,0,5)\n",
    "\n",
    "\n",
    "y_axis_text_pigment<-data.frame(x = rep(-0.05,5),\n",
    "                                y = c(0.5,1.5,2.5,3.5,4.5), \n",
    "                                lab = c(\"Rhodophyceae\",\n",
    "                                    \"Phaeophyceae\",\n",
    "                                    \"Bacillariophy.\",\n",
    "                                    \"Chlorophyceae\",\n",
    "                                    \"Magnoliopsida\")\n",
    "                                )\n",
    "x_axis_text_pigment<-data.frame(y = rep(5.5,9),\n",
    "                                x = c(0.5,1.5,2.5,3.5,4.5,5.5,6.5,7.5,8.5),\n",
    "                                lab = c(\"Chlb\",\"Chlc\",\"Fuco\",\"Zea\",\"Diad\",\"Lut\",\"Neo\",\"PE\",\"PC\")\n",
    "                                )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "min_x_line = -1.5\n",
    "max_y_line = 6\n",
    "\n",
    "line_grid<-data.frame(xmin=NA,\n",
    "                      xmax=NA,\n",
    "                      ymin=NA,\n",
    "                      ymax=NA,\n",
    "                      group=NA)\n",
    "\n",
    "for(i in 0:max_y_line){\n",
    "  line_grid[i+1,1]=min_x_line\n",
    "  line_grid[i+1,2]=9\n",
    "  line_grid[i+1,3]=i\n",
    "  line_grid[i+1,4]=i\n",
    "  line_grid[i+1,5]=i+1\n",
    "}\n",
    "nrowgrid=nrow(line_grid)\n",
    "for(i in 0:9){\n",
    "  line_grid[i+1+nrowgrid,1]=i\n",
    "  line_grid[i+1+nrowgrid,2]=i\n",
    "  line_grid[i+1+nrowgrid,3]=0\n",
    "  line_grid[i+1+nrowgrid,4]=max_y_line\n",
    "  line_grid[i+1+nrowgrid,5]=i+1+nrowgrid\n",
    "  \n",
    "}\n",
    "\n",
    " line_grid[nrow(line_grid)+1,1] = min_x_line\n",
    " line_grid[nrow(line_grid),2] = min_x_line\n",
    " line_grid[nrow(line_grid),3] = 0\n",
    " line_grid[nrow(line_grid),4] = max_y_line\n",
    " line_grid[nrow(line_grid),5] = nrow(line_grid)\n",
    "\n",
    " \n",
    " line_grid<- line_grid %>% \n",
    "   rbind(c(-1.5,9,-0.6,-0.6,19))%>% \n",
    "   rbind(c(-1.5,-1.5,-0.6,0,20))%>% \n",
    "   rbind(c(0,0,-0.6,0,21))%>% \n",
    "   rbind(c(1,1,-0.6,0,22))%>% \n",
    "   rbind(c(2,2,-0.6,0,23))%>% \n",
    "   rbind(c(3,3,-0.6,0,24))%>% \n",
    "   rbind(c(4,4,-0.6,0,25))%>% \n",
    "   rbind(c(5,5,-0.6,0,26))%>% \n",
    "   rbind(c(6,6,-0.6,0,27))%>% \n",
    "   rbind(c(7,7,-0.6,0,28))%>% \n",
    "   rbind(c(8,8,-0.6,0,29))%>% \n",
    "   rbind(c(9,9,-0.6,0,30))\n",
    " \n",
    " \n",
    " Absorption_wv<- data.frame(x = c(0.5,1.5,2.5,3.5,4.5,5.5,6.5,7.5,8.5),\n",
    "                            y = rep(-0.3,9),\n",
    "                            lab = c(\"650\",\"636\",\"550\",\"489\",\"496\",\"490\",\"450\",\"566\",\"615\"))\n",
    " \n",
    " \n",
    "rast_pigment<-ggplot()+\n",
    "  geom_spatraster(data = (df_pigment),\n",
    "                  # show.legend = F,\n",
    "                  na.rm = TRUE\n",
    "                  )+\n",
    "  coord_fixed(ratio = 0.7)+\n",
    "  scale_fill_gradientn(\n",
    "    # colours = c(\"black\",\"firebrick4\",\"goldenrod\",\"yellow\",\"chartreuse3\"),\n",
    "    # colours = c(\"khaki1\",\"goldenrod\",\"firebrick3\",\"chartreuse4\"),\n",
    "    colours = c(\"firebrick4\",\"orange\",\"#007d00\"),\n",
    "    # colours = c(\"#7d0000\",\"#00b300\",\"#00b300\",\"#00b300\",\"#007d00\"),\n",
    "    na.value = \"white\",\n",
    "    labels = c(\"Absence\",\"no se\",\"Presence\"),\n",
    "    breaks = c(0,0.5,1))+\n",
    "  geom_text(\n",
    "    data = y_axis_text_pigment,\n",
    "    aes(x = x, y = y, label = lab),\n",
    "    # fill = \"#ffffff99\",\n",
    "    hjust = 1,\n",
    "    size = 6,\n",
    "    # angle = 20\n",
    "    )+\n",
    "  geom_text(\n",
    "    data = data.frame(x = -0.05, y = -0.3, lab = \"Absorption (nm)\"),\n",
    "    aes(x = x, y = y, label = lab),\n",
    "    # fill = \"#ffffff99\",\n",
    "    hjust = 1,\n",
    "    size = 5,\n",
    "    # angle = 20\n",
    "    )+\n",
    "  geom_text(\n",
    "    data = Absorption_wv,\n",
    "    aes(x = x, y = y, label = lab),\n",
    "    # fill = \"#ffffff99\",\n",
    "    hjust = 0.5,\n",
    "    size = 6,\n",
    "    # angle = 20\n",
    "    )+\n",
    "  geom_text(\n",
    "    data = x_axis_text_pigment,\n",
    "    aes(x = x, y = y, label = lab),\n",
    "    # fill = \"#ffffff99\",\n",
    "    hjust = 0.5,\n",
    "    size = 7,\n",
    "    # angle = 20\n",
    "    )+\n",
    "  geom_segment(data = line_grid, \n",
    "               aes(x = xmin, xend = xmax, y = ymin, yend = ymax, group = group),\n",
    "               linewidth = 1)+\n",
    "  xlim(c(-2,9))+\n",
    "  ylim(c(-1,6.1))+\n",
    "  theme_Bede()+\n",
    "  theme(\n",
    "  # axis.title.x = element_text(size = 35),\n",
    "        # axis.title.y = element_text(size = 35),\n",
    "        axis.title.y = element_blank(),\n",
    "        axis.title.x = element_blank(),\n",
    "\n",
    "        # axis.text.x = element_text(size = 30),\n",
    "        axis.text.x = element_blank(),\n",
    "        # axis.text.y = element_text(size = 25),\n",
    "         axis.text.y = element_blank(),\n",
    "\n",
    "        axis.ticks.x = element_blank(),\n",
    "        axis.ticks.y = element_blank(),\n",
    "        panel.border = element_blank(),\n",
    "        axis.line=element_blank(),\n",
    "        # legend.position = \"top\",\n",
    "        legend.position = \"none\",\n",
    "        legend.direction=\"horizontal\",\n",
    "        legend.key.width = unit(2,\"cm\"),\n",
    "        legend.text = element_text(size = 20),\n",
    "        legend.title = element_text(size = 20, vjust = 2),\n",
    "        legend.title.align = 0.5,\n",
    "        legend.justification = \"center\")\n",
    "\n",
    "# rast_pigment\n",
    "\n",
    "ggsave(\"./Figures/Disc_Pigment_Table.png\",rast_pigment, width= 1066*4, height=516*4, units = \"px\")"
   ],
   "id": "5863d015-af29-4afd-aa16-36e7b4844afe"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "# Pigment composition of the different vegetation classes. Green indicate the presence of the pigment in the class while red indicate the abscence. Chlb : Chlorophyll-b ; Chlc : Chlorophyll-c ; Fuco : Fucoxanthin ; Zea : Zeaxanthin ; Diad : Diadinoxanthin ; Lut : Lutein ; Neo : Neoxanthin ; PE : Phycoerythrin ; PC : Phycocyanin.\n",
    "\n",
    "knitr::include_graphics(\"./Figures/Low_res/Disc_Pigment_Table.png\")"
   ],
   "id": "cell-fig-Pigm"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The primary objective of this study was to develop a method for the accurate classification of macrophytes on intertidal mudflats, specifically focusing on distinguishing between Chlorophyceae (green algae) and marine Magnoliopsida (seagrasses) using multispectral drone data. The ability to differentiate between various types of vegetation, as demonstrated in this study, plays a critical role in ecological monitoring and management. By distinguishing between seagrasses and macroalgae, our approach facilitates targeted conservation strategies, enabling more effective preservation and restoration efforts in coastal ecosystems. The discrimination of seagrasses from green macroalgae presents significant challenges, primarily due to the shared pigment composition between these taxa, as depicted in <a href=\"#fig-Pigm\" class=\"quarto-xref\">Figure 11</a> but also by the frequent spatial mixing of these green macrophytes. This similarity in pigment composition, including photosynthetic and carotenoid pigments like chlorophyll b and c, fucoxanthin, and neoxanthin, results in spectral signatures that are often indistinguishable to non exper eyes. This spectral overlap is further complicated by the intermingling of seagrasses and green macroalgae in the same spatial locations, posing a significant challenge for remote sensing aimed at accurately mapping and monitoring coastal ecosystems. Our study addresses these complexities by utilizing high-resolution drone imagery with 10 spectral bands. Davies et al. ([2023a](#ref-Davies2023)) has shown that having at least 8 spectral bands ranging between 500 nm to 850 nm with including a band at 530 nm and another one at 730 nm is crucial to accurately discriminate green macrophytes.\n",
    "\n",
    "``` r\n",
    "knitr::include_graphics(\"./Figures/Low_res/Validation/ConfusionMatrixGreen.png\")\n",
    "```\n",
    "\n",
    "<figure id=\"fig-ValidationGreen\">\n",
    "<img src=\"attachment:./Figures/Low_res/Validation/ConfusionMatrixGreen.png\" />\n",
    "<figcaption>Figure 12: Sample of <a href=\"#fig-Validation\" class=\"quarto-xref\">Figure 8</a> focusing on green macrophytes. The labels inside the matrice indicate the number of pixels.</figcaption>\n",
    "</figure>\n",
    "\n",
    "The Micasense RedEdge-MX DUAL camera used in this study meets these two criteria, enabling the classifier to achieve 97% of accuracy between these two classes (<a href=\"#fig-ValidationGreen\" class=\"quarto-xref\">Figure 12</a>). Even if the pigment compostion of green macrophytes is similar, differences in the spectral shape can still be observed (<a href=\"#fig-vegetation\" class=\"quarto-xref\">Figure 2</a>). Several factors can explain these differences such has different pigment concentration or proportion, different cellular organisation and 3D disposition of the plant a whole on the intertidal mudflat Hedley et al. ([2018](#ref-hedley2018influence)).\n",
    "\n",
    "The VIP analysis of the Neural Network model (<a href=\"#fig-VIP\" class=\"quarto-xref\">Figure 9</a>) shows that the 531 nm band is the most important spectral band for accurately identifying Chlorophyceae. In fact, at this wavelength, Chlorophyceae exhibits the highest reflectance among all other classes, highlighting the difference in accessories carotenoid proportion between seagrasses and green algae ([Repolho et al., 2017](#ref-repolho2017seagrass)).\n",
    "\n",
    "Concerning Phaeophyceae, the thick cell walls of plant of this class make it really reflective in the infrared part of the spectra whereas the presence of Fucoxanthin and Zeaxanthin result in a low reflectance in the visible part of the electromagnetic spectra. These two key features has been identified by the Neural Network as the two principal predictors to accurately identify Brown algae (<a href=\"#fig-VIP\" class=\"quarto-xref\">Figure 9</a>). Similarly, the presence of phycoerythrin and phycocyanin in Rhodophyceae contributes to the lowest reflectance among all classes in the spectral range of 560 to 615 nm (<a href=\"#fig-VIP\" class=\"quarto-xref\">Figure 9</a>). The band at 560 nm has been identified as important for accurately identifying this class Resolution impact on the prediction. Regarding Bacillariophyceae, the VIP analysis (<a href=\"#fig-VIP\" class=\"quarto-xref\">Figure 9</a>) indicates that 475 nm is the most important predictor for this class. This could be attributed to the absence of lutein, neoxanthin, and zeaxanthin in diatoms, absorbing in the blue part of spectra, resulting in the highest visible reflectance among all classes. Furethemore, it is vegetation with the lowest concentration of chlorphyll-a, pigment absorbing light both in the blue and the red. The transparency of Bacillariophyceae makes the reflectance of the sediment part of the overall reflectance of Bacillariophyceae, further explaining the high reflectance in the blue.\n",
    "\n",
    "## 4.2 Spectral Spatial Temporal Resolution impact on the prediction"
   ],
   "attachments": {
    "./Figures/Low_res/Validation/ConfusionMatrixGreen.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAA3YAAAJ4CAYAAAAp0HFJAAAABGdBTUEAALGPC/xhBQAACklpQ0NQ\nc1JHQiBJRUM2MTk2Ni0yLjEAAEiJnVN3WJP3Fj7f92UPVkLY8LGXbIEAIiOsCMgQWaIQkgBhhBAS\nQMWFiApWFBURnEhVxILVCkidiOKgKLhnQYqIWotVXDjuH9yntX167+3t+9f7vOec5/zOec8PgBES\nJpHmomoAOVKFPDrYH49PSMTJvYACFUjgBCAQ5svCZwXFAADwA3l4fnSwP/wBr28AAgBw1S4kEsfh\n/4O6UCZXACCRAOAiEucLAZBSAMguVMgUAMgYALBTs2QKAJQAAGx5fEIiAKoNAOz0ST4FANipk9wX\nANiiHKkIAI0BAJkoRyQCQLsAYFWBUiwCwMIAoKxAIi4EwK4BgFm2MkcCgL0FAHaOWJAPQGAAgJlC\nLMwAIDgCAEMeE80DIEwDoDDSv+CpX3CFuEgBAMDLlc2XS9IzFLiV0Bp38vDg4iHiwmyxQmEXKRBm\nCeQinJebIxNI5wNMzgwAABr50cH+OD+Q5+bk4eZm52zv9MWi/mvwbyI+IfHf/ryMAgQAEE7P79pf\n5eXWA3DHAbB1v2upWwDaVgBo3/ldM9sJoFoK0Hr5i3k4/EAenqFQyDwdHAoLC+0lYqG9MOOLPv8z\n4W/gi372/EAe/tt68ABxmkCZrcCjg/1xYW52rlKO58sEQjFu9+cj/seFf/2OKdHiNLFcLBWK8ViJ\nuFAiTcd5uVKRRCHJleIS6X8y8R+W/QmTdw0ArIZPwE62B7XLbMB+7gECiw5Y0nYAQH7zLYwaC5EA\nEGc0Mnn3AACTv/mPQCsBAM2XpOMAALzoGFyolBdMxggAAESggSqwQQcMwRSswA6cwR28wBcCYQZE\nQAwkwDwQQgbkgBwKoRiWQRlUwDrYBLWwAxqgEZrhELTBMTgN5+ASXIHrcBcGYBiewhi8hgkEQcgI\nE2EhOogRYo7YIs4IF5mOBCJhSDSSgKQg6YgUUSLFyHKkAqlCapFdSCPyLXIUOY1cQPqQ28ggMor8\nirxHMZSBslED1AJ1QLmoHxqKxqBz0XQ0D12AlqJr0Rq0Hj2AtqKn0UvodXQAfYqOY4DRMQ5mjNlh\nXIyHRWCJWBomxxZj5Vg1Vo81Yx1YN3YVG8CeYe8IJAKLgBPsCF6EEMJsgpCQR1hMWEOoJewjtBK6\nCFcJg4Qxwicik6hPtCV6EvnEeGI6sZBYRqwm7iEeIZ4lXicOE1+TSCQOyZLkTgohJZAySQtJa0jb\nSC2kU6Q+0hBpnEwm65Btyd7kCLKArCCXkbeQD5BPkvvJw+S3FDrFiOJMCaIkUqSUEko1ZT/lBKWf\nMkKZoKpRzame1AiqiDqfWkltoHZQL1OHqRM0dZolzZsWQ8ukLaPV0JppZ2n3aC/pdLoJ3YMeRZfQ\nl9Jr6Afp5+mD9HcMDYYNg8dIYigZaxl7GacYtxkvmUymBdOXmchUMNcyG5lnmA+Yb1VYKvYqfBWR\nyhKVOpVWlX6V56pUVXNVP9V5qgtUq1UPq15WfaZGVbNQ46kJ1Bar1akdVbupNq7OUndSj1DPUV+j\nvl/9gvpjDbKGhUaghkijVGO3xhmNIRbGMmXxWELWclYD6yxrmE1iW7L57Ex2Bfsbdi97TFNDc6pm\nrGaRZp3mcc0BDsax4PA52ZxKziHODc57LQMtPy2x1mqtZq1+rTfaetq+2mLtcu0W7eva73VwnUCd\nLJ31Om0693UJuja6UbqFutt1z+o+02PreekJ9cr1Dund0Uf1bfSj9Rfq79bv0R83MDQINpAZbDE4\nY/DMkGPoa5hpuNHwhOGoEctoupHEaKPRSaMnuCbuh2fjNXgXPmasbxxirDTeZdxrPGFiaTLbpMSk\nxeS+Kc2Ua5pmutG003TMzMgs3KzYrMnsjjnVnGueYb7ZvNv8jYWlRZzFSos2i8eW2pZ8ywWWTZb3\nrJhWPlZ5VvVW16xJ1lzrLOtt1ldsUBtXmwybOpvLtqitm63Edptt3xTiFI8p0in1U27aMez87Ars\nmuwG7Tn2YfYl9m32zx3MHBId1jt0O3xydHXMdmxwvOuk4TTDqcSpw+lXZxtnoXOd8zUXpkuQyxKX\ndpcXU22niqdun3rLleUa7rrStdP1o5u7m9yt2W3U3cw9xX2r+00umxvJXcM970H08PdY4nHM452n\nm6fC85DnL152Xlle+70eT7OcJp7WMG3I28Rb4L3Le2A6Pj1l+s7pAz7GPgKfep+Hvqa+It89viN+\n1n6Zfgf8nvs7+sv9j/i/4XnyFvFOBWABwQHlAb2BGoGzA2sDHwSZBKUHNQWNBbsGLww+FUIMCQ1Z\nH3KTb8AX8hv5YzPcZyya0RXKCJ0VWhv6MMwmTB7WEY6GzwjfEH5vpvlM6cy2CIjgR2yIuB9pGZkX\n+X0UKSoyqi7qUbRTdHF09yzWrORZ+2e9jvGPqYy5O9tqtnJ2Z6xqbFJsY+ybuIC4qriBeIf4RfGX\nEnQTJAntieTE2MQ9ieNzAudsmjOc5JpUlnRjruXcorkX5unOy553PFk1WZB8OIWYEpeyP+WDIEJQ\nLxhP5aduTR0T8oSbhU9FvqKNolGxt7hKPJLmnVaV9jjdO31D+miGT0Z1xjMJT1IreZEZkrkj801W\nRNberM/ZcdktOZSclJyjUg1plrQr1zC3KLdPZisrkw3keeZtyhuTh8r35CP5c/PbFWyFTNGjtFKu\nUA4WTC+oK3hbGFt4uEi9SFrUM99m/ur5IwuCFny9kLBQuLCz2Lh4WfHgIr9FuxYji1MXdy4xXVK6\nZHhp8NJ9y2jLspb9UOJYUlXyannc8o5Sg9KlpUMrglc0lamUycturvRauWMVYZVkVe9ql9VbVn8q\nF5VfrHCsqK74sEa45uJXTl/VfPV5bdra3kq3yu3rSOuk626s91m/r0q9akHV0IbwDa0b8Y3lG19t\nSt50oXpq9Y7NtM3KzQM1YTXtW8y2rNvyoTaj9nqdf13LVv2tq7e+2Sba1r/dd3vzDoMdFTve75Ts\nvLUreFdrvUV99W7S7oLdjxpiG7q/5n7duEd3T8Wej3ulewf2Re/ranRvbNyvv7+yCW1SNo0eSDpw\n5ZuAb9qb7Zp3tXBaKg7CQeXBJ9+mfHvjUOihzsPcw83fmX+39QjrSHkr0jq/dawto22gPaG97+iM\no50dXh1Hvrf/fu8x42N1xzWPV56gnSg98fnkgpPjp2Snnp1OPz3Umdx590z8mWtdUV29Z0PPnj8X\ndO5Mt1/3yfPe549d8Lxw9CL3Ytslt0utPa49R35w/eFIr1tv62X3y+1XPK509E3rO9Hv03/6asDV\nc9f41y5dn3m978bsG7duJt0cuCW69fh29u0XdwruTNxdeo94r/y+2v3qB/oP6n+0/rFlwG3g+GDA\nYM/DWQ/vDgmHnv6U/9OH4dJHzEfVI0YjjY+dHx8bDRq98mTOk+GnsqcTz8p+Vv9563Or59/94vtL\nz1j82PAL+YvPv655qfNy76uprzrHI8cfvM55PfGm/K3O233vuO+638e9H5ko/ED+UPPR+mPHp9BP\n9z7nfP78L/eE8/stRzjPAAAAIGNIUk0AAHomAACAhAAA+gAAAIDoAAB1MAAA6mAAADqYAAAXcJy6\nUTwAAAAJcEhZcwAALiMAAC4jAXilP3YAANFvSURBVHic7N13fM1n/8fx1znZiQghxN57r1btPWvW\nVtTqotqq6p5KB1qKKkVL1ai9apTaWnvWHhUrSMjeOef3R+6cnyQnU+JIvJ+PRx538v1e3+t8Tm40\n71zLYDabfYBiQCAiIiIiIiKpMwDmbPD65v+1Tav0ts/s5zPyWh7ANXsg9/9ueDyiAkRERERERCTz\n5La3dQUiIiIiIpJtZXSE6sHRtsTPmxNdNyf6PKVnnkQGAKOtqxARERERkWwrowHLkOh/E98zJPra\nbOWeOdH/Gqy0sfZ1jqRgJyIiIiIimSW1AJXSSF3i+8n1a06mn8RBz1r75IJktqdgJyIiIiIij8qD\nIcxaiEspwCXXT/zXD47YmbE+Umct4D3KDU+yjIKdiIiIiIjYQmphytq0ywefs3YvrX1bG9XL1hTs\nRERERETkUUvvCFni4w3iR/OSG5Wztv4u8bq7HEXBTkREREREHicPhrLk1uTFB7XEI23WRvmsjexl\n+xG6RMwKdiIiIiIi8jhJbndLa2vu0jKdM/7ZxO1T2y0z8b3HeZTPoHPsRESyyIoVK5g+fTre3t6Y\nzWZMJhMAnp6ePP300/Tr1w8nJ6d09Xnx4kUmTZrEyZMnadq0KePHj8dgeJz/OyMiIpJprE25TC24\npbQbptlKO2v3k/v6sWIwm80BgIetCxERyWk++OADJkyYkOz9UaNGMXXq1HT12ahRI/bu3QtAqVKl\nuHDhAnZ2dg9Vp4iISDaU3DRLa7tdWguAPND+wXbW7j3uDECgRuxERLKIg4MDAFWqVGHAgAEEBgaS\nK1cu9u3bx4YNG/j+++/p3LkzLVu2TFN/p0+fZu/evTg7O3Pu3DkKFiyI0agZ9SIikq1ZC2jWwlri\nIwnSM00y8fEH1l7D2uhdSqEupR05bUI/EYiIZLEGDRrwzjvvMGHCBN5//33Wr19Pw4YNAdiyZYul\nXUBAAK+99hq5c+emcuXKrFixwnLv1q1bfPzxxwDkzp2b9evXc/v2bQwGAyaTiYkTJ1KoUCGKFi2a\nZBTw7NmzDB8+nMOHD7NlyxYKFSrEpEmTAAgPD+ftt98mT548lC9fnsWLFyd49tSpUwwbNoyLFy9y\n6tQpSpUqRenSpfn222+TvM8jR47QpUsX3NzcqFu3LitXrkzSZtWqVVStWpXcuXPz2muv4e/vn+D+\nlStXGDhwIHnz5qVQoUK89NJLXL9+PT3fbhERyV5S2q0y8dl08RKvtUvua4OVe5D+kbjkpnE+NqEO\nALPZHGAWEZFM98knn5gB85AhQxJcDw8PNxcvXtwMmGfOnGk2m81mPz8/c5EiRcyA2cnJyWw0Gs2A\n+f333zebzWbz0qVL4/8DYvmYOHGi2Ww2m1u2bGkGzA4ODmYHBwczYO7cubPl9X7++WczYK5YsaK5\ncOHCZsD82muvmc1ms7lChQqW17SzszMD5tdff93y7KxZs8yAuWrVquYCBQokeP13333X0m7FihWW\n6y4uLmaDwWAGzCNGjLC0ef/9982A2WAwmJ2cnMyAuXDhwub79++bzWaz2dfX1+zh4WEGzHnz5jU7\nOztb+jx9+nTm/R8jIiI5hel/H9aumZK5Zu0jLW0e5w+z2WwO0IidiEgWW7FiBTVr1qRmzZrUr1+f\n4sWL4+PjQ7FixRg0aBAAn3zyCTdu3OCjjz4iIiKC2NhYypQpw4QJE7h16xa9evXi559/BqBQoULs\n37+fMWPGsHDhQrZt20bnzp2JiooiKiqK3r17s3btWtavXw+Aq6srEDdyV6NGDd544w3ef/99vv/+\ne86dO8ebb75JREQEMTEx1KhRg6lTp3L69GkAnJ2dgbiRu99++w2z2czkyZMB2Lp1q+U9vvPOOwD8\n+uuvhIWFcfDgQQDLCOD169eZMGECefLkISgoiIiICH744Qdu3rxpWYe4aNEiAgMDGTRoEPfu3SM8\nPJxZs2ZRs2ZN7t69m3X/B4mIyOMotY1LEp9hl/hcO2tTO5PbbZNUrmcLCnYiIlksMDCQ48ePc/z4\ncfbv309QUBDjxo3j6NGjuLi4APDXX38BULlyZQIDA4mKiqJx48YArF69GoAaNWoA4ObmxlNPPQXA\npk2bAKhVqxYRERGEh4dTu3ZtANatWwdg2TXTzc2NhQsX8t133+Ht7c1vv/0GQLVq1QgODiY6OppG\njRoBsGHDhgTv4emnn6ZVq1YA1K1bF4B79+4BcYHx4sWL5MuXj169egFQvXp1bt++zc2bNwH4448/\ngLhpqWazmbt371KuXDkgbnpm/HsAmD9/Pk8//TRffvklzZo14+jRozRp0iSD330REclmktvFMi1f\nWzvE/MH7yR1antz6vWwV8hTsRESyWOfOnTGbzcydOxeAyMhI6tWrR758+QCIiYkhJCQEgP79+5Mn\nTx6cnJz45ZdfALh//z4AZnPcf4eio6MtfcevURs3bhwuLi64urpaRs8CAgKA/w92FStWxNPT0/Js\nYGAgAMOGDSN37tw4OjoyY8aMBK8Zz8vLy/J5/ChefD3xo2klS5bE0dERiNs4pkCBApbjHO7cuQPA\nxo0b8fDwoECBArRu3dry/QBo1qwZy5cvx9PTkwMHDvD+++9ToUIFihYtyvnz59PyrRYRkezvYcNU\nSufUPTialzjkJV7ykJYNWzKj3szqQ8FORCSrFSxYEIAhQ4Ywffp0ANq1a2eZKmlvb2+ZLrl69Wqi\noqK4cOEC/v7+BAUFMWrUKOD/g9SD8ubNC8CECRMwm82cPXuW27dvExwcbAmS8c89GOoA3N3dgbgp\nkLGxsZw/fx4/Pz+CgoJ4//33E7SN3+ETSHJuXoECBQA4d+4coaGhlutLly611FC8eHEAunTpgslk\n4syZM9y4cYOQkBBOnz5tmX7apUsX7ty5w7///su0adOoW7cuN27cSFKPiIg88ZIbmUuuTVpZ2y3T\n2rTOh3mNxDJlExYFOxGRLBY/IgUwYsQIOnfuDMArr7xCeHg4AE2bNgVgz549ODg4ULZsWT799FOe\nf/55y8ibNfHTNXfs2AFAhQoVWLBgAW3atLGsk4sPdokDWfz0xt27d2M0GilXrhxfffUVvXr1soyw\nxYs/XN1afxUqVKBixYqEhISwYMECAA4dOkSfPn0YNmwY0dHRtG3bFoBt27bh7+9PxYoVOXz4MI0a\nNWLdunU4Ozszfvx4HBwcGDp0KJUrV2bkyJH069cPiBvVFBEReYC1kTRru1+m5fnE11J6LvH0zseG\ngp2ISBaJnzL5YCgCmDZtGq6urly/fp3+/fsD8NVXX1GwYEG++eYbjEYjRqORadOmcfLkScuoXHxA\nfLC/V155hUaNGrFp0yYMBgNGo5G3336bf/75xzINMj48xsbGJqjjo48+omTJksyYMcPympMmTeLY\nsWN4eHgAEBUVleC1ASIiIpLU8dVXXwHw6quvYjQaqVevHhA3RdTBwYGCBQsyZswYgoODyZ8/P0aj\nkc6dO3Ps2DHLNNS+ffvi4uLC/PnzLfWMHj0agDFjxqT32y8iIjlPaiNbqQUta9MuE2+08mBf1tbt\nWesvudd9pJuxKNiJiGSRsmXLAlCpUqUE14sXL86sWbMoWLAgZ86c4fLly+TNm5d///2XMWPGUKRI\nEcqVK8d3333H+fPncXNzAyB//vyUKFGC+vXrJ+hv+/btfPfdd1SsWJEiRYrw9ttvc+fOHctmK8WL\nF8fNzc3ydbw8efJw7NgxPvjgA4oVK0bp0qX56quvuHLlimX9X7FixXB2drZsyAJx6+2KFi2aoI4u\nXbpw4sQJevfuTb58+WjatCkbN27kww8/tLSZOHEiq1evpmHDhnh6etK7d29OnTrF8OHDAShXrhw3\nb97k448/ply5cnh7ezNo0CDOnDlj2dRFRESeaOkJSdbCWHK7Ylpr+2D4s/a8tdey1u8jO+vOYDab\nAwCPR/WCIiLy/2JjYzGbzdjb29u6FBERkezO2hq4tD6HlWeT21UzNclNDU3cJrNCnwEI1E8SIiI2\nZGdnZ+sSREREsjtrUyoTj5wlF/gSh7rEXxsSXTOTfCiLv56WwJbpI3maiikiIiIiItmZtTPt0jra\nZm2NnbUQmNL5eNauP3IasXtM7du2jcO7dpEn0fbkIiKZxRwTS4yTE4VL5MHNEWJibfrfIxHJoQxA\naEgQ1eq1pHipCrYuR3K2B0fmkjvQPLkz6pIb1Usp7KUlQGbmlMsUKdg9po7u3MmJyd9htrcDQ3Ij\nxyIiGWcODCKgfFXe/q4t9b2j8QuMTf0hEZF0cnYwEOJi4sq1Ugp2ktVSmm6ZOPAlN60ypY1QUnqt\n5ALcI/utqYLdY8rRyQmzvR1mOwU7EckaJjt7MBqJjTUTFW0mOkYjdiKS+YwGiI4xY++kHzvFZqxt\nivLgUQbWzsSLb2ft+cTtH+nul8nR37DHmcHw/x8iIpnM8L9/X2JNZmJNEGtK/RkRkfTSvy/yGLC2\nIQpWPk/pGWvtrYXCB6+lZ1fMh56yqWAnIiIiIiI5WXJr59IT6pKbspk4yKU1oCVu89Ajfgp2IiIi\nIiLyJLG2gYq1dXjWAp61HTETH4eQnLSM5GWYjjsQEREREZEnSUpTMpPbJCXxWXnJnZmX0udZuhZP\nwU5ERERERJ40KZ1rFx/Akgt+1q4nPsjc2oYtaZHhzTUU7ERERERE5EmU0qHjiadfpnSUQXIbraR2\nREJqNaWL1tiJiIiIiMiTJi2Hkyf3XHLHJaR2OLm10JZpB5hrxE5ERERERJ4UyYWr1NomDm/WNkxJ\naRdNa6+RaaEOFOxEREREROTJkZ41bMltsmJtDV3i+4nX2yW3w2ZaNl5JEwU7ERERERGRtO9aaSZh\n29R21Eytr/RcT5aCnYiIiIiISMojcam1e3CKZnKHjyc3cpe4XXKvmSIFOxERERERkeSlZfQscfCy\nFuISb6SS3Lq79IY+QMFOREREREQkJcmNziUObKmtv0urDG2oomAnIiIiIiKSNikdb/Bgm/h2aZl6\naW2TlnRTsBMREREREUmZtVG65AJZ4jZmrE+9fHB0L7WwmCoFOxERERERkZSldOZcamEscdBLfORB\n4rPxIAOjd/bpfUBEREREROQJFB/uEoez5DZBwUr75KZfJh7Rs3Z0QooBUiN2IiIiIiIi6ZfS0QYP\nfp7edskFuJQCpIKdiIiIiIhIBljbATO5KZRpnVqZXLvkAp8lFCrYiYiIiIiIpE1aAlpKxx9Y23Al\npamd1tpZDZQKdiIiIiIiIilLaYpkYtbOvXtwumVym6c8+HVyu2QmW4eCnYiIiIiISMpSO3jc2s6W\niZ9NLbBZ22SFZK4loWAnIiIiIiKSPtZ2rXzwXnIB78HPE0/ZTMtrJdtOwU5ERERERCR9kjt+wNoa\nuuRG8x4cyUvcNrm1dNbagTZPERERERERyZDkzp1LHNasnVmXOOwlt+Nlauv6LAFPB5SLiIiIiIhk\nHmsbpCR3QHlynycX6JINehqxExEREREReTjJBbXkpllaa5tS36mej6dgJyIiIiIikjUeDGXp2SQl\nvv2D9xL3leAZTcUUERERERHJXKlNvyTRPWv3rU3pfFCCtXwasRMREREREck8iQNZ4nuJDyu3tqlK\ncmv04u8l2VhFI3YiIiIiIiKZJ3GoS236ZeIpmslNu3zwfhIasRMREREREcl6ad08JdWNUqy10Yid\niIiIiIhI1kvrSF6Ka+mSaaNgJyIiIiIiksVSW3cHKex4aeXZJMFOUzFFRERERESyVkrr7NJzL37T\nlCTPKNiJiIiIiIjYRnK7YCa+lvh6kjaaiikiIiIiIvJ4SHz+nbUQZ62NRuxERERERERsJKVDzOND\nXeLjEKx+rmAnIiIiIiJiG8lNuyTRdWtn1yXYXEXBTkRERERExDasBTprIc7akQcPtjVrjZ2IiIiI\niMjj48EpmNZYHeXTiJ2IiIiIiMjjw1qoe/Ba/Jq7BCN4CnYiIiIiIiKPD2sjdSmddQco2ImIiIiI\niGQ3SUb1FOxEREREREQeL9Y2SnlQkhE8BTsREREREZHHS6pTLxNTsBMREREREcmeLCN7CnaZLCoq\nivv37xMbG2vrUkREREREJHtKbSpmfBvLyJ6CXSY4cOAAgwYNomrVqhQsWBAvLy/OnTsHwLhx42jf\nvj1BQUE2rlJERERERLKJtE7FtARAHVD+kAYMGMDChQuTXDca4zLzvn372LRpE4ULF2bHjh3UrVv3\nUZcoIiIiIiI5T4Kz7DRi9xA6d+5sCXUNGjTgiy++oECBAgCYzXHf45o1awIQGhpK27ZtuXPnjk1q\nFRERERGRnEvBLoN++eUX1q1bB8C6devYu3cvH3zwgSXYRUdHA/Dll19y8uRJ8uXLx7179/jqq69s\nVrOIiIiIiOQoWmP3sObOnQvATz/9xLPPPmu5Hj9S96CqVasyb948AP78889HU6CIiIiIiDwxFOwy\n4N69e+zZswdHR0cGDBiQpmc6d+6Mh4cHp06d4ubNm1lcoYiIiIiIPEkU7DIgMDAQgPz58+Pk5JTm\n51xcXIC4IxFERERERETSIcUjEBTsMiB//vwYjUZu3rzJ3bt30/TM7du38fX1xcHBAS8vryyuUERE\nREREcpgEu2AmpmCXAe7u7rRu3RqAWbNmpemZGTNmANC8eXPc3NyyrDYREREREcnRrIY7BbsMGjp0\nKAAfffQR3333neW6wZD0LMGvv/6acePGATBo0KBHU6CIiIiIiOQ0BpI5vFzBLoN69uxJ3759ARg9\nejQVKlTgrbfespxT9/vvvzNu3DjKlCnDu+++C0DHjh3p16+fzWoWEREREZGcyd7WBWRnixYtwsPD\ngx9//JHz58/z7bffWu6NHz8+Qdvu3buzYsWKR12iiIiIiIg8ATRi95BmzpzJsWPHGDZsGNWrV8fT\n0xODwYCbmxsVKlSgT58+7NixQ6FORERERESyjEbsMkGNGjX46aefAAgLCyMgIAA3Nzc8PDxsXJmI\niIiIiDwBDAp2mczV1RVXV1dblyEiIiIiIk8Os4JdJggPDycsLAyTyZRgV0yz2YzJZCI2Npbw8HA2\nb95MYGAg7733ng2rFRERERGRnEbB7iEcPXqUN954g0OHDmE2mzEajZZgZzKZLO2io6OJjo4G4s6x\nU7ATEREREZHMpGCXQefOnaNevXrExsamqb3BYMBsNlO1atUsrkxERERERJ402hUzgyZNmkRsbCwO\nDg588MEHHDx4kNdeew2AXr168c8//7Bo0SJGjBgBxE3LXLlyJd9//70tyxYRERERkRxIwS6D9uzZ\nA8C0adP44osvqFu3Li+//DIAQUFBPP300/Tt25fp06dz9uxZAN544w1blSsiIiIiIjmYgl0G3Lx5\nk7Nnz2JnZ0efPn0s1ytXrkyuXLnYtGkTYWFhlusVKlTgt99+w8fHhy+//NIWJYuIiIiISA6mYJcB\nkZGRABQsWDDJWXXly5cH4OLFiwmu9+zZE8By3p2IiIiIiEhmUbDLgFy5cgEQFRWV5F6tWrUAOH/+\nfILrDg4OVK1alStXruDn55f1RYqIiIiIyBNDwS4DvLy8KFOmDH5+fly5ciXBveLFiwNxRyEkFh8E\nQ0JCsr5IERERERF5YijYZVDLli0B+PDDDxNcr127NgBLlixJcP3IkSOcP38eFxcX8ufP/2iKFBER\nERGRJ4KCXQa99NJLACxatIi8efOyY8cOIO4A8nz58nH58mVat27N9u3bWbRoEV26dAGgVatWlqmc\nIiIiIiIimUHBLoNq167NzJkzAQgICODChQsAuLm58fnnnwOwdetWWrRoQf/+/bl+/ToAb731lm0K\nFhERERGRHEvB7iG8/PLLHDt2jK5du1KnTh3L9VdffZU5c+bg5ORkuebp6cmGDRto2rSpLUoVERER\nEZEczN7WBWR3NWrUYNWqVUmuDx06lP79+7N//34cHR2pV68e9vb6douIiIiISOZT0shCzs7OGqET\nEREREZEsp2CXSaKjozlz5gz//fcfISEhdOzYEQ8PD27duoWrq2uSg8xFREREREQyi9bYPaTQ0FDe\nfvttChQoQI0aNejSpQv9+/fn2rVrALz33nvkyZOHDRs22LhSERERERHJqRTsHsKZM2coXrw4kyZN\nIiAgIME9BwcHAK5evQrAs88+yzfffPOoSxQRERERkSeAgl0GBQUF0bp1a+7du0fBggWZOnUq586d\no3z58gDExMQA8NFHH1muvfPOO+zevdtmNYuIiIiISM6kYJdBX3zxBTdu3KBq1apcvXqVUaNGUb58\neRwdHQGIjY0FoEWLFpw7d4727dsDWM6+ExERERERySwKdhm0Zs0aAH788ccE59WZzWar7b/99lsA\ntm3blvXFiYiIiIjIE0XBLgOuXbvG+fPn8fb2pmHDhml6pmLFinh7e3Pnzh3++++/rC1QRERERESe\nKAp2GRC/fs5o1LdPRERERERsT8kkA0qVKkWhQoW4efMm//77b5qeOX78OL6+vhQpUoSSJUtmbYEi\nIiIiIvJEUbDLoDZt2gBx59SlxdixYwFo2bJlltUkIiIiIiJPJgW7DHrjjTcAWLduHY0bN+bcuXPA\n/59fZ29vD8CJEyd45pln2LJlCwCvvvrqoy9WRERERERyNHtbF5Bd1axZk59++onhw4ezZ88eKlas\nSNGiRfHz8wNg+PDh+Pn5cf78ecszEydO5Omnn7ZVySIiIiIikkMp2D2EYcOGUaZMGZ5//nlu3rzJ\n9evXLff27dtn+dzd3Z158+bRo0cPW5QpIiIiIiI5nILdQ2revDk3btxg165dbN68mStXruDn54eL\niwvVqlWjbdu2NGjQADs7O1uXKiIiIiIiOZSCXSZp0qQJTZo0sXUZIiIiIiLyBNLmKQ/pxIkTVK1a\nlS+//DLJvc2bN2MwGGjfvj1nz561QXUiIiIiIvIkULB7CDNmzKBGjRr8+++/7N69O8n9ixcvArBp\n0yYqVarEkiVLHnWJIiIiIiLyBFCwy6AdO3YwcuRIAIoWLcrzzz+fpE2PHj2YN28etWrVAqBv374c\nO3bsUZYpIiIiIiJPAAW7DJo2bRoQF96uXbtGv379krQpWLAggwcP5siRI/Tt2xeAyZMnP9I6RURE\nREQk51Owy4DIyEg2b94MwHfffZemZ+LX4K1fv56YmJgsq01ERERERJ48CnYZcPPmTUJDQylcuDBF\nixZN0zMlSpTA29ubgICABOfdiYiIiIiIPCwFuwxwc3MDICAggOjo6DQ/FxERAYCTk1OW1CUiIiIi\nIk8mBbsMKFCgAHXr1iUsLIwVK1ak6Zm1a9cSEBBAjRo1KFSoUBZXKCIiIiIiTxIFuwzq1asXAEOH\nDuXgwYMptv33338ZPHgwAM8991yW1yYiIiIiIk8We1sXkF29/fbbzJ8/n3///ZennnqKbt260a5d\nO7y9vcmdOzdBQUHcvn2bP//8k2XLlgFQpUoVPvjgAxtXLiIiIiIiOY2C3UP4888/ad68OefOnWPV\nqlWsWrUq2bYVKlRg69atGI0aJBURERERkcyllPEQChUqxNmzZ5kzZw4NGzbE3d09wX1XV1fq1avH\nrFmzOHv2LN7e3jaqVEREREREcjKN2GWCoUOHMnToUCIjI7lx4wYBAQG4ublRrFgxXF1dbV2eiIiI\niIjkcAp2GWA2m7l8+TKFCxfGxcXFct3JyYnSpUvbsDIREREREXkSaSpmBhgMBvr164ebmxtfffUV\nN27csHVJIiIiIiLyBFOwyyBXV1fMZjPvvfceRYsWpXHjxixYsIDQ0FBblyYiIiIiIk8YBbsM+vXX\nX/niiy8oX748AHv27GHQoEHkyZOH3r178+eff9q4QhEREREReVIo2GVQ0aJF+eCDDzh37hzHjx9n\n9OjRFCxYkJiYGH7//XfatGlD/vz5GT16NCdPnrR1uSIiIiIikoMp2GWC6tWrM3nyZHx9fdm+fTuD\nBg3CxcUFf39/vvvuO6pXr07lypWZNGkSvr6+ti5XRERERERyGAW7TNasWTN++eUXgoODOXjwIN99\n9x2VKlXizJkzvP322xQqVIhmzZqxYsUKzGazrcsVEREREZEcQMEui9jZ2VG3bl1q1aqFp6dngns7\nd+6kR48e5M6dm+nTp9uoQhERERERySkU7DLZ3bt3mT59OvXq1cNgMNCsWTP27t0LQKdOnVi/fj3T\np0+ndOnShISE8Nprr9GlSxcbVy0iIiIiItmZDijPBCEhIaxcuZL58+fz119/JbhXsWJFhg8fzoAB\nA/Dy8rJcHzFiBHPnzmXYsGGsXbuWuXPnMnTo0EdduoiIiIiI5AAKdhkUHR3N+vXrWbBgARs2bCA6\nOtpyz9PTk+eff55hw4ZRrVq1ZPsYOnQoGzduZMWKFaxbt07BTkREREREMkTBLoO6d+/O+vXrLV8b\njUY6derEiy++SPv27TEYDGnqJ379XeJ1eCIiIiIiImmlYJdBAQEBANSuXZthw4bRt29f8uTJk+5+\n6tevT0REBGPHjs3cAkVERERE5ImhYJdBL7zwAp9//jnNmzd/qH6GDBnCkCFDMqkqERERERF5EinY\nZZDWw4mIiIiIyONCwS6TBAUFER4enuLautjYWFxcXDI0ZVNERERERCQ5CnYP4fbt27z33nts2bKF\ngIAAoqKiAJINd1FRUbRu3ZotW7Y8yjJFRERERCSHU7DLoDt37lCrVi1u3bqVrudiYmKyqCIRERER\nEXlSKdhl0Mcff2wJdX369KFbt27kypULk8mE2Wy2+kxoaCglSpR4lGWKiIiIiMgTQMEug7Zu3QrA\n+PHjef/9921cjYiIiIiIPMmMti4gO7py5QqXLl3CYDAwfPhwW5cjIiKSbgZD3IeIiOQMGrHLAGdn\nZwC8vLzw8vKycTWSHZgBN6MRRyDGZCY2NnuttTSbzdjb2WGwtyfUZCLWbEY/D4pkP2azGQMGCnjl\nwtHVAaKiwRQLj+pvtNkMDvaYMXL7djCR0bHYGfWviYhIZlCwy4BChQpRo0YNjh8/zunTp6lcubKt\nS5LHmMFgIK/BwKF7/uyLjcUAFCIu7Flfjfn4MQBhgAfQOk9enBydCIuNUbgTyUbMZjMO9vZ4e+fi\n99V7mbFoP2aDAxVKexMV/Wh+2WQ2GwgLD6NG2Ty8/VIHomOduBcYpnAnIpIJFOwyaOzYsfTv35/x\n48fz22+/2boceUyZgbxGI5vu3Mb8VD1+fHss9nZ2uOdyi7ufzEY7jxt7e3uM9g4sWLCAlXPn0s0j\nD3YODpiySf0iEse7kDsffLWS1UfdeX7IDJwc7PDwcCc2NvaRvL69vT1mgx2//forh0cvYtGkPjg7\n2hMd82heX0QkJ1Owy6B+/fqxfv16Fi1aRGxsLF9++SXFihXD3l7fUvl/zgYDp+7cxlSxInv37iWv\nvYOtS3oozzZpwlD3XPw5ZSqdvAoQrGAnkm14ebqxZsPfbD6Viz1795I3l53NahnavxNdeg5lwg8b\n+eLt7ly/FWSzWkREcgqlkFT4+/vTtGlT/P39cXd3B+LOosufP7/lN5xLly5l6dKl5MmTBy8vLwwG\ng9WRmKtXr9KrVy9+/fXXR/oexHYcDQZOA3169Mj2oS5er569+HPKVCIjIsDR0dbliEgaOedyZPv+\nK7RsN8ymoS5e7969WDlzBzGRERgMccvvREQk4xTsUhETE8OFCxeIiorC19fXcv3KlStJ2gYEBBAQ\nEJBifzdu3MjsEuUxZwacXVxsXUam8fDwwMnBgeiYGAyOjtlmnaCIQEysiQJ5PGxdBgB58uTBaO9I\n9CNa3yciktMp2KUid+7cfPvtt4SFheHk5PRQfQUEBFC1atVMqkyyE7PJlOJ9f39/7t69S7ly5bCz\nS/qb9PDwcK5evUrJkiUtu7Imdv/+fe7cuUPZsmWt9nHnzh0CAgIoX758snVcuXIFJycnChcunGyb\nmJhoTLGxGBycFOpEshmDAaKjolNsc/36dWJiYihZsqTV+1FRUVy5coWiRYvi5uaWbD+RkZGEhobi\n6elp9X50TAwGzBiMRiDlfyNFRCR1CnapcHFxYcSIEbYuQ3KwXr16sWLFCvLmzUtYWBjz58+nZ8+e\nlvuLFi1i8ODBODo6EhkZyaJFi+jRo0eCPt544w2mT5+Oi4sLRqORdevW0aRJE8v9YcOGMXfuXNzd\n3fH09GTt2rVUr17dcj8wMJDWrVtz9OhRYmJi6NWrF0uXLs36Ny8ij42IiAjat2/Prl27sLe3p2rV\nqmzatCnBsT6LFi2if//+5M6dG5PJxJdffsnIkSOt9vfqq6+ybt067ty586jegojIE00HlD+mNBLy\nZOjUqRPLli1j7969+Pn5MWTIEHr16sXu3bsBOHDgAP379+fdd98lODiYUaNG0bNnT44dO2bpY8yY\nMUydOpUNGzYQHBxMmzZtaNq0KZcuXQKgR48ezJ07lz179hAUFET58uWpVasWwcHBlj6aNGnCf//9\nx7179zh27Bi///47w4YNe6TfCxGxrY4dO7Jjxw4uXrzIrVu3uHbtGg0bNrTc37RpE/379+fll18m\nMDCQH374gddee42ff/45SV/bt29n3rx5eHg8HtM+RUSeBAp2mSQmJobTp0+zfv16Fi9ezJYtW6yu\nw0srO4NBC8lzuLt377J+/Xrmz59P/fr1AZg+fTplypRh6tSpAHz00UdUqVKFzz77DIBJkyZRoUIF\nPvjgAyBu+uWUKVP47bffaNu2LQC//vorw4YNIygobpe5K1euMHnyZMsPaHPmzMFkMvHXX38BsHPn\nTk6cOMH27dtxd3enRo0azJs3j7lz53L//v1H9w0REZu5f/8+f/31FwsWLKBUqVJ4enqydOlSLly4\nwIULFwB45513qFWrFjNnzgRgwIABvPXWW7zzzjsJ+goJCWH06NEUKlTIsumYiIhkPQW7h3Tv3j3L\nf8CqVKlCp06d6NevH23btqV06dKUKlWKiRMnpvuMIDNg1IGtOZqbmxt79uyhW7duCa4HBwdbpj7t\n27ePjh07JrjfqlUrjh49CsT9Bt3V1ZWePXuyZMkSBgwYwOHDh/npp5+oVasWAIcOHWL06NGW51es\nWAFAjRo1AFi/fj1ubm5UqVIlwWvEv76I5Hx58+alWLFiln8fANatW4eDgwPlypUD4OLFi1SqVCnB\ncy1btuTu3bv4+PhYrrVr146nn36aGTNmcO3atUfzBkRERGvsHsaOHTto3bo1MTHJ7+j133//MXbs\nWKZPn86ePXsoVqyY1XaHDx/m66+/xtPTE4OrK+FnzuBqtCNSkzJzLFdX1wTTnADGjRvHnTt3+PDD\nD4G433wn3nigTJky3Lp1C4hbG+fk5ETPnj05e/YsuXPnplGjRowZM4aJEycCYDDE/YLgl19+YfDg\nwQDMnTvXsjHCtWvXLG3ixW+ecu7cuSTBUkRypgMHDlChQgVcXFzIlSsXfn5+HD582HK/QoUKnDt3\nLsEz8SP/8VO7p0yZwsGDB9mzZw8zZ84kOjrljVpERCTzaMQug86dO0erVq2IiYkhb968fP755+ze\nvZsLFy5w8+ZNTp8+zbZt2xgzZgzOzs74+PjQvHlzwsLCrPYXGBjI5cuXOXniBPsPH+b+vXsYNGD3\nRPn+++/5+OOP+fHHHylSpIjlByKjMeFfU3v7//99jLu7O35+fhQuXJizZ89y4MABVqxYwaRJk9i+\nfXuC58qWLcuXX35Jt27dGDlyJEeOHAHAbDYn2UUz/jWjoqIy/X2KyOPpk08+IW/evIwfP55PPvmE\nsmXL8tVXX1nuT5kyhcOHD9O9e3fOnDnDxx9/zIwZMwDw9PTEbDbz2WefsWHDBgDCwsJw1FmXIiKP\njIJdBn399dfExsZSs2ZNbty4wUcffUSjRo0oW7YshQoVolKlSrRo0YKJEydy6dIlChcuzKVLl/ji\niy+s9teiRQsOHTrE3n37OLJzJx07dyY6ndM3Jft6/fXXef3115k7dy4vvfQSAA4OcQeaR0ZGJmgb\nGBho2WI8JCQEgM8//9xyv3v37hQsWJBp06YleK5Ro0a8++67rFy5kuLFizNo0CAg7ly6xKPO8b+A\nKFCgQGa9RRF5jC1dupTZs2fzxx9/MHr0aEaOHMn+/ftZtmwZ48ePB+I2Wdq9ezc7duygTp063Lhx\ng5kzZ+Ll5UXu3Llp3rw5lStXpnnz5vj5+XH16lXCw8M5fPgwERERNn6HIiI5n4JdBphMJtauXQvA\n4sWLcUnl8OnChQvz66+/ArB8+fI0vYaDwUisSdMwnwTdu3fn+++/56+//mLIkCEJ7lWuXDnBDpgQ\nt+6tePHiABQpUgQgwQ6XAHZ2dpZNC6ytc2nQoAFnzpwBoGnTpoSFhSUYTT558iQAtWvXfsh3JyLZ\nQfwGKQ+uofP09CRv3rwcOHAAiDtvs06dOty7d4+wsDDmzp2Lj48Pbm5uuLm54ePjw759+3BycsLL\ny4tp06YREhJC3bp1WbNmjU3el4jIk0TBLgOuXr2Kv78/VapUoWLFiml6pkWLFnh7e3PhwgX++++/\nVNtHm02aivkEGDx4MKtWreLu3bs0b948yf0uXbqwYsUKy8jc/fv32bhxI/379wfg2Wefxd7eno8/\n/tjyzJYtW7h58yZvvfUWACNHjuT999+33L937x4///wzr776KgCtW7fGbDbz448/Wtp8++23eHl5\nUbNmzUx/zyLy+In/9+fBzVOOHz/O/fv36d69OwCfffYZBQsWtNyPjY3liy++4NlnnwVg165dnDx5\nksOHDxMUFMTrr7+Om5sbx48fT7JJlIiIZD4FuwyIP5cnNDQ0Xc/FT3dzdnZOta0yXc63e/dufvnl\nFwoXLkz//v0pU6YM5cuXx8HBgXHjxgEwYcIE6tWrR+7cuenTpw+enp40btzYctwBxO2MuXDhQqpW\nrUrXrl1p27YtI0aMsBxAvmbNGhYuXEj16tXp168f+fLlo27dunz//fdA3HTLH374gbfeeotWrVpR\ns2ZNli1bpt+wizxBGjZsyPvvv0/Pnj1p0aIFXbt2pWbNmvTp08cybfv999/HbDZToEABBg4caDnE\nPH7ad9GiRalatSo1atTA3d2dIkWK4ODgQPXq1bXWTkTkEVCwywBPT0/atGnDf//9x5YtW9L0zPbt\n2/Hz86Np06Z4e3tncYXyuDCb4/6SJd4ABeLW0L388sv06NEDNzc36tatS61atahfv75liiXE7VQ3\nadIkrl27xrfffsuuXbsS9NOyZUuuX79OgwYNCAoKYu3atUyfPt1yv3Pnzvz333/UrVuXW7du8csv\nv3Dw4MEEfbzyyiv8/fffODo6UqlSJS5fvswzzzyTud8MEbE5A4Yku+DGGz9+PHv37iVPnjxERkay\nbt06Fi9ebLnv7e2Nr68vgwYN4vLly8ycOTPBrpmJ1a9fnzFjxmT6exAREet03EEGzZw5kzJlytC2\nbVsOHjxI3bp1k21748YNOnXqBMCsWbMeVYnyGDAaIBKItLLld/369S0Hk6dm9OjRCc6iS6xIkSLM\nnj072fslSpRg3rx5Kb5G/fr1+eOPP1Ktxc7ODoPRiFlHcYhkOzGxphTvN2jQgJUrVyZ7383NzXKU\nSmoaN25M48aNk71vb2eHGQNms/4tERHJDAp2GVS6dGk++eQTPvvsM+rVq0f37t3p06cPpUqVwt7e\nHrPZzM2bN9mwYQM///wzERERdOjQgdu3b7Nz584EvzENDAykc+fOlC9f3obvSLKCwWAgGpI95iI7\nioiIJCYmBjtnF7Rvq0j2YjQYiIyKTL3hIxARGUFsTDRGoxYfiIhkBgW7DGrevDk7duywfL1y5coU\nf8sJ8McffyQ7IpI/f34Fuxwo0mymEnDif7vK5QR/bd9OFODq7EKQOeXf/ovI4yM2IoYaFQry255d\nwNu2LofNm7eSzy0SJzc3zAGBti5HRCTbU7DLIE9PTwCKFy9uGX2LiYlJcB6YnZ1dgsOkTSaT1Skn\nN27cwNXVNYsrFluIMJl4Kr8Xy3bsoE2/fkz79ltcHB2z1ZlOdnZ2cWfpGQys2bSJme+/T1snZ8Jt\nXZiIpIuvXyjD+jRj64GFNGzVm2++/gIXJ4ckZ2VmFaPRjujoSEwmM4t+X8vBLbNZ930X7t/XvyYi\nIplBwS6Dli1blqnrAqxtriHZnxkIMxjp5pmPPxYvpvHyZRTIlZtYU/YZ6bI3GgmKiiIgNAQnoHUu\ndwq4uREQE6Pdl0SykdhYE0FhsfzydW8+mryGj15ugUeevBh4NP8eOTjYc9c/iNt+96lZ2oWVUzrh\nXbgg12/ex95O/5qIiDwsBbsMUhCTtDAAUaZYYo1GOhcoSEhICKGBgdnqz4/ZbMZgNOLg6kZuFxci\n7ewU6kSyIaPRQEBwOC5O9kz6pC9hQUEEB4dj9whDlckEjo4O5MnvQcD9CIU6EZFMpGCXiW7dusWF\nCxeIjIzE3d2dsmXLkj9/fluXJTZmAExmM4GxsRhcXHB1cbF1SRkWDBAbq1Ankk3ZGQ1ERcdy/XoA\ndnYGDEYXYh7hppQGI0TGwI3rgZhBoU5EJBMp2GWCadOmMWPGDM6dO5fgutFopHz58gwZMoTRo0dj\nZ2dnowrlcWH+34eIiK3FxupfJBGRnES/KnsI/v7+VKtWjVGjRiUJdRC3WcrZs2cZO3YslSpV4vbt\n2zaoUkREREREcjqN2D2Edu3acerUKQC6d+/Oc889R+nSpXFzcyMwMJDLly+zcuVK1qxZw4ULF3j2\n2Wc5ePCgjasWEREREZGcRsEug6ZMmcKhQ4cwGAz89ddfNGvWLEmbRo0aMXDgQHbu3EnLli05dOgQ\nM2bMYMSIEY++YBERERERybE0FTODFi1aBMCcOXOshroHNW3alDlz5iR4TkREREREJLMo2GWAn58f\nBw8exMnJiX79+qXpmb59++Lo6Mi+ffu4d+9eFlcoIiIiIiJPEgW7DAgODgYgX758ODs7p+kZJycn\nPD09AQgKCsqy2kRERERE5MmjYJcBXl5eGI1Gbt68ib+/f5qeuXfvHr6+vkBcIBQREREREcksCnYZ\nkCtXLlq3bg3Ajz/+mKZn4tu1atUKd3f3LKtNRERERESePAp2GTR48GAAPvzwQ6ZMmZJi26lTp/LB\nBx8A8MILL2RxZSIiIiIi8qTRcQcZ1Lt3b5YuXcqqVat48803mT59Ol27dqVChQo4OzsTHh7OuXPn\nWLNmDZcuXQKgc+fO9O/f38aVi4iIiIhITqNg9xBWrlxJnz59WLp0KZcuXWLy5MnJtu3VqxdLly59\nhNWJiIiIiMiTQlMxH9KSJUvYvXs3AwYMoHz58uTKlcvyUb58eQYMGMCuXbsU6kREREREJMtoxC6D\nbty4gZ2dHd7e3jRq1IhGjRoBEBoaSlRUFI6Ojri5udm4ShEREREReRJoxC6DRo4cSaFChfj2228T\nXHdzcyNv3rwKdSIiIiIi8sgo2GWAv78/W7duBeDOnTs2rkZERERERJ50CnYZEB4eTkhICADVqlWz\ncTUiIiIiIvKkU7DLgKJFizJgwAAAfvvtNxtXIyIiIiIiTzptnpJBCxYsoGDBgkyaNImmTZsybtw4\nypUrR548ebC3t8dkMiV5xmw2Y2dnh4ODgw0qFhERERGRnErBLoO++eYbTp06hbOzM7t27aJp06YA\n5M2bFxcXF6vBztfXl+eee47ly5c/6nJFRERERCQHU7DLoI0bN7Jjx44k1+/fv8/9+/eTfe7evXtZ\nWJWIiIiIiDyJFOwy6N1336Vbt27kzp07zc8EBQVRvnz5LKxKRERERESeRAp2GdS2bVvatm1r6zJE\nRERERES0K6aIiIiIiEh2pxG7THDv3j0WLlzInj17uHz5MpGRkeTOnZvSpUtTt25d+vTpQ8GCBW1d\npoiIiIiI5FAKdg/phx9+YMSIEVbv7du3j4ULF/LGG28wadIk3nrrrUdcnYiIiIiIPAkU7B7CRx99\nxBdffAGAo6MjdevWpWrVqri7u3P//n1OnTrF0aNHiY6OZsyYMfj7+zNhwgQbVy0iIiIiIjmNgl0G\n7dixwxLqevbsyZw5c6zukBkQEMCLL77IsmXL+PLLL2nbtq3lzDsREREREZHMoGCXQTNnzgRg8ODB\nzJs3L9l2efLk4ffff2fYsGHMnTuX2bNnK9iJiIiIZDKz2cyxY8cIDQ3FwcEhyX2TyYSTkxP58+en\ncOHC2Nvrx+CMiIiIsMxIq1WrFu7u7rYuSf5Hf6IzIDY2li1btgDwySefpOmZjz/+mLlz57J582bM\nZjMGgyErSxQRERF5okRFRfHss89y8+bNVNt6eHjQqFEjPvvsM+rUqfMIqst+IiIimDx5Mn369KFM\nmTKW61evXqVBgwYAbNu2jRYtWtiqRElExx1kgI+PDwEBARQuXJgSJUqk6ZnixYvj7e2Nv78/V69e\nzeIKRURERJ4sRqMRozFtP9oGBgayYcMG6taty9y5c7O4suzn2rVreHt78+GHH+Lq6prgXmxsrOXz\nqKioR12apEAjdhng6OgIpP8Pc0xMDICG/kVERESyQPyMqPilMpcvX8ZsNlvuxQe/VatW8c477xAd\nHc2wYcOoV68e1atXt1ndj5tjx44RGBiI0WhMMsuscOHCTJw4kfDwcKpVq2ajCsUaJYwMKFKkCJUr\nV+b06dNs3ryZtm3bpvrMn3/+iZ+fH5UqVaJo0aKPoEoRERGRJ1P8+cGlS5e2ev/NN9+kZMmSdO/e\nHYBZs2YxY8aMR1bf4+7BQYjEwS5PnjyMGTPmUZckaaCpmBnUsWNHAF555RXu37+fYtv79+/z8ssv\nA9CuXbssr01ERETkSRYZGZlqm27dulG3bl0A9u/fn2LbmzdvsnHjRnx9fZNtExQUxK5du9i0aRNX\nrlxJX8HE/by4detWdu3aRUhISJqfM5lMnDx5kj/++IPDhw+ne0bZvXv32LhxY4KaH5xuGT/imRE+\nPj5s3ryZHTt2pPrzsjX+/v5s3bqV7du3ExgYmOE6nhQKdhn0zjvvkD9/fq5cuUKpUqWYOHEiZ8+e\nJTQ0lOjoaEJDQzlz5gwTJ06kVKlSXL58mbx58zJ27Fhbly4iIiIiQO3atYG4NWWhoaGW68899xwG\ng4EjR46wd+9eihQpQocOHShUqBDvvPNOgj7OnDlD+/bt8fDwoGnTprRv357SpUtTvHhxFi5caPV1\n169fj8FgYODAgQC89NJLeHp60rp1a5o2bYq7uztDhw4lKCgo2dqjo6P5+OOPyZMnD9WrV6djx47U\nrVuXXLlyMWLEiGSfHTVqFAaDgQ0bNnDt2jW8vLzo0KEDZcuWpUOHDpQrV45OnToBcaGxZMmSGAwG\npk2bBsCVK1fw9vbG3t6eAwcOWH2NZcuWUbZsWUqUKEG7du1o3rw5np6etGjRghMnTlh95rXXXsNg\nMLBy5UoA2rRpQ/78+WndujUtWrQgT5489O/fn4CAgGS/J086BbsMypcvH6tXr8ZoNBIYGMjYsWOp\nVKkShQsXpnjx4hQuXJjKlSszduxYAgMDMRgMrFmzBm9vb1uXLiIiIiLEBTqI+7nOzc3Ncv3WrVsA\nzJ071xK+4j04mrVo0SIqV67Mpk2bsLOzo1GjRnTp0oUCBQpw7do1BgwYQK9evZK8bnBwMBAXknr0\n6MHs2bMpVKgQL774Iq1atQJg3rx5VK9eHT8/vyTP3717l0qVKjFu3DiCg4MpV64cPXr0oFq1akRH\nR/PDDz9QsmRJzp07l+TZ27dvA7B27Vp69uyJyWQC4kJcWFgYTk5OCdpHR0cD/z9yFxUVxe3bt4mN\njbW8jwfFv+dLly6RL18+OnXqRNOmTbG3t2f79u3UqFGDn3/+Odm6duzYQdOmTfnzzz8pUKAArVu3\nplChQpbvd6NGjSz7VkhCCnYPoWHDhly9epWePXtargUFBeHr65vgtyQ9evTAx8eHxo0b26JMERER\nEUnk6tWrbN26FSDJxinxa8x++OEHwsLCWLVqFadOneLzzz/nhRdeAODgwYP0798fgMaNG3P37l12\n797N6tWruX37NuPGjQPiRq/efvvtBP3b2dkBsGfPHlasWMGbb77JzZs3mTVrFn/++ScHDx7E3d2d\nq1evWl7vQd26dePSpUs4OzuzYcMGzp8/z7Jlyzhx4gT79+/Hy8uL+/fv07FjxyRTM+Pf2+zZszl5\n8iRz587l4sWLfPHFF4wbN45Tp06xYsUKIG593ZUrVzCbzYwYMcJyLV7i8wI/+eQTyyjlRx99hJ+f\nH2vXrmXHjh3cu3ePli1bAjBkyBB2795t9Xsybdo0Dh48yLp167h9+zZbtmzh5s2blu/hv//+yw8/\n/JCh6a45nTZPeUhFixbl999/5/79++zbt4+rV68SGxuL0WikRIkSNGzYkLx589q6TBEREZEnhouL\nS4r37927R69evSyjUYlH5R4ML8uWLaNRo0YAVKlSxXL9888/B6BixYps377dEkziffjhh4SHhzNh\nwgQmTZrEyy+/nOA8uHjdu3fn22+/TXCtbt26/Prrr3Tt2pUNGzawa9cumjRpAsCvv/7K3r17AVi+\nfDkdOnRI8OxTTz3Fpk2bqFOnDpcuXeK7775LMn003ty5c+nTpw8AH3zwgeV6/BEHBoPBMoKX+P0l\n5uPjw1dffQXA6NGjLd+feO7u7mzdupWaNWty/PhxPvvsM0uwTsza+/rmm2/YsWMHBw8eZOfOneTL\nl49SpUqlWNOTRsEuk+TNm9eyoYqIiIiI2M7q1asJCwtLsB7LYDBgMBi4evUqO3futEzn69evX5IQ\nEa9evXqWUPega9eusX79egBef/31ZEPP6NGjmTZtGsHBwaxYscLqXgtvvPGG1We7dOlCzZo1OXbs\nGGvWrLEEu/g1aM2aNUv2Z8/atWszcOBAFixYwO+//2412BUqVIiuXbtafT4+8ELaN09Zv349UVFR\nODg4JPueAMaMGcOAAQPYtm0b586do0KFCgnuV6pUKdn/P6pWrcrBgwcJDg5Ocf3hk0rBTkRERERy\nlNOnT3P69OlU273++utMmTIlyfX4MFO+fHmrzx08eNDyecOGDZPtP1++fNSpU4cdO3ZY3WjE29ub\n+vXrJ/t87dq1OXbsGCdPnrTUFf/a1gLngxo2bMiCBQs4efIkN2/epHDhwgnulylTBmdn5xT7SI+/\n//4bgDp16lCsWLFk2z3zzDOWzw8dOpQk2BUvXjzZZz08PIC4dX46HD0pBbtU3L9/n65du3L//n1y\n5cr1UH1du3aNLl26MH369EyqTkREREQSK1++PE2aNEkwqhMbG4ujoyPFihWjYsWKtGvXzrIpR3KK\nFCli9frdu3cB8PLyomzZsin2UblyZXbs2GF1TVjp0qWTrFN7UMmSJQG4dOkSEHeMQ/xmKpUrV07x\ndStVqgTEjb5dv349SbBL7r1l1J07d4C4qakpKVWqFMWLF8fHx8fq9yR//vzJPhs/Mho/+ioJKdil\nIioqin/++SfTfiuQlt8eiYiIiEjG9ejRg/Hjxz90P8mFrvipivb29ikGM0h9vV9K4te6RUREAHE7\nV8bvyuno6Jjis6mNxqVWd3rFT21NrS6j0WhZt2dtmmdKge1hztR7EijYpcLNzY3333+f0NDQhx6u\n9vf3txyEKSIiIiJZIzw8PFP6iT8KILH4YBIaGkpwcHCKG+XFH53w4HEKqfUf7/r16wCW0TaDwYCD\ngwMxMTGpnuf24GHq1sJlaq+dXvGBLrWDxAMDAy0jntZmwym8ZZyCXSpy5crFJ598YusyREREROQx\nUa5cOSDumKvTp0+nuM4ufrZWtWrVkty7cOECYWFhlpG5xK5evQpAiRIlgLiAVrZsWU6ePMmxY8dS\nrDH+db29vZNdK5iZKlSowKZNm5I9gDzepUuXLKHU2vdEMk7n2ImIiIiIpEO9evXw8vICYN26dcm2\nO3TokCWAWZu15e/vbzm6IDFfX182bdoEQIsWLSzX40Pkpk2bUhx1W7NmDQA1atRIcuh4eqR1LVv8\nrp1nzpyxbKRizapVq4C4kKpgl7kU7ERERERE0sHNzY2+ffsC8PXXX3P48GGr7eKPN/D09KR79+5W\n23z66adWr3/++eeEh4eTP39+y2sBPP/88wBcvHiRjz/+2OqzP/zwgyVc9evXL/U3lEj8IeYmk8my\ndi41Xbt2pXTp0kDckQbW/Pvvv3z55ZcAPPfccxQsWDDdtUnyNBXzIV2/fp1Nmzbh6+tLVFQURmPy\nWfnevXvUrl2bF1544dEVKCIiIiKZ7vPPP2fNmjVcvXqVJk2aMGXKFHr06IGjoyNnzpxh5MiR7N+/\nH4Dvv//eslV/Yvv27eOpp55i3rx5VKhQgdu3bzN27FgWL14MwJdffplgDV/Dhg0ZOXIk06dPZ/z4\n8Vy7do3PPvuMAgUKEBgYyMSJE/nuu+8A6NatW5LD19Mid+7cls/Hjx9PmzZtLEczJDeCZzQamTx5\nMt26dWPfvn3UqVOH6dOnU716daKjo1mzZg2vvfYasbGxFChQgMmTJ6e7LkmZgt1D+Prrr3n33XfT\n9Uzz5s0V7EREREQyWUxMjGW3yofdGCR+N/T43Sit8fDwYM+ePbRu3ZqzZ8/y4osv8uKLLyZp99tv\nvyU7alamTBmaNm3KvHnzrE5LnDx5MsOGDUtyfdq0aRiNRr7//nsWLFjAggULkrTp168fv/32W5Lr\nkZGRqb63hg0bUqVKFf79919mzpzJzJkz6dKlC6tXr07wvU28SU3Xrl1ZunQpvXv35siRIzRo0CBJ\n32XLlmXbtm0UKFAgwfX4/+9S2vgmNDQUiDu6Iq0jiU8SBbsMWrZsmSXUGQwGSpYsaRlOTm43n2vX\nrlnOFBERERGRzOPo6MjQoUP5+++/ady48UP11bt3b1xcXGjWrFmK7YoWLcqZM2dYsGABS5Ys4dCh\nQ4SGhlK1alU6duzIK6+8YlmLZ01gYCBz587lxRdfZPTo0Rw6dIiCBQvSvn17Ro8eneTw7gdNnTqV\nAQMG8OOPP7J7926uXr1K0aJFadKkCcOHD09wEPiDOnbsyO3bt+nUqVOK723NmjWMHTuWvXv3EhYW\nRr169YC4c+aGDx9OcHAwZcqUSfJcr169aNq0KTNmzGDz5s2cOnUKNzc3nnrqKfr165dsyE1LXW3a\ntOHMmTN069aNmjVrplj/k8hgNpsDAOtjw5Ksli1b8tdff1G2bFnWr1+f4l+8jPhp/Hj2T5yE2c4O\ndACjiGQBc2AwgeUq8/qXLahXKAb/oFhblyQiOZCTg4HQCBN33btQr0FrW5fzWPj999/p3bs3uXPn\nxsfHJ9lpmiLpEKjNUzLAz8+PXbt2ATB37txMD3UiIiIiIiLpoWCXAcHBwcTExFCgQAHL1q4iIiIi\nIiK2omCXAcWLF8fV1ZU7d+4QEhJi63JEREREJBtKbl8GkYxQsMsAOzs7nnvuOQDWrl1r42pERERE\nJDuJ31nSZDIRG6v1zZI5FOwyaObMmeTJk4f+/ftz+fJlW5cjIiIiItlEmzZt2LZtGzt27NDGKZJp\ndNxBKiIiIli+fDnh4eE4OjoCceekVK1ald69ezNr1iwqVapE//79qVWrFrlz58ZoNFo9PyUoKIjy\n5cvTtm3bR/02REREROQx4enpSYsWLWxdhuQwCnapCAwMZOjQoZaDKq2Jiori559/5ueff061v+bN\nmyvYiYiIiIhIplKwS4WdnR0FChTAz8+P3LlzJ7hnMBgwGhPOZk1pnvSdO3fIkydPVpQpIiIiIiJP\nMAW7VOTLl4+LFy9mSl9msxk7O7tM6UtEREREHn+3bt1iwoQJ7Nu3j5CQELy8vPjuu++oV69elr3m\nli1bmDdvHm3atGHIkCFZ9jppsWrVKhYvXsxzzz1H7969bVpLTqdglwqDwYCTk5OtyxARERGRbObO\nnTvUqFGDu3fvWq6dP3+eoKCgLH3dP/74g6VLl3L+/HmbB7vff/+dZcuWERkZqWCXxRTsskhMTAxG\nozHJVE0REREReTLMmDHDEupefvllWrZsyc2bN6lWrVqWvq6DgwMArq6uWfo6aeHs7AyggZJHQMEu\nncLCwvj0009p06YNrVq1SrbdK6+8wqpVq5g7dy5dunR5hBWKiIiIyOPg7NmzALRq1YqZM2c+stcd\nO3Ysffv2JX/+/I/sNcX2FOzS4Y8//qBnz56EhYURFBSUbLAzmUxs2bIFf39/unbtSp06dfjjjz8o\nUKDAI65YRERERGwlMjISgJo1az7S1/Xy8sLLy+uRvqbYnoJdGv3yyy8MHjzY8rWvr2+ybaOiomjZ\nsiULFy4kOjqaw4cPU7lyZY4cOULx4sUfRbkiIiIikolCQkLYsmULhw4dIjAwkNKlS9O6dWuqV6+e\npO3x48eJjY3Fz88PgAsXLnDp0iUuXrxItWrVKFy4cIqvFRERwb59+zAajTRt2hSDwcD27dvZsGED\nYWFh1KlTh86dOycb3i5fvsyxY8coXbq0JVTu2LEDf39/8ufPT926dXFzc0vy3P379zl27Bi3bt2i\nWrVqVqeM+vr68scff3Dy5EkAqlatyrPPPkvBggVTfE/JCQ8PZ9u2bRw4cAA/Pz+KFClC3bp1ad68\nueUMaUkbBbs0OHnypCXUubu7M2vWLPr27Ztse2dnZ+bNm8ecOXMYPXo0U6dOxd/fn759+7J3795H\nVbaIiIiIZILZs2czatQoywjcgxo3bsySJUsShLUhQ4Zw5MgRy9dr1qxhzZo1AHz11Ve88847Kb7e\nrVu3aNmyJW5ubpw7d46RI0eyevXqJO2+/fZb3nzzzSTXZ8yYwbfffkv9+vX5+++/AThx4gSvv/46\nEDdV8+uvv07y3Isvvsjy5csBOHz4cJL77777rtXnAD766CM+//zzFN9XYqtXr6Z///6EhYUluefu\n7s78+fPp1q1buvp8kmlnjzSYOHEiAPnz5+fkyZMphroHGY1GpkyZwtSpUwHYt28fv//+e5bVKSIi\nIiKZ64033uCll14iMjKSOnXqMH36dBYuXMhrr72Gg4MDu3fvpkKFCpw/f97yzLPPPkvXrl0pUqQI\nAGXKlOHFF1+kZcuW1K1bN9XXNBgMQNxmfG3atGH16tV06NCB9evXs3btWlq3bg3A6NGj+eyzz5I8\nH795n739/4/hjBo1ij59+gDwzTff8M8//yR4Zvbs2ZZQN2fOHGrXrp3gfpcuXSyhrl27dsyZM4f5\n8+fTo0cPAMaNG0e/fv1SfW/xtm3bRrdu3QgLC6NJkybMnDmT+fPnM3HiREqXLk1wcDDdu3dn586d\nae7zSadgl4qAgADLb0i+/PJLSpQoke4+Ro0aRdu2bQFYtGhRZpYnIiIiIllk+fLlll/Qv/322xw6\ndIgRI0bQv39/vv/+e06fPo2XlxchISEMGDDA8txnn33GqlWrePrppwHo2LEjs2bNYuvWrbRs2TLV\n1zUYDNjb2xMZGcnp06cZN24cGzZsoGPHjnTq1IktW7bw9ttvA/Dpp59y/PjxJM8DSXZnnz17tiVs\njho1ynL9/PnzjBw5EoCBAwcydOjQBM9NnDiRtWvXAnHLkzZu3MjQoUMZOHAgy5YtY8WKFQAsXryY\nn376KdX3B1i+r4MGDWLnzp28/PLLDBw4kDFjxnD+/HmqVKkCkO5RwCeZgl0qTp48SXBwMM7Ozjz3\n3HMZ7mfQoEEAHDlyBJPJlFnliYiIiEgWmTJlCgCNGjXim2++SXK/bNmy/PLLLwAcOHCAZcuWJbgf\nGxsLxO2/kFHNmzfnww8/THL9m2++oWrVqgBpDlPu7u6WtgcPHmTatGkADB8+nOjoaMqWLcusWbMS\nPBMZGcn3338PwGuvvWb5mfZB3bt354033gDipoGmxaVLlwCoUaNGknt2dnZ88skndOjQgeeffz5N\n/YmCXaouXrwIQLly5cibN2+G+6lcuTIA165dw8fHJ1NqExEREZGs4ePjY5muaC3MxOvQoYNlA5Vt\n27ZZbWM2m9P9+vEDASm9dvw0yM2bN6e53/bt2/Paa68BMHPmTF5++WV27doFxI3GxZ87F2/79u1c\nv34dgJdeeinZfuNHLI8fP87BgwdTrcPb2xuIGw08dOhQkvs9e/Zkw4YNCTYvlJRp85RUxA9h29nZ\nPVQ/j8MBkSIiIiKSNkeOHLGMuD311FMptq1ZsyYnTpywuuFIRsUHu/jpnNZUqFABgKtXr3Ljxg3L\nNMvUTJkyha1bt3LmzBnOnDkDwIQJE2jYsGGStqdPnwbifpY9c+YMt2/fJiQkJEEbDw8Pbty4gdFo\nxGQycfnyZerVq5diDa+88gp//fUXt27dol69epQrV44OHTrQpk0bmjdvjouLS5rei/w/BbtUlC9f\nHoArV64QFRWV4W1Xb968CcT9wddhkSIiIiKPt/idGt3c3FINTGXLlgXijgvIbCmdR1esWDEAoqOj\nuX79epqDndFoZOrUqbRp0waIm5n23nvvWW174cIFIO770bNnzzT1f+XKlVTb9OjRg7lz5zJ8+HBM\nJhMXLlxg6tSpTJ06FTc3N9q1a8cbb7xBo0aN0vSaomCXqooVK+Ls7ExgYCC7du1K9lDy1MQPkVeu\nXJlcuXJlZokiIiIikoXiNyNJTvzoWmrtMlv8iCKkf3bZvn37LJ9fuHCB9evX8+yzzyZpFxMTA8Tt\nDt+6dWuio6OT3S/CaDTi6+tr2fgkNUOGDKFv376sXr2adevWsXXrVu7evUtoaCgrVqxgxYoVfP75\n53z00Ufpem9PKgW7VOTNm5fOnTvz+++/M3Xq1AwFO39/f3788UcAq0PcIiIiIvJ4iZ8KGBoayo0b\nN/D09Ey2bfyeDAUKFMj0OsLCwsiXL5/Ve/H7Nri6ulK8ePE097llyxY+/fRTjEYjFSpU4MyZMwwc\nOJCzZ88meQ/xo4Kurq4sWLAgwREKmcHFxYW+fftajhM7fPgwK1asYPr06QQHB/Pxxx/TsWPHJMcv\nSFLaPCUNXnzxRQDWr19vdVei1PTq1csyNP/CCy9kZmkiIiIikgVq1aplGYGztrnHg06ePAlAtWrV\nMr2Oo0ePJnsvfv1b6dKl0xwqw8LCLD/bDhw4kNOnT1OiRAnu37/PK6+8kqR9/AaAPj4+nDp1Ktl+\ng4OD2blzJ6dPn06yBi+xqKgorl69yq5duwgICEhwr06dOkyYMIEdO3ZYrsUfsi4pU7BLg5YtWzJw\n4EAAxo8fT4cOHSy/mUnJ5s2bKVu2LH/99RcQd85IWoemRURERMR2SpYsadm45Ndff0223ZYtWyzn\nyDVu3DjTXj9+ZCz+0PDETCYTv//+OxB3eHhajRw5kqtXr+Lk5MSnn34KwFdffQXAypUrmTNnToL2\njRo1ws3NDcBytIM1v/zyC82aNaNKlSqcPXs2xRpOnz5NyZIladq0KQcOHLDaplq1ahQqVCiN70pA\nwS7N5s+fbzlQcuPGjZQrV47q1avz6quv8v333zNv3jx++uknxo8fT9++fSlSpAjt2rWznNHx8ssv\n88knn9jyLYiIiIhIOsQf2r19+3Y+++yzJPdv3rzJsGHDgLiRrf79+2d6Db/++isLFixIcv2ll17i\n0qVL2NvbW2pIzdKlS/n5558BmDx5MiVKlACgT58+lo1RRo0alWDzE29vb8sI39SpU1m4cGGSfk+d\nOsU777wDxIXbunXrplhHzZo1qVixIoAlXCY2a9Ysbt26BUCDBg3S9P6edFpjlw5bt27lnXfesRxQ\nefLkScvQe3I8PDyYOXOmZd6wiIiIiGQP/fv3Z+fOnfz00098+umnbN++nSFDhuDu7s6RI0eYMmUK\nISEh2NnZWQ1fGTm/LvGzdnZ2DBo0iC1bttCzZ0+Cg4P54YcfLNMTf/jhB0qWLJliHwC3bt3i5Zdf\nBqB169aMGDEiQdspU6awadMmgoODeemll9iyZYvl3qRJk9izZw8HDx5kwIABLF++nF69euHs7Mzf\nf//N999/T1RUFG5ubklG/OJrSPy9+Oabb+jcuTN///03ZcqUYeTIkRQtWpSYmBg2bNjAb7/9BsCr\nr75KrVq10vvteyIp2KXT119/zciRI5kxYwbr16/n4sWLREZGWu4bDAa8vb2pX78+PXv2pGfPnpm+\nyFREREREHo3Zs2dTunRp3nvvPXbu3MnOnTsT3K9Tpw5Lly6lTJkySZ6Njo4GSHYXyZTE73j5888/\ns2jRIn777TdL2AFwdnZmwYIFVo8giH82KirKcq13794EBARgNBqZNm1akmcKFy7MpEmTeOmll/jz\nzz959913LVM0jUYj+/btY9iwYcyfP581a9awZs2aBM9Xq1aNlStXWo5+iBe/q+aDO3gCdOrUicWL\nFzNgwAAuX77M6NGjk9T0zjvvWGqQ1BnMZnMA4GHrQrKroKAgrl69ir+/Py4uLhQtWpTChQs/9Ha3\nP40fz/6JkzDb2cEj3jpXRJ4M5sBgAstV5vUvW1CvUAz+QbGpPyQikk5ODgZCI0zcde9CvQatbV1O\nhvn5+bF+/XqOHDlCcHAw5cqVo02bNilOOzx8+DAnT56kXr166dpn4erVq5ZRuD179tCwYUNWrFjB\n5s2bMRgMNG7cmG7dulnWviV29uxZDhw4QIUKFXj66acJCgpi9erVODk5Ua5cuWR3mIyNjWXXrl1c\nu3YNDw8Pq2v3Ll++zMaNG/n333+JiIigTJkyNGvWLNmd30+cOMHRo0epXr261ZG3kJAQNm7cyPHj\nx7l+/Tru7u7UqFGDjh07ao1d+gQq2D2mFOxEJKsp2InIo5BTgt2j9GCw27p1q2WfB5EUBGrzFBER\nERERkWxOwU5ERERERCSbU7ATEREREXmMPLjRSPwGLCKpUbATEREREXmMODs7U6ZMGZycnMidO7et\ny5FsQvvwi4iIiIg8RgoXLszFixdtXYZkMxqxExERERERyeYU7ERERERERLI5BTsREREREZFsTsFO\nREREREQkm1OwExERERERyeYU7ERERERERLK5HH3cwfHjx5kxYwbBwcEYDAYMBgMA4eHh5MqVi2nT\npuHh4WHjKkVERERERB5Ojg12169f55lnniE8PNzqfQcHB7755hsFOxERERERyfZybLCbP38+4eHh\n1K1bl+HDhxMbGwuA2WwmKioKZ2dncufObeMqRUREREREHl6ODXZhYWEALFy4kAoVKti4GhERERGR\nrFOzZk3s7Ow4fPiwrUsRG8mxm6eMGTMGd3d3/v77b1uXIiIiIiKSZUaPHs3x48e5d++erUsRG8qx\nI3Z58+alV69eDB48mL1791KqVCns7e0xmUxERkbi4uLCiBEjcHNzs3WpIiIiIiIZ8sknn/Ddd98B\n4OLiYuNqxJZybLD7559/mDt3LgBz5sxJct/BwYGBAwcq2ImIiIhIttS/f38WLVpk6zLkMZFjg13B\nggUZOHAgrq6uODg4YDabMZvNAERERODm5oarq6uNqxQRERERSZ81a9bw0ksvcfv2bQoUKEChQoU4\nfvy4rcsSG8uxwa5UqVLMnz/f1mWIiIiIiGSaZcuW0atXLwAKFy7M2bNnWbNmDQMGDLBxZWJrOXbz\nlHj+/v5MmTKFfv368dprr7F7925blyQiIiIikiH+/v4AfPXVV9y4cQN3d3euX79u46rkcZBjR+wA\n9uzZQ7NmzSxn2AFMnz6dvn37aj6yiIiIiGQ7HTt2xMfHh2LFitm6FHnM5NhgFxAQQOfOnYmNjaV1\n69a0atWKu3fvsnDhQhYvXkyFChX45JNPbF2miIiIiEiaKdBJcnJssFu5ciX3799nwoQJvPfee5br\nn332GV5eXnz33XcKdiIiIiLZnNlsJiY6GoPBAIC9gwPB9+8THBgIZnPcR1r6MRgwGgzkL1wYe3t7\nYmNjMRgMmGJjcXByysq3IJIpcmywO3bsGADPPfdcguuurq506dKFxYsX4+vri7e3tw2qExEREZHM\nEHTvHlPfegs7e3vMJhOuuXLhc+ECB/78E0cnJwzGtG0pER0ZiZuHBw3ad8DZ1YXoqCjsHRzwLl6c\nFz74IIvfhcjDy7HBrmrVqgD88ccflC9f3nI9JiaGjRs34ubmplAnIiIiks1FR0ezbMkSIiIjMQAx\nQC6jEW83N0xmM2kbr4s73DsyNJRFvy4glrgdBs1Ay44deSGLahfJTDk22HXr1o033niDN998k507\nd9K2bVvu3r3LvHnzCAgIYMyYMbYuUUREREQekp3RSHH33EQ5RCS4bjaDAQOGtHZkMuNgZ0/xXO4J\nLudzds6cQkWyWI4Ndl5eXqxYsYIOHTqwevVqVq9ebbnXpEkTvv76a9sVJyIiIiKZxs5gwM6Q5giX\nLsa0R0MRm8qxwQ6gffv23Lx5k1mzZnHixAk8PDzo2LEjPXr0sHVpIiIiIpIJDMSFr6wKYAp2kl3k\n6GAHUKhQIT799FNblyEiIiIiWcRoiPvICoa07b0iYnM5KtgdOHCAU6dOUblyZapVq8aGDRswGo0Y\njUYMBgPm/213Gx0djZOTE+3atcNZ86ZFREREsjUjWTeyplwn2UWOCnaTJ0/m999/p2vXrnzzzTf0\n7t072bYODg74+PhoZ0wRERGRbM5oALssGrFTsJPsIkcFu2rVqrF//36qVauGm5sbjRo1ws7ODnv7\n/3+bZrOZiIgI8uTJg6Ojow2rFREREZGHZ8jSNXbp2FfTZkwmEwCRkZE2rkRsKUcFuw8//JAPP/zQ\n8vXu3bttWI2IiIiIZDUDYAeYnuARO09PTwDKlClj40rElrLDn9UMuXbtGr169WL79u1J7k2ZMoW+\nffvaoCoRERERyWzGB0btMv/j8ffyyy9jNpvZsmWLrUsRG8oOf1bT7MKFC+zbt4+goCD27t3LsmXL\nWLt2LSaTiTNnznD+/HkCAgJYtmwZS5YswdfX19Yli4iIiMjD+N/6uqz6MGbR+XgimS1HTcX877//\nmDRpUoLfVkyZMoUpU6YkaVu4cGHy5Mnz6IoTERERkSxhhwFTFq2Fs8sGa+xEIIcFu9atW+Pt7U3z\n5s0JDg4mKioKNzc3PDw8iImJwWw2ExMTg9FoZOLEiTrqQERERCSbMxA3spZVa+wU6yS7yFHBDuJ2\nxvTz8+Po0aPUrl2badOmMXjwYFuXJSIiIiJZJCt3xcyqg89FMluOC3bxatWqhdls5vr16wQGBuLh\n4UFYWBjHjh2jTp06ODk52bpEEREREckEWTlip2An2UWO2jwlsVmzZlGsWDH27t0LwL1792jYsCEl\nS5bEz8/PxtWJiIiISGawMxiy7COrRgJFMluODXanT5/m5ZdfBsDwv92MHB0dqVChAr6+vrz11lu2\nLE9EREREMokxCz8U6yS7yLHBbvny5QD88ccftG/fHoACBQpw9uxZnn76aRYsWEBISIgtSxQRERGR\nhxS/eUrWHXdg63cokjY5NtjFT7UsU6ZMknulS5cGULATERERyQGy9oByJTvJHnJssGvdujUAr7zy\nCj4+PoSHhxMYGMjs2bNZvHgxVatWxdvb28ZVioiIiMjDMvxvZC0rPnQ+uWQXOXZXzE6dOtGgQQP+\n+usvSpQokeT+u+++a4OqRERERCSz2WMgq1bD2WVJryKZL8cGO4AtW7bw5ptv8ssvvxAdHQ1AuXLl\n+Oabb+jatatti0uF2QzRMSYw61dFIpI1TDEmYmNNODk54OJqwDk6x07iEBEbcnIwEGswWTazy2zx\na+zMWXXcQdZ0K5LpcnSwc3NzY/bs2fz444/4+vri5uaGh4eHrctKEzuDGQ8HM2Y7s7ZjEpGs4Wgm\nxt5AWHAYoffCCQs227oiEcmBYh0MhEWYMefPun9j4newzKq+RbKDHB3s4hmNRgoXLmzrMtLFPq87\nbm0aAgb0o5aIZAUHTMQa3Zi76DKLokMxGzXhSEQynwEzZoz0fCkyy17DzmDAnCUjgmaMhsyNdoGB\ngRw9epSoqChq1qxJgQIFMrX/jIqOjubEiRP4+vpSvnx5ypUrZ+uSJJ1yVLAbM2YM3377LSNGjODt\nt9+madOm2Nvb4+jomKBdSEgIefLkYdu2beTPn99G1aYsl7ORWoWNxJoV7EQka+R2NHAq3IEVa89i\nvH0Te1cXW5ckIjmQKTYWOwcHol+IzpL+DWThiJ3ZkGkrYu7evcuLL77I6tWrE1xv3rw5s2fPpmzZ\nshnq988//+TZZ58lf/782Ntb/9Hex8eHiRMnMmbMGKv3P//8c8aPH09UVJTlWvHixZk+fTqdOnXK\nUF3y6OWoYHft2jXMZjPXr18nMjKS//77L9m2t27dIiYm5tEVl07O5lg8TYHEmI2YbF2MiORI+WJj\n8cERQ3gIzpEB2JnCbF2SiORA5pgYDI6O2GXh0hI7DJizYu2KIXMCo4+PD7Vr18bf3x97e3uaNGmC\nvb09u3btYvv27ZQvX55//vmHp556Kt19//vvv0RFRXHz5s0U2wUGBlq93rFjR/744w8AatSoQZky\nZdi3bx8+Pj507tw5xUAoj5ccFew+++wzBg8eTPHixSlWrBjbt2/HYDBgNCb8KxkbG4u9vT2enp42\nqjR1JgyEm+yIMRsU7EQkS4SazESbDNg5OmA0GDE6Odm6JBHJgcx2dhgcHbN0BtLjvnnKgAED8Pf3\np0KFCuzatcsy/fL+/fu0bt2aw4cPM2DAAM6dO5fuvv/++28AmjRpwtixY/H19U3S5t69e7Ro0SLJ\n9c8//9wS6jZs2ECHDh0s90aOHMmMGTN4++23adSoEfXr1093bfJo5ahgV7FiRSpWrGj5ulmzZrYr\nRkREREQeiSwbsePhjztYsWIFu3btAmDJkiUJ1tTlzZuX9evXU65cOc6fP8/MmTN55ZVX0tX/xYsX\nAejevTsdO3ZM83N+fn588803AEyePDlBqAOYPn06ly9fZuPGjXzzzTesXLkyXXXJo5ejgt2KFSs4\nePAg9vb22NvbJ9hW12yO+z2RwWAgMjISFxcX3nzzTXLlymWrckVERETkIT3uxx0sW7YMiJvyWLNm\nzST3vb29GTRoEDNmzGDRokXpCna+vr4cOXIEgOrVq6errrVr1xIaGoqrqysvvPCC1TYvvfQSGzdu\nZNWqVVy9etXq2dDy+MhRwW758uUsWbIkTW0dHBwYPny4gp2IiIiIZJnDhw8D0LBhw2TbtGrVihkz\nZrB3715u3bpFoUKF0tT3mTNngLgjvurUqZOuug4ePAjAU089lezypMaNG+Pu7k5wcDB79+5VsHvM\n5ahg16lTJ/Lnz4/BYKBs2bIsWLCAw4cP07FjR2rXrk1oaChLly7lxo0bfPrpp3h7e9u6ZBERERHJ\noc6fP2+ZKpnSiFqRIkWAuBlmV65cSXOwO3/+PBAXwHLnzs26dev4448/8PX1pWzZsnTp0oVGjRpZ\nfTY+2KVUl6enJ97e3gQHB1vehzy+clSw69evH/369QPA39+fN954g2+++Ya3337b0mby5MnkyZPH\nMmwtIiIiItmc2Rz38aCMnlOQuJ+H8OBOlCmFtWLFiuHs7ExERATnz5+nQYMGaep///79ANy+fZsq\nVapw+vTpBPcnTZpE06ZNWb16NXny5ElwLygoCCDVs55LlizJhQsXLKOD8vjKUcHuQZMmTcJsNtO5\nc+ck99q1a8fSpUu5e/cuXl5eNqhORERERDKD2WwmLCCAmPDwuK9jY7F3ccHZwwOzKe17ixsMBkyx\nsYQHBGA2mTAYDJhNJqLCMn4UzIPBzs4u+W1YcuXKhZubGxEREURERKS5/7NnzwJw9OhR8ubNy6RJ\nk6hevTohISH8/vvvLFmyhJ07d1KvXj2OHz+Oq6srAJGRkYSEhKRaF4C7uztAuuoS28ixwc7DwwOA\nr7/+mnnz5lmuHzx4kKVLl2Jvb2/5wy0iIiIi2ZOdkxPPvP46Rjs7zCYTzh4e+J44wZmVK7F3ccFg\nTNv2J9Hh4TjnyUOdoUNxzJWLmMhI7BwdccmfP8O1hf8vbELc/g7Jid/4D+KO5UqLgIAALly4AEDb\ntm3ZtGlTgvvdunWja9eu9OnTh4sXLzJq1CjmzJkDkCBAOjo6pvg68XWntS6xncw4muOxNGjQINzc\n3Pj555/x8PCgUaNGlC1b1nLw49ixY3Fzc7NxlSIiIiLyMMyxsdw+fpxbR45w6+hRrv3zD/cvX8bo\n4JCu6ZhGe3tMMTH4njjBjUOHuHX0KDcPHSLQxyfDtT14lrIphdFDs9mcYAf3tMidOzfHjh1jx44d\nzJ8/32qb3r178+KLLwLw66+/Wg4xNxqNltpSqiu+tsTvRR5POXbErlChQmzZsoXnnnsOX19f9u7d\na7k3ZswYxo8fb8PqRERERCQzmGJiuLJzJzH/G4Eym0zYOznhlDt3uqZi2jk4EBsdzaVt2yxTMTGb\nMWd0rR7/P40RUh7xioqKIioqCsAycpcao9FIkSJFLBuvJKdv377Mnj2bqKgojhw5QuHChXF1dcXV\n1RV/f39iYmJSfD4yMhJIfcqm2F6ODXYADRo04NatW4SEhHDv3j1u375N8eLFKViwoK1LExEREZFM\nYDAYcMmTh5j/BZB46Ql1EDcyFd/XgxxdXDJcW/zSIMAS3Kzx9/cnICAAIMEB5pnhwc1R4nfRtLOz\nsxz5lVJdAHfv3gXQbvLZQI4fUzWbzfz333/8888/FC1aFA8PD8sfUBERERGRrFKoUCHLGrZLly4l\n2+7atWuWzytWrJhl9Ty4DKl48eIAKR5jEBMTw3///QdA5cqVs6wuyRw5Otjt2LEDFxcXqlWrRu/e\nvTl16hR79+6lQIEClj+kIiIiIiJZoUCBAlSrVg2AEydOJNsu/iiBAgUKpPkQ8IULF+Lh4cHTTz+d\n4qDF1atXLZ9XqVLF8nnt2rUBOHnyZLLPXrp0ybIur3z58mmqS2wnxwY7X19f2rRpQ2RkJD169AAg\nNDTUcn/MmDG2Kk1EREREnhDNmzcHYP369cm22bBhAwCtW7fGJY1TPw0GA0FBQRw4cCDJjpgPWrZs\nGRA3lbJu3bpJ6jp06FCyoXPz5s1A3HTOJk2apKkusZ0cG+wWL15MdHQ0mzdvtvyBPnz4MC1btqRZ\ns2asWLHCMpdZRERERCQrDBo0CIgbGZs2bVqS+6tWrbKEvhdeeCHJfV9fX86cOZNguiZAv379KFOm\nDBB3vJe13S3Xr1/PTz/9BMBrr72Gs7Oz5V7r1q0to3Yff/xxkmfv3LnDhAkTAHj++edxcnJK9b2K\nbeXYYHflyhUAqlatarkWv+tP/OJPHbQoIiIiIlmpatWqjB49GoBRo0bxwQcf4Ovri7+/P99++y3d\nu3cHoH///rRq1SrJ84MGDaJy5co0btw4wc6aBoOBL7/8EoB///2XatWqsXXrVvz8/PDx8eGTTz6h\nU6dOADRp0oT3338/Sd/jxo0DYM2aNXTo0IEzZ84QGBjI5s2bqVevnmXjwQ8//DBzvymSJXJssIv/\ni/H8889z9uxZ7O3tMZvNbN++nSVLllCjRg3t7iMiIiIiWW7y5Mk8//zzAEyYMIFChQqRP39+3nrr\nLSDugPGFCxdafTZ+YCIkJCTJoETPnj1ZsGABAKdPn6Z169Z4eXlRokQJPv/8cwDat2/Ptm3brPbd\noUMHZs+eDcDGjRupXLkyefLkoV27dvj4+JA/f342b96c4NgGeXzl2OMOOnfuTMeOHdmwYQOVKlUC\n4oapv/76awA++ugjW5YnIiIiIk+QX3/9lYEDB/LTTz9x4MABTCYTtWrVYuDAgTz33HPJPte/f3+8\nvLyoVasWrq6uSe4PGDCAVq1aMWvWLP766y/OnTtH3rx5qVevHgMHDqR169Yp1jV8+HCaNm3K9OnT\n2bFjB76+vlSqVIlnn32WV199NcFOmvJ4M5jN5gDAI7WG2dWMGTMYP348oaGhmEwmKlSowKRJk2jW\nrJmtS0vRurlTcTu2khizgfSdwiIikjZ5HWL5J8qLJWtO4HzzCnb6j7eIZAFzTAwGR0dG//IzHbp1\ny/T+w+7eZfbTTyc5xy6zlG7Rgu6//polfYtkosAcO2Ln4+PD9evXGTFiBCNGjCAiIgI7OzscHBxs\nXZqIiIiIiEimyrFr7BYvXkzDhg0tW7k6Ozsr1ImIiIiISI6UY4Nd0aJFAVI8sFFERERERCQnyLFT\nMfv378+qVatYsWIFzz77LL169cLNzQ2j0Uh0dDSOjo60a9cuwXkeIiIiIiJZJTIykm3btnHw4EGi\noqKoXbs2bdu2JVeuXA/d95UrV9ixYwcXL17EaDRSqVIlWrRokeou8D4+Pvj4+CR7MHpsbCwODg5U\nq1YNe/scGx1yhBz7/84///zDihUrANiwYQMbNmxIcN/BwQEfHx8deSAiIiIiWW7z5s306tWLoKCg\nBNft7OyYP38+/fv3z1C/9+/fZ8CAAUl+1o03cuRIqwejx3vuuec4dOhQiq/h4uJiOf5AHl85Nti5\nu7tTo0YNPDw8cHBwwGQyYTKZMJvNhIaG4uHhgaOjo63LFBEREZEcbvPmzbRr1w6AGjVqMHjwYOzt\n7fn55585fPgwzz//PKGhobz44ovp6tff35/atWvj4+MDQO/evWnSpAnR0dGsXLmSXbt2MX36dM6e\nPcuff/6Z5PmQkBD++++/h35/8njIscGuSpUqHDt2zNZliIiIiMgTLDIykuHDhwPQt29fFi1aZLk3\nYsQIXnrpJWbPns0bb7xBhw4dLPtEpMXYsWPx8fHB2dmZ3bt3U7duXcu9119/nVmzZvHyyy+zdetW\nvv32W0aPHp3g+VOnTuHn5wfApk2bqF+/Prdv307QxmQyYWdnR968edP93uXRypGbp5jNZvbv38/G\njRvx9fW1dTkiIiIi8oRasGAB165dI1euXEyfPj3J/VmzZlGlShXCw8OZMWNGmvu9efMmv/7vfL2v\nvvoqQaiL99JLL9G7d28AS9sH/fvvvwCUL1+etm3b4uHhQfny5RN8VKxYkXLlymFnZ5fm2sQ2clyw\nu3btGpUrV6Z+/fp06NCBQoUKMWHCBFuXJSIiIiJPoGXLlgHQrVs3PD09rbbp1asXAOvXr09zv//8\n8w/R0dEAtGjRItl2TZo0AeD8+fOW0bl4Z86cAaBSpUppfl15fOWoYLd161aGDBnC2bNnMRgMlt8s\nfPDBB6xbt87G1YmIiIjIkyQyMpKTJ08CUKdOnWTbNWjQAIibGpnWNW+NGzdm7dq1rFixgvLlyyfb\nLiYmBgB7e/sko24HDhwA/j/8SfaW44Ld1q1bAVi3bh0xMTGW4eeNGzfasjQRERERecKcP3/esiyo\nRo0aybYrWLBggmfSwsvLi06dOtG9e3ecnJySbRe/W2bFihUTrJMzmUyW16pRowZbt27lmWeeIXfu\n3BQpUoTnnnuOzZs3p6kWeTzkqGB3/fp1ACpXrkzHjh2BuLnF8P9DzSIiIiKSc5jNZiKDgogMDCQy\nMJCI+/eJDg3FYEzfj7kGg+H/+woIiOsvIICYiIgM1xYSEmL5PKXNRwoVKoS7uzsAV69ezfDrJbZm\nzRq2bNkCwPPPP5/g3vHjxy0bpfTu3ZvWrVvzzz//EBwczM2bN1m5ciXt2rWjb9++mVaPZK0ctStm\n/DzjBw9PdHBwAP5/GFpEREREcg6jnR1Fnn4as8kEZjN2Tk6E+flx78IFjPb2YDCkqZ/YmBjsnZwo\n9swz2Dk4YIqNxWhnR94yZTJc24Nn1hlSqMPFxQUnJyeCg4Mz7WfWK1euMHDgQACqV6/Oa6+9luD+\ng7vHh4SEMHPmTHr27ImjoyM+Pj588sknrFixgiVLlmBvb2918xV5vOSoYBf/FyY2NtZyLTw8PMG1\n8PBwoqOjMRgM5MqVK8W/ZCIiIiLyeDPa21Ps6acx2NtjNplw9vDA98QJ7vz7Lw52dmn+Wc8UHY0x\nVy6KPfMMjrlyERMZiZ2jIy7JbHiSFpGRkZbP4wcbrLGzs7Osf3vw59iMunHjBo0aNSIoKAgnJyeW\nLFmSbNsiRYqwc+dOyjwQYKtUqcLy5ct54403mDp1KgsXLuT555+nbdu2D12bZJ0cGez+/fdfXFxc\nMBgMlr8c+/fvx9XVFbPZTEREBO7u7ly4cCHBnGYRERERyV5iIiP5+/vvifnfL/PNJhN2Tk44ubvH\njeKlkYOLC9FhYeydPBmzyRT3c6XZTLmOHakzbFiGanN0dPz/OlMYiTOZTJj+V+vDHitw5swZGjdu\njL+/P3Z2duzYscPqrpeDBw9m0KBBREVF4ezsbLWvcePGsWTJEm7fvs1vv/2mYPeYy1HB7kERieZD\nm0wmy+hd/H2z2fyoyxIRERGRzGQw4OTujt0DIQpIV6iDuLV6BoMBp9y5E1y3Tyb0pEXuB/oypVBP\nRESEZXTvwSVF6fXXX3/Rtm1bYmJicHd3Z8+ePVSvXj3Z9kajMdlQB+Du7k7Lli1ZtGgRJ06cyHBd\n8mjkqGD35Zdf8tZbb2EwGDCmsGA2NjYWe3t78uXL9wirExEREZHMZiBu1lZmLa9J3M/D9Bq/IQpA\naGhosu18fX0t6/FKliyZodeaOXMmr776KgClSpVi9+7dFClSJEN9PSg+nN66dYuwsDBcXV0fuk/J\nGjkq2JUsWTLDfxlERERERDJTmTJlyJ8/P35+fpw6dcpyXl1i8UciAJQrVy7dr/Pxxx8zbtw4ABo1\nasTWrVtTPAIhNjaWwMBAbt++jaenZ4pLk+KnkHp5eSnUPeZy1HEHIiIiIiKPC1dXV8tUyOPHjyfb\n7p9//gGgfPny6R6k+Oijjyyhrm/fvuzevTvFUAdxUzDr169P5cqV+eijj1Jse+rUKSDuHDx5vCnY\niYiIiIhkkc6dOwOwfPnyBPs9PGjVqlUAtGvXLl19//LLL3zxxRcAfPjhhyxatChNzxkMBipUqADE\nHWCeeG+KeH/99ZcldHbr1i1dtcmjp2AnIiIiIpJFBgwYgLu7O3fu3ElylhzABx98wKFDhwB4+eWX\nk9zfunUrkyZNYuXKlQk2/rtz5w5vvfUWAPXq1WPcuHGYzWZu376Nn59fgo+7d+/i5+eXYAOXV155\nBYCbN28yatSoJK978+ZNBgwYYOm/f//+D/FdkEchR62xExERERF5nHh6ejJlyhSGDh3K3LlzOXv2\nLCNHjsTR0ZF58+axYcMGAL7++murxxJ89dVXbNu2jYIFC9K+fXtcXFwAmD59Ovfu3QPiDiP38vLi\n3r17VnfVjIqKwsXFhfPnz1O0aFEAOnTowKhRo/j+++/56aefOHr0KK+++ip58+bl4MGDTJkyhbCw\nMHLlysWCBQuy6tsjmUjBTkREREQkCw0ZMgSj0cjgwYPZu3cve/fuTXD/q6++YuzYsVafjd+l02g0\nJtj1fc+ePZbP/fz8LJ9HRUVZ7Sc8PDzJlMupU6dSqFAh3nvvPQ4dOsSQIUMS3K9Tpw7Lly/X5oTZ\nhIKdiIiIiEgWe+GFF+jUqRMrV67k0KFDmEwmatSoQdeuXS2jaNZMmDCBwYMHU7hw4QQHnn/++efc\nvXs3xXPo4sXExODg4GD1dd59912GDh3K+vXrOXr0KEFBQZQtW5bWrVvz9NNPZ+zNik0o2ImIiIiI\nPAL58uVj+PDhDB8+PM3P1KtXj3r16iW53qhRo0yry8vLi8GDBzN48OBM61MePW2eIiIiIiIiks0p\n2ImIiIiIiGRzCnYiIiIiIiLZnIKdiIiIiIhINqdgJyIiIiIiks0p2ImIiIiIiGRzCnYiIiIiIiLZ\nnIKdiIiIiIhINqcDyh9SREQEp0+fJjQ0FLPZjMFgsNwzm82YTCZiYmIICwtj5cqV2NvbM2fOHBtW\nLCIiIiIiOY2C3UP4/fffGTJkCKGhoWl+pnnz5llYkYiIiIiIPIkU7DJo586d9O7dO83tDQYD1apV\no0ePHllYlYiIiIiIPIkU7DLohx9+AMDb25vZs2fTqVMnpk2bxqhRo3jxxRf58ccfOXv2LAcPHuS9\n997j5s2bfPjhh/Ts2dPGlYuIiIiILWzbto05c+Zw4MABoqKiqFu3Li+88AJdunR56L4XLlzI4sWL\nOX78OK6urjRs2JBXXnmFp556KsXnFixYwJ49e8ifP7/V+xEREbi7uzN27Fjc3Nweuk7JOgp2GfTP\nP/8A8OOPP9KpUycAOnTowKhRozh27BgGg4FKlSpRqVIl+vbtS8mSJenVqxe3bt3C29vblqWLiIiI\nyCM2fPjwJPssXL9+ndWrV9OxY0fWr1+foX7DwsJo0aIF+/fvT3D9woUL/PLLL7z//vuMHz8+2ecn\nTJjAuXPnUnwNFxcXRo4cqWD3mNOumBng4+ODj48PuXPnpkOHDpbrZcqUoVChQhw4cAA/Pz/LdQcH\nB3799VcAPv/880der4iIiIjYzocffmgJdePGjcPX15fbt2/zzTffALBhwwYGDBiQob6fe+459u/f\nj4eHB4sWLcLf35+rV68yfPhwIC64TZ061eqz165d48KFCwDUrl2bVq1a0bhx4wQftWrVomHDhjg4\nOGSoPnl0NGKXAbGxsQDkypUryR/ykiVLcuvWLa5fv55gSLtFixY4OTmxYsUKyzROEREREcnZzp07\nZxkx++GHH3jllVcs995++23Kli1L9+7dWbhwIYMHD6ZFixZp7nvx4sVs2rQJgE2bNlG/fn0APD09\nmT17NgUKFGD8+PF8+OGHPP/88+TLly/B85cvX8ZkMmEwGNiwYYNmlWVzGrHLAE9PTwDCw8OT3Ktb\nty4Ax48fT3KvSpUq3Llzhxs3bmRtgSIiIiLyWIgfqStXrlyCUBevW7duPPvss0Dcerf0mD17NgB9\n+vSxhLoHffzxxxQqVIiQkBCWLl2a5H780qKnnnpKoS4HULDLAA8PD6pXr879+/c5cuRIgnuFChUC\nSHId4N69ewBER0dnfZEiIiIiYnN//fUXAJ07d062TceOHQHYuHGj1YEDa3x9fdm9ezcQFw6tcXR0\npF27dgCsXbs2yf2LFy8CUL169TS9pjzeFOwyKP4vybBhw7h//77l+jPPPAPA3LlzuXv3ruX6mjVr\n+O+///D29qZYsWKPtlgRERGRHMoMmM3mrPt4iNpu377NyZMnAahWrVqy7SpVqgTAnTt3uHLlSpr6\nPn78uGV5UOXKlVPt+/z580nu/f333wCW0b7bt2+zadMmtm/fTnBwcJrqkMeHgl0GjRo1Cjc3N44e\nPYqnpyeLFy8GoFmzZlSpUoXQ0FCeeuopvvrqK15//XX69OkDQOvWrbGzs7Nl6SIiIiI5hsEAdk52\nWfZhdMj4j8u+vr6WmVrxAcuaB3/pf/r06TT1ff36dQC8vLwoU6ZMsu1Kly4NwJUrV7h8+bLl+r17\n9ywjdiEhIbRv3x5vb2/at29PixYtyJ07NwMGDMDf3z9N9YjtafOUDCpSpAgbN26kVatWREVFYTKZ\nLPe+++472rRpw3///cd7771nue7s7MyHH35oi3JFREREcqzI4LvEhIWCwYApJgbHXO645C+AOdYE\naRxzMxiNmGJiCLtzC1NsLAajEbPJRGxkSIbrCgwMtHzu5OSUbDtPT0/y5MlDQECAZelOWvt2cHBI\nsW8vLy/L5wEBAZbPL1y4QGRkJACvv/46BoOBunXrUqJECU6cOMGFCxdYuHAhGzZsYN++fVSsWDFN\ndYntKNg9hMaNG3P37l1mzZpl2TQF4kbl/v77b9566y327duHg4MDLVu2ZMqUKZQvX96GFYuIiIjk\nLEYHB2oPHY7BYIfZbMLBLRcBVy7is2c79k6OGIxpmykVGxWFS153KnUfgYOzK7HRURgdHHAvWCjD\ntYWE/H8oTGnGlpOTE46OjgDExMSkqe/4qZL29vYYjcmPKjo7O1s+f7DvHTt2WD7v0KEDy5Ytw9XV\n1XJt9erV9OrVi/v379O9e3dOnDiBvb2iw+NM/+88pNy5c/P2228nuV6/fn327t3L/fv3sbOzI3fu\n3DaoTkRERCRnMxjAwdneEuwcne2xd7bD3tmInaMdhhRCT4J+jHbYO9vh4GSPg7M9RjsTRgd77Bwz\nvoQmfg0ckGL4MhgMGAwGIG69YHr6Tm2JT3y/iftu3Lgx/fv3x9vbm6+//jpJP127dmXJkiU899xz\nnDlzhunTp/PGG2+kqTaxDQW7LJY3b15blyAiIiKSY5liYji9Yj4xYeFg4H9TMXPhms6pmI4ujphi\novl36Zz/n4ppNlG65bPUeuHVDNXm4uJi+TylXdFjYmIso2lpHRWL7zt+SVBywfHB132w7wYNGtCg\nQYMUX6N79+7UrVuXQ4cOsWXLFgW7x5yCXSpMJhOBgYGYzeYEv/HIaF+Ojo64u7tnUnUiIiIiTzaD\n0UCe4kWJjYpMejOdG5/YOTqQp2TC3ctd82f8l/R58uSxfP7gfgyJhYSEEBoaGvd6D0yHTEn84EFs\nbGyKwS4oKMjy+YNBM61q1qzJoUOHOHfuXIqvI7anYJcKf39/KlasyL1793Bzc0ty/8E/3CaTiYiI\nCMvQuIuLCw4ODkDc0HdwcDDNmze3nGciIiIiIg/P3tEOgyFrfqy1s8/4VMwHl+I8eAxWYteuXSMi\nIgIgzfsxxIfGwMBA7t27R4ECBay2iz8+wcvLi1KlSqWp7wfFj/KFhYURFRWVYM2ePF4U7FJhMpks\nC1/jf5OSVuHh4ZZDJh92tE9EREREkjIYDJY1cVnB/iHW2JUvX55y5cpx4cIFjh8/Tps2bay2u3Hj\nBhA3YFCiRIk09V2nTh0g7ufN06dPJxvsrl27BsQdqRA/YhcdHc3y5cs5evQoXbt2TXFKZvzPv6VL\nl1aoe8wp2KUiT548LF68mMjISMvoW0xMDFWqVGHTpk2MHTsWgBEjRtC+fXuKFStGnjx5iIyM5MaN\nG2zfvp3JkycTGhrKG2+8wXfffWfLtyMiIiKSsxjA6GQPxqwJdnYPcY4dxAWwCxcu8M8//yTbZuvW\nrUDc5nve3t5p6rdcuXKUL1+e8+fPs3//fpo1a2a13bZt2wBo1KiR5ZqDgwMfffQRly5d4sKFC6xa\ntcrqs9HR0ezcuROAWrVqpakusR0Fu1Q4OTnRvXv3JNcjIyPp0KEDdnZ2XLx4kZIlSyZpU65cOZo1\na8Zbb71FpUqVmDJlCn379uWpp556BJWLiIiIPBkcnOwwpfFYg/QyOjxcv927d2fJkiWsXLmSEydO\nUL169QT3fX19mT9/PgC9evVKV99dunRh4sSJzJkzh9GjR1sGIeJt2bKFAwcOANCzZ88E9zp06MC0\nadNYvXo1x48fp0aNGkn6nzJlCj4+PgD07ds3XbXJo6fVjxnUv39/rl+/zsaNG62Gugflzp2b5cuX\nA3EHQIqIiIhI5jAYDNg72WfZx8OO2PXs2ZOnn34aiAtiV69etdwLCgqiU6dOhISEUKZMGUaOHJnk\n+c8++4wKFSrw4osvJtmAZdSoUbi5uXHx4kU6d+6c4HiFY8eO0adPHwB69OiRYMQO4K233rIcbP7c\nc89x+fLlBPdnzJhhmZk2bNgwGjZsmNFvgTwiGrHLgICAAFasWIGnpyetW7dO0zPPPPMM+fPn559/\n/uHWrVsUKpTxwy5FRERE5H8MYO9ojynLpmI+/Ejg/PnzqVOnDv/99x9ly5aladOmODo6snPnTsLC\nwgD49ddfrZ5Jt3PnTs6fP4+fnx/ffvstuXLlstwrWrQoP//8M7169WLTpk3ky5ePZs2a4efnx969\newEoVaoUc+bMSdJviRIlWL58OZ06deLSpUuUKVOGBg0aULhwYfbv329Zm9exY0d++umnh/4eSNbT\niF0G3L9/HyDdC0jjd9CMjLSyHa+IiIiIZEjWjtg9fLCrUKECFy9epFOnTsTExLBt2zY2btxIWFgY\njRo14uzZszzzzDNWny1cuDAQtxGLtaMQevbsyYEDB6hRowaBgYGsWbPGEuoGDhzIsWPH8PDwsNr3\ns88+y9mzZ2nbti0A+/btY/ny5Vy7do2CBQsyc+ZM1q9f/9DvXx4NjdhlQKlSpShatCjXr19n165d\nNGnSJNVn1q1bx507d6hQoUKqUzdFREREJG3ip2KasmhXzMwIdgDe3t6sXbuWe/fuceTIEUwmEzVr\n1kx2N8t4CxcuZOHChSm2qVevHseOHePatWucPHmSXLlyUbt27QSje8mpUKECmzZtIjAwkOPHjxMc\nHGzZmEWyFwW7DHr22Wf58ccfGTBgAP/880+KUytPnDjBgAEDAKxuxCIiIiIiGWfvZIfJLovOscuk\nYBfP09OTVq1aZWqf8YoVK0axYsVSb2iFh4dHmgYr5PGlYJdBH3zwAYsWLcLHx4fSpUvzwgsv0KbN\n/7V332FRXOsfwL+7y1KkSRMFC6CiYgWsUbFFxdiNHfUa9UaNMfYWExNjYpKbaJqJLbbEGrsmYu9e\ne2/YwAJYUHqH3fP7g9/OZdldQASWhe/neXiUOXNm3ll2Z+edc+acTqhWrRosLS2RkpKC0NBQ7Ny5\nE+vWrQOQdUfks88+M3LkRERERKWITAaFuRlkClEkm3/TUTGJigsTuwKqXLkyjh8/jnbt2iEmJgZL\nlizBkiVLDK7v4+ODw4cPS6MPEREREdGbk8kAhYUZ5KqiSezedFRMouLCd+obaNiwISIjI/H111+j\nbt26kMlkWuUymQyNGjXCsmXLcPPmTbi6uhopUiIiIqLSqygHT2GLHZkKtti9IUtLS8ycORMzZ87E\n8+fPce/ePWRkZECpVKJWrVpwcXExdohEREREpdf/D54iiqrFzpyJHZkGJnaFyNXVla1yRERERMVI\nJgMUlkWX2LHFjkwFEzsiIiIiMmlmSjOIIho8RWHGxI5MAxO7PMTGxmL48OG4cuUKpk+fjg8++AAA\nMGvWLOzdu1eaNDI/IiMjERgYiK+//rqowiUiIiIqU2QyGcwszSDURTV4ChM7Mg1M7PKQlpaG4OBg\npKen48aNG9LyM2fO4MqVK7hy5cprbc/BwaGQIyQiIiIqw2RZg6eIopmfnIkdmQwmdnkoV64cRo8e\njZCQEK1JG3v27AlnZ+fXGhzlxYsXaNWqVVGESURERFRGyZjYEYGJXZ5sbW3x888/6yyfOHEiJk6c\nWPwBEREREdH//P88dqJoemJCzlExyUQwsSMiIiIik6WZoBxFlNhxugMyFUzsitDVq1dx//591K5d\nG3Xr1jV2OERERESlkAwKCyUgZEWydTlHxSQTwcTuDS1evBi3b9/W6q6ZlpaG7t2748CBA9Kytm3b\nYsuWLXBycjJGmERERESl0/8PnlJULXacx45MBRO7N9C3b19s3boVSqUSM2fOlKY+GDFihFZSBwBH\njx5F69atcf36dSgUPEEQERERFQYZVFBYygAoi2b7bLEjE8HEroDWrFmDrVu3AgCaNWsGa2trAMCD\nBw+wfv16AECPHj2wdu1aXL58Gf3798ft27fx9ddf45NPPjFa3ERERESliswCCqsGANJR6Je2IglQ\neBbuNomKCBO7Avrzzz8BALNnz8aXX34pLf/7778BAAqFAosWLYKtrS0CAgLw559/olOnTti6dSsT\nOyIiIqLCIrMAzDsZOwoio5MbOwBTlJKSghMnTgAAJk+erFW2e/duAEDHjh1RpUoVaXnHjh1hZ2eH\nK1euICoqqviCJSIiIiKiUo+JXQE8e/YM6enpcHNzg6Ojo7T86dOnOHnyJADg7bff1qlXrlw5AEBS\nUlLxBEpERERERGUCE7sCsLCwAACoVCqt5cePH0daWhoAoE2bNlplaWlpiImJAQDY2NgUQ5RERERE\nRFRWMLErADc3N3h7e+P58+e4evWqtFzTDdPb2xuNGzfWqrNjxw6kpaWhUaNGcHZ2LtZ4iYiIiIio\ndGNiV0Bdu3YFAIwaNQq3b9/Gjz/+iHXr1gEA+vTpo7Xuzp07MXLkSABZz9oREREREREVJo6KWUCz\nZ8/Gxo0bceHCBfj4+EjLnZ2dMXXqVOn33r17Y8eOHQAAd3d3zJo1q7hDJSIiIiKiUo4tdgXk5OSE\n8+fPo0uXLtKyt956C6dOnYKTk5O0zMXFBUDWXHdnz56Fg4NDscdKRERERESlG1vs3oC7uzv27NmD\nmJgYqNVqrYROY9iwYejVqxfeeecdI0RIRERERERlARO7QpBbK1yrVq2KMRIiIiIiIiqLmNgVknPn\nzuHkyZN49OgR4uPj4ezsDE9PT7Rr1w516tQxdnhERERERFSKMbF7Q3v27MG4cePw8OFDg+s0btwY\nK1asQIMGDYovMCIiIiIiKjM4eMobmDt3Lrp27Soldc7OzmjSpAnefvttNGrUCHZ2dgCACxcuoGHD\nhti4caMRoyUiIiIiotKKiV0BBQcH4/PPPwcAdOrUCVeuXEFUVBTOnTuHAwcO4PLly4iNjcWJEyfQ\ntGlTAMCgQYNw7do1I0ZNRERERESlERO7Alq0aBEAYOzYsdi3bx8aNmyos45MJkOrVq1w9uxZ9OjR\nAwCwcOHCYo2TiIiIiIhKPyZ2BRAXF4c9e/YAAL7++ut81fnpp58AADt27EBGRkaRxUZERERERGUP\nE7sCiI6OBgA0aNAA9vb2+arj4eGBihUrIi4uDhEREUUZHhERERERlTFM7AqgSpUqUCqVuHv3LtRq\ndb7qpKenIyoqCjKZDK6urkUcIRERERERlSVM7ArAzMwMQ4cORWpqKr777rt81Vm+fDlUKhWGDRsG\nKyurIo6QiIiIiIjKEs5jV0ArVqzA48ePMXPmTCQlJeGLL74wuO7vv/+ODz/8ENWrV8eyZcuKMUoi\nIiIiIioLmNgV0NmzZ+Hp6QkAmDdvHhYvXoyuXbuidu3aUCgUEELgyZMn+OeffxAWFgYAcHd3x6ef\nfoq7d+9CqVRK23r69CnmzZuHtm3bGuNQiIiIiIjIxDGxK6CZM2fi6NGj0u8vX77EmjVrcq1z/Phx\nHD9+XG+ZZpJzIiIiIiKi18XEroACAwNhaWlZKAOhPHv2DDVq1CiEqIiIiIiIqCxiYldAM2bMwIwZ\nM4wdBhEREREREUfFJCIiIiIiMnVssSskoaGhOH78OO7fv4/U1FTY2dmhRo0aaN68Oby8vIwdHhER\nERERlWJM7N7QvXv3MGrUKIODogBAkyZNsGLFCtSvX78YIyMiIiIiorKCXTHfwN69e+Ht7a2V1FWt\nWhUNGjSAu7u7tOz8+fNo0KAB/v77b2OESUREREREpRwTuwJ68uQJ3n33XQCAm5sb1qxZg2fPnuHR\no0e4evUqHj9+jIiICKxatQpubm4AgIEDByI8PNyYYRMRERERUSnExK6A/vOf/yA5ORl169bFgwcP\nMGzYMK2pD+RyOdzc3DB8+HCEhYWhXr16SEpKwsKFC40YNRERERERlUZM7Apo3759AIDffvsNlpaW\nua5rbm6O3377DQAQHBxc5LEREREREVHZwsSuAB4/fox79+6hYsWKCAgIyFed1q1bo0KFCggJCWF3\nTCIiIiIiKlRM7ApApVIByOpu+To062dmZhZ6TEREREREVHYxsSsAT09PVKhQAZGRkbh3716+6jx4\n8ADPnj2Di4sLqlWrVsQREhERERFRWcLEroA6d+4MAPj444/ztb5mvY4dO0ImkxVZXEREREREVPZw\ngvICGjduHP78809s2bIFHTt2xLJly+Dp6amzXlhYGEaPHo0DBw4AAD744IPiDpVKILUQUMrlcLIx\nh5XSDJkqNYSxg8pGJgPUaiA2JR1J6ZmQ82YEUYkkkHWH1kImg0Img1qUpDNJ4ZAh6zjTAKiEAM9G\nRET6MbEroGbNmuHbb7/FjBkzcPDgQXh5eaFOnTqoV68eLCwskJqaips3b+L27dtSnfnz56Nly5ZG\njJpKgky1gKu9NdJTEvH73ycQHpuC8tYWUMhlKCnXZCq1ANSZ6NqkNurV8MTjVwmQ82qKqEQRAKwV\nCiSlpuB4WhqSAFgCpS7xUf//TwMzJdzLWSNerSp1x0hEVBiY2L2B6dOno3Llyhg1ahRSUlJw+/Zt\nrUROw8rKCsuXL0dQUJARoqSSRAjAxdYKTyKf4fPgO6jWthcsytnAqaIrlEozqNVqY4cIAFCamyMh\nMQmz1q/EsPBo9G/nj0cv49lyR1SCWMnleJ6YgBtCoMeQIbBzcoKLoyOUZiXnXFIYLCwt8erFC6xc\nsQLR0dGoZ18eCSomd0REOTGxe0ODBw9Gnz59sGvXLhw9ehSPHz+GEAIymQxVq1ZF27Zt0b17d1hZ\nWRk7VCoBFHIZFOp0fPHPdfT4dCm+nDDC2CHlauSYcRgYGADPm/dQvVplJKVxRFeikkAGICMtDZdU\nKizauAGjBww0dkhFbujo0ejk64fIuFg42NpBVVK6OBARlRBM7AqBpaUl+vfvj/79+xs7FCrhypkr\ncOpaCCo2alXikzoA8K3ujm4DhyN4/Tx8UsuDiR1RCaGQyXAnLRUt3nqrTCR1ANCoenX0HzkCq378\nEe1lMqQwsSMi0sJRMYmKkVIhx6uEFJR3dDZ2KPlWyb0y0oScz9gRlSByACkAnCtUMHYoxaqSWyWk\nofQ9R0hEVBjYYpeH9PR0HDt2DK9evUKDBg3g4+MDADh//jzCw8NRrly5fG8rOTkZlStXRpMmTYoq\nXCrhhAAUChlyGyUlOTkZ0dHRUpfezMxMlC9fHuXLl9daLzQ0FDY2Nqhg4MIuISEBz58/h6enJxQK\nhU55ZmYmHj58iEqVKsHa2tpgPJkZGZDJeGecqKSRAVnD1+ZBrVYjMTERNjY2kMuz7ufGxMQgLi4O\nZmZZlwFCCJibmwMA7OzstB4fePXqFeLi4uDl5ZXrfhITE6FWq2FnZ6dT9vLlSyQkJOgdPVqz/4cP\nH8LR0RH29vYG9yHUPBcRERnCFrs8xMTEoFu3bhg0aBAWLVokLZ8+fTr69OmDwMDAfP/06dMHM2bM\nMOLRUEkgkwbv1m/AgAGoUqUKqlatiipVqsDT0xNz586Vyk+cOIGKFSuicePGcHd3R6dOnZCamqq1\njblz58LOzg6NGzeGk5MT9u7dq1V+/PhxuLq6olGjRrCxscFvv/1mMB4BkVu4RGRE+flobt26FQ0b\nNkRycrK0bMaMGfD09ESVKlVQpUoV1KxZEz4+PqhYsSJOnTolrTdu3Dg4OzvD19cXXl5euHbtmsH9\n1K5dG2PHjtVZPmTIELi4uKB+/fqoWbMm7ty5o1V+/PhxODg4oH79+qhcuTI+//zzNzpeIqKyiold\nHmQymXSHM/vE4tmX5ffndfDLq2zStBBPmjQJt2/fxsGDB7F9+3aMGTMGAHD37l0EBASgZcuWiIiI\nwIULF3D48GG0a9dO2sby5cvx+eefY8WKFYiNjUWvXr3QpUsX3Lt3DwDw7NkztGnTBoGBgUhMTMTC\nhQsxbtw4/P3330Y5ZiIqOjt37kT//v2RkJAACwsLafmkSZOwfft27Nu3D0ePHsX169dhY2MDNzc3\nvP322wCyErLffvsNhw4dQkxMDGrUqIFGjRohPj5eZz9z5sxBREQEnJ21u5mPGTMG69atw/HjxxEV\nFQVbW1s0adIEKpUKAHDr1i20adMGzZo1w6tXr7B582bMnTtX62YWERHlD7ti5sHJyQn3799Hamoq\nnJycpOUbN25EUlKS1I0lPzIzM3Pt8padTCbLainhkwRlyvHjx5GSkoJp06ahUqVKqF27tlb5L7/8\nAgcHB2zduhUA0LBhQwQHB6NTp064d+8eatasiR9++AHt2rXDiBFZg7OsXr0amzZtwpo1a/Dll19i\n3rx5sLW1xbp16wBkXeAFBwdj2rRp6NatW/EeMBEVmV69emHnzp3w8vLSucFYp04d1KlTR/p9/fr1\nePz4Me7fvy8tu337NubPn4/27dsDAFatWoXKlStj//796Nu3r7Terl27sH79ejg6OiI2NlYrht9/\n/x2zZs1C69atAWS1Hnp5eSE4OBjdunXDZ599BicnJ+zbtw8AEBgYiMWLF2PixIn4+OOPoVQqC/11\nISIqrZjY5UGhUMDd3V1nuYuLC1xcXIpsv0qF/P+77FFZEhISAiEEvvjiCxw5cgSWlpZYtGgRWrVq\nBQAYO3aslLBpJCYmAoB0N75t27ZYvXo1VCoVFAoFzp07h9TUVDRv3hwAEBwcjA4dOmhto2vXrpg2\nbRoyMzNf62YFEZVMr169wqtXr/DkyROcP38e/fr1gzDwbG9cXByCgoIwa9YsVK9eHUDWM28XLlzQ\nSgZ37twJIOuGkkZqaioGDRqEzZs3Y+vWrYiIiNDadpMmTbBnzx7Mnz8fALB//34AQN26dQFkJY85\nn7tr27Yt0tLScPr0aQQEBLzJy0BEVKawK2YJ8ejRIyxevBhr1qzBX9t2IOTOHVia8wK7rLl+/TpU\nKhWSk5MxY8YMODo6onXr1lILnY+PD3x9fbXqDB06FO3bt0fVqlUBAL/99hvat28PMzMz+Pj4oFmz\nZpg3b57UGhcZGakzOEGNGjWQkZGB8PDwYjhKIipqTk5OOHHiBCpXrozIyEiDSR0ATJkyBWZmZlLy\nBUCrhW/dunWQyWQYN24cli5dipo1a0rrBQYGws/PD++88w5CQkJ0Hjs4efIkkpKSIJPJ4OXlhTFj\nxmDHjh1SMufj44MnT55o1Tl+/DiA/920IiKi/GHmkIf09HScPn0a6enpb9ySkZKSgkqVKulcmAPA\nnTt3sGLFCshlMkQmZCCwpgPeqqMEoHqjfZJpmTp1Kj766CPpbvZ7772Hvn37YujQoXj33Xe11hVC\nSBdH2Z+P27x5M86dO4cxY8bA398fGzduxLp16zBq1ChUrFgRAHRGydS8tzMyMors2Iio5MnIyMCK\nFSswc+ZMg+tUq1YNX3zxBS5duoTJkyfD398f/v7+WLp0KUJCQhAZGQkgKxnM+T35zTffICEhAZ98\n8gkqVaqEpUuX4pdffkFgYCAsLCzw1VdfwdvbG61atcKyZctw6NAhKcHMLRklIiJdTOzyEBMTg06d\nOiE9Pb1QtteuXTscPnxYZ3nr1q2xY8d2yBVmqFSxItb//DUyHh0slH2S6ch+J1yjT58+2Lp1K8LD\nw1G5cmUAwNOnT1GtWjXY2Njg+fPn0tDkycnJ6N+/P2bPno0vv/wSADBq1Cj4+vqia9euuHjxIqyt\nrZGWlqa1D81gCA4ODkV5eERUwqxfvx4A8MEHHxhcp1WrVlJ38Dp16mDKlCk4evQoxowZg2+//RZy\nuRwvXrxAfHw8VCqVNLXClStX8Mknn2Dnzp3o0aOHtB9LS0uMGTMGq1atQs2aNXHz5k106dIFvr6+\naNeuHX755ReMGTMGlpaWRf8CEBGVIuyKWUJYWVmhcuUqcKtUCTKZDM4OdkjNyDR2WFTMZs+ejSNH\njmgtS0lJAfC/Z+hu3rwJNzc3NGrUCNHR0VoD8jx79gwA0LRpU61ttGnTBpcuXQIABAQEICQkRKv8\nv//9L2xsbHRGtCOi0m3FihWoXbs2qlSporU8MTERS5cu1ekmGRAQgPDwcJw5cwYA8Nlnn0Emk8HV\n1RXXr1/HmTNnpLk14+LiAAB+fn5a22jQoIFUPz4+Hh4eHnj06BFSUlKwd+9eCCEQFxcHf3//Ijlm\nIqLSii12eXB2dsa9e/egUqmkKQ4KKjMzEzY2NvlaNz1DxT9OGbRjxw4sW7YMUVFR0rKZM2eiTZs2\ncHFxQXR0NOrVq4dRo0Zh+fLlOvW9vLxgaWmJP/74Q7pDDmSN4qoZ2a5nz5547733EBoaKk04vHLl\nSvTq1atoD46ISpxr165h+PDhOss1rWpBQUFYu3YtACAhIQHLli3DuHHj0Lx5c9y+fRspKSmQyWRw\nd3dHu3btYG1tLY1wqemBsGHDBkybNg1AVi+Y8+fPS10/16xZgwkTJiA+Pl76fpw1axbatWuH8uXL\nF/HRExGVLswd8qBQKKRBKYpT1kQHHBWzrAkODkbt2rXh4OCAXr16YcuWLbCwsMD27dsBQJr89969\ne2jZsiWioqKQnp4OCwsL7NmzB9WrV8eePXvQvn171KhRAwEBAfjrr79gaWmJTZs2AQCGDx+OrVu3\nonr16hg8eDB27doFJycnrFy50mjHTURFRy6XQ61WIz09XWv6gKdPnyIuLk7vc99mZmbYuXMnevbs\nievXr6Nhw4b4888/0aBBAyxatAgAdKZjsbW1hYODg5SQubm54ddff8W4ceOwdetW1KhRA+vWrUOL\nFi3w9ddfAwBGjx6NH374ARUqVMDAgQOxbds2WFlZYdu2bUX0ahARlV7sillCMaUrrbIGAzDU+lu1\nalVERUVh7NixCAkJwZQpUxAZGSk9+9asWTO8//77cHd3h7OzM+rXr4+6deuiTp06UlfNdu3aISIi\nAp06dcLdu3cxa9YsPH36VKub5e7du7Fs2TLcv38f48aNw507dzhfFJEJUuSjJ4mvry8+/vhjnWfW\n5HI5Jk2aJD0/l1OPHj0QFhaGRo0aISwsDEuXLsXVq1cN7mfo0KHo37+/1rIPPvgAV69ehYeHB8LD\nw7FixQr897//lcrNzc1x7949zJgxA7du3cLkyZPx5MmTfPduISKi/5EJIWIB2Oe1IhkWExODdevW\n4eTJk3jw4AHS0tJgb28PLy8vNG7cGAMGDJCeOciv3St+gvWVbcgUMqiLKG4qfg7lzBF86hKuu7bG\n1m3bjR1Ovny7aDnOLPwQXw5uhyexqcYOhwqRg1KFM+ku2LjzGiwjw6DI9rwmlWzmMhkuJ8SjSbdu\n2LV7t7HDKTZfffcffDZ9Bt6xs0eymt+OpkJkZkJmbo7Jq1fhnd69jR0OUWkVx66Yb2jJkiVS97ic\nTp48iT/++AMfffQRFi5ciEmTJhVzdFTSKBUypKgEHkU8M3Yo+SeTQc1Rx4lKFDlkUAGIio42dijF\njj1aiIj0Y2L3BubMmYN58+YBAJRKJfz9/VGvXj3Y2toiNjYWN27cwOXLl5GZmYnJkyfj1atX0hD0\nVDYlpmWitrsTjtyLxYvENFSwsTB2SHm6fvUKnCzlyODNcaISIwMCLgDin79AiloNqzcc3MtU3Lh2\nDXYAVJzjjohIR9n4JigCx44dk5K6Pn364MWLFzh9+jSWL1+OhQsXYuXKlTh37hyeP3+OPn36AAC+\n+uorHD9+3Jhhk5ElpavQtE51VE19jJ6D30OSsQPKwyc/r8L5LSsQ1NYXr5IKZy5HInpzKiHgZWOL\n8Af3MXzMGGOHUyy+WbYMu9aug5/SHOlM7IiIdLDFroAWL14MAPjXv/6F1atXG1zP0dERW7duxYgR\nI7Bq1SosW7YMAQEBxRQllUTPE9Mxs1cLfLN1G96qUQ2e3rUgVJkQJeRCxcxMgZiEFMTHx0HxNARf\nda0HB0cHPItNhFzGTlBEJUW6TIaW1jb4Z/lyVPtnD2rVrIHMzJJzLikMcrkcqRkZiI6KwoPQUDSx\ntISluQWS1Wp2ySQiyoGJXQGoVCrs378fAPD555/nq85nn32GVatWSfP7UNkkA5CakQkhzDB3cHuE\nPHyCR89uQKkwg6yEtJ/LASRZqGBb1QItAtsiBWaIjEmEmZyXUUQlSaYQgEKBDja2eBIZgZDIiFL5\npZ4JoByAjpZWEOYWSFarmNQREelRGr8Ditzjx48RExMDNzc3eHh45KtOtWrVULFiRTx79gwPHz7M\ndz0qfeQyGTJUKjyJSYajkwtcK7iiJN1fFwDM5DKo1AJPkzIhkMmkjqgEkgFQqdVIAuBiY4cKpfRj\nKgOgBpAmBMCkjojIICZ2BaCZ7ysjI+O16mVmZgLImviVCADSMtVIyyzZo5LwIoqo5FNBoETdISIi\nomJXQjp/mZbKlSvDx8cHUVFRUpfMvBw4cAAvX75EnTp1ULly5SKOkIiIiIiIyhImdgXUtWtXAMDY\nsWMRExOT67qxsbHSXHedO3cu8tiIiIiIiKhsYWJXQNOmTYOTkxNCQ0Ph5eWFhQsX4s6dO0hOToZK\npUJycjLu3LmDBQsWwMvLCw8ePIC9vT2mT59u7NCJiIiIiKiU4cNeBeTi4oLt27cjICAAsbGxmDJl\nCqZMmQJ7e3tYW1sjKSkJcXFxWnV27tyJSpUqGSliIiIiIiIqrdhi9wZat26Nx48f491335WWxcXF\nITIyUiup69OnDx49eoQ2bdoYI0wiIiIiIirl2GL3hqpUqYItW7YgOjoap06dwqNHj6BSqSCXy1Gt\nWjW0atUKjo6Oxg6TiIiIiIhKMSZ2hcTR0RHdu3c3dhhERERERFQGsSsmERERERGRiWOL3RvaunUr\n/vrrLyQkJECtVkMmMzydc2RkJAIDA/Htt98WY4RERERERFTaMbF7A127dsWePXteq46Tk1MRRUNE\nRERERGUVE7sC+vbbb6WkzsPDA61bt4a1tTXMzMygVqv11omKikLLli2LM0wiIiIiIioDmNgV0KZN\nmwAAPXv2xI4dO4wbDBERERERlWkcPKUAnjx5gsuXLwMAvvvuOyNHQ0REREREZR0TuwLIzMwEALi5\nuaFmzZpGjoaIiIiIiMo6JnYF4OnpCTc3N0RGRiIyMtLY4RARERERURnHxK6ABg8eDADYsGGDkSMh\nIiIiIqKyjoOnFNB3332HGzduYOrUqahUqZKU6BERERERERU3JnYFdOXKFTRu3Bh79+5FUFAQPv30\nUzRs2BAODg6Qy+V6pzx4/vw52rRpg2nTphkhYiIiIiIiKq2Y2BXQpEmTcPToUen30NBQhIaG5lkv\nOTmZiR0RERERERUqJnYFFBAQgLS0NFSqVCnfdZ4+fYrWrVsXYVRERERERFQWMbEroLlz52Lu3LnG\nDoOIiIiIiIijYhIREREREZk6JnZEREREREQmjl0xC+j777/H8ePHUaFChXytL4SAXC6HtbU1bG1t\nUa9ePXTp0gV2dnZFHCkREREREZV2TOwK6J9//tEaFbMgZDIZfv/9d4wYMaJwgiIiIiIiojKJiV0B\nzZ07Fz/++CO2b98OALCzs8Pbb78Nb29vKJVKJCUl4cqVKzhx4gQyMjIgl8vRt29fJCYmIjw8HA8e\nPEBSUhJGjhyJzMxMvP/++0Y+IiIiIiIiMlVM7AooICAAc+bMAQBMnjwZCxYs0LtefHw83n33XRw8\neBB+fn6YMWMGACA9PR0jR47E2rVrMXPmTPTv3x/ly5cvrvCJiIiIiKgU4eApBTRu3DgcO3YMH3/8\nscGkDshqyTtw4AD8/Pwwc+ZMnDx5EgBgbm6OP//8E+3bt0dMTAx27dpVXKETEREREVEpw8SugDZs\n2AAgq7UuP2bOnAkA2LRpk9byIUOGAADOnTtXiNEREREREVFZwsSuAMLCwhATEwM3Nzc4OTnlq06l\nSpUAAJcvX9Za7urqCgB49OhR4QZJRERERERlBhO7ArC1tQUAREZGIjo6Ol91wsPDAQD29vZ6y83N\nzQsnOCIiIiIiKnOY2BWAs7Mz2rRpAwD44Ycf8lVHs56mnsahQ4cAAI0bNy7ECImIiIiIqCxhYldA\no0aNAgB8+eWXmDRpEoQQeteLjIxEmzZtcO7cOVhaWmLw4MFS2aZNm7Bw4UIAQKtWrYo+aCIiIiIi\nKpU43UEBDRkyBIcPH8aqVavw448/Yvny5QgICICnpyfkcjlSUlJw5coVXLx4UaqzfPlyVK5cGQAw\nZcoUKanr1q0bWrdubZTjICIiIiIi08fE7g2sXLkSdevWxdSpU5GUlITg4GC967m5uWHTpk1arXJn\nzpwBALRv3x4bN24slniJiIiIiKh0YmL3hqZMmYIRI0Zg27ZtOHToEB4/fgy1Wg2FQgFvb2/07t0b\n3bp106qjUqkwdepUuLu7o2nTpkaKnIiIiIiISgsmdoXAwcEBI0eOxMiRI/O1vkKhQO/evYs4KiIi\nIiIiKis4eAoREREREZGJY4tdIUhJSUFycjLUajVkMpm0XAgBtVoNlUqFlJQU7Nu3D3FxcZg1a5YR\noyUiIiIiotKGid0buHz5MiZOnIgLFy5ACAG5XC4ldmq1WlovIyMDGRkZAIB27doxsSMiIiIiokLF\nxK6AQkJC0KRJE6hUqnytL5PJIIRA/fr1izgyIiIiIiIqa/iMXQEtWLAAKpUKSqUSn3zyCS5evIjx\n48cDAAYMGICzZ89iw4YN+PDDDwFkdcvcvn07fvrpJ2OGTUREREREpRATuwI6efIkAGDRokWYN28e\n/Pz8MGbMGABAXFwcmjZtioEDB+KXX35BSEgIAGDChAlGi5eIiIiIiEovJnYFEBkZiZCQECgUCgwY\nMEBa7uPjA1tbW+zduxfJycnS8lq1amHdunV4/Pgx5s+fb4yQiYiIiIioFGNiVwBpaWkAAFdXV9jb\n22uV1axZEwBw//59reX9+vUDAPz+++/FECEREREREZUlTOwKwMbGBgCQnp6uU+br6wsAuHv3rtZy\npVKJevXqISwsDC9fviz6IImIiIiIqMxgYlcALi4uqFGjBl6+fImwsDCtsqpVqwLImgohJ00imJiY\nWPRBEhERERFRmcHEroDat28PAJg9e7bWcn9/fwDAxo0btZZfunQJd+/ehZWVFVxcXIonSCIiIiIi\nKhOY2BXQ6NGjAQAbNmxA+fLlcfToUQBA27Zt4eTkhNDQUHTs2BGHDx/G+vXr0bNnTwBAx44dYW1t\nbaywiYiIiIioFGJiV0B+fn5YsmQJgKzpDe7duwcAsLa2xrx58wAABw8eRIcOHRAUFITw8HAAwOTJ\nk40TMBERERERlVpM7N7A6NGjcfXqVfTu3VvqggkAY8eOxe+//w5LS0tpmaOjI/755x+0adPGGKES\nEREREVEpZmbsAExdgwYNsG3bNp3lI0eORFBQEM6ePQsLCws0adIECoXCCBESEREREVFpx8SuCFla\nWrKFjoiIiIiIihy7YhIREREREZk4ttjlITY2FsOHD0dsbOwbj2YZGRmJwMBAfP3114UUHRERERER\nERO7PKWlpSE4OFiaXPxNOTg4FMp2iIiIiIiINJjYvSYvLy9069YN5cuXx9OnT/HixQuoVCrIZLI8\n60ZGRqJFixbFECUREREREZUlTOzyYGdnh5o1a+LmzZsAgLCwMERERKBr167o1KmTkaMjIiIiIiLi\n4Cl5srKywo0bN3D69Gl89NFHcHR0xNatW9G5c2fIZDJ06NABmzZtQmpqqrFDJSIiIiKiMoqJXT41\nb94cP/30E16+fImTJ0/i/fffh7W1NQ4fPoyBAwfCysoKHTt2xF9//cUkj4iIiIiIihUTuwJo2bIl\nli5disTERBw5cgT/+te/YGFhgYMHD2LAgAGwsrJCjx49sGvXLmRmZho7XCIiIiIiKuWY2L2htm3b\nYvXq1UhJScH+/fsxaNAgmJmZYffu3ejZsyfMzc3Rv39/7N+/H0IIY4dLRERERESlEBO7QiKTydCx\nY0esX78eqamp2L17N4KCgmBjY4PNmzejc+fOUCqV+P77740dKhERERERlTJM7IqAQqFAt27dsHbt\nWoSFhSEwMBAAoFKpsGfPHiNHR0REREREpQ2nOygCMTEx2LZtG7Zv345jx44hMTFRKqtataoRIyMi\nIiIiotKIiV0hefXqFbZv346NGzfi0KFDWmXVq1dHUFAQBgwYAB8fHyNFSEREREREpRUTuzcQFRUl\nJXNHjhzRKvPy8pKSubp16xopQiIiIiIiKguY2L2m58+fY9u2bdi0aROOHTumVebp6YnBgwdjwIAB\nqF+/vpEiJCIiIiKisoaJXT48ffpUSuZOnDihVebh4YFBgwZh4MCBaNCggZEiJCIiIiKisoyJXR6i\no6NRp04dxMXFAQDMzMzQsWNHDBgwAH379oW1tbWRIyQiIiIiorKOiV0eMjIykJKSIv3erFkzvP32\n27Czs8P27dsRERGBpKQkyGQyyGSyXLcVHR0NPz8/DB8+vIijJiIiIiKisoSJ3Ws6deoUTp06VeD6\n7dq1Y2JHRERERESFioldHhQKBapUqYJXr17B1tYWQogCbys8PBwuLi6FGB0RERERERETuzw5Ozvj\n9u3bhbItIQTkcnmhbIuIiIiIiEiDiV0+KJVKY4dARERERERkEJuPiIiIiIiITBwTOyIiIiIiIhPH\nxI6IiIiIiMjEMbEjIiIiIiIycUzsiIiIiIiITBwTOyIiIiIiIhPHxI6IiIiIiMjEMbEjIiIiIiIy\ncUzsiIiIiIiITBwTOyIiIiIiIhPHxI6IiIiIiMjEMbEjIiIiIiIycUzsiIiIiIiITBwTOyIiIiIi\nIhPHxI6IiIiIiMjEMbEjIiIiIiIycUzsiIiIiIiITBwTOyIiIiIiIhPHxI6IiIiIiMjEMbEjIiIi\nIiIycUzsiIiIiIiITBwTOyIiIiIiIhPHxI6IiIiIiMjEMbEjIiIiIiIycUzsiIiIiIiITBwTOyIi\nIiIiIhPHxI6IiIiIiMjEMbEjIiIiIiIycUzsiIiIiIiITBwTOyIiIiIiIhPHxI6IiIiIiMjEMbEj\nIiIiIiIycUzsiIiIiIiITBwTOyIiIiIiIhPHxI6IiIiIiMjEMbEjIiIiIiIycUzsiIiIiIiITBwT\nOyIiIiIiIhPHxI6IiIiIiMjEMbEjIiIiIiIycUzsiIiIiIiITBwTOyIiIiIiIhPHxI6IiIiIiMjE\nMbEjIiIiIiIycUzsiIiIiIiITBwTOyIiIiIiIhPHxI6IiIiIiMjEMbEjIiIiIiIycUzsiIiIiIiI\nTBwTOyIiIiIiIhPHxI6IiIiIiMjEmRk7ANJPlZkBa3kmMoQMasiMHQ6VcEIASjM55DIZ0jNVxg6H\nTISNXAWlXECVlgG1UEOWkmrskMhEyOUyCAEIIYwdCpkAocqETKWGit9PREWKiV0JpbAsh5A0a8gU\n5uDXJuVGCAELpRLp6ZlIV2XCxsKCyR3lS5RcIDrDHOXdXWFpI4fMwsLYIVEJJ5PJIIcMqRkZsFCa\nQQYZMgVvP1Lu1CoVVGo1LCx5jiEqSjIhRCwAe2MHQtrS01KRnJwMhVxh7FCoBBMQkMvlsLG1w+Tp\nM7F77wFcPHEYlpYWSE1N48UW5U4GqFRqCJHVAkOUF/vy5XHr5k0MHjkSMyZOxKCBAxEXG2vssMgE\nqNVq2NjYQGlubuxQiEqrOLbYlVDmFpYwt7A0dhhkQmQQUEIFO/us+zR8/xBRUbAsVw5IT4eZUgkg\nK9kjIiLj4+ApRKWEAgJWSjlUarWxQyGiUiwzIwMKmQzqzExjh0JERNmwxY6olKhbty7S0tKgkPN+\nDREVHXt7e9SqVQvu7u7GDoWIiLLhM3ZERERERESmLY639omIiIiIiEwcu2KSUahUKhw9ehTnzp3D\n/fv3kZqaimrVqsHPzw+dO3eGra2twbohISE4duwYvL290a5du9fa7+PHj3HgwAFUqlQJgYGBkJeB\nbotRUVHYt28frKys0KVLF5QrV87YIREVm/379+PVq1dZo8fa2KBz584wM8vfV9+pU6cQFhYGAKhW\nrRpat25dlKGWeBEREdi/fz+cnZ0RGBgI5f8PnqJSqbB37148e/YMHTt2RNWqVY0cacHdvXsXR48e\nRfXq1dGhQ4fXqnv79m0cOXIEderUee3vJiKiQiGEiBVExejHH38Ujo6OAoDeHzMzM/H1118brP/Z\nZ58JAKJevXqvve/ly5cLAMLZ2VkkJia+yWGYjAMHDkiv7Z07d4wdDlGxSU9PF5UrV9Y6v5w6dSpf\nddPS0oS7u7tUr0WLFkUcbcm3fv16AUCUK1dOvHz5UlqekJAgndOXLVtmxAjf3Pz58wUAUaNGjdeu\n++mnnwoAws/PrwgiIyLKUyxb7KjYxMXFoW3btrhy5QoAwNraGm3btoW3tzcUCgXu3r2Lw4cPIzEx\nEbNmzcKRI0ewb98+ne1o7rabF2AuHM0dZrlcDoWibMwRmP0489tSQVQamJmZwSLHpOu7d+/GW2+9\nlWfd/fv3IyIiQvo953bKIs35U6FQlJnzJxGRKeFVHhULIQRatWqFGzduAACmTp2Kb775RufiIC0t\nDcOGDcNff/2F/fv3Y/jw4Vi9enWhxyOTlZ3JmLMfa1k6bqLs73crKyukpKTg77//xtdff51n3e3b\ntwPIugmk5hQiubKyssKqVasQGRmJTp06GTucN9KnTx/Y2dmhZs2axg6FiOi1MbGjYjF+/HgpqVu+\nfDlGjRqldz0LCwts2rQJarUaW7ZswZo1azBw4EAEBgYWZ7hEVMp0794dp0+fxo0bN3DmzBk0b97c\n4LopKSnYuHEjHB0dUa9ePRw/frwYIzU9CoUCPXr0MHYYhaJWrVqoVauWscMgIioQJnZU5O7fv49f\nf/0VADB48GCDSV128+fPx5YtWwAA69atyzOxe/ToEU6dOoXU1FT4+vrC19f3jWJ+8uQJzp07h+jo\naHh4eKBZs2aws7MzuH50dDTS0tJQoUIFKBQKnD59GteuXUNAQADq1Kmjt87Nmzdx+fJlpKenw8fH\nB/7+/lJXp5xSUlIQHR0NGxsb2NtnzU5y7do1nDt3Dra2tmjZsiUqV66c7+OLjY3F8ePH8fTpU9Ss\nWRNvvfUWLC0ttdZJTU3FixcvYGFhAXt7e53y7NLS0vDy5UvIZDJUqlRJp2UwJiYGZ86cQXh4OCpW\nrIjmzZvDxcUlzzifP3+Os2fP4tmzZ6hSpQpatGiB8uXL5+sYIyIicP36dURGRkKpVMLLywu+vr75\nHjzmzp07uHDhAlJTU1G3bl00btyYXVlNWLNmzeDg4IClS5di9+7duSZ2+/btQ3JyMkaOHAkrK6t8\nJXb37t3DrVu38PLlS1hbW6NWrVqoX79+vt4z9+7dw5kzZ6BWq9GyZUvUqFEDQNbnNCkpCU5OTlqf\nv4SEBMTFxcHR0VF6P9++fRvnzp2TjrV27dp57hcAwsLCcP78ecTHx6N69epo2rQprK2t81VXQwgh\nnQMdHR0NnisSEhJw9uxZhIWFwcHBAU2bNs11oJWkpCTExMSgfPnysLGxAQCcO3cOly9fhouLC1q1\naoUKFSrkGd/Dhw9x8eJFREdHo3LlyvDz84Orq6vedZOTkxEbGwsrKys4ODjoXSctLQ2nTp3C3bt3\nUbt2bbRp0+a1ekO8ePEC165dQ3h4OGQyGapVqwZfX1/p3E5EVGCCg6dQEZs3b95rD1wghBDLli0T\nv/76qwgLC9O7vU6dOgmVSiV69OihMwBLtWrV9O5r9erVAoBwdXUVKSkpOuXXr18XLVu21Duoy9ix\nY0VycrLeWOvWrSsAiFu3bolffvlFq96MGTO01t21a5eoWrWqzvZtbGzEzz//rHf7ixcvlmIQQgh/\nf3+d+j169BBRUVE6dY8cOSIACHNzc/H8+XOxYMECnboWFhZ6Bz3w8fERAMS4ceP0xqURFBQkAIjh\nw4drLX/58qUYMmSI3tdzyJAhBgewef78uejVq5feemPGjBFpaWkGY9m/f7+oV6+e3roKhULMmTMn\n12M5ceKEqFWrlk5dW1tbsWLFilzrUslTvXp1AUCsXr1a7NmzRwAQ9evXz7WO5v18/PhxMWfOHAFA\ntG3bVu+6K1euFG5ubnrfb+XLlxerVq0yuJ8bN24IPz8/nXodO3YUQgjRu3dvAUCsW7dOq97o0aMF\nAOn9qPmcZv/x8/MTISEhBvd95swZ4evrq1NPLpeLGTNmCJVKpVNn8+bN0mchJiZGWp6cnCwNNJMz\nViGESEpKEuPGjdP7GrVo0UJcv35db4ya114zmJaHh4dO/eHDh+s9lwshxLVr1/S+vgDE22+/Le7f\nv69TR3N+NDQAym+//SasrKy0tuXs7CzUarX49ddfc6179uxZ0aJFC4MDh40bN06kp6frrUtElA+x\nTOyoyLVu3VoAELVq1SqU7WkSu+rVq4s6deoIAMLT01OMHj1atGvXTuuL8vz581p1c0vsNBctAISl\npaXo16+fmDp1qmjVqpW03NPTU7x48UInppo1awoAIigoSJQrV04rhl9//VVab8aMGdLyihUrin//\n+99iwoQJonbt2tLynj176mz/559/FgDEW2+9JZo2bSoAiICAAPHdd9+JYcOGCYVCIQAId3d3nfg0\niZ2tra10IWdnZyeCgoJEv379tC5SciaWU6dOFQCEm5ubSE1N1fv3iIiIEEqlUgAQW7ZskZaHhoYK\nBwcHrQup6dOni8DAQGlZvXr1RGZmptb2rl+/rvUadunSRUybNk107txZWlarVi0RG6t76sqeVFes\nWFEMGzZMjB8/XgQFBYny5ctLZaNHj9Z7LJoLM81F+fDhw8WkSZNEgwYNpOWjRo3SW5dKJk1i9+OP\nPwohhPR3PHPmjN71ExIShEwmEwCEEEJKSPQldmPGjJG25+3tLf7973+L8ePHi549e0qfCQDil19+\n0al78uRJqdzKykoMHjxYTJo0Sbqp0LBhQ+n/q1ev1qo7fPhwAUAMGjRISlwaNmwoxo4dK5o0aSJt\n18HBQTx9+lRn38uWLdO6YREUFCSmTJkinVs028t5I8tQYpeYmCicnJwEALFy5UqtOk+fPtW6kdWq\nVSsxdepU0b9/f2FhYSEt3759u06cM2fOFABE+/btpZs177zzjli4cKF49913pboNGjTQOY9cvnxZ\n+jtWqFBBDB8+XEyYMEF069ZNyOVyAUBYW1vr3Dj85ptvBABRu3ZtnXg+/PBDaZ8+Pj5i6tSpUhze\n3t7STUZ9id2WLVu0zi0DBw4UH374oXjvvfdExYoVpbIePXro1CUiyicmdlT0PD09BQDRp0+fQtle\n9hZAAGLz5s1a5QcPHpTKhgwZolVmKLG7c+eOlBw1bdpUJCQkaNU7cuSIsLS0FABEhw4ddGLKfse8\nadOm4s6dO+Lp06di5cqVIjw8XAghxLp166R1/v3vf+ts44cffpDKZ82apVWWsxUwZyvA9evXhY2N\njQAgevXqpRN79ro5WxCjoqKki9/q1atrlV27dk2qt2PHDp2YhfhfMpRzeHDNnWknJydx7do1rbLD\nhw9L2504caK0PCMjQ3h7ewsAwsPDQ+ei6+bNm8LZ2VlvAnzz5k1pmzlbDjU6duwogKzWyydPnmiV\nnThxQqrfv39/nbqaqTIMXahTyaR5b3/66adCCCE6dOggAIhPPvlE7/qa5GXgwIFCiP+13uVM7LJf\nqP/nP//R2c6LFy+El5eXACDq1q2rVZaeni6V1atXT7x69Uqr/KuvvtL6zP7xxx9a5SNGjJDKXFxc\nxIULF7TKsydumuPWOH36tFTWqVMnkZGRoVW+Y8cOqbxv3756Xxt9LXaaz2XOWNu0aSN95o4cOaJV\nFh8fL5o1ayYACKVSKe7evatV/vHHH2u9DsHBwVrl2c9tmt4MGprWzs6dO4ucrl+/LqytrfV+R3z7\n7bd6/2Zbt26V9pXzvRMVFSWdt/Qldi9fvpRuLAUGBurEI4QQQ4cOlernvCFJRJRPTOyoaEVERAhz\nc3MBQMyePbtQtpk9sTN0ga3pqpSzldBQYvfee+9Jd3ajo6P1bnPbtm3Sfrdu3apVpknsLCwspEQu\nJ013TX0XGhrTp0+XLnIePXokLc+e2E2bNk1v3ewtjhcvXpSWZ7/40ZewCCHE2rVrpXVu3bqlVaa5\nMBs2bJjeum3bttVJGHft2iVt78CBA3rraS7aHB0dRXx8vM5x5kwGNbIn7kePHpWWa7qrli9fXm+X\nVCGEOHXqlFT32LFjWmXdunUTAESTJk301hXify2uVatW1WkhoJJJk9jNnDlTCCHEokWLBJDVIqXP\ngAEDBACxbds2rd9zJnaalrymTZsa3PfSpUulpOb58+fSck0MAHSSGY3syVtuid2+ffv01u/UqZPe\n803Pnj2lFqacSZ1G9psY2T9jr5vYZU+GsrfmZxcdHS21WI0YMUKrLHtip2lxzUnTm0Eul2udezU3\nFJcuXaq33jfffCNsbGx0umYbSuw058F33nlH7/Zu3bplMLHLfj68efOm3voPHjyQ1lm/fr3edYiI\n8hArB1ERSkxMRHp6OgAYHBikoKytrTFgwAC9ZQ0aNACQNfhAampqrtvJzMzE/v37AQCjRo0y+MB8\n79690aRJEwD/Gwo9p8DAQLi7u+ssP3/+PG7evAkA+PDDDw3G8uGHH8LMzAwZGRn4559/9K4zevRo\nvcv79u0rDbqwa9cuvesMGzZM7/L69etL/3/16pVWWb9+/QAAO3fuRGxsrFbZ7du3cfToUQDAu+++\nKy3fvXs3gKxBHN5++229+xwxYgTmzp2LxYsXS3OEbdq0SYoze0zZdejQAS1atJBiyr69R48e4ezZ\ns3B2dtZbt2LFitL/U1JSpP8/fPgQf//9NwBg0qRJeusCwMiRIwEAjx8/xoEDBwyuRyVX165dAQBX\nr17F+fPntcpiY2Ol92CXLl1y3c6CBQtw9+5daX19KlWqBADIyMjQOg/t2bMHADBgwACDw+qPHTs2\njyMBvL29DU4v4O3tDSDrHKzx8uVLHDx4EEDWecTQwC4jR45E9erVARg+1+WH5hzm6+urdX7IzsHB\nAe+//z6ArPNG9s+lhpOTk/TZyykoKAg2NjZQq9Va50wrKysAwIYNG/TWmzFjBhISEjB37tw8j+PR\no0c4duwYAOC9997Tu06dOnWkc2VOgYGBiIiIwK1btwwOalOhQgVpYCh9rwERUX4wsaMiZW9vL33B\nahK8wuLq6mpwZEUnJycAWSM7ZmRk5Lqdy5cvSxMRt2zZMtd1NYmdZpJ1DSEEAEgXQzmdOXMGQNbF\nRtOmTQ1uv0qVKlJSeuHCBZ3yxo0bG9wHANSrVw8AcOvWLb3lhkbOzD7SZM6/U+/evaFUKhEXFycl\nbBo7duwAADRt2lR6bYCsESUBwN/f32Cs1atXx5w5c9C/f3+Ym5sjOTlZqufh4QEACA0NRVhYmPQT\nGhqqVZ79OM3NzVG1alXpglYjLi4ON27cwLZt2zB9+nRpueZvBmQlqBqurq5IT0/X2m9YWBieP38O\nS0tLKXHPXodMh4eHBzp27AgAUjKvsXfvXgBZo/fmNgoskDU1S82aNaX3osbLly9x8eJFrF27Ft99\n9x2ArPn0NO+3pKQknD17FgDQunVrg9v39fXNc6RbNzc3g2WaERazf54vXryIpKQkAJBujugjk8nQ\nuHFjAMClS5dyjSE3muPMbV9A1g0gAIiKisL169d1ylu2bCmNipmTo6MjfHx8AGifD4KCggAAR48e\nhb29PYYMGYL169fj+fPnr30cly9flv6vOT/roznfZT+3AFk3Nd3c3FCnTh3I5f+77EpMTERISAh2\n7dqF8ePHSzfOctYnIsovjt1NRcrV1RXu7u64f/8+7t+/X6jbdnR0NFimmVBYJpPlOQx1XFyc9H9N\ni5chmqkLIiMjkZSUJA0Lrvki1tdaB2QN9w9kXVTmNTx3rVq1cOnSJSnZzK5KlSq51s2eEOljaDjt\n7BMw53y93Nzc0LNnT2zZsgWbN2/G0KFDpbKtW7cCyGot1BBC4MmTJwBgcEhxfcLCwhAVFQUA+OKL\nL/DFF1/kWefRo0c6y+7du4eVK1fi7NmzCAkJwdOnT/XWzX6cISEh0v87dOiQr3gfPHiQr/Wo5Ond\nuzcOHDiAXbt2abXYbN68GQAMtrzoc/bsWaxZswZXrlzBnTt3EB0drbNO9vdaXFwc4uPjARg+XwBZ\nc8O5u7sjPDzc4DqaG1j6aBKI7PvWJA6Wlpa53iACtM91BaFWq6Vza86bLTllb7V89uyZTnm1atVy\nre/h4YFz587h3r170rKPP/4Yd+7cwR9//IH4+HisW7cO69atA5CVaPbr1w/Dhg3L9TXU0LwGFStW\nzDUWzXFmZmbqLY+IiMDKlStx4sQJhISESOdJIqLCwsSOipyHhwfu37+vdfGcH3fv3sWjR4/Qtm1b\nvd04c7urqSnLz9xCKpVK+n9ec07lVa5QKPQu13zRGyp/nX3kxtzcHADybKXMKa87xIMHD8aWLVuw\ne/duPH78GFWrVsW5c+dw8eJFAFkXyhppaWlSt7PXmQ8re/cjNzc3yOVyrYRTQ61Ww8rKCnFxcToX\nWfPmzcOcOXO0lpmbm8PPzw/9+vVDpUqVMHjwYJ1taloxLC0t4erqqvf1E0JAJpPB2toajx49ytc8\nfFQyabpjXrlyBRcvXoS/vz+io6Oxbds2KJXKPOfN1BgyZIiULGjY2tqiadOmGDVqFCIjIzFlyhSt\n97FarZZ+z+tzl9f5Kz/nwOw05yG5XJ7vc11BW49UKpV0bn2d82pB9mfovLdmzRpMnDgRq1atwp49\ne6SbMadPn8bp06cxefJkBAcH5/n31vy95HJ5rudwTauivmNYvny51OVUQ6FQoGHDhujatSsaN26M\nUaNGSTe3iIgKgokdFblOnTrh4MGDuHbtGm7cuCF1F8zLwoULsXTpUri6uuL69es6F9KF1V1F83wX\nkPV8mZeXl8F1NXdY7ezspC6m2elLRABI3bpiY2ORkZGR6/OGmjv0+lrXsieh+mhasAxN+lvQ16xn\nz57w8PDAw4cPsXPnTowfP1569qZTp05aLZ2WlpZwcnLC06dPX+tuf/YW2P/85z9SV6r8Wr16tZTU\ntW/fHjNnzkS9evWk55wAwy2ZmlZUlUqF06dPa9Wh0qdq1aoIDAzE3r17sXv3bvj7+yM4OBgAMHTo\n0Dy7YQLA1KlTpaRu6NChGDt2LGrVqqX1PtZ0Xc7+uStXrhwsLS2RlJSUZ7fAly9f5lr+up9nzXEl\nJyfj1atXufZ6ePz4MQDDrfx5USqVUsKV13Fk751gZ2enU26oBUxDc97L2S0WyOrS6uvri59//hn3\n79/HgQMHsHHjRmnS+cGDB+POnTu53qjRTAIfExODV69eGeyJoEnKsne3BID9+/dLSZ2/vz8+++wz\nNGrUSKsHRmJi4mvfkCMiyonP2FGR69mzp/T/5cuX56tOVFSUNChB1apVi7R1pFatWlJyd/Xq1VzX\nvXHjBoCsZ9lyfnnnRjMQSHh4uPQcmT7p6elSy2bDhg317j+3i7mwsDAAeXcpfV1yuVwaqEYz+IJm\nAIhBgwbprK/pYpZbN7JXr15hwIAB6NevHx4/fgwvLy94enoCMPyMoIa+i6Dff/8dANC8eXMcOnQI\nHTt21EnQsg/+kv111NxsyMjIyLXLcGZmptZgFGS6NK3Mmvex5nm7/HTDTEhIwOrVqwFkDdrzxx9/\noEWLFjqJUs7BhoCsGxiaZ7FyDt6S3b179wq9+7qmeyXwv3OZIZrBnnJ7piwvdevWBZD3c3qa51Wt\nrKyk5+Wyu3btmsG6arVaSuw0XTo1n+2YmBikpaVJ69aoUQNjx47FsWPHpL9fTEyM9B4wRPP3SklJ\nyfX8fffuXQC6LZSac5OnpydOnTqF7t2763Srj4uLk3oOEBEVFBM7KnLe3t5SUvDzzz/j8OHDedaZ\nPXu2dFE0derUogwPlSpVkgY0+euvvwyuFxYWJo26ln2gkPxo3ry5dNdX8xyPPjt37pSeCdO3j9DQ\nUJw+fVpv3QsXLuDcuXMAYHCkvDeheY7uwoUL2LlzJ65du4Zy5cppdcPUCAgIAAAEBwcbvFt/4cIF\n/PXXX9iyZYt0zJpBFDZs2JBrAvvWW2/B3Nwc48ePB5B1oa0ZdEGzDX2yX8Blb/1s1KiR1Gr3559/\nGqx/4cIF2NraQiaTYcuWLQbXo5JP0x3z/PnzOHz4MPbt2wd7e3t07tw5z7q3b9+WRo/NbcCl7KM0\nZn+/aT6f27ZtM9j1LmcXz8Lg4+Mj3cTI7Vx36dIlnDp1CgByHewpL9nPAw8fPjS4nuYmXoMGDfS2\nhv33v/812NoeHBwstS5qRuA9d+4crK2t4ejoKCWoOfXv3196vi6vnhANGzaUbpZpnivWx9BIxprB\nV/z9/bV6iGS3f/9+6WZVXi2URESGMLGjYvHtt99Kz1t16NABa9euNbju+++/L7Xsde/eHf379y/y\n+DRDWB84cABLlizRu86UKVOgUqmgVCpfu5ugi4uLVOeLL77QO+JlTEyMNGqjn5+fNHJfTtOmTdO7\nfMaMGVLdvIZqL4jGjRujWbNmiIyMxKeffgogqxuTvq5aQUFBMDc3z3U48W+//RZA1sWYJhkbNWoU\ngKwkevLkyXrrzZ07V0riNCP32draSq1zhu6oX716FfPmzZN+z37xVK5cOek9sHz5cp3RPzU0fx8L\nC4vXTu6pZHF3d5eSu4kTJyImJgbDhg3L13O5rq6uUou9odFR//rrL62pELK/38aOHQsXFxfExsZi\n+PDhOnX/+eeffA3DXxD/+te/AAAbN240OFWD5rPn6OhocEqZ/BgyZAisrKyQkZGBiRMn6l1nyZIl\nOHTokFZsOQkh9N7gS09Pl85777zzDnx9fQEAXl5e0t/xl19+0bvNPXv2SMl5fh4PGDFiBICsm5Mn\nT57UKf/pp5/0nteB//Vg0LTo5RQZGYmPP/5Y+j2vRJOIyCDBCcqpmBw6dEiagBWAqFOnjpg9e7ZY\ntWqVWL58uZgwYYI0US0AUb9+fZGUlKSzHc0E5Tkngc1u3bp1AoBwcHAQCQkJ0nJDE5QLIUSfPn2k\nfY8aNUpcuXJFhIaGikOHDokmTZpIZYsXL9bZX+3atQUAsXDhQoMxxcfHCw8PD2kC8m+++UbcuXNH\nPHjwQKxdu1ZUqFBB2sf58+e16mafuFvz2hw+fFg8efJEHDt2TPj5+emdUFgI7QnKQ0ND9cYWGhoq\nrXP48GGDx7BgwQKtOA4ePGhw3R9++EFar0ePHuL8+fPi4cOH4vjx46Jly5ZS2ZkzZ7TqffTRR1JZ\np06dxMmTJ0VoaKi4dOmS+Ne//mVwsnXN5OEAxPvvvy8eP34soqOjxf3798Xs2bMFAGFnZyccHR0F\nALFo0SKdmOvVqydtY8qUKeL69esiNDRUHDhwQPj7+0tly5YtM3jcVLLknKA8uxUrVmi9n3NOWi+E\n4QnK33nnHane/PnzxYsXL8SrV6/EtWvXxLBhwwQAUaFCBWFubq73c6mZ7BuAqFKlivjkk0/Ejz/+\nKLp06SIACEdHR2FtbS0AiDVr1mjV1UxQ3rt3b4PH/cknnwgYmEC9bdu20r4nTpwobty4IUJDQ8We\nPXuEj4+PVLZp0ya9Med3gnIh/jdJuyaWgwcPitDQUHH16lUxatQoqaxnz546cWafoByAaN26tTh7\n9qwIDw8Xe/bskf62SqVS3L59W6vul19+KdXr3bu3OH36tLhz5464deuW+P7774WFhYUAIPr06aNV\nz9AE5UII0bx5cwFAmJmZiQULFoiHDx+Ku3fvig8++EAAEJaWlgKA8PX11aqX/fzdt29fcf/+fREd\nHS0ePnwovv/+e6mu5jtg1qxZev+mRER5iGViR8Xqzp07olGjRlpf1vp+RowYITIyMvRuY86cOVJi\naMjvv/8uXchnvwDRXGQ4OjqK+Ph4nXrZE4ecPzKZTOcCS0NzgTF//vxcj//FixfC19fX4D4qVaok\nLly4oFNPc2HQtGlTMX36dL11LSwsxL59+3Tq7t+/X1rn1q1beuO6ffu2tI6+bWg8fvxYWFlZCQCi\ndu3auR6rEEIsXLjQ4LGamZkZ3NfMmTNzfX8MHjxYb73sF6w5f+rXry/UarUYOnSoACC6d++uUz8u\nLk4EBATkum99iT2VTKmpqaJSpUoCgJgwYYJO+dOnT6X3s4+Pj95t9OjRQwAQTZo00VoeEREhKleu\nbPB9MmDAAJGWliYaNGhg8GI9ODhYlC9fXqduQECAuH79uqhWrZoAINavX69VT/MeDgwMNHjskyZN\nEgCEv7+/TllGRobo2bOnwdgtLCzEjh07dOr9+eefAoCwsrISL168kJbHx8dLN0yWLFmiU2/lypW5\nfqZGjhyp9xg0iV1gYKCYMGGC3rpOTk7i0qVLeutrEi5DP126dNH5nvnqq68EAOHl5aWzvYSEBNGi\nRQu92+rdu7dYvny5ACAaNGigU1dzg8DQeT8pKUk63saNG+s9HiKiPMQqPv/885kA8h4CjKgQODk5\nYcyYMQgMDETFihVRoUIFlCtXDhUrVkTLli0xZMgQLF++HMOHDzc4OIm5ubk0mIem601OFhYWkMlk\n6N27N1q1aiUNUW1ubg6ZTIY+ffqgZcuWOvvo1asXunfvDisrK8jlcpQrVw7NmzfHuHHjsG7dOjRv\n3tzg/tzc3PDuu+/mOjeVtbU1Ro8ejSZNmkgjYzo7O6NDhw6YPXs2VqxYobf+uXPnEBwcDBsbG+zc\nuRO9evVCVFQUMjIyUL9+fUyYMAEbNmzQO/CAmZmZNIR7hw4d9I74Z2ZmBmtra/j7++Odd97RmrA8\nO3t7e+zZswdPnjzB+++/n+ecby1atMB7770HW1tbaYj1Ro0aYezYsdi4caM0uEJOb7/9NoKCgmBv\nbw+lUgkzMzP4+PigV69e+Omnn/DRRx/prTd8+HB4e3tLUyW4uLggMDAQ//nPf7BgwQLIZDJUrFgR\ncrkc/v7+aNasmdZ7wMLCAu+99x4CAgJgZWUFhUIBGxsbNG3aFMOHD8cff/yB9u3b53rMVHKYmZmh\nXLlycHd3R9++faXBeTRsbGzg4eEBGxsbTJkyRe+gQ1ZWVrC2tsagQYO0BhKxtbXFmDFjpG6ZKpUK\n1apVQ58+fbB48WJ89NFHUCgUcHNzg7m5OZo0aYJGjRppbbtGjRqYPHkyWrdujdq1a6NDhw744osv\nMGfOHKhUKmmU15EjR2rN9WZubg6lUolBgwbp/cwDWSNgyuVy9OvXT2e/crkcAwcORKdOnWBpaQmZ\nTAYbGxu0atUKkyZNwtq1a/UOmqI59/bq1QsBAQHSICEKhQJWVlaoXr06evfurfOcnK+vLz744AO4\nuLjAzMwMMpkMPj4+GDhwIJYsWSJ1wc7p8OHDOHHiBKpWrYoNGzYgICAAUVFRUKvVaNy4MWbMmIE/\n//zT4ETuXbt2Rffu3WFrawtra2solUpUr14d3bt3x/fff4/Zs2frfAcolUooFAr0799fGjQl+/GP\nHDkSvr6+SE9PR2ZmJmrVqoXp06dj4cKFcHJyQmpqKnr16gU/Pz+tun379oWfnx/kcjmEELC3t5f+\n3itWrIBSqYSrqyuUSqU0kqeh5/GIiAxIkwkhYgHYGzsSIjJs0aJFGD9+PLy8vIw6MXZERASqVKkC\nIQQuXryoc/FCRHl7+PAhbG1tc50c++7du6hVqxaArMFDWrRoUVzhlRizZ8/G/Pnz0bJlS73PtRER\nkZY4Dp5CRPm2efNmCCHQqlUrJnVEBfTtt9/C2dlZGpBDn127dgHIGngpZ4sbERGRPkzsiMigtLQ0\npKenA8iaS0rTNWzs2LHGDIvIpGla31atWiUlcNmdPn0aM2fOBJA1r56VlVWxxkdERKbJLO9ViMjY\nxP/P6aZWq6FWq19rcvQ3ERcXh+bNmyM+Ph6JiYlIS0tDy5YtMXjw4GLZP1FpNGzYMPz+++84ceIE\nevbsiTZt2uCtt96CEAJnz57FkSNHAGTNn/bTTz8ZOVrjUavVWv8SEVHumNgRmQDNnExyubzYkjog\na2CX9PR0ab6n6tWr5zrBOhHlz8GDBzFhwgQsWbIEx44dw7Fjx7TKx48fjwULFkgDlJRF2c97RESU\nNw6eQmQC7t27h71798LT0xPdunUr1n2vWLECW7duhYeHByZOnAhvb+9i3T9RabZv3z4cPXoU9+/f\nh0wmg7e3Nzp06IB27doZOzSju3btGg4dOoQGDRrkOQIvEREhjokdERERERGRaeOomERERERERKaO\niR0REREREZGJY2JHRERERERk4pjYERERERERmTgmdkRERERERCbODEA8skbFjDNyLERERERERPR6\n7AHE/x/72rgmzZqENgAAAABJRU5ErkJggg==\n"
    }
   },
   "id": "f537fc7a-41ed-4e26-8f5d-5edbea03b643"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "high_alt<- \"Data/image/micasense/Gafanha_high_multi.tif\" %>% \n",
    "  rast()\n",
    "\n",
    "# low_alt<- \"Data/image/micasense/Gafanha_Low_Multi_rsp.tif\" %>% \n",
    "#   rast() \n",
    "\n",
    "low_alt<- \"Data/image/micasense/Gafanha_low_multi.tif\" %>% \n",
    "  rast() %>% \n",
    "  resample(high_alt, method = \"med\")\n",
    "\n",
    "\n",
    "low_alt_pred<- \"Data/image/prediction/NewPreds/Gafanha_Low_Pred_rsp.tif\" %>% \n",
    "  rast() \n",
    "\n",
    "mask<-\"Data/shp/Mask_Alt_Compare_All.shp\" %>% \n",
    "  vect()\n",
    "  \n",
    "high_alt_masked<-crop(high_alt,mask)\n",
    "high_alt_masked<-mask(high_alt_masked,mask)\n",
    "\n",
    "low_alt_masked<-crop(low_alt,mask)\n",
    "low_alt_masked<-mask(low_alt_masked,mask)\n",
    "\n",
    "low_alt_pred_masked<-crop(low_alt_pred,mask)\n",
    "low_alt_pred_masked<-mask(low_alt_pred_masked,mask)\n",
    "\n",
    "\n",
    "stk<-rast(list(c(high_alt_masked,low_alt_masked,low_alt_pred_masked)))%>% \n",
    "  as.data.frame()\n",
    "\n",
    "names(stk)<-c(\"B1_high\",\"B2_high\",\"B3_high\",\"B4_high\",\"B5_high\",\"B6_high\",\"B7_high\",\"B8_high\",\"B9_high\",\"B10_high\",\n",
    "              \"B1_low\",\"B2_low\",\"B3_low\",\"B4_low\",\"B5_low\",\"B6_low\",\"B7_low\",\"B8_low\",\"B9_low\",\"B10_low\",\n",
    "              \"pred\")\n",
    "\n",
    "stk_long<-stk %>% \n",
    "  mutate(px_ID = c(1:nrow(stk))) %>% \n",
    "  pivot_longer(-c(px_ID,pred),values_to = \"Value\",names_to = \"Bands\") %>% \n",
    "  mutate(alt = case_when(str_detect(Bands, \"low\") ~ \"Low\",\n",
    "                         str_detect(Bands, \"high\") ~ \"High\",\n",
    "                         TRUE ~ \"NA\"),\n",
    "         Bands = gsub(\"_high\",\"\",Bands),\n",
    "         Bands = gsub(\"_low\",\"\",Bands))\n",
    "\n",
    "\n",
    "stk_summarized<-stk_long %>% \n",
    "  # group_by(px_ID,alt) %>% \n",
    "  # reframe(Value = mean(Value),\n",
    "  #         pred = unique(pred)) %>% \n",
    "  pivot_wider(names_from = \"alt\", values_from = \"Value\")\n",
    "\n",
    "df<-stk_summarized %>% \n",
    "  ungroup() %>% \n",
    "  group_by(Bands) %>% \n",
    "  filter(Low<10000,\n",
    "         High < 10000,\n",
    "         Low < quantile(Low,probs = 0.99),\n",
    "         High < quantile(High,probs = 0.99),\n",
    "         Low > quantile(Low,probs = 0.01),\n",
    "         High > quantile(High,probs = 0.01),\n",
    "         !pred %in% c(7,9,3,10)) %>%\n",
    "  ungroup()\n",
    " \n",
    "plot <- df %>% \n",
    "  group_by(pred) %>% \n",
    "  sample_n(10000) %>% \n",
    "  mutate(Low = Low /10000, \n",
    "         High = High/10000) %>% \n",
    "  ggplot(aes(x=Low, y=High))+\n",
    "  # geom_point(aes(color = as.factor(pred)))\n",
    "  ggpointdensity::geom_pointdensity(adjust = .03)+\n",
    "  viridis::scale_color_viridis(limits = c(0,6000),\n",
    "                               breaks = c(0,2000,4000,6000),\n",
    "                               labels = c(\"1\",\"2000\",\"4000\",\"6000\"))+\n",
    "  # geom_smooth(method = \"loess\")+\n",
    "  geom_abline(intercept = 0, slope = 1, linewidth = 1, linetype = \"dashed\", color = \"black\" )+\n",
    "  geom_smooth(method = \"lm\", color = \"red\", linetype = \"dashed\",linewidth = 1)+\n",
    "  theme_Bede()+\n",
    "  ylim(c(0.05,0.6))+\n",
    "  xlim(c(0.05,0.6))+\n",
    "  ylab(\"Reflectance of the 120m flight\")+\n",
    "  xlab(\"Reflectance of the 12m flight\")+\n",
    "  labs(color = \"Pixel density\")+\n",
    "  theme(\n",
    "    legend.position = \"top\",\n",
    "        legend.key.width = unit(3,\"line\"))\n",
    "plot\n",
    "ggsave(\"./Figures/Compare_reflectance.png\",plot, width= 472*4, height=402*4, units = \"px\")"
   ],
   "id": "4a567a1d-b917-44ab-a076-f3da02a0dc78"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: [rast] unknown extent"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "SpatRaster resampled to ncells = 500851"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "Warning: [rast] unknown extent"
     ]
    },
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "SpatRaster resampled to ncells = 500851"
     ]
    },
    {
     "output_type": "display_data",
     "metadata": {},
     "data": {}
    }
   ],
   "source": [
    "RAW <- ggplot()+\n",
    "  geom_spatraster_rgb(data = rast(\"./Figures/Low_res/Compare_reflectance.png\"))+\n",
    "  coord_equal()+\n",
    "  geom_label(aes(x = 300,y= 1200, label = \"A\") ,size = 10)+\n",
    "    theme_void()+\n",
    "  theme(axis.text = element_blank(), \n",
    "        axis.title = element_blank(), \n",
    "        axis.ticks = element_blank(), \n",
    "        axis.ticks.length = unit(0, \"pt\"),\n",
    "        panel.grid.major=element_blank(), \n",
    "        panel.grid.minor=element_blank(), \n",
    "        plot.margin = margin(0, 0, 0, 0, \"pt\"))"
   ],
   "id": "cell-fig-CompareRef"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this study, a key innovation lies in the utilization of two different altitudes (12m and 120m) for constructing the neural network model. The lower altitude flights (8mm spatial resolution) enable precise selection of pure pixels representing the classes used in the neural network model. This methodology implies a consistency between the reflectance of both altitude. <a href=\"#fig-CompareRef\" class=\"quarto-xref\">Figure 13</a> (A) depicts the relationship between reflectance from a low-altitude flight and a high-altitude flight conducted over the same area. An apparent underestimation of reflectance values is observed in the high-altitude flight, particularly for higher values. Since both flights were conducted over vegetation areas, the highest reflectance values correspond to the infrared part of the spectrum.\n",
    "\n",
    "The disparity in infrared reflectance may stem from temporal differences between the flights, possibly resulting in a slightly drier intertidal area and consequently higher infrared reflectance. This disparity poses an issue for the methodology followed in the present study, relying solely on one flight height for training. To address this issue, we employed min/max standardized reflectance spectra as predictors for the model (<a href=\"#eq-std\" class=\"quarto-xref\">Equation 1</a> ; <a href=\"#fig-CompareRef\" class=\"quarto-xref\">Figure 13</a> (B) ; ([Cao et al., 2017](#ref-Cao2017))). This approach allows us to eliminate differences in the level of reflectance, focusing instead on the shape of the spectra. This is a key feature in building a model that can reliably predict vegetation across geographical sites and seasons. It enables consistent prediction of vegetation classes across variations in biomass and variability in light conditions ([Fyfe, 2003](#ref-fyfe2003spatial) ; [Costa et al., 2021](#ref-COSTA2021107018) ; [Piaser et al., 2023](#ref-piaser2023impact)).\n",
    "\n",
    "The <a href=\"#fig-upscaling\" class=\"quarto-xref\">Figure 10</a> demonstrates that a seagrass cover percentage of 90 % is necessary for confident prediction of seagrass presence. This highlights a limitation of the methodology used to construct the training dataset for the model. The dataset was composed exclusively of pure pixels, which has resulted in the model’s reduced confidence when faced with lower percentages of seagrass cover. Intertidal seagrasses exhibit strong phenology, with varying pigment composition throughout the year([Bargain et al., 2013](#ref-bargain2013seasonal) ; [Légaré et al., 2022](#ref-legare2022remote)). This suggests that outside of the biomass peak of seagrasses, this model may be less accurate. Further investigation is warranted to explore this aspect.\n",
    "\n",
    "## 4.3 Application of the model\n",
    "\n",
    "Climate change, global warming, alien and invasive species development, coastal erosion, sealevel rise, while further impact coastal ecosystems in the future ([Holon et al., 2018](#ref-holon2018predictive) ; [Marquet et al., 2024](#ref-marquet2024global); [Schibalski et al., 2022](#ref-SCHIBALSKI2022101414)). The demand for meaningful and efficient monitoring methods for these habitats has therefore never been greater ([Muller-Karger et al., 2018](#ref-muller2018satellite) ; [Villalobos Perna et al., 2023](#ref-villalobos2023remote) ; [Oiry and Barillé, 2021](#ref-oiry2021using)). Our findings, particularly the improved discrimination of vegetation types, underscore the potential of drone-based remote sensing to support diverse applications, from biodiversity conservation to climate change adaptation strategies**.** Employing traditional sampling methods to monitor these ecosystems is time and resource-intensive, and the findings are often difficult to scale up. Earth observation methods can bridge this gap and meet the needs for monitoring coastal ecosystems. ([Papathanasopoulou et al., 2019](#ref-papathanasopoulou2019satellite)). The retrieval of Essential Biodiversity Variables (EBVs) and Essential Ocean Variables (EOVs) through satellite observations is increasingly common, enabling comprehensive monitoring of entire ecosystems over extended time periods ([Ratnarajah et al., 2023](#ref-ratnarajah2023monitoring) ; [Zoffoli et al., 2020](#ref-ZOFFOLI2020112020)). The significance of monitoring waters, particularly in light of the Water Framework Directive (WFD ; ([European-Comission, 2000](#ref-WFD2000))), cannot be overstated in environmental management and conservation efforts. The WFD mandates the achievement and maintenance of “good ecological status” for all European waters, which necessitates a comprehensive understanding and monitoring of aquatic ecosystems, including coastal habitats like seagrass ([Foden and Brazier, 2007](#ref-foden2007angiosperms) ; [Nordlund et al., 2024](#ref-nordlund2024one) ; [Zoffoli et al., 2021](#ref-Zoffoli2021)).\n",
    "\n",
    "Effective and efficient monitoring tools are essential for identifying the impacts of human activities and natural changes on these ecosystems. Drones, with their capability for high-resolution, multispectral imagery, provide a novel and powerful means to rapidly and accurately acquire ground truth data (i.e. output of the model used in the present study). This data is critical for calibrating and validating satellite remote sensing applications, thereby enhancing our capacity to monitor vast coastal areas systematically under the WFD framework. The integration of drone technology facilitates a scalable approach to environmental surveillance, offering significant advancements in the spatial and temporal resolution of data collection. This, in turn, supports the directive’s objectives by enabling more informed and timely management decisions for the preservation and restoration of aquatic ecosystems.\n",
    "\n",
    "# 5. Conclusion\n",
    "\n",
    "The utilization of drone-based remote sensing coupled with machine learning techniques has proven to be an effective method for the discrimination of seagrasses and macroalgae in coastal ecosystems. This study has highlighted the potential for these technologies to enhance the accuracy and efficiency of habitat monitoring, which is essential for environmental management and conservation strategies. Through the application of high-resolution multispectral imagery, the research contributes to the broader understanding and surveillance of aquatic ecosystems, aligning with objectives outlined in policies such as the Water Framework Directive. The findings underscore the importance of adopting advanced remote sensing tools in ecological studies and environmental monitoring, providing a foundation for future research and policy implementation aimed at ecosystem preservation and restoration.\n",
    "\n",
    "Adade, R., Aibinu, A.M., Ekumah, B., Asaana, J., 2021. Unmanned aerial vehicle (UAV) applications in coastal zone management—a review. Environmental Monitoring and Assessment 193, 1–12.\n",
    "\n",
    "Angnuureng, D.B., Brempong, K., Jayson-Quashigah, P., Dada, O., Akuoko, S., Frimpomaa, J., Mattah, P., Almar, R., 2022. Satellite, drone and video camera multi-platform monitoring of coastal erosion at an engineered pocket beach: A showcase for coastal management at elmina bay, ghana (west africa). Regional Studies in Marine Science 53, 102437.\n",
    "\n",
    "Bannari, A., Ali, T.S., Abahussain, A., 2022. The capabilities of sentinel-MSI (2A/2B) and landsat-OLI (8/9) in seagrass and algae species differentiation using spectral reflectance. Ocean Science 18, 361–388.\n",
    "\n",
    "Bargain, A., Robin, M., Méléder, V., Rosa, P., Le Menn, E., Harin, N., Barillé, L., 2013. Seasonal spectral variation of zostera noltii and its influence on pigment-based vegetation indices. Journal of experimental marine biology and ecology 446, 86–94.\n",
    "\n",
    "Beach, K., Borgeas, H., Nishimura, N., Smith, C., 1997. In vivo absorbance spectra and the ecophysiology of reef macroalgae. Coral Reefs 16, 21–28.\n",
    "\n",
    "Brunier, G., Oiry, S., Gruet, Y., Dubois, S.F., Barillé, L., 2022. Topographic analysis of intertidal polychaete reefs (sabellaria alveolata) at a very high spatial resolution. Remote Sensing 2022, Vol. 14, Page 307 14, 307. <https://doi.org/10.3390/RS14020307>\n",
    "\n",
    "Cao, J., Thorson, J.T., Richards, R.A., Chen, Y., 2017. Spatiotemporal index standardization improves the stock assessment of northern shrimp in the gulf of maine. Canadian Journal of Fisheries and Aquatic Sciences 74, 1781–1793. <https://doi.org/10.1139/cjfas-2016-0137>\n",
    "\n",
    "Casella, E., Drechsel, J., Winter, C., Benninghoff, M., Rovere, A., 2020. Accuracy of sand beach topography surveying by drones and photogrammetry. Geo-Marine Letters 40, 255–268.\n",
    "\n",
    "Coffer, M.M., Graybill, D.D., Whitman, P.J., Schaeffer, B.A., Salls, W.B., Zimmerman, R.C., Hill, V., Lebrasse, M.C., Li, J., Keith, D.J., others, 2023. Providing a framework for seagrass mapping in united states coastal ecosystems using high spatial resolution satellite imagery. Journal of Environmental Management 337, 117669.\n",
    "\n",
    "Collin, A., Dubois, S., James, D., Houet, T., 2019. Improving intertidal reef mapping using UAV surface, red edge, and near-infrared data. Drones 3, 67.\n",
    "\n",
    "Costa, V., Serôdio, J., Lillebø, A.I., Sousa, A.I., 2021. Use of hyperspectral reflectance to non-destructively estimate seagrass zostera noltei biomass. Ecological Indicators 121, 107018. https://doi.org/<https://doi.org/10.1016/j.ecolind.2020.107018>\n",
    "\n",
    "Davies, B.F.R., Gernez, P., Geraud, A., Oiry, Simon, Rosa, P., Zoffoli, M.L., Barillé, L., 2023a. Multi- and hyperspectral classification of soft-bottom intertidal vegetation using a spectral library for coastal biodiversity remote sensing. Remote Sensing of Environment 290, 113554. <https://doi.org/10.1016/j.rse.2023.113554>\n",
    "\n",
    "Davies, B.F.R., Sousa, A.I., Figueira, R., Oiry, S., Gernez, P., Barillé, L., 2023b. Benthic intertidal vegetation from the tagus estuary and aveiro lagoon. <https://doi.org/10.15468/n4ak6x>\n",
    "\n",
    "Devlin, M., Brodie, J., 2023. Nutrients and eutrophication, in: Marine Pollution–Monitoring, Management and Mitigation. Springer, pp. 75–100.\n",
    "\n",
    "Douay, F., Verpoorter, C., Duong, G., Spilmont, N., Gevaert, F., 2022. New hyperspectral procedure to discriminate intertidal macroalgae. Remote Sensing 14. <https://doi.org/10.3390/rs14020346>\n",
    "\n",
    "Duffy, J.E., Benedetti-Cecchi, L., Trinanes, J., Muller-Karger, F.E., Ambo-Rappe, R., Boström, C., Buschmann, A.H., Byrnes, J., Coles, R.G., Creed, J., others, 2019. Toward a coordinated global observing system for seagrasses and marine macroalgae. Frontiers in Marine Science 6, 317.\n",
    "\n",
    "European-Comission, 2000. <span class=\"nocase\">“DIRECTIVE 2000/60/EC OF THE EUROPEAN PARLIAMENT AND OF THE COUNCIL of 23 October 2000 establishing a framework for Community action in the field of water policy” or, in short, the EU Water Framework Directive</span>. Official Journal of the European Communities L 327, 1–72.\n",
    "\n",
    "Fairley, I., Williamson, B.J., McIlvenny, J., King, N., Masters, I., Lewis, M., Neill, S., Glasby, D., Coles, D., Powell, B., others, 2022. Drone-based large-scale particle image velocimetry applied to tidal stream energy resource assessment. Renewable Energy 196, 839–855.\n",
    "\n",
    "Foden, J., Brazier, D., 2007. Angiosperms (seagrass) within the EU water framework directive: A UK perspective. Marine Pollution Bulletin 55, 181–195.\n",
    "\n",
    "Fyfe, S., 2003. Spatial and temporal variation in spectral reflectance: Are seagrass species spectrally distinct? Limnology and Oceanography 48, 464–479.\n",
    "\n",
    "Gardner, R.C., Finlayson, C., 2018. Global wetland outlook: State of the world’s wetlands and their services to people. Stetson Law.\n",
    "\n",
    "Hedley, J.D., Mirhakak, M., Wentworth, A., Dierssen, H.M., 2018. Influence of three-dimensional coral structures on hyperspectral benthic reflectance and water-leaving reflectance. Applied Sciences 8, 2688.\n",
    "\n",
    "Holon, F., Marre, G., Parravicini, V., Mouquet, N., Bockel, T., Descamp, P., Tribot, A.-S., Boissery, P., Deter, J., 2018. A predictive model based on multiple coastal anthropogenic pressures explains the degradation status of a marine ecosystem: Implications for management and conservation. Biological Conservation 222, 125–135.\n",
    "\n",
    "Jankowska, E., Michel, L.N., Lepoint, G., Włodarska-Kowalczuk, M., 2019. Stabilizing effects of seagrass meadows on coastal water benthic food webs. Journal of Experimental Marine Biology and Ecology 510, 54–63.\n",
    "\n",
    "Joyce, K.E., Fickas, K.C., Kalamandeen, M., 2023. The unique value proposition for using drones to map coastal ecosystems. Cambridge Prisms: Coastal Futures 1, e6.\n",
    "\n",
    "Kirk, J.T., 1994. Light and photosynthesis in aquatic ecosystems. Cambridge university press.\n",
    "\n",
    "Légaré, B., Bélanger, S., Singh, R.K., Bernatchez, P., Cusson, M., 2022. Remote sensing of coastal vegetation phenology in a cold temperate intertidal system: Implications for classification of coastal habitats. Remote Sensing 14, 3000.\n",
    "\n",
    "Lin, H., Sun, T., Zhou, Y., Gu, R., Zhang, X., Yang, W., 2018. Which genes in a typical intertidal seagrass (zostera japonica) indicate copper-, lead-, and cadmium pollution? Frontiers in Plant Science 9, 1545.\n",
    "\n",
    "Los Santos, C.B. de, Krause-Jensen, D., Alcoverro, T., Marbà, N., Duarte, C.M., Van Katwijk, M.M., Pérez, M., Romero, J., Sánchez-Lizaso, J.L., Roca, G., others, 2019. Recent trend reversal for declining european seagrass meadows. Nature communications 10, 3356.\n",
    "\n",
    "Marquet, P.A., Buschmann, A.H., Corcoran, D., Dı́az, P.A., Fuentes-Castillo, T., Garreaud, R., Pliscoff, P., Salazar, A., 2024. Global change and acceleration of anthropic pressures on patagonian ecosystems, in: Conservation in Chilean Patagonia: Assessing the State of Knowledge, Opportunities, and Challenges. Springer International Publishing Cham, pp. 33–65.\n",
    "\n",
    "Miloslavich, P., Bax, N.J., Simmons, S.E., Klein, E., Appeltans, W., Aburto-Oropeza, O., Garcia, M.A., Batten, S.D., Benedetti-Cecchi, L., Checkley, D.M., Chiba, S., Duffy, J.E., Dunn, D.C., Fischer, A., Gunn, J., Kudela, R., Marsac, F., Muller-Karger, F.E., Obura, D., Shin, Y.J., 2018. Essential ocean variables for global sustained observations of biodiversity and ecosystem changes. Global Change Biology 24, 2416–2433. <https://doi.org/10.1111/GCB.14108>\n",
    "\n",
    "Muller-Karger, F.E., Hestir, E., Ade, C., Turpie, K., Roberts, D.A., Siegel, D., Miller, R.J., Humm, D., Izenberg, N., Keller, M., others, 2018. Satellite sensor requirements for monitoring essential biodiversity variables of coastal ecosystems. Ecological applications 28, 749–760.\n",
    "\n",
    "Nguyen, H.M., Ralph, P.J., Marı́n-Guirao, L., Pernice, M., Procaccini, G., 2021. Seagrasses in an era of ocean warming: A review. Biological Reviews 96, 2009–2030.\n",
    "\n",
    "Nijland, W., Reshitnyk, L., Rubidge, E., 2019. Satellite remote sensing of canopy-forming kelp on a complex coastline: A novel procedure using the landsat image archive. Remote Sensing of Environment 220, 41–50.\n",
    "\n",
    "Nordlund, L.M., Unsworth, R.K., Wallner-Hahn, S., Ratnarajah, L., Beca-Carretero, P., Boikova, E., Bull, J.C., Chefaoui, R.M., Santos, C.B. de los, Gagnon, K., others, 2024. One hundred priority questions for advancing seagrass conservation in europe. Plants, People, Planet.\n",
    "\n",
    "Oh, J., Kim, D., Lee, H., 2017. Use of a drone for mapping and time series image acquisition of tidal zones. Journal of the Korean Institute of Intelligent Systems 27, 119–125.\n",
    "\n",
    "Oiry, S., Barillé, L., 2021. Using sentinel-2 satellite imagery to develop microphytobenthos-based water quality indices in estuaries. Ecological Indicators 121, 107184.\n",
    "\n",
    "Orth, R.J., Carruthers, T.J., Dennison, W.C., Duarte, C.M., Fourqurean, J.W., Heck, K.L., Hughes, A.R., Kendrick, G.A., Kenworthy, W.J., Olyarnik, S., others, 2006. A global crisis for seagrass ecosystems. Bioscience 56, 987–996.\n",
    "\n",
    "Papathanasopoulou, E., Simis, S., Alikas, K., Ansper, A., Anttila, J., Barillé, A., Barillé, L., Brando, V., Bresciani, M., Bučas, M., others, 2019. Satellite-assisted monitoring of water quality to support the implementation of the water framework directive. EOMORES white paper.\n",
    "\n",
    "Pereira, H.M., Ferrier, S., Walters, M., Geller, G.N., Jongman, R.H., Scholes, R.J., Bruford, M.W., Brummitt, N., Butchart, S.H., Cardoso, A., others, 2013. Essential biodiversity variables. Science 339, 277–278.\n",
    "\n",
    "Piaser, E., Berton, A., Bolpagni, R., Caccia, M., Castellani, M.B., Coppi, A., Dalla Vecchia, A., Gallivanone, F., Sona, G., Villa, P., 2023. Impact of radiometric variability on ultra-high resolution hyperspectral imagery over aquatic vegetation: Preliminary results. IEEE Journal of Selected Topics in Applied Earth Observations and Remote Sensing.\n",
    "\n",
    "Ralph, P., Polk, S., Moore, K., Orth, R., Smith Jr, W., 2002. Operation of the xanthophyll cycle in the seagrass zostera marina in response to variable irradiance. Journal of Experimental Marine Biology and Ecology 271, 189–207.\n",
    "\n",
    "Ratnarajah, L., Abu-Alhaija, R., Atkinson, A., Batten, S., Bax, N.J., Bernard, K.S., Canonico, G., Cornils, A., Everett, J.D., Grigoratou, M., others, 2023. Monitoring and modelling marine zooplankton in a changing climate. Nature Communications 14, 564.\n",
    "\n",
    "Repolho, T., Duarte, B., Dionı́sio, G., Paula, J.R., Lopes, A.R., Rosa, I.C., Grilo, T.F., Caçador, I., Calado, R., Rosa, R., 2017. Seagrass ecophysiological performance under ocean warming and acidification. Scientific Reports 7, 41443.\n",
    "\n",
    "Roca, M., Dunbar, M.B., Román, A., Caballero, I., Zoffoli, M.L., Gernez, P., Navarro, G., 2022. Monitoring the marine invasive alien species rugulopteryx okamurae using unmanned aerial vehicles and satellites. Frontiers in Marine Science 9. <https://doi.org/10.3389/fmars.2022.1004012>\n",
    "\n",
    "Román, A., Tovar-Sánchez, A., Olivé, I., Navarro, G., 2021. Using a UAV-mounted multispectral camera for the monitoring of marine macrophytes. Frontiers in Marine Science 1225.\n",
    "\n",
    "Rossiter, T., Furey, T., McCarthy, T., Stengel, D.B., 2020. UAV-mounted hyperspectral mapping of intertidal macroalgae. Estuarine, Coastal and Shelf Science 242, 106789.\n",
    "\n",
    "Schibalski, A., Kleyer, M., Maier, M., Schröder, B., 2022. Spatiotemporally explicit prediction of future ecosystem service provisioning in response to climate change, sea level rise, and adaptation strategies. Ecosystem Services 54, 101414. https://doi.org/<https://doi.org/10.1016/j.ecoser.2022.101414>\n",
    "\n",
    "Soissons, L.M., Haanstra, E.P., Van Katwijk, M.M., Asmus, R., Auby, I., Barillé, L., Brun, F.G., Cardoso, P.G., Desroy, N., Fournier, J., others, 2018. Latitudinal patterns in european seagrass carbon reserves: Influence of seasonal fluctuations versus short-term stress and disturbance events. Frontiers in Plant Science 9, 88.\n",
    "\n",
    "Tallam, K., Nguyen, N., Ventura, J., Fricker, A., Calhoun, S., O’Leary, J., Fitzgibbons, M., Robbins, I., Walter, R.K., 2023. Application of deep learning for classification of intertidal eelgrass from drone-acquired imagery. Remote Sensing 15, 2321.\n",
    "\n",
    "Traganos, D., Reinartz, P., 2018. Mapping mediterranean seagrasses with sentinel-2 imagery. Marine Pollution Bulletin 134, 197–209. <https://doi.org/10.1016/j.marpolbul.2017.06.075>\n",
    "\n",
    "Tuya, F., Hernandez-Zerpa, H., Espino, F., Haroun, R., 2013. Drastic decadal decline of the seagrass cymodocea nodosa at gran canaria (eastern atlantic): Interactions with the green algae caulerpa prolifera. Aquatic Botany 105, 1–6.\n",
    "\n",
    "Unsworth, R.K., Cullen-Unsworth, L.C., Jones, B.L., Lilley, R.J., 2022. The planetary role of seagrass conservation. Science 377, 609–613.\n",
    "\n",
    "Villalobos Perna, P., Di Febbraro, M., Carranza, M.L., Marzialetti, F., Innangi, M., 2023. Remote sensing and invasive plants in coastal ecosystems: What we know so far and future prospects. Land 12, 341.\n",
    "\n",
    "Wang, Z., Fang, Z., Liang, J., Song, X., 2022. Assessment of global habitat suitability and risk of ocean green tides. Harmful Algae 119, 102324.\n",
    "\n",
    "Xu, S., Xu, S., Zhou, Y., Yue, S., Zhang, X., Gu, R., Zhang, Y., Qiao, Y., Liu, M., 2021. Long-term changes in the unique and largest seagrass meadows in the bohai sea (china) using satellite (1974–2019) and sonar data: Implication for conservation and restoration. Remote Sensing 13, 856.\n",
    "\n",
    "Zoffoli, M.L., Gernez, P., Godet, L., Peters, S., Oiry, S., Barillé, L., 2021. Decadal increase in the ecological status of a north-atlantic intertidal seagrass meadow observed with multi-mission satellite time-series. Ecological Indicators 130, 108033. <https://doi.org/10.1016/j.ecolind.2021.108033>\n",
    "\n",
    "Zoffoli, M.L., Gernez, P., Oiry, S., Godet, L., Dalloyau, S., Davies, B.F.R., Barillé, L., 2022. Remote sensing in seagrass ecology: Coupled dynamics between migratory herbivorous birds and intertidal meadows observed by satellite during four decades. Remote Sensing in Ecology and Conservation. <https://doi.org/10.1002/rse2.319>\n",
    "\n",
    "Zoffoli, M.L., Gernez, P., Rosa, P., Le Bris, A., Brando, V.E., Barillé, A.-L., Harin, N., Peters, S., Poser, K., Spaias, L., Peralta, G., Barillé, L., 2020. Sentinel-2 remote sensing of zostera noltei-dominated intertidal seagrass meadows. Remote Sensing of Environment 251, 112020. https://doi.org/<https://doi.org/10.1016/j.rse.2020.112020>"
   ],
   "id": "39277d89-7163-4c07-addf-e8910b9e2e1b"
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
